{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8439389d-c89d-4d0a-90ae-fbe6706cd498",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-08-09 12:08:15.086219: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2022-08-09 12:08:15.086314: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import time\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import h5py\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "import random\n",
    "import json\n",
    "\n",
    "import tensorflow as tf\n",
    "from tensorflow import keras"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0256fa07-f9b1-4e18-bed2-a0ab93ef20f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-08-09 12:08:24.150485: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory\n",
      "2022-08-09 12:08:24.150944: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)\n",
      "2022-08-09 12:08:24.151232: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (anupkumar-HP-ProBook-450-G5): /proc/driver/nvidia/version does not exist\n",
      "2022-08-09 12:08:24.154735: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    }
   ],
   "source": [
    "fig_size = (15, 15)\n",
    "font = {'family': 'serif', 'size': 12}\n",
    "plt.rc('font', **font)\n",
    "\n",
    "batch_size = 100\n",
    "test_batches = 1\n",
    "n_topk = 2\n",
    "max_seq_len = 25\n",
    "\n",
    "base_path = \"../log_08_08_22_2/\"\n",
    "predict_rnn = False\n",
    "\n",
    "#base_path = \"log_08_08_22_rnn/\"\n",
    "#predict_rnn = True # set to True for RNN model\n",
    "\n",
    "# log_08_08_22_2 (finish time: 40,000 steps in 158683.60082054138 seconds)\n",
    "# log_08_08_22_rnn (finish time: 40,000 steps in 173480.83078551292 seconds)\n",
    "\n",
    "#\"log_03_08_22_1/\" Balanced data with really selection of low freq tools - random choice\n",
    "# RNN: log_01_08_22_3_rnn\n",
    "# Transformer: log_01_08_22_0\n",
    "\n",
    "model_number = 40000\n",
    "model_path = base_path + \"saved_model/\" + str(model_number) + \"/tf_model/\"\n",
    "\n",
    "\n",
    "def remove_file(file_path):\n",
    "    if os.path.exists(file_path):\n",
    "        os.remove(file_path)\n",
    "\n",
    "\n",
    "def read_file(file_path):\n",
    "    \"\"\"\n",
    "    Read a file\n",
    "    \"\"\"\n",
    "    with open(file_path, \"r\") as json_file:\n",
    "        file_content = json.loads(json_file.read())\n",
    "    return file_content\n",
    "\n",
    "\n",
    "def write_file(file_path, content):\n",
    "    \"\"\"\n",
    "    Write a file\n",
    "    \"\"\"\n",
    "    remove_file(file_path)\n",
    "    with open(file_path, \"w\") as json_file:\n",
    "        json_file.write(json.dumps(content))\n",
    "\n",
    "\n",
    "r_dict = read_file(base_path + \"data/rev_dict.txt\")\n",
    "f_dict = read_file(base_path + \"data/f_dict.txt\")\n",
    "\n",
    "path_test_data = base_path + \"saved_data/test.h5\"\n",
    "\n",
    "file_obj = h5py.File(path_test_data, 'r')\n",
    "\n",
    "test_input = np.array(file_obj[\"input\"])\n",
    "test_target = np.array(file_obj[\"target\"])\n",
    "\n",
    "class_weights = read_file(base_path + \"data/class_weights.txt\")\n",
    "compatible_tools = read_file(base_path + \"data/compatible_tools.txt\")\n",
    "\n",
    "#tool_freq = utils.read_file(base_path + \"data/freq_dict_names.txt\")\n",
    "published_connections = read_file(base_path + \"data/published_connections.txt\")\n",
    "\n",
    "c_weights = list(class_weights.values())\n",
    "\n",
    "c_weights = tf.convert_to_tensor(c_weights, dtype=tf.float32)\n",
    "\n",
    "te_lowest_t_ids = read_file(base_path + \"data/te_lowest_t_ids.txt\")\n",
    "lowest_t_ids = [int(item) for item in te_lowest_t_ids.split(\",\")]\n",
    "tool_tr_freq = read_file(base_path + \"data/all_sel_tool_ids.txt\")\n",
    "\n",
    "all_sel_tool_ids = read_file(base_path + \"data/all_sel_tool_ids.txt\").split(\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0524ee9c-2b57-432b-a05c-5533271ebf2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.]])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "27ae2b83-55f4-4075-8c63-babcf9b86d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generated_attention(attention_weights, i_names, f_dict, r_dict):\n",
    "\n",
    "    #print(attention_weights.shape)\n",
    "    attention_heads = tf.squeeze(attention_weights, 0)\n",
    "    n_heads = attention_heads.shape[1]\n",
    "    i_names = i_names.split(\",\")\n",
    "    in_tokens = i_names\n",
    "    out_tokens = i_names\n",
    "    fig = plt.figure(figsize=(16, 8))\n",
    "\n",
    "    for h, head in enumerate(attention_heads):\n",
    "      ax = fig.add_subplot(2, n_heads, h+1)\n",
    "      plot_attention_head(in_tokens, out_tokens, head)\n",
    "      ax.set_xlabel(f'Head {h+1}')\n",
    "    plt.tight_layout()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9390e063-d2c1-4a5b-83f2-1ec4235a81d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_attention_head(in_tokens, out_tokens, attention):\n",
    "  # The plot is of the attention when a token was generated.\n",
    "  # The model didn't generate `<START>` in the output. Skip it.\n",
    "  #translated_tokens = translated_tokens[1:]\n",
    "  #print(attention)\n",
    "  ax = plt.gca()\n",
    "  ax.matshow(attention[:len(in_tokens), :len(out_tokens)])\n",
    "  #ax.matshow(attention[:len(in_tokens), :])\n",
    "\n",
    "  ax.set_xticks(range(len(in_tokens)))\n",
    "  ax.set_yticks(range(len(out_tokens)))\n",
    "\n",
    "  ax.set_xticklabels(in_tokens, rotation=90)\n",
    "  ax.set_yticklabels(out_tokens)\n",
    "\n",
    "\n",
    "def visualize_loss_acc():\n",
    "    epo_tr_batch_loss = read_file(base_path + \"data/epo_tr_batch_loss.txt\").split(\",\")\n",
    "\n",
    "    #print(len(epo_tr_batch_loss))\n",
    "    epo_tr_batch_loss = [np.round(float(item), 4) for item in epo_tr_batch_loss]\n",
    "\n",
    "    epo_tr_batch_acc = read_file(base_path + \"data/epo_tr_batch_acc.txt\").split(\",\")\n",
    "    epo_tr_batch_acc = [np.round(float(item), 4) for item in epo_tr_batch_acc]\n",
    "\n",
    "    epo_te_batch_loss = read_file(base_path + \"data/epo_te_batch_loss.txt\").split(\",\")\n",
    "    epo_te_batch_loss = [np.round(float(item), 4) for item in epo_te_batch_loss]\n",
    "\n",
    "    epo_te_batch_acc = read_file(base_path + \"data/epo_te_precision.txt\").split(\",\")\n",
    "    epo_te_batch_acc = [np.round(float(item), 4) for item in epo_te_batch_acc]\n",
    "\n",
    "    plot_loss_acc(epo_tr_batch_loss, epo_tr_batch_acc, \"training\")\n",
    "    plot_loss_acc(epo_te_batch_loss, epo_te_batch_acc, \"test\")\n",
    "\n",
    "    epo_te_low_batch_acc = read_file(base_path + \"data/epo_low_te_precision.txt\").split(\",\")\n",
    "    epo_te_low_batch_acc = [np.round(float(item), 4) for item in epo_te_low_batch_acc]\n",
    "\n",
    "    plot_low_te_prec(epo_te_low_batch_acc, \"Low test\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fa590092-8a14-485f-ab29-612852d95e5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_model():\n",
    "    return tf.saved_model.load(model_path)\n",
    "tf_loaded_model = load_model()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e90a9344-73c6-4577-a751-aed336aa5765",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_balanced_tr_y(x_seqs, y_labels, ulabels_tr_y_dict):\n",
    "    batch_y_tools = list(ulabels_tr_y_dict.keys())\n",
    "    random.shuffle(batch_y_tools)\n",
    "    label_tools = list()\n",
    "    rand_batch_indices = list()\n",
    "\n",
    "    for l_tool in batch_y_tools:\n",
    "        seq_indices = ulabels_tr_y_dict[l_tool]\n",
    "        random.shuffle(seq_indices)\n",
    "        \n",
    "        if seq_indices[0] not in rand_batch_indices:\n",
    "            rand_batch_indices.append(seq_indices[0])\n",
    "            label_tools.append(l_tool)\n",
    "        if len(rand_batch_indices) == batch_size:\n",
    "            break\n",
    "    \n",
    "    x_batch_train = x_seqs[rand_batch_indices]\n",
    "    y_batch_train = y_labels[rand_batch_indices]\n",
    "\n",
    "    unrolled_x = tf.convert_to_tensor(x_batch_train, dtype=tf.int64)\n",
    "    unrolled_y = tf.convert_to_tensor(y_batch_train, dtype=tf.int64)\n",
    "    return unrolled_x, unrolled_y, label_tools, rand_batch_indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ad68747f-5586-46d3-bb27-ce601e6907b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_custom_seq():\n",
    "\n",
    "    n_topk_ind = 20\n",
    "    t_ip = np.zeros((25))\n",
    "    t_ip[0] = int(f_dict[\"snpEff_build_gb\"])\n",
    "    t_ip[1] = int(f_dict[\"bwa_mem\"])\n",
    "    t_ip[2] = int(f_dict[\"samtools_view\"])\n",
    "\n",
    "    last_tool_name = \"samtools_view\"\n",
    "    \n",
    "    t_ip = tf.convert_to_tensor(t_ip, dtype=tf.int64)\n",
    "    if predict_rnn is True:\n",
    "        prediction = tf_loaded_model([t_ip], training=False)\n",
    "    else:\n",
    "        prediction, att_weights = tf_loaded_model([t_ip], training=False)\n",
    "    prediction_cwts = tf.math.multiply(c_weights, prediction)\n",
    "\n",
    "    top_k = tf.math.top_k(prediction, k=n_topk_ind, sorted=True)\n",
    "    top_k_wts = tf.math.top_k(prediction_cwts, k=n_topk_ind, sorted=True)\n",
    "\n",
    "    t_ip = t_ip.numpy()\n",
    "    label_pos = np.where(t_ip > 0)[0]\n",
    "\n",
    "    i_names = \",\".join([r_dict[str(item)] for item in t_ip[label_pos]  if item not in [0, \"0\"]])\n",
    "\n",
    "    pred_tools = [r_dict[str(item)] for item in top_k.indices.numpy()[0]  if item not in [0, \"0\"]]\n",
    "    pred_tools_wts = [r_dict[str(item)] for item in top_k_wts.indices.numpy()[0]  if item not in [0, \"0\"]]\n",
    "\n",
    "    c_tools = [r_dict[str(item)] for item in compatible_tools[str(f_dict[last_tool_name])]]\n",
    "\n",
    "    pred_intersection = list(set(pred_tools).intersection(set(c_tools)))\n",
    "    prd_te_prec = len(pred_intersection) / float(n_topk_ind)\n",
    "\n",
    "    print(\"Tool sequence: {}\".format([r_dict[str(item)] for item in t_ip[label_pos]]))\n",
    "    print()\n",
    "    print(\"Compatible true tools: {}, size: {}\".format(c_tools, len(c_tools)))\n",
    "    print()\n",
    "    print(\"Predicted top {} tools: {}\".format(n_topk_ind, pred_tools))\n",
    "    print()\n",
    "    print(\"Predicted precision: {}\".format(prd_te_prec))\n",
    "    print()\n",
    "    print(\"Correctly predicted tools: {}\".format(pred_intersection))\n",
    "    print()\n",
    "    print(\"Predicted top {} tools with weights: {}\".format(n_topk_ind, pred_tools_wts))\n",
    "    print()\n",
    "    if predict_rnn is False:\n",
    "        generated_attention(att_weights, i_names, f_dict, r_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "991ec9f5-77a9-4278-903e-2ee8ac1d5ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_low_freq_seq():\n",
    "    \n",
    "    \n",
    "    low_te_data = test_input[lowest_t_ids]\n",
    "    low_te_labels = test_target[lowest_t_ids]\n",
    "    #print(\"Test lowest ids\", low_te_data.shape, low_te_labels.shape)\n",
    "    #low_te_pred_batch, low_att_weights = tf_loaded_model([low_te_data], training=False)\n",
    "    low_topk = 5\n",
    "    low_te_precision = list()\n",
    "    for i, (low_inp, low_tar) in enumerate(zip(low_te_data, low_te_labels)):\n",
    "        if predict_rnn is True:\n",
    "            low_prediction = tf_loaded_model([low_inp], training=False)\n",
    "        else:\n",
    "            low_prediction, att_weights = tf_loaded_model([low_inp], training=False)\n",
    "        #print(low_prediction.shape)\n",
    "        \n",
    "        low_label_pos = np.where(low_tar > 0)[0]\n",
    "        low_topk = len(low_label_pos)\n",
    "        low_topk_pred = tf.math.top_k(low_prediction, k=low_topk, sorted=True)\n",
    "        low_topk_pred = low_topk_pred.indices.numpy()[0]\n",
    "        \n",
    "        low_label_pos_tools = [r_dict[str(item)] for item in low_label_pos if item not in [0, \"0\"]]\n",
    "        low_pred_label_pos_tools = [r_dict[str(item)] for item in low_topk_pred if item not in [0, \"0\"]]\n",
    "\n",
    "        low_intersection = list(set(low_label_pos_tools).intersection(set(low_pred_label_pos_tools)))\n",
    "        low_pred_precision = float(len(low_intersection)) / len(low_label_pos)\n",
    "        low_te_precision.append(low_pred_precision)\n",
    "        \n",
    "        #low_inp = low_inp.numpy()\n",
    "        low_inp_pos = np.where(low_inp > 0)[0]\n",
    "\n",
    "        print(\"Low: test tool sequence: {}\".format([r_dict[str(int(item))] for item in low_inp[low_inp_pos]]))\n",
    "        print()\n",
    "        print(\"Low: True labels: {}\".format(low_label_pos_tools))\n",
    "        print()\n",
    "        print(\"Low: Predicted labels: {}, Precision: {}\".format(low_pred_label_pos_tools, low_pred_precision))\n",
    "        print(\"-----------------\")\n",
    "        print()\n",
    "    \n",
    "    if len(lowest_t_ids) > 0:\n",
    "        print(\"Low test prediction precision: {}\".format(np.mean(low_te_precision)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d23c48f7-8ff3-4342-85c2-efe1e3ffe42b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def verify_tool_in_tr(r_dict):\n",
    "    \n",
    "\n",
    "    freq_dict = dict()\n",
    "    freq_dict_names = dict()\n",
    "\n",
    "    for tool_id in all_sel_tool_ids:\n",
    "        if tool_id not in freq_dict:\n",
    "            freq_dict[tool_id] = 0\n",
    "\n",
    "        if tool_id not in freq_dict_names:\n",
    "            freq_dict_names[r_dict[str(int(tool_id))]] = 0\n",
    "\n",
    "        freq_dict[tool_id] += 1\n",
    "        freq_dict_names[r_dict[str(int(tool_id))]] += 1\n",
    "\n",
    "    s_freq = dict(sorted(freq_dict.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    s_freq_names = dict(sorted(freq_dict_names.items(), key=lambda kv: kv[1], reverse=True))\n",
    "\n",
    "\n",
    "    write_file(base_path + \"data/s_freq_names.txt\", s_freq_names)\n",
    "    write_file(base_path + \"data/s_freq.txt\", s_freq)\n",
    "\n",
    "    return s_freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0e7626fc-8703-4aac-afa0-c72a41aa6364",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_u_tr_labels(y_tr):\n",
    "    labels = list()\n",
    "    labels_pos_dict = dict()\n",
    "    for i, item in enumerate(y_tr):\n",
    "        label_pos = np.where(item > 0)[0]\n",
    "        labels.extend(label_pos)\n",
    "        for label in label_pos:\n",
    "            if label not in labels_pos_dict:\n",
    "                labels_pos_dict[label] = list()\n",
    "            labels_pos_dict[label].append(i)\n",
    "\n",
    "    u_labels = list(set(labels))\n",
    "    \n",
    "    for item in labels_pos_dict:\n",
    "        labels_pos_dict[item] = list(set(labels_pos_dict[item]))\n",
    "    return u_labels, labels_pos_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "06683e24-e729-458e-b9fb-a1d4cd1f650d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_seq():\n",
    "\n",
    "    #visualize_loss_acc()\n",
    "\n",
    "    \n",
    "    #sys.exit()\n",
    "\n",
    "    #verify_training_sampling(tool_tr_freq, r_dict)\n",
    "    \n",
    "    #r_dict = read_file(base_path + \"data/rev_dict.txt\")\n",
    "    #f_dict = read_file(base_path + \"data/f_dict.txt\")\n",
    "    \n",
    "    all_tr_label_tools = verify_tool_in_tr(r_dict)\n",
    "\n",
    "    all_tr_label_tools_ids = list(all_tr_label_tools.keys())\n",
    "    all_tr_label_tools_ids = [int(t) for t in all_tr_label_tools_ids]\n",
    "    \n",
    "    \n",
    "    print(test_input.shape, test_target.shape)\n",
    "\n",
    "    \n",
    "    u_te_y_labels, u_te_y_labels_dict = get_u_tr_labels(test_target)\n",
    "\n",
    "    precision = list()\n",
    "    pub_prec_list = list()\n",
    "    error_label_tools = list()\n",
    "\n",
    "    for j in range(test_batches):\n",
    "        #te_x_batch, y_train_batch = sample_balanced(test_input, test_target, ulabels_te_dict)\n",
    "        te_x_batch, y_train_batch, selected_label_tools, bat_ind = sample_balanced_tr_y(test_input, test_target, u_te_y_labels_dict)\n",
    "        #print(j * batch_size, j * batch_size + batch_size)\n",
    "        #te_x_batch = test_input[j * batch_size : j * batch_size + batch_size, :]\n",
    "        #y_train_batch = test_target[j * batch_size : j * batch_size + batch_size, :]\n",
    "\n",
    "        for i, (inp, tar) in enumerate(zip(te_x_batch, y_train_batch)):\n",
    "\n",
    "            t_ip = inp\n",
    "            if len(np.where(inp > 0)[0]) < max_seq_len:\n",
    "                real_prediction = np.where(tar > 0)[0]\n",
    "                target_pos = list(set(all_tr_label_tools_ids).intersection(set(real_prediction)))\n",
    "                #print(\"Real predicted tools: {}, Actual trained labels: {}\".format(real_prediction, target_pos))\n",
    "                #print()\n",
    "                t_ip = tf.convert_to_tensor(t_ip, dtype=tf.int64)\n",
    "                if predict_rnn is True:\n",
    "                    prediction = tf_loaded_model([t_ip], training=False)\n",
    "                else:\n",
    "                    prediction, att_weights = tf_loaded_model([t_ip], training=False)\n",
    "                prediction_wts = tf.math.multiply(c_weights, prediction)\n",
    "\n",
    "                n_topk = len(target_pos)\n",
    "                top_k = tf.math.top_k(prediction, k=n_topk, sorted=True)\n",
    "                top_k_wts = tf.math.top_k(prediction_wts, k=n_topk, sorted=True)\n",
    "\n",
    "                t_ip = t_ip.numpy()\n",
    "                label_pos = np.where(t_ip > 0)[0]\n",
    "\n",
    "                i_names = \",\".join([r_dict[str(item)] for item in t_ip[label_pos]  if item not in [0, \"0\"]])\n",
    "                t_names = \",\".join([r_dict[str(int(item))] for item in target_pos  if item not in [0, \"0\"]])\n",
    "\n",
    "                last_i_tool = [r_dict[str(item)] for item in t_ip[label_pos]][-1]\n",
    "\n",
    "                true_tools = [r_dict[str(int(item))] for item in target_pos]\n",
    "\n",
    "                #print(\"Selected tool in true tools: {}\".format(s_label_tool in true_tools))\n",
    "\n",
    "                pred_tools = [r_dict[str(item)] for item in top_k.indices.numpy()[0]  if item not in [0, \"0\"]]\n",
    "                pred_tools_wts = [r_dict[str(item)] for item in top_k_wts.indices.numpy()[0]  if item not in [0, \"0\"]]\n",
    "\n",
    "                intersection = list(set(true_tools).intersection(set(pred_tools)))\n",
    "\n",
    "                pub_prec = 0.0\n",
    "                pub_prec_wt = 0.0\n",
    "\n",
    "                if last_i_tool in published_connections:\n",
    "                    true_pub_conn = published_connections[last_i_tool]\n",
    "                    #print(\"Test batch {}, True published tools: {}\".format(j+1, true_pub_conn)) \n",
    "                    #print()\n",
    "                    if len(pred_tools) > 0:\n",
    "                        intersection_pub = list(set(true_pub_conn).intersection(set(pred_tools)))\n",
    "                        intersection_pub_wt = list(set(true_pub_conn).intersection(set(pred_tools_wts)))\n",
    "                        pub_prec = float(len(intersection_pub)) / len(pred_tools)\n",
    "                        pub_prec_list.append(pub_prec)\n",
    "                        pub_prec_wt = float(len(intersection_pub_wt)) / len(pred_tools)\n",
    "                    else:\n",
    "                        pub_prec = False\n",
    "                        pub_prec_wt = False\n",
    "\n",
    "                if len(pred_tools) > 0:\n",
    "                    pred_precision = float(len(intersection)) / len(pred_tools)\n",
    "                    precision.append(pred_precision)\n",
    "\n",
    "                if pred_precision < 2.0:\n",
    "            \n",
    "                    print(\"Test batch {}, Tool sequence: {}\".format(j+1, [r_dict[str(item)] for item in t_ip[label_pos]]))\n",
    "                    print()\n",
    "                    print(\"Test batch {}, True tools: {}\".format(j+1, true_tools))\n",
    "                    print()\n",
    "                    print(\"Test batch {}, Predicted top {} tools: {}\".format(j+1, n_topk, pred_tools))\n",
    "                    print()\n",
    "                    print(\"Test batch {}, Predicted top {} tools with weights: {}\".format(j+1, n_topk, pred_tools_wts))\n",
    "                    print()\n",
    "                    print(\"Test batch {}, Precision: {}\".format(j+1, pred_precision)) \n",
    "                    print()\n",
    "                    print(\"Test batch {}, Published precision: {}\".format(j+1, pub_prec))\n",
    "                    print()\n",
    "                    print(\"Test batch {}, Published precision with weights: {}\".format(j+1, pub_prec_wt))\n",
    "                    #error_label_tools.append(select_tools[i])\n",
    "                    print(\"=========================\")\n",
    "                print(\"--------------------------\")\n",
    "                #generated_attention(att_weights, i_names, f_dict, r_dict)\n",
    "        \n",
    "                print(\"Batch {} prediction finished ...\".format(j+1))\n",
    "\n",
    "    if test_batches > 0:\n",
    "        print(\"Precision@{}: {}\".format(n_topk, np.mean(precision)))\n",
    "        print(\"Published Precision@{}: {}\".format(n_topk, np.mean(pub_prec_list)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "34f50e20-a934-463a-81b6-4393d36a9d25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(103152, 25) (103152, 2198)\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'deeptools_bam_coverage', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'bedtools_subtractbed', 'tp_sort_header_tool', 'tp_head_tool', 'Extract genomic DNA 1', 'macs2_callpeak', 'bowtie2', 'macs2_callpeak', 'samtools_rmdup', 'macs2_callpeak', 'diffbind', 'chipseeker', 'Cut1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'join_files_on_column_fuzzy', 'regexColumn1', 'collection_column_join', 'plotly_ml_performance_plots', 'bg_uniq', 'venn_list', 'scanpy_read_10x', 'tp_find_and_replace', 'bedtools_slopbed', 'Paste1', 'Show beginning1', 'ggplot_histogram', 'krona-text', 'stacks_denovomap', 'XY_Plot_1', 'random_lines1', 'unipept', 'query_tabular', 'ChangeCase', 'tp_tail_tool', 'collapse_dataset', 'annotatemyids', 'join1', 'goseq', 'heinz_scoring', 'tp_uniq_tool', 'export2graphlan', 'tp_cut_tool', 'seqtk_subseq', 'wc_gnu', 'bedtools_genomecoveragebed', 'wig_to_bigWig', 'clusterprofiler_bitr', 'bedtools_intersectbed', 'crossmap_bed', '__FLATTEN__', 'barchart_gnuplot', 'gtftobed12', 'Remove beginning1', 'macs2_callpeak', 'picard_FilterSamReads', 'pygenomeTracks', 'get_feature_info', 'bedtools_mergebed', 'ggplot_point', 'tp_replace_in_line', 'stringtie_merge', '__FILTER_FROM_FILE__', 'deseq2', 'mergeCols1', 'biotranslator', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'ggplot2_heatmap', 'tab2fasta', 'Grep1', 'msstats', 'table_compute', 'tp_grep_tool', 'egsea', 'add_line_to_file', 'fgsea', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'annotateMyIDs', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'goenrichment', 'regex1', 'tp_multijoin_tool', 'CONVERTER_interval_to_bed_0', 'gops_concat_1', 'bedtools_intersectbed_bam', 'tp_cat', 'charts', 'bcftools_consensus', 'fasterq_dump', 'ggplot2_heatmap2', 'rna_star', 'combine_metaphlan2_humann2', 'stacks_populations', 'Add_a_column1', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'w4mclassfilter', 'get_sequences', 'Kraken2Tax', 'cshl_find_and_replace', 'ont_fast5_api_fast5_subset', 'interactions', 'gops_coverage_1', 'heinz_bum', 'addName', 'mass_spectrometry_imaging_ion_images', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'Show tail1', 'trimmer', 'comp1', 'secure_hash_message_digest', 'volcanoplot', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'circos', 'gops_subtract_1', 'cat1', 'scatterplot_rpy', 'regex_replace', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'histogram_rpy', 'diff', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'interactive_tool_simtext_app', 'sklearn_ensemble', 'addValue', 'dexseq_count', 'cardinal_filtering', 'MassCalculator', 'split_file_to_collection', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'CONVERTER_bed_to_gff_0', 'Krona', 'bedtools_sortbed', 'param_value_from_file', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools: ['get_flanks1', 'join_files_on_column_fuzzy', 'ggplot_histogram', 'tp_uniq_tool', 'barchart_gnuplot', 'deseq2', 'mergeCols1', 'Grouping1', 'annotateMyIDs', 'bedtools_getfastabed', 'tp_sorted_uniq', 'goenrichment', 'rna_star', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'gops_coverage_1', 'comp1', 'regex_replace', '__TAG_FROM_FILE__', 'sklearn_ensemble', 'dexseq_count', 'regexColumn1', 'scanpy_read_10x', 'krona-text', 'heinz_scoring', 'wig_to_bigWig', 'clusterprofiler_bitr', 'crossmap_bed', 'ggplot_point', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'tab2fasta', 'msstats', 'table_compute', 'featurecounts', 'tp_cat', 'bcftools_consensus', 'Add_a_column1', 'w4mclassfilter', 'cshl_find_and_replace', 'addName', 'mass_spectrometry_imaging_ion_images', 'secure_hash_message_digest', 'MassCalculator', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'param_value_from_file', 'plotly_ml_performance_plots', 'stacks_denovomap', 'random_lines1', 'query_tabular', 'ChangeCase', 'goseq', 'seqtk_subseq', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', '__FLATTEN__', 'gtftobed12', 'picard_FilterSamReads', 'stringtie_merge', 'biotranslator', 'egsea', 'tp_replace_in_column', 'seq_filter_by_id', 'fasterq_dump', 'combine_metaphlan2_humann2', 'stacks_populations', 'get_sequences', 'Kraken2Tax', 'ont_fast5_api_fast5_subset', 'heinz_bum', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'sort1', 'addValue', 'ncbi_acc_download', 'venn_list', 'Paste1', 'join1', 'tp_cut_tool', 'get_feature_info', 'add_line_to_file', 'fgsea', 'regex1', 'bedtools_intersectbed_bam', 'interactions', 'Convert characters1', 'circos', 'cat1', 'scatterplot_rpy', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'diff', 'tp_head_tool', 'interactive_tool_simtext_app', 'CONVERTER_bed_to_gff_0', 'Krona', 'cshl_word_list_grep', 'collection_column_join', 'bg_uniq', 'XY_Plot_1', 'unipept', 'annotatemyids', 'export2graphlan', 'tp_easyjoin_tool', 'ggplot2_heatmap', 'tp_grep_tool', 'gops_join_1', 'tp_awk_tool', 'pygenomeTracks', 'trimmer', 'bedtools_sortbed', 'collapse_dataset', 'taxonomy_krona_chart', 'Show beginning1', 'Grep1', 'tp_find_and_replace', 'tp_tail_tool', 'Remove beginning1', 'hgv_david', 'volcanoplot', 'split_file_to_collection', 'bedtools_slopbed', 'Summary_Statistics1', 'macs2_callpeak', 'Show tail1', 'umi_tools_extract', '__FILTER_FROM_FILE__', 'tp_multijoin_tool', 'wc_gnu', '__RELABEL_FROM_FILE__', 'ggplot2_heatmap2', 'tp_sed_tool', 'deeptools_compute_matrix', 'datamash_ops', 'Count1', 'histogram_rpy', 'bg_column_arrange_by_header', 'bedtools_mergebed', 'CONVERTER_interval_to_bed_0', 'gops_subtract_1', 'cardinal_filtering', 'gops_concat_1', 'tp_split_on_column', 'charts']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'tp_replace_in_line', 'datamash_ops', 'bedtools_genomecoveragebed', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'rna_star', 'fasterq_dump', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'volcanoplot', 'tp_grep_tool', 'circos', 'tp_awk_tool', 'annotatemyids', 'taxonomy_krona_chart', 'tp_replace_in_column', 'filter_by_fasta_ids', 'goseq', 'gops_intersect_1', 'collection_column_join', 'limma_voom', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'Kraken2Tax', 'addValue', 'combine_metaphlan2_humann2', 'pygenomeTracks', 'addName', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'regexColumn1', 'seqtk_subseq', 'fgsea', 'XY_Plot_1', 'bedtools_mergebed', 'export2graphlan', 'goenrichment', 'query_tabular', 'bg_uniq', 'krona-text', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'gtftobed12', 'umi_tools_extract', 'stringtie_merge', 'dexseq_count', 'regex1', 'msstats', 'picard_FilterSamReads', 'ggplot2_heatmap', 'mergeCols1', 'tp_split_on_column', 'histogram_rpy', 'ncbi_blastdbcmd_wrapper', 'scatterplot_rpy', 'venn_list', 'egsea', 'join_files_on_column_fuzzy', 'add_line_to_file', 'bedtools_slopbed', 'dropletutils_read_10x', 'plotly_ml_performance_plots', 'get_flanks1', 'tp_uniq_tool', 'regex_replace', 'bg_column_arrange_by_header', 'sklearn_ensemble', 'stacks_denovomap', 'diff', 'gops_coverage_1', 'get_feature_info', 'checkFormat', 'get_sequences', 'heinz_scoring', 'heinz_bum', 'MassCalculator', 'gops_merge_1', 'macs2_bdgdiff', 'uniprot', 'edger', 'chipeakanno_annopeaks', 'xarray_select', 'deg_annotate', 'metilene', 'tabular_to_fastq', 'filter_tabular', 'bedtools_subtractbed', 'raceid_filtnormconf', 'htseq_count', 'tables_arithmetic_operations', 'Heatmap', 'chipseeker', 'cardinal_spectra_plots', 'fastq_groomer', 'column_remove_by_header', 'cummeRbund', 'plink', 'macs2_bdgpeakcall', 'tbl2gff3', 'humann2', 'scater_create_qcmetric_ready_sce', 'datamash_reverse', 'multiqc', 'column_order_header_sort', 'macs2_bdgcmp', 'hicexplorer_hicfindrestrictionsites', 'melt', 'vegan_diversity', 'get_pdb', 'graphprot_predict_profile', 'cor2', 'gops_cluster_1', 'bedtools_multiintersectbed', 'idr', 'intervene_upset', 'intervene_pairwise', 'sam_merge2', 'stoceps_trend_indic', 'sklearn_feature_selection', 'samtools_idxstats', 'rseqc_infer_experiment', 'vegan_rarefaction', 'ctb_chemfp_mol2fps', 'climate_stripes', 'Fetch Taxonomic Ranks', 'subtract_query1', 'sklearn_svm_classifier', 'flair_correct', 'transit_gumbel', 'kc-align', 'cardinal_classification']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3092105263157895\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.21710526315789475\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'hisat2', 'stringtie', 'tp_multijoin_tool', 'tp_cut_tool', 'table_compute', 'Filter1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'pampa_communitymetrics', 'tp_easyjoin_tool', 'join_files_on_column_fuzzy', 'htseq_count', 'regexColumn1', 'snpSift_filter', 'hgv_snpFreq', 'collection_column_join', 'seq_select_by_id', 'gprofiler_gost', 'bg_uniq', 'venn_list', 'tp_find_and_replace', 'Show beginning1', 'XY_Plot_1', 'changeGffFeatures', 'random_lines1', 'unipept', 'cshl_sort_header', 'ChangeCase', 'uniprot', 'aggregate_scores_in_intervals2', 'tp_tail_tool', 'bedtools_subtractbed', 'collapse_dataset', 'annotatemyids', 'join1', 'deg_annotate', 'tp_cut_tool', 'wc_gnu', 'bedtools_intersectbed', 'cardinal_combine', 'repeat_counter', 'transpose', 'dexseq_annotate', 'Remove beginning1', 'cardinal_mz_images', 'column_remove_by_header', 'bedtools_mergebed', 'tp_replace_in_line', 'deseq2', 'mergeCols1', 'vegan_rarefaction', 'Interval_Maf_Merged_Fasta2', 'tp_sort_header_tool', 'Grouping1', 'pandas_rolling_window', 'naive_variant_caller', 'gops_intersect_1', 'samtools_rmdup', 'Fetch Taxonomic Ranks', 'ggplot2_heatmap', 'trinity', 'tab2fasta', 'Grep1', 'SnpEff-cds-report', 'samtool_filter2', 'table_compute', 'samtools_view', 'tabular_to_csv', 'table_annovar', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'cooler_csort_pairix', '__FILTER_EMPTY_DATASETS__', 'gops_cluster_1', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'filter_tabular', 'goenrichment', 'snpEff', 'cardinal_spectra_plots', 'gdal_ogr2ogr', 'tp_multijoin_tool', 'gops_concat_1', 'tp_cat', 'charts', 'bcftools_consensus', 'vcfbedintersect', 'sam_merge2', 'ggplot2_heatmap2', 'Add_a_column1', 'stoceps_glm', 'datamash_transpose', 'stoceps_trend_indic', 'gops_coverage_1', 'chipseeker', 'snpSift_rmInfo', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'diapysef', 'Show tail1', 'trimmer', 'comp1', 'ggplot2_point', 'volcanoplot', 'multiqc', 'tbl2gff3', 'Cut1', 'CONVERTER_vcf_bgzip_to_tabix_0', 'jbrowse', 'gops_subtract_1', 'minimap2', 'cat1', 'Extract genomic DNA 1', 'tab2csv_R', 'vegan_diversity', 'pathview', 'pampa_glmsp', 'histogram_rpy', '__BUILD_LIST__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'addValue', 'bcftools_norm', 'dexseq_count', 'cardinal_filtering', 'split_file_to_collection', '__EXTRACT_DATASET__', 'samtools_fastx', 'bedtools_sortbed', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'bedtools_maskfastabed', 'fastq_filter', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 135 tools: ['tp_easyjoin_tool', 'join_files_on_column_fuzzy', 'Show beginning1', 'Grouping1', 'samtools_rmdup', 'Grep1', 'samtool_filter2', 'table_annovar', 'datamash_ops', 'tp_sorted_uniq', 'tp_multijoin_tool', 'datamash_transpose', 'snpSift_rmInfo', 'bg_column_arrange_by_header', 'tp_find_and_replace', 'changeGffFeatures', 'tp_tail_tool', 'cardinal_combine', 'cardinal_mz_images', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'featurecounts', 'charts', 'bcftools_consensus', 'Add_a_column1', 'chipseeker', 'hgv_david', 'random_lines1', 'ChangeCase', 'bedtools_intersectbed', 'Fetch Taxonomic Ranks', 'vcfbedintersect', 'ggplot2_heatmap2', 'stoceps_trend_indic', 'jbrowse', 'sort1', 'addValue', 'bcftools_norm', 'gprofiler_gost', 'bedtools_subtractbed', 'join1', 'tp_cut_tool', 'column_remove_by_header', 'gops_concat_1', 'tp_sed_tool', 'Cut1', 'cat1', 'bedtools_sortbed', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'Summary_Statistics1', 'collection_column_join', 'bg_uniq', 'XY_Plot_1', 'unipept', 'uniprot', 'annotatemyids', 'dexseq_annotate', 'mergeCols1', '__FILTER_EMPTY_DATASETS__', 'gops_cluster_1', 'goenrichment', 'cardinal_filtering', 'bedtools_maskfastabed', 'transpose', 'SnpEff-cds-report', 'snpEff', 'deg_annotate', 'tp_replace_in_column', 'CONVERTER_vcf_bgzip_to_tabix_0', 'split_file_to_collection', 'pathview', 'cshl_word_list_grep', 'collapse_dataset', 'cshl_sort_header', 'repeat_counter', 'pandas_rolling_window', 'bedtools_getfastabed', 'sam_merge2', 'stoceps_glm', 'gops_coverage_1', 'dexseq_count', 'htseq_count', 'regexColumn1', 'hgv_snpFreq', 'volcanoplot', 'tab2csv_R', 'cooler_csort_pairix', 'diapysef', 'ggplot2_point', 'tbl2gff3', 'gops_join_1', 'tp_awk_tool', 'venn_list', 'bedtools_mergebed', 'Interval_Maf_Merged_Fasta2', 'gdal_ogr2ogr', 'snpSift_filter', 'seq_select_by_id', 'aggregate_scores_in_intervals2', 'naive_variant_caller', 'trinity', 'comp1', 'cardinal_spectra_plots', 'tp_cat', 'tabular_to_csv', 'minimap2', 'fastq_filter', 'wc_gnu', 'gops_subtract_1', 'Remove beginning1', 'Convert characters1', 'vegan_diversity', 'vegan_rarefaction', 'deseq2', 'ggplot2_heatmap', 'Show tail1', 'multiqc', 'histogram_rpy', 'table_compute', 'ncbi_acc_download', 'pampa_glmsp', 'samtools_fastx', 'pampa_communitymetrics', 'filter_tabular', 'Count1', 'tp_head_tool', 'seq_filter_by_id', '__BUILD_LIST__', 'samtools_view', '__EXTRACT_DATASET__', 'trimmer', 'tab2fasta', 'Extract genomic DNA 1', 'deeptools_compute_matrix']\n",
      "\n",
      "Test batch 1, Predicted top 135 tools with weights: ['Add_a_column1', 'snpSift_filter', 'tp_find_and_replace', 'tp_replace_in_line', 'samtools_view', 'datamash_ops', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'filter_tabular', 'multiqc', 'minimap2', 'htseq_count', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'samtools_rmdup', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'snpEff', 'volcanoplot', 'tp_awk_tool', 'annotatemyids', 'fastq_filter', 'deg_annotate', 'tp_replace_in_column', 'gops_intersect_1', 'collection_column_join', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'samtool_filter2', 'addValue', 'bedtools_maskfastabed', 'trinity', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'sam_merge2', 'jbrowse', 'regexColumn1', 'XY_Plot_1', 'bedtools_mergebed', 'goenrichment', 'bcftools_norm', 'bg_uniq', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'tbl2gff3', 'dexseq_count', 'chipseeker', 'bedtools_subtractbed', 'ggplot2_heatmap', 'mergeCols1', 'histogram_rpy', 'naive_variant_caller', 'venn_list', 'cardinal_mz_images', 'join_files_on_column_fuzzy', 'cardinal_spectra_plots', 'gprofiler_gost', 'vcfbedintersect', 'column_remove_by_header', 'bg_column_arrange_by_header', 'cardinal_combine', 'uniprot', 'vegan_diversity', 'SnpEff-cds-report', 'gops_coverage_1', 'diapysef', 'Fetch Taxonomic Ranks', 'vegan_rarefaction', 'pampa_glmsp', 'stoceps_trend_indic', 'stoceps_glm', 'gops_cluster_1', 'gtftobed12', 'fastqc', 'samtools_slice_bam', 'addName', 'filter_by_fasta_ids', 'tp_split_on_column', 'cutadapt', 'goseq', 'plotly_ml_performance_plots', 'cardinal_classification', 'xarray_select', 'rna_star', 'eggnog_mapper', 'picard_FilterSamReads', 'chipeakanno_annopeaks', 'picard_MarkDuplicates', 'snpfreqplot', 'cummeRbund', 'rcas', 'circos_wiggle_to_stacked', 'gops_merge_1', 'egsea', 'metaspades', 'qualimap_bamqc', 'blockbuster', 'taxonomy_krona_chart', 'fgsea', 'bedtools_windowbed', 'snpSift_extractFields', 'gemini_query', 'macs2_bdgdiff', 'msstats', 'get_sequences', 'samtools_depth', 'rseqc_infer_experiment', 'ctb_multi_obgrep', 'diffbind', 'samtools_idxstats', 'tp_sort_rows', 'ncbi_blastdbcmd_wrapper', 'regex_replace', 'transit_gumbel', 'bedtools_makewindowsbed', 'dropletutils_read_10x', 'heinz_bum', 'kraken2', 'gd_rank_terms']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.1925925925925926\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.1111111111111111\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['picard_NormalizeFasta', 'bowtie2', 'hicexplorer_hicbuildmatrix', 'hicexplorer_hicadjustmatrix', 'hicexplorer_hiccorrectmatrix']\n",
      "\n",
      "Test batch 1, True tools: ['hicexplorer_hiccomparematrices', 'hicexplorer_hictransform', 'hicexplorer_hicplotdistvscounts', 'hicexplorer_hicfindtads', 'hicexplorer_hicnormalize', 'hicexplorer_hicplottads', 'hicexplorer_hicplotmatrix', 'pygenomeTracks', 'hicexplorer_hicpca', 'hicexplorer_hicinfo']\n",
      "\n",
      "Test batch 1, Predicted top 10 tools: ['hicexplorer_hicpca', 'hicexplorer_hicplottads', 'pygenomeTracks', 'hicexplorer_hicnormalize', 'hicexplorer_hicplotdistvscounts', 'hicexplorer_hiccomparematrices', 'hicexplorer_hicplotmatrix', 'hicexplorer_hicinfo', 'hicexplorer_hictransform', 'hicexplorer_hicfindtads']\n",
      "\n",
      "Test batch 1, Predicted top 10 tools with weights: ['pygenomeTracks', 'hicexplorer_hicplotmatrix', 'hicexplorer_hicfindtads', 'hicexplorer_hicnormalize', 'hicexplorer_hicinfo', 'hicexplorer_hicpca', 'hicexplorer_hiccomparematrices', 'hicexplorer_hicplotdistvscounts', 'hicexplorer_hictransform', 'hicexplorer_hicmergematrixbins']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.2\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'fastqc', 'Grep1', '__FLATTEN__', 'collapse_dataset', 'Convert characters1', 'Cut1', 'collection_column_join', 'regex_replace', 'melt', 'join1', 'Add_a_column1', 'Filter1', 'wc_gnu', 'addValue', 'Grouping1', 'cat1']\n",
      "\n",
      "Test batch 1, True tools: ['pureclip', 'uclust2otutable', 'tp_easyjoin_tool', 'mothur_unique_seqs', 'join_files_on_column_fuzzy', 'regexColumn1', 'ggplot2_pca', 'anndata_manipulate', 'cardinal_classification', 'bg_uniq', 'tp_find_and_replace', 'porechop', 'Paste1', 'ggplot_histogram', 'purge_dups', 'fastq_paired_end_interlacer', 'XY_Plot_1', 'rbc_mafft', 'random_lines1', 'unipept', 'query_tabular', 'search_gui', 'p_clip_peaks', 'targetfinder', 'ChangeCase', 'uniprot', 'mothur_phylotype', 'gops_merge_1', 'CONVERTER_fasta_to_tabular', 'annotatemyids', 'join1', 'qiime_pick_closed_reference_otus', 'salsa', 'tp_cut_tool', 'cshl_fastx_renamer', 'mass_spectrometry_imaging_combine', 'cshl_fastx_barcode_splitter', 'bedtools_intersectbed', 'cardinal_combine', '__FLATTEN__', 'minia', 'fastq_groomer', 'cd_hit', '_ensembl_gtf2gene_list', 'barchart_gnuplot', 'cshl_fastx_trimmer', 'cd_hit_est', 'fastq_to_fasta_python', 'quast', 'mergeCols1', 'umi_tools_extract', 'tp_sort_header_tool', 'msstatstmt', 'Grouping1', 'regionalgam_flight_curve', 'filter_by_fasta_ids', 'trinity', 'tab2fasta', 'Grep1', 'table_compute', 'regionalgam_ab_index', 'add_line_to_file', 'tp_replace_in_column', 'bedtools_getfastabed', 'datamash_ops', 'trimmomatic', 'tp_sorted_uniq', 'cardinal_spectra_plots', 'seq_filter_by_mapping', 'fasta2tab', 'fastq_to_tabular', 'heatmap2', 'ggplot2_heatmap2', 'rna_star', 'bwa_mem', 'Add_a_column1', 'fastqc', 'cutadapt', 'cshl_fastx_clipper', 'melt', 'checkFormat', 'rbc_mirdeep2_mapper', 'datamash_transpose', 'w4mclassfilter', 'stoceps_trend_indic', 'snpfreqplot', 'CONVERTER_interval_to_bedstrict_0', 'cshl_fastx_collapser', 'salmon', 'Convert characters1', 'fastq_paired_end_deinterlacer', 'ggplot2_point', 'volcanoplot', 'multiqc', 'Cut1', 'jbrowse', 'minimap2', 'column_order_header_sort', 'ete_homology_classifier', 'ggplot2_histogram', 'merqury', 'uc2otutable', 'Filter1', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'sort1', 'ctb_chemfp_mol2fps', 'bismark_bowtie2', 'bowtie2', 'rna_starsolo', 'addValue', 'cardinal_filtering', 'cast', 'proteomics_moff', 'busco', 'bedtools_sortbed', 'cshl_fastq_to_fasta', 'fastq_filter', 'hisat2', 'Summary_Statistics1', 'trim_galore']\n",
      "\n",
      "Test batch 1, Predicted top 121 tools: ['tp_easyjoin_tool', 'barchart_gnuplot', 'Grep1', 'datamash_ops', 'fastq_to_tabular', '__TAG_FROM_FILE__', 'hisat2', 'regexColumn1', 'fastq_groomer', 'filter_by_fasta_ids', 'bwa_mem', 'Add_a_column1', 'query_tabular', 'bedtools_intersectbed', 'tp_replace_in_column', 'trimmomatic', 'sort1', 'Paste1', 'tp_cut_tool', 'add_line_to_file', 'Filter1', 'XY_Plot_1', 'tp_sorted_uniq', 'rbc_mafft', 'cshl_fastq_to_fasta', 'Cut1', 'minimap2', 'fastq_filter', 'unipept', 'join_files_on_column_fuzzy', 'porechop', 'fastq_to_fasta_python', 'Grouping1', 'rna_star', 'merqury', 'table_compute', 'random_lines1', 'ChangeCase', 'seq_filter_by_mapping', 'addValue', 'fastq_paired_end_interlacer', 'join1', 'cutadapt', 'bedtools_sortbed', 'ggplot_histogram', 'mergeCols1', 'bedtools_getfastabed', '__RELABEL_FROM_FILE__', 'ggplot2_heatmap2', 'melt', 'Summary_Statistics1', 'annotatemyids', 'multiqc', 'cardinal_filtering', 'w4mclassfilter', 'volcanoplot', '__FLATTEN__', 'minia', 'cshl_fastx_renamer', 'column_order_header_sort', 'cd_hit_est', 'cshl_fastx_collapser', 'umi_tools_extract', 'fastq_paired_end_deinterlacer', 'pureclip', 'trinity', 'checkFormat', 'ggplot2_point', 'bg_uniq', 'datamash_transpose', 'busco', 'targetfinder', 'mothur_phylotype', 'uniprot', 'bismark_bowtie2', 'quast', 'cshl_fastx_clipper', 'CONVERTER_interval_to_bedstrict_0', 'purge_dups', 'tab2fasta', 'ete_homology_classifier', 'mass_spectrometry_imaging_combine', 'tp_sort_header_tool', 'trim_galore', 'uclust2otutable', 'mothur_unique_seqs', 'uc2otutable', 'p_clip_peaks', 'cshl_fastx_barcode_splitter', 'msstatstmt', 'cardinal_classification', 'anndata_manipulate', 'CONVERTER_fasta_to_tabular', 'fasta2tab', 'rna_starsolo', 'salsa', 'cshl_fastx_trimmer', 'tp_find_and_replace', 'rbc_mirdeep2_mapper', 'gops_merge_1', 'jbrowse', 'Convert characters1', 'qiime_pick_closed_reference_otus', 'proteomics_moff', 'cd_hit', '_ensembl_gtf2gene_list', 'stoceps_trend_indic', 'ggplot2_histogram', 'cast', 'cardinal_combine', 'regionalgam_flight_curve', 'ggplot2_pca', 'ctb_chemfp_mol2fps', 'bowtie2', 'regionalgam_ab_index', 'snpfreqplot', 'heatmap2', 'fastqc', 'search_gui', 'salmon', 'cardinal_spectra_plots']\n",
      "\n",
      "Test batch 1, Predicted top 121 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'bwa_mem', 'datamash_ops', 'gops_merge_1', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'trimmomatic', 'hisat2', 'multiqc', 'minimap2', 'fastq_to_fasta_python', 'trim_galore', 'datamash_transpose', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'fastq_groomer', 'bedtools_intersectbed', 'porechop', 'fasta2tab', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'salmon', 'busco', 'quast', 'volcanoplot', 'annotatemyids', 'fastq_filter', 'tp_replace_in_column', 'fastq_paired_end_interlacer', 'filter_by_fasta_ids', 'mothur_unique_seqs', 'cshl_fastq_to_fasta', 'rbc_mirdeep2_mapper', 'addValue', 'snpfreqplot', 'trinity', 'bedtools_getfastabed', 'bedtools_sortbed', 'cshl_fastx_trimmer', 'cd_hit', 'ggplot2_point', 'anndata_manipulate', 'cshl_fastx_renamer', 'jbrowse', 'regexColumn1', 'fastq_paired_end_deinterlacer', 'fastq_to_tabular', 'XY_Plot_1', 'cd_hit_est', 'rna_starsolo', 'query_tabular', 'bg_uniq', 'rbc_mafft', 'cshl_fastx_collapser', 'tp_sorted_uniq', 'cardinal_filtering', 'purge_dups', 'umi_tools_extract', 'cshl_fastx_barcode_splitter', 'seq_filter_by_mapping', 'mergeCols1', 'search_gui', 'merqury', 'bismark_bowtie2', 'cshl_fastx_clipper', '_ensembl_gtf2gene_list', 'ggplot2_pca', 'pureclip', 'cardinal_classification', 'salsa', 'join_files_on_column_fuzzy', 'add_line_to_file', 'cardinal_spectra_plots', 'minia', 'targetfinder', 'column_order_header_sort', 'cardinal_combine', 'uniprot', 'ggplot2_histogram', 'msstatstmt', 'ctb_chemfp_mol2fps', 'melt', 'qiime_pick_closed_reference_otus', 'checkFormat', 'cast', 'ete_homology_classifier', 'stoceps_trend_indic', 'regionalgam_ab_index', 'regionalgam_flight_curve', 'mothur_phylotype', 'tp_grep_tool', 'split_file_to_collection', 'goseq', 'fasterq_dump', 'tp_head_tool', 'tp_tail_tool', 'filter_tabular', 'combine_metaphlan2_humann2', 'gtftobed12', 'taxonomy_krona_chart', 'tp_replace_in_line', 'addName', 'stacks_denovomap', 'tp_split_on_column', 'featurecounts', 'export2graphlan', 'kraken2', 'ncbi_acc_download', 'fgsea', 'ncbi_blastdbcmd_wrapper', 'pathview', 'egsea', 'tp_cat', 'limma_voom', 'msstats', 'dropletutils_read_10x', 'Heatmap']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.14049586776859505\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.09917355371900827\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['cutadapt', 'hifiasm', 'gfa_to_fa', 'cat1', 'salsa', 'bwa_mem']\n",
      "\n",
      "Test batch 1, True tools: ['rseqc_bam_stat', 'samtools_fixmate', 'picard_AddOrReplaceReadGroups', 'htseq_count', 'bellerophon', 'ococo', 'picard_ReplaceSamHeader', 'CONVERTER_bam_to_coodinate_sorted_bam', 'ngsutils_bam_filter', 'rseqc_geneBody_coverage', 'picard_MarkDuplicates', 'samtools_filter', 'hicexplorer_hicquickqc', 'pilon', 'bamFilter', 'ivar_trim', 'hypo', 'picard_CollectInsertSizeMetrics', 'picard_ValidateSamFile', 'deeptools_plot_fingerprint', 'picard_MergeSamFiles', 'bedtools_genomecoveragebed', 'picard_BamIndexStats', 'bcftools_mpileup', 'lumpy_prep', 'bedtools_intersectbed', 'PicardASMetrics', 'samtools_stats', 'CONVERTER_Bam_Bai_0', 'picard_FixMateInformation', 'macs2_callpeak', 'picard_FilterSamReads', 'gemini_load', 'deeptools_bam_coverage', 'picard_ARRG', 'gatk4_mutect2', 'samtools_reheader', 'naive_variant_caller', 'lofreq_indelqual', 'samtools_rmdup', 'picard_SortSam', 'samtool_filter2', 'ivar_variants', 'samtools_view', 'samtools_markdup', 'hicexplorer_hicbuildmatrix', 'featurecounts', 'picard_DownsampleSam', 'pretext_map', 'samtools_depth', 'je_markdupes', 'lofreq_viterbi', 'stringtie', 'samtools_sort', 'sam_merge2', 'deeptools_multi_bam_summary', 'samtools_slice_bam', 'fastqc', 'deeptools_plot_coverage', 'exomedepth', 'samtools_flagstat', 'ivar_consensus', 'picard_CollectWgsMetrics', 'picard_CASM', 'gemini_annotate', 'jvarkit_wgscoverageplotter', 'read2mut', 'picard_CleanSam', 'pe_histogram', 'picard_ReorderSam', 'bam_to_sam', 'samtools_merge', 'bamtools', 'samtools_mpileup', 'sam_pileup', 'lofreq_call', 'jbrowse', 'deeptools_bam_compare', 'sam_to_bam', 'samtools_idxstats', 'bamleftalign', '__BUILD_LIST__', 'qualimap_bamqc', 'picard_MarkDuplicatesWithMateCigar', 'freebayes', 'picard_MergeBamAlignment', 'samtools_fastx']\n",
      "\n",
      "Test batch 1, Predicted top 87 tools: ['CONVERTER_bam_to_coodinate_sorted_bam', 'samtools_filter', 'bcftools_mpileup', 'macs2_callpeak', 'deeptools_bam_coverage', 'picard_ARRG', 'samtools_rmdup', 'samtool_filter2', 'sam_merge2', 'ivar_consensus', 'samtools_merge', 'samtools_mpileup', 'htseq_count', 'ococo', 'picard_ReplaceSamHeader', 'rseqc_geneBody_coverage', 'ivar_trim', 'picard_CollectInsertSizeMetrics', 'picard_BamIndexStats', 'picard_SortSam', 'samtools_markdup', 'featurecounts', 'je_markdupes', 'samtools_slice_bam', 'bamtools', 'deeptools_bam_compare', 'qualimap_bamqc', 'rseqc_bam_stat', 'ngsutils_bam_filter', 'bamFilter', 'deeptools_plot_fingerprint', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', 'samtools_stats', 'picard_FixMateInformation', 'picard_FilterSamReads', 'picard_DownsampleSam', 'lofreq_viterbi', 'samtools_sort', 'fastqc', 'exomedepth', 'picard_CollectWgsMetrics', 'picard_CleanSam', 'picard_ReorderSam', 'bam_to_sam', 'jbrowse', 'sam_to_bam', 'samtools_fixmate', 'samtools_view', 'deeptools_multi_bam_summary', 'samtools_flagstat', 'jvarkit_wgscoverageplotter', 'samtools_idxstats', 'bamleftalign', 'freebayes', 'picard_AddOrReplaceReadGroups', 'bellerophon', 'samtools_reheader', 'naive_variant_caller', 'lumpy_prep', 'hypo', 'gemini_annotate', 'CONVERTER_Bam_Bai_0', 'read2mut', 'lofreq_call', 'lofreq_indelqual', 'sam_pileup', 'samtools_fastx', 'gatk4_mutect2', 'ivar_variants', 'picard_MergeBamAlignment', 'hicexplorer_hicquickqc', 'picard_ValidateSamFile', 'picard_MergeSamFiles', 'gemini_load', '__BUILD_LIST__', 'pilon', 'PicardASMetrics', 'hicexplorer_hicbuildmatrix', 'deeptools_plot_coverage', 'picard_MarkDuplicatesWithMateCigar', 'pe_histogram', 'picard_MarkDuplicates', 'pretext_map', 'picard_CASM', 'samtools_depth', 'stringtie']\n",
      "\n",
      "Test batch 1, Predicted top 87 tools with weights: ['lofreq_call', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'ivar_trim', 'lofreq_indelqual', 'lofreq_viterbi', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'bedtools_intersectbed', 'macs2_callpeak', 'samtools_rmdup', 'ivar_consensus', 'bamFilter', 'freebayes', 'samtools_flagstat', 'samtools_idxstats', 'bamleftalign', 'ivar_variants', 'ngsutils_bam_filter', 'samtools_sort', 'rseqc_geneBody_coverage', 'samtool_filter2', 'samtools_slice_bam', 'deeptools_bam_compare', 'pe_histogram', 'bcftools_mpileup', 'bamtools', 'rseqc_bam_stat', 'picard_SortSam', 'sam_merge2', 'jbrowse', 'jvarkit_wgscoverageplotter', 'samtools_merge', 'picard_CollectInsertSizeMetrics', 'samtools_markdup', 'deeptools_multi_bam_summary', 'deeptools_plot_fingerprint', 'samtools_depth', 'picard_CASM', 'picard_MergeSamFiles', 'picard_FilterSamReads', 'PicardASMetrics', 'bellerophon', 'pretext_map', 'sam_pileup', 'deeptools_plot_coverage', 'naive_variant_caller', 'samtools_fixmate', 'pilon', 'samtools_mpileup', 'ococo', 'lumpy_prep', 'picard_BamIndexStats', 'hicexplorer_hicbuildmatrix', 'gatk4_mutect2', 'picard_DownsampleSam', 'bam_to_sam', 'picard_ValidateSamFile', 'picard_MergeBamAlignment', 'hicexplorer_hicquickqc', 'picard_AddOrReplaceReadGroups', 'hypo', 'picard_CollectWgsMetrics', 'exomedepth', 'picard_FixMateInformation', 'picard_ReorderSam', 'picard_ARRG', 'picard_ReplaceSamHeader', 'read2mut', 'samtools_filter', 'deeptools_alignmentsieve', 'rseqc_infer_experiment', 'mimodd_varcall', 'diffbind', 'pileometh', 'bedtools_bamtobed', 'strelka_germline', 'mut2sscs', 'rseqc_read_distribution', 'medaka_variant', 'bedtools_multicovtbed']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.10344827586206896\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.09195402298850575\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['FeatureFinderMultiplex', 'IDMapper', 'FileFilter', 'IDConflictResolver', 'FileMerger']\n",
      "\n",
      "Test batch 1, True tools: ['FileInfo', 'MzTabExporter', 'ProteinQuantifier', 'Epifany']\n",
      "\n",
      "Test batch 1, Predicted top 4 tools: ['Epifany', 'FileInfo', 'ProteinQuantifier', 'MzTabExporter']\n",
      "\n",
      "Test batch 1, Predicted top 4 tools with weights: ['Epifany', 'MzTabExporter', 'tbvcfreport', 'ConsensusMapNormalizer']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.25\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.25\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'hisat2', 'samtools_flagstat', 'Show beginning1', 'Show tail1', 'Convert characters1', 'Grouping1', 'join1', 'Filter1', 'comp1', 'wc_gnu', 'addValue', 'Cut1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'join_files_on_column_fuzzy', 'regexColumn1', 'collection_column_join', 'plotly_ml_performance_plots', 'bg_uniq', 'venn_list', 'scanpy_read_10x', 'tp_find_and_replace', 'bedtools_slopbed', 'Paste1', 'Show beginning1', 'ggplot_histogram', 'krona-text', 'stacks_denovomap', 'XY_Plot_1', 'random_lines1', 'unipept', 'query_tabular', 'ChangeCase', 'tp_tail_tool', 'collapse_dataset', 'annotatemyids', 'join1', 'goseq', 'heinz_scoring', 'tp_uniq_tool', 'export2graphlan', 'tp_cut_tool', 'seqtk_subseq', 'wc_gnu', 'bedtools_genomecoveragebed', 'wig_to_bigWig', 'clusterprofiler_bitr', 'bedtools_intersectbed', 'crossmap_bed', '__FLATTEN__', 'barchart_gnuplot', 'gtftobed12', 'Remove beginning1', 'macs2_callpeak', 'picard_FilterSamReads', 'pygenomeTracks', 'get_feature_info', 'bedtools_mergebed', 'ggplot_point', 'tp_replace_in_line', 'stringtie_merge', '__FILTER_FROM_FILE__', 'deseq2', 'mergeCols1', 'biotranslator', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'ggplot2_heatmap', 'tab2fasta', 'Grep1', 'msstats', 'table_compute', 'tp_grep_tool', 'egsea', 'add_line_to_file', 'fgsea', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'annotateMyIDs', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'goenrichment', 'regex1', 'tp_multijoin_tool', 'CONVERTER_interval_to_bed_0', 'gops_concat_1', 'bedtools_intersectbed_bam', 'tp_cat', 'charts', 'bcftools_consensus', 'fasterq_dump', 'ggplot2_heatmap2', 'rna_star', 'combine_metaphlan2_humann2', 'stacks_populations', 'Add_a_column1', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'w4mclassfilter', 'get_sequences', 'Kraken2Tax', 'cshl_find_and_replace', 'ont_fast5_api_fast5_subset', 'interactions', 'gops_coverage_1', 'heinz_bum', 'addName', 'mass_spectrometry_imaging_ion_images', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'Show tail1', 'trimmer', 'comp1', 'secure_hash_message_digest', 'volcanoplot', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'circos', 'gops_subtract_1', 'cat1', 'scatterplot_rpy', 'regex_replace', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'histogram_rpy', 'diff', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'interactive_tool_simtext_app', 'sklearn_ensemble', 'addValue', 'dexseq_count', 'cardinal_filtering', 'MassCalculator', 'split_file_to_collection', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'CONVERTER_bed_to_gff_0', 'Krona', 'bedtools_sortbed', 'param_value_from_file', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools: ['get_flanks1', 'join_files_on_column_fuzzy', 'ggplot_histogram', 'tp_uniq_tool', 'barchart_gnuplot', 'deseq2', 'mergeCols1', 'Grouping1', 'ggplot2_heatmap', 'annotateMyIDs', 'bedtools_getfastabed', 'tp_sorted_uniq', 'goenrichment', 'rna_star', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'gops_coverage_1', 'comp1', 'regex_replace', '__TAG_FROM_FILE__', 'sklearn_ensemble', 'dexseq_count', 'regexColumn1', 'scanpy_read_10x', 'krona-text', 'tp_tail_tool', 'heinz_scoring', 'wig_to_bigWig', 'clusterprofiler_bitr', 'crossmap_bed', 'Remove beginning1', 'ggplot_point', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'tab2fasta', 'msstats', 'table_compute', 'tp_cat', 'bcftools_consensus', 'Add_a_column1', 'w4mclassfilter', 'cshl_find_and_replace', 'addName', 'mass_spectrometry_imaging_ion_images', 'secure_hash_message_digest', 'MassCalculator', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'param_value_from_file', 'plotly_ml_performance_plots', 'stacks_denovomap', 'random_lines1', 'query_tabular', 'ChangeCase', 'goseq', 'seqtk_subseq', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', '__FLATTEN__', 'gtftobed12', 'picard_FilterSamReads', 'stringtie_merge', 'biotranslator', 'tp_grep_tool', 'egsea', 'tp_replace_in_column', 'seq_filter_by_id', 'fasterq_dump', 'combine_metaphlan2_humann2', 'stacks_populations', 'get_sequences', 'Kraken2Tax', 'ont_fast5_api_fast5_subset', 'heinz_bum', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'gops_join_1', 'sort1', 'tp_awk_tool', 'addValue', 'ncbi_acc_download', 'Paste1', 'join1', 'tp_cut_tool', 'get_feature_info', 'add_line_to_file', 'fgsea', 'regex1', 'interactions', 'trimmer', 'circos', 'cat1', 'scatterplot_rpy', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'diff', 'tp_head_tool', 'interactive_tool_simtext_app', 'CONVERTER_bed_to_gff_0', 'Krona', 'cshl_word_list_grep', 'collection_column_join', 'bg_uniq', 'XY_Plot_1', 'annotatemyids', 'export2graphlan', 'taxonomy_krona_chart', 'Show beginning1', 'featurecounts', 'hgv_david', 'ggplot2_heatmap2', 'split_file_to_collection', 'venn_list', 'bedtools_slopbed', 'pygenomeTracks', 'Convert characters1', 'bedtools_sortbed', 'Summary_Statistics1', 'unipept', 'tp_easyjoin_tool', 'macs2_callpeak', 'tp_multijoin_tool', 'Show tail1', 'tp_find_and_replace', 'volcanoplot', '__FILTER_FROM_FILE__', 'bedtools_intersectbed_bam', 'tp_sed_tool', 'Count1', 'collapse_dataset', 'Grep1', '__RELABEL_FROM_FILE__', 'deeptools_compute_matrix', 'datamash_ops', 'umi_tools_extract', 'wc_gnu', 'histogram_rpy', 'bedtools_mergebed', 'gops_subtract_1', 'bg_column_arrange_by_header', 'CONVERTER_interval_to_bed_0', 'cardinal_filtering', 'gops_concat_1', 'charts', 'tp_split_on_column']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'tp_replace_in_line', 'datamash_ops', 'bedtools_genomecoveragebed', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'rna_star', 'fasterq_dump', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'volcanoplot', 'tp_grep_tool', 'circos', 'tp_awk_tool', 'annotatemyids', 'taxonomy_krona_chart', 'tp_replace_in_column', 'filter_by_fasta_ids', 'goseq', 'gops_intersect_1', 'collection_column_join', 'limma_voom', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'Kraken2Tax', 'addValue', 'combine_metaphlan2_humann2', 'pygenomeTracks', 'addName', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'regexColumn1', 'seqtk_subseq', 'fgsea', 'XY_Plot_1', 'bedtools_mergebed', 'export2graphlan', 'goenrichment', 'query_tabular', 'bg_uniq', 'krona-text', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'gtftobed12', 'umi_tools_extract', 'stringtie_merge', 'dexseq_count', 'regex1', 'msstats', 'picard_FilterSamReads', 'ggplot2_heatmap', 'mergeCols1', 'tp_split_on_column', 'histogram_rpy', 'ncbi_blastdbcmd_wrapper', 'scatterplot_rpy', 'venn_list', 'egsea', 'join_files_on_column_fuzzy', 'add_line_to_file', 'bedtools_slopbed', 'dropletutils_read_10x', 'plotly_ml_performance_plots', 'get_flanks1', 'tp_uniq_tool', 'regex_replace', 'bg_column_arrange_by_header', 'sklearn_ensemble', 'stacks_denovomap', 'diff', 'gops_coverage_1', 'get_feature_info', 'checkFormat', 'get_sequences', 'heinz_scoring', 'heinz_bum', 'MassCalculator', 'gops_merge_1', 'macs2_bdgdiff', 'uniprot', 'edger', 'xarray_select', 'chipeakanno_annopeaks', 'deg_annotate', 'metilene', 'tabular_to_fastq', 'filter_tabular', 'raceid_filtnormconf', 'bedtools_subtractbed', 'htseq_count', 'Heatmap', 'tables_arithmetic_operations', 'chipseeker', 'column_remove_by_header', 'cardinal_spectra_plots', 'cummeRbund', 'macs2_bdgpeakcall', 'multiqc', 'plink', 'tbl2gff3', 'scater_create_qcmetric_ready_sce', 'datamash_reverse', 'humann2', 'macs2_bdgcmp', 'rseqc_infer_experiment', 'fastq_groomer', 'vegan_diversity', 'graphprot_predict_profile', 'melt', 'idr', 'bedtools_multiintersectbed', 'intervene_upset', 'gops_cluster_1', 'intervene_pairwise', 'column_order_header_sort', 'samtools_idxstats', 'vegan_rarefaction', 'get_pdb', 'stoceps_trend_indic', 'hicexplorer_hicfindrestrictionsites', 'sklearn_feature_selection', 'sklearn_svm_classifier', 'Fetch Taxonomic Ranks', 'subtract_query1', 'ggplot2_pca', 'transit_gumbel', 'flair_correct', 'ctb_chemfp_mol2fps', 'trim_galore', 'kc-align', 'transit_hmm', 'gd_rank_terms']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3092105263157895\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.21710526315789475\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['hisat2', 'hicexplorer_hicbuildmatrix']\n",
      "\n",
      "Test batch 1, True tools: ['hicexplorer_hicsummatrices', 'hicexplorer_hictransform', 'hicexplorer_hicmergematrixbins', 'hicexplorer_hicfindtads', 'hicexplorer_hicplotviewpoint', 'hicexplorer_chicqualitycontrol', 'multiqc', 'hicexplorer_hiccorrectmatrix', 'hicexplorer_hicplottads', 'hicexplorer_hicadjustmatrix', 'hicexplorer_hicplotmatrix', 'tp_sed_tool', 'hicexplorer_chicviewpoint', 'hicexplorer_hicpca', 'hicexplorer_chicviewpointbackgroundmodel']\n",
      "\n",
      "Test batch 1, Predicted top 15 tools: ['hicexplorer_chicviewpoint', 'hicexplorer_chicviewpointbackgroundmodel', 'multiqc', 'hicexplorer_hicpca', 'hicexplorer_hicplottads', 'hicexplorer_hicadjustmatrix', 'hicexplorer_chicqualitycontrol', 'hicexplorer_hiccorrectmatrix', 'hicexplorer_hicplotviewpoint', 'hicexplorer_hicsummatrices', 'hicexplorer_hictransform', 'hicexplorer_hicplotmatrix', 'hicexplorer_hicmergematrixbins', 'hicexplorer_hicfindtads', 'tp_sed_tool']\n",
      "\n",
      "Test batch 1, Predicted top 15 tools with weights: ['multiqc', 'tp_sed_tool', 'hicexplorer_hicplotmatrix', 'hicexplorer_hicmergematrixbins', 'hicexplorer_hicfindtads', 'hicexplorer_hicplotviewpoint', 'hicexplorer_hicpca', 'hicexplorer_hicsummatrices', 'hicexplorer_hicadjustmatrix', 'hicexplorer_chicviewpoint', 'hicexplorer_chicqualitycontrol', 'hicexplorer_hictransform', 'lumpy_sv', 'tbvcfreport', 'bismark_pretty_report']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.06666666666666667\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.06666666666666667\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['methtools_calling', 'cshl_sort_header', 'cshl_uniq_tool', 'Filter1', 'Add_a_column1', 'Cut1', 'bedtools_intersectbed_bam', 'methtools_dmr']\n",
      "\n",
      "Test batch 1, True tools: ['methtools_plot', 'deeptools_computeMatrix']\n",
      "\n",
      "Test batch 1, Predicted top 2 tools: ['deeptools_computeMatrix', 'methtools_plot']\n",
      "\n",
      "Test batch 1, Predicted top 2 tools with weights: ['glob_report', 'sklearn_numeric_clustering']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['hisat2', 'picard_MarkDuplicates']\n",
      "\n",
      "Test batch 1, True tools: ['gatk2_realigner_target_creator', 'gatk_indel_realigner', 'picard_AddOrReplaceReadGroups', 'htseq_count', 'control_freec', '__MERGE_COLLECTION__', 'deeptools_alignmentsieve', 'bedtools_bamtobed', 'snpEff_build_gb', 'ngsutils_bam_filter', 'rseqc_geneBody_coverage', 'gatk2_base_recalibrator', 'basil', 'samtools_filter', 'delly_call', 'pilon', 'CONVERTER_bam_to_qname_sorted_bam', 'bamFilter', 'rseqc_read_duplication', 'gatk_table_recalibration', 'gatk2_indel_realigner', 'picard_CollectInsertSizeMetrics', 'deeptools_plot_fingerprint', 'picard_MergeSamFiles', 'NFS_transfer', 'gatk2_unified_genotyper', 'bedtools_genomecoveragebed', 'bcftools_mpileup', 'picard_CollectGcBiasMetrics', 'deeptools_bamCoverage', 'bedtools_intersectbed', 'samtools_stats', '__FLATTEN__', 'picard_FixMateInformation', 'macs2_callpeak', 'deeptools_bam_coverage', 'genrich', 'delly_cnv', 'gatk_count_covariates', 'gatk4_mutect2', 'gatk_realigner_target_creator', 'picard_CollectHsMetrics', 'naive_variant_caller', 'rseqc_read_distribution', 'samtools_rmdup', 'Grep1', 'picard_SortSam', 'samtool_filter2', 'samtools_view', 'featurecounts', 'igvtools_count', 'samtools_depth', 'lofreq_viterbi', 'stringtie', 'samtools_sort', 'deeptools_bam_pe_fragmentsize', 'pileometh', 'gatk2_haplotype_caller', 'strelka_germline', 'freebayes_wrapper', 'sam_merge2', 'bedtools_genomecoveragebed_histogram', 'deeptools_multi_bam_summary', 'fastqc', 'deeptools_plot_coverage', 'exomedepth', 'samtools_flagstat', 'picard_CASM', 'diffbind', 'picard_CleanSam', 'gatk2_print_reads', 'pe_histogram', 'picard_ReorderSam', 'bam_to_sam', 'macs2_predictd', 'samtools_merge', 'bamtools', 'samtools_mpileup', 'sam_pileup', 'lofreq_call', 'multiqc', 'gatk_unified_genotyper', 'deeptools_bam_compare', 'lofreq_alnqual', 'mimodd_varcall', 'samtools_idxstats', 'bamleftalign', 'rseqc_infer_experiment', 'mimodd_delcall', 'qualimap_bamqc', 'modencode_peakcalling_spp', 'freebayes', 'picard_MergeBamAlignment', 'bedtools_genomecoveragebed_bedgraph', 'bedtools_coveragebed']\n",
      "\n",
      "Test batch 1, Predicted top 95 tools: ['deeptools_alignmentsieve', 'bedtools_bamtobed', 'gatk2_base_recalibrator', 'samtools_filter', 'pilon', 'NFS_transfer', 'bcftools_mpileup', 'picard_CollectGcBiasMetrics', 'macs2_callpeak', 'deeptools_bam_coverage', 'naive_variant_caller', 'samtools_rmdup', 'samtool_filter2', 'freebayes_wrapper', 'macs2_predictd', 'samtools_merge', 'lofreq_call', 'multiqc', 'lofreq_alnqual', 'bedtools_genomecoveragebed_bedgraph', 'bedtools_coveragebed', 'gatk2_realigner_target_creator', 'htseq_count', 'picard_CollectInsertSizeMetrics', 'gatk_realigner_target_creator', 'rseqc_read_distribution', 'picard_SortSam', 'featurecounts', 'samtools_depth', 'sam_pileup', 'deeptools_bam_compare', 'rseqc_infer_experiment', 'qualimap_bamqc', 'gatk_indel_realigner', '__MERGE_COLLECTION__', 'ngsutils_bam_filter', 'gatk2_indel_realigner', 'deeptools_plot_fingerprint', 'bedtools_intersectbed', 'samtools_stats', '__FLATTEN__', 'picard_FixMateInformation', 'gatk4_mutect2', 'lofreq_viterbi', 'deeptools_bam_pe_fragmentsize', 'fastqc', 'exomedepth', 'picard_CASM', 'picard_CleanSam', 'picard_ReorderSam', 'bam_to_sam', 'mimodd_delcall', 'picard_MergeBamAlignment', 'stringtie', 'samtools_flagstat', 'diffbind', 'mimodd_varcall', 'samtools_idxstats', 'bamleftalign', 'freebayes', 'picard_MergeSamFiles', 'control_freec', 'rseqc_geneBody_coverage', 'pe_histogram', 'samtools_view', 'bedtools_genomecoveragebed_histogram', 'deeptools_plot_coverage', 'gatk_table_recalibration', 'samtools_mpileup', 'CONVERTER_bam_to_qname_sorted_bam', 'igvtools_count', 'bamtools', 'bedtools_genomecoveragebed', 'samtools_sort', 'pileometh', 'modencode_peakcalling_spp', 'deeptools_bamCoverage', 'picard_CollectHsMetrics', 'rseqc_read_duplication', 'gatk2_unified_genotyper', 'delly_cnv', 'delly_call', 'gatk2_print_reads', 'strelka_germline', 'deeptools_multi_bam_summary', 'snpEff_build_gb', 'gatk_unified_genotyper', 'basil', 'genrich', 'sam_merge2', 'gatk2_haplotype_caller', 'bamFilter', 'picard_AddOrReplaceReadGroups', 'gatk_count_covariates', 'Grep1']\n",
      "\n",
      "Test batch 1, Predicted top 95 tools with weights: ['lofreq_call', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'lofreq_viterbi', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'multiqc', 'stringtie', 'htseq_count', 'deeptools_bam_coverage', 'bedtools_intersectbed', 'macs2_callpeak', 'samtools_rmdup', 'bamFilter', 'rseqc_infer_experiment', 'freebayes', 'samtools_flagstat', 'samtools_idxstats', 'bamleftalign', 'bedtools_bamtobed', 'rseqc_read_distribution', 'ngsutils_bam_filter', 'samtools_sort', 'rseqc_geneBody_coverage', 'samtool_filter2', 'deeptools_bam_compare', 'pileometh', 'pe_histogram', 'bcftools_mpileup', 'bamtools', 'picard_SortSam', 'sam_merge2', 'samtools_merge', 'diffbind', 'picard_CollectInsertSizeMetrics', 'deeptools_multi_bam_summary', 'deeptools_plot_fingerprint', 'samtools_depth', 'picard_CASM', 'picard_MergeSamFiles', 'deeptools_alignmentsieve', 'bedtools_genomecoveragebed_bedgraph', 'bedtools_coveragebed', 'sam_pileup', 'genrich', 'deeptools_plot_coverage', 'naive_variant_caller', 'delly_call', 'mimodd_varcall', 'snpEff_build_gb', 'pilon', 'samtools_mpileup', 'rseqc_read_duplication', 'gatk4_mutect2', 'bam_to_sam', 'basil', 'picard_MergeBamAlignment', 'bedtools_genomecoveragebed_histogram', 'picard_CollectGcBiasMetrics', 'strelka_germline', 'picard_AddOrReplaceReadGroups', 'exomedepth', 'picard_FixMateInformation', 'lofreq_alnqual', 'delly_cnv', 'control_freec', 'gatk_count_covariates', 'picard_ReorderSam', 'freebayes_wrapper', 'samtools_filter', 'mimodd_delcall', 'jbrowse', 'rna_star', 'picard_MarkDuplicates', 'tp_sed_tool', 'peakachu', 'picard_DownsampleSam', 'Add_a_column1', 'bismark_deduplicate', 'rseqc_bam_stat', 'snippy_core', 'samtools_fixmate', 'bedtools_makewindowsbed', 'jvarkit_wgscoverageplotter', 'medaka_variant', 'stacks2_refmap', 'chipeakanno_annopeaks', 'bedtools_nucbed', 'rseqc_RPKM_count', 'samtools_slice_bam', 'ivar_trim', 'ococo', 'deeptools_plot_enrichment']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.16842105263157894\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.15789473684210525\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['gmx_setup', 'gmx_merge_topology_files', 'gmx_solvate', 'gmx_sim', 'md_converter']\n",
      "\n",
      "Test batch 1, True tools: ['bio3d_rmsf', 'bio3d_pca', 'vmd_hbonds', 'mmpbsa_mmgbsa', 'bio3d_rmsd', 'mdanalysis_cosine_analysis', 'bio3d_pca_visualize']\n",
      "\n",
      "Test batch 1, Predicted top 7 tools: ['bio3d_pca', 'mmpbsa_mmgbsa', 'bio3d_rmsf', 'mdanalysis_cosine_analysis', 'bio3d_pca_visualize', 'bio3d_rmsd', 'vmd_hbonds']\n",
      "\n",
      "Test batch 1, Predicted top 7 tools with weights: ['bio3d_rmsd', 'bio3d_rmsf', 'mmpbsa_mmgbsa', 'bio3d_pca', 'vmd_hbonds', 'mdanalysis_cosine_analysis', 'bio3d_pca_visualize']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.8571428571428571\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.8571428571428571\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'samtools_rmdup', 'deeptools_bam_coverage', 'bowtie2', 'macs2_callpeak', 'macs2_callpeak', 'diffbind', 'chipseeker', 'Remove beginning1', 'bedtools_intersectbed', 'Cut1', 'Extract genomic DNA 1', 'macs2_callpeak', 'macs2_callpeak']\n",
      "\n",
      "Test batch 1, True tools: ['wig_to_bigWig', 'bed2gff1', 'comp1', 'bedtools_intersectbed', 'crossmap_bed', 'tp_easyjoin_tool', 'multiqc', 'gops_concat_1', 'Cut1', 'jbrowse', 'tp_cat', 'intervene_pairwise', 'pygenomeTracks', 'Remove beginning1', 'macs2_callpeak', 'cat1', 'deeptools_bam_coverage', 'bedtools_mergebed', 'Extract genomic DNA 1', 'bedtools_slopbed', 'Filter1', 'bedtools_multiintersectbed', 'Add_a_column1', 'bedtools_makewindowsbed', 'CONVERTER_bedgraph_to_bigwig', 'histogram_rpy', 'macs2_bdgdiff', 'idr', 'tp_sort_header_tool', 'gops_join_1', 'macs2_bdgpeakcall', 'gops_intersect_1', 'sort1', 'samtools_rmdup', 'deeptools_compute_matrix', 'tp_awk_tool', 'bowtie2', 'tp_head_tool', 'intervene_upset', 'chipseeker', 'addValue', 'bedtools_subtractbed', 'samtools_view', 'tp_grep_tool', 'join1', 'diffbind', 'deeptools_plot_enrichment', 'CONVERTER_interval_to_bedstrict_0', 'chipeakanno_annopeaks', 'macs2_bdgcmp', 'bedtools_sortbed', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'tp_cut_tool', 'Convert characters1', 'bundle_collection', 'Summary_Statistics1', 'bedtools_getfastabed']\n",
      "\n",
      "Test batch 1, Predicted top 57 tools: ['tp_easyjoin_tool', 'macs2_bdgdiff', 'bundle_collection', 'histogram_rpy', 'crossmap_bed', 'tp_sort_header_tool', 'Add_a_column1', 'bedtools_makewindowsbed', 'deeptools_compute_matrix', 'intervene_pairwise', 'deeptools_plot_enrichment', 'CONVERTER_bedgraph_to_bigwig', 'idr', 'intervene_upset', 'bowtie2', 'addValue', 'bedtools_slopbed', 'tp_cut_tool', 'diffbind', 'macs2_bdgcmp', 'Convert characters1', 'Cut1', 'Summary_Statistics1', 'macs2_bdgpeakcall', 'bedtools_getfastabed', 'comp1', 'wig_to_bigWig', 'gops_intersect_1', 'bedtools_intersectbed', 'sort1', 'tp_awk_tool', 'pygenomeTracks', 'cat1', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'deeptools_bam_coverage', 'chipeakanno_annopeaks', 'bedtools_subtractbed', 'join1', 'bedtools_multiintersectbed', 'samtools_rmdup', 'Filter1', 'tp_grep_tool', 'Extract genomic DNA 1', 'bedtools_mergebed', 'jbrowse', 'gops_join_1', 'tp_head_tool', 'multiqc', 'bed2gff1', 'CONVERTER_interval_to_bedstrict_0', 'tp_cat', 'gops_concat_1', 'bedtools_sortbed', 'chipseeker', 'samtools_view', 'macs2_callpeak', 'Remove beginning1']\n",
      "\n",
      "Test batch 1, Predicted top 57 tools with weights: ['Add_a_column1', 'samtools_view', 'gops_concat_1', 'bowtie2', 'multiqc', 'deeptools_bam_coverage', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'samtools_rmdup', 'tp_cat', 'deeptools_compute_matrix', 'tp_grep_tool', 'tp_awk_tool', 'gops_intersect_1', 'tp_head_tool', 'addValue', 'pygenomeTracks', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'bedtools_sortbed', 'jbrowse', 'diffbind', 'bedtools_mergebed', 'chipseeker', 'bedtools_subtractbed', 'bedtools_multiintersectbed', 'histogram_rpy', 'macs2_bdgpeakcall', 'macs2_bdgcmp', 'bedtools_slopbed', 'idr', 'chipeakanno_annopeaks', 'bedtools_makewindowsbed', 'intervene_pairwise', 'intervene_upset', 'macs2_bdgdiff', 'deeptools_plot_enrichment', 'featurecounts', 'samtools_idxstats', 'htseq_count', 'samtools_slice_bam', 'sklearn_pca', 'deeptools_multi_bigwig_summary', 'sam_pileup', 'seacr', 'tp_uniq_tool', 'lofreq_call', 'rseqc_infer_experiment', 'bigwig_to_bedgraph', 'fgsea', 'bedtools_coveragebed', 'deeptools_bigwig_compare', 'query_tabular', 'deeptools_bam_compare', 'fastqc']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.07017543859649122\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.05263157894736842\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['hca_matrix_downloader', 'seurat_read10x']\n",
      "\n",
      "Test batch 1, True tools: ['seurat_filter_cells', 'seurat_create_seurat_object', 'seurat_normalise_data', 'seurat_find_variable_genes']\n",
      "\n",
      "Test batch 1, Predicted top 4 tools: ['seurat_filter_cells', 'seurat_normalise_data', 'seurat_create_seurat_object', 'seurat_find_variable_genes']\n",
      "\n",
      "Test batch 1, Predicted top 4 tools with weights: ['seurat_find_variable_genes', 'cp_display_data_on_image', 'cardinal_segmentations', 'iframe']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.25\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['bioext_bealign', 'tn93_filter', 'iqtree', 'hyphy_annotate']\n",
      "\n",
      "Test batch 1, True tools: ['hyphy_summary', 'hyphy_fade', 'hyphy_bgm', 'hyphy_fel', 'hyphy_cfel', 'hyphy_busted', 'hyphy_absrel', 'hyphy_meme', 'hyphy_relax', 'hyphy_prime', 'hyphy_slac']\n",
      "\n",
      "Test batch 1, Predicted top 11 tools: ['hyphy_summary', 'hyphy_fel', 'hyphy_meme', 'hyphy_slac', 'hyphy_prime', 'hyphy_cfel', 'hyphy_bgm', 'hyphy_busted', 'hyphy_absrel', 'hyphy_relax', 'hyphy_fade']\n",
      "\n",
      "Test batch 1, Predicted top 11 tools with weights: ['hyphy_slac', 'hyphy_fade', 'hyphy_meme', 'hyphy_relax', 'hyphy_summary', 'hyphy_bgm', 'hyphy_cfel', 'hyphy_fel', 'hyphy_busted', 'hyphy_prime', 'hyphy_absrel']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.8181818181818182\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.8181818181818182\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'bowtie2', 'samtools_rmdup', 'sam_merge2', 'deeptools_bam_coverage', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'bedtools_intersectbed', 'tp_sort_header_tool']\n",
      "\n",
      "Test batch 1, True tools: ['datamash_ops', 'wig_to_bigWig', 'metilene', 'tp_sorted_uniq', 'Show tail1', 'filter_tabular', 'goenrichment', 'comp1', 'gd_rank_terms', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'regex1', 'tp_multijoin_tool', 'multiqc', 'seurat', 'Cut1', 'tp_cat', 'barchart_gnuplot', 'bedtools_clusterbed', '__MERGE_COLLECTION__', 'Remove beginning1', 'column_order_header_sort', 'column_remove_by_header', 'bg_uniq', 'bedtools_mergebed', 'tp_find_and_replace', 'Filter1', 'samtools_idxstats', 'samtools_slice_bam', 'Add_a_column1', 'Show beginning1', 'histogram_rpy', 'bedtools_complementbed', 'XY_Plot_1', 'datamash_transpose', 'gops_join_1', 'Grouping1', 'gops_intersect_1', 'sort1', 'tp_awk_tool', 'tp_head_tool', 'Grep1', 'tp_tail_tool', 'table_compute', 'addValue', 'annotatemyids', 'tp_grep_tool', 'bed_to_bigBed', 'join1', 'tp_uniq_tool', 'bedtools_genomecoveragebed_bedgraph', 'fgsea', 'blockbuster', 'cshl_awk_tool', 'hgv_david', 'tp_cut_tool', 'Convert characters1', 'cummeRbund', 'Summary_Statistics1', 'wc_gnu', 'transit_gumbel']\n",
      "\n",
      "Test batch 1, Predicted top 61 tools: ['bedtools_complementbed', 'tp_uniq_tool', 'metilene', 'barchart_gnuplot', 'Grep1', 'comp1', 'multiqc', 'bedtools_genomecoveragebed_bedgraph', 'cummeRbund', 'bed_to_bigBed', 'cshl_awk_tool', 'tp_cat', 'samtools_slice_bam', 'Add_a_column1', 'gd_rank_terms', 'bedtools_clusterbed', 'tp_awk_tool', 'join1', 'tp_cut_tool', 'fgsea', 'filter_tabular', 'Cut1', 'Filter1', 'blockbuster', 'Summary_Statistics1', 'annotatemyids', 'transit_gumbel', 'addValue', 'regex1', 'tp_easyjoin_tool', 'datamash_transpose', 'wig_to_bigWig', 'seurat', 'samtools_idxstats', 'bedtools_mergebed', 'Show beginning1', 'histogram_rpy', 'tp_find_and_replace', 'Remove beginning1', 'gops_intersect_1', '__MERGE_COLLECTION__', 'tp_multijoin_tool', 'gops_join_1', 'XY_Plot_1', 'table_compute', 'tp_grep_tool', 'column_order_header_sort', 'goenrichment', 'bg_uniq', 'Show tail1', 'sort1', 'taxonomy_krona_chart', 'Grouping1', 'tp_head_tool', 'hgv_david', 'datamash_ops', 'Convert characters1', 'column_remove_by_header', 'tp_sorted_uniq', 'tp_tail_tool', 'wc_gnu']\n",
      "\n",
      "Test batch 1, Predicted top 61 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'datamash_ops', 'filter_tabular', 'multiqc', 'datamash_transpose', 'tp_easyjoin_tool', 'tp_cat', 'table_compute', 'samtools_idxstats', 'tp_grep_tool', 'tp_awk_tool', 'annotatemyids', 'taxonomy_krona_chart', 'gops_intersect_1', 'tp_head_tool', 'samtools_slice_bam', 'addValue', 'tp_tail_tool', 'fgsea', 'XY_Plot_1', 'bedtools_mergebed', 'goenrichment', 'bg_uniq', 'tp_sorted_uniq', 'regex1', 'bedtools_genomecoveragebed_bedgraph', 'metilene', 'histogram_rpy', 'column_remove_by_header', 'tp_uniq_tool', 'seurat', 'column_order_header_sort', 'cummeRbund', 'blockbuster', 'transit_gumbel', 'gd_rank_terms', 'freebayes', 'scatterplot_rpy', 'pathview', 'deseq2', 'pygenomeTracks', 'tp_replace_in_line', 'chipeakanno_annopeaks', 'bedtools_intersectbed', 'samtools_view', 'tab2fasta', 'bedtools_sortbed', 'bedtools_slopbed', 'tp_sort_header_tool', 'scanpy_plot', 'cummerbund_to_cuffdiff', 'ggplot2_heatmap2', 'bedtools_multiintersectbed', 'goseq', 'macs2_bdgpeakcall', 'regexColumn1', 'egsea', 'macs2_callpeak', 'macs2_bdgdiff', 'bedtools_map']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.11475409836065574\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.04918032786885246\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['picard_FastqToSam', 'picard_SamToFastq', 'bwa_mem', 'picard_MergeBamAlignment', 'picard_MarkDuplicates', 'picard_SortSam']\n",
      "\n",
      "Test batch 1, True tools: ['bcftools_mpileup', 'bamtools', 'samtools_mpileup', 'sam_pileup', 'samtools_stats', 'picard_AddOrReplaceReadGroups', 'htseq_count', 'racon', 'macs2_callpeak', 'bedtools_bamtobed', 'rgPicardMarkDups', 'deeptools_bam_coverage', 'sam_merge2', 'samtools_idxstats', 'fastqc', 'blockclust', 'flaimapper', 'picard_MarkDuplicates', 'samtools_flagstat', 'samtools_rmdup', 'hicexplorer_hicquickqc', 'pilon', 'picard_CASM', 'samtool_filter2', 'samtools_view', 'freebayes', 'samtools_depth', 'samtools_fastx', 'picard_CollectRnaSeqMetrics', 'hicexplorer_hicbuildmatrix', 'bam_to_sam']\n",
      "\n",
      "Test batch 1, Predicted top 31 tools: ['bedtools_bamtobed', 'pilon', 'deeptools_bam_coverage', 'sam_merge2', 'samtools_mpileup', 'samtools_depth', 'sam_pileup', 'picard_CollectRnaSeqMetrics', 'samtools_stats', 'fastqc', 'samtools_view', 'freebayes', 'racon', 'hicexplorer_hicquickqc', 'picard_AddOrReplaceReadGroups', 'samtools_flagstat', 'flaimapper', 'bcftools_mpileup', 'samtool_filter2', 'bam_to_sam', 'samtools_rmdup', 'htseq_count', 'samtools_idxstats', 'blockclust', 'bamtools', 'samtools_fastx', 'macs2_callpeak', 'hicexplorer_hicbuildmatrix', 'picard_MarkDuplicates', 'rgPicardMarkDups', 'picard_CASM']\n",
      "\n",
      "Test batch 1, Predicted top 31 tools with weights: ['samtools_view', 'samtools_stats', 'fastqc', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'macs2_callpeak', 'samtools_rmdup', 'freebayes', 'samtools_flagstat', 'samtools_idxstats', 'bedtools_bamtobed', 'samtool_filter2', 'racon', 'bcftools_mpileup', 'bamtools', 'sam_merge2', 'samtools_depth', 'picard_CASM', 'sam_pileup', 'picard_CollectRnaSeqMetrics', 'pilon', 'samtools_mpileup', 'hicexplorer_hicbuildmatrix', 'bam_to_sam', 'hicexplorer_hicquickqc', 'picard_AddOrReplaceReadGroups', 'rgPicardMarkDups', 'qualimap_bamqc', 'deeptools_multi_bam_summary', 'samtools_sort']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.06451612903225806\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.06451612903225806\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'bowtie2', 'macs2_callpeak', 'macs2_callpeak', 'samtools_rmdup', 'macs2_callpeak', 'diffbind', 'Remove beginning1', 'bedtools_intersectbed', 'bedtools_intersectbed']\n",
      "\n",
      "Test batch 1, True tools: ['datamash_ops', 'samtools_merge', 'bamtools', 'bed2gff1', 'samtools_mpileup', 'comp1', 'bedtools_intersectbed', 'tp_easyjoin_tool', 'tp_multijoin_tool', 'htseq_count', 'gops_concat_1', 'Cut1', 'deeptools_bam_compare', 'deeptools_plot_heatmap', 'tp_cat', 'gops_subtract_1', '__MERGE_COLLECTION__', 'deeptools_alignmentsieve', 'macs2_callpeak', 'cat1', 'deeptools_bam_coverage', 'sam_merge2', 'bedtools_mergebed', 'Extract genomic DNA 1', 'Filter1', 'bedtools_multiintersectbed', 'bedtools_windowbed', 'ngsutils_bam_filter', 'fastqc', 'bedtools_makewindowsbed', 'mergeCols1', '__RELABEL_FROM_FILE__', 'picard_MarkDuplicates', '__BUILD_LIST__', 'tp_sort_header_tool', 'samtools_filter', 'Grouping1', 'gops_join_1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'snpSift_extractFields', 'chipseeker', 'addValue', 'bedtools_subtractbed', 'bcftools_annotate', 'samtools_view', 'freebayes', 'join1', 'tabular_to_csv', 'samtools_fastx', 'picard_CollectInsertSizeMetrics', 'CONVERTER_interval_to_bedstrict_0', 'bedtools_sortbed', 'tp_replace_in_column', 'tp_cut_tool', 'Convert characters1', 'bedtools_getfastabed', 'wc_gnu']\n",
      "\n",
      "Test batch 1, Predicted top 59 tools: ['tp_easyjoin_tool', 'deeptools_bam_coverage', 'deeptools_plot_heatmap', 'comp1', 'bedtools_windowbed', 'tp_sort_header_tool', 'bedtools_makewindowsbed', 'bedtools_intersectbed', 'sort1', 'addValue', 'join1', 'tp_cut_tool', 'bedtools_multiintersectbed', 'cat1', 'Filter1', 'bed2gff1', 'deeptools_alignmentsieve', 'Grouping1', 'CONVERTER_interval_to_bedstrict_0', 'gops_subtract_1', 'tp_cat', 'gops_concat_1', 'Convert characters1', 'deeptools_compute_matrix', 'ngsutils_bam_filter', 'tp_replace_in_column', 'bedtools_subtractbed', 'freebayes', '__RELABEL_FROM_FILE__', 'samtools_fastx', 'picard_MarkDuplicates', 'wc_gnu', 'bedtools_getfastabed', 'chipseeker', 'deeptools_bam_compare', 'gops_join_1', 'tp_awk_tool', 'bcftools_annotate', 'tabular_to_csv', 'Extract genomic DNA 1', 'picard_CollectInsertSizeMetrics', 'fastqc', 'Cut1', 'samtools_filter', 'macs2_callpeak', 'bedtools_sortbed', 'sam_merge2', 'samtools_view', '__MERGE_COLLECTION__', 'bedtools_mergebed', 'mergeCols1', 'samtools_merge', 'snpSift_extractFields', 'tp_multijoin_tool', 'bamtools', 'samtools_mpileup', 'datamash_ops', 'htseq_count', '__BUILD_LIST__']\n",
      "\n",
      "Test batch 1, Predicted top 59 tools with weights: ['snpSift_extractFields', 'samtools_view', 'datamash_ops', 'gops_concat_1', 'gops_subtract_1', 'fastqc', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'tp_cat', 'deeptools_compute_matrix', 'bcftools_annotate', 'freebayes', 'deeptools_plot_heatmap', 'tp_awk_tool', 'tp_replace_in_column', 'ngsutils_bam_filter', 'deeptools_bam_compare', 'addValue', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'bedtools_sortbed', 'bamtools', 'sam_merge2', 'samtools_merge', 'picard_CollectInsertSizeMetrics', 'bedtools_mergebed', 'chipseeker', 'deeptools_alignmentsieve', 'bedtools_subtractbed', 'mergeCols1', 'bedtools_multiintersectbed', 'bedtools_windowbed', 'samtools_mpileup', 'bedtools_makewindowsbed', 'samtools_filter', 'Add_a_column1', 'gops_intersect_1', 'featurecounts', 'tp_sed_tool', 'deseq2', 'jbrowse', 'samtools_rmdup', 'bg_column_arrange_by_header', 'chipeakanno_annopeaks', 'collection_column_join', 'get_flanks1', 'samtools_idxstats', 'histogram_rpy', 'goseq', 'samtools_stats', 'annotatemyids', 'macs2_bdgpeakcall', 'circos_wiggle_to_scatter', 'bedtools_mergebedgraph']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.13559322033898305\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.06779661016949153\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'hisat2', 'featurecounts', 'tp_multijoin_tool', 'join1', 'tp_cut_tool', 'tp_replace_in_column', 'Filter1', 'addValue', 'tp_cat']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'rbc_mirdeep2_quantifier', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'regexColumn1', 'samtools_bedcov', 'lastz_wrapper_2', 'staramr_search', 'ggplot2_pca', 'gprofiler_gost', 'anndata_manipulate', 'miniasm', 'kraken2', 'porechop', 'fastp', 'Paste1', 'Show beginning1', 'maxquant_mqpar', 'nspdk_sparse', 'XY_Plot_1', 'rbc_mafft', 'unipept', 'bwa', 'query_tabular', 'meryl', 'sample_seqs', 'uniprot', 'tp_tail_tool', 'annotatemyids', 'join1', 'qiime_pick_closed_reference_otus', 'tp_uniq_tool', 'humann2', 'spades', 'tp_cut_tool', 'wc_gnu', 'nextclade', 'metilene', 'bedtools_intersectbed', 'je_demultiplex', '__FLATTEN__', 'fastq_groomer', '_ensembl_gtf2gene_list', 'Remove beginning1', 'racon', 'cshl_fastx_trimmer', 'DecoyDatabase', 'cd_hit_est', 'cardinal_mz_images', 'fastq_to_fasta_python', 'tp_replace_in_line', 'mergeCols1', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'regionalgam_flight_curve', 'filter_by_fasta_ids', 'trinity', 'tab2fasta', 'export_to_cluster', 'table_compute', 'regionalgam_ab_index', 'tp_grep_tool', 'PlasFlow', 'tp_replace_in_column', 'sRNAPipe', 'datamash_ops', 'trimmomatic', 'tp_sorted_uniq', 'mass_spectrometry_imaging_filtering', 'tombo_resquiggle', 'tp_multijoin_tool', 'fasta2tab', 'circos_interval_to_tile', 'fasta_concatenate0', 'clustalw', 'blast_parser', 'bcftools_consensus', 'sam_merge2', 'bg_sortmerna', 'ggplot2_heatmap2', 'rna_star', 'bwa_mem', 'Add_a_column1', 'fastqc', 'cutadapt', 'cshl_fastx_clipper', 'filtlong', 'csv_to_tabular', 'flashlfq', 'rbc_mirdeep2_mapper', 'fasta-stats', 'w4mclassfilter', 'trinity_align_and_estimate_abundance', 'ont_fast5_api_fast5_subset', 'miranda', 'pangolin', 'canu', 'umi_tools_whitelist', 'gmx_sim', 'openbabel_filter', 'cshl_fastx_collapser', 'jellyfish', 'bwameth', 'tp_sed_tool', 'medaka_consensus_pipeline', 'fastq_paired_end_deinterlacer', 'ggplot2_point', 'volcanoplot', 'iReport', 'circos', 'multiqc', 'Cut1', 'jbrowse', 'minimap2', 'column_order_header_sort', 'cat1', 'scatterplot_rpy', 'Filter1', 'prinseq', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'rdock_sort_filter', 'Count1', 'sort1', 'ctb_chemfp_mol2fps', 'ncbi_blastn_wrapper', 'bowtie2', 'metaphlan2', 'qiime_align_seqs', 'snp_sites', 'addValue', 'cardinal_filtering', 'circos_interval_to_text', 'cast', 'split_file_to_collection', 'prokka', 'bedtools_sortbed', 'nanoplot', 'cshl_fastq_to_fasta', 'fasta_compute_length', 'hisat2', 'drep_dereplicate', 'trim_galore']\n",
      "\n",
      "Test batch 1, Predicted top 145 tools: ['tp_easyjoin_tool', 'get_flanks1', 'staramr_search', 'humann2', 'metilene', 'cshl_fastx_trimmer', 'fastq_to_fasta_python', 'mergeCols1', 'Grouping1', 'trinity', 'datamash_ops', 'tp_sorted_uniq', 'mass_spectrometry_imaging_filtering', 'tp_multijoin_tool', 'fasta2tab', 'blast_parser', 'jellyfish', 'bwameth', 'multiqc', '__RELABEL_FROM_FILE__', 'metaphlan2', 'nanoplot', 'hisat2', 'bg_column_arrange_by_header', 'regexColumn1', 'meryl', 'tp_tail_tool', 'fastq_groomer', 'DecoyDatabase', 'tp_sort_header_tool', 'filter_by_fasta_ids', 'table_compute', 'PlasFlow', 'bcftools_consensus', 'bwa_mem', 'Add_a_column1', 'filtlong', 'w4mclassfilter', 'miranda', 'medaka_consensus_pipeline', 'volcanoplot', 'prinseq', 'ncbi_blastn_wrapper', 'cast', 'fasta_compute_length', 'miniasm', 'query_tabular', 'bedtools_intersectbed', 'racon', 'export_to_cluster', 'tp_replace_in_column', 'trimmomatic', 'ggplot2_heatmap2', 'fastqc', 'umi_tools_whitelist', 'fastq_paired_end_deinterlacer', 'jbrowse', 'column_order_header_sort', 'sort1', 'qiime_align_seqs', 'snp_sites', 'addValue', 'prokka', 'cshl_fastq_to_fasta', 'trim_galore', 'lastz_wrapper_2', 'ggplot2_pca', 'Paste1', 'join1', 'sRNAPipe', 'cutadapt', 'cshl_fastx_clipper', 'csv_to_tabular', 'tp_sed_tool', 'minimap2', 'cat1', 'scatterplot_rpy', 'Filter1', 'bedtools_sortbed', 'rbc_mirdeep2_quantifier', 'samtools_bedcov', 'anndata_manipulate', 'XY_Plot_1', 'uniprot', 'spades', 'tp_uniq_tool', '__TAG_FROM_FILE__', 'rdock_sort_filter', 'ctb_chemfp_mol2fps', 'rbc_mafft', 'bwa', 'fasta_concatenate0', 'pangolin', 'maxquant_mqpar', 'clustalw', 'ont_fast5_api_fast5_subset', 'ggplot2_point', 'bowtie2', 'nspdk_sparse', 'tp_cut_tool', 'Show beginning1', 'bg_sortmerna', 'regionalgam_flight_curve', 'regionalgam_ab_index', 'tp_grep_tool', 'circos_interval_to_text', 'cd_hit_est', 'cshl_fastx_collapser', 'nextclade', 'rna_star', 'qiime_pick_closed_reference_otus', 'tp_replace_in_line', 'gmx_sim', 'iReport', 'Cut1', 'annotatemyids', '_ensembl_gtf2gene_list', 'porechop', 'cardinal_filtering', 'openbabel_filter', 'circos_interval_to_tile', 'tab2fasta', '__FLATTEN__', 'rbc_mirdeep2_mapper', 'sample_seqs', 'Remove beginning1', 'trinity_align_and_estimate_abundance', 'split_file_to_collection', 'drep_dereplicate', 'circos', 'je_demultiplex', 'tp_split_on_column', 'umi_tools_extract', 'canu', 'unipept', 'fastp', 'wc_gnu', 'fasta-stats', 'Count1', 'kraken2', 'tombo_resquiggle', 'gprofiler_gost', 'flashlfq', 'cardinal_mz_images', 'sam_merge2']\n",
      "\n",
      "Test batch 1, Predicted top 145 tools with weights: ['Add_a_column1', 'tp_replace_in_line', 'fastp', 'bwa_mem', 'datamash_ops', 'bcftools_consensus', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'trimmomatic', 'hisat2', 'prokka', 'multiqc', 'minimap2', 'kraken2', 'fastq_to_fasta_python', 'ncbi_blastn_wrapper', 'trim_galore', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'fastq_groomer', 'bedtools_intersectbed', 'porechop', 'fasta2tab', 'spades', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'PlasFlow', 'volcanoplot', 'tp_grep_tool', 'fasta-stats', 'nextclade', 'circos', 'annotatemyids', 'metaphlan2', 'nanoplot', 'tp_replace_in_column', 'filter_by_fasta_ids', 'bg_sortmerna', 'split_file_to_collection', 'tp_sed_tool', 'cshl_fastq_to_fasta', 'rbc_mirdeep2_mapper', 'addValue', 'gmx_sim', 'trinity', 'pangolin', 'racon', 'clustalw', 'filtlong', 'bedtools_sortbed', 'cshl_fastx_trimmer', 'bwameth', 'staramr_search', 'meryl', 'tp_tail_tool', 'bwa', 'ggplot2_point', 'anndata_manipulate', 'humann2', 'trinity_align_and_estimate_abundance', 'sam_merge2', 'jbrowse', 'regexColumn1', 'fastq_paired_end_deinterlacer', 'fasta_compute_length', 'XY_Plot_1', 'cd_hit_est', 'query_tabular', 'rbc_mafft', 'cshl_fastx_collapser', 'tp_sorted_uniq', 'cardinal_filtering', 'umi_tools_extract', 'miniasm', 'rbc_mirdeep2_quantifier', 'je_demultiplex', 'medaka_consensus_pipeline', 'sample_seqs', 'metilene', 'mergeCols1', 'tp_split_on_column', 'nspdk_sparse', 'snp_sites', 'cshl_fastx_clipper', '_ensembl_gtf2gene_list', 'ggplot2_pca', 'qiime_align_seqs', 'canu', 'scatterplot_rpy', 'fasta_concatenate0', 'cardinal_mz_images', 'drep_dereplicate', 'sRNAPipe', 'prinseq', 'gprofiler_gost', 'get_flanks1', 'tp_uniq_tool', 'bg_column_arrange_by_header', 'jellyfish', 'column_order_header_sort', 'maxquant_mqpar', 'uniprot', 'DecoyDatabase', 'ctb_chemfp_mol2fps', 'samtools_bedcov', 'flashlfq', 'blast_parser', 'openbabel_filter', 'rdock_sort_filter', 'umi_tools_whitelist', 'circos_interval_to_text', 'qiime_pick_closed_reference_otus', 'cast', 'mass_spectrometry_imaging_filtering', 'regionalgam_ab_index', 'regionalgam_flight_curve', 'tp_find_and_replace', 'tp_cat', 'sklearn_ensemble', 'seqtk_subseq', 'fastq_paired_end_interlacer', 'stringtie', 'fastq_to_tabular', 'salmon', 'megahit', 'plotly_ml_performance_plots', 'edger', 'seq_filter_by_id', 'quast', 'kallisto_quant', 'limma_voom', 'goenrichment', 'bowtie_wrapper', 'targetfinder', 'ncbi_acc_download', 'minia', 'egsea', 'bismark_bowtie2', 'fasterq_dump', 'fgsea', 'featurecounts', 'rbc_mirdeep2']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.09655172413793103\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.08275862068965517\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'hisat2', 'featurecounts', 'deseq2', 'Filter1', 'tp_replace_in_column', 'Cut1', 'tp_cat']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'rbc_mirdeep2_quantifier', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'regexColumn1', 'samtools_bedcov', 'lastz_wrapper_2', 'staramr_search', 'ggplot2_pca', 'gprofiler_gost', 'anndata_manipulate', 'miniasm', 'kraken2', 'porechop', 'fastp', 'Paste1', 'Show beginning1', 'maxquant_mqpar', 'nspdk_sparse', 'XY_Plot_1', 'rbc_mafft', 'unipept', 'bwa', 'query_tabular', 'meryl', 'sample_seqs', 'uniprot', 'tp_tail_tool', 'annotatemyids', 'join1', 'qiime_pick_closed_reference_otus', 'tp_uniq_tool', 'humann2', 'spades', 'tp_cut_tool', 'wc_gnu', 'nextclade', 'metilene', 'bedtools_intersectbed', 'je_demultiplex', '__FLATTEN__', 'fastq_groomer', '_ensembl_gtf2gene_list', 'Remove beginning1', 'racon', 'cshl_fastx_trimmer', 'DecoyDatabase', 'cd_hit_est', 'cardinal_mz_images', 'fastq_to_fasta_python', 'tp_replace_in_line', 'mergeCols1', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'regionalgam_flight_curve', 'filter_by_fasta_ids', 'trinity', 'tab2fasta', 'export_to_cluster', 'table_compute', 'regionalgam_ab_index', 'tp_grep_tool', 'PlasFlow', 'tp_replace_in_column', 'sRNAPipe', 'datamash_ops', 'trimmomatic', 'tp_sorted_uniq', 'mass_spectrometry_imaging_filtering', 'tombo_resquiggle', 'tp_multijoin_tool', 'fasta2tab', 'circos_interval_to_tile', 'fasta_concatenate0', 'clustalw', 'blast_parser', 'bcftools_consensus', 'sam_merge2', 'bg_sortmerna', 'ggplot2_heatmap2', 'rna_star', 'bwa_mem', 'Add_a_column1', 'fastqc', 'cutadapt', 'cshl_fastx_clipper', 'filtlong', 'csv_to_tabular', 'flashlfq', 'rbc_mirdeep2_mapper', 'fasta-stats', 'w4mclassfilter', 'trinity_align_and_estimate_abundance', 'ont_fast5_api_fast5_subset', 'miranda', 'pangolin', 'canu', 'umi_tools_whitelist', 'gmx_sim', 'openbabel_filter', 'cshl_fastx_collapser', 'jellyfish', 'bwameth', 'tp_sed_tool', 'medaka_consensus_pipeline', 'fastq_paired_end_deinterlacer', 'ggplot2_point', 'volcanoplot', 'iReport', 'circos', 'multiqc', 'Cut1', 'jbrowse', 'minimap2', 'column_order_header_sort', 'cat1', 'scatterplot_rpy', 'Filter1', 'prinseq', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'rdock_sort_filter', 'Count1', 'sort1', 'ctb_chemfp_mol2fps', 'ncbi_blastn_wrapper', 'bowtie2', 'metaphlan2', 'qiime_align_seqs', 'snp_sites', 'addValue', 'cardinal_filtering', 'circos_interval_to_text', 'cast', 'split_file_to_collection', 'prokka', 'bedtools_sortbed', 'nanoplot', 'cshl_fastq_to_fasta', 'fasta_compute_length', 'hisat2', 'drep_dereplicate', 'trim_galore']\n",
      "\n",
      "Test batch 1, Predicted top 145 tools: ['tp_easyjoin_tool', 'humann2', 'metilene', 'cshl_fastx_trimmer', 'fastq_to_fasta_python', 'mergeCols1', 'Grouping1', 'trinity', 'datamash_ops', 'tp_sorted_uniq', 'mass_spectrometry_imaging_filtering', 'fasta2tab', 'blast_parser', 'bwameth', 'multiqc', '__RELABEL_FROM_FILE__', 'metaphlan2', 'nanoplot', 'hisat2', 'bg_column_arrange_by_header', 'regexColumn1', 'meryl', 'tp_tail_tool', 'fastq_groomer', 'DecoyDatabase', 'filter_by_fasta_ids', 'table_compute', 'bwa_mem', 'Add_a_column1', 'filtlong', 'w4mclassfilter', 'medaka_consensus_pipeline', 'volcanoplot', 'prinseq', 'ncbi_blastn_wrapper', 'miniasm', 'maxquant_mqpar', 'query_tabular', 'bedtools_intersectbed', 'racon', 'export_to_cluster', 'tp_replace_in_column', 'trimmomatic', 'ggplot2_heatmap2', 'fastqc', 'umi_tools_whitelist', 'fastq_paired_end_deinterlacer', 'jbrowse', 'column_order_header_sort', 'sort1', 'bowtie2', 'qiime_align_seqs', 'snp_sites', 'addValue', 'prokka', 'cshl_fastq_to_fasta', 'trim_galore', 'lastz_wrapper_2', 'ggplot2_pca', 'Paste1', 'join1', 'cutadapt', 'cshl_fastx_clipper', 'csv_to_tabular', 'tp_sed_tool', 'minimap2', 'cat1', 'scatterplot_rpy', 'Filter1', 'bedtools_sortbed', 'rbc_mirdeep2_quantifier', 'samtools_bedcov', 'anndata_manipulate', 'XY_Plot_1', 'uniprot', 'spades', 'tp_uniq_tool', 'tp_multijoin_tool', 'jellyfish', 'PlasFlow', 'cast', 'fasta_compute_length', 'regionalgam_flight_curve', 'regionalgam_ab_index', 'ont_fast5_api_fast5_subset', 'circos_interval_to_text', 'nspdk_sparse', 'cd_hit_est', 'sRNAPipe', 'get_flanks1', 'staramr_search', 'circos_interval_to_tile', '__TAG_FROM_FILE__', 'rdock_sort_filter', 'ctb_chemfp_mol2fps', 'rbc_mafft', 'bwa', 'je_demultiplex', 'tp_replace_in_line', 'fasta_concatenate0', 'bcftools_consensus', 'miranda', 'clustalw', 'tp_cut_tool', 'porechop', 'qiime_pick_closed_reference_otus', 'pangolin', '_ensembl_gtf2gene_list', 'nextclade', 'cardinal_filtering', 'tp_sort_header_tool', 'bg_sortmerna', 'Cut1', 'wc_gnu', '__FLATTEN__', 'cshl_fastx_collapser', 'gmx_sim', 'sample_seqs', 'openbabel_filter', 'rna_star', 'rbc_mirdeep2_mapper', 'ggplot2_point', 'Remove beginning1', 'umi_tools_extract', 'tp_grep_tool', 'canu', 'tab2fasta', 'iReport', 'tp_split_on_column', 'drep_dereplicate', 'circos', 'fastp', 'annotatemyids', 'Show beginning1', 'trinity_align_and_estimate_abundance', 'kraken2', 'split_file_to_collection', 'flashlfq', 'gprofiler_gost', 'fasta-stats', 'unipept', 'Count1', 'tombo_resquiggle', 'cardinal_mz_images', 'sam_merge2']\n",
      "\n",
      "Test batch 1, Predicted top 145 tools with weights: ['Add_a_column1', 'tp_replace_in_line', 'fastp', 'bwa_mem', 'datamash_ops', 'bcftools_consensus', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'trimmomatic', 'hisat2', 'prokka', 'multiqc', 'minimap2', 'kraken2', 'fastq_to_fasta_python', 'ncbi_blastn_wrapper', 'trim_galore', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'fastq_groomer', 'bedtools_intersectbed', 'porechop', 'fasta2tab', 'spades', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'PlasFlow', 'volcanoplot', 'tp_grep_tool', 'fasta-stats', 'nextclade', 'circos', 'annotatemyids', 'metaphlan2', 'nanoplot', 'tp_replace_in_column', 'filter_by_fasta_ids', 'bg_sortmerna', 'split_file_to_collection', 'tp_sed_tool', 'cshl_fastq_to_fasta', 'rbc_mirdeep2_mapper', 'addValue', 'gmx_sim', 'trinity', 'pangolin', 'racon', 'clustalw', 'filtlong', 'bedtools_sortbed', 'cshl_fastx_trimmer', 'bwameth', 'staramr_search', 'meryl', 'tp_tail_tool', 'bwa', 'ggplot2_point', 'anndata_manipulate', 'humann2', 'trinity_align_and_estimate_abundance', 'sam_merge2', 'jbrowse', 'regexColumn1', 'fastq_paired_end_deinterlacer', 'fasta_compute_length', 'XY_Plot_1', 'cd_hit_est', 'query_tabular', 'rbc_mafft', 'cshl_fastx_collapser', 'tp_sorted_uniq', 'cardinal_filtering', 'umi_tools_extract', 'miniasm', 'rbc_mirdeep2_quantifier', 'je_demultiplex', 'medaka_consensus_pipeline', 'sample_seqs', 'metilene', 'mergeCols1', 'tp_split_on_column', 'nspdk_sparse', 'snp_sites', 'cshl_fastx_clipper', '_ensembl_gtf2gene_list', 'ggplot2_pca', 'qiime_align_seqs', 'canu', 'scatterplot_rpy', 'fasta_concatenate0', 'cardinal_mz_images', 'drep_dereplicate', 'sRNAPipe', 'prinseq', 'gprofiler_gost', 'get_flanks1', 'tp_uniq_tool', 'bg_column_arrange_by_header', 'jellyfish', 'column_order_header_sort', 'maxquant_mqpar', 'uniprot', 'DecoyDatabase', 'ctb_chemfp_mol2fps', 'samtools_bedcov', 'flashlfq', 'blast_parser', 'openbabel_filter', 'rdock_sort_filter', 'umi_tools_whitelist', 'circos_interval_to_text', 'qiime_pick_closed_reference_otus', 'cast', 'mass_spectrometry_imaging_filtering', 'regionalgam_ab_index', 'regionalgam_flight_curve', 'tp_find_and_replace', 'tp_cat', 'sklearn_ensemble', 'fastq_paired_end_interlacer', 'seqtk_subseq', 'fastq_to_tabular', 'megahit', 'seq_filter_by_id', 'quast', 'salmon', 'bowtie_wrapper', 'stringtie', 'kallisto_quant', 'plotly_ml_performance_plots', 'targetfinder', 'edger', 'bismark_bowtie2', 'limma_voom', 'goenrichment', 'minia', 'metaeuk_easy_predict', 'mothur_make_contigs', 'cshl_fastq_quality_filter', 'tbvcfreport', 'mothur_align_seqs', 'rbc_mirdeep2']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.09655172413793103\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.08275862068965517\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['mothur_merge_files', 'mothur_count_seqs', 'mothur_summary_seqs', 'mothur_screen_seqs', 'mothur_filter_seqs', 'mothur_unique_seqs', 'mothur_pre_cluster']\n",
      "\n",
      "Test batch 1, True tools: ['mothur_remove_seqs', 'mothur_classify_seqs', 'mothur_classify_otu', 'mothur_remove_lineage', 'mothur_cluster_split', 'mothur_make_shared', 'mothur_summary_seqs', 'mothur_chimera_uchime', 'mothur_chimera_vsearch']\n",
      "\n",
      "Test batch 1, Predicted top 9 tools: ['mothur_classify_seqs', 'mothur_chimera_vsearch', 'mothur_cluster_split', 'mothur_remove_lineage', 'mothur_classify_otu', 'mothur_chimera_uchime', 'mothur_make_shared', 'mothur_remove_seqs', 'mothur_summary_seqs']\n",
      "\n",
      "Test batch 1, Predicted top 9 tools with weights: ['mothur_summary_seqs', 'mothur_cluster_split', 'mothur_classify_seqs', 'mothur_make_shared', 'mothur_remove_lineage', 'mothur_chimera_vsearch', 'mothur_remove_seqs', 'mothur_classify_otu', 'mothur_remove_rare']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3333333333333333\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.2222222222222222\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['paired_fastq_filtering']\n",
      "\n",
      "Test batch 1, True tools: []\n",
      "\n",
      "Test batch 1, Predicted top 0 tools: []\n",
      "\n",
      "Test batch 1, Predicted top 0 tools with weights: []\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'bowtie2', 'macs2_callpeak', 'bedtools_intersectbed', 'sort1', 'tp_head_tool', 'Extract genomic DNA 1', 'macs2_callpeak', 'diffbind', 'Remove beginning1', 'bedtools_subtractbed', 'bedtools_intersectbed', 'Cut1', 'Filter1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'pampa_communitymetrics', 'tp_easyjoin_tool', 'join_files_on_column_fuzzy', 'htseq_count', 'regexColumn1', 'snpSift_filter', 'hgv_snpFreq', 'collection_column_join', 'seq_select_by_id', 'gprofiler_gost', 'bg_uniq', 'venn_list', 'tp_find_and_replace', 'Show beginning1', 'XY_Plot_1', 'changeGffFeatures', 'random_lines1', 'unipept', 'cshl_sort_header', 'ChangeCase', 'uniprot', 'aggregate_scores_in_intervals2', 'tp_tail_tool', 'bedtools_subtractbed', 'collapse_dataset', 'annotatemyids', 'join1', 'deg_annotate', 'tp_cut_tool', 'wc_gnu', 'bedtools_intersectbed', 'cardinal_combine', 'repeat_counter', 'transpose', 'dexseq_annotate', 'Remove beginning1', 'cardinal_mz_images', 'column_remove_by_header', 'bedtools_mergebed', 'tp_replace_in_line', 'deseq2', 'mergeCols1', 'vegan_rarefaction', 'Interval_Maf_Merged_Fasta2', 'tp_sort_header_tool', 'Grouping1', 'pandas_rolling_window', 'naive_variant_caller', 'gops_intersect_1', 'samtools_rmdup', 'Fetch Taxonomic Ranks', 'ggplot2_heatmap', 'trinity', 'tab2fasta', 'Grep1', 'SnpEff-cds-report', 'samtool_filter2', 'table_compute', 'samtools_view', 'tabular_to_csv', 'table_annovar', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'cooler_csort_pairix', '__FILTER_EMPTY_DATASETS__', 'gops_cluster_1', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'filter_tabular', 'goenrichment', 'snpEff', 'cardinal_spectra_plots', 'gdal_ogr2ogr', 'tp_multijoin_tool', 'gops_concat_1', 'tp_cat', 'charts', 'bcftools_consensus', 'vcfbedintersect', 'sam_merge2', 'ggplot2_heatmap2', 'Add_a_column1', 'stoceps_glm', 'datamash_transpose', 'stoceps_trend_indic', 'gops_coverage_1', 'chipseeker', 'snpSift_rmInfo', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'diapysef', 'Show tail1', 'trimmer', 'comp1', 'ggplot2_point', 'volcanoplot', 'multiqc', 'tbl2gff3', 'Cut1', 'CONVERTER_vcf_bgzip_to_tabix_0', 'jbrowse', 'gops_subtract_1', 'minimap2', 'cat1', 'Extract genomic DNA 1', 'tab2csv_R', 'vegan_diversity', 'pathview', 'pampa_glmsp', 'histogram_rpy', '__BUILD_LIST__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'addValue', 'bcftools_norm', 'dexseq_count', 'cardinal_filtering', 'split_file_to_collection', '__EXTRACT_DATASET__', 'samtools_fastx', 'bedtools_sortbed', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'bedtools_maskfastabed', 'fastq_filter', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 135 tools: ['tp_easyjoin_tool', 'join_files_on_column_fuzzy', 'Show beginning1', 'mergeCols1', 'Grouping1', 'samtools_rmdup', 'Grep1', 'samtool_filter2', 'table_annovar', 'datamash_ops', 'tp_multijoin_tool', 'datamash_transpose', 'snpSift_rmInfo', 'bg_column_arrange_by_header', 'tp_find_and_replace', 'changeGffFeatures', 'tp_tail_tool', 'cardinal_combine', 'cardinal_mz_images', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'featurecounts', 'snpEff', 'charts', 'bcftools_consensus', 'Add_a_column1', 'chipseeker', 'random_lines1', 'ChangeCase', 'bedtools_intersectbed', 'Fetch Taxonomic Ranks', 'vcfbedintersect', 'stoceps_trend_indic', 'jbrowse', 'gops_join_1', 'sort1', 'addValue', 'bcftools_norm', 'gprofiler_gost', 'bedtools_subtractbed', 'join1', 'tp_cut_tool', 'bedtools_mergebed', 'gops_concat_1', 'tp_sed_tool', 'Cut1', 'cat1', 'bedtools_sortbed', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'Summary_Statistics1', 'cshl_word_list_grep', 'collection_column_join', 'XY_Plot_1', 'unipept', 'uniprot', 'annotatemyids', 'naive_variant_caller', 'trinity', '__FILTER_EMPTY_DATASETS__', 'bedtools_getfastabed', 'tp_sorted_uniq', 'sam_merge2', 'gops_coverage_1', 'dexseq_count', 'cardinal_filtering', 'regexColumn1', 'SnpEff-cds-report', 'hgv_david', 'samtools_fastx', 'tp_replace_in_column', 'ggplot2_heatmap2', 'venn_list', 'column_remove_by_header', 'tabular_to_csv', 'fastq_filter', 'wc_gnu', 'cshl_sort_header', 'repeat_counter', 'dexseq_annotate', 'deseq2', 'pandas_rolling_window', 'gops_cluster_1', 'goenrichment', 'stoceps_glm', 'gops_subtract_1', 'htseq_count', 'hgv_snpFreq', 'transpose', 'tp_cat', 'tab2csv_R', 'cooler_csort_pairix', 'diapysef', 'CONVERTER_vcf_bgzip_to_tabix_0', 'split_file_to_collection', 'Interval_Maf_Merged_Fasta2', 'gdal_ogr2ogr', 'Convert characters1', 'pathview', 'snpSift_filter', 'seq_select_by_id', 'bg_uniq', 'collapse_dataset', 'Show tail1', 'comp1', 'histogram_rpy', 'bedtools_maskfastabed', 'tbl2gff3', 'tp_awk_tool', 'ncbi_acc_download', 'samtools_view', 'minimap2', 'aggregate_scores_in_intervals2', 'cardinal_spectra_plots', 'volcanoplot', 'deg_annotate', 'seq_filter_by_id', 'filter_tabular', 'tp_head_tool', 'ggplot2_point', 'ggplot2_heatmap', 'vegan_diversity', 'vegan_rarefaction', 'pampa_glmsp', 'Remove beginning1', 'pampa_communitymetrics', 'multiqc', 'Count1', 'deeptools_compute_matrix', '__EXTRACT_DATASET__', 'table_compute', 'trimmer', 'tab2fasta', 'Extract genomic DNA 1', '__BUILD_LIST__']\n",
      "\n",
      "Test batch 1, Predicted top 135 tools with weights: ['Add_a_column1', 'snpSift_filter', 'tp_find_and_replace', 'tp_replace_in_line', 'samtools_view', 'datamash_ops', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'filter_tabular', 'multiqc', 'minimap2', 'htseq_count', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'samtools_rmdup', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'snpEff', 'volcanoplot', 'tp_awk_tool', 'annotatemyids', 'fastq_filter', 'deg_annotate', 'tp_replace_in_column', 'gops_intersect_1', 'collection_column_join', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'samtool_filter2', 'addValue', 'bedtools_maskfastabed', 'trinity', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'sam_merge2', 'jbrowse', 'regexColumn1', 'XY_Plot_1', 'bedtools_mergebed', 'goenrichment', 'bcftools_norm', 'bg_uniq', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'tbl2gff3', 'dexseq_count', 'chipseeker', 'bedtools_subtractbed', 'ggplot2_heatmap', 'mergeCols1', 'histogram_rpy', 'naive_variant_caller', 'venn_list', 'cardinal_mz_images', 'join_files_on_column_fuzzy', 'cardinal_spectra_plots', 'gprofiler_gost', 'vcfbedintersect', 'column_remove_by_header', 'bg_column_arrange_by_header', 'cardinal_combine', 'uniprot', 'vegan_diversity', 'SnpEff-cds-report', 'gops_coverage_1', 'diapysef', 'Fetch Taxonomic Ranks', 'vegan_rarefaction', 'pampa_glmsp', 'stoceps_trend_indic', 'stoceps_glm', 'gops_cluster_1', 'fastqc', 'gtftobed12', 'samtools_slice_bam', 'tp_split_on_column', 'filter_by_fasta_ids', 'addName', 'cutadapt', 'picard_MarkDuplicates', 'rna_star', 'plotly_ml_performance_plots', 'cardinal_classification', 'xarray_select', 'goseq', 'picard_FilterSamReads', 'eggnog_mapper', 'circos_wiggle_to_stacked', 'chipeakanno_annopeaks', 'cummeRbund', 'fgsea', 'taxonomy_krona_chart', 'metaspades', 'snpfreqplot', 'blockbuster', 'bedtools_windowbed', 'gops_merge_1', 'gemini_query', 'egsea', 'rcas', 'qualimap_bamqc', 'samtools_depth', 'msstats', 'ncbi_blastdbcmd_wrapper', 'get_sequences', 'snpSift_extractFields', 'macs2_bdgdiff', 'macs2_callpeak', 'regex_replace', 'kraken2', 'heinz_bum', 'ctb_multi_obgrep', 'regex1', 'heinz_scoring', 'bedtools_makewindowsbed', 'transit_gumbel', 'tp_uniq_tool', 'get_feature_info', 'deeptools_bam_coverage']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.1925925925925926\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.1111111111111111\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['mothur_make_group', 'mothur_count_seqs', 'mothur_screen_seqs', 'mothur_pre_cluster', 'mothur_remove_seqs', 'mothur_remove_lineage', 'mothur_get_groups', 'mothur_dist_seqs', 'mothur_cluster', 'mothur_make_shared', 'mothur_sub_sample']\n",
      "\n",
      "Test batch 1, True tools: ['mothur_unifrac_unweighted', 'mothur_dist_shared', 'mothur_unifrac_weighted', 'mothur_rarefaction_single', 'mothur_corr_axes', 'mothur_venn', 'mycrobiota_subsample_add_discarded_samples', 'mothur_count_groups', 'mothur_summary_tax', 'mothur_summary_single', 'mothur_make_biom', 'mothur_get_communitytype', 'mothur_metastats', 'collapse_dataset']\n",
      "\n",
      "Test batch 1, Predicted top 14 tools: ['mothur_count_groups', 'mothur_summary_tax', 'mycrobiota_subsample_add_discarded_samples', 'mothur_unifrac_weighted', 'collapse_dataset', 'mothur_unifrac_unweighted', 'mothur_dist_shared', 'mothur_venn', 'mothur_summary_single', 'mothur_rarefaction_single', 'mothur_corr_axes', 'mothur_metastats', 'mothur_get_communitytype', 'mothur_make_biom']\n",
      "\n",
      "Test batch 1, Predicted top 14 tools with weights: ['collapse_dataset', 'mothur_make_biom', 'mothur_count_groups', 'mothur_venn', 'mothur_summary_single', 'mothur_dist_shared', 'mothur_rarefaction_single', 'mothur_corr_axes', 'mothur_get_communitytype', 'mothur_unifrac_weighted', 'mothur_metastats', 'mothur_sub_sample', 'regex1', 'mothur_heatmap_bin']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.14285714285714285\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.14285714285714285\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'bismark_bowtie2']\n",
      "\n",
      "Test batch 1, True tools: ['pileometh', 'bismark_pretty_report', 'bismark_methylation_extractor', 'perpi_methylextract', 'samtools_rmdup', 'methtools_calling', 'multiqc', 'bismark_deduplicate', 'qualimap_bamqc', 'deeptools_plot_coverage', 'jbrowse', 'samtool_filter2']\n",
      "\n",
      "Test batch 1, Predicted top 12 tools: ['samtools_rmdup', 'samtool_filter2', 'multiqc', 'qualimap_bamqc', 'bismark_pretty_report', 'pileometh', 'jbrowse', 'deeptools_plot_coverage', 'perpi_methylextract', 'bismark_deduplicate', 'bismark_methylation_extractor', 'methtools_calling']\n",
      "\n",
      "Test batch 1, Predicted top 12 tools with weights: ['qualimap_bamqc', 'multiqc', 'samtools_rmdup', 'samtool_filter2', 'pileometh', 'jbrowse', 'deeptools_plot_coverage', 'bismark_deduplicate', 'bismark_pretty_report', 'tbvcfreport', 'fastqc', 'macs2_callpeak']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'samtools_rmdup', 'deeptools_bam_coverage', 'bowtie2', 'macs2_callpeak', 'Cut1', 'Filter1', 'Remove beginning1', 'bedtools_intersectbed', 'sort1']\n",
      "\n",
      "Test batch 1, True tools: ['datamash_ops', 'Show tail1', 'tp_sorted_uniq', 'ggplot2_point', 'comp1', 'pampa_communitymetrics', 'tp_easyjoin_tool', 'pampa_presabs', 'tp_split_on_column', 'tp_multijoin_tool', 'Cut1', 'transit_hmm', 'tp_cat', 'gops_subtract_1', 'cat1', 'Extract genomic DNA 1', 'tp_find_and_replace', 'ggplot2_heatmap2', 'Filter1', 'kraken2', 'Add_a_column1', 'Show beginning1', 'datamash_transpose', 'tp_sort_header_tool', 'Grouping1', 'gops_intersect_1', 'tp_awk_tool', 'tp_head_tool', 'tab2fasta', 'Grep1', 'tp_tail_tool', 'addValue', 'annotatemyids', 'join1', 'deg_annotate', 'blockbuster', 'cshl_awk_tool', 'salsa', 'tp_cut_tool', 'create_opson_pipeline_summary', 'cummeRbund', 'cshl_uniq_tool', 'bedtools_getfastabed']\n",
      "\n",
      "Test batch 1, Predicted top 43 tools: ['tp_easyjoin_tool', 'Grouping1', 'Grep1', 'create_opson_pipeline_summary', 'comp1', 'tp_cat', 'Add_a_column1', 'tp_awk_tool', 'join1', 'tp_cut_tool', 'transit_hmm', 'Filter1', 'tp_multijoin_tool', 'tp_sort_header_tool', 'cshl_uniq_tool', 'datamash_ops', 'cshl_awk_tool', 'addValue', 'pampa_presabs', 'salsa', 'cummeRbund', 'gops_intersect_1', 'blockbuster', 'tp_split_on_column', 'kraken2', 'cat1', 'tp_sorted_uniq', 'tp_head_tool', 'Extract genomic DNA 1', 'Cut1', 'pampa_communitymetrics', 'datamash_transpose', 'gops_subtract_1', 'ggplot2_point', 'Show beginning1', 'deg_annotate', 'tab2fasta', 'ggplot2_heatmap2', 'bedtools_getfastabed', 'Show tail1', 'annotatemyids', 'tp_tail_tool', 'tp_find_and_replace']\n",
      "\n",
      "Test batch 1, Predicted top 43 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'datamash_ops', 'gops_subtract_1', 'kraken2', 'datamash_transpose', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'tp_cat', 'tab2fasta', 'ggplot2_heatmap2', 'tp_awk_tool', 'annotatemyids', 'deg_annotate', 'gops_intersect_1', 'tp_head_tool', 'addValue', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'tp_tail_tool', 'ggplot2_point', 'tp_sorted_uniq', 'tp_split_on_column', 'salsa', 'cummeRbund', 'blockbuster', 'transit_hmm', 'regexColumn1', 'tp_replace_in_column', 'jbrowse', 'table_compute', 'volcanoplot', 'bg_uniq', 'xarray_select', 'addName', 'gops_concat_1', 'pathview', 'bedtools_intersectbed', 'featurecounts', 'XY_Plot_1', 'bedtools_sortbed', 'bowtie2', 'taxonomy_krona_chart']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.09302325581395349\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.06976744186046512\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'bowtie2', 'macs2_callpeak', 'bedtools_intersectbed', 'chipseeker', 'samtools_rmdup', 'sam_merge2', 'deeptools_bam_coverage', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'Cut1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'join_files_on_column_fuzzy', 'regexColumn1', 'collection_column_join', 'plotly_ml_performance_plots', 'bg_uniq', 'venn_list', 'scanpy_read_10x', 'tp_find_and_replace', 'bedtools_slopbed', 'Paste1', 'Show beginning1', 'ggplot_histogram', 'krona-text', 'stacks_denovomap', 'XY_Plot_1', 'random_lines1', 'unipept', 'query_tabular', 'ChangeCase', 'tp_tail_tool', 'collapse_dataset', 'annotatemyids', 'join1', 'goseq', 'heinz_scoring', 'tp_uniq_tool', 'export2graphlan', 'tp_cut_tool', 'seqtk_subseq', 'wc_gnu', 'bedtools_genomecoveragebed', 'wig_to_bigWig', 'clusterprofiler_bitr', 'bedtools_intersectbed', 'crossmap_bed', '__FLATTEN__', 'barchart_gnuplot', 'gtftobed12', 'Remove beginning1', 'macs2_callpeak', 'picard_FilterSamReads', 'pygenomeTracks', 'get_feature_info', 'bedtools_mergebed', 'ggplot_point', 'tp_replace_in_line', 'stringtie_merge', '__FILTER_FROM_FILE__', 'deseq2', 'mergeCols1', 'biotranslator', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'ggplot2_heatmap', 'tab2fasta', 'Grep1', 'msstats', 'table_compute', 'tp_grep_tool', 'egsea', 'add_line_to_file', 'fgsea', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'annotateMyIDs', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'goenrichment', 'regex1', 'tp_multijoin_tool', 'CONVERTER_interval_to_bed_0', 'gops_concat_1', 'bedtools_intersectbed_bam', 'tp_cat', 'charts', 'bcftools_consensus', 'fasterq_dump', 'ggplot2_heatmap2', 'rna_star', 'combine_metaphlan2_humann2', 'stacks_populations', 'Add_a_column1', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'w4mclassfilter', 'get_sequences', 'Kraken2Tax', 'cshl_find_and_replace', 'ont_fast5_api_fast5_subset', 'interactions', 'gops_coverage_1', 'heinz_bum', 'addName', 'mass_spectrometry_imaging_ion_images', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'Show tail1', 'trimmer', 'comp1', 'secure_hash_message_digest', 'volcanoplot', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'circos', 'gops_subtract_1', 'cat1', 'scatterplot_rpy', 'regex_replace', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'histogram_rpy', 'diff', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'interactive_tool_simtext_app', 'sklearn_ensemble', 'addValue', 'dexseq_count', 'cardinal_filtering', 'MassCalculator', 'split_file_to_collection', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'CONVERTER_bed_to_gff_0', 'Krona', 'bedtools_sortbed', 'param_value_from_file', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools: ['get_flanks1', 'join_files_on_column_fuzzy', 'ggplot_histogram', 'tp_uniq_tool', 'barchart_gnuplot', 'deseq2', 'mergeCols1', 'Grouping1', 'ggplot2_heatmap', 'annotateMyIDs', 'bedtools_getfastabed', 'tp_sorted_uniq', 'goenrichment', 'rna_star', 'dropletutils_read_10x', 'datamash_transpose', 'openbabel_remIons', 'comp1', 'regex_replace', '__TAG_FROM_FILE__', 'sklearn_ensemble', 'dexseq_count', 'regexColumn1', 'scanpy_read_10x', 'krona-text', 'tp_tail_tool', 'heinz_scoring', 'wig_to_bigWig', 'clusterprofiler_bitr', 'crossmap_bed', 'ggplot_point', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'tab2fasta', 'msstats', 'table_compute', 'featurecounts', 'tp_cat', 'bcftools_consensus', 'Add_a_column1', 'w4mclassfilter', 'cshl_find_and_replace', 'addName', 'mass_spectrometry_imaging_ion_images', 'hgv_david', 'secure_hash_message_digest', 'MassCalculator', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'param_value_from_file', 'plotly_ml_performance_plots', 'stacks_denovomap', 'random_lines1', 'query_tabular', 'ChangeCase', 'goseq', 'seqtk_subseq', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', '__FLATTEN__', 'gtftobed12', 'picard_FilterSamReads', 'stringtie_merge', 'biotranslator', 'tp_grep_tool', 'egsea', 'tp_replace_in_column', 'seq_filter_by_id', 'fasterq_dump', 'combine_metaphlan2_humann2', 'stacks_populations', 'get_sequences', 'Kraken2Tax', 'ont_fast5_api_fast5_subset', 'heinz_bum', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'gops_join_1', 'sort1', 'tp_awk_tool', 'addValue', 'ncbi_acc_download', 'venn_list', 'Paste1', 'join1', 'tp_cut_tool', 'get_feature_info', 'add_line_to_file', 'fgsea', 'regex1', 'interactions', 'Convert characters1', 'circos', 'cat1', 'scatterplot_rpy', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'diff', 'tp_head_tool', 'interactive_tool_simtext_app', 'CONVERTER_bed_to_gff_0', 'Krona', 'cshl_word_list_grep', 'collection_column_join', 'bg_uniq', 'XY_Plot_1', 'collapse_dataset', 'annotatemyids', 'export2graphlan', 'tp_easyjoin_tool', 'Show beginning1', 'checkFormat', 'gops_coverage_1', 'tp_find_and_replace', 'Remove beginning1', 'bedtools_slopbed', 'pygenomeTracks', 'bedtools_intersectbed_bam', 'trimmer', 'bedtools_sortbed', 'Summary_Statistics1', 'unipept', 'taxonomy_krona_chart', 'macs2_callpeak', 'tp_multijoin_tool', 'Show tail1', 'volcanoplot', 'ggplot2_heatmap2', 'split_file_to_collection', '__FILTER_FROM_FILE__', 'Grep1', 'Count1', 'wc_gnu', 'datamash_ops', 'umi_tools_extract', 'tp_sed_tool', '__RELABEL_FROM_FILE__', 'deeptools_compute_matrix', 'bedtools_mergebed', 'histogram_rpy', 'gops_subtract_1', 'bg_column_arrange_by_header', 'cardinal_filtering', 'CONVERTER_interval_to_bed_0', 'gops_concat_1', 'charts', 'tp_split_on_column']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'tp_replace_in_line', 'datamash_ops', 'bedtools_genomecoveragebed', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'rna_star', 'fasterq_dump', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'volcanoplot', 'tp_grep_tool', 'circos', 'tp_awk_tool', 'annotatemyids', 'taxonomy_krona_chart', 'tp_replace_in_column', 'filter_by_fasta_ids', 'goseq', 'gops_intersect_1', 'collection_column_join', 'limma_voom', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'Kraken2Tax', 'addValue', 'combine_metaphlan2_humann2', 'pygenomeTracks', 'addName', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'regexColumn1', 'seqtk_subseq', 'fgsea', 'XY_Plot_1', 'bedtools_mergebed', 'export2graphlan', 'goenrichment', 'query_tabular', 'bg_uniq', 'krona-text', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'gtftobed12', 'umi_tools_extract', 'stringtie_merge', 'dexseq_count', 'regex1', 'msstats', 'picard_FilterSamReads', 'ggplot2_heatmap', 'mergeCols1', 'tp_split_on_column', 'histogram_rpy', 'ncbi_blastdbcmd_wrapper', 'scatterplot_rpy', 'venn_list', 'egsea', 'join_files_on_column_fuzzy', 'add_line_to_file', 'bedtools_slopbed', 'dropletutils_read_10x', 'plotly_ml_performance_plots', 'get_flanks1', 'tp_uniq_tool', 'regex_replace', 'bg_column_arrange_by_header', 'sklearn_ensemble', 'stacks_denovomap', 'diff', 'gops_coverage_1', 'get_feature_info', 'checkFormat', 'get_sequences', 'heinz_scoring', 'heinz_bum', 'MassCalculator', 'macs2_bdgdiff', 'gops_merge_1', 'uniprot', 'edger', 'chipeakanno_annopeaks', 'xarray_select', 'deg_annotate', 'metilene', 'tabular_to_fastq', 'bedtools_subtractbed', 'filter_tabular', 'htseq_count', 'raceid_filtnormconf', 'chipseeker', 'tables_arithmetic_operations', 'Heatmap', 'cardinal_spectra_plots', 'column_remove_by_header', 'cummeRbund', 'fastq_groomer', 'macs2_bdgpeakcall', 'multiqc', 'tbl2gff3', 'humann2', 'plink', 'macs2_bdgcmp', 'scater_create_qcmetric_ready_sce', 'column_order_header_sort', 'datamash_reverse', 'vegan_diversity', 'bedtools_multiintersectbed', 'idr', 'intervene_upset', 'intervene_pairwise', 'gops_cluster_1', 'hicexplorer_hicfindrestrictionsites', 'melt', 'sam_merge2', 'graphprot_predict_profile', 'rseqc_infer_experiment', 'samtools_idxstats', 'stoceps_trend_indic', 'vegan_rarefaction', 'get_pdb', 'ctb_chemfp_mol2fps', 'cor2', 'Fetch Taxonomic Ranks', 'transit_gumbel', 'sklearn_feature_selection', 'climate_stripes', 'sklearn_svm_classifier', 'flair_correct', 'subtract_query1', 'gd_rank_terms', 'transit_hmm']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3092105263157895\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.21710526315789475\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['fastq_groomer', 'unicycler', 'bwa_mem', 'samtools_sort', 'samtools_mpileup']\n",
      "\n",
      "Test batch 1, True tools: ['varscan_somatic', 'varscan', 'OverhangTrimming_V3', 'snpEff', 'bcftools_view', 'CONVERTER_interval_to_bgzip_0', 'varscan_copynumber', 'Filter1', 'medaka_variant', 'pileup_interval', 'Cut1', 'vcf_filter', 'bcftools_call', 'varscan_mpileup', 'pileup_parser']\n",
      "\n",
      "Test batch 1, Predicted top 15 tools: ['pileup_parser', 'varscan', 'pileup_interval', 'snpEff', 'varscan_copynumber', 'varscan_somatic', 'OverhangTrimming_V3', 'bcftools_call', 'CONVERTER_interval_to_bgzip_0', 'medaka_variant', 'vcf_filter', 'varscan_mpileup', 'bcftools_view', 'Cut1', 'Filter1']\n",
      "\n",
      "Test batch 1, Predicted top 15 tools with weights: ['snpEff', 'medaka_variant', 'pileup_parser', 'bcftools_call', 'varscan', 'bcftools_view', 'varscan_copynumber', 'vcf_filter', 'bcftools_consensus', 'snpeff_sars_cov_2', 'addName', 'tp_cat', 'plotdexseq', 'tp_find_and_replace', 'query_tabular']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.06666666666666667\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'cat1', 'trinity', 'trinity_filter_low_expr_transcripts', 'trinity_align_and_estimate_abundance', 'trinity_abundance_estimates_to_matrix']\n",
      "\n",
      "Test batch 1, True tools: ['trinity_analyze_diff_expr', 'trinity_filter_low_expr_transcripts', 'trinity_contig_exn50_statistic', 'trinity_samples_qccheck', 'trinity_run_de_analysis']\n",
      "\n",
      "Test batch 1, Predicted top 5 tools: ['trinity_run_de_analysis', 'trinity_analyze_diff_expr', 'trinity_filter_low_expr_transcripts', 'trinity_contig_exn50_statistic', 'trinity_samples_qccheck']\n",
      "\n",
      "Test batch 1, Predicted top 5 tools with weights: ['trinity_run_de_analysis', 'trinity_contig_exn50_statistic', 'trinity_samples_qccheck', 'trinity_analyze_diff_expr', 'trinity_filter_low_expr_transcripts']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['retrieve_scxa', 'scanpy_read_10x', 'scanpy_filter_cells', 'scanpy_filter_genes', 'scanpy_normalise_data', 'scanpy_regress_variable', 'scanpy_scale_data', 'scanpy_find_variable_genes', 'scanpy_run_pca', 'scanpy_compute_graph', 'scanpy_find_cluster', 'scanpy_run_umap', 'run_sccaf', 'sccaf_asses']\n",
      "\n",
      "Test batch 1, True tools: ['sccaf_asses_merger']\n",
      "\n",
      "Test batch 1, Predicted top 1 tools: ['sccaf_asses']\n",
      "\n",
      "Test batch 1, Predicted top 1 tools with weights: ['sccaf_asses']\n",
      "\n",
      "Test batch 1, Precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['fastq_filter', 'fastq_trimmer', 'fastq_to_fasta_python', 'cshl_fastx_trimmer', 'fasta2tab', 'collection_column_join', 'mergeCols1', 'Cut1', 'tab2fasta']\n",
      "\n",
      "Test batch 1, True tools: ['cshl_fastx_barcode_splitter', 'megablast_wrapper', 'mothur_get_lineage', 'circos', 'meme_chip', 'cshl_fasta_formatter', 'clustalw', 'tp_cat', 'drep_dereplicate', 'cd_hit_est', 'meme_meme', 'EMBOSS: transeq101', 'DecoyDatabase', 'cat1', 'circos_gc_skew', 'bg_protein_properties', 'ncbi_blastp_wrapper', 'fasta_merge_files_and_filter_unique_sequences', 'graphmap_align', 'cshl_fastx_reverse_complement', 'cutadapt', 'abricate', 'fasta-stats', 'rbc_mafft', 'ncbi_blastn_wrapper', 'infernal_cmsearch', 'transdecoder', 'bowtie2', 'rgweblogo3', 'tp_tail_tool', 'CONVERTER_fasta_to_tabular', 'collapse_dataset', 'canu', 'fasta_filter_by_length', 'picard_NormalizeFasta', 'blast_reciprocal_best_hits', 'cshl_fastx_collapser', 'seq_filter_by_id', 'fasta_compute_length', 'mlocarna']\n",
      "\n",
      "Test batch 1, Predicted top 40 tools: ['megablast_wrapper', 'infernal_cmsearch', 'abricate', 'graphmap_align', 'tp_cat', 'meme_meme', 'bg_protein_properties', 'circos_gc_skew', 'picard_NormalizeFasta', 'seq_filter_by_id', 'canu', 'cshl_fasta_formatter', 'cshl_fastx_barcode_splitter', 'mlocarna', 'cat1', 'fasta_filter_by_length', 'cshl_fastx_reverse_complement', 'collapse_dataset', 'mothur_get_lineage', 'EMBOSS: transeq101', 'bowtie2', 'drep_dereplicate', 'ncbi_blastp_wrapper', 'DecoyDatabase', 'fasta_merge_files_and_filter_unique_sequences', 'cshl_fastx_collapser', 'rbc_mafft', 'fasta_compute_length', 'clustalw', 'cd_hit_est', 'transdecoder', 'blast_reciprocal_best_hits', 'CONVERTER_fasta_to_tabular', 'meme_chip', 'tp_tail_tool', 'rgweblogo3', 'cutadapt', 'ncbi_blastn_wrapper', 'circos', 'fasta-stats']\n",
      "\n",
      "Test batch 1, Predicted top 40 tools with weights: ['cutadapt', 'bowtie2', 'abricate', 'ncbi_blastn_wrapper', 'tp_cat', 'ncbi_blastp_wrapper', 'collapse_dataset', 'fasta-stats', 'circos', 'fasta_merge_files_and_filter_unique_sequences', 'infernal_cmsearch', 'blast_reciprocal_best_hits', 'seq_filter_by_id', 'fasta_filter_by_length', 'clustalw', 'tp_tail_tool', 'fasta_compute_length', 'cd_hit_est', 'rbc_mafft', 'cshl_fastx_collapser', 'cshl_fastx_barcode_splitter', 'meme_meme', 'cshl_fastx_reverse_complement', 'picard_NormalizeFasta', 'meme_chip', 'canu', 'drep_dereplicate', 'circos_gc_skew', 'cshl_fasta_formatter', 'graphmap_align', 'DecoyDatabase', 'megablast_wrapper', 'mlocarna', 'rgweblogo3', 'transdecoder', 'fasta2tab', 'maker', 'iqtree', 'viennarna_rnafold', 'bedtools_intersectbed']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.05\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.05\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['tp_cat', 'ncbi_blastn_wrapper', 'tp_split_on_column', 'Grouping1', 'table_compute', 'Filter1', '__FILTER_EMPTY_DATASETS__', 'collapse_dataset', 'filter_tabular', 'tp_easyjoin_tool', 'tab2fasta', 'EMBOSS: transeq101']\n",
      "\n",
      "Test batch 1, True tools: ['EMBOSS: pepinfo69', 'rbc_mafft', 'interproscan', 'busco', 'ncbi_blastp_wrapper', 'tp_multijoin_tool', 'filter_by_fasta_ids', 'fasta2tab', 'ncbi_makeblastdb', 'EMBOSS: pepwindow73', 't_coffee', 'EMBOSS: pepstats71']\n",
      "\n",
      "Test batch 1, Predicted top 12 tools: ['interproscan', 'busco', 'tp_multijoin_tool', 'ncbi_blastp_wrapper', 'EMBOSS: pepstats71', 'fasta2tab', 'EMBOSS: pepinfo69', 'EMBOSS: pepwindow73', 'filter_by_fasta_ids', 't_coffee', 'ncbi_makeblastdb', 'rbc_mafft']\n",
      "\n",
      "Test batch 1, Predicted top 12 tools with weights: ['fasta2tab', 'ncbi_blastp_wrapper', 'busco', 'filter_by_fasta_ids', 'ncbi_makeblastdb', 'interproscan', 'rbc_mafft', 't_coffee', 'EMBOSS: pepinfo69', 'EMBOSS: pepwindow73', 'bamtools', 'hhsearch']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.16666666666666666\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.16666666666666666\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['ctb_pubchem_download_as_smiles', 'openbabel_compound_convert']\n",
      "\n",
      "Test batch 1, True tools: ['padel', 'prepare_box', 'prepare_ligands_for_docking', 'ctb_frankenstein_ligand', 'tp_cat', 'openbabel_svg_depiction', 'cat1', 'gmx_setup', 'tp_find_and_replace', 'docking', 'openbabel_genProp', '__TAG_FROM_FILE__', 'rdock_rbdock', 'ctb_simsearch', 'xchem_transfs_scoring', 'collection_element_identifiers', 'ctb_chemfp_mol2fps', 'rxdock_rbcavity', 'ambertools_acpype', 'rxdock_rbdock', 'tp_grep_tool', 'split_file_to_collection', 'ambertools_antechamber', 'chembl', 'ctb_rdkit_descriptors', 'tp_cut_tool', 'rdock_rbcavity']\n",
      "\n",
      "Test batch 1, Predicted top 27 tools: ['ctb_chemfp_mol2fps', 'xchem_transfs_scoring', 'tp_cut_tool', 'openbabel_genProp', 'ctb_rdkit_descriptors', 'collection_element_identifiers', 'tp_find_and_replace', 'rxdock_rbdock', 'docking', 'prepare_ligands_for_docking', 'rdock_rbdock', 'padel', 'prepare_box', 'rdock_rbcavity', 'split_file_to_collection', '__TAG_FROM_FILE__', 'ambertools_acpype', 'ctb_frankenstein_ligand', 'chembl', 'cat1', 'rxdock_rbcavity', 'openbabel_svg_depiction', 'tp_cat', 'gmx_setup', 'tp_grep_tool', 'ambertools_antechamber', 'ctb_simsearch']\n",
      "\n",
      "Test batch 1, Predicted top 27 tools with weights: ['tp_find_and_replace', 'tp_cat', 'tp_grep_tool', 'split_file_to_collection', 'docking', 'ambertools_acpype', 'chembl', 'gmx_setup', 'prepare_box', 'openbabel_svg_depiction', 'collection_element_identifiers', 'rxdock_rbdock', 'ambertools_antechamber', 'prepare_ligands_for_docking', 'ctb_rdkit_descriptors', 'rdock_rbcavity', 'xchem_transfs_scoring', 'ctb_chemfp_mol2fps', 'rxdock_rbcavity', 'openbabel_genProp', 'padel', 'ctb_simsearch', 'collection_column_join', 'addName', 'query_tabular', 'tximport', 'tp_easyjoin_tool']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.14814814814814814\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.14814814814814814\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'macs2_callpeak', 'samtools_rmdup', 'sam_merge2', 'deeptools_bam_coverage', 'bowtie2', 'macs2_callpeak', 'macs2_callpeak', 'diffbind', 'Filter1', 'Remove beginning1', 'bedtools_subtractbed', 'bedtools_intersectbed', 'bedtools_intersectbed', 'Cut1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'join_files_on_column_fuzzy', 'regexColumn1', 'collection_column_join', 'plotly_ml_performance_plots', 'bg_uniq', 'venn_list', 'scanpy_read_10x', 'tp_find_and_replace', 'bedtools_slopbed', 'Paste1', 'Show beginning1', 'ggplot_histogram', 'krona-text', 'stacks_denovomap', 'XY_Plot_1', 'random_lines1', 'unipept', 'query_tabular', 'ChangeCase', 'tp_tail_tool', 'collapse_dataset', 'annotatemyids', 'join1', 'goseq', 'heinz_scoring', 'tp_uniq_tool', 'export2graphlan', 'tp_cut_tool', 'seqtk_subseq', 'wc_gnu', 'bedtools_genomecoveragebed', 'wig_to_bigWig', 'clusterprofiler_bitr', 'bedtools_intersectbed', 'crossmap_bed', '__FLATTEN__', 'barchart_gnuplot', 'gtftobed12', 'Remove beginning1', 'macs2_callpeak', 'picard_FilterSamReads', 'pygenomeTracks', 'get_feature_info', 'bedtools_mergebed', 'ggplot_point', 'tp_replace_in_line', 'stringtie_merge', '__FILTER_FROM_FILE__', 'deseq2', 'mergeCols1', 'biotranslator', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'ggplot2_heatmap', 'tab2fasta', 'Grep1', 'msstats', 'table_compute', 'tp_grep_tool', 'egsea', 'add_line_to_file', 'fgsea', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'annotateMyIDs', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'goenrichment', 'regex1', 'tp_multijoin_tool', 'CONVERTER_interval_to_bed_0', 'gops_concat_1', 'bedtools_intersectbed_bam', 'tp_cat', 'charts', 'bcftools_consensus', 'fasterq_dump', 'ggplot2_heatmap2', 'rna_star', 'combine_metaphlan2_humann2', 'stacks_populations', 'Add_a_column1', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'w4mclassfilter', 'get_sequences', 'Kraken2Tax', 'cshl_find_and_replace', 'ont_fast5_api_fast5_subset', 'interactions', 'gops_coverage_1', 'heinz_bum', 'addName', 'mass_spectrometry_imaging_ion_images', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'Show tail1', 'trimmer', 'comp1', 'secure_hash_message_digest', 'volcanoplot', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'circos', 'gops_subtract_1', 'cat1', 'scatterplot_rpy', 'regex_replace', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'histogram_rpy', 'diff', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'interactive_tool_simtext_app', 'sklearn_ensemble', 'addValue', 'dexseq_count', 'cardinal_filtering', 'MassCalculator', 'split_file_to_collection', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'CONVERTER_bed_to_gff_0', 'Krona', 'bedtools_sortbed', 'param_value_from_file', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools: ['tp_easyjoin_tool', 'get_flanks1', 'join_files_on_column_fuzzy', 'ggplot_histogram', 'tp_uniq_tool', 'barchart_gnuplot', 'deseq2', 'mergeCols1', 'Grouping1', 'ggplot2_heatmap', 'annotateMyIDs', 'bedtools_getfastabed', 'tp_sorted_uniq', 'goenrichment', 'rna_star', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'gops_coverage_1', 'comp1', 'regex_replace', '__TAG_FROM_FILE__', 'sklearn_ensemble', 'dexseq_count', 'regexColumn1', 'scanpy_read_10x', 'krona-text', 'heinz_scoring', 'wig_to_bigWig', 'clusterprofiler_bitr', 'crossmap_bed', 'Remove beginning1', 'ggplot_point', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'tab2fasta', 'msstats', 'table_compute', 'featurecounts', 'tp_cat', 'bcftools_consensus', 'Add_a_column1', 'w4mclassfilter', 'cshl_find_and_replace', 'addName', 'mass_spectrometry_imaging_ion_images', 'secure_hash_message_digest', 'MassCalculator', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'param_value_from_file', 'plotly_ml_performance_plots', 'stacks_denovomap', 'random_lines1', 'query_tabular', 'ChangeCase', 'goseq', 'seqtk_subseq', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', '__FLATTEN__', 'gtftobed12', 'picard_FilterSamReads', 'stringtie_merge', 'biotranslator', 'tp_grep_tool', 'egsea', 'tp_replace_in_column', 'seq_filter_by_id', 'fasterq_dump', 'combine_metaphlan2_humann2', 'stacks_populations', 'get_sequences', 'Kraken2Tax', 'ont_fast5_api_fast5_subset', 'heinz_bum', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'sort1', 'tp_awk_tool', 'addValue', 'ncbi_acc_download', 'bedtools_slopbed', 'Paste1', 'join1', 'tp_cut_tool', 'get_feature_info', 'add_line_to_file', 'fgsea', 'regex1', 'interactions', 'circos', 'cat1', 'scatterplot_rpy', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'diff', 'interactive_tool_simtext_app', 'CONVERTER_bed_to_gff_0', 'Krona', 'bedtools_sortbed', 'cshl_word_list_grep', 'collection_column_join', 'bg_uniq', 'XY_Plot_1', 'unipept', 'collapse_dataset', 'annotatemyids', 'export2graphlan', 'taxonomy_krona_chart', 'Show beginning1', 'tp_find_and_replace', 'tp_tail_tool', 'hgv_david', 'gops_join_1', 'split_file_to_collection', 'venn_list', 'pygenomeTracks', 'trimmer', 'tp_head_tool', 'macs2_callpeak', 'Grep1', 'tp_multijoin_tool', 'Show tail1', 'volcanoplot', '__FILTER_FROM_FILE__', 'bedtools_intersectbed_bam', 'Convert characters1', 'Summary_Statistics1', '__RELABEL_FROM_FILE__', 'umi_tools_extract', 'tp_sed_tool', 'Count1', 'wc_gnu', 'datamash_ops', 'deeptools_compute_matrix', 'ggplot2_heatmap2', 'histogram_rpy', 'bg_column_arrange_by_header', 'bedtools_mergebed', 'cardinal_filtering', 'gops_subtract_1', 'CONVERTER_interval_to_bed_0', 'gops_concat_1', 'charts', 'tp_split_on_column']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'tp_replace_in_line', 'datamash_ops', 'bedtools_genomecoveragebed', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'rna_star', 'fasterq_dump', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'volcanoplot', 'tp_grep_tool', 'circos', 'tp_awk_tool', 'annotatemyids', 'taxonomy_krona_chart', 'tp_replace_in_column', 'filter_by_fasta_ids', 'goseq', 'gops_intersect_1', 'collection_column_join', 'limma_voom', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'Kraken2Tax', 'addValue', 'combine_metaphlan2_humann2', 'pygenomeTracks', 'addName', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'regexColumn1', 'seqtk_subseq', 'fgsea', 'XY_Plot_1', 'bedtools_mergebed', 'export2graphlan', 'goenrichment', 'query_tabular', 'bg_uniq', 'krona-text', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'gtftobed12', 'umi_tools_extract', 'stringtie_merge', 'dexseq_count', 'regex1', 'msstats', 'picard_FilterSamReads', 'ggplot2_heatmap', 'mergeCols1', 'tp_split_on_column', 'histogram_rpy', 'ncbi_blastdbcmd_wrapper', 'scatterplot_rpy', 'venn_list', 'egsea', 'join_files_on_column_fuzzy', 'add_line_to_file', 'bedtools_slopbed', 'dropletutils_read_10x', 'plotly_ml_performance_plots', 'get_flanks1', 'tp_uniq_tool', 'regex_replace', 'bg_column_arrange_by_header', 'sklearn_ensemble', 'stacks_denovomap', 'diff', 'gops_coverage_1', 'get_feature_info', 'checkFormat', 'get_sequences', 'heinz_scoring', 'heinz_bum', 'MassCalculator', 'gops_merge_1', 'macs2_bdgdiff', 'uniprot', 'edger', 'chipeakanno_annopeaks', 'xarray_select', 'deg_annotate', 'metilene', 'tabular_to_fastq', 'filter_tabular', 'bedtools_subtractbed', 'htseq_count', 'raceid_filtnormconf', 'Heatmap', 'tables_arithmetic_operations', 'chipseeker', 'cardinal_spectra_plots', 'column_remove_by_header', 'fastq_groomer', 'cummeRbund', 'macs2_bdgpeakcall', 'tbl2gff3', 'humann2', 'plink', 'scater_create_qcmetric_ready_sce', 'multiqc', 'macs2_bdgcmp', 'vegan_diversity', 'column_order_header_sort', 'datamash_reverse', 'gops_cluster_1', 'bedtools_multiintersectbed', 'graphprot_predict_profile', 'idr', 'melt', 'get_pdb', 'intervene_upset', 'intervene_pairwise', 'rseqc_infer_experiment', 'vegan_rarefaction', 'hicexplorer_hicfindrestrictionsites', 'stoceps_trend_indic', 'samtools_idxstats', 'ctb_chemfp_mol2fps', 'cor2', 'Fetch Taxonomic Ranks', 'sam_merge2', 'sklearn_feature_selection', 'sklearn_svm_classifier', 'transit_gumbel', 'climate_stripes', 'flair_correct', 'kc-align', 'subtract_query1', 'ggplot2_pca']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3092105263157895\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.21710526315789475\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['fastp', 'hisat2']\n",
      "\n",
      "Test batch 1, True tools: ['samtools_merge', 'varscan_somatic', 'samtools_mpileup', 'rseqc_bam_stat', 'ggplot2_point', 'PicardASMetrics', 'samtools_stats', 'multiqc', 'picard_AddOrReplaceReadGroups', 'htseq_count', 'stringtie', 'jbrowse', 'tp_cat', 'lastz_wrapper_2', 'pipmir', 'samtools_sort', 'macs2_callpeak', 'ococo', 'deeptools_bam_coverage', 'sam_merge2', 'deeptools_correct_gc_bias', 'custom_pro_db', 'rna_star', 'deeptools_multi_bam_summary', 'samtools_idxstats', 'samtools_slice_bam', 'ngsutils_bam_filter', 'fastqc', 'rseqc_geneBody_coverage', 'rseqc_infer_experiment', 'rseqc_geneBody_coverage2', 'cufflinks', 'picard_MarkDuplicates', 'samtools_filter', 'cuffdiff', 'cuffquant', 'samtools_flagstat', 'rseqc_read_distribution', 'samtools_rmdup', 'rseqc_junction_annotation', 'deeptools_compute_gc_bias', 'bedtools_bamtofastq', 'hicexplorer_hicquickqc', 'qualimap_bamqc', 'picard_MarkDuplicatesWithMateCigar', 'pilon', 'trinity', 'ivar_consensus', 'bamFilter', 'picard_SortSam', 'picard_CASM', 'samtool_filter2', 'rseqc_read_duplication', 'samtools_view', 'dexseq_count', 'freebayes', 'umi_tools_dedup', 'picard_QualityScoreDistribution', 'megahit', 'qualimap_rnaseq', 'picard_CollectInsertSizeMetrics', 'picard_CollectRnaSeqMetrics', 'hicexplorer_hicbuildmatrix', 'featurecounts', 'picard_MergeSamFiles', 'umi_tools_count', 'spades', 'pe_histogram', 'rseqc_bam2wig', 'rseqc_inner_distance', 'express', 'bam_to_bigwig', 'je_markdupes', 'bedtools_genomecoveragebed', 'bioext_bam2msa']\n",
      "\n",
      "Test batch 1, Predicted top 75 tools: ['custom_pro_db', 'pilon', 'qualimap_rnaseq', 'umi_tools_count', 'macs2_callpeak', 'deeptools_bam_coverage', 'trinity', 'samtool_filter2', 'sam_merge2', 'deeptools_correct_gc_bias', 'rna_star', 'rseqc_geneBody_coverage2', 'samtools_merge', 'multiqc', 'dexseq_count', 'htseq_count', 'rseqc_geneBody_coverage', 'cuffdiff', 'deeptools_compute_gc_bias', 'picard_CollectInsertSizeMetrics', 'featurecounts', 'rseqc_bam2wig', 'je_markdupes', 'samtools_slice_bam', 'rseqc_inner_distance', 'rseqc_infer_experiment', 'qualimap_bamqc', 'rseqc_bam_stat', 'ngsutils_bam_filter', 'picard_MarkDuplicates', 'bamFilter', 'samtools_stats', 'cufflinks', 'megahit', 'bam_to_bigwig', 'fastqc', 'ggplot2_point', 'lastz_wrapper_2', 'stringtie', 'deeptools_multi_bam_summary', 'cuffquant', 'samtools_flagstat', 'samtools_idxstats', 'freebayes', 'rseqc_junction_annotation', 'rseqc_read_duplication', 'picard_MergeSamFiles', 'spades', 'samtools_filter', 'tp_cat', 'PicardASMetrics', 'ococo', 'rseqc_read_distribution', 'picard_SortSam', 'bioext_bam2msa', 'picard_CollectRnaSeqMetrics', 'express', 'picard_QualityScoreDistribution', 'samtools_sort', 'picard_CASM', 'hicexplorer_hicquickqc', 'samtools_view', 'pe_histogram', 'samtools_mpileup', 'bedtools_bamtofastq', 'jbrowse', 'pipmir', 'picard_MarkDuplicatesWithMateCigar', 'samtools_rmdup', 'ivar_consensus', 'bedtools_genomecoveragebed', 'picard_AddOrReplaceReadGroups', 'varscan_somatic', 'hicexplorer_hicbuildmatrix', 'umi_tools_dedup']\n",
      "\n",
      "Test batch 1, Predicted top 75 tools with weights: ['samtools_view', 'samtools_stats', 'qualimap_bamqc', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'rna_star', 'multiqc', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'macs2_callpeak', 'samtools_rmdup', 'tp_cat', 'spades', 'ivar_consensus', 'bamFilter', 'rseqc_infer_experiment', 'freebayes', 'samtools_flagstat', 'samtools_idxstats', 'rseqc_read_distribution', 'ngsutils_bam_filter', 'bedtools_bamtofastq', 'samtools_sort', 'cufflinks', 'rseqc_geneBody_coverage', 'samtool_filter2', 'samtools_slice_bam', 'pe_histogram', 'trinity', 'bioext_bam2msa', 'ggplot2_point', 'rseqc_bam_stat', 'picard_SortSam', 'sam_merge2', 'jbrowse', 'megahit', 'samtools_merge', 'picard_CollectInsertSizeMetrics', 'deeptools_multi_bam_summary', 'picard_CASM', 'cuffdiff', 'dexseq_count', 'picard_MergeSamFiles', 'PicardASMetrics', 'custom_pro_db', 'umi_tools_dedup', 'qualimap_rnaseq', 'picard_CollectRnaSeqMetrics', 'rseqc_bam2wig', 'pilon', 'samtools_mpileup', 'rseqc_read_duplication', 'ococo', 'rseqc_geneBody_coverage2', 'hicexplorer_hicbuildmatrix', 'rseqc_junction_annotation', 'hicexplorer_hicquickqc', 'picard_AddOrReplaceReadGroups', 'picard_QualityScoreDistribution', 'rseqc_inner_distance', 'pipmir', 'express', 'samtools_filter', 'deeptools_plot_fingerprint', 'peakachu', 'fastq_to_fasta_python', 'Add_a_column1', 'bam_to_sam', 'bedtools_bamtobed', 'fastq_groomer', 'samtools_bedcov', 'samtools_calmd']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.09333333333333334\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.09333333333333334\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['unicycler', 'regex1', 'tp_cat', 'minimap2']\n",
      "\n",
      "Test batch 1, True tools: ['rseqc_bam_stat', 'medaka_consensus', 'htseq_count', 'bedtools_bamtobed', 'ococo', 'miniasm', 'rseqc_insertion_profile', 'medaka_variant', 'ngsutils_bam_filter', 'nanopolish_eventalign', 'rseqc_geneBody_coverage', 'purge_dups', 'picard_MarkDuplicates', 'bandage_image', 'rseqc_junction_annotation', 'pilon', 'bamFilter', 'ivar_trim', 'collapse_dataset', 'join1', 'bamtools_split', 'hypo', 'qualimap_rnaseq', 'picard_MergeSamFiles', 'bedtools_genomecoveragebed', 'rseqc_RNA_fragment_size', 'samtools_stats', 'racon', 'rseqc_FPKM_count', 'deeptools_bam_coverage', 'cufflinks', 'rseqc_read_distribution', 'bedtools_bamtofastq', 'nanopolish_variants', 'samtool_filter2', 'ivar_variants', 'samtools_view', 'samtools_markdup', 'featurecounts', 'samtools_depth', '__APPLY_RULES__', 'medaka_variant_pipeline', 'stringtie', 'samtools_sort', 'sam_merge2', 'bedtools_genomecoveragebed_histogram', 'rseqc_junction_saturation', 'deeptools_multi_bam_summary', 'samtools_slice_bam', 'fastqc', 'deeptools_plot_coverage', 'samtools_flagstat', 'ivar_consensus', 'jvarkit_wgscoverageplotter', 'pe_histogram', 'salmon', 'bam_to_sam', 'samtools_merge', 'bamtools', 'rseqc_clipping_profile', 'samtools_mpileup', 'sam_pileup', 'lofreq_call', 'Cut1', 'jbrowse', 'samtools_idxstats', 'rseqc_infer_experiment', 'tabular_to_fastq', 'rseqc_mismatch_profile', 'pycoqc', 'samtools_calmd', 'qualimap_bamqc', 'freebayes', 'samtools_fastx', 'nanoplot', 'CONVERTER_bam_to_bigwig_0', 'bedtools_coveragebed']\n",
      "\n",
      "Test batch 1, Predicted top 77 tools: ['qualimap_rnaseq', 'rseqc_RNA_fragment_size', 'rseqc_FPKM_count', 'deeptools_bam_coverage', 'ivar_consensus', 'samtools_merge', 'pycoqc', 'htseq_count', 'rseqc_geneBody_coverage', 'ivar_trim', 'samtools_markdup', 'samtools_slice_bam', 'rseqc_infer_experiment', 'rseqc_mismatch_profile', 'qualimap_bamqc', 'miniasm', 'rseqc_insertion_profile', 'ngsutils_bam_filter', 'bamFilter', 'hypo', 'samtools_stats', 'racon', 'cufflinks', 'nanopolish_variants', 'samtools_sort', 'fastqc', 'bam_to_sam', 'nanopolish_eventalign', 'samtools_view', '__APPLY_RULES__', 'stringtie', 'bedtools_genomecoveragebed_histogram', 'samtools_flagstat', 'jvarkit_wgscoverageplotter', 'rseqc_clipping_profile', 'freebayes', 'CONVERTER_bam_to_bigwig_0', 'rseqc_junction_annotation', 'bamtools_split', 'picard_MergeSamFiles', 'pilon', 'samtool_filter2', 'nanoplot', 'ococo', 'rseqc_read_distribution', 'samtools_depth', 'jbrowse', 'bedtools_coveragebed', 'sam_pileup', 'rseqc_bam_stat', 'picard_MarkDuplicates', 'salmon', 'medaka_variant_pipeline', 'rseqc_junction_saturation', 'Cut1', 'collapse_dataset', 'samtools_mpileup', 'deeptools_multi_bam_summary', 'bedtools_bamtofastq', 'medaka_consensus', 'purge_dups', 'tabular_to_fastq', 'bedtools_genomecoveragebed', 'bedtools_bamtobed', 'lofreq_call', 'ivar_variants', 'sam_merge2', 'samtools_calmd', 'featurecounts', 'join1', 'bamtools', 'pe_histogram', 'deeptools_plot_coverage', 'bandage_image', 'medaka_variant', 'samtools_idxstats', 'samtools_fastx']\n",
      "\n",
      "Test batch 1, Predicted top 77 tools with weights: ['lofreq_call', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'ivar_trim', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'bandage_image', 'ivar_consensus', 'bamFilter', 'rseqc_infer_experiment', 'collapse_dataset', 'freebayes', 'salmon', 'samtools_flagstat', 'samtools_idxstats', 'bedtools_bamtobed', 'ivar_variants', 'nanoplot', 'rseqc_read_distribution', 'ngsutils_bam_filter', 'bedtools_bamtofastq', 'samtools_sort', 'cufflinks', 'rseqc_geneBody_coverage', 'samtool_filter2', 'samtools_slice_bam', 'pe_histogram', 'medaka_variant', 'racon', 'medaka_consensus', 'bamtools', 'rseqc_bam_stat', 'sam_merge2', 'jbrowse', 'jvarkit_wgscoverageplotter', 'samtools_merge', 'samtools_markdup', 'deeptools_multi_bam_summary', 'samtools_depth', 'purge_dups', 'miniasm', 'tabular_to_fastq', 'picard_MergeSamFiles', 'bedtools_coveragebed', 'sam_pileup', 'deeptools_plot_coverage', 'qualimap_rnaseq', 'pycoqc', 'samtools_calmd', 'pilon', 'samtools_mpileup', 'ococo', 'bamtools_split', 'rseqc_FPKM_count', 'bam_to_sam', 'medaka_variant_pipeline', 'rseqc_junction_annotation', 'bedtools_genomecoveragebed_histogram', 'nanopolish_variants', 'hypo', 'rseqc_junction_saturation', 'rseqc_mismatch_profile', 'nanopolish_eventalign', 'rseqc_clipping_profile', 'rseqc_insertion_profile', 'rseqc_RNA_fragment_size', 'samtools_rmdup', 'rna_star', 'picard_CollectInsertSizeMetrics', 'macs2_callpeak', 'bwa']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.22077922077922077\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.2077922077922078\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'macs2_callpeak', 'diffbind', 'chipseeker', 'Remove beginning1', 'bedtools_subtractbed', 'tp_sort_header_tool', 'join1', 'Add_a_column1', 'Cut1', 'Show beginning1', 'Extract genomic DNA 1', 'macs2_callpeak', 'samtools_rmdup', 'sam_merge2']\n",
      "\n",
      "Test batch 1, True tools: ['samtools_mpileup', 'sam_pileup', 'gatk_indel_realigner', 'deeptools_bamCompare', 'samtools_stats', 'bedtools_multicovtbed', 'picard_AddOrReplaceReadGroups', 'htseq_count', 'jbrowse', 'samtools_bedcov', 'deeptools_bam_compare', 'deeptools_alignmentsieve', 'macs2_callpeak', 'samtools_sort', 'bedtools_bamtobed', 'deeptools_bam_coverage', 'deeptools_multi_bam_summary', 'samtools_idxstats', 'ngsutils_bam_filter', 'fastqc', 'deeptools_plot_coverage', 'cufflinks', 'macs2_filterdup', 'picard_MarkDuplicates', 'naive_variant_caller', 'piranha', 'samtools_flagstat', 'samtools_rmdup', 'ivar_consensus', 'bamFilter', 'picard_SortSam', 'samtool_filter2', 'freebayes', 'samtools_markdup', 'umi_tools_dedup', 'hicexplorer_hicbuildmatrix', 'featurecounts', 'rseqc_bam2wig', 'je_markdupes']\n",
      "\n",
      "Test batch 1, Predicted top 39 tools: ['macs2_callpeak', 'naive_variant_caller', 'ivar_consensus', 'samtools_markdup', 'featurecounts', 'je_markdupes', 'bedtools_multicovtbed', 'ngsutils_bam_filter', 'macs2_filterdup', 'samtools_flagstat', 'deeptools_bamCompare', 'picard_AddOrReplaceReadGroups', 'samtools_bedcov', 'htseq_count', 'deeptools_alignmentsieve', 'gatk_indel_realigner', 'cufflinks', 'hicexplorer_hicbuildmatrix', 'deeptools_plot_coverage', 'deeptools_multi_bam_summary', 'samtools_idxstats', 'deeptools_bam_coverage', 'bamFilter', 'samtools_sort', 'piranha', 'picard_SortSam', 'umi_tools_dedup', 'samtools_rmdup', 'samtools_mpileup', 'sam_pileup', 'deeptools_bam_compare', 'rseqc_bam2wig', 'samtools_stats', 'jbrowse', 'samtool_filter2', 'fastqc', 'bedtools_bamtobed', 'picard_MarkDuplicates', 'freebayes']\n",
      "\n",
      "Test batch 1, Predicted top 39 tools with weights: ['samtools_stats', 'fastqc', 'featurecounts', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'macs2_callpeak', 'samtools_rmdup', 'ivar_consensus', 'bamFilter', 'freebayes', 'samtools_flagstat', 'samtools_idxstats', 'bedtools_bamtobed', 'ngsutils_bam_filter', 'samtools_sort', 'cufflinks', 'samtool_filter2', 'deeptools_bam_compare', 'picard_SortSam', 'jbrowse', 'samtools_markdup', 'deeptools_multi_bam_summary', 'deeptools_alignmentsieve', 'sam_pileup', 'umi_tools_dedup', 'deeptools_plot_coverage', 'naive_variant_caller', 'rseqc_bam2wig', 'piranha', 'samtools_mpileup', 'bedtools_multicovtbed', 'hicexplorer_hicbuildmatrix', 'samtools_bedcov', 'picard_AddOrReplaceReadGroups', 'macs2_filterdup', 'picard_MergeSamFiles', 'pe_histogram', 'rseqc_bam_stat']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.07692307692307693\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.07692307692307693\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['cat1', 'cutadapt', 'fastq_filter']\n",
      "\n",
      "Test batch 1, True tools: ['bg_diamond', 'fastq_manipulation', 'fastq_trimmer', 'medaka_consensus', 'tp_cat', 'minimap2', 'cshl_fastx_trimmer', 'cat1', 'seqtk_sample', 'fastq_to_tabular', 'fastq_to_fasta_python', 'bg_sortmerna', 'rna_star', 'porechop', 'bwa_mem', 'kraken2', 'fastqc', 'cutadapt', 'rbc_mirdeep2_mapper', 'umi_tools_extract', 'bowtie_wrapper', 'bwa', 'bowtie2', 'velveth', 'unicycler', 'canu', 'megahit', 'fastq collapser', 'cshl_fastx_collapser', 'cshl_fastq_to_fasta', 'bwameth', 'flye', 'hisat2']\n",
      "\n",
      "Test batch 1, Predicted top 33 tools: ['porechop', 'hisat2', 'unicycler', 'bowtie_wrapper', 'kraken2', 'bwa_mem', 'bg_diamond', 'rna_star', 'fastq collapser', 'fastqc', 'fastq_to_fasta_python', 'bwameth', 'bowtie2', 'tp_cat', 'fastq_to_tabular', 'bwa', 'megahit', 'fastq_trimmer', 'umi_tools_extract', 'fastq_manipulation', 'canu', 'minimap2', 'flye', 'cutadapt', 'cshl_fastq_to_fasta', 'medaka_consensus', 'rbc_mirdeep2_mapper', 'cshl_fastx_trimmer', 'cshl_fastx_collapser', 'velveth', 'seqtk_sample', 'bg_sortmerna', 'cat1']\n",
      "\n",
      "Test batch 1, Predicted top 33 tools with weights: ['bwa_mem', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'hisat2', 'minimap2', 'kraken2', 'fastq_to_fasta_python', 'porechop', 'tp_cat', 'unicycler', 'bowtie_wrapper', 'bg_sortmerna', 'cshl_fastq_to_fasta', 'bg_diamond', 'rbc_mirdeep2_mapper', 'flye', 'cshl_fastx_trimmer', 'bwameth', 'medaka_consensus', 'bwa', 'megahit', 'fastq_to_tabular', 'seqtk_sample', 'cshl_fastx_collapser', 'fastq_trimmer', 'umi_tools_extract', 'canu', 'filter_by_fasta_ids', 'medaka_consensus_pipeline', 'shovill', 'trimmomatic']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.030303030303030304\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.030303030303030304\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['bowtie2', 'hicexplorer_hicbuildmatrix', 'hicexplorer_chicviewpointbackgroundmodel', 'hicexplorer_chicviewpoint']\n",
      "\n",
      "Test batch 1, True tools: ['hicexplorer_chicsignificantinteractions', 'hicexplorer_chicaggregatestatistic']\n",
      "\n",
      "Test batch 1, Predicted top 2 tools: ['hicexplorer_chicaggregatestatistic', 'hicexplorer_chicsignificantinteractions']\n",
      "\n",
      "Test batch 1, Predicted top 2 tools with weights: ['hicexplorer_chicsignificantinteractions', 'stacks2_tsv2bam']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['__SORTLIST__', 'XTandemAdapter', 'IDMerger', 'IDPosteriorErrorProbability', 'ConsensusID', 'PeptideIndexer', 'FalseDiscoveryRate', 'IDFilter']\n",
      "\n",
      "Test batch 1, True tools: ['FileInfo', 'QCCalculator', 'IDMassAccuracy', 'MzTabExporter', 'ProteinQuantifier', 'TextExporter', 'SequenceCoverageCalculator', 'IDScoreSwitcher', 'IDMerger', 'FalseDiscoveryRate', 'IDMapper', 'ProteinResolver', 'IDConflictResolver', 'IDFileConverter', 'ConsensusID']\n",
      "\n",
      "Test batch 1, Predicted top 15 tools: ['FileInfo', 'MzTabExporter', 'ConsensusID', 'FalseDiscoveryRate', 'IDFileConverter', 'SequenceCoverageCalculator', 'ProteinResolver', 'IDMassAccuracy', 'QCCalculator', 'TextExporter', 'IDScoreSwitcher', 'IDConflictResolver', 'ProteinQuantifier', 'IDMerger', 'IDMapper']\n",
      "\n",
      "Test batch 1, Predicted top 15 tools with weights: ['TextExporter', 'MzTabExporter', 'IDMerger', 'ConsensusID', 'IDFileConverter', 'SequenceCoverageCalculator', 'QCCalculator', 'ProteinResolver', 'graphlan_annotate', 'seq_composition', 'EMBOSS: extractfeat34', 'ip_overlay_segmentation', 'FeatureLinkerUnlabeledQT', 'PSMFeatureExtractor', 'gmx_energy']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.06666666666666667\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'bowtie2', 'pilon', 'prokka', 'gff2bed1', 'bed2gff1']\n",
      "\n",
      "Test batch 1, True tools: ['gffread', 'bedtools_intersectbed', 'tp_awk_tool', 'tp_replace_in_column', 'bedtools_nucbed', 'featurecounts', 'htseq_count', 'tp_tail_tool', 'resize_coordinate_window']\n",
      "\n",
      "Test batch 1, Predicted top 9 tools: ['resize_coordinate_window', 'bedtools_nucbed', 'featurecounts', 'tp_awk_tool', 'htseq_count', 'tp_tail_tool', 'bedtools_intersectbed', 'gffread', 'tp_replace_in_column']\n",
      "\n",
      "Test batch 1, Predicted top 9 tools with weights: ['featurecounts', 'htseq_count', 'bedtools_intersectbed', 'tp_awk_tool', 'tp_replace_in_column', 'tp_tail_tool', 'gffread', 'bedtools_nucbed', 'bedtools_mergebedgraph']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.1111111111111111\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.1111111111111111\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['sarscov2genomes', 'remove_nucleotide_deletions', 'ncbi_blastn_wrapper']\n",
      "\n",
      "Test batch 1, True tools: ['datamash_ops', 'blastxml_to_tabular', 'tp_sorted_uniq', 'filter_tabular', 'comp1', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'tp_split_on_column', 'multiqc', 'Cut1', 'tp_cat', 'blast_parser', 'plasmid_profiler', 'bg_uniq', 'Filter1', 'Add_a_column1', 'Paste1', 'Show beginning1', 'tabular_to_fastq', 'krona-text', 'blastxml_to_top_descr', '__BUILD_LIST__', 'tp_sort_header_tool', 'Grouping1', 'sort1', 'BlastParser_and_hits', 'tp_awk_tool', 'Fetch Taxonomic Ranks', 'tp_head_tool', 'megan_blast2lca', 'tab2fasta', 'blast_to_fasta', 'table_compute', 'join1', 'blast2go', 'add_line_to_file', 'seq_filter_by_id', 'tp_cut_tool']\n",
      "\n",
      "Test batch 1, Predicted top 38 tools: ['Filter1', 'comp1', 'Cut1', 'tp_split_on_column', 'datamash_ops', 'tp_easyjoin_tool', 'krona-text', 'Add_a_column1', 'blast_parser', 'tp_sort_header_tool', 'tp_cut_tool', 'join1', 'megan_blast2lca', 'Paste1', 'plasmid_profiler', 'blast_to_fasta', 'add_line_to_file', 'BlastParser_and_hits', 'filter_tabular', 'sort1', 'blastxml_to_tabular', 'blast2go', 'blastxml_to_top_descr', 'Show beginning1', 'tp_head_tool', 'tabular_to_fastq', 'Grouping1', '__BUILD_LIST__', 'seq_filter_by_id', 'tp_cat', 'taxonomy_krona_chart', 'multiqc', 'Fetch Taxonomic Ranks', 'tp_awk_tool', 'tp_sorted_uniq', 'table_compute', 'bg_uniq', 'tab2fasta']\n",
      "\n",
      "Test batch 1, Predicted top 38 tools with weights: ['Add_a_column1', 'datamash_ops', 'filter_tabular', 'multiqc', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'tp_cat', 'tab2fasta', 'table_compute', 'tp_awk_tool', 'taxonomy_krona_chart', 'tp_head_tool', 'seq_filter_by_id', 'bg_uniq', 'krona-text', 'blastxml_to_top_descr', 'tp_sorted_uniq', 'tabular_to_fastq', 'tp_split_on_column', 'add_line_to_file', 'blastxml_to_tabular', 'megan_blast2lca', 'blast_parser', 'Fetch Taxonomic Ranks', 'blast2go', 'column_remove_by_header', 'tp_find_and_replace', 'query_tabular', 'ncbi_blastn_wrapper', 'scatterplot_rpy', 'tp_replace_in_column', 'deeptools_compute_matrix', 'addName', 'ggplot2_heatmap2', 'featurecounts', 'deeptools_plot_correlation', 'sklearn_svm_classifier', 'goseq']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.10526315789473684\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.05263157894736842\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['make_families', 'correct_barcodes', 'align_families']\n",
      "\n",
      "Test batch 1, True tools: ['mut2read']\n",
      "\n",
      "Test batch 1, Predicted top 1 tools: ['hisat2']\n",
      "\n",
      "Test batch 1, Predicted top 1 tools with weights: ['hisat2']\n",
      "\n",
      "Test batch 1, Precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['1_create_conf']\n",
      "\n",
      "Test batch 1, True tools: ['conf4circos', '5_calc_stat', '4_filter_sam', '3_filter_single_pair']\n",
      "\n",
      "Test batch 1, Predicted top 4 tools: ['3_filter_single_pair', 'diffbind-embl', 'ia_coordinates_of_roi', 'NSPDK_candidateClust']\n",
      "\n",
      "Test batch 1, Predicted top 4 tools with weights: ['plotdexseq', 'pyprophet_score', 'msnbase_readmsdata', 'ip_scale_image']\n",
      "\n",
      "Test batch 1, Precision: 0.25\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['decoyfasta', 'XTandemAdapter']\n",
      "\n",
      "Test batch 1, True tools: ['FileInfo', 'PeptideIndexer', 'idconvert', 'FidoAdapter', 'IDScoreSwitcher', 'IDPosteriorErrorProbability', 'IDMerger', 'proteomics_search_peptide_prophet_1', 'IDFileConverter']\n",
      "\n",
      "Test batch 1, Predicted top 9 tools: ['FileInfo', 'proteomics_search_peptide_prophet_1', 'IDScoreSwitcher', 'PeptideIndexer', 'FidoAdapter', 'IDMerger', 'idconvert', 'IDFileConverter', 'IDPosteriorErrorProbability']\n",
      "\n",
      "Test batch 1, Predicted top 9 tools with weights: ['IDMerger', 'idconvert', 'IDFileConverter', 'seq_composition', 'hyphy_fade', 'TextExporter', 'LuciphorAdapter', 'ConsensusID', 'newick_display']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'hisat2', 'samtools_idxstats', 'Grep1', 'collapse_dataset', 'join1', 'Add_a_column1', 'Cut1', 'Filter1', 'wc_gnu', 'addValue', 'Grouping1', 'cat1']\n",
      "\n",
      "Test batch 1, True tools: ['pureclip', 'uclust2otutable', 'tp_easyjoin_tool', 'mothur_unique_seqs', 'join_files_on_column_fuzzy', 'regexColumn1', 'ggplot2_pca', 'anndata_manipulate', 'cardinal_classification', 'bg_uniq', 'tp_find_and_replace', 'porechop', 'Paste1', 'ggplot_histogram', 'purge_dups', 'fastq_paired_end_interlacer', 'XY_Plot_1', 'rbc_mafft', 'random_lines1', 'unipept', 'query_tabular', 'search_gui', 'p_clip_peaks', 'targetfinder', 'ChangeCase', 'uniprot', 'mothur_phylotype', 'gops_merge_1', 'CONVERTER_fasta_to_tabular', 'annotatemyids', 'join1', 'qiime_pick_closed_reference_otus', 'salsa', 'tp_cut_tool', 'cshl_fastx_renamer', 'mass_spectrometry_imaging_combine', 'cshl_fastx_barcode_splitter', 'bedtools_intersectbed', 'cardinal_combine', '__FLATTEN__', 'minia', 'fastq_groomer', 'cd_hit', '_ensembl_gtf2gene_list', 'barchart_gnuplot', 'cshl_fastx_trimmer', 'cd_hit_est', 'fastq_to_fasta_python', 'quast', 'mergeCols1', 'umi_tools_extract', 'tp_sort_header_tool', 'msstatstmt', 'Grouping1', 'regionalgam_flight_curve', 'filter_by_fasta_ids', 'trinity', 'tab2fasta', 'Grep1', 'table_compute', 'regionalgam_ab_index', 'add_line_to_file', 'tp_replace_in_column', 'bedtools_getfastabed', 'datamash_ops', 'trimmomatic', 'tp_sorted_uniq', 'cardinal_spectra_plots', 'seq_filter_by_mapping', 'fasta2tab', 'fastq_to_tabular', 'heatmap2', 'ggplot2_heatmap2', 'rna_star', 'bwa_mem', 'Add_a_column1', 'fastqc', 'cutadapt', 'cshl_fastx_clipper', 'melt', 'checkFormat', 'rbc_mirdeep2_mapper', 'datamash_transpose', 'w4mclassfilter', 'stoceps_trend_indic', 'snpfreqplot', 'CONVERTER_interval_to_bedstrict_0', 'cshl_fastx_collapser', 'salmon', 'Convert characters1', 'fastq_paired_end_deinterlacer', 'ggplot2_point', 'volcanoplot', 'multiqc', 'Cut1', 'jbrowse', 'minimap2', 'column_order_header_sort', 'ete_homology_classifier', 'ggplot2_histogram', 'merqury', 'uc2otutable', 'Filter1', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'sort1', 'ctb_chemfp_mol2fps', 'bismark_bowtie2', 'bowtie2', 'rna_starsolo', 'addValue', 'cardinal_filtering', 'cast', 'proteomics_moff', 'busco', 'bedtools_sortbed', 'cshl_fastq_to_fasta', 'fastq_filter', 'hisat2', 'Summary_Statistics1', 'trim_galore']\n",
      "\n",
      "Test batch 1, Predicted top 121 tools: ['tp_easyjoin_tool', 'Grep1', 'datamash_ops', 'fastq_to_tabular', '__TAG_FROM_FILE__', 'hisat2', 'regexColumn1', 'fastq_groomer', 'filter_by_fasta_ids', 'bwa_mem', 'Add_a_column1', 'query_tabular', 'bedtools_intersectbed', 'tp_replace_in_column', 'trimmomatic', 'sort1', 'cshl_fastq_to_fasta', 'Paste1', 'tp_cut_tool', 'add_line_to_file', 'Filter1', 'XY_Plot_1', 'barchart_gnuplot', 'fastq_to_fasta_python', 'mergeCols1', 'Grouping1', 'tp_sorted_uniq', 'rna_star', 'merqury', 'table_compute', 'seq_filter_by_mapping', 'melt', 'fastq_paired_end_interlacer', 'cutadapt', 'Cut1', 'minimap2', 'fastq_filter', 'unipept', 'join_files_on_column_fuzzy', 'porechop', 'ggplot_histogram', 'rbc_mafft', 'random_lines1', 'ChangeCase', 'addValue', 'join1', 'bedtools_sortbed', 'Summary_Statistics1', 'annotatemyids', 'cshl_fastx_renamer', 'bedtools_getfastabed', 'multiqc', '__RELABEL_FROM_FILE__', 'cardinal_filtering', 'w4mclassfilter', 'volcanoplot', 'ggplot2_heatmap2', 'column_order_header_sort', 'cshl_fastx_collapser', 'pureclip', 'uniprot', 'minia', 'ggplot2_point', 'cd_hit_est', 'bg_uniq', 'checkFormat', 'datamash_transpose', 'busco', 'targetfinder', '__FLATTEN__', 'fastq_paired_end_deinterlacer', 'mothur_phylotype', 'cshl_fastx_barcode_splitter', 'quast', 'cshl_fastx_clipper', 'trinity', 'fasta2tab', 'umi_tools_extract', 'bismark_bowtie2', 'CONVERTER_interval_to_bedstrict_0', 'cardinal_classification', 'ete_homology_classifier', 'trim_galore', 'uclust2otutable', 'CONVERTER_fasta_to_tabular', 'mothur_unique_seqs', 'tab2fasta', 'purge_dups', 'p_clip_peaks', 'mass_spectrometry_imaging_combine', 'uc2otutable', 'msstatstmt', 'gops_merge_1', 'rna_starsolo', 'salsa', 'stoceps_trend_indic', 'anndata_manipulate', 'ggplot2_pca', 'cd_hit', 'qiime_pick_closed_reference_otus', 'tp_sort_header_tool', 'proteomics_moff', '_ensembl_gtf2gene_list', 'tp_find_and_replace', 'cast', 'rbc_mirdeep2_mapper', 'ggplot2_histogram', 'cshl_fastx_trimmer', 'cardinal_combine', 'jbrowse', 'Convert characters1', 'bowtie2', 'regionalgam_flight_curve', 'regionalgam_ab_index', 'ctb_chemfp_mol2fps', 'heatmap2', 'snpfreqplot', 'fastqc', 'cardinal_spectra_plots', 'search_gui', 'salmon']\n",
      "\n",
      "Test batch 1, Predicted top 121 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'bwa_mem', 'datamash_ops', 'gops_merge_1', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'trimmomatic', 'hisat2', 'multiqc', 'minimap2', 'fastq_to_fasta_python', 'trim_galore', 'datamash_transpose', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'fastq_groomer', 'bedtools_intersectbed', 'porechop', 'fasta2tab', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'salmon', 'busco', 'quast', 'volcanoplot', 'annotatemyids', 'fastq_filter', 'tp_replace_in_column', 'fastq_paired_end_interlacer', 'filter_by_fasta_ids', 'mothur_unique_seqs', 'cshl_fastq_to_fasta', 'rbc_mirdeep2_mapper', 'addValue', 'snpfreqplot', 'trinity', 'bedtools_getfastabed', 'bedtools_sortbed', 'cshl_fastx_trimmer', 'cd_hit', 'ggplot2_point', 'anndata_manipulate', 'cshl_fastx_renamer', 'jbrowse', 'regexColumn1', 'fastq_paired_end_deinterlacer', 'fastq_to_tabular', 'XY_Plot_1', 'cd_hit_est', 'rna_starsolo', 'query_tabular', 'bg_uniq', 'rbc_mafft', 'cshl_fastx_collapser', 'tp_sorted_uniq', 'cardinal_filtering', 'purge_dups', 'umi_tools_extract', 'cshl_fastx_barcode_splitter', 'seq_filter_by_mapping', 'mergeCols1', 'search_gui', 'merqury', 'bismark_bowtie2', 'cshl_fastx_clipper', '_ensembl_gtf2gene_list', 'ggplot2_pca', 'pureclip', 'cardinal_classification', 'salsa', 'join_files_on_column_fuzzy', 'add_line_to_file', 'cardinal_spectra_plots', 'minia', 'targetfinder', 'column_order_header_sort', 'cardinal_combine', 'uniprot', 'ggplot2_histogram', 'msstatstmt', 'ctb_chemfp_mol2fps', 'melt', 'qiime_pick_closed_reference_otus', 'checkFormat', 'cast', 'ete_homology_classifier', 'stoceps_trend_indic', 'regionalgam_ab_index', 'regionalgam_flight_curve', 'mothur_phylotype', 'tp_grep_tool', 'split_file_to_collection', 'goseq', 'tp_tail_tool', 'fasterq_dump', 'filter_tabular', 'combine_metaphlan2_humann2', 'tp_head_tool', 'gtftobed12', 'taxonomy_krona_chart', 'tp_replace_in_line', 'addName', 'kraken2', 'stacks_denovomap', 'tp_split_on_column', 'export2graphlan', 'featurecounts', 'Heatmap', 'ncbi_acc_download', 'fgsea', 'pathview', 'ncbi_blastdbcmd_wrapper', 'tp_cat', 'limma_voom', 'egsea', 'msstats', 'cardinal_mz_images']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.14049586776859505\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.09917355371900827\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'fastqc', 'Grep1', 'collapse_dataset', 'Cut1', 'Filter1', 'wc_gnu', 'Remove beginning1', 'Show beginning1', 'datamash_transpose', 'join1', 'Add_a_column1', 'comp1', 'cat1']\n",
      "\n",
      "Test batch 1, True tools: ['pureclip', 'uclust2otutable', 'tp_easyjoin_tool', 'mothur_unique_seqs', 'join_files_on_column_fuzzy', 'regexColumn1', 'ggplot2_pca', 'anndata_manipulate', 'cardinal_classification', 'bg_uniq', 'tp_find_and_replace', 'porechop', 'Paste1', 'ggplot_histogram', 'purge_dups', 'fastq_paired_end_interlacer', 'XY_Plot_1', 'rbc_mafft', 'random_lines1', 'unipept', 'query_tabular', 'search_gui', 'p_clip_peaks', 'targetfinder', 'ChangeCase', 'uniprot', 'mothur_phylotype', 'gops_merge_1', 'CONVERTER_fasta_to_tabular', 'annotatemyids', 'join1', 'qiime_pick_closed_reference_otus', 'salsa', 'tp_cut_tool', 'cshl_fastx_renamer', 'mass_spectrometry_imaging_combine', 'cshl_fastx_barcode_splitter', 'bedtools_intersectbed', 'cardinal_combine', '__FLATTEN__', 'minia', 'fastq_groomer', 'cd_hit', '_ensembl_gtf2gene_list', 'barchart_gnuplot', 'cshl_fastx_trimmer', 'cd_hit_est', 'fastq_to_fasta_python', 'quast', 'mergeCols1', 'umi_tools_extract', 'tp_sort_header_tool', 'msstatstmt', 'Grouping1', 'regionalgam_flight_curve', 'filter_by_fasta_ids', 'trinity', 'tab2fasta', 'Grep1', 'table_compute', 'regionalgam_ab_index', 'add_line_to_file', 'tp_replace_in_column', 'bedtools_getfastabed', 'datamash_ops', 'trimmomatic', 'tp_sorted_uniq', 'cardinal_spectra_plots', 'seq_filter_by_mapping', 'fasta2tab', 'fastq_to_tabular', 'heatmap2', 'ggplot2_heatmap2', 'rna_star', 'bwa_mem', 'Add_a_column1', 'fastqc', 'cutadapt', 'cshl_fastx_clipper', 'melt', 'checkFormat', 'rbc_mirdeep2_mapper', 'datamash_transpose', 'w4mclassfilter', 'stoceps_trend_indic', 'snpfreqplot', 'CONVERTER_interval_to_bedstrict_0', 'cshl_fastx_collapser', 'salmon', 'Convert characters1', 'fastq_paired_end_deinterlacer', 'ggplot2_point', 'volcanoplot', 'multiqc', 'Cut1', 'jbrowse', 'minimap2', 'column_order_header_sort', 'ete_homology_classifier', 'ggplot2_histogram', 'merqury', 'uc2otutable', 'Filter1', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'sort1', 'ctb_chemfp_mol2fps', 'bismark_bowtie2', 'bowtie2', 'rna_starsolo', 'addValue', 'cardinal_filtering', 'cast', 'proteomics_moff', 'busco', 'bedtools_sortbed', 'cshl_fastq_to_fasta', 'fastq_filter', 'hisat2', 'Summary_Statistics1', 'trim_galore']\n",
      "\n",
      "Test batch 1, Predicted top 121 tools: ['tp_easyjoin_tool', 'barchart_gnuplot', 'Grep1', 'datamash_ops', 'fastq_to_tabular', '__TAG_FROM_FILE__', 'hisat2', 'regexColumn1', 'fastq_groomer', 'filter_by_fasta_ids', 'table_compute', 'bwa_mem', 'Add_a_column1', 'query_tabular', 'bedtools_intersectbed', 'tp_replace_in_column', 'trimmomatic', 'sort1', 'cshl_fastq_to_fasta', 'Paste1', 'join1', 'tp_cut_tool', 'add_line_to_file', 'cutadapt', 'Filter1', 'XY_Plot_1', 'porechop', 'fastq_to_fasta_python', 'mergeCols1', 'Grouping1', 'rna_star', 'merqury', 'rbc_mafft', 'random_lines1', 'ChangeCase', 'seq_filter_by_mapping', 'fastq_paired_end_interlacer', 'Cut1', 'minimap2', 'bedtools_sortbed', 'fastq_filter', 'unipept', 'join_files_on_column_fuzzy', 'ggplot_histogram', 'bedtools_getfastabed', 'tp_sorted_uniq', 'multiqc', '__RELABEL_FROM_FILE__', 'cardinal_filtering', 'w4mclassfilter', 'volcanoplot', 'ggplot2_heatmap2', 'melt', 'addValue', 'Summary_Statistics1', 'annotatemyids', 'cshl_fastx_renamer', '__FLATTEN__', 'column_order_header_sort', 'cd_hit_est', 'minia', 'checkFormat', 'fastq_paired_end_deinterlacer', 'ggplot2_point', 'cshl_fastx_collapser', 'pureclip', 'bg_uniq', 'uniprot', 'trinity', 'busco', 'umi_tools_extract', 'targetfinder', 'quast', 'fasta2tab', 'datamash_transpose', 'mothur_phylotype', 'cshl_fastx_clipper', 'CONVERTER_interval_to_bedstrict_0', 'bismark_bowtie2', 'cshl_fastx_barcode_splitter', 'mothur_unique_seqs', 'ete_homology_classifier', 'purge_dups', 'trim_galore', 'anndata_manipulate', 'uclust2otutable', 'CONVERTER_fasta_to_tabular', 'mass_spectrometry_imaging_combine', 'cardinal_classification', 'tab2fasta', 'uc2otutable', 'p_clip_peaks', 'salsa', 'msstatstmt', 'gops_merge_1', 'rna_starsolo', 'tp_sort_header_tool', 'cshl_fastx_trimmer', 'tp_find_and_replace', 'stoceps_trend_indic', 'jbrowse', 'qiime_pick_closed_reference_otus', 'rbc_mirdeep2_mapper', 'cd_hit', 'proteomics_moff', '_ensembl_gtf2gene_list', 'bowtie2', 'cardinal_combine', 'cast', 'ggplot2_pca', 'Convert characters1', 'ggplot2_histogram', 'regionalgam_flight_curve', 'ctb_chemfp_mol2fps', 'regionalgam_ab_index', 'snpfreqplot', 'heatmap2', 'search_gui', 'fastqc', 'cardinal_spectra_plots', 'salmon']\n",
      "\n",
      "Test batch 1, Predicted top 121 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'bwa_mem', 'datamash_ops', 'gops_merge_1', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'trimmomatic', 'hisat2', 'multiqc', 'minimap2', 'fastq_to_fasta_python', 'trim_galore', 'datamash_transpose', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'fastq_groomer', 'bedtools_intersectbed', 'porechop', 'fasta2tab', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'salmon', 'busco', 'quast', 'volcanoplot', 'annotatemyids', 'fastq_filter', 'tp_replace_in_column', 'fastq_paired_end_interlacer', 'filter_by_fasta_ids', 'mothur_unique_seqs', 'cshl_fastq_to_fasta', 'rbc_mirdeep2_mapper', 'addValue', 'snpfreqplot', 'trinity', 'bedtools_getfastabed', 'bedtools_sortbed', 'cshl_fastx_trimmer', 'cd_hit', 'ggplot2_point', 'anndata_manipulate', 'cshl_fastx_renamer', 'jbrowse', 'regexColumn1', 'fastq_paired_end_deinterlacer', 'fastq_to_tabular', 'XY_Plot_1', 'cd_hit_est', 'rna_starsolo', 'query_tabular', 'bg_uniq', 'rbc_mafft', 'cshl_fastx_collapser', 'tp_sorted_uniq', 'cardinal_filtering', 'purge_dups', 'umi_tools_extract', 'cshl_fastx_barcode_splitter', 'seq_filter_by_mapping', 'mergeCols1', 'search_gui', 'merqury', 'bismark_bowtie2', 'cshl_fastx_clipper', '_ensembl_gtf2gene_list', 'ggplot2_pca', 'pureclip', 'cardinal_classification', 'salsa', 'join_files_on_column_fuzzy', 'add_line_to_file', 'cardinal_spectra_plots', 'minia', 'targetfinder', 'column_order_header_sort', 'cardinal_combine', 'uniprot', 'ggplot2_histogram', 'msstatstmt', 'ctb_chemfp_mol2fps', 'melt', 'qiime_pick_closed_reference_otus', 'checkFormat', 'cast', 'ete_homology_classifier', 'stoceps_trend_indic', 'regionalgam_ab_index', 'regionalgam_flight_curve', 'mothur_phylotype', 'tp_grep_tool', 'split_file_to_collection', 'goseq', 'fasterq_dump', 'tp_tail_tool', 'tp_head_tool', 'combine_metaphlan2_humann2', 'filter_tabular', 'gtftobed12', 'taxonomy_krona_chart', 'tp_replace_in_line', 'addName', 'stacks_denovomap', 'export2graphlan', 'featurecounts', 'kraken2', 'ncbi_acc_download', 'tp_split_on_column', 'fgsea', 'ncbi_blastdbcmd_wrapper', 'pathview', 'tp_cat', 'egsea', 'limma_voom', 'msstats', 'Heatmap', 'dropletutils_read_10x']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.14049586776859505\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.09917355371900827\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'macs2_callpeak', 'diffbind', 'chipseeker', 'samtools_rmdup', 'deeptools_bam_coverage', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'bedtools_intersectbed', 'Cut1', 'Extract genomic DNA 1', 'macs2_callpeak', 'bowtie2']\n",
      "\n",
      "Test batch 1, True tools: ['velvetoptimiser', 'rseqc_bam_stat', 'gatk2_realigner_target_creator', 'deeptools_bamCompare', 'stacks2_refmap', 'get_read_pipeline', 'bedtools_multicovtbed', 'picard_AddOrReplaceReadGroups', 'htseq_count', '__MERGE_COLLECTION__', 'deeptools_alignmentsieve', 'bedtools_bamtobed', 'ococo', 'bamSplit', 'rgPicardMarkDups', 'kraken2', 'ngsutils_bam_filter', 'fastq_paired_end_interlacer', 'picard_MarkDuplicates', 'peakachu', 'basil', 'samtools_filter', 'collection_element_identifiers', 'cuffdiff', 'bwa', 'deeptools_compute_gc_bias', 'pilon', 'bamFilter', 'ivar_trim', 'tp_tail_tool', 'deeptools_plot_enrichment', 'picard_CollectInsertSizeMetrics', 'humann2', 'deeptools_plot_fingerprint', 'picard_MergeSamFiles', 'mageck_count', 'umi_tools_count', 'spades', 'tophat2', 'samtools_split', 'bundle_collection', 'bedtools_genomecoveragebed', 'deeptools_computeGCBias', 'bcftools_mpileup', 'deeptools_bamCoverage', 'bedtools_intersectbed', 'PicardASMetrics', 'samtools_stats', '__FLATTEN__', 'minia', 'macs2_callpeak', 'racon', 'picard_FilterSamReads', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'picard_ARRG', 'gatk4_mutect2', 'cufflinks', 'tp_sort_header_tool', 'Grouping1', 'naive_variant_caller', 'lofreq_indelqual', 'samtools_rmdup', 'deeptools_bamFingerprint', 'trinity', 'picard_SortSam', 'picard_artifact_metrics', 'samtool_filter2', 'ivar_variants', 'samtools_view', 'picard_QualityScoreDistribution', 'paralyzer', 'megahit', 'rnaspades', 'hicexplorer_hicbuildmatrix', 'featurecounts', 'picard_DownsampleSam', 'samtools_depth', 'bam_to_bigwig', 'je_markdupes', 'bioext_bam2msa', 'trimmomatic', 'lofreq_viterbi', 'seq_filter_by_mapping', 'stringtie', 'tp_cat', 'deeptools_bamCorrelate', 'samtools_sort', 'fastq_to_tabular', 'sam_merge2', 'deeptools_correct_gc_bias', 'bg_sortmerna', 'bam-filter', 'rna_star', 'bwa_mem', 'deeptools_multi_bam_summary', 'samtools_slice_bam', 'fastqc', 'peakcalling_macs', 'rbc_mirdeep2_mapper', 'samtools_flagstat', 'fastq_paired_end_joiner', 'ivar_consensus', 'metaphlan', 'picard_CASM', 'metaphlan2_hutlab', 'umi_tools_dedup', 'varscan', 'jvarkit_wgscoverageplotter', 'diffbind', 'picard_CleanSam', 'cufflinks_prok', 'pe_histogram', 'salmon', 'rseqc_inner_distance', 'bam_to_sam', 'macs2_predictd', 'samtools_merge', 'bamtools', 'samtools_mpileup', 'sam_pileup', 'lofreq_call', 'multiqc', 'jbrowse', 'deeptools_bam_compare', 'PicardInsertSize', 'cat1', 'mimodd_varcall', 'sam_to_bam', 'Extract genomic DNA 1', 'metaspades', 'samtools_idxstats', 'bamleftalign', 'rseqc_infer_experiment', 'mimodd_delcall', '__BUILD_LIST__', 'qualimap_bamqc', 'metaphlan2', 'unicycler', 'freebayes', 'samtools_fastx', 'bedtools_genomecoveragebed_bedgraph', 'cshl_fastq_to_fasta', 'CONVERTER_bam_to_bigwig_0', 'bedtools_coveragebed', 'hisat2']\n",
      "\n",
      "Test batch 1, Predicted top 146 tools: ['samtools_filter', 'pilon', 'humann2', 'umi_tools_count', 'tophat2', 'samtools_split', 'deeptools_computeGCBias', 'bcftools_mpileup', 'macs2_callpeak', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'picard_ARRG', 'trinity', 'picard_artifact_metrics', 'samtool_filter2', 'paralyzer', 'deeptools_bamCorrelate', 'fastq_to_tabular', 'deeptools_correct_gc_bias', 'rna_star', 'varscan', 'samtools_merge', 'samtools_mpileup', 'lofreq_call', 'multiqc', 'metaspades', 'metaphlan2', 'bedtools_genomecoveragebed_bedgraph', 'gatk2_realigner_target_creator', 'stacks2_refmap', 'htseq_count', 'peakachu', 'cuffdiff', 'bwa', 'ivar_trim', 'tp_tail_tool', 'picard_CollectInsertSizeMetrics', 'mageck_count', 'tp_sort_header_tool', 'rnaspades', 'featurecounts', 'samtools_depth', 'je_markdupes', 'tp_cat', 'bg_sortmerna', 'bwa_mem', 'samtools_slice_bam', 'fastq_paired_end_joiner', 'metaphlan', 'sam_pileup', 'deeptools_bam_compare', 'rseqc_infer_experiment', 'qualimap_bamqc', 'rseqc_bam_stat', 'bedtools_multicovtbed', 'ngsutils_bam_filter', 'bamFilter', 'deeptools_plot_fingerprint', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', 'PicardASMetrics', 'samtools_stats', '__FLATTEN__', 'gatk4_mutect2', 'deeptools_bamFingerprint', 'picard_QualityScoreDistribution', 'megahit', 'bam_to_bigwig', 'trimmomatic', 'lofreq_viterbi', 'seq_filter_by_mapping', 'samtools_sort', 'bam-filter', 'fastqc', 'rbc_mirdeep2_mapper', 'picard_CASM', 'metaphlan2_hutlab', 'picard_CleanSam', 'salmon', 'bam_to_sam', 'PicardInsertSize', 'unicycler', 'cshl_fastq_to_fasta', 'velvetoptimiser', 'bamSplit', 'fastq_paired_end_interlacer', 'samtools_view', 'stringtie', 'deeptools_multi_bam_summary', 'samtools_flagstat', 'jvarkit_wgscoverageplotter', 'diffbind', 'cufflinks_prok', 'cat1', 'mimodd_varcall', 'Extract genomic DNA 1', 'samtools_idxstats', 'bamleftalign', 'freebayes', 'CONVERTER_bam_to_bigwig_0', 'deeptools_bamCompare', 'get_read_pipeline', 'kraken2', 'collection_element_identifiers', 'spades', 'minia', 'bedtools_bamtobed', 'naive_variant_caller', 'sam_merge2', 'macs2_predictd', 'ococo', 'deeptools_compute_gc_bias', 'picard_SortSam', 'rseqc_inner_distance', 'racon', 'mimodd_delcall', 'deeptools_bamCoverage', 'picard_AddOrReplaceReadGroups', 'deeptools_alignmentsieve', 'ivar_consensus', 'lofreq_indelqual', 'pe_histogram', '__MERGE_COLLECTION__', 'rgPicardMarkDups', 'sam_to_bam', 'hisat2', 'deeptools_plot_enrichment', 'basil', 'jbrowse', 'samtools_rmdup', 'hicexplorer_hicbuildmatrix', 'picard_DownsampleSam', 'bioext_bam2msa', 'picard_MarkDuplicates', 'cufflinks', 'bundle_collection', 'picard_FilterSamReads', 'bedtools_coveragebed', 'bamtools', 'peakcalling_macs', 'picard_MergeSamFiles', 'Grouping1', 'ivar_variants', 'umi_tools_dedup', 'samtools_fastx', '__BUILD_LIST__']\n",
      "\n",
      "Test batch 1, Predicted top 146 tools with weights: ['lofreq_call', 'bwa_mem', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'ivar_trim', 'lofreq_indelqual', 'lofreq_viterbi', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'rna_star', 'trimmomatic', 'hisat2', 'multiqc', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'kraken2', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'samtools_rmdup', 'tp_cat', 'spades', 'ivar_consensus', 'bamFilter', 'rseqc_infer_experiment', 'unicycler', 'freebayes', 'salmon', 'samtools_flagstat', 'samtools_idxstats', 'bamleftalign', 'bedtools_bamtobed', 'metaphlan2', 'ivar_variants', 'fastq_paired_end_interlacer', 'ngsutils_bam_filter', 'bg_sortmerna', 'samtools_sort', 'cufflinks', 'cshl_fastq_to_fasta', 'samtool_filter2', 'rbc_mirdeep2_mapper', 'samtools_slice_bam', 'deeptools_bam_compare', 'pe_histogram', 'trinity', 'Extract genomic DNA 1', 'racon', 'bioext_bam2msa', 'tp_tail_tool', 'tophat2', 'bcftools_mpileup', 'bwa', 'bamtools', 'humann2', 'rseqc_bam_stat', 'picard_SortSam', 'sam_merge2', 'jbrowse', 'jvarkit_wgscoverageplotter', 'megahit', 'fastq_to_tabular', 'peakachu', 'samtools_merge', 'metaphlan', 'diffbind', 'picard_CollectInsertSizeMetrics', 'fastq_paired_end_joiner', 'deeptools_multi_bam_summary', 'deeptools_plot_fingerprint', 'samtools_depth', 'picard_CASM', 'cuffdiff', 'velvetoptimiser', 'picard_MergeSamFiles', 'picard_FilterSamReads', 'deeptools_alignmentsieve', 'metaspades', 'PicardASMetrics', 'bedtools_genomecoveragebed_bedgraph', 'mageck_count', 'bedtools_coveragebed', 'sam_pileup', 'seq_filter_by_mapping', 'umi_tools_dedup', 'naive_variant_caller', 'varscan', 'mimodd_varcall', 'collection_element_identifiers', 'pilon', 'samtools_mpileup', 'bedtools_multicovtbed', 'rnaspades', 'ococo', 'minia', 'hicexplorer_hicbuildmatrix', 'gatk4_mutect2', 'picard_DownsampleSam', 'bam_to_sam', 'samtools_split', 'basil', 'picard_artifact_metrics', 'stacks2_refmap', 'deeptools_plot_enrichment', 'picard_AddOrReplaceReadGroups', 'picard_QualityScoreDistribution', 'rseqc_inner_distance', 'paralyzer', 'picard_ARRG', 'get_read_pipeline', 'deeptools_bamCorrelate', 'rgPicardMarkDups', 'samtools_filter', 'bamSplit', 'mimodd_delcall', 'rseqc_read_distribution', 'qualimap_rnaseq', 'rseqc_geneBody_coverage', 'fastq_groomer', 'macs2_bdgpeakcall', 'idr', 'macs2_bdgcmp', 'intervene_pairwise', 'samtools_fixmate', 'intervene_upset', 'rseqc_read_duplication', 'pileometh', 'gemini_query', 'macs2_bdgdiff', 'exomedepth', 'bowtie2', 'rseqc_bam2wig', 'picard_ReorderSam', 'deeptools_plot_coverage', 'bellerophon', 'Add_a_column1', 'delly_call', 'cutadapt', 'picard_CollectGcBiasMetrics', 'control_freec', 'bedtools_genomecoveragebed_histogram']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.10273972602739725\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0958904109589041\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['Grep1', 'computeMatrix']\n",
      "\n",
      "Test batch 1, True tools: ['dt_profiler', 'heatmapper_deepTools']\n",
      "\n",
      "Test batch 1, Predicted top 2 tools: ['dt_profiler', 'heatmapper_deepTools']\n",
      "\n",
      "Test batch 1, Predicted top 2 tools with weights: ['EMBOSS: infoseq46', 'MetaboliteAdductDecharger']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['fastq_join', 'cutadapt', 'tp_cat']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'rbc_mirdeep2_quantifier', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'regexColumn1', 'samtools_bedcov', 'lastz_wrapper_2', 'staramr_search', 'ggplot2_pca', 'gprofiler_gost', 'anndata_manipulate', 'miniasm', 'kraken2', 'porechop', 'fastp', 'Paste1', 'Show beginning1', 'maxquant_mqpar', 'nspdk_sparse', 'XY_Plot_1', 'rbc_mafft', 'unipept', 'bwa', 'query_tabular', 'meryl', 'sample_seqs', 'uniprot', 'tp_tail_tool', 'annotatemyids', 'join1', 'qiime_pick_closed_reference_otus', 'tp_uniq_tool', 'humann2', 'spades', 'tp_cut_tool', 'wc_gnu', 'nextclade', 'metilene', 'bedtools_intersectbed', 'je_demultiplex', '__FLATTEN__', 'fastq_groomer', '_ensembl_gtf2gene_list', 'Remove beginning1', 'racon', 'cshl_fastx_trimmer', 'DecoyDatabase', 'cd_hit_est', 'cardinal_mz_images', 'fastq_to_fasta_python', 'tp_replace_in_line', 'mergeCols1', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'regionalgam_flight_curve', 'filter_by_fasta_ids', 'trinity', 'tab2fasta', 'export_to_cluster', 'table_compute', 'regionalgam_ab_index', 'tp_grep_tool', 'PlasFlow', 'tp_replace_in_column', 'sRNAPipe', 'datamash_ops', 'trimmomatic', 'tp_sorted_uniq', 'mass_spectrometry_imaging_filtering', 'tombo_resquiggle', 'tp_multijoin_tool', 'fasta2tab', 'circos_interval_to_tile', 'fasta_concatenate0', 'clustalw', 'blast_parser', 'bcftools_consensus', 'sam_merge2', 'bg_sortmerna', 'ggplot2_heatmap2', 'rna_star', 'bwa_mem', 'Add_a_column1', 'fastqc', 'cutadapt', 'cshl_fastx_clipper', 'filtlong', 'csv_to_tabular', 'flashlfq', 'rbc_mirdeep2_mapper', 'fasta-stats', 'w4mclassfilter', 'trinity_align_and_estimate_abundance', 'ont_fast5_api_fast5_subset', 'miranda', 'pangolin', 'canu', 'umi_tools_whitelist', 'gmx_sim', 'openbabel_filter', 'cshl_fastx_collapser', 'jellyfish', 'bwameth', 'tp_sed_tool', 'medaka_consensus_pipeline', 'fastq_paired_end_deinterlacer', 'ggplot2_point', 'volcanoplot', 'iReport', 'circos', 'multiqc', 'Cut1', 'jbrowse', 'minimap2', 'column_order_header_sort', 'cat1', 'scatterplot_rpy', 'Filter1', 'prinseq', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'rdock_sort_filter', 'Count1', 'sort1', 'ctb_chemfp_mol2fps', 'ncbi_blastn_wrapper', 'bowtie2', 'metaphlan2', 'qiime_align_seqs', 'snp_sites', 'addValue', 'cardinal_filtering', 'circos_interval_to_text', 'cast', 'split_file_to_collection', 'prokka', 'bedtools_sortbed', 'nanoplot', 'cshl_fastq_to_fasta', 'fasta_compute_length', 'hisat2', 'drep_dereplicate', 'trim_galore']\n",
      "\n",
      "Test batch 1, Predicted top 145 tools: ['tp_easyjoin_tool', 'get_flanks1', 'staramr_search', 'tp_uniq_tool', 'humann2', 'metilene', 'cshl_fastx_trimmer', 'fastq_to_fasta_python', 'mergeCols1', 'Grouping1', 'trinity', 'datamash_ops', 'tp_sorted_uniq', 'mass_spectrometry_imaging_filtering', 'tp_multijoin_tool', 'fasta2tab', 'blast_parser', 'jellyfish', 'bwameth', 'multiqc', '__RELABEL_FROM_FILE__', 'rdock_sort_filter', 'metaphlan2', 'nanoplot', 'hisat2', 'bg_column_arrange_by_header', 'regexColumn1', 'rbc_mafft', 'meryl', 'tp_tail_tool', 'fastq_groomer', 'DecoyDatabase', 'tp_sort_header_tool', 'filter_by_fasta_ids', 'table_compute', 'PlasFlow', 'fasta_concatenate0', 'bwa_mem', 'Add_a_column1', 'filtlong', 'w4mclassfilter', 'medaka_consensus_pipeline', 'volcanoplot', 'prinseq', 'ncbi_blastn_wrapper', 'cast', 'fasta_compute_length', 'miniasm', 'maxquant_mqpar', 'query_tabular', 'bedtools_intersectbed', 'racon', 'export_to_cluster', 'regionalgam_ab_index', 'tp_replace_in_column', 'trimmomatic', 'clustalw', 'ggplot2_heatmap2', 'fastqc', 'ont_fast5_api_fast5_subset', 'umi_tools_whitelist', 'fastq_paired_end_deinterlacer', 'jbrowse', 'column_order_header_sort', 'sort1', 'bowtie2', 'qiime_align_seqs', 'snp_sites', 'addValue', 'circos_interval_to_text', 'prokka', 'cshl_fastq_to_fasta', 'trim_galore', 'lastz_wrapper_2', 'ggplot2_pca', 'Paste1', 'nspdk_sparse', 'join1', 'tp_cut_tool', 'cd_hit_est', 'sRNAPipe', 'cutadapt', 'cshl_fastx_clipper', 'csv_to_tabular', 'tp_sed_tool', 'minimap2', 'cat1', 'scatterplot_rpy', 'Filter1', 'bedtools_sortbed', 'rbc_mirdeep2_quantifier', 'samtools_bedcov', 'anndata_manipulate', 'XY_Plot_1', 'uniprot', 'spades', 'circos_interval_to_tile', 'ctb_chemfp_mol2fps', 'bwa', 'tp_replace_in_line', 'bcftools_consensus', 'bg_sortmerna', 'miranda', 'pangolin', '_ensembl_gtf2gene_list', 'porechop', 'nextclade', 'rna_star', '__TAG_FROM_FILE__', 'cardinal_filtering', 'qiime_pick_closed_reference_otus', 'je_demultiplex', '__FLATTEN__', 'regionalgam_flight_curve', 'tp_grep_tool', 'ggplot2_point', 'cshl_fastx_collapser', 'Cut1', 'gmx_sim', 'rbc_mirdeep2_mapper', 'sample_seqs', 'openbabel_filter', 'iReport', 'annotatemyids', 'Show beginning1', 'Remove beginning1', 'umi_tools_extract', 'tab2fasta', 'trinity_align_and_estimate_abundance', 'canu', 'wc_gnu', 'drep_dereplicate', 'tp_split_on_column', 'circos', 'split_file_to_collection', 'fastp', 'unipept', 'kraken2', 'gprofiler_gost', 'flashlfq', 'fasta-stats', 'tombo_resquiggle', 'Count1', 'cardinal_mz_images', 'sam_merge2']\n",
      "\n",
      "Test batch 1, Predicted top 145 tools with weights: ['Add_a_column1', 'tp_replace_in_line', 'fastp', 'bwa_mem', 'datamash_ops', 'bcftools_consensus', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'trimmomatic', 'hisat2', 'prokka', 'multiqc', 'minimap2', 'kraken2', 'fastq_to_fasta_python', 'ncbi_blastn_wrapper', 'trim_galore', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'fastq_groomer', 'bedtools_intersectbed', 'porechop', 'fasta2tab', 'spades', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'PlasFlow', 'volcanoplot', 'tp_grep_tool', 'fasta-stats', 'nextclade', 'circos', 'annotatemyids', 'metaphlan2', 'nanoplot', 'tp_replace_in_column', 'filter_by_fasta_ids', 'bg_sortmerna', 'split_file_to_collection', 'tp_sed_tool', 'cshl_fastq_to_fasta', 'rbc_mirdeep2_mapper', 'addValue', 'gmx_sim', 'trinity', 'pangolin', 'racon', 'clustalw', 'filtlong', 'bedtools_sortbed', 'cshl_fastx_trimmer', 'bwameth', 'staramr_search', 'meryl', 'tp_tail_tool', 'bwa', 'ggplot2_point', 'anndata_manipulate', 'humann2', 'trinity_align_and_estimate_abundance', 'sam_merge2', 'jbrowse', 'regexColumn1', 'fastq_paired_end_deinterlacer', 'fasta_compute_length', 'XY_Plot_1', 'cd_hit_est', 'query_tabular', 'rbc_mafft', 'cshl_fastx_collapser', 'tp_sorted_uniq', 'cardinal_filtering', 'umi_tools_extract', 'miniasm', 'rbc_mirdeep2_quantifier', 'je_demultiplex', 'medaka_consensus_pipeline', 'sample_seqs', 'metilene', 'mergeCols1', 'tp_split_on_column', 'nspdk_sparse', 'snp_sites', 'cshl_fastx_clipper', '_ensembl_gtf2gene_list', 'ggplot2_pca', 'qiime_align_seqs', 'canu', 'scatterplot_rpy', 'fasta_concatenate0', 'cardinal_mz_images', 'drep_dereplicate', 'sRNAPipe', 'prinseq', 'gprofiler_gost', 'get_flanks1', 'tp_uniq_tool', 'bg_column_arrange_by_header', 'jellyfish', 'column_order_header_sort', 'maxquant_mqpar', 'uniprot', 'DecoyDatabase', 'ctb_chemfp_mol2fps', 'samtools_bedcov', 'flashlfq', 'blast_parser', 'openbabel_filter', 'rdock_sort_filter', 'umi_tools_whitelist', 'circos_interval_to_text', 'qiime_pick_closed_reference_otus', 'cast', 'mass_spectrometry_imaging_filtering', 'regionalgam_ab_index', 'regionalgam_flight_curve', 'tp_find_and_replace', 'tp_cat', 'sklearn_ensemble', 'fastq_paired_end_interlacer', 'seqtk_subseq', 'fastq_to_tabular', 'megahit', 'salmon', 'stringtie', 'seq_filter_by_id', 'quast', 'kallisto_quant', 'bowtie_wrapper', 'plotly_ml_performance_plots', 'edger', 'targetfinder', 'limma_voom', 'bismark_bowtie2', 'goenrichment', 'minia', 'mothur_make_contigs', 'rbc_mirdeep2', 'abyss-pe', 'tbvcfreport', 'ncbi_acc_download', 'bismark_bowtie']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.09655172413793103\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.08275862068965517\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['schicexplorer_schicqualitycontrol', 'schicexplorer_schicnormalize', 'schicexplorer_schicclustersvl']\n",
      "\n",
      "Test batch 1, True tools: ['schicexplorer_schicplotclusterprofiles', 'schicexplorer_schicconsensusmatrices']\n",
      "\n",
      "Test batch 1, Predicted top 2 tools: ['schicexplorer_schicplotclusterprofiles', 'schicexplorer_schicclustersvl']\n",
      "\n",
      "Test batch 1, Predicted top 2 tools with weights: ['cp_display_data_on_image', 'cp_export_to_spreadsheet']\n",
      "\n",
      "Test batch 1, Precision: 0.5\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['dimspy_process_scans', 'dimspy_replicate_filter']\n",
      "\n",
      "Test batch 1, True tools: []\n",
      "\n",
      "Test batch 1, Predicted top 0 tools: []\n",
      "\n",
      "Test batch 1, Predicted top 0 tools with weights: []\n",
      "\n",
      "Test batch 1, Precision: 0.5\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'bowtie2', 'samtools_rmdup', 'macs2_callpeak', 'macs2_callpeak', 'bedtools_intersectbed', 'Cut1', 'Filter1', 'Extract genomic DNA 1', 'macs2_callpeak', 'deeptools_compute_matrix']\n",
      "\n",
      "Test batch 1, True tools: ['deeptools_plot_profile', 'deeptools_compute_matrix_operations', 'bedtools_intersectbed', 'deeptools_plot_pca', 'deeptools_plot_heatmap', 'Remove beginning1']\n",
      "\n",
      "Test batch 1, Predicted top 6 tools: ['deeptools_plot_pca', 'deeptools_plot_heatmap', 'deeptools_plot_profile', 'deeptools_compute_matrix_operations', 'Remove beginning1', 'bedtools_intersectbed']\n",
      "\n",
      "Test batch 1, Predicted top 6 tools with weights: ['bedtools_intersectbed', 'deeptools_plot_heatmap', 'deeptools_plot_profile', 'deeptools_plot_pca', 'deeptools_compute_matrix_operations', 'tp_tail_tool']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.16666666666666666\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.16666666666666666\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['Filter1', '__BUILD_LIST__', 'cardinal_combine']\n",
      "\n",
      "Test batch 1, True tools: ['cardinal_preprocessing', 'cardinal_filtering', 'cardinal_classification', 'cardinal_spectra_plots', 'Filter1', 'tp_replace_in_column', 'cardinal_quality_report', 'maldi_quant_peak_detection', 'cardinal_data_exporter', 'maldi_quant_preprocessing']\n",
      "\n",
      "Test batch 1, Predicted top 10 tools: ['cardinal_spectra_plots', 'cardinal_quality_report', 'cardinal_preprocessing', 'cardinal_classification', 'cardinal_filtering', 'maldi_quant_peak_detection', 'Filter1', 'tp_replace_in_column', 'maldi_quant_preprocessing', 'cardinal_data_exporter']\n",
      "\n",
      "Test batch 1, Predicted top 10 tools with weights: ['tp_replace_in_column', 'cardinal_filtering', 'cardinal_quality_report', 'cardinal_classification', 'cardinal_spectra_plots', 'cardinal_data_exporter', 'maldi_quant_peak_detection', 'cardinal_mz_images', 'Add_a_column1', 'datamash_ops']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.2\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['PeakPickerHiRes', 'FileFilter', 'xcms-find-peaks', 'xcms-collect-peaks', 'xcms-group-peaks', 'xcms-blankfilter', 'xcms-dilutionfilter', 'camera-annotate-peaks', 'camera-group-fwhm', 'camera-find-adducts', 'camera-find-isotopes', 'msms2metfrag-multiple', 'metfrag-cli-batch-multiple']\n",
      "\n",
      "Test batch 1, True tools: ['passatutto']\n",
      "\n",
      "Test batch 1, Predicted top 1 tools: ['passatutto']\n",
      "\n",
      "Test batch 1, Predicted top 1 tools with weights: ['hyphy_slac']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['mothur_align_seqs', 'mothur_screen_seqs', 'mothur_filter_seqs', 'mothur_unique_seqs']\n",
      "\n",
      "Test batch 1, True tools: ['datamash_ops', 'mothur_remove_seqs', 'mothur_get_lineage', 'mothur_cluster', 'mothur_cluster_split', 'fasta2tab', 'mothur_chimera_vsearch', 'minimap2', 'mothur_pre_cluster', 'hmmer_hmmsearch', 'mothur_classify_seqs', 'column_remove_by_header', 'mothur_screen_seqs', 'Paste1', 'mothur_split_groups', 'Show beginning1', 'mothur_count_seqs', 'mothur_split_abund', 'mothur_trim_seqs', 'mothur_pairwise_seqs', 'mothur_align_seqs', 'ncbi_blastn_wrapper', 'mothur_chimera_uchime', 'mycrobiota-qc-report', 'mothur_remove_lineage', 'mothur_summary_seqs', 'seq_filter_by_id', 'mothur_get_oturep']\n",
      "\n",
      "Test batch 1, Predicted top 28 tools: ['datamash_ops', 'hmmer_hmmsearch', 'mothur_align_seqs', 'mothur_split_abund', 'mothur_remove_lineage', 'Paste1', 'mothur_trim_seqs', 'mycrobiota-qc-report', 'mothur_summary_seqs', 'mothur_get_lineage', 'mothur_pre_cluster', 'ncbi_blastn_wrapper', 'mothur_chimera_vsearch', 'mothur_screen_seqs', 'mothur_remove_seqs', 'Show beginning1', 'mothur_cluster_split', 'fasta2tab', 'mothur_count_seqs', 'mothur_classify_seqs', 'seq_filter_by_id', 'mothur_chimera_uchime', 'minimap2', 'column_remove_by_header', 'mothur_split_groups', 'mothur_get_oturep', 'mothur_pairwise_seqs', 'mothur_cluster']\n",
      "\n",
      "Test batch 1, Predicted top 28 tools with weights: ['datamash_ops', 'minimap2', 'ncbi_blastn_wrapper', 'fasta2tab', 'mothur_summary_seqs', 'mothur_screen_seqs', 'mothur_count_seqs', 'mothur_align_seqs', 'seq_filter_by_id', 'mothur_cluster_split', 'mothur_classify_seqs', 'mothur_pre_cluster', 'mothur_remove_lineage', 'hmmer_hmmsearch', 'mothur_chimera_vsearch', 'mothur_remove_seqs', 'mothur_cluster', 'column_remove_by_header', 'mothur_get_oturep', 'mothur_trim_seqs', 'mothur_pairwise_seqs', 'mothur_split_abund', 'mothur_split_groups', 'fastqc', 'tp_easyjoin_tool', 'meryl', 'mothur_classify_otu', 'megahit']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.14285714285714285\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.14285714285714285\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['fastq_dump', 'cshl_fastq_quality_filter', 'fastq_quality_trimmer', 'tophat2']\n",
      "\n",
      "Test batch 1, True tools: ['samtools_mpileup', 'tophat_fusion_post', 'samtools_stats', 'multiqc', 'htseq_count', 'CONVERTER_bed_to_tabix_0', 'jbrowse', 'tp_cat', 'stringtie', 'samtools_sort', 'freebayes_wrapper', 'rgPicardMarkDups', 'deeptools_bam_coverage', 'sam_merge2', 'deeptools_multi_bam_summary', 'fastqc', 'filter_bed_on_splice_junctions', 'cufflinks', 'picard_MarkDuplicates', 'cuffdiff', 'naive_variant_caller', 'cuffquant', 'samtools_flagstat', 'samtools_rmdup', 'bedtools_bamtofastq', 'cuffnorm', 'picard_SortSam', 'samtool_filter2', 'samtools_view', 'freebayes', 'picard_QualityScoreDistribution', 'CONVERTER_interval_to_bedstrict_0', 'featurecounts', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'picard_MergeSamFiles', 'rseqc_bam2wig', 'bam_to_sam']\n",
      "\n",
      "Test batch 1, Predicted top 37 tools: ['filter_bed_on_splice_junctions', 'samtools_mpileup', 'htseq_count', 'CONVERTER_bed_to_tabix_0', 'samtools_stats', 'bedtools_bamtofastq', 'picard_QualityScoreDistribution', 'samtools_sort', 'samtools_view', 'deeptools_multi_bam_summary', 'cuffquant', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'cuffnorm', 'tophat_fusion_post', 'picard_MarkDuplicates', 'samtools_rmdup', 'CONVERTER_interval_to_bedstrict_0', 'rseqc_bam2wig', 'deeptools_bam_coverage', 'cuffdiff', 'featurecounts', 'cufflinks', 'samtool_filter2', 'picard_SortSam', 'jbrowse', 'naive_variant_caller', 'stringtie', 'samtools_flagstat', 'multiqc', 'rgPicardMarkDups', 'tp_cat', 'freebayes', 'picard_MergeSamFiles', 'fastqc', 'bam_to_sam', 'freebayes_wrapper', 'sam_merge2']\n",
      "\n",
      "Test batch 1, Predicted top 37 tools with weights: ['samtools_view', 'samtools_stats', 'fastqc', 'featurecounts', 'multiqc', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'samtools_rmdup', 'tp_cat', 'freebayes', 'samtools_flagstat', 'bedtools_bamtofastq', 'samtools_sort', 'cufflinks', 'samtool_filter2', 'picard_SortSam', 'sam_merge2', 'jbrowse', 'deeptools_multi_bam_summary', 'cuffdiff', 'picard_MergeSamFiles', 'naive_variant_caller', 'rseqc_bam2wig', 'samtools_mpileup', 'bam_to_sam', 'cuffnorm', 'picard_QualityScoreDistribution', 'tophat_fusion_post', 'freebayes_wrapper', 'rgPicardMarkDups', 'filter_bed_on_splice_junctions', 'bamFilter', 'rseqc_bam_stat', 'ivar_consensus', 'tp_sort_header_tool']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.10810810810810811\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.08108108108108109\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'fastqc', 'Grep1', 'collapse_dataset', 'Cut1', 'collection_column_join', 'regex_replace', 'melt', 'Show beginning1', 'datamash_transpose', 'Remove beginning1', 'datamash_ops', 'join1', 'Add_a_column1', 'Filter1', 'addValue', 'cat1']\n",
      "\n",
      "Test batch 1, True tools: ['pureclip', 'uclust2otutable', 'tp_easyjoin_tool', 'mothur_unique_seqs', 'join_files_on_column_fuzzy', 'regexColumn1', 'ggplot2_pca', 'anndata_manipulate', 'cardinal_classification', 'bg_uniq', 'tp_find_and_replace', 'porechop', 'Paste1', 'ggplot_histogram', 'purge_dups', 'fastq_paired_end_interlacer', 'XY_Plot_1', 'rbc_mafft', 'random_lines1', 'unipept', 'query_tabular', 'search_gui', 'p_clip_peaks', 'targetfinder', 'ChangeCase', 'uniprot', 'mothur_phylotype', 'gops_merge_1', 'CONVERTER_fasta_to_tabular', 'annotatemyids', 'join1', 'qiime_pick_closed_reference_otus', 'salsa', 'tp_cut_tool', 'cshl_fastx_renamer', 'mass_spectrometry_imaging_combine', 'cshl_fastx_barcode_splitter', 'bedtools_intersectbed', 'cardinal_combine', '__FLATTEN__', 'minia', 'fastq_groomer', 'cd_hit', '_ensembl_gtf2gene_list', 'barchart_gnuplot', 'cshl_fastx_trimmer', 'cd_hit_est', 'fastq_to_fasta_python', 'quast', 'mergeCols1', 'umi_tools_extract', 'tp_sort_header_tool', 'msstatstmt', 'Grouping1', 'regionalgam_flight_curve', 'filter_by_fasta_ids', 'trinity', 'tab2fasta', 'Grep1', 'table_compute', 'regionalgam_ab_index', 'add_line_to_file', 'tp_replace_in_column', 'bedtools_getfastabed', 'datamash_ops', 'trimmomatic', 'tp_sorted_uniq', 'cardinal_spectra_plots', 'seq_filter_by_mapping', 'fasta2tab', 'fastq_to_tabular', 'heatmap2', 'ggplot2_heatmap2', 'rna_star', 'bwa_mem', 'Add_a_column1', 'fastqc', 'cutadapt', 'cshl_fastx_clipper', 'melt', 'checkFormat', 'rbc_mirdeep2_mapper', 'datamash_transpose', 'w4mclassfilter', 'stoceps_trend_indic', 'snpfreqplot', 'CONVERTER_interval_to_bedstrict_0', 'cshl_fastx_collapser', 'salmon', 'Convert characters1', 'fastq_paired_end_deinterlacer', 'ggplot2_point', 'volcanoplot', 'multiqc', 'Cut1', 'jbrowse', 'minimap2', 'column_order_header_sort', 'ete_homology_classifier', 'ggplot2_histogram', 'merqury', 'uc2otutable', 'Filter1', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'sort1', 'ctb_chemfp_mol2fps', 'bismark_bowtie2', 'bowtie2', 'rna_starsolo', 'addValue', 'cardinal_filtering', 'cast', 'proteomics_moff', 'busco', 'bedtools_sortbed', 'cshl_fastq_to_fasta', 'fastq_filter', 'hisat2', 'Summary_Statistics1', 'trim_galore']\n",
      "\n",
      "Test batch 1, Predicted top 121 tools: ['tp_easyjoin_tool', 'Grep1', 'datamash_ops', 'fastq_to_tabular', '__TAG_FROM_FILE__', 'hisat2', 'regexColumn1', 'fastq_groomer', 'filter_by_fasta_ids', 'bwa_mem', 'Add_a_column1', 'query_tabular', 'bedtools_intersectbed', 'tp_replace_in_column', 'trimmomatic', 'sort1', 'cshl_fastq_to_fasta', 'Paste1', 'join1', 'tp_cut_tool', 'add_line_to_file', 'Filter1', 'XY_Plot_1', 'barchart_gnuplot', 'rna_star', 'rbc_mafft', 'seq_filter_by_mapping', 'Cut1', 'unipept', 'join_files_on_column_fuzzy', 'porechop', 'ggplot_histogram', 'fastq_to_fasta_python', 'mergeCols1', 'Grouping1', 'tp_sorted_uniq', 'merqury', 'table_compute', 'random_lines1', 'ChangeCase', 'addValue', 'fastq_paired_end_interlacer', 'cutadapt', 'minimap2', 'bedtools_sortbed', 'fastq_filter', 'bedtools_getfastabed', '__RELABEL_FROM_FILE__', 'cardinal_filtering', 'volcanoplot', 'ggplot2_heatmap2', 'melt', 'Summary_Statistics1', 'annotatemyids', 'multiqc', 'w4mclassfilter', '__FLATTEN__', 'cshl_fastx_collapser', 'minia', 'cshl_fastx_renamer', 'column_order_header_sort', 'cd_hit_est', 'pureclip', 'checkFormat', 'umi_tools_extract', 'fastq_paired_end_deinterlacer', 'ggplot2_point', 'bg_uniq', 'trinity', 'datamash_transpose', 'targetfinder', 'uniprot', 'busco', 'bismark_bowtie2', 'mothur_phylotype', 'quast', 'cshl_fastx_clipper', 'CONVERTER_interval_to_bedstrict_0', 'tab2fasta', 'ete_homology_classifier', 'purge_dups', 'mass_spectrometry_imaging_combine', 'trim_galore', 'uclust2otutable', 'cardinal_classification', 'tp_sort_header_tool', 'uc2otutable', 'p_clip_peaks', 'cshl_fastx_barcode_splitter', 'msstatstmt', 'mothur_unique_seqs', 'anndata_manipulate', 'CONVERTER_fasta_to_tabular', 'fasta2tab', 'rna_starsolo', 'salsa', 'cshl_fastx_trimmer', 'tp_find_and_replace', 'rbc_mirdeep2_mapper', 'jbrowse', 'Convert characters1', 'gops_merge_1', 'qiime_pick_closed_reference_otus', 'cd_hit', '_ensembl_gtf2gene_list', 'stoceps_trend_indic', 'proteomics_moff', 'cast', 'ggplot2_histogram', 'cardinal_combine', 'regionalgam_flight_curve', 'ggplot2_pca', 'regionalgam_ab_index', 'ctb_chemfp_mol2fps', 'bowtie2', 'snpfreqplot', 'heatmap2', 'fastqc', 'search_gui', 'salmon', 'cardinal_spectra_plots']\n",
      "\n",
      "Test batch 1, Predicted top 121 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'bwa_mem', 'datamash_ops', 'gops_merge_1', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'trimmomatic', 'hisat2', 'multiqc', 'minimap2', 'fastq_to_fasta_python', 'trim_galore', 'datamash_transpose', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'fastq_groomer', 'bedtools_intersectbed', 'porechop', 'fasta2tab', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'salmon', 'busco', 'quast', 'volcanoplot', 'annotatemyids', 'fastq_filter', 'tp_replace_in_column', 'fastq_paired_end_interlacer', 'filter_by_fasta_ids', 'mothur_unique_seqs', 'cshl_fastq_to_fasta', 'rbc_mirdeep2_mapper', 'addValue', 'snpfreqplot', 'trinity', 'bedtools_getfastabed', 'bedtools_sortbed', 'cshl_fastx_trimmer', 'cd_hit', 'ggplot2_point', 'anndata_manipulate', 'cshl_fastx_renamer', 'jbrowse', 'regexColumn1', 'fastq_paired_end_deinterlacer', 'fastq_to_tabular', 'XY_Plot_1', 'cd_hit_est', 'rna_starsolo', 'query_tabular', 'bg_uniq', 'rbc_mafft', 'cshl_fastx_collapser', 'tp_sorted_uniq', 'cardinal_filtering', 'purge_dups', 'umi_tools_extract', 'cshl_fastx_barcode_splitter', 'seq_filter_by_mapping', 'mergeCols1', 'search_gui', 'merqury', 'bismark_bowtie2', 'cshl_fastx_clipper', '_ensembl_gtf2gene_list', 'ggplot2_pca', 'pureclip', 'cardinal_classification', 'salsa', 'join_files_on_column_fuzzy', 'add_line_to_file', 'cardinal_spectra_plots', 'minia', 'targetfinder', 'column_order_header_sort', 'cardinal_combine', 'uniprot', 'ggplot2_histogram', 'msstatstmt', 'ctb_chemfp_mol2fps', 'melt', 'qiime_pick_closed_reference_otus', 'checkFormat', 'cast', 'ete_homology_classifier', 'stoceps_trend_indic', 'regionalgam_ab_index', 'regionalgam_flight_curve', 'mothur_phylotype', 'tp_grep_tool', 'split_file_to_collection', 'goseq', 'fasterq_dump', 'tp_head_tool', 'tp_tail_tool', 'combine_metaphlan2_humann2', 'filter_tabular', 'gtftobed12', 'taxonomy_krona_chart', 'tp_replace_in_line', 'addName', 'stacks_denovomap', 'tp_split_on_column', 'featurecounts', 'export2graphlan', 'ncbi_acc_download', 'kraken2', 'fgsea', 'ncbi_blastdbcmd_wrapper', 'pathview', 'egsea', 'tp_cat', 'limma_voom', 'msstats', 'dropletutils_read_10x', 'cardinal_mz_images']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.14049586776859505\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.09917355371900827\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'fastqc', 'Grep1', 'collapse_dataset', 'join1', 'Cut1', 'Filter1', 'addValue', 'Grouping1', 'cat1', 'cast', 'comp1', 'wc_gnu', 'Remove beginning1']\n",
      "\n",
      "Test batch 1, True tools: ['datamash_ops', 'bg_column_arrange_by_header', 'wig_to_bigWig', 'tp_sorted_uniq', 'mass_spectrometry_imaging_filtering', 'comp1', 'bedtools_intersectbed', 'cardinal_spectra_plots', 'ggplot2_point', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'iReport', 'tp_split_on_column', 'tp_cut_tool', 'join_files_on_column_fuzzy', 'Cut1', 'regexColumn1', 'tp_cat', 'collection_column_join', 'gtftobed12', '__MERGE_COLLECTION__', 'edger', 'cat1', 'plotly_ml_performance_plots', 'cardinal_classification', 'climate_stripes', 'xarray_select', 'bg_uniq', 'rna_star', 'tp_replace_in_line', 'Filter1', 'Add_a_column1', 'fastqc', 'Show beginning1', 'deseq2', 'Paste1', 'pathview', 'plotly_regression_performance_plots', 'scanpy_filter_genes', 'csv_to_tabular', 'mergeCols1', 'cor2', '__TAG_FROM_FILE__', 'XY_Plot_1', 'datamash_transpose', 'random_lines1', 'tp_sort_header_tool', 'unipept', 'Grouping1', 'get_sequences', 'sort1', 'eggnog_mapper', 'tp_head_tool', 'tab2fasta', 'Grep1', 'chipseeker', 'addValue', 'bedtools_subtractbed', 'collapse_dataset', 'split_file_to_collection', 'join1', 'goseq', 'addName', 'bedtools_sortbed', 'chipeakanno_annopeaks', 'tp_replace_in_column', 'mass_spectrometry_imaging_ion_images', 'featurecounts', 'openbabel_filter', 'tp_sed_tool', 'Convert characters1', 'mass_spectrometry_imaging_qc', 'Summary_Statistics1', 'wc_gnu']\n",
      "\n",
      "Test batch 1, Predicted top 74 tools: ['xarray_select', 'eggnog_mapper', 'deseq2', 'mass_spectrometry_imaging_filtering', 'rna_star', 'datamash_transpose', 'chipeakanno_annopeaks', 'plotly_regression_performance_plots', '__TAG_FROM_FILE__', 'cardinal_classification', 'wig_to_bigWig', 'tp_replace_in_line', 'tp_sort_header_tool', 'Add_a_column1', 'scanpy_filter_genes', 'addName', 'mass_spectrometry_imaging_ion_images', 'mass_spectrometry_imaging_qc', '__MERGE_COLLECTION__', 'edger', 'plotly_ml_performance_plots', 'random_lines1', 'goseq', 'gtftobed12', 'climate_stripes', 'fastqc', 'cor2', 'get_sequences', 'ggplot2_point', 'iReport', 'addValue', 'Paste1', 'join1', 'tp_cut_tool', 'csv_to_tabular', 'openbabel_filter', 'Convert characters1', 'cat1', 'pathview', 'tp_head_tool', 'Summary_Statistics1', 'collection_column_join', 'bg_uniq', 'XY_Plot_1', 'unipept', 'tp_easyjoin_tool', 'bedtools_intersectbed', 'sort1', 'Filter1', 'bedtools_sortbed', 'join_files_on_column_fuzzy', 'Grep1', 'comp1', 'regexColumn1', 'tab2fasta', 'tp_cat', 'taxonomy_krona_chart', 'mergeCols1', 'featurecounts', 'cardinal_spectra_plots', 'bedtools_subtractbed', 'Show beginning1', 'chipseeker', 'tp_replace_in_column', 'Cut1', 'bg_column_arrange_by_header', 'collapse_dataset', 'wc_gnu', 'tp_sed_tool', 'tp_split_on_column', 'tp_sorted_uniq', 'split_file_to_collection', 'datamash_ops', 'Grouping1']\n",
      "\n",
      "Test batch 1, Predicted top 74 tools with weights: ['Add_a_column1', 'tp_replace_in_line', 'datamash_ops', 'fastqc', 'featurecounts', 'rna_star', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'tp_cat', 'tab2fasta', 'collapse_dataset', 'taxonomy_krona_chart', 'tp_replace_in_column', 'edger', 'goseq', 'collection_column_join', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'addValue', 'addName', 'bedtools_sortbed', 'ggplot2_point', 'regexColumn1', 'XY_Plot_1', 'eggnog_mapper', 'bg_uniq', 'tp_sorted_uniq', 'pathview', 'gtftobed12', 'xarray_select', 'chipseeker', 'bedtools_subtractbed', 'mergeCols1', 'tp_split_on_column', 'cardinal_classification', 'join_files_on_column_fuzzy', 'cardinal_spectra_plots', 'plotly_ml_performance_plots', 'chipeakanno_annopeaks', 'bg_column_arrange_by_header', 'scanpy_filter_genes', 'cor2', 'openbabel_filter', 'get_sequences', 'climate_stripes', 'mass_spectrometry_imaging_filtering', 'pygenomeTracks', 'filter_by_fasta_ids', 'tp_find_and_replace', 'filter_tabular', 'macs2_callpeak', 'query_tabular', 'stacks_denovomap', 'fasta2tab', 'regex_replace', 'export2graphlan', 'cutadapt', 'snpfreqplot', 'ncbi_blastdbcmd_wrapper', 'sam_merge2', 'sklearn_ensemble', 'bedtools_genomecoveragebed', 'deeptools_compute_matrix', 'plink', 'stringtie_merge', 'heinz_scoring', 'get_feature_info', 'egsea', 'heinz_bum', 'imagecoordinates_flipaxis']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.20270270270270271\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.08108108108108109\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'macs2_callpeak', 'diffbind', 'Filter1', 'chipseeker', 'samtools_rmdup', 'macs2_callpeak', 'macs2_callpeak', 'deeptools_bam_coverage', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'bedtools_intersectbed', 'bedtools_subtractbed', 'tp_sort_header_tool', 'tp_head_tool']\n",
      "\n",
      "Test batch 1, True tools: ['comp1', 'bedtools_intersectbed', 'tp_easyjoin_tool', 'circos', 'meme_chip', 'Cut1', 'tp_cat', 'barchart_gnuplot', 'Remove beginning1', 'cat1', 'cardinal_mz_images', 'Extract genomic DNA 1', 'Add_a_column1', 'Paste1', 'datamash_transpose', 'tab2fasta', 'sklearn_ensemble', 'tp_tail_tool', 'table_compute', 'addValue', 'collapse_dataset', 'tp_grep_tool', 'annotatemyids', 'join1', 'fasta_filter_by_length', 'openbabel_filter', 'tp_replace_in_column', 'mass_spectrometry_imaging_ion_images', 'xarray_mapplot', 'seq_filter_by_id', 'tp_cut_tool', 'Convert characters1', 'bedtools_getfastabed']\n",
      "\n",
      "Test batch 1, Predicted top 33 tools: ['datamash_transpose', 'Remove beginning1', 'tab2fasta', 'tp_cat', 'mass_spectrometry_imaging_ion_images', 'addValue', 'xarray_mapplot', 'Paste1', 'openbabel_filter', 'barchart_gnuplot', 'table_compute', 'tp_replace_in_column', 'cat1', 'collapse_dataset', 'Convert characters1', 'Add_a_column1', 'join1', 'sklearn_ensemble', 'Extract genomic DNA 1', 'meme_chip', 'comp1', 'tp_cut_tool', 'tp_tail_tool', 'circos', 'cardinal_mz_images', 'tp_easyjoin_tool', 'tp_grep_tool', 'annotatemyids', 'fasta_filter_by_length', 'seq_filter_by_id', 'bedtools_intersectbed', 'Cut1', 'bedtools_getfastabed']\n",
      "\n",
      "Test batch 1, Predicted top 33 tools with weights: ['Add_a_column1', 'datamash_transpose', 'tp_easyjoin_tool', 'bedtools_intersectbed', 'tp_cat', 'tab2fasta', 'table_compute', 'collapse_dataset', 'tp_grep_tool', 'circos', 'annotatemyids', 'tp_replace_in_column', 'addValue', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'fasta_filter_by_length', 'tp_tail_tool', 'meme_chip', 'cardinal_mz_images', 'xarray_mapplot', 'sklearn_ensemble', 'openbabel_filter', 'tp_find_and_replace', 'collection_column_join', 'tp_sort_header_tool', 'cutadapt', 'gtftobed12', 'fastqc', 'scatterplot_rpy', 'combine_metaphlan2_humann2', 'bowtie2', 'mergeCols1']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.12121212121212122\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.06060606060606061\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['normalize_readsizes', 'minimap2']\n",
      "\n",
      "Test batch 1, True tools: ['rseqc_bam_stat', 'medaka_consensus', 'htseq_count', 'bedtools_bamtobed', 'ococo', 'miniasm', 'rseqc_insertion_profile', 'medaka_variant', 'ngsutils_bam_filter', 'nanopolish_eventalign', 'rseqc_geneBody_coverage', 'purge_dups', 'picard_MarkDuplicates', 'bandage_image', 'rseqc_junction_annotation', 'pilon', 'bamFilter', 'ivar_trim', 'collapse_dataset', 'join1', 'bamtools_split', 'hypo', 'qualimap_rnaseq', 'picard_MergeSamFiles', 'bedtools_genomecoveragebed', 'rseqc_RNA_fragment_size', 'samtools_stats', 'racon', 'rseqc_FPKM_count', 'deeptools_bam_coverage', 'cufflinks', 'rseqc_read_distribution', 'bedtools_bamtofastq', 'nanopolish_variants', 'samtool_filter2', 'ivar_variants', 'samtools_view', 'samtools_markdup', 'featurecounts', 'samtools_depth', '__APPLY_RULES__', 'medaka_variant_pipeline', 'stringtie', 'samtools_sort', 'sam_merge2', 'bedtools_genomecoveragebed_histogram', 'rseqc_junction_saturation', 'deeptools_multi_bam_summary', 'samtools_slice_bam', 'fastqc', 'deeptools_plot_coverage', 'samtools_flagstat', 'ivar_consensus', 'jvarkit_wgscoverageplotter', 'pe_histogram', 'salmon', 'bam_to_sam', 'samtools_merge', 'bamtools', 'rseqc_clipping_profile', 'samtools_mpileup', 'sam_pileup', 'lofreq_call', 'Cut1', 'jbrowse', 'samtools_idxstats', 'rseqc_infer_experiment', 'tabular_to_fastq', 'rseqc_mismatch_profile', 'pycoqc', 'samtools_calmd', 'qualimap_bamqc', 'freebayes', 'samtools_fastx', 'nanoplot', 'CONVERTER_bam_to_bigwig_0', 'bedtools_coveragebed']\n",
      "\n",
      "Test batch 1, Predicted top 77 tools: ['qualimap_rnaseq', 'rseqc_RNA_fragment_size', 'rseqc_FPKM_count', 'deeptools_bam_coverage', 'ivar_consensus', 'samtools_merge', 'pycoqc', 'htseq_count', 'ococo', 'rseqc_geneBody_coverage', 'ivar_trim', 'samtools_markdup', 'rseqc_infer_experiment', 'rseqc_mismatch_profile', 'qualimap_bamqc', 'miniasm', 'rseqc_insertion_profile', 'ngsutils_bam_filter', 'bamFilter', 'hypo', 'samtools_stats', 'racon', 'cufflinks', 'nanopolish_variants', 'samtools_sort', 'fastqc', 'bam_to_sam', 'nanopolish_eventalign', 'samtools_view', '__APPLY_RULES__', 'medaka_variant_pipeline', 'stringtie', 'bedtools_genomecoveragebed_histogram', 'samtools_flagstat', 'jvarkit_wgscoverageplotter', 'rseqc_clipping_profile', 'freebayes', 'CONVERTER_bam_to_bigwig_0', 'rseqc_junction_annotation', 'bamtools_split', 'picard_MergeSamFiles', 'pilon', 'samtool_filter2', 'nanoplot', 'samtools_slice_bam', 'sam_pileup', 'rseqc_bam_stat', 'picard_MarkDuplicates', 'jbrowse', 'Cut1', 'bedtools_coveragebed', 'rseqc_read_distribution', 'samtools_depth', 'salmon', 'rseqc_junction_saturation', 'deeptools_multi_bam_summary', 'collapse_dataset', 'samtools_mpileup', 'medaka_consensus', 'bedtools_bamtobed', 'purge_dups', 'bedtools_bamtofastq', 'lofreq_call', 'bedtools_genomecoveragebed', 'ivar_variants', 'tabular_to_fastq', 'samtools_calmd', 'featurecounts', 'sam_merge2', 'bamtools', 'bandage_image', 'deeptools_plot_coverage', 'pe_histogram', 'join1', 'medaka_variant', 'samtools_idxstats', 'samtools_fastx']\n",
      "\n",
      "Test batch 1, Predicted top 77 tools with weights: ['lofreq_call', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'ivar_trim', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'bandage_image', 'ivar_consensus', 'bamFilter', 'rseqc_infer_experiment', 'collapse_dataset', 'freebayes', 'salmon', 'samtools_flagstat', 'samtools_idxstats', 'bedtools_bamtobed', 'ivar_variants', 'nanoplot', 'rseqc_read_distribution', 'ngsutils_bam_filter', 'bedtools_bamtofastq', 'samtools_sort', 'cufflinks', 'rseqc_geneBody_coverage', 'samtool_filter2', 'samtools_slice_bam', 'pe_histogram', 'medaka_variant', 'racon', 'medaka_consensus', 'bamtools', 'rseqc_bam_stat', 'sam_merge2', 'jbrowse', 'jvarkit_wgscoverageplotter', 'samtools_merge', 'samtools_markdup', 'deeptools_multi_bam_summary', 'samtools_depth', 'purge_dups', 'miniasm', 'tabular_to_fastq', 'picard_MergeSamFiles', 'bedtools_coveragebed', 'sam_pileup', 'deeptools_plot_coverage', 'qualimap_rnaseq', 'pycoqc', 'samtools_calmd', 'pilon', 'samtools_mpileup', 'ococo', 'bamtools_split', 'rseqc_FPKM_count', 'bam_to_sam', 'medaka_variant_pipeline', 'rseqc_junction_annotation', 'bedtools_genomecoveragebed_histogram', 'nanopolish_variants', 'hypo', 'rseqc_junction_saturation', 'rseqc_mismatch_profile', 'nanopolish_eventalign', 'rseqc_clipping_profile', 'rseqc_insertion_profile', 'rseqc_RNA_fragment_size', 'samtools_rmdup', 'rna_star', 'picard_CollectInsertSizeMetrics', 'macs2_callpeak', 'bwa']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.22077922077922077\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.2077922077922078\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'deeptools_bam_coverage', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'bedtools_intersectbed', 'Cut1', 'macs2_callpeak', 'diffbind', 'chipseeker', 'samtools_rmdup', 'sam_merge2', 'deeptools_bam_coverage', 'bowtie2']\n",
      "\n",
      "Test batch 1, True tools: ['velvetoptimiser', 'rseqc_bam_stat', 'gatk2_realigner_target_creator', 'deeptools_bamCompare', 'stacks2_refmap', 'get_read_pipeline', 'bedtools_multicovtbed', 'picard_AddOrReplaceReadGroups', 'htseq_count', '__MERGE_COLLECTION__', 'deeptools_alignmentsieve', 'bedtools_bamtobed', 'ococo', 'bamSplit', 'rgPicardMarkDups', 'kraken2', 'ngsutils_bam_filter', 'fastq_paired_end_interlacer', 'picard_MarkDuplicates', 'peakachu', 'basil', 'samtools_filter', 'collection_element_identifiers', 'cuffdiff', 'bwa', 'deeptools_compute_gc_bias', 'pilon', 'bamFilter', 'ivar_trim', 'tp_tail_tool', 'deeptools_plot_enrichment', 'picard_CollectInsertSizeMetrics', 'humann2', 'deeptools_plot_fingerprint', 'picard_MergeSamFiles', 'mageck_count', 'umi_tools_count', 'spades', 'tophat2', 'samtools_split', 'bundle_collection', 'bedtools_genomecoveragebed', 'deeptools_computeGCBias', 'bcftools_mpileup', 'deeptools_bamCoverage', 'bedtools_intersectbed', 'PicardASMetrics', 'samtools_stats', '__FLATTEN__', 'minia', 'macs2_callpeak', 'racon', 'picard_FilterSamReads', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'picard_ARRG', 'gatk4_mutect2', 'cufflinks', 'tp_sort_header_tool', 'Grouping1', 'naive_variant_caller', 'lofreq_indelqual', 'samtools_rmdup', 'deeptools_bamFingerprint', 'trinity', 'picard_SortSam', 'picard_artifact_metrics', 'samtool_filter2', 'ivar_variants', 'samtools_view', 'picard_QualityScoreDistribution', 'paralyzer', 'megahit', 'rnaspades', 'hicexplorer_hicbuildmatrix', 'featurecounts', 'picard_DownsampleSam', 'samtools_depth', 'bam_to_bigwig', 'je_markdupes', 'bioext_bam2msa', 'trimmomatic', 'lofreq_viterbi', 'seq_filter_by_mapping', 'stringtie', 'tp_cat', 'deeptools_bamCorrelate', 'samtools_sort', 'fastq_to_tabular', 'sam_merge2', 'deeptools_correct_gc_bias', 'bg_sortmerna', 'bam-filter', 'rna_star', 'bwa_mem', 'deeptools_multi_bam_summary', 'samtools_slice_bam', 'fastqc', 'peakcalling_macs', 'rbc_mirdeep2_mapper', 'samtools_flagstat', 'fastq_paired_end_joiner', 'ivar_consensus', 'metaphlan', 'picard_CASM', 'metaphlan2_hutlab', 'umi_tools_dedup', 'varscan', 'jvarkit_wgscoverageplotter', 'diffbind', 'picard_CleanSam', 'cufflinks_prok', 'pe_histogram', 'salmon', 'rseqc_inner_distance', 'bam_to_sam', 'macs2_predictd', 'samtools_merge', 'bamtools', 'samtools_mpileup', 'sam_pileup', 'lofreq_call', 'multiqc', 'jbrowse', 'deeptools_bam_compare', 'PicardInsertSize', 'cat1', 'mimodd_varcall', 'sam_to_bam', 'Extract genomic DNA 1', 'metaspades', 'samtools_idxstats', 'bamleftalign', 'rseqc_infer_experiment', 'mimodd_delcall', '__BUILD_LIST__', 'qualimap_bamqc', 'metaphlan2', 'unicycler', 'freebayes', 'samtools_fastx', 'bedtools_genomecoveragebed_bedgraph', 'cshl_fastq_to_fasta', 'CONVERTER_bam_to_bigwig_0', 'bedtools_coveragebed', 'hisat2']\n",
      "\n",
      "Test batch 1, Predicted top 146 tools: ['samtools_filter', 'pilon', 'humann2', 'umi_tools_count', 'tophat2', 'deeptools_computeGCBias', 'bcftools_mpileup', 'macs2_callpeak', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'picard_ARRG', 'trinity', 'picard_artifact_metrics', 'samtool_filter2', 'paralyzer', 'deeptools_bamCorrelate', 'fastq_to_tabular', 'deeptools_correct_gc_bias', 'rna_star', 'varscan', 'samtools_merge', 'samtools_mpileup', 'lofreq_call', 'multiqc', 'metaspades', 'metaphlan2', 'bedtools_genomecoveragebed_bedgraph', 'gatk2_realigner_target_creator', 'stacks2_refmap', 'htseq_count', 'peakachu', 'bwa', 'deeptools_compute_gc_bias', 'ivar_trim', 'picard_CollectInsertSizeMetrics', 'mageck_count', 'tp_sort_header_tool', 'picard_SortSam', 'rnaspades', 'featurecounts', 'samtools_depth', 'je_markdupes', 'tp_cat', 'bg_sortmerna', 'bwa_mem', 'samtools_slice_bam', 'fastq_paired_end_joiner', 'metaphlan', 'rseqc_inner_distance', 'sam_pileup', 'deeptools_bam_compare', 'rseqc_infer_experiment', 'qualimap_bamqc', 'rseqc_bam_stat', 'bedtools_multicovtbed', 'ngsutils_bam_filter', 'bamFilter', 'deeptools_plot_fingerprint', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', 'samtools_stats', '__FLATTEN__', 'gatk4_mutect2', 'deeptools_bamFingerprint', 'picard_QualityScoreDistribution', 'megahit', 'bam_to_bigwig', 'trimmomatic', 'lofreq_viterbi', 'seq_filter_by_mapping', 'samtools_sort', 'bam-filter', 'fastqc', 'rbc_mirdeep2_mapper', 'picard_CASM', 'metaphlan2_hutlab', 'picard_CleanSam', 'salmon', 'bam_to_sam', 'PicardInsertSize', 'unicycler', 'cshl_fastq_to_fasta', 'velvetoptimiser', 'bamSplit', 'fastq_paired_end_interlacer', 'samtools_view', 'stringtie', 'deeptools_multi_bam_summary', 'samtools_flagstat', 'jvarkit_wgscoverageplotter', 'diffbind', 'cufflinks_prok', 'cat1', 'mimodd_varcall', 'Extract genomic DNA 1', 'samtools_idxstats', 'bamleftalign', 'freebayes', 'CONVERTER_bam_to_bigwig_0', 'deeptools_bamCompare', 'get_read_pipeline', 'kraken2', 'collection_element_identifiers', 'spades', 'minia', 'bedtools_bamtobed', 'samtools_split', 'naive_variant_caller', 'macs2_predictd', 'cuffdiff', 'tp_tail_tool', 'PicardASMetrics', 'racon', 'sam_to_bam', 'mimodd_delcall', 'deeptools_bamCoverage', 'picard_AddOrReplaceReadGroups', 'sam_merge2', 'ivar_consensus', 'ococo', 'lofreq_indelqual', 'pe_histogram', '__MERGE_COLLECTION__', 'rgPicardMarkDups', 'deeptools_alignmentsieve', 'hisat2', 'deeptools_plot_enrichment', 'basil', 'jbrowse', 'bamtools', 'samtools_rmdup', 'hicexplorer_hicbuildmatrix', 'bioext_bam2msa', 'picard_DownsampleSam', 'bedtools_coveragebed', 'cufflinks', 'picard_FilterSamReads', 'bundle_collection', 'picard_MarkDuplicates', 'peakcalling_macs', 'picard_MergeSamFiles', 'ivar_variants', 'Grouping1', 'umi_tools_dedup', 'samtools_fastx', '__BUILD_LIST__']\n",
      "\n",
      "Test batch 1, Predicted top 146 tools with weights: ['lofreq_call', 'bwa_mem', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'ivar_trim', 'lofreq_indelqual', 'lofreq_viterbi', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'rna_star', 'trimmomatic', 'hisat2', 'multiqc', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'kraken2', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'samtools_rmdup', 'tp_cat', 'spades', 'ivar_consensus', 'bamFilter', 'rseqc_infer_experiment', 'unicycler', 'freebayes', 'salmon', 'samtools_flagstat', 'samtools_idxstats', 'bamleftalign', 'bedtools_bamtobed', 'metaphlan2', 'ivar_variants', 'fastq_paired_end_interlacer', 'ngsutils_bam_filter', 'bg_sortmerna', 'samtools_sort', 'cufflinks', 'cshl_fastq_to_fasta', 'samtool_filter2', 'rbc_mirdeep2_mapper', 'samtools_slice_bam', 'deeptools_bam_compare', 'pe_histogram', 'trinity', 'Extract genomic DNA 1', 'racon', 'bioext_bam2msa', 'tp_tail_tool', 'tophat2', 'bcftools_mpileup', 'bwa', 'bamtools', 'humann2', 'rseqc_bam_stat', 'picard_SortSam', 'sam_merge2', 'jbrowse', 'jvarkit_wgscoverageplotter', 'megahit', 'fastq_to_tabular', 'peakachu', 'samtools_merge', 'metaphlan', 'diffbind', 'picard_CollectInsertSizeMetrics', 'fastq_paired_end_joiner', 'deeptools_multi_bam_summary', 'deeptools_plot_fingerprint', 'samtools_depth', 'picard_CASM', 'cuffdiff', 'velvetoptimiser', 'picard_MergeSamFiles', 'picard_FilterSamReads', 'deeptools_alignmentsieve', 'metaspades', 'PicardASMetrics', 'bedtools_genomecoveragebed_bedgraph', 'mageck_count', 'bedtools_coveragebed', 'sam_pileup', 'seq_filter_by_mapping', 'umi_tools_dedup', 'naive_variant_caller', 'varscan', 'mimodd_varcall', 'collection_element_identifiers', 'pilon', 'samtools_mpileup', 'bedtools_multicovtbed', 'rnaspades', 'ococo', 'minia', 'hicexplorer_hicbuildmatrix', 'gatk4_mutect2', 'picard_DownsampleSam', 'bam_to_sam', 'samtools_split', 'basil', 'picard_artifact_metrics', 'stacks2_refmap', 'deeptools_plot_enrichment', 'picard_AddOrReplaceReadGroups', 'picard_QualityScoreDistribution', 'rseqc_inner_distance', 'paralyzer', 'picard_ARRG', 'get_read_pipeline', 'deeptools_bamCorrelate', 'rgPicardMarkDups', 'samtools_filter', 'bamSplit', 'mimodd_delcall', 'rseqc_read_distribution', 'qualimap_rnaseq', 'rseqc_geneBody_coverage', 'fastq_groomer', 'macs2_bdgpeakcall', 'idr', 'macs2_bdgcmp', 'samtools_fixmate', 'intervene_pairwise', 'intervene_upset', 'rseqc_read_duplication', 'pileometh', 'gemini_query', 'exomedepth', 'macs2_bdgdiff', 'bowtie2', 'picard_ReorderSam', 'deeptools_plot_coverage', 'rseqc_bam2wig', 'delly_call', 'bellerophon', 'picard_CollectGcBiasMetrics', 'control_freec', 'bedtools_genomecoveragebed_histogram', 'cutadapt', 'Add_a_column1']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.10273972602739725\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0958904109589041\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['collapse_dataset', 'porechop', 'cutadapt', 'minimap2', 'ngsutils_bam_filter', 'samtools_fastx', 'canu']\n",
      "\n",
      "Test batch 1, True tools: ['megablast_wrapper', 'khmer_abundance_distribution_single', 'fasta2tab', 'medaka_consensus', 'minimap2', 'racon', 'bwa_mem', 'fasta_merge_files_and_filter_unique_sequences', 'mummer_dnadiff', 'abricate', 'quast', 'purge_dups', 'fasta-stats', 'bandage_image', 'augustus', 'ncbi_blastn_wrapper', 'bowtie2', 'pilon', 'trinity', 'nanopolish_variants', 'prokka', 'blast_reciprocal_best_hits', 'jellyfish', 'spades', 'bandage_info', 'flye', 'medaka_consensus_pipeline']\n",
      "\n",
      "Test batch 1, Predicted top 27 tools: ['ncbi_blastn_wrapper', 'fasta2tab', 'pilon', 'bwa_mem', 'prokka', 'medaka_consensus_pipeline', 'racon', 'fasta_merge_files_and_filter_unique_sequences', 'khmer_abundance_distribution_single', 'jellyfish', 'quast', 'mummer_dnadiff', 'bandage_image', 'megablast_wrapper', 'bandage_info', 'abricate', 'purge_dups', 'augustus', 'bowtie2', 'medaka_consensus', 'spades', 'nanopolish_variants', 'minimap2', 'trinity', 'fasta-stats', 'blast_reciprocal_best_hits', 'flye']\n",
      "\n",
      "Test batch 1, Predicted top 27 tools with weights: ['bwa_mem', 'bowtie2', 'prokka', 'abricate', 'minimap2', 'ncbi_blastn_wrapper', 'fasta2tab', 'spades', 'bandage_image', 'quast', 'fasta-stats', 'fasta_merge_files_and_filter_unique_sequences', 'blast_reciprocal_best_hits', 'flye', 'trinity', 'racon', 'medaka_consensus', 'purge_dups', 'bandage_info', 'mummer_dnadiff', 'medaka_consensus_pipeline', 'augustus', 'pilon', 'jellyfish', 'megablast_wrapper', 'nanopolish_variants', 'khmer_abundance_distribution_single']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'hisat2', 'samtools_idxstats', 'Grep1', 'datamash_transpose', 'Remove beginning1', 'cat1', 'cast', 'comp1', 'addValue', 'Grouping1', 'join1', 'Add_a_column1', 'Filter1', 'wc_gnu', 'collapse_dataset', 'Cut1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'join_files_on_column_fuzzy', 'regexColumn1', 'collection_column_join', 'plotly_ml_performance_plots', 'bg_uniq', 'venn_list', 'scanpy_read_10x', 'tp_find_and_replace', 'bedtools_slopbed', 'Paste1', 'Show beginning1', 'ggplot_histogram', 'krona-text', 'stacks_denovomap', 'XY_Plot_1', 'random_lines1', 'unipept', 'query_tabular', 'ChangeCase', 'tp_tail_tool', 'collapse_dataset', 'annotatemyids', 'join1', 'goseq', 'heinz_scoring', 'tp_uniq_tool', 'export2graphlan', 'tp_cut_tool', 'seqtk_subseq', 'wc_gnu', 'bedtools_genomecoveragebed', 'wig_to_bigWig', 'clusterprofiler_bitr', 'bedtools_intersectbed', 'crossmap_bed', '__FLATTEN__', 'barchart_gnuplot', 'gtftobed12', 'Remove beginning1', 'macs2_callpeak', 'picard_FilterSamReads', 'pygenomeTracks', 'get_feature_info', 'bedtools_mergebed', 'ggplot_point', 'tp_replace_in_line', 'stringtie_merge', '__FILTER_FROM_FILE__', 'deseq2', 'mergeCols1', 'biotranslator', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'ggplot2_heatmap', 'tab2fasta', 'Grep1', 'msstats', 'table_compute', 'tp_grep_tool', 'egsea', 'add_line_to_file', 'fgsea', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'annotateMyIDs', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'goenrichment', 'regex1', 'tp_multijoin_tool', 'CONVERTER_interval_to_bed_0', 'gops_concat_1', 'bedtools_intersectbed_bam', 'tp_cat', 'charts', 'bcftools_consensus', 'fasterq_dump', 'ggplot2_heatmap2', 'rna_star', 'combine_metaphlan2_humann2', 'stacks_populations', 'Add_a_column1', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'w4mclassfilter', 'get_sequences', 'Kraken2Tax', 'cshl_find_and_replace', 'ont_fast5_api_fast5_subset', 'interactions', 'gops_coverage_1', 'heinz_bum', 'addName', 'mass_spectrometry_imaging_ion_images', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'Show tail1', 'trimmer', 'comp1', 'secure_hash_message_digest', 'volcanoplot', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'circos', 'gops_subtract_1', 'cat1', 'scatterplot_rpy', 'regex_replace', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'histogram_rpy', 'diff', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'interactive_tool_simtext_app', 'sklearn_ensemble', 'addValue', 'dexseq_count', 'cardinal_filtering', 'MassCalculator', 'split_file_to_collection', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'CONVERTER_bed_to_gff_0', 'Krona', 'bedtools_sortbed', 'param_value_from_file', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools: ['get_flanks1', 'join_files_on_column_fuzzy', 'ggplot_histogram', 'tp_uniq_tool', 'barchart_gnuplot', 'deseq2', 'mergeCols1', 'Grouping1', 'ggplot2_heatmap', 'annotateMyIDs', 'bedtools_getfastabed', 'tp_sorted_uniq', 'goenrichment', 'rna_star', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'gops_coverage_1', 'comp1', 'regex_replace', '__TAG_FROM_FILE__', 'sklearn_ensemble', 'dexseq_count', 'regexColumn1', 'scanpy_read_10x', 'krona-text', 'heinz_scoring', 'wig_to_bigWig', 'clusterprofiler_bitr', 'crossmap_bed', 'ggplot_point', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'tab2fasta', 'msstats', 'table_compute', 'featurecounts', 'tp_cat', 'Add_a_column1', 'w4mclassfilter', 'cshl_find_and_replace', 'addName', 'mass_spectrometry_imaging_ion_images', 'secure_hash_message_digest', 'MassCalculator', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'param_value_from_file', 'plotly_ml_performance_plots', 'stacks_denovomap', 'random_lines1', 'query_tabular', 'ChangeCase', 'goseq', 'seqtk_subseq', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', '__FLATTEN__', 'gtftobed12', 'picard_FilterSamReads', 'stringtie_merge', 'biotranslator', 'egsea', 'tp_replace_in_column', 'seq_filter_by_id', 'fasterq_dump', 'combine_metaphlan2_humann2', 'stacks_populations', 'get_sequences', 'Kraken2Tax', 'ont_fast5_api_fast5_subset', 'heinz_bum', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'sort1', 'addValue', 'ncbi_acc_download', 'venn_list', 'Paste1', 'join1', 'tp_cut_tool', 'get_feature_info', 'add_line_to_file', 'fgsea', 'regex1', 'bedtools_intersectbed_bam', 'interactions', 'Convert characters1', 'cat1', 'scatterplot_rpy', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'diff', 'interactive_tool_simtext_app', 'CONVERTER_bed_to_gff_0', 'Krona', 'cshl_word_list_grep', 'collection_column_join', 'bg_uniq', 'XY_Plot_1', 'unipept', 'annotatemyids', 'export2graphlan', 'Grep1', 'bcftools_consensus', 'tp_grep_tool', 'gops_join_1', 'tp_awk_tool', 'pygenomeTracks', 'trimmer', 'circos', 'tp_head_tool', 'bedtools_sortbed', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'Show beginning1', 'macs2_callpeak', 'tp_find_and_replace', 'tp_tail_tool', 'Remove beginning1', 'umi_tools_extract', 'hgv_david', 'volcanoplot', 'split_file_to_collection', 'bedtools_slopbed', 'Summary_Statistics1', 'collapse_dataset', 'Show tail1', '__FILTER_FROM_FILE__', 'wc_gnu', 'tp_sed_tool', 'tp_multijoin_tool', '__RELABEL_FROM_FILE__', 'datamash_ops', 'deeptools_compute_matrix', 'ggplot2_heatmap2', 'Count1', 'histogram_rpy', 'bg_column_arrange_by_header', 'CONVERTER_interval_to_bed_0', 'bedtools_mergebed', 'gops_subtract_1', 'cardinal_filtering', 'gops_concat_1', 'tp_split_on_column', 'charts']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'tp_replace_in_line', 'datamash_ops', 'bedtools_genomecoveragebed', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'rna_star', 'fasterq_dump', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'volcanoplot', 'tp_grep_tool', 'circos', 'tp_awk_tool', 'annotatemyids', 'taxonomy_krona_chart', 'tp_replace_in_column', 'filter_by_fasta_ids', 'goseq', 'gops_intersect_1', 'collection_column_join', 'limma_voom', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'Kraken2Tax', 'addValue', 'combine_metaphlan2_humann2', 'pygenomeTracks', 'addName', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'regexColumn1', 'seqtk_subseq', 'fgsea', 'XY_Plot_1', 'bedtools_mergebed', 'export2graphlan', 'goenrichment', 'query_tabular', 'bg_uniq', 'krona-text', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'gtftobed12', 'umi_tools_extract', 'stringtie_merge', 'dexseq_count', 'regex1', 'msstats', 'picard_FilterSamReads', 'ggplot2_heatmap', 'mergeCols1', 'tp_split_on_column', 'histogram_rpy', 'ncbi_blastdbcmd_wrapper', 'scatterplot_rpy', 'venn_list', 'egsea', 'join_files_on_column_fuzzy', 'add_line_to_file', 'bedtools_slopbed', 'dropletutils_read_10x', 'plotly_ml_performance_plots', 'get_flanks1', 'tp_uniq_tool', 'regex_replace', 'bg_column_arrange_by_header', 'sklearn_ensemble', 'stacks_denovomap', 'diff', 'gops_coverage_1', 'get_feature_info', 'checkFormat', 'get_sequences', 'heinz_scoring', 'heinz_bum', 'MassCalculator', 'gops_merge_1', 'macs2_bdgdiff', 'uniprot', 'edger', 'chipeakanno_annopeaks', 'xarray_select', 'deg_annotate', 'metilene', 'tabular_to_fastq', 'filter_tabular', 'bedtools_subtractbed', 'raceid_filtnormconf', 'htseq_count', 'tables_arithmetic_operations', 'chipseeker', 'Heatmap', 'cardinal_spectra_plots', 'fastq_groomer', 'column_remove_by_header', 'plink', 'cummeRbund', 'scater_create_qcmetric_ready_sce', 'macs2_bdgpeakcall', 'humann2', 'tbl2gff3', 'datamash_reverse', 'column_order_header_sort', 'multiqc', 'melt', 'macs2_bdgcmp', 'hicexplorer_hicfindrestrictionsites', 'vegan_diversity', 'cor2', 'get_pdb', 'graphprot_predict_profile', 'gops_cluster_1', 'sklearn_feature_selection', 'bedtools_multiintersectbed', 'sam_merge2', 'idr', 'stoceps_trend_indic', 'intervene_upset', 'intervene_pairwise', 'climate_stripes', 'vegan_rarefaction', 'samtools_idxstats', 'ctb_chemfp_mol2fps', 'rseqc_infer_experiment', 'Fetch Taxonomic Ranks', 'subtract_query1', 'sklearn_svm_classifier', 'flair_correct', 'transit_gumbel', 'cardinal_classification', 'kc-align']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3092105263157895\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.21710526315789475\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'samtools_rmdup', 'sam_merge2', 'deeptools_bam_coverage', 'bowtie2', 'macs2_callpeak', 'join1', 'Add_a_column1', 'Cut1', 'macs2_callpeak', 'diffbind', 'chipseeker', 'Remove beginning1', 'bedtools_intersectbed', 'bedtools_subtractbed', 'tp_sort_header_tool', 'tp_head_tool', 'Extract genomic DNA 1']\n",
      "\n",
      "Test batch 1, True tools: ['meme_dreme', 'meme_chip', 'fasta2tab', 'clustalw', 'deeptools_bam_compare', 'macs2_callpeak', 'EMBOSS: transeq101', 'meme_meme', 'translate_bed_sequences', 'Filter1', 'fasta_merge_files_and_filter_unique_sequences', 'meme_fimo', 'tp_awk_tool', 'ncbi_blastn_wrapper', 'rgweblogo3', 'maker', 'fasta_filter_by_length', 'sed_stream_editor', 'fasta_compute_length', 'viennarna_rnafold', 'graphprot_predict_profile']\n",
      "\n",
      "Test batch 1, Predicted top 21 tools: ['translate_bed_sequences', 'meme_fimo', 'sed_stream_editor', 'fasta_merge_files_and_filter_unique_sequences', 'meme_dreme', 'maker', 'meme_chip', 'meme_meme', 'viennarna_rnafold', 'graphprot_predict_profile', 'deeptools_bam_compare', 'ncbi_blastn_wrapper', 'rgweblogo3', 'EMBOSS: transeq101', 'fasta2tab', 'clustalw', 'fasta_compute_length', 'fasta_filter_by_length', 'Filter1', 'tp_awk_tool', 'macs2_callpeak']\n",
      "\n",
      "Test batch 1, Predicted top 21 tools with weights: ['ncbi_blastn_wrapper', 'macs2_callpeak', 'fasta2tab', 'tp_awk_tool', 'fasta_merge_files_and_filter_unique_sequences', 'deeptools_bam_compare', 'fasta_filter_by_length', 'clustalw', 'fasta_compute_length', 'meme_meme', 'meme_chip', 'meme_fimo', 'viennarna_rnafold', 'maker', 'meme_dreme', 'rgweblogo3', 'graphprot_predict_profile', 'sed_stream_editor', 'translate_bed_sequences', 'intarna', 'hyphy_slac']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.09523809523809523\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.09523809523809523\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['macs2_callpeak', 'bedtools_multiintersectbed', 'Filter1', 'Cut1', 'bedtools_sortbed']\n",
      "\n",
      "Test batch 1, True tools: ['wig_to_bigWig', 'bedtools_intersectbed', 'Cut1', 'bedtools_bed12tobed6', 'bedtools_genomecoveragebed', 'pygenomeTracks', 'bedtools_mergebedgraph', 'bedtools_map', 'bedtools_mergebed', 'bedtools_multiintersectbed', 'Add_a_column1', 'bedtools_complementbed', 'deeptools_compute_matrix', 'tp_awk_tool', 'bedtools_spacingbed', 'bedtools_bedtobam', 'addValue', 'bcftools_view', 'bedtools_genomecoveragebed_bedgraph', 'wc_gnu']\n",
      "\n",
      "Test batch 1, Predicted top 20 tools: ['wig_to_bigWig', 'bedtools_bedtobam', 'Add_a_column1', 'deeptools_compute_matrix', 'bedtools_map', 'bedtools_bed12tobed6', 'pygenomeTracks', 'bedtools_mergebedgraph', 'bcftools_view', 'bedtools_spacingbed', 'bedtools_complementbed', 'bedtools_genomecoveragebed_bedgraph', 'bedtools_mergebed', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', 'bedtools_multiintersectbed', 'wc_gnu', 'addValue', 'Cut1', 'tp_awk_tool']\n",
      "\n",
      "Test batch 1, Predicted top 20 tools with weights: ['Add_a_column1', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', 'deeptools_compute_matrix', 'tp_awk_tool', 'addValue', 'pygenomeTracks', 'bedtools_mergebed', 'bedtools_genomecoveragebed_bedgraph', 'bedtools_multiintersectbed', 'bedtools_bedtobam', 'bedtools_map', 'bcftools_view', 'bedtools_mergebedgraph', 'bedtools_bed12tobed6', 'bedtools_spacingbed', 'tp_sort_header_tool', 'wsdl_hmdb', 'jbrowse', 'bedtools_sortbed']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.2\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.15\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['filter_tabular', 'Cut1', 'param_value_from_file', 'bioext_bam2msa', 'fastq_combiner', 'fastq_paired_end_joiner', 'fastq_to_fasta_python', 'filter_by_fasta_ids', 'translate_nucleotides']\n",
      "\n",
      "Test batch 1, True tools: ['ncbi_tblastn_wrapper', 'tp_find_and_replace', 'fasta2tab', 'muscle', 'rgweblogo3']\n",
      "\n",
      "Test batch 1, Predicted top 5 tools: ['muscle', 'tp_find_and_replace', 'ncbi_tblastn_wrapper', 'rgweblogo3', 'fasta2tab']\n",
      "\n",
      "Test batch 1, Predicted top 5 tools with weights: ['tp_find_and_replace', 'fasta2tab', 'ncbi_tblastn_wrapper', 'muscle', 'rgweblogo3']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'macs2_callpeak', 'diffbind', 'Remove beginning1', 'bedtools_intersectbed', 'chipseeker', 'samtools_rmdup', 'sam_merge2', 'deeptools_bam_coverage', 'bowtie2', 'macs2_callpeak', 'Cut1', 'Filter1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'pampa_communitymetrics', 'tp_easyjoin_tool', 'join_files_on_column_fuzzy', 'htseq_count', 'regexColumn1', 'snpSift_filter', 'hgv_snpFreq', 'collection_column_join', 'seq_select_by_id', 'gprofiler_gost', 'bg_uniq', 'venn_list', 'tp_find_and_replace', 'Show beginning1', 'XY_Plot_1', 'changeGffFeatures', 'random_lines1', 'unipept', 'cshl_sort_header', 'ChangeCase', 'uniprot', 'aggregate_scores_in_intervals2', 'tp_tail_tool', 'bedtools_subtractbed', 'collapse_dataset', 'annotatemyids', 'join1', 'deg_annotate', 'tp_cut_tool', 'wc_gnu', 'bedtools_intersectbed', 'cardinal_combine', 'repeat_counter', 'transpose', 'dexseq_annotate', 'Remove beginning1', 'cardinal_mz_images', 'column_remove_by_header', 'bedtools_mergebed', 'tp_replace_in_line', 'deseq2', 'mergeCols1', 'vegan_rarefaction', 'Interval_Maf_Merged_Fasta2', 'tp_sort_header_tool', 'Grouping1', 'pandas_rolling_window', 'naive_variant_caller', 'gops_intersect_1', 'samtools_rmdup', 'Fetch Taxonomic Ranks', 'ggplot2_heatmap', 'trinity', 'tab2fasta', 'Grep1', 'SnpEff-cds-report', 'samtool_filter2', 'table_compute', 'samtools_view', 'tabular_to_csv', 'table_annovar', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'cooler_csort_pairix', '__FILTER_EMPTY_DATASETS__', 'gops_cluster_1', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'filter_tabular', 'goenrichment', 'snpEff', 'cardinal_spectra_plots', 'gdal_ogr2ogr', 'tp_multijoin_tool', 'gops_concat_1', 'tp_cat', 'charts', 'bcftools_consensus', 'vcfbedintersect', 'sam_merge2', 'ggplot2_heatmap2', 'Add_a_column1', 'stoceps_glm', 'datamash_transpose', 'stoceps_trend_indic', 'gops_coverage_1', 'chipseeker', 'snpSift_rmInfo', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'diapysef', 'Show tail1', 'trimmer', 'comp1', 'ggplot2_point', 'volcanoplot', 'multiqc', 'tbl2gff3', 'Cut1', 'CONVERTER_vcf_bgzip_to_tabix_0', 'jbrowse', 'gops_subtract_1', 'minimap2', 'cat1', 'Extract genomic DNA 1', 'tab2csv_R', 'vegan_diversity', 'pathview', 'pampa_glmsp', 'histogram_rpy', '__BUILD_LIST__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'addValue', 'bcftools_norm', 'dexseq_count', 'cardinal_filtering', 'split_file_to_collection', '__EXTRACT_DATASET__', 'samtools_fastx', 'bedtools_sortbed', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'bedtools_maskfastabed', 'fastq_filter', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 135 tools: ['tp_easyjoin_tool', 'join_files_on_column_fuzzy', 'Show beginning1', 'mergeCols1', 'Grouping1', 'samtools_rmdup', 'samtool_filter2', 'table_annovar', 'datamash_ops', 'goenrichment', 'tp_multijoin_tool', 'sam_merge2', 'datamash_transpose', 'snpSift_rmInfo', 'bg_column_arrange_by_header', 'htseq_count', 'tp_find_and_replace', 'changeGffFeatures', 'tp_tail_tool', 'cardinal_combine', 'cardinal_mz_images', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'featurecounts', 'charts', 'bcftools_consensus', 'Add_a_column1', 'chipseeker', 'random_lines1', 'ChangeCase', 'bedtools_intersectbed', 'vcfbedintersect', 'ggplot2_heatmap2', 'stoceps_trend_indic', 'jbrowse', 'addValue', 'bcftools_norm', 'gprofiler_gost', 'bedtools_subtractbed', 'join1', 'tp_cut_tool', 'gops_concat_1', 'tp_sed_tool', 'Convert characters1', 'Cut1', 'cat1', 'bedtools_sortbed', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'Summary_Statistics1', 'collection_column_join', 'XY_Plot_1', 'unipept', 'uniprot', 'annotatemyids', '__FILTER_EMPTY_DATASETS__', 'gops_cluster_1', 'tp_sorted_uniq', 'gops_coverage_1', 'gops_subtract_1', 'dexseq_count', 'cardinal_filtering', 'bedtools_maskfastabed', 'regexColumn1', 'transpose', 'SnpEff-cds-report', 'snpEff', 'Fetch Taxonomic Ranks', 'gops_join_1', 'sort1', 'split_file_to_collection', 'venn_list', 'column_remove_by_header', 'bedtools_mergebed', 'cshl_word_list_grep', 'bg_uniq', 'cshl_sort_header', 'repeat_counter', 'dexseq_annotate', 'pandas_rolling_window', 'naive_variant_caller', 'trinity', 'Grep1', 'bedtools_getfastabed', 'stoceps_glm', 'hgv_snpFreq', 'hgv_david', 'tab2csv_R', 'samtools_fastx', 'deg_annotate', 'tp_replace_in_column', 'cooler_csort_pairix', 'diapysef', 'CONVERTER_vcf_bgzip_to_tabix_0', 'Interval_Maf_Merged_Fasta2', 'samtools_view', 'gdal_ogr2ogr', 'minimap2', 'pathview', 'fastq_filter', 'snpSift_filter', 'seq_select_by_id', 'collapse_dataset', 'aggregate_scores_in_intervals2', 'deseq2', 'tbl2gff3', 'tp_awk_tool', 'ncbi_acc_download', 'tabular_to_csv', 'comp1', 'histogram_rpy', 'cardinal_spectra_plots', 'Show tail1', 'tp_cat', 'vegan_diversity', 'filter_tabular', 'volcanoplot', 'vegan_rarefaction', 'wc_gnu', 'ggplot2_heatmap', 'ggplot2_point', 'pampa_glmsp', 'tp_head_tool', 'pampa_communitymetrics', 'seq_filter_by_id', 'multiqc', 'table_compute', 'Remove beginning1', '__EXTRACT_DATASET__', 'deeptools_compute_matrix', 'Count1', '__BUILD_LIST__', 'Extract genomic DNA 1', 'trimmer', 'tab2fasta']\n",
      "\n",
      "Test batch 1, Predicted top 135 tools with weights: ['Add_a_column1', 'snpSift_filter', 'tp_find_and_replace', 'tp_replace_in_line', 'samtools_view', 'datamash_ops', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'filter_tabular', 'multiqc', 'minimap2', 'htseq_count', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'samtools_rmdup', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'snpEff', 'volcanoplot', 'tp_awk_tool', 'annotatemyids', 'fastq_filter', 'deg_annotate', 'tp_replace_in_column', 'gops_intersect_1', 'collection_column_join', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'samtool_filter2', 'addValue', 'bedtools_maskfastabed', 'trinity', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'sam_merge2', 'jbrowse', 'regexColumn1', 'XY_Plot_1', 'bedtools_mergebed', 'goenrichment', 'bcftools_norm', 'bg_uniq', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'tbl2gff3', 'dexseq_count', 'chipseeker', 'bedtools_subtractbed', 'ggplot2_heatmap', 'mergeCols1', 'histogram_rpy', 'naive_variant_caller', 'venn_list', 'cardinal_mz_images', 'join_files_on_column_fuzzy', 'cardinal_spectra_plots', 'gprofiler_gost', 'vcfbedintersect', 'column_remove_by_header', 'bg_column_arrange_by_header', 'cardinal_combine', 'uniprot', 'vegan_diversity', 'SnpEff-cds-report', 'gops_coverage_1', 'diapysef', 'Fetch Taxonomic Ranks', 'vegan_rarefaction', 'pampa_glmsp', 'stoceps_trend_indic', 'stoceps_glm', 'gops_cluster_1', 'fastqc', 'samtools_slice_bam', 'gtftobed12', 'cutadapt', 'addName', 'filter_by_fasta_ids', 'picard_MarkDuplicates', 'tp_split_on_column', 'rna_star', 'plotly_ml_performance_plots', 'xarray_select', 'goseq', 'circos_wiggle_to_stacked', 'cardinal_classification', 'picard_FilterSamReads', 'eggnog_mapper', 'cummeRbund', 'chipeakanno_annopeaks', 'rcas', 'snpfreqplot', 'gops_merge_1', 'bedtools_windowbed', 'gemini_query', 'fgsea', 'blockbuster', 'metaspades', 'samtools_depth', 'qualimap_bamqc', 'egsea', 'macs2_bdgdiff', 'macs2_callpeak', 'taxonomy_krona_chart', 'snpSift_extractFields', 'ncbi_blastdbcmd_wrapper', 'diffbind', 'ctb_multi_obgrep', 'bedtools_makewindowsbed', 'get_sequences', 'bam_to_sam', 'msstats', 'heinz_bum', 'kraken2', 'heinz_scoring', 'transit_gumbel', 'rseqc_infer_experiment', 'tp_sort_rows', 'samtools_idxstats']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.1925925925925926\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.1111111111111111\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'bowtie2', 'macs2_callpeak', 'chipseeker', 'Remove beginning1', 'bedtools_subtractbed', 'tp_sort_header_tool', 'join1', 'Add_a_column1', 'Cut1', 'Extract genomic DNA 1', 'macs2_callpeak', 'samtools_rmdup', 'fastqc']\n",
      "\n",
      "Test batch 1, True tools: ['varscan_somatic', 'velvetoptimiser', 'genomespace_exporter', 'ggplot2_point', 'snpEff', 'volcanoplot', 'CONVERTER_gz_to_uncompressed', 'tp_multijoin_tool', 'multiqc', '__FLATTEN__', 'Cut1', 'stringtie', 'regexColumn1', 'tp_cat', '__MERGE_COLLECTION__', 'cshl_fastx_trimmer', 'bcftools_consensus', 'snpSift_annotate', 'fastq_to_fasta_python', 'tp_find_and_replace', 'rna_star', 'bwa_mem', 'stringtie_merge', 'Filter1', 'Add_a_column1', 'deseq2', 'cutadapt', 'bamleftalign', 'diff', 'bwa', '__UNZIP_COLLECTION__', 'export_remote', 'samtools_calmd', 'bowtie2', 'bamFilter', 'Grep1', 'unicycler', 'ivar_variants', 'freebayes', 'gemini_annotate', 'join1', 'deg_annotate', 'picard_NormalizeFasta', 'featurecounts', 'NFS_transfer', 'hisat2', 'create_opson_pipeline_summary']\n",
      "\n",
      "Test batch 1, Predicted top 47 tools: ['genomespace_exporter', '__UNZIP_COLLECTION__', 'cshl_fastx_trimmer', 'create_opson_pipeline_summary', 'samtools_calmd', 'tp_find_and_replace', 'bwa', 'featurecounts', 'bwa_mem', 'Add_a_column1', 'volcanoplot', 'snpSift_annotate', 'picard_NormalizeFasta', 'gemini_annotate', 'velvetoptimiser', 'join1', 'varscan_somatic', 'cutadapt', 'Filter1', 'bamleftalign', 'diff', 'CONVERTER_gz_to_uncompressed', 'deseq2', 'rna_star', 'hisat2', '__MERGE_COLLECTION__', 'bamFilter', 'deg_annotate', 'stringtie_merge', 'bowtie2', 'stringtie', 'freebayes', 'multiqc', 'unicycler', 'export_remote', 'ivar_variants', 'NFS_transfer', 'Grep1', 'Cut1', '__FLATTEN__', 'tp_cat', 'tp_multijoin_tool', 'ggplot2_point', 'fastq_to_fasta_python', 'bcftools_consensus', 'snpEff', 'regexColumn1']\n",
      "\n",
      "Test batch 1, Predicted top 47 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'bwa_mem', 'bcftools_consensus', 'featurecounts', 'rna_star', 'cutadapt', 'bowtie2', 'hisat2', 'multiqc', 'stringtie', 'fastq_to_fasta_python', 'deseq2', 'tp_cat', 'bamFilter', 'unicycler', 'freebayes', 'snpEff', 'bamleftalign', 'volcanoplot', 'deg_annotate', 'ivar_variants', 'cshl_fastx_trimmer', 'bwa', 'ggplot2_point', 'regexColumn1', 'stringtie_merge', 'velvetoptimiser', 'picard_NormalizeFasta', 'samtools_calmd', 'snpSift_annotate', 'diff', 'ncbi_blastn_wrapper', 'fastqc', 'jbrowse', 'collapse_dataset', 'tp_tail_tool', 'bedtools_intersectbed', 'tp_sort_header_tool', 'samtools_mpileup', 'ecology_presence_abs_abund', 'tp_text_file_with_recurring_lines', 'gemini_query', 'kraken2', 'samtools_idxstats', 'ggplot2_pca', 'datamash_transpose']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0425531914893617\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.02127659574468085\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'macs2_callpeak', 'diffbind', 'chipseeker', 'Remove beginning1', 'bedtools_intersectbed', 'bedtools_intersectbed', 'Cut1', 'Filter1', 'Extract genomic DNA 1', 'macs2_callpeak', 'bowtie2']\n",
      "\n",
      "Test batch 1, True tools: ['velvetoptimiser', 'rseqc_bam_stat', 'gatk2_realigner_target_creator', 'deeptools_bamCompare', 'stacks2_refmap', 'get_read_pipeline', 'bedtools_multicovtbed', 'picard_AddOrReplaceReadGroups', 'htseq_count', '__MERGE_COLLECTION__', 'deeptools_alignmentsieve', 'bedtools_bamtobed', 'ococo', 'bamSplit', 'rgPicardMarkDups', 'kraken2', 'ngsutils_bam_filter', 'fastq_paired_end_interlacer', 'picard_MarkDuplicates', 'peakachu', 'basil', 'samtools_filter', 'collection_element_identifiers', 'cuffdiff', 'bwa', 'deeptools_compute_gc_bias', 'pilon', 'bamFilter', 'ivar_trim', 'tp_tail_tool', 'deeptools_plot_enrichment', 'picard_CollectInsertSizeMetrics', 'humann2', 'deeptools_plot_fingerprint', 'picard_MergeSamFiles', 'mageck_count', 'umi_tools_count', 'spades', 'tophat2', 'samtools_split', 'bundle_collection', 'bedtools_genomecoveragebed', 'deeptools_computeGCBias', 'bcftools_mpileup', 'deeptools_bamCoverage', 'bedtools_intersectbed', 'PicardASMetrics', 'samtools_stats', '__FLATTEN__', 'minia', 'macs2_callpeak', 'racon', 'picard_FilterSamReads', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'picard_ARRG', 'gatk4_mutect2', 'cufflinks', 'tp_sort_header_tool', 'Grouping1', 'naive_variant_caller', 'lofreq_indelqual', 'samtools_rmdup', 'deeptools_bamFingerprint', 'trinity', 'picard_SortSam', 'picard_artifact_metrics', 'samtool_filter2', 'ivar_variants', 'samtools_view', 'picard_QualityScoreDistribution', 'paralyzer', 'megahit', 'rnaspades', 'hicexplorer_hicbuildmatrix', 'featurecounts', 'picard_DownsampleSam', 'samtools_depth', 'bam_to_bigwig', 'je_markdupes', 'bioext_bam2msa', 'trimmomatic', 'lofreq_viterbi', 'seq_filter_by_mapping', 'stringtie', 'tp_cat', 'deeptools_bamCorrelate', 'samtools_sort', 'fastq_to_tabular', 'sam_merge2', 'deeptools_correct_gc_bias', 'bg_sortmerna', 'bam-filter', 'rna_star', 'bwa_mem', 'deeptools_multi_bam_summary', 'samtools_slice_bam', 'fastqc', 'peakcalling_macs', 'rbc_mirdeep2_mapper', 'samtools_flagstat', 'fastq_paired_end_joiner', 'ivar_consensus', 'metaphlan', 'picard_CASM', 'metaphlan2_hutlab', 'umi_tools_dedup', 'varscan', 'jvarkit_wgscoverageplotter', 'diffbind', 'picard_CleanSam', 'cufflinks_prok', 'pe_histogram', 'salmon', 'rseqc_inner_distance', 'bam_to_sam', 'macs2_predictd', 'samtools_merge', 'bamtools', 'samtools_mpileup', 'sam_pileup', 'lofreq_call', 'multiqc', 'jbrowse', 'deeptools_bam_compare', 'PicardInsertSize', 'cat1', 'mimodd_varcall', 'sam_to_bam', 'Extract genomic DNA 1', 'metaspades', 'samtools_idxstats', 'bamleftalign', 'rseqc_infer_experiment', 'mimodd_delcall', '__BUILD_LIST__', 'qualimap_bamqc', 'metaphlan2', 'unicycler', 'freebayes', 'samtools_fastx', 'bedtools_genomecoveragebed_bedgraph', 'cshl_fastq_to_fasta', 'CONVERTER_bam_to_bigwig_0', 'bedtools_coveragebed', 'hisat2']\n",
      "\n",
      "Test batch 1, Predicted top 146 tools: ['bedtools_bamtobed', 'samtools_filter', 'pilon', 'humann2', 'umi_tools_count', 'tophat2', 'samtools_split', 'deeptools_computeGCBias', 'bcftools_mpileup', 'macs2_callpeak', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'picard_ARRG', 'trinity', 'picard_artifact_metrics', 'samtool_filter2', 'paralyzer', 'deeptools_bamCorrelate', 'fastq_to_tabular', 'sam_merge2', 'deeptools_correct_gc_bias', 'rna_star', 'varscan', 'samtools_merge', 'samtools_mpileup', 'lofreq_call', 'multiqc', 'metaspades', 'metaphlan2', 'bedtools_genomecoveragebed_bedgraph', 'gatk2_realigner_target_creator', 'stacks2_refmap', 'htseq_count', 'ococo', 'peakachu', 'cuffdiff', 'bwa', 'deeptools_compute_gc_bias', 'ivar_trim', 'picard_CollectInsertSizeMetrics', 'mageck_count', 'tp_sort_header_tool', 'picard_SortSam', 'rnaspades', 'featurecounts', 'samtools_depth', 'je_markdupes', 'tp_cat', 'bg_sortmerna', 'bwa_mem', 'samtools_slice_bam', 'fastq_paired_end_joiner', 'metaphlan', 'pe_histogram', 'rseqc_inner_distance', 'sam_pileup', 'deeptools_bam_compare', 'rseqc_infer_experiment', 'qualimap_bamqc', 'rseqc_bam_stat', 'bedtools_multicovtbed', '__MERGE_COLLECTION__', 'ngsutils_bam_filter', 'bamFilter', 'deeptools_plot_fingerprint', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', 'PicardASMetrics', 'samtools_stats', '__FLATTEN__', 'racon', 'gatk4_mutect2', 'deeptools_bamFingerprint', 'picard_QualityScoreDistribution', 'megahit', 'bam_to_bigwig', 'trimmomatic', 'lofreq_viterbi', 'seq_filter_by_mapping', 'samtools_sort', 'bam-filter', 'fastqc', 'rbc_mirdeep2_mapper', 'picard_CASM', 'metaphlan2_hutlab', 'picard_CleanSam', 'salmon', 'bam_to_sam', 'PicardInsertSize', 'unicycler', 'cshl_fastq_to_fasta', 'velvetoptimiser', 'bamSplit', 'fastq_paired_end_interlacer', 'samtools_view', 'stringtie', 'deeptools_multi_bam_summary', 'samtools_flagstat', 'jvarkit_wgscoverageplotter', 'diffbind', 'cufflinks_prok', 'cat1', 'mimodd_varcall', 'Extract genomic DNA 1', 'samtools_idxstats', 'bamleftalign', 'freebayes', 'CONVERTER_bam_to_bigwig_0', 'deeptools_bamCompare', 'get_read_pipeline', 'picard_AddOrReplaceReadGroups', 'kraken2', 'collection_element_identifiers', 'spades', 'minia', 'ivar_consensus', 'hisat2', 'tp_tail_tool', 'lofreq_indelqual', 'mimodd_delcall', 'deeptools_bamCoverage', 'deeptools_alignmentsieve', 'naive_variant_caller', 'samtools_rmdup', 'macs2_predictd', 'deeptools_plot_enrichment', 'sam_to_bam', 'rgPicardMarkDups', 'picard_MarkDuplicates', 'hicexplorer_hicbuildmatrix', 'basil', 'cufflinks', 'bioext_bam2msa', 'picard_FilterSamReads', 'jbrowse', 'picard_DownsampleSam', 'bedtools_coveragebed', 'picard_MergeSamFiles', 'bundle_collection', 'bamtools', 'peakcalling_macs', 'Grouping1', 'ivar_variants', 'umi_tools_dedup', '__BUILD_LIST__', 'samtools_fastx']\n",
      "\n",
      "Test batch 1, Predicted top 146 tools with weights: ['lofreq_call', 'bwa_mem', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'ivar_trim', 'lofreq_indelqual', 'lofreq_viterbi', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'rna_star', 'trimmomatic', 'hisat2', 'multiqc', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'kraken2', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'samtools_rmdup', 'tp_cat', 'spades', 'ivar_consensus', 'bamFilter', 'rseqc_infer_experiment', 'unicycler', 'freebayes', 'salmon', 'samtools_flagstat', 'samtools_idxstats', 'bamleftalign', 'bedtools_bamtobed', 'metaphlan2', 'ivar_variants', 'fastq_paired_end_interlacer', 'ngsutils_bam_filter', 'bg_sortmerna', 'samtools_sort', 'cufflinks', 'cshl_fastq_to_fasta', 'samtool_filter2', 'rbc_mirdeep2_mapper', 'samtools_slice_bam', 'deeptools_bam_compare', 'pe_histogram', 'trinity', 'Extract genomic DNA 1', 'racon', 'bioext_bam2msa', 'tp_tail_tool', 'tophat2', 'bcftools_mpileup', 'bwa', 'bamtools', 'humann2', 'rseqc_bam_stat', 'picard_SortSam', 'sam_merge2', 'jbrowse', 'jvarkit_wgscoverageplotter', 'megahit', 'fastq_to_tabular', 'peakachu', 'samtools_merge', 'metaphlan', 'diffbind', 'picard_CollectInsertSizeMetrics', 'fastq_paired_end_joiner', 'deeptools_multi_bam_summary', 'deeptools_plot_fingerprint', 'samtools_depth', 'picard_CASM', 'cuffdiff', 'velvetoptimiser', 'picard_MergeSamFiles', 'picard_FilterSamReads', 'deeptools_alignmentsieve', 'metaspades', 'PicardASMetrics', 'bedtools_genomecoveragebed_bedgraph', 'mageck_count', 'bedtools_coveragebed', 'sam_pileup', 'seq_filter_by_mapping', 'umi_tools_dedup', 'naive_variant_caller', 'varscan', 'mimodd_varcall', 'collection_element_identifiers', 'pilon', 'samtools_mpileup', 'bedtools_multicovtbed', 'rnaspades', 'ococo', 'minia', 'hicexplorer_hicbuildmatrix', 'gatk4_mutect2', 'picard_DownsampleSam', 'bam_to_sam', 'samtools_split', 'basil', 'picard_artifact_metrics', 'stacks2_refmap', 'deeptools_plot_enrichment', 'picard_AddOrReplaceReadGroups', 'picard_QualityScoreDistribution', 'rseqc_inner_distance', 'paralyzer', 'picard_ARRG', 'get_read_pipeline', 'deeptools_bamCorrelate', 'rgPicardMarkDups', 'samtools_filter', 'bamSplit', 'mimodd_delcall', 'rseqc_read_distribution', 'qualimap_rnaseq', 'rseqc_geneBody_coverage', 'fastq_groomer', 'samtools_fixmate', 'macs2_bdgpeakcall', 'idr', 'macs2_bdgcmp', 'intervene_pairwise', 'intervene_upset', 'rseqc_read_duplication', 'exomedepth', 'bowtie2', 'pileometh', 'macs2_bdgdiff', 'rseqc_bam2wig', 'gemini_query', 'cutadapt', 'picard_ReorderSam', 'deeptools_plot_coverage', 'bellerophon', 'control_freec', 'delly_call', 'bedtools_bamtofastq', 'bedtools_genomecoveragebed_histogram', 'hicexplorer_hicquickqc']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.10273972602739725\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0958904109589041\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['tp_cut_tool', 'tp_tail_tool', 'fastq_dump']\n",
      "\n",
      "Test batch 1, True tools: ['trimmomatic', 'fastq_paired_end_deinterlacer', 'fastq_manipulation', 'fastq_trimmer', 'kallisto_quant', 'regex1', 'khmer_abundance_distribution_single', 'multiqc', 'CONVERTER_gz_to_uncompressed', 'fastq_groomer', 'minimap2', 'cshl_fastx_trimmer', 'cat1', 'fastq_to_fasta_python', 'rna_star', 'bwa_mem', 'kraken2', 'fastp', 'fastqc', 'cutadapt', 'cshl_fastx_clipper', 'predict_pipeline', 'cufflinks', 'bowtie_wrapper', 'gffread', 'fastq_paired_end_splitter', 'bismark_bowtie2', 'bowtie2', 'cshl_fastq_quality_filter', 'sample_seqs', 'snippy', 'rna_starsolo', 'fastq_stats', 'sickle', 'hifiasm', 'megahit', '__EXTRACT_DATASET__', 'cshl_fastq_to_fasta', 'bwameth', 'spades', 'gffcompare', 'hisat2', 'trim_galore']\n",
      "\n",
      "Test batch 1, Predicted top 43 tools: ['multiqc', 'cshl_fastq_quality_filter', 'bowtie2', 'hifiasm', 'CONVERTER_gz_to_uncompressed', 'fastq_paired_end_splitter', 'snippy', 'sickle', 'fastqc', 'bwameth', 'kallisto_quant', 'bismark_bowtie2', 'fastq_to_fasta_python', 'cshl_fastx_clipper', 'fastp', 'predict_pipeline', 'khmer_abundance_distribution_single', 'fastq_manipulation', 'sample_seqs', 'bwa_mem', 'cufflinks', 'kraken2', 'fastq_groomer', 'bowtie_wrapper', 'megahit', 'regex1', 'trimmomatic', 'fastq_trimmer', 'fastq_paired_end_deinterlacer', 'trim_galore', 'gffcompare', 'cshl_fastx_trimmer', '__EXTRACT_DATASET__', 'rna_starsolo', 'cshl_fastq_to_fasta', 'cat1', 'fastq_stats', 'rna_star', 'cutadapt', 'gffread', 'spades', 'minimap2', 'hisat2']\n",
      "\n",
      "Test batch 1, Predicted top 43 tools with weights: ['fastp', 'bwa_mem', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'trimmomatic', 'hisat2', 'multiqc', 'minimap2', 'kraken2', 'fastq_to_fasta_python', 'trim_galore', 'fastq_groomer', 'spades', 'snippy', 'bowtie_wrapper', 'cufflinks', 'cshl_fastq_to_fasta', 'kallisto_quant', 'cshl_fastx_trimmer', 'bwameth', 'cshl_fastq_quality_filter', 'gffread', 'megahit', 'fastq_paired_end_deinterlacer', 'rna_starsolo', 'fastq_trimmer', 'sickle', 'regex1', 'gffcompare', 'sample_seqs', 'hifiasm', 'bismark_bowtie2', 'cshl_fastx_clipper', 'fastq_stats', 'fastq_paired_end_splitter', 'khmer_abundance_distribution_single', 'predict_pipeline', 'seqtk_sample', 'fastq_filter', 'samtools_rmdup', 'cshl_fastx_reverse_complement']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.023255813953488372\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.023255813953488372\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['locarna_multiple', 'infernal_cmbuild']\n",
      "\n",
      "Test batch 1, True tools: ['infernal_cmstat', 'infernal_cmsearch', '__FILTER_FAILED_DATASETS__', 'cmv']\n",
      "\n",
      "Test batch 1, Predicted top 4 tools: ['infernal_cmsearch', 'cmv', 'infernal_cmstat', '__FILTER_FAILED_DATASETS__']\n",
      "\n",
      "Test batch 1, Predicted top 4 tools with weights: ['infernal_cmsearch', 'infernal_cmstat', 'cmv', 'scanpy_plot']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.75\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.75\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['fastp', 'bwa_mem', 'samtools_view', 'lofreq_viterbi', 'lofreq_indelqual', 'ivar_trim', 'ivar_removereads', 'lofreq_call', 'vcfvcfintersect', 'tp_replace_in_line', 'snpeff_sars_cov_2', 'snpSift_filter', 'snpSift_extractFields', 'replace_column_with_key_value_file', 'datamash_ops', 'tp_find_and_replace', 'Filter1', 'gops_concat_1', 'gops_merge_1', 'gops_subtract_1']\n",
      "\n",
      "Test batch 1, True tools: ['gene2exon1', 'join1', 'flanking_features_1', 'Count1', 'Extract genomic DNA 1', 'gops_intersect_1', 'deeptools_compute_matrix', 'Filter1', 'tp_replace_in_column', 'Add_a_column1', 'bedtools_maskfastabed', 'tp_head_tool', 'bedtools_getfastabed']\n",
      "\n",
      "Test batch 1, Predicted top 13 tools: ['flanking_features_1', 'gene2exon1', 'bedtools_maskfastabed', 'bedtools_getfastabed', 'gops_intersect_1', 'Add_a_column1', 'tp_replace_in_column', 'tp_head_tool', 'Count1', 'join1', 'Extract genomic DNA 1', 'Filter1', 'deeptools_compute_matrix']\n",
      "\n",
      "Test batch 1, Predicted top 13 tools with weights: ['Add_a_column1', 'deeptools_compute_matrix', 'tp_replace_in_column', 'gops_intersect_1', 'tp_head_tool', 'bedtools_maskfastabed', 'bedtools_getfastabed', 'Extract genomic DNA 1', 'tp_tail_tool', 'table_compute', 'tp_awk_tool', 'gops_subtract_1', 'annotatemyids']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.07692307692307693\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.07692307692307693\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'mothur_make_contigs', 'mothur_screen_seqs', 'seq_filter_by_id', 'mothur_unique_seqs', 'mothur_pre_cluster', 'mothur_chimera_vsearch', 'mothur_remove_seqs', 'mothur_classify_seqs', 'mothur_remove_lineage', 'mothur_make_shared', 'mothur_sub_sample']\n",
      "\n",
      "Test batch 1, True tools: ['mothur_unifrac_unweighted', 'mothur_dist_shared', 'mothur_unifrac_weighted', 'mothur_rarefaction_single', 'mothur_corr_axes', 'mothur_venn', 'mycrobiota_subsample_add_discarded_samples', 'mothur_count_groups', 'mothur_summary_tax', 'mothur_summary_single', 'mothur_make_biom', 'mothur_get_communitytype', 'mothur_metastats', 'collapse_dataset']\n",
      "\n",
      "Test batch 1, Predicted top 14 tools: ['mothur_count_groups', 'mothur_summary_tax', 'mycrobiota_subsample_add_discarded_samples', 'collapse_dataset', 'mothur_summary_single', 'mothur_rarefaction_single', 'mothur_unifrac_weighted', 'mothur_unifrac_unweighted', 'mothur_venn', 'mothur_corr_axes', 'mothur_dist_shared', 'mothur_metastats', 'mothur_get_communitytype', 'mothur_make_biom']\n",
      "\n",
      "Test batch 1, Predicted top 14 tools with weights: ['collapse_dataset', 'mothur_make_biom', 'mothur_count_groups', 'mothur_venn', 'mothur_summary_single', 'mothur_dist_shared', 'mothur_rarefaction_single', 'mothur_corr_axes', 'mothur_get_communitytype', 'mothur_unifrac_weighted', 'mothur_metastats', 'sansa_annotate', 'deeptools_alignmentsieve', 'fastqc']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.14285714285714285\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.14285714285714285\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'macs2_callpeak', 'bowtie2', 'samtools_rmdup', 'macs2_callpeak', 'join1', 'Add_a_column1', 'Cut1', 'Filter1', 'chipseeker', 'Remove beginning1', 'bedtools_intersectbed', 'bedtools_subtractbed', 'tp_sort_header_tool']\n",
      "\n",
      "Test batch 1, True tools: ['datamash_ops', 'wig_to_bigWig', 'metilene', 'tp_sorted_uniq', 'Show tail1', 'filter_tabular', 'goenrichment', 'comp1', 'gd_rank_terms', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'regex1', 'tp_multijoin_tool', 'multiqc', 'seurat', 'Cut1', 'tp_cat', 'barchart_gnuplot', 'bedtools_clusterbed', '__MERGE_COLLECTION__', 'Remove beginning1', 'column_order_header_sort', 'column_remove_by_header', 'bg_uniq', 'bedtools_mergebed', 'tp_find_and_replace', 'Filter1', 'samtools_idxstats', 'samtools_slice_bam', 'Add_a_column1', 'Show beginning1', 'histogram_rpy', 'bedtools_complementbed', 'XY_Plot_1', 'datamash_transpose', 'gops_join_1', 'Grouping1', 'gops_intersect_1', 'sort1', 'tp_awk_tool', 'tp_head_tool', 'Grep1', 'tp_tail_tool', 'table_compute', 'addValue', 'annotatemyids', 'tp_grep_tool', 'bed_to_bigBed', 'join1', 'tp_uniq_tool', 'bedtools_genomecoveragebed_bedgraph', 'fgsea', 'blockbuster', 'cshl_awk_tool', 'hgv_david', 'tp_cut_tool', 'Convert characters1', 'cummeRbund', 'Summary_Statistics1', 'wc_gnu', 'transit_gumbel']\n",
      "\n",
      "Test batch 1, Predicted top 61 tools: ['bedtools_complementbed', 'tp_uniq_tool', 'metilene', 'barchart_gnuplot', 'Grep1', 'datamash_transpose', 'comp1', 'multiqc', 'bedtools_genomecoveragebed_bedgraph', 'cummeRbund', 'wig_to_bigWig', 'bed_to_bigBed', 'cshl_awk_tool', 'tp_cat', 'samtools_slice_bam', 'Add_a_column1', 'seurat', 'gd_rank_terms', 'bedtools_clusterbed', 'tp_awk_tool', 'addValue', 'join1', 'tp_cut_tool', 'fgsea', 'filter_tabular', 'Cut1', 'Filter1', 'samtools_idxstats', 'blockbuster', 'Summary_Statistics1', 'annotatemyids', 'transit_gumbel', 'gops_join_1', 'bedtools_mergebed', 'regex1', 'XY_Plot_1', 'tp_easyjoin_tool', 'Show beginning1', 'goenrichment', 'tp_find_and_replace', 'gops_intersect_1', 'bg_uniq', 'Remove beginning1', 'table_compute', '__MERGE_COLLECTION__', 'column_order_header_sort', 'histogram_rpy', 'tp_grep_tool', 'tp_multijoin_tool', 'taxonomy_krona_chart', 'hgv_david', 'sort1', 'Convert characters1', 'tp_head_tool', 'Grouping1', 'Show tail1', 'column_remove_by_header', 'datamash_ops', 'tp_tail_tool', 'tp_sorted_uniq', 'wc_gnu']\n",
      "\n",
      "Test batch 1, Predicted top 61 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'datamash_ops', 'filter_tabular', 'multiqc', 'datamash_transpose', 'tp_easyjoin_tool', 'tp_cat', 'table_compute', 'samtools_idxstats', 'tp_grep_tool', 'tp_awk_tool', 'annotatemyids', 'taxonomy_krona_chart', 'gops_intersect_1', 'tp_head_tool', 'samtools_slice_bam', 'addValue', 'tp_tail_tool', 'fgsea', 'XY_Plot_1', 'bedtools_mergebed', 'goenrichment', 'bg_uniq', 'tp_sorted_uniq', 'regex1', 'bedtools_genomecoveragebed_bedgraph', 'metilene', 'histogram_rpy', 'column_remove_by_header', 'tp_uniq_tool', 'seurat', 'column_order_header_sort', 'cummeRbund', 'blockbuster', 'transit_gumbel', 'gd_rank_terms', 'freebayes', 'scatterplot_rpy', 'pathview', 'deseq2', 'bedtools_intersectbed', 'pygenomeTracks', 'samtools_view', 'chipeakanno_annopeaks', 'scanpy_plot', 'tp_replace_in_line', 'bedtools_sortbed', 'tab2fasta', 'bedtools_slopbed', 'cummerbund_to_cuffdiff', 'bedtools_multiintersectbed', 'bedtools_map', 'ggplot2_heatmap2', 'regexColumn1', 'macs2_bdgpeakcall', 'macs2_callpeak', 'tp_sort_header_tool', 'bedtools_bed12tobed6', 'macs2_bdgdiff', 'goseq']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.11475409836065574\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.04918032786885246\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['get_pdb', 'tp_grep_tool', 'prepare_receptor', 'docking', 'split_file_to_collection', 'openbabel_compound_convert']\n",
      "\n",
      "Test batch 1, True tools: ['padel', 'prepare_box', 'prepare_ligands_for_docking', 'ctb_frankenstein_ligand', 'tp_cat', 'openbabel_svg_depiction', 'cat1', 'gmx_setup', 'tp_find_and_replace', 'docking', 'openbabel_genProp', '__TAG_FROM_FILE__', 'rdock_rbdock', 'ctb_simsearch', 'xchem_transfs_scoring', 'collection_element_identifiers', 'ctb_chemfp_mol2fps', 'rxdock_rbcavity', 'ambertools_acpype', 'rxdock_rbdock', 'tp_grep_tool', 'split_file_to_collection', 'ambertools_antechamber', 'chembl', 'ctb_rdkit_descriptors', 'tp_cut_tool', 'rdock_rbcavity']\n",
      "\n",
      "Test batch 1, Predicted top 27 tools: ['xchem_transfs_scoring', 'tp_grep_tool', 'rxdock_rbdock', 'ambertools_acpype', 'ctb_rdkit_descriptors', 'collection_element_identifiers', 'docking', 'gmx_setup', 'prepare_ligands_for_docking', 'tp_cut_tool', 'padel', 'ctb_chemfp_mol2fps', 'ambertools_antechamber', 'ctb_frankenstein_ligand', 'chembl', 'prepare_box', 'rdock_rbcavity', 'cat1', 'openbabel_svg_depiction', 'ctb_simsearch', 'rdock_rbdock', '__TAG_FROM_FILE__', 'rxdock_rbcavity', 'openbabel_genProp', 'tp_find_and_replace', 'split_file_to_collection', 'tp_cat']\n",
      "\n",
      "Test batch 1, Predicted top 27 tools with weights: ['tp_find_and_replace', 'tp_cat', 'tp_grep_tool', 'split_file_to_collection', 'docking', 'ambertools_acpype', 'chembl', 'gmx_setup', 'prepare_box', 'openbabel_svg_depiction', 'collection_element_identifiers', 'rxdock_rbdock', 'ambertools_antechamber', 'prepare_ligands_for_docking', 'ctb_rdkit_descriptors', 'rdock_rbcavity', 'xchem_transfs_scoring', 'ctb_chemfp_mol2fps', 'rxdock_rbcavity', 'openbabel_genProp', 'padel', 'ctb_simsearch', 'plotdexseq', 'collection_column_join', 'addName', 'query_tabular', 'gmx_sim']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.14814814814814814\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.14814814814814814\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'macs2_callpeak', 'diffbind', 'Filter1', 'Remove beginning1', 'Cut1', 'Extract genomic DNA 1', 'macs2_callpeak', 'bowtie2']\n",
      "\n",
      "Test batch 1, True tools: ['velvetoptimiser', 'rseqc_bam_stat', 'gatk2_realigner_target_creator', 'deeptools_bamCompare', 'stacks2_refmap', 'get_read_pipeline', 'bedtools_multicovtbed', 'picard_AddOrReplaceReadGroups', 'htseq_count', '__MERGE_COLLECTION__', 'deeptools_alignmentsieve', 'bedtools_bamtobed', 'ococo', 'bamSplit', 'rgPicardMarkDups', 'kraken2', 'ngsutils_bam_filter', 'fastq_paired_end_interlacer', 'picard_MarkDuplicates', 'peakachu', 'basil', 'samtools_filter', 'collection_element_identifiers', 'cuffdiff', 'bwa', 'deeptools_compute_gc_bias', 'pilon', 'bamFilter', 'ivar_trim', 'tp_tail_tool', 'deeptools_plot_enrichment', 'picard_CollectInsertSizeMetrics', 'humann2', 'deeptools_plot_fingerprint', 'picard_MergeSamFiles', 'mageck_count', 'umi_tools_count', 'spades', 'tophat2', 'samtools_split', 'bundle_collection', 'bedtools_genomecoveragebed', 'deeptools_computeGCBias', 'bcftools_mpileup', 'deeptools_bamCoverage', 'bedtools_intersectbed', 'PicardASMetrics', 'samtools_stats', '__FLATTEN__', 'minia', 'macs2_callpeak', 'racon', 'picard_FilterSamReads', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'picard_ARRG', 'gatk4_mutect2', 'cufflinks', 'tp_sort_header_tool', 'Grouping1', 'naive_variant_caller', 'lofreq_indelqual', 'samtools_rmdup', 'deeptools_bamFingerprint', 'trinity', 'picard_SortSam', 'picard_artifact_metrics', 'samtool_filter2', 'ivar_variants', 'samtools_view', 'picard_QualityScoreDistribution', 'paralyzer', 'megahit', 'rnaspades', 'hicexplorer_hicbuildmatrix', 'featurecounts', 'picard_DownsampleSam', 'samtools_depth', 'bam_to_bigwig', 'je_markdupes', 'bioext_bam2msa', 'trimmomatic', 'lofreq_viterbi', 'seq_filter_by_mapping', 'stringtie', 'tp_cat', 'deeptools_bamCorrelate', 'samtools_sort', 'fastq_to_tabular', 'sam_merge2', 'deeptools_correct_gc_bias', 'bg_sortmerna', 'bam-filter', 'rna_star', 'bwa_mem', 'deeptools_multi_bam_summary', 'samtools_slice_bam', 'fastqc', 'peakcalling_macs', 'rbc_mirdeep2_mapper', 'samtools_flagstat', 'fastq_paired_end_joiner', 'ivar_consensus', 'metaphlan', 'picard_CASM', 'metaphlan2_hutlab', 'umi_tools_dedup', 'varscan', 'jvarkit_wgscoverageplotter', 'diffbind', 'picard_CleanSam', 'cufflinks_prok', 'pe_histogram', 'salmon', 'rseqc_inner_distance', 'bam_to_sam', 'macs2_predictd', 'samtools_merge', 'bamtools', 'samtools_mpileup', 'sam_pileup', 'lofreq_call', 'multiqc', 'jbrowse', 'deeptools_bam_compare', 'PicardInsertSize', 'cat1', 'mimodd_varcall', 'sam_to_bam', 'Extract genomic DNA 1', 'metaspades', 'samtools_idxstats', 'bamleftalign', 'rseqc_infer_experiment', 'mimodd_delcall', '__BUILD_LIST__', 'qualimap_bamqc', 'metaphlan2', 'unicycler', 'freebayes', 'samtools_fastx', 'bedtools_genomecoveragebed_bedgraph', 'cshl_fastq_to_fasta', 'CONVERTER_bam_to_bigwig_0', 'bedtools_coveragebed', 'hisat2']\n",
      "\n",
      "Test batch 1, Predicted top 146 tools: ['bedtools_bamtobed', 'samtools_filter', 'pilon', 'humann2', 'umi_tools_count', 'tophat2', 'samtools_split', 'deeptools_computeGCBias', 'bcftools_mpileup', 'macs2_callpeak', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'picard_ARRG', 'trinity', 'picard_artifact_metrics', 'samtool_filter2', 'paralyzer', 'deeptools_bamCorrelate', 'fastq_to_tabular', 'sam_merge2', 'deeptools_correct_gc_bias', 'rna_star', 'varscan', 'samtools_merge', 'samtools_mpileup', 'lofreq_call', 'multiqc', 'metaspades', 'metaphlan2', 'bedtools_genomecoveragebed_bedgraph', 'gatk2_realigner_target_creator', 'stacks2_refmap', 'htseq_count', 'ococo', 'peakachu', 'bwa', 'deeptools_compute_gc_bias', 'ivar_trim', 'tp_tail_tool', 'picard_CollectInsertSizeMetrics', 'mageck_count', 'tp_sort_header_tool', 'picard_SortSam', 'rnaspades', 'featurecounts', 'samtools_depth', 'je_markdupes', 'tp_cat', 'bg_sortmerna', 'bwa_mem', 'samtools_slice_bam', 'fastq_paired_end_joiner', 'metaphlan', 'pe_histogram', 'rseqc_inner_distance', 'sam_pileup', 'deeptools_bam_compare', 'rseqc_infer_experiment', 'qualimap_bamqc', 'rseqc_bam_stat', 'bedtools_multicovtbed', 'ngsutils_bam_filter', 'bamFilter', 'deeptools_plot_fingerprint', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', 'PicardASMetrics', 'samtools_stats', '__FLATTEN__', 'racon', 'gatk4_mutect2', 'deeptools_bamFingerprint', 'picard_QualityScoreDistribution', 'megahit', 'bam_to_bigwig', 'trimmomatic', 'lofreq_viterbi', 'seq_filter_by_mapping', 'samtools_sort', 'bam-filter', 'fastqc', 'rbc_mirdeep2_mapper', 'picard_CASM', 'metaphlan2_hutlab', 'picard_CleanSam', 'salmon', 'bam_to_sam', 'PicardInsertSize', 'unicycler', 'cshl_fastq_to_fasta', 'velvetoptimiser', 'bamSplit', 'fastq_paired_end_interlacer', 'samtools_view', 'stringtie', 'deeptools_multi_bam_summary', 'samtools_flagstat', 'jvarkit_wgscoverageplotter', 'diffbind', 'cufflinks_prok', 'cat1', 'mimodd_varcall', 'Extract genomic DNA 1', 'samtools_idxstats', 'bamleftalign', 'freebayes', 'CONVERTER_bam_to_bigwig_0', 'deeptools_bamCompare', 'get_read_pipeline', 'picard_AddOrReplaceReadGroups', 'kraken2', 'collection_element_identifiers', 'spades', 'minia', 'macs2_predictd', 'cuffdiff', 'lofreq_indelqual', '__MERGE_COLLECTION__', 'sam_to_bam', 'mimodd_delcall', 'deeptools_bamCoverage', 'deeptools_alignmentsieve', 'naive_variant_caller', 'ivar_consensus', 'hisat2', 'rgPicardMarkDups', 'deeptools_plot_enrichment', 'basil', 'samtools_rmdup', 'cufflinks', 'hicexplorer_hicbuildmatrix', 'jbrowse', 'picard_MarkDuplicates', 'picard_DownsampleSam', 'bioext_bam2msa', 'picard_FilterSamReads', 'bedtools_coveragebed', 'bundle_collection', 'bamtools', 'picard_MergeSamFiles', 'peakcalling_macs', 'ivar_variants', 'Grouping1', 'umi_tools_dedup', '__BUILD_LIST__', 'samtools_fastx']\n",
      "\n",
      "Test batch 1, Predicted top 146 tools with weights: ['lofreq_call', 'bwa_mem', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'ivar_trim', 'lofreq_indelqual', 'lofreq_viterbi', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'rna_star', 'trimmomatic', 'hisat2', 'multiqc', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'kraken2', 'fastq_to_fasta_python', 'deeptools_bam_coverage', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'samtools_rmdup', 'tp_cat', 'spades', 'ivar_consensus', 'bamFilter', 'rseqc_infer_experiment', 'unicycler', 'freebayes', 'salmon', 'samtools_flagstat', 'samtools_idxstats', 'bamleftalign', 'bedtools_bamtobed', 'metaphlan2', 'ivar_variants', 'fastq_paired_end_interlacer', 'ngsutils_bam_filter', 'bg_sortmerna', 'samtools_sort', 'cufflinks', 'cshl_fastq_to_fasta', 'samtool_filter2', 'rbc_mirdeep2_mapper', 'samtools_slice_bam', 'deeptools_bam_compare', 'pe_histogram', 'trinity', 'Extract genomic DNA 1', 'racon', 'bioext_bam2msa', 'tp_tail_tool', 'tophat2', 'bcftools_mpileup', 'bwa', 'bamtools', 'humann2', 'rseqc_bam_stat', 'picard_SortSam', 'sam_merge2', 'jbrowse', 'jvarkit_wgscoverageplotter', 'megahit', 'fastq_to_tabular', 'peakachu', 'samtools_merge', 'metaphlan', 'diffbind', 'picard_CollectInsertSizeMetrics', 'fastq_paired_end_joiner', 'deeptools_multi_bam_summary', 'deeptools_plot_fingerprint', 'samtools_depth', 'picard_CASM', 'cuffdiff', 'velvetoptimiser', 'picard_MergeSamFiles', 'picard_FilterSamReads', 'deeptools_alignmentsieve', 'metaspades', 'PicardASMetrics', 'bedtools_genomecoveragebed_bedgraph', 'mageck_count', 'bedtools_coveragebed', 'sam_pileup', 'seq_filter_by_mapping', 'umi_tools_dedup', 'naive_variant_caller', 'varscan', 'mimodd_varcall', 'collection_element_identifiers', 'pilon', 'samtools_mpileup', 'bedtools_multicovtbed', 'rnaspades', 'ococo', 'minia', 'hicexplorer_hicbuildmatrix', 'gatk4_mutect2', 'picard_DownsampleSam', 'bam_to_sam', 'samtools_split', 'basil', 'picard_artifact_metrics', 'stacks2_refmap', 'deeptools_plot_enrichment', 'picard_AddOrReplaceReadGroups', 'picard_QualityScoreDistribution', 'rseqc_inner_distance', 'paralyzer', 'picard_ARRG', 'get_read_pipeline', 'deeptools_bamCorrelate', 'rgPicardMarkDups', 'samtools_filter', 'bamSplit', 'mimodd_delcall', 'rseqc_read_distribution', 'qualimap_rnaseq', 'rseqc_geneBody_coverage', 'fastq_groomer', 'macs2_bdgpeakcall', 'samtools_fixmate', 'idr', 'macs2_bdgcmp', 'intervene_pairwise', 'intervene_upset', 'rseqc_read_duplication', 'bowtie2', 'exomedepth', 'macs2_bdgdiff', 'pileometh', 'gemini_query', 'cutadapt', 'rseqc_bam2wig', 'picard_ReorderSam', 'deeptools_plot_coverage', 'bellerophon', 'Add_a_column1', 'control_freec', 'stacks2_gstacks', 'bedtools_genomecoveragebed_histogram', 'delly_call']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.10273972602739725\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0958904109589041\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'rna_star', 'rseqc_junction_annotation', 'gops_join_1', 'Filter1', 'annotatemyids', 'Cut1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'join_files_on_column_fuzzy', 'regexColumn1', 'collection_column_join', 'plotly_ml_performance_plots', 'bg_uniq', 'venn_list', 'scanpy_read_10x', 'tp_find_and_replace', 'bedtools_slopbed', 'Paste1', 'Show beginning1', 'ggplot_histogram', 'krona-text', 'stacks_denovomap', 'XY_Plot_1', 'random_lines1', 'unipept', 'query_tabular', 'ChangeCase', 'tp_tail_tool', 'collapse_dataset', 'annotatemyids', 'join1', 'goseq', 'heinz_scoring', 'tp_uniq_tool', 'export2graphlan', 'tp_cut_tool', 'seqtk_subseq', 'wc_gnu', 'bedtools_genomecoveragebed', 'wig_to_bigWig', 'clusterprofiler_bitr', 'bedtools_intersectbed', 'crossmap_bed', '__FLATTEN__', 'barchart_gnuplot', 'gtftobed12', 'Remove beginning1', 'macs2_callpeak', 'picard_FilterSamReads', 'pygenomeTracks', 'get_feature_info', 'bedtools_mergebed', 'ggplot_point', 'tp_replace_in_line', 'stringtie_merge', '__FILTER_FROM_FILE__', 'deseq2', 'mergeCols1', 'biotranslator', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'ggplot2_heatmap', 'tab2fasta', 'Grep1', 'msstats', 'table_compute', 'tp_grep_tool', 'egsea', 'add_line_to_file', 'fgsea', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'annotateMyIDs', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'goenrichment', 'regex1', 'tp_multijoin_tool', 'CONVERTER_interval_to_bed_0', 'gops_concat_1', 'bedtools_intersectbed_bam', 'tp_cat', 'charts', 'bcftools_consensus', 'fasterq_dump', 'ggplot2_heatmap2', 'rna_star', 'combine_metaphlan2_humann2', 'stacks_populations', 'Add_a_column1', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'w4mclassfilter', 'get_sequences', 'Kraken2Tax', 'cshl_find_and_replace', 'ont_fast5_api_fast5_subset', 'interactions', 'gops_coverage_1', 'heinz_bum', 'addName', 'mass_spectrometry_imaging_ion_images', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'Show tail1', 'trimmer', 'comp1', 'secure_hash_message_digest', 'volcanoplot', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'circos', 'gops_subtract_1', 'cat1', 'scatterplot_rpy', 'regex_replace', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'histogram_rpy', 'diff', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'interactive_tool_simtext_app', 'sklearn_ensemble', 'addValue', 'dexseq_count', 'cardinal_filtering', 'MassCalculator', 'split_file_to_collection', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'CONVERTER_bed_to_gff_0', 'Krona', 'bedtools_sortbed', 'param_value_from_file', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools: ['get_flanks1', 'join_files_on_column_fuzzy', 'ggplot_histogram', 'tp_uniq_tool', 'barchart_gnuplot', 'deseq2', 'mergeCols1', 'Grouping1', 'ggplot2_heatmap', 'annotateMyIDs', 'bedtools_getfastabed', 'tp_sorted_uniq', 'goenrichment', 'rna_star', 'dropletutils_read_10x', 'datamash_transpose', 'openbabel_remIons', 'gops_coverage_1', 'comp1', 'regex_replace', 'sklearn_ensemble', 'dexseq_count', 'regexColumn1', 'scanpy_read_10x', 'krona-text', 'tp_tail_tool', 'heinz_scoring', 'wig_to_bigWig', 'clusterprofiler_bitr', 'crossmap_bed', 'Remove beginning1', 'ggplot_point', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'tab2fasta', 'msstats', 'table_compute', 'tp_cat', 'Add_a_column1', 'w4mclassfilter', 'cshl_find_and_replace', 'addName', 'mass_spectrometry_imaging_ion_images', 'secure_hash_message_digest', 'MassCalculator', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'param_value_from_file', 'plotly_ml_performance_plots', 'stacks_denovomap', 'random_lines1', 'query_tabular', 'ChangeCase', 'goseq', 'seqtk_subseq', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', '__FLATTEN__', 'gtftobed12', 'picard_FilterSamReads', 'stringtie_merge', 'biotranslator', 'egsea', 'tp_replace_in_column', 'seq_filter_by_id', 'fasterq_dump', 'combine_metaphlan2_humann2', 'stacks_populations', 'get_sequences', 'Kraken2Tax', 'ont_fast5_api_fast5_subset', 'heinz_bum', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'gops_join_1', 'sort1', 'tp_awk_tool', 'addValue', 'ncbi_acc_download', 'venn_list', 'Paste1', 'join1', 'tp_cut_tool', 'get_feature_info', 'add_line_to_file', 'fgsea', 'regex1', 'interactions', 'circos', 'cat1', 'scatterplot_rpy', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'diff', 'tp_head_tool', 'interactive_tool_simtext_app', 'CONVERTER_bed_to_gff_0', 'Krona', 'cshl_word_list_grep', 'collection_column_join', 'bg_uniq', 'XY_Plot_1', 'annotatemyids', 'export2graphlan', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'checkFormat', 'featurecounts', 'bcftools_consensus', 'hgv_david', 'tp_grep_tool', 'split_file_to_collection', 'bedtools_slopbed', 'Convert characters1', 'trimmer', 'bedtools_sortbed', 'unipept', 'collapse_dataset', 'Show beginning1', 'tp_multijoin_tool', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'tp_find_and_replace', 'volcanoplot', 'ggplot2_heatmap2', 'pygenomeTracks', '__FILTER_FROM_FILE__', 'bedtools_intersectbed_bam', 'tp_sed_tool', 'Count1', 'macs2_callpeak', 'datamash_ops', 'Show tail1', 'Summary_Statistics1', 'deeptools_compute_matrix', 'wc_gnu', 'histogram_rpy', 'Grep1', 'umi_tools_extract', 'bedtools_mergebed', 'bg_column_arrange_by_header', 'gops_subtract_1', 'gops_concat_1', 'CONVERTER_interval_to_bed_0', 'cardinal_filtering', 'charts', 'tp_split_on_column']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'tp_replace_in_line', 'datamash_ops', 'bedtools_genomecoveragebed', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'rna_star', 'fasterq_dump', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'volcanoplot', 'tp_grep_tool', 'circos', 'tp_awk_tool', 'annotatemyids', 'taxonomy_krona_chart', 'tp_replace_in_column', 'filter_by_fasta_ids', 'goseq', 'gops_intersect_1', 'collection_column_join', 'limma_voom', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'Kraken2Tax', 'addValue', 'combine_metaphlan2_humann2', 'pygenomeTracks', 'addName', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'regexColumn1', 'seqtk_subseq', 'fgsea', 'XY_Plot_1', 'bedtools_mergebed', 'export2graphlan', 'goenrichment', 'query_tabular', 'bg_uniq', 'krona-text', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'gtftobed12', 'umi_tools_extract', 'stringtie_merge', 'dexseq_count', 'regex1', 'msstats', 'picard_FilterSamReads', 'ggplot2_heatmap', 'mergeCols1', 'tp_split_on_column', 'histogram_rpy', 'ncbi_blastdbcmd_wrapper', 'scatterplot_rpy', 'venn_list', 'egsea', 'join_files_on_column_fuzzy', 'add_line_to_file', 'bedtools_slopbed', 'dropletutils_read_10x', 'plotly_ml_performance_plots', 'get_flanks1', 'tp_uniq_tool', 'regex_replace', 'bg_column_arrange_by_header', 'sklearn_ensemble', 'stacks_denovomap', 'diff', 'gops_coverage_1', 'get_feature_info', 'checkFormat', 'get_sequences', 'heinz_scoring', 'heinz_bum', 'MassCalculator', 'macs2_bdgdiff', 'gops_merge_1', 'edger', 'uniprot', 'xarray_select', 'tabular_to_fastq', 'deg_annotate', 'chipeakanno_annopeaks', 'metilene', 'bedtools_subtractbed', 'filter_tabular', 'htseq_count', 'raceid_filtnormconf', 'Heatmap', 'tables_arithmetic_operations', 'chipseeker', 'column_remove_by_header', 'cardinal_spectra_plots', 'cummeRbund', 'macs2_bdgpeakcall', 'multiqc', 'tbl2gff3', 'rseqc_infer_experiment', 'datamash_reverse', 'humann2', 'macs2_bdgcmp', 'graphprot_predict_profile', 'bedtools_multiintersectbed', 'vegan_diversity', 'plink', 'gops_cluster_1', 'melt', 'idr', 'scater_create_qcmetric_ready_sce', 'intervene_upset', 'intervene_pairwise', 'fastq_groomer', 'vegan_rarefaction', 'sklearn_svm_classifier', 'column_order_header_sort', 'samtools_idxstats', 'ggplot2_pca', 'stoceps_trend_indic', 'get_pdb', 'hicexplorer_hicfindrestrictionsites', 'subtract_query1', 'Fetch Taxonomic Ranks', 'trim_galore', 'ctb_chemfp_mol2fps', 'transit_gumbel', 'flair_correct', 'sklearn_feature_selection', 'kc-align', 'bedtools_genomecoveragebed_bedgraph', 'gd_rank_terms']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3092105263157895\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.21710526315789475\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['fastp', 'bowtie2', 'samtools_view', 'samtools_fastx', 'collapse_dataset', 'seqtk_sample', 'unicycler']\n",
      "\n",
      "Test batch 1, True tools: ['samtools_mpileup', 'medaka_consensus_pipeline', 'regex1', 'circos', 'fasta2tab', 'kraken', 'jbrowse', 'lastz_wrapper_2', 'chromeister', 'staramr_search', 'minimap2', 'racon', 'bcftools_consensus', 'miniasm', 'sam_to_bam', 'ncbi_blastp_wrapper', 'bwa_mem', 'kraken2', 'EMBOSS: megamerger53', 'abricate', 'quast', 'fasta-stats', 'antismash', 'bandage_image', 'fastani', 'mummer_mummer', 'bwa', 'directory_uri', 'filter_by_fasta_ids', 'ncbi_blastn_wrapper', 'transdecoder', 'pilon', 'metaphlan2', 'mlst', 'bowtie2', 'glimmer_knowlegde-based', 'freebayes', 'fasta_filter_by_length', 'prokka', 'busco', 'glimmer_build-icm', 'PlasFlow', 'nanoplot', 'progressivemauve', 'fasta_compute_length', 'bandage_info', 'mob_recon', 'bio_hansel']\n",
      "\n",
      "Test batch 1, Predicted top 48 tools: ['staramr_search', 'bandage_image', 'bwa_mem', 'medaka_consensus_pipeline', 'ncbi_blastn_wrapper', 'mummer_mummer', 'quast', 'bandage_info', 'fasta-stats', 'filter_by_fasta_ids', 'mob_recon', 'bio_hansel', 'PlasFlow', 'prokka', 'circos', 'pilon', 'abricate', 'progressivemauve', 'fasta_compute_length', 'fastani', 'lastz_wrapper_2', 'miniasm', 'EMBOSS: megamerger53', 'chromeister', 'antismash', 'fasta2tab', 'busco', 'glimmer_build-icm', 'sam_to_bam', 'glimmer_knowlegde-based', 'racon', 'bcftools_consensus', 'kraken', 'fasta_filter_by_length', 'transdecoder', 'metaphlan2', 'freebayes', 'kraken2', 'ncbi_blastp_wrapper', 'mlst', 'bwa', 'samtools_mpileup', 'regex1', 'minimap2', 'directory_uri', 'bowtie2', 'nanoplot', 'jbrowse']\n",
      "\n",
      "Test batch 1, Predicted top 48 tools with weights: ['bwa_mem', 'bcftools_consensus', 'bowtie2', 'prokka', 'abricate', 'minimap2', 'kraken2', 'ncbi_blastn_wrapper', 'fasta2tab', 'bandage_image', 'ncbi_blastp_wrapper', 'PlasFlow', 'freebayes', 'busco', 'quast', 'fasta-stats', 'circos', 'metaphlan2', 'nanoplot', 'filter_by_fasta_ids', 'racon', 'fasta_filter_by_length', 'staramr_search', 'bwa', 'jbrowse', 'mlst', 'fasta_compute_length', 'antismash', 'mob_recon', 'miniasm', 'bandage_info', 'regex1', 'mummer_mummer', 'medaka_consensus_pipeline', 'fastani', 'chromeister', 'progressivemauve', 'kraken', 'pilon', 'samtools_mpileup', 'bio_hansel', 'transdecoder', 'glimmer_knowlegde-based', 'EMBOSS: megamerger53', 'glimmer_build-icm', 'fasta_merge_files_and_filter_unique_sequences', 'seq_filter_by_id', 'meme_meme']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.25\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.25\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['__UNZIP_COLLECTION__', 'cutadapt', 'trim_galore', 'umi_tools_extract', 'rna_star', 'bam_to_sam', 'tp_awk_tool', 'tp_sed_tool', 'sam_to_bam', 'umi_tools_dedup', 'piranha', 'bed2gff1', 'bedtools_intersectbed', 'wc_gnu', 'tp_awk_tool', 'bedtools_sortbed']\n",
      "\n",
      "Test batch 1, True tools: ['wig_to_bigWig', 'bedtools_intersectbed', 'Cut1', 'bedtools_bed12tobed6', 'bedtools_genomecoveragebed', 'pygenomeTracks', 'bedtools_mergebedgraph', 'bedtools_map', 'bedtools_mergebed', 'bedtools_multiintersectbed', 'Add_a_column1', 'bedtools_complementbed', 'deeptools_compute_matrix', 'tp_awk_tool', 'bedtools_spacingbed', 'bedtools_bedtobam', 'addValue', 'bcftools_view', 'bedtools_genomecoveragebed_bedgraph', 'wc_gnu']\n",
      "\n",
      "Test batch 1, Predicted top 20 tools: ['bedtools_mergebedgraph', 'Add_a_column1', 'bedtools_spacingbed', 'bedtools_map', 'bedtools_bed12tobed6', 'bedtools_bedtobam', 'bedtools_complementbed', 'bcftools_view', 'bedtools_genomecoveragebed_bedgraph', 'deeptools_compute_matrix', 'wig_to_bigWig', 'bedtools_mergebed', 'pygenomeTracks', 'bedtools_multiintersectbed', 'wc_gnu', 'bedtools_intersectbed', 'bedtools_genomecoveragebed', 'addValue', 'Cut1', 'tp_awk_tool']\n",
      "\n",
      "Test batch 1, Predicted top 20 tools with weights: ['Add_a_column1', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', 'deeptools_compute_matrix', 'tp_awk_tool', 'addValue', 'pygenomeTracks', 'bedtools_mergebed', 'bedtools_genomecoveragebed_bedgraph', 'bedtools_multiintersectbed', 'bedtools_bedtobam', 'bedtools_map', 'bcftools_view', 'bedtools_mergebedgraph', 'bedtools_bed12tobed6', 'bedtools_spacingbed', 'tp_sort_header_tool', 'pileup_parser', 'bedtools_sortbed', 'bigwig_to_bedgraph']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.2\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.15\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['OpenSwathWorkflow', 'pyprophet_merge', 'pyprophet_score', 'pyprophet_peptide']\n",
      "\n",
      "Test batch 1, True tools: ['pyprophet_protein']\n",
      "\n",
      "Test batch 1, Predicted top 1 tools: ['pyprophet_protein']\n",
      "\n",
      "Test batch 1, Predicted top 1 tools with weights: ['pyprophet_protein']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['openbabel_compound_convert', 'rdock_rbdock', 'collapse_dataset', 'sucos_docking_scoring']\n",
      "\n",
      "Test batch 1, True tools: ['rdock_sort_filter', 'sucos_max_score', 'rxdock_sort_filter', 'tp_cat']\n",
      "\n",
      "Test batch 1, Predicted top 4 tools: ['rxdock_sort_filter', 'tp_cat', 'sucos_max_score', 'rdock_sort_filter']\n",
      "\n",
      "Test batch 1, Predicted top 4 tools with weights: ['tp_cat', 'rdock_sort_filter', 'sucos_max_score', 'tp_find_and_replace']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['porechop', 'canu', 'minimap2']\n",
      "\n",
      "Test batch 1, True tools: ['rseqc_bam_stat', 'medaka_consensus', 'htseq_count', 'bedtools_bamtobed', 'ococo', 'miniasm', 'rseqc_insertion_profile', 'medaka_variant', 'ngsutils_bam_filter', 'nanopolish_eventalign', 'rseqc_geneBody_coverage', 'purge_dups', 'picard_MarkDuplicates', 'bandage_image', 'rseqc_junction_annotation', 'pilon', 'bamFilter', 'ivar_trim', 'collapse_dataset', 'join1', 'bamtools_split', 'hypo', 'qualimap_rnaseq', 'picard_MergeSamFiles', 'bedtools_genomecoveragebed', 'rseqc_RNA_fragment_size', 'samtools_stats', 'racon', 'rseqc_FPKM_count', 'deeptools_bam_coverage', 'cufflinks', 'rseqc_read_distribution', 'bedtools_bamtofastq', 'nanopolish_variants', 'samtool_filter2', 'ivar_variants', 'samtools_view', 'samtools_markdup', 'featurecounts', 'samtools_depth', '__APPLY_RULES__', 'medaka_variant_pipeline', 'stringtie', 'samtools_sort', 'sam_merge2', 'bedtools_genomecoveragebed_histogram', 'rseqc_junction_saturation', 'deeptools_multi_bam_summary', 'samtools_slice_bam', 'fastqc', 'deeptools_plot_coverage', 'samtools_flagstat', 'ivar_consensus', 'jvarkit_wgscoverageplotter', 'pe_histogram', 'salmon', 'bam_to_sam', 'samtools_merge', 'bamtools', 'rseqc_clipping_profile', 'samtools_mpileup', 'sam_pileup', 'lofreq_call', 'Cut1', 'jbrowse', 'samtools_idxstats', 'rseqc_infer_experiment', 'tabular_to_fastq', 'rseqc_mismatch_profile', 'pycoqc', 'samtools_calmd', 'qualimap_bamqc', 'freebayes', 'samtools_fastx', 'nanoplot', 'CONVERTER_bam_to_bigwig_0', 'bedtools_coveragebed']\n",
      "\n",
      "Test batch 1, Predicted top 77 tools: ['qualimap_rnaseq', 'rseqc_RNA_fragment_size', 'rseqc_FPKM_count', 'deeptools_bam_coverage', 'ivar_consensus', 'samtools_merge', 'pycoqc', 'htseq_count', 'ococo', 'ivar_trim', 'rseqc_read_distribution', 'samtools_markdup', 'samtools_slice_bam', 'rseqc_infer_experiment', 'rseqc_mismatch_profile', 'qualimap_bamqc', 'rseqc_insertion_profile', 'ngsutils_bam_filter', 'bamFilter', 'hypo', 'samtools_stats', 'racon', 'cufflinks', 'nanopolish_variants', 'samtools_sort', 'fastqc', 'bam_to_sam', 'nanopolish_eventalign', 'samtools_view', '__APPLY_RULES__', 'stringtie', 'bedtools_genomecoveragebed_histogram', 'samtools_flagstat', 'jvarkit_wgscoverageplotter', 'rseqc_clipping_profile', 'freebayes', 'CONVERTER_bam_to_bigwig_0', 'rseqc_junction_annotation', 'bamtools_split', 'picard_MergeSamFiles', 'samtool_filter2', 'nanoplot', 'rseqc_geneBody_coverage', 'samtools_depth', 'sam_pileup', 'miniasm', 'picard_MarkDuplicates', 'jbrowse', 'medaka_variant_pipeline', 'pilon', 'rseqc_bam_stat', 'Cut1', 'collapse_dataset', 'samtools_mpileup', 'bedtools_coveragebed', 'salmon', 'rseqc_junction_saturation', 'deeptools_multi_bam_summary', 'medaka_consensus', 'lofreq_call', 'bedtools_bamtofastq', 'bedtools_bamtobed', 'purge_dups', 'ivar_variants', 'bedtools_genomecoveragebed', 'featurecounts', 'samtools_calmd', 'sam_merge2', 'tabular_to_fastq', 'bamtools', 'pe_histogram', 'deeptools_plot_coverage', 'bandage_image', 'join1', 'medaka_variant', 'samtools_idxstats', 'samtools_fastx']\n",
      "\n",
      "Test batch 1, Predicted top 77 tools with weights: ['lofreq_call', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'ivar_trim', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'bandage_image', 'ivar_consensus', 'bamFilter', 'rseqc_infer_experiment', 'collapse_dataset', 'freebayes', 'salmon', 'samtools_flagstat', 'samtools_idxstats', 'bedtools_bamtobed', 'ivar_variants', 'nanoplot', 'rseqc_read_distribution', 'ngsutils_bam_filter', 'bedtools_bamtofastq', 'samtools_sort', 'cufflinks', 'rseqc_geneBody_coverage', 'samtool_filter2', 'samtools_slice_bam', 'pe_histogram', 'medaka_variant', 'racon', 'medaka_consensus', 'bamtools', 'rseqc_bam_stat', 'sam_merge2', 'jbrowse', 'jvarkit_wgscoverageplotter', 'samtools_merge', 'samtools_markdup', 'deeptools_multi_bam_summary', 'samtools_depth', 'purge_dups', 'miniasm', 'tabular_to_fastq', 'picard_MergeSamFiles', 'bedtools_coveragebed', 'sam_pileup', 'deeptools_plot_coverage', 'qualimap_rnaseq', 'pycoqc', 'samtools_calmd', 'pilon', 'samtools_mpileup', 'ococo', 'bamtools_split', 'rseqc_FPKM_count', 'bam_to_sam', 'medaka_variant_pipeline', 'rseqc_junction_annotation', 'bedtools_genomecoveragebed_histogram', 'nanopolish_variants', 'hypo', 'rseqc_junction_saturation', 'rseqc_mismatch_profile', 'nanopolish_eventalign', 'rseqc_clipping_profile', 'rseqc_insertion_profile', 'rseqc_RNA_fragment_size', 'samtools_rmdup', 'rna_star', 'picard_CollectInsertSizeMetrics', 'macs2_callpeak', 'bwa']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.22077922077922077\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.2077922077922078\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'bowtie2', 'macs2_callpeak', 'join1', 'Add_a_column1', 'Cut1', 'Show beginning1', 'Extract genomic DNA 1', 'macs2_callpeak', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'bedtools_intersectbed', 'chipseeker', 'samtools_rmdup']\n",
      "\n",
      "Test batch 1, True tools: ['varscan_somatic', 'rseqc_bam_stat', 'samtools_mpileup', 'sam_pileup', 'gatk2_realigner_target_creator', 'gatk_indel_realigner', 'lofreq_call', 'bedtools_intersectbed', 'deeptools_bamCoverage', 'samtools_stats', 'multiqc', 'bedtools_multicovtbed', 'htseq_count', 'stringtie', 'jbrowse', 'deeptools_bam_compare', 'macs2_callpeak', 'bedtools_bamtobed', 'samtools_sort', 'pileometh', 'deeptools_bam_coverage', 'sam_merge2', 'mimodd_varcall', 'gatk4_mutect2', 'deeptools_multi_bam_summary', 'samtools_idxstats', 'samtools_slice_bam', 'gatk_realigner_target_creator', 'ngsutils_bam_filter', 'fastqc', 'bamleftalign', 'rseqc_infer_experiment', 'peakcalling_macs', 'picard_MarkDuplicates', 'gemini_query', 'cuffdiff', 'samtools_flagstat', 'rseqc_read_distribution', 'samtools_calmd', 'qualimap_bamqc', 'picard_CASM', 'samtool_filter2', 'samtools_view', 'gatk2_indel_realigner', 'freebayes', 'samtools_depth', 'diffbind', 'qualimap_rnaseq', 'picard_CollectInsertSizeMetrics', 'deeptools_plot_fingerprint', 'featurecounts', 'CONVERTER_bam_to_bigwig_0', 'pe_histogram', 'bam_to_sam']\n",
      "\n",
      "Test batch 1, Predicted top 54 tools: ['bedtools_bamtobed', 'qualimap_rnaseq', 'macs2_callpeak', 'deeptools_bam_coverage', 'samtool_filter2', 'sam_merge2', 'samtools_mpileup', 'lofreq_call', 'multiqc', 'gatk2_realigner_target_creator', 'htseq_count', 'cuffdiff', 'picard_CollectInsertSizeMetrics', 'gatk_realigner_target_creator', 'rseqc_read_distribution', 'featurecounts', 'samtools_depth', 'samtools_slice_bam', 'pe_histogram', 'sam_pileup', 'deeptools_bam_compare', 'rseqc_infer_experiment', 'qualimap_bamqc', 'rseqc_bam_stat', 'gatk_indel_realigner', 'bedtools_multicovtbed', 'ngsutils_bam_filter', 'picard_MarkDuplicates', 'gatk2_indel_realigner', 'deeptools_plot_fingerprint', 'bedtools_intersectbed', 'samtools_stats', 'gatk4_mutect2', 'samtools_sort', 'pileometh', 'fastqc', 'picard_CASM', 'bam_to_sam', 'deeptools_bamCoverage', 'samtools_view', 'varscan_somatic', 'stringtie', 'deeptools_multi_bam_summary', 'peakcalling_macs', 'gemini_query', 'samtools_flagstat', 'diffbind', 'mimodd_varcall', 'samtools_idxstats', 'bamleftalign', 'freebayes', 'CONVERTER_bam_to_bigwig_0', 'jbrowse', 'samtools_calmd']\n",
      "\n",
      "Test batch 1, Predicted top 54 tools with weights: ['lofreq_call', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'fastqc', 'featurecounts', 'multiqc', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'bedtools_intersectbed', 'macs2_callpeak', 'rseqc_infer_experiment', 'freebayes', 'samtools_flagstat', 'samtools_idxstats', 'bamleftalign', 'bedtools_bamtobed', 'rseqc_read_distribution', 'ngsutils_bam_filter', 'samtools_sort', 'samtool_filter2', 'samtools_slice_bam', 'deeptools_bam_compare', 'pileometh', 'pe_histogram', 'rseqc_bam_stat', 'sam_merge2', 'jbrowse', 'diffbind', 'picard_CollectInsertSizeMetrics', 'deeptools_multi_bam_summary', 'deeptools_plot_fingerprint', 'samtools_depth', 'picard_CASM', 'cuffdiff', 'sam_pileup', 'qualimap_rnaseq', 'gemini_query', 'mimodd_varcall', 'samtools_calmd', 'samtools_mpileup', 'bedtools_multicovtbed', 'gatk4_mutect2', 'bam_to_sam', 'samtools_rmdup', 'Add_a_column1', 'samtools_merge', 'jvarkit_wgscoverageplotter', 'varscan', 'deeptools_alignmentsieve', 'rna_star', 'stacks2_refmap']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.037037037037037035\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.037037037037037035\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['__FILTER_FROM_FILE__', 'trim_galore', '__UNZIP_COLLECTION__']\n",
      "\n",
      "Test batch 1, True tools: ['dada2_dada', 'bedtools_intersectbed', 'je_demultiplex', 'dada2_learnErrors', '__APPLY_RULES__', 'multiqc', 'fastq_groomer', '__MERGE_COLLECTION__', 'seqtk_sample', 'bg_sortmerna', 'rna_star', 'dada2_mergePairs', 'tp_replace_in_line', 'bwa_mem', 'kraken2', 'fastqc', 'cutadapt', 'fastp', 'flexbar_no_split', '__SORTLIST__', 'fastq_paired_end_interlacer', 'umi_tools_extract', 'dada2_filterAndTrim', 'bismark_bowtie2', 'bowtie2', 'extract_bcs.py', 'collapse_dataset', '__EXTRACT_DATASET__', 'remove_tail.py', 'seqtk_subseq']\n",
      "\n",
      "Test batch 1, Predicted top 30 tools: ['flexbar_no_split', 'extract_bcs.py', 'rna_star', 'dada2_mergePairs', 'dada2_filterAndTrim', 'je_demultiplex', 'fastq_groomer', 'bg_sortmerna', 'bwa_mem', 'remove_tail.py', '__SORTLIST__', 'bismark_bowtie2', 'dada2_learnErrors', 'cutadapt', 'dada2_dada', 'seqtk_sample', 'kraken2', 'fastp', 'fastqc', 'multiqc', '__MERGE_COLLECTION__', 'tp_replace_in_line', 'umi_tools_extract', 'fastq_paired_end_interlacer', 'bedtools_intersectbed', 'bowtie2', 'seqtk_subseq', '__EXTRACT_DATASET__', '__APPLY_RULES__', 'collapse_dataset']\n",
      "\n",
      "Test batch 1, Predicted top 30 tools with weights: ['tp_replace_in_line', 'fastp', 'bwa_mem', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'multiqc', 'kraken2', 'fastq_groomer', 'bedtools_intersectbed', 'collapse_dataset', 'dada2_filterAndTrim', 'fastq_paired_end_interlacer', 'dada2_mergePairs', 'bg_sortmerna', 'dada2_dada', 'seqtk_subseq', 'seqtk_sample', 'umi_tools_extract', 'je_demultiplex', 'dada2_learnErrors', 'bismark_bowtie2', 'tp_sed_tool', 'hisat2', 'trim_galore', 'tp_cat', 'gmx_fep', 'stacks2_denovomap', 'FeatureLinkerUnlabeledQT']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.2\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.2\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'hisat2', 'samtools_flagstat', 'Show beginning1', 'datamash_transpose', 'join1', 'Filter1', 'addValue', 'cat1', 'Cut1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'join_files_on_column_fuzzy', 'regexColumn1', 'collection_column_join', 'plotly_ml_performance_plots', 'bg_uniq', 'venn_list', 'scanpy_read_10x', 'tp_find_and_replace', 'bedtools_slopbed', 'Paste1', 'Show beginning1', 'ggplot_histogram', 'krona-text', 'stacks_denovomap', 'XY_Plot_1', 'random_lines1', 'unipept', 'query_tabular', 'ChangeCase', 'tp_tail_tool', 'collapse_dataset', 'annotatemyids', 'join1', 'goseq', 'heinz_scoring', 'tp_uniq_tool', 'export2graphlan', 'tp_cut_tool', 'seqtk_subseq', 'wc_gnu', 'bedtools_genomecoveragebed', 'wig_to_bigWig', 'clusterprofiler_bitr', 'bedtools_intersectbed', 'crossmap_bed', '__FLATTEN__', 'barchart_gnuplot', 'gtftobed12', 'Remove beginning1', 'macs2_callpeak', 'picard_FilterSamReads', 'pygenomeTracks', 'get_feature_info', 'bedtools_mergebed', 'ggplot_point', 'tp_replace_in_line', 'stringtie_merge', '__FILTER_FROM_FILE__', 'deseq2', 'mergeCols1', 'biotranslator', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'ggplot2_heatmap', 'tab2fasta', 'Grep1', 'msstats', 'table_compute', 'tp_grep_tool', 'egsea', 'add_line_to_file', 'fgsea', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'annotateMyIDs', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'goenrichment', 'regex1', 'tp_multijoin_tool', 'CONVERTER_interval_to_bed_0', 'gops_concat_1', 'bedtools_intersectbed_bam', 'tp_cat', 'charts', 'bcftools_consensus', 'fasterq_dump', 'ggplot2_heatmap2', 'rna_star', 'combine_metaphlan2_humann2', 'stacks_populations', 'Add_a_column1', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'w4mclassfilter', 'get_sequences', 'Kraken2Tax', 'cshl_find_and_replace', 'ont_fast5_api_fast5_subset', 'interactions', 'gops_coverage_1', 'heinz_bum', 'addName', 'mass_spectrometry_imaging_ion_images', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'Show tail1', 'trimmer', 'comp1', 'secure_hash_message_digest', 'volcanoplot', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'circos', 'gops_subtract_1', 'cat1', 'scatterplot_rpy', 'regex_replace', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'histogram_rpy', 'diff', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'interactive_tool_simtext_app', 'sklearn_ensemble', 'addValue', 'dexseq_count', 'cardinal_filtering', 'MassCalculator', 'split_file_to_collection', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'CONVERTER_bed_to_gff_0', 'Krona', 'bedtools_sortbed', 'param_value_from_file', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools: ['tp_easyjoin_tool', 'get_flanks1', 'join_files_on_column_fuzzy', 'ggplot_histogram', 'tp_uniq_tool', 'barchart_gnuplot', 'deseq2', 'mergeCols1', 'Grouping1', 'ggplot2_heatmap', 'annotateMyIDs', 'bedtools_getfastabed', 'tp_sorted_uniq', 'goenrichment', 'rna_star', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'gops_coverage_1', 'comp1', 'regex_replace', '__TAG_FROM_FILE__', 'sklearn_ensemble', 'dexseq_count', 'regexColumn1', 'scanpy_read_10x', 'krona-text', 'heinz_scoring', 'wig_to_bigWig', 'clusterprofiler_bitr', 'crossmap_bed', 'ggplot_point', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'tab2fasta', 'msstats', 'table_compute', 'featurecounts', 'tp_cat', 'Add_a_column1', 'w4mclassfilter', 'cshl_find_and_replace', 'addName', 'mass_spectrometry_imaging_ion_images', 'secure_hash_message_digest', 'MassCalculator', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'param_value_from_file', 'plotly_ml_performance_plots', 'stacks_denovomap', 'random_lines1', 'query_tabular', 'ChangeCase', 'goseq', 'seqtk_subseq', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', '__FLATTEN__', 'gtftobed12', 'picard_FilterSamReads', 'stringtie_merge', 'biotranslator', 'tp_grep_tool', 'egsea', 'tp_replace_in_column', 'seq_filter_by_id', 'fasterq_dump', 'combine_metaphlan2_humann2', 'stacks_populations', 'get_sequences', 'Kraken2Tax', 'ont_fast5_api_fast5_subset', 'heinz_bum', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'sort1', 'tp_awk_tool', 'addValue', 'ncbi_acc_download', 'venn_list', 'Paste1', 'join1', 'tp_cut_tool', 'get_feature_info', 'add_line_to_file', 'fgsea', 'regex1', 'interactions', 'Convert characters1', 'circos', 'cat1', 'scatterplot_rpy', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'diff', 'tp_head_tool', 'interactive_tool_simtext_app', 'CONVERTER_bed_to_gff_0', 'Krona', 'cshl_word_list_grep', 'collection_column_join', 'bg_uniq', 'XY_Plot_1', 'unipept', 'collapse_dataset', 'annotatemyids', 'export2graphlan', 'Grep1', 'tp_find_and_replace', 'tp_tail_tool', 'bcftools_consensus', 'volcanoplot', 'pygenomeTracks', 'bedtools_intersectbed_bam', 'trimmer', 'bedtools_sortbed', 'Summary_Statistics1', 'Show beginning1', 'datamash_ops', 'tp_multijoin_tool', 'Remove beginning1', 'umi_tools_extract', 'hgv_david', 'gops_join_1', 'bedtools_slopbed', 'wc_gnu', 'taxonomy_krona_chart', 'Show tail1', 'tp_sed_tool', 'ggplot2_heatmap2', 'split_file_to_collection', '__FILTER_FROM_FILE__', 'Count1', 'macs2_callpeak', 'histogram_rpy', '__RELABEL_FROM_FILE__', 'deeptools_compute_matrix', 'bg_column_arrange_by_header', 'bedtools_mergebed', 'cardinal_filtering', 'gops_concat_1', 'gops_subtract_1', 'CONVERTER_interval_to_bed_0', 'charts', 'tp_split_on_column']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'tp_replace_in_line', 'datamash_ops', 'bedtools_genomecoveragebed', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'rna_star', 'fasterq_dump', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'volcanoplot', 'tp_grep_tool', 'circos', 'tp_awk_tool', 'annotatemyids', 'taxonomy_krona_chart', 'tp_replace_in_column', 'filter_by_fasta_ids', 'goseq', 'gops_intersect_1', 'collection_column_join', 'limma_voom', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'Kraken2Tax', 'addValue', 'combine_metaphlan2_humann2', 'pygenomeTracks', 'addName', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'regexColumn1', 'seqtk_subseq', 'fgsea', 'XY_Plot_1', 'bedtools_mergebed', 'export2graphlan', 'goenrichment', 'query_tabular', 'bg_uniq', 'krona-text', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'gtftobed12', 'umi_tools_extract', 'stringtie_merge', 'dexseq_count', 'regex1', 'msstats', 'picard_FilterSamReads', 'ggplot2_heatmap', 'mergeCols1', 'tp_split_on_column', 'histogram_rpy', 'ncbi_blastdbcmd_wrapper', 'scatterplot_rpy', 'venn_list', 'egsea', 'join_files_on_column_fuzzy', 'add_line_to_file', 'bedtools_slopbed', 'dropletutils_read_10x', 'plotly_ml_performance_plots', 'get_flanks1', 'tp_uniq_tool', 'regex_replace', 'bg_column_arrange_by_header', 'sklearn_ensemble', 'stacks_denovomap', 'diff', 'gops_coverage_1', 'get_feature_info', 'checkFormat', 'get_sequences', 'heinz_scoring', 'heinz_bum', 'MassCalculator', 'gops_merge_1', 'macs2_bdgdiff', 'uniprot', 'edger', 'chipeakanno_annopeaks', 'deg_annotate', 'xarray_select', 'filter_tabular', 'metilene', 'tabular_to_fastq', 'bedtools_subtractbed', 'chipseeker', 'htseq_count', 'tables_arithmetic_operations', 'raceid_filtnormconf', 'Heatmap', 'cardinal_spectra_plots', 'column_remove_by_header', 'cummeRbund', 'fastq_groomer', 'tbl2gff3', 'macs2_bdgpeakcall', 'bedtools_multiintersectbed', 'plink', 'ctb_chemfp_mol2fps', 'column_order_header_sort', 'scater_create_qcmetric_ready_sce', 'macs2_bdgcmp', 'hicexplorer_hicfindrestrictionsites', 'datamash_reverse', 'multiqc', 'humann2', 'melt', 'idr', 'graphprot_predict_profile', 'cor2', 'intervene_upset', 'gops_cluster_1', 'vegan_diversity', 'stoceps_trend_indic', 'get_pdb', 'intervene_pairwise', 'cardinal_classification', 'climate_stripes', 'subtract_query1', 'flair_correct', 'trim_galore', 'transit_gumbel', 'Fetch Taxonomic Ranks', 'vegan_rarefaction', 'samtools_idxstats', 'salsa', 'sam_merge2', 'sklearn_feature_selection', 'metaspades']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3092105263157895\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.21710526315789475\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'bowtie2', 'macs2_callpeak', 'deeptools_bam_coverage', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'bedtools_intersectbed', 'bedtools_subtractbed', 'Cut1', 'macs2_callpeak', 'diffbind', 'Filter1', 'samtools_rmdup', 'fastqc']\n",
      "\n",
      "Test batch 1, True tools: ['varscan_somatic', 'velvetoptimiser', 'genomespace_exporter', 'ggplot2_point', 'snpEff', 'volcanoplot', 'CONVERTER_gz_to_uncompressed', 'tp_multijoin_tool', 'multiqc', '__FLATTEN__', 'Cut1', 'stringtie', 'regexColumn1', 'tp_cat', '__MERGE_COLLECTION__', 'cshl_fastx_trimmer', 'bcftools_consensus', 'snpSift_annotate', 'fastq_to_fasta_python', 'tp_find_and_replace', 'rna_star', 'bwa_mem', 'stringtie_merge', 'Filter1', 'Add_a_column1', 'deseq2', 'cutadapt', 'bamleftalign', 'diff', 'bwa', '__UNZIP_COLLECTION__', 'export_remote', 'samtools_calmd', 'bowtie2', 'bamFilter', 'Grep1', 'unicycler', 'ivar_variants', 'freebayes', 'gemini_annotate', 'join1', 'deg_annotate', 'picard_NormalizeFasta', 'featurecounts', 'NFS_transfer', 'hisat2', 'create_opson_pipeline_summary']\n",
      "\n",
      "Test batch 1, Predicted top 47 tools: ['genomespace_exporter', '__UNZIP_COLLECTION__', 'create_opson_pipeline_summary', 'samtools_calmd', 'tp_find_and_replace', 'bwa', 'featurecounts', 'Add_a_column1', 'snpSift_annotate', 'picard_NormalizeFasta', 'velvetoptimiser', 'join1', 'varscan_somatic', 'Filter1', 'bamleftalign', 'diff', 'CONVERTER_gz_to_uncompressed', 'cshl_fastx_trimmer', 'volcanoplot', 'gemini_annotate', 'bowtie2', 'stringtie', 'cutadapt', 'deseq2', 'multiqc', 'hisat2', 'bwa_mem', 'bamFilter', 'stringtie_merge', 'unicycler', 'rna_star', 'deg_annotate', 'freebayes', '__MERGE_COLLECTION__', 'export_remote', 'Grep1', 'ivar_variants', 'NFS_transfer', 'tp_cat', 'tp_multijoin_tool', 'Cut1', 'ggplot2_point', '__FLATTEN__', 'fastq_to_fasta_python', 'bcftools_consensus', 'snpEff', 'regexColumn1']\n",
      "\n",
      "Test batch 1, Predicted top 47 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'bwa_mem', 'bcftools_consensus', 'featurecounts', 'rna_star', 'cutadapt', 'bowtie2', 'hisat2', 'multiqc', 'stringtie', 'fastq_to_fasta_python', 'deseq2', 'tp_cat', 'bamFilter', 'unicycler', 'freebayes', 'snpEff', 'bamleftalign', 'volcanoplot', 'deg_annotate', 'ivar_variants', 'cshl_fastx_trimmer', 'bwa', 'ggplot2_point', 'regexColumn1', 'stringtie_merge', 'velvetoptimiser', 'picard_NormalizeFasta', 'samtools_calmd', 'snpSift_annotate', 'diff', 'ncbi_blastn_wrapper', 'fastqc', 'jbrowse', 'collapse_dataset', 'tp_tail_tool', 'bedtools_intersectbed', 'tp_sort_header_tool', 'ecology_presence_abs_abund', 'tp_text_file_with_recurring_lines', 'samtools_mpileup', 'macs2_callpeak', 'gemini_query', 'fasta2tab', 'datamash_transpose', 'kraken2']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0425531914893617\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.02127659574468085\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['qiime_validate_mapping_file', 'qiime_split_libraries', 'qiime_pick_otus', 'qiime_pick_rep_set', 'qiime_assign_taxonomy', 'qiime_make_otu_table']\n",
      "\n",
      "Test batch 1, True tools: ['qiime_make_otu_heatmap', 'qiime_summarize_taxa_through_plots']\n",
      "\n",
      "Test batch 1, Predicted top 2 tools: ['qiime_make_otu_heatmap', 'qiime_summarize_taxa_through_plots']\n",
      "\n",
      "Test batch 1, Predicted top 2 tools with weights: ['qiime_make_otu_heatmap', 'qiime_summarize_taxa_through_plots']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'fastqc', 'Grep1', 'Add_a_column1', 'addValue', 'cat1', 'cast', 'comp1', 'wc_gnu', 'collapse_dataset', 'Convert characters1', 'Grouping1', 'join1', 'Cut1', 'collection_column_join', 'regex_replace', 'melt', 'Show beginning1']\n",
      "\n",
      "Test batch 1, True tools: ['ctb_ob_genProp', 'Show tail1', 'comp1', 'cardinal_spectra_plots', 'volcanoplot', 'tp_easyjoin_tool', 'Cut1', 'tp_cat', 'collection_column_join', 'barchart_gnuplot', 'Remove beginning1', 'cat1', 'cardinal_mz_images', 'plotly_ml_performance_plots', 'column_remove_by_header', 'sam_merge2', 'Extract genomic DNA 1', 'tp_find_and_replace', 'rna_star', 'ggplot2_heatmap2', 'Add_a_column1', 'fastqc', 'deseq2', 'Paste1', '__FILTER_FROM_FILE__', 'cutadapt', 'ctb_multi_obgrep', 'datamash_transpose', '__BUILD_LIST__', 'tp_sort_header_tool', 'random_lines1', 'ctb_chemfp_mol2fps', 'ctb_subsearch', 'ncbi_blastn_wrapper', 'bowtie2', 'tab2fasta', 'tables_arithmetic_operations', 'sklearn_ensemble', 'table_compute', 'addValue', 'join1', 'tophat', 'tp_replace_in_column', 'param_value_from_file', 'hgv_david', 'tp_cut_tool', 'Convert characters1']\n",
      "\n",
      "Test batch 1, Predicted top 47 tools: ['tables_arithmetic_operations', 'datamash_transpose', 'comp1', 'ctb_chemfp_mol2fps', 'plotly_ml_performance_plots', 'ctb_ob_genProp', 'Paste1', 'join1', 'tophat', 'cutadapt', 'ctb_subsearch', 'Convert characters1', 'ctb_multi_obgrep', 'Add_a_column1', 'random_lines1', 'tab2fasta', 'param_value_from_file', 'Cut1', 'cardinal_mz_images', 'tp_cat', 'collection_column_join', 'deseq2', 'rna_star', 'sklearn_ensemble', 'cardinal_spectra_plots', 'cat1', 'Remove beginning1', '__FILTER_FROM_FILE__', 'tp_cut_tool', 'barchart_gnuplot', 'tp_sort_header_tool', 'addValue', 'Extract genomic DNA 1', 'sam_merge2', 'tp_replace_in_column', 'bowtie2', 'volcanoplot', 'hgv_david', 'table_compute', 'column_remove_by_header', '__BUILD_LIST__', 'tp_find_and_replace', 'ggplot2_heatmap2', 'tp_easyjoin_tool', 'Show tail1', 'ncbi_blastn_wrapper', 'fastqc']\n",
      "\n",
      "Test batch 1, Predicted top 47 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'ncbi_blastn_wrapper', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'tp_cat', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'volcanoplot', 'tp_replace_in_column', 'collection_column_join', 'addValue', 'Extract genomic DNA 1', 'sam_merge2', 'cardinal_mz_images', 'cardinal_spectra_plots', 'plotly_ml_performance_plots', 'column_remove_by_header', 'sklearn_ensemble', 'ctb_chemfp_mol2fps', 'tables_arithmetic_operations', 'ctb_multi_obgrep', 'bedtools_intersectbed', 'regexColumn1', 'datamash_ops', 'gtftobed12', 'collapse_dataset', 'cor2', 'bg_uniq', 'macs2_callpeak', 'tp_sed_tool', 'seq_filter_by_id', 'trimmomatic', 'tp_awk_tool', 'gmx_solvate', 'circos', 'addName', 'ggplot2_point', 'fasta_filter_by_length', 'combine_metaphlan2_humann2']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.1276595744680851\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.10638297872340426\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'samtools_rmdup', 'macs2_callpeak', 'macs2_callpeak', 'diffbind', 'Filter1', 'Extract genomic DNA 1', 'macs2_callpeak', 'deeptools_bam_coverage', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'bedtools_subtractbed', 'join1']\n",
      "\n",
      "Test batch 1, True tools: ['datamash_ops', 'bg_column_arrange_by_header', 'trimmer', 'ggplot2_point', 'comp1', 'volcanoplot', 'datamash_reverse', 'taxonomy_krona_chart', 'get_flanks1', 'Cut1', 'tp_cat', 'cat1', 'column_remove_by_header', 'bg_uniq', 'regex_replace', 'Extract genomic DNA 1', 'ggplot2_heatmap2', 'tp_replace_in_line', 'Filter1', 'Add_a_column1', 'Show beginning1', 'tabular_to_fastq', 'mergeCols1', '__BUILD_LIST__', 'tp_sort_header_tool', 'Grouping1', 'Count1', 'sort1', 'tp_awk_tool', 'ggplot2_heatmap', 'tab2fasta', 'collapse_dataset', 'table_compute', 'addValue', 'annotatemyids', 'raceid_filtnormconf', 'goseq', 'deg_annotate', 'tp_replace_in_column', 'tp_cut_tool', 'create_opson_pipeline_summary']\n",
      "\n",
      "Test batch 1, Predicted top 41 tools: ['create_opson_pipeline_summary', 'datamash_reverse', 'raceid_filtnormconf', 'tp_replace_in_column', 'collapse_dataset', 'Filter1', 'tp_sort_header_tool', 'deg_annotate', 'column_remove_by_header', 'Cut1', 'get_flanks1', 'tp_cut_tool', 'sort1', 'tabular_to_fastq', 'cat1', 'tp_awk_tool', 'trimmer', 'tp_cat', 'addValue', 'volcanoplot', 'ggplot2_point', 'bg_column_arrange_by_header', 'Count1', 'Show beginning1', 'annotatemyids', 'Extract genomic DNA 1', 'goseq', 'datamash_ops', 'regex_replace', 'Add_a_column1', 'bg_uniq', 'mergeCols1', 'comp1', 'tp_replace_in_line', 'ggplot2_heatmap', 'taxonomy_krona_chart', 'Grouping1', '__BUILD_LIST__', 'table_compute', 'ggplot2_heatmap2', 'tab2fasta']\n",
      "\n",
      "Test batch 1, Predicted top 41 tools with weights: ['Add_a_column1', 'tp_replace_in_line', 'datamash_ops', 'tp_sort_header_tool', 'tp_cat', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'volcanoplot', 'tp_awk_tool', 'annotatemyids', 'deg_annotate', 'taxonomy_krona_chart', 'tp_replace_in_column', 'goseq', 'addValue', 'datamash_reverse', 'Extract genomic DNA 1', 'ggplot2_point', 'bg_uniq', 'tabular_to_fastq', 'ggplot2_heatmap', 'mergeCols1', 'get_flanks1', 'column_remove_by_header', 'regex_replace', 'bg_column_arrange_by_header', 'raceid_filtnormconf', 'featurecounts', 'add_line_to_file', 'tp_sed_tool', 'deeptools_compute_matrix', 'collection_column_join', 'gprofiler_gost', 'datamash_transpose', 'tp_easyjoin_tool', 'tp_grep_tool', 'freebayes', 'bedtools_intersectbed', 'htseq_count']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.14634146341463414\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.07317073170731707\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['__UNZIP_COLLECTION__', 'seqtk_sample', 'trim_galore', 'bowtie2', 'picard_MarkDuplicates']\n",
      "\n",
      "Test batch 1, True tools: ['gatk2_realigner_target_creator', 'gatk_indel_realigner', 'picard_AddOrReplaceReadGroups', 'htseq_count', 'control_freec', '__MERGE_COLLECTION__', 'deeptools_alignmentsieve', 'bedtools_bamtobed', 'snpEff_build_gb', 'ngsutils_bam_filter', 'rseqc_geneBody_coverage', 'gatk2_base_recalibrator', 'basil', 'samtools_filter', 'delly_call', 'pilon', 'CONVERTER_bam_to_qname_sorted_bam', 'bamFilter', 'rseqc_read_duplication', 'gatk_table_recalibration', 'gatk2_indel_realigner', 'picard_CollectInsertSizeMetrics', 'deeptools_plot_fingerprint', 'picard_MergeSamFiles', 'NFS_transfer', 'gatk2_unified_genotyper', 'bedtools_genomecoveragebed', 'bcftools_mpileup', 'picard_CollectGcBiasMetrics', 'deeptools_bamCoverage', 'bedtools_intersectbed', 'samtools_stats', '__FLATTEN__', 'picard_FixMateInformation', 'macs2_callpeak', 'deeptools_bam_coverage', 'genrich', 'delly_cnv', 'gatk_count_covariates', 'gatk4_mutect2', 'gatk_realigner_target_creator', 'picard_CollectHsMetrics', 'naive_variant_caller', 'rseqc_read_distribution', 'samtools_rmdup', 'Grep1', 'picard_SortSam', 'samtool_filter2', 'samtools_view', 'featurecounts', 'igvtools_count', 'samtools_depth', 'lofreq_viterbi', 'stringtie', 'samtools_sort', 'deeptools_bam_pe_fragmentsize', 'pileometh', 'gatk2_haplotype_caller', 'strelka_germline', 'freebayes_wrapper', 'sam_merge2', 'bedtools_genomecoveragebed_histogram', 'deeptools_multi_bam_summary', 'fastqc', 'deeptools_plot_coverage', 'exomedepth', 'samtools_flagstat', 'picard_CASM', 'diffbind', 'picard_CleanSam', 'gatk2_print_reads', 'pe_histogram', 'picard_ReorderSam', 'bam_to_sam', 'macs2_predictd', 'samtools_merge', 'bamtools', 'samtools_mpileup', 'sam_pileup', 'lofreq_call', 'multiqc', 'gatk_unified_genotyper', 'deeptools_bam_compare', 'lofreq_alnqual', 'mimodd_varcall', 'samtools_idxstats', 'bamleftalign', 'rseqc_infer_experiment', 'mimodd_delcall', 'qualimap_bamqc', 'modencode_peakcalling_spp', 'freebayes', 'picard_MergeBamAlignment', 'bedtools_genomecoveragebed_bedgraph', 'bedtools_coveragebed']\n",
      "\n",
      "Test batch 1, Predicted top 95 tools: ['control_freec', 'bedtools_bamtobed', 'gatk2_base_recalibrator', 'samtools_filter', 'pilon', 'bcftools_mpileup', 'picard_CollectGcBiasMetrics', 'macs2_callpeak', 'deeptools_bam_coverage', 'naive_variant_caller', 'samtools_rmdup', 'macs2_predictd', 'samtools_merge', 'lofreq_call', 'multiqc', 'lofreq_alnqual', 'bedtools_genomecoveragebed_bedgraph', 'bedtools_coveragebed', 'gatk2_realigner_target_creator', 'htseq_count', 'picard_CollectInsertSizeMetrics', 'rseqc_read_distribution', 'picard_SortSam', 'featurecounts', 'samtools_depth', 'sam_pileup', 'deeptools_bam_compare', 'rseqc_infer_experiment', 'qualimap_bamqc', 'gatk_indel_realigner', '__MERGE_COLLECTION__', 'ngsutils_bam_filter', 'deeptools_plot_fingerprint', 'bedtools_intersectbed', 'samtools_stats', '__FLATTEN__', 'gatk4_mutect2', 'lofreq_viterbi', 'fastqc', 'picard_CASM', 'picard_CleanSam', 'picard_ReorderSam', 'bam_to_sam', 'mimodd_delcall', 'picard_MergeBamAlignment', 'stringtie', 'samtools_flagstat', 'diffbind', 'mimodd_varcall', 'samtools_idxstats', 'bamleftalign', 'freebayes', 'picard_MergeSamFiles', 'deeptools_alignmentsieve', 'NFS_transfer', 'samtool_filter2', 'freebayes_wrapper', 'gatk_realigner_target_creator', 'gatk2_indel_realigner', 'picard_FixMateInformation', 'samtools_sort', 'deeptools_bam_pe_fragmentsize', 'gatk_table_recalibration', 'CONVERTER_bam_to_qname_sorted_bam', 'igvtools_count', 'pe_histogram', 'bamtools', 'bedtools_genomecoveragebed', 'exomedepth', 'modencode_peakcalling_spp', 'deeptools_bamCoverage', 'picard_CollectHsMetrics', 'samtools_view', 'bedtools_genomecoveragebed_histogram', 'deeptools_plot_coverage', 'delly_cnv', 'samtools_mpileup', 'rseqc_geneBody_coverage', 'delly_call', 'rseqc_read_duplication', 'pileometh', 'gatk2_print_reads', 'strelka_germline', 'gatk2_unified_genotyper', 'gatk_unified_genotyper', 'deeptools_multi_bam_summary', 'basil', 'snpEff_build_gb', 'sam_merge2', 'genrich', 'bamFilter', 'gatk2_haplotype_caller', 'picard_AddOrReplaceReadGroups', 'gatk_count_covariates', 'Grep1']\n",
      "\n",
      "Test batch 1, Predicted top 95 tools with weights: ['lofreq_call', 'samtools_view', 'samtools_stats', 'qualimap_bamqc', 'lofreq_viterbi', 'bedtools_genomecoveragebed', 'fastqc', 'featurecounts', 'multiqc', 'stringtie', 'htseq_count', 'deeptools_bam_coverage', 'bedtools_intersectbed', 'macs2_callpeak', 'samtools_rmdup', 'bamFilter', 'rseqc_infer_experiment', 'freebayes', 'samtools_flagstat', 'samtools_idxstats', 'bamleftalign', 'bedtools_bamtobed', 'rseqc_read_distribution', 'ngsutils_bam_filter', 'samtools_sort', 'rseqc_geneBody_coverage', 'samtool_filter2', 'deeptools_bam_compare', 'pileometh', 'pe_histogram', 'bcftools_mpileup', 'bamtools', 'picard_SortSam', 'sam_merge2', 'samtools_merge', 'diffbind', 'picard_CollectInsertSizeMetrics', 'deeptools_multi_bam_summary', 'deeptools_plot_fingerprint', 'samtools_depth', 'picard_CASM', 'picard_MergeSamFiles', 'deeptools_alignmentsieve', 'bedtools_genomecoveragebed_bedgraph', 'bedtools_coveragebed', 'sam_pileup', 'genrich', 'deeptools_plot_coverage', 'naive_variant_caller', 'delly_call', 'mimodd_varcall', 'snpEff_build_gb', 'pilon', 'samtools_mpileup', 'rseqc_read_duplication', 'gatk4_mutect2', 'bam_to_sam', 'basil', 'picard_MergeBamAlignment', 'bedtools_genomecoveragebed_histogram', 'picard_CollectGcBiasMetrics', 'strelka_germline', 'picard_AddOrReplaceReadGroups', 'exomedepth', 'picard_FixMateInformation', 'lofreq_alnqual', 'delly_cnv', 'control_freec', 'gatk_count_covariates', 'picard_ReorderSam', 'freebayes_wrapper', 'samtools_filter', 'mimodd_delcall', 'jbrowse', 'rna_star', 'picard_MarkDuplicates', 'peakachu', 'picard_DownsampleSam', 'tp_sed_tool', 'snippy_core', 'Add_a_column1', 'rseqc_bam_stat', 'bismark_deduplicate', 'samtools_fixmate', 'jvarkit_wgscoverageplotter', 'medaka_variant', 'stacks2_refmap', 'bedtools_makewindowsbed', 'samtools_slice_bam', 'bedtools_nucbed', 'chipeakanno_annopeaks', 'rseqc_RPKM_count', 'ococo', 'deeptools_plot_enrichment', 'ivar_trim']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.16842105263157894\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.15789473684210525\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['mothur_pre_cluster', 'mothur_remove_seqs', 'mothur_classify_seqs', 'mothur_remove_lineage', 'mothur_remove_rare']\n",
      "\n",
      "Test batch 1, True tools: ['mothur_classify_otu', 'mothur_dist_shared', 'mothur_make_shared', 'mothur_summary_single', 'mothur_make_biom']\n",
      "\n",
      "Test batch 1, Predicted top 5 tools: ['mothur_classify_otu', 'mothur_make_shared', 'mothur_make_biom', 'mothur_dist_shared', 'mothur_summary_single']\n",
      "\n",
      "Test batch 1, Predicted top 5 tools with weights: ['mothur_make_biom', 'mothur_make_shared', 'mothur_classify_otu', 'mothur_summary_single', 'mothur_dist_shared']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.0\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'bowtie2', 'samtools_rmdup', 'deeptools_bam_coverage', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'bedtools_subtractbed', 'Cut1', 'bedtools_intersectbed', 'Extract genomic DNA 1']\n",
      "\n",
      "Test batch 1, True tools: ['meme_dreme', 'meme_chip', 'fasta2tab', 'clustalw', 'deeptools_bam_compare', 'macs2_callpeak', 'EMBOSS: transeq101', 'meme_meme', 'translate_bed_sequences', 'Filter1', 'fasta_merge_files_and_filter_unique_sequences', 'meme_fimo', 'tp_awk_tool', 'ncbi_blastn_wrapper', 'rgweblogo3', 'maker', 'fasta_filter_by_length', 'sed_stream_editor', 'fasta_compute_length', 'viennarna_rnafold', 'graphprot_predict_profile']\n",
      "\n",
      "Test batch 1, Predicted top 21 tools: ['translate_bed_sequences', 'meme_fimo', 'sed_stream_editor', 'meme_dreme', 'maker', 'meme_chip', 'meme_meme', 'deeptools_bam_compare', 'viennarna_rnafold', 'clustalw', 'fasta_merge_files_and_filter_unique_sequences', 'graphprot_predict_profile', 'EMBOSS: transeq101', 'rgweblogo3', 'Filter1', 'ncbi_blastn_wrapper', 'fasta_filter_by_length', 'fasta2tab', 'tp_awk_tool', 'fasta_compute_length', 'macs2_callpeak']\n",
      "\n",
      "Test batch 1, Predicted top 21 tools with weights: ['ncbi_blastn_wrapper', 'macs2_callpeak', 'fasta2tab', 'tp_awk_tool', 'fasta_merge_files_and_filter_unique_sequences', 'deeptools_bam_compare', 'fasta_filter_by_length', 'clustalw', 'fasta_compute_length', 'meme_meme', 'meme_chip', 'meme_fimo', 'viennarna_rnafold', 'maker', 'meme_dreme', 'rgweblogo3', 'graphprot_predict_profile', 'sed_stream_editor', 'translate_bed_sequences', 'tp_cat', 'macs2_bdgpeakcall']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.09523809523809523\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.09523809523809523\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['FeatureFinderMultiplex', 'HighResPrecursorMassCorrector', 'MSGFPlusAdapter', 'PeptideIndexer', 'FalseDiscoveryRate', 'IDFilter', 'IDMapper', 'IDConflictResolver', 'FileFilter']\n",
      "\n",
      "Test batch 1, True tools: ['FileInfo', 'FileMerger', 'xcms-find-peaks', 'MSGFPlusAdapter', 'MapRTTransformer', 'MapAlignerPoseClustering', 'IDConflictResolver', 'MultiplexResolver']\n",
      "\n",
      "Test batch 1, Predicted top 8 tools: ['MapRTTransformer', 'MultiplexResolver', 'MSGFPlusAdapter', 'xcms-find-peaks', 'FileInfo', 'FileMerger', 'IDConflictResolver', 'MapAlignerPoseClustering']\n",
      "\n",
      "Test batch 1, Predicted top 8 tools with weights: ['sansa_annotate', 'tn93_cluster', 'FeatureLinkerUnlabeledQT', 'chira_extract', 'mimodd_reheader', 'gemini_db_info', 'minfi_rset', 'PSMFeatureExtractor']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.25\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.0\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'macs2_callpeak', 'deeptools_compute_matrix', 'deeptools_plot_heatmap', 'Remove beginning1', 'bedtools_subtractbed', 'tp_sort_header_tool', 'Cut1', 'bedtools_intersectbed', 'chipseeker', 'samtools_rmdup', 'deeptools_bam_coverage', 'bowtie2', 'macs2_callpeak', 'diffbind', 'Filter1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'pampa_communitymetrics', 'tp_easyjoin_tool', 'join_files_on_column_fuzzy', 'htseq_count', 'regexColumn1', 'snpSift_filter', 'hgv_snpFreq', 'collection_column_join', 'seq_select_by_id', 'gprofiler_gost', 'bg_uniq', 'venn_list', 'tp_find_and_replace', 'Show beginning1', 'XY_Plot_1', 'changeGffFeatures', 'random_lines1', 'unipept', 'cshl_sort_header', 'ChangeCase', 'uniprot', 'aggregate_scores_in_intervals2', 'tp_tail_tool', 'bedtools_subtractbed', 'collapse_dataset', 'annotatemyids', 'join1', 'deg_annotate', 'tp_cut_tool', 'wc_gnu', 'bedtools_intersectbed', 'cardinal_combine', 'repeat_counter', 'transpose', 'dexseq_annotate', 'Remove beginning1', 'cardinal_mz_images', 'column_remove_by_header', 'bedtools_mergebed', 'tp_replace_in_line', 'deseq2', 'mergeCols1', 'vegan_rarefaction', 'Interval_Maf_Merged_Fasta2', 'tp_sort_header_tool', 'Grouping1', 'pandas_rolling_window', 'naive_variant_caller', 'gops_intersect_1', 'samtools_rmdup', 'Fetch Taxonomic Ranks', 'ggplot2_heatmap', 'trinity', 'tab2fasta', 'Grep1', 'SnpEff-cds-report', 'samtool_filter2', 'table_compute', 'samtools_view', 'tabular_to_csv', 'table_annovar', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'cooler_csort_pairix', '__FILTER_EMPTY_DATASETS__', 'gops_cluster_1', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'filter_tabular', 'goenrichment', 'snpEff', 'cardinal_spectra_plots', 'gdal_ogr2ogr', 'tp_multijoin_tool', 'gops_concat_1', 'tp_cat', 'charts', 'bcftools_consensus', 'vcfbedintersect', 'sam_merge2', 'ggplot2_heatmap2', 'Add_a_column1', 'stoceps_glm', 'datamash_transpose', 'stoceps_trend_indic', 'gops_coverage_1', 'chipseeker', 'snpSift_rmInfo', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'diapysef', 'Show tail1', 'trimmer', 'comp1', 'ggplot2_point', 'volcanoplot', 'multiqc', 'tbl2gff3', 'Cut1', 'CONVERTER_vcf_bgzip_to_tabix_0', 'jbrowse', 'gops_subtract_1', 'minimap2', 'cat1', 'Extract genomic DNA 1', 'tab2csv_R', 'vegan_diversity', 'pathview', 'pampa_glmsp', 'histogram_rpy', '__BUILD_LIST__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'addValue', 'bcftools_norm', 'dexseq_count', 'cardinal_filtering', 'split_file_to_collection', '__EXTRACT_DATASET__', 'samtools_fastx', 'bedtools_sortbed', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'bedtools_maskfastabed', 'fastq_filter', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 135 tools: ['tp_easyjoin_tool', 'Show beginning1', 'Grouping1', 'table_annovar', 'datamash_ops', 'datamash_transpose', 'snpSift_rmInfo', 'tp_find_and_replace', 'changeGffFeatures', 'tp_tail_tool', 'cardinal_combine', 'cardinal_mz_images', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'featurecounts', 'charts', 'bcftools_consensus', 'Add_a_column1', 'random_lines1', 'bedtools_intersectbed', 'vcfbedintersect', 'jbrowse', 'addValue', 'bcftools_norm', 'gprofiler_gost', 'bedtools_subtractbed', 'join1', 'tp_cut_tool', 'gops_concat_1', 'tp_sed_tool', 'Cut1', 'cat1', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'Summary_Statistics1', 'collection_column_join', 'XY_Plot_1', 'unipept', 'annotatemyids', 'Grep1', 'samtool_filter2', 'bedtools_sortbed', 'join_files_on_column_fuzzy', 'mergeCols1', 'samtools_rmdup', 'tp_sorted_uniq', 'bg_column_arrange_by_header', 'snpEff', 'chipseeker', 'ChangeCase', 'Fetch Taxonomic Ranks', 'ggplot2_heatmap2', 'stoceps_trend_indic', 'gops_join_1', 'sort1', 'uniprot', 'collapse_dataset', 'tp_multijoin_tool', 'sam_merge2', 'cardinal_filtering', 'SnpEff-cds-report', 'hgv_david', 'CONVERTER_vcf_bgzip_to_tabix_0', 'column_remove_by_header', 'bedtools_mergebed', 'cshl_word_list_grep', 'bg_uniq', 'naive_variant_caller', '__FILTER_EMPTY_DATASETS__', 'gops_cluster_1', 'goenrichment', 'regexColumn1', 'transpose', 'samtools_fastx', 'minimap2', 'fastq_filter', 'snpSift_filter', 'bedtools_getfastabed', 'gops_coverage_1', 'tbl2gff3', 'venn_list', 'Convert characters1', 'wc_gnu', 'cshl_sort_header', 'repeat_counter', 'dexseq_annotate', 'pandas_rolling_window', 'stoceps_glm', 'hgv_snpFreq', 'cardinal_spectra_plots', 'tp_cat', 'tab2csv_R', 'tp_replace_in_column', 'cooler_csort_pairix', 'diapysef', 'Interval_Maf_Merged_Fasta2', 'gdal_ogr2ogr', 'seq_select_by_id', 'trinity', 'gops_subtract_1', 'dexseq_count', 'bedtools_maskfastabed', 'deg_annotate', 'pathview', 'htseq_count', 'volcanoplot', 'split_file_to_collection', 'tabular_to_csv', 'aggregate_scores_in_intervals2', 'Show tail1', 'histogram_rpy', 'filter_tabular', 'comp1', 'Remove beginning1', 'ggplot2_point', 'ncbi_acc_download', 'deseq2', 'vegan_diversity', 'vegan_rarefaction', 'tp_awk_tool', 'samtools_view', 'ggplot2_heatmap', 'pampa_glmsp', 'multiqc', 'tp_head_tool', 'pampa_communitymetrics', 'table_compute', 'seq_filter_by_id', 'Count1', '__EXTRACT_DATASET__', 'tab2fasta', 'trimmer', '__BUILD_LIST__', 'deeptools_compute_matrix', 'Extract genomic DNA 1']\n",
      "\n",
      "Test batch 1, Predicted top 135 tools with weights: ['Add_a_column1', 'snpSift_filter', 'tp_find_and_replace', 'tp_replace_in_line', 'samtools_view', 'datamash_ops', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'filter_tabular', 'multiqc', 'minimap2', 'htseq_count', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'samtools_rmdup', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'snpEff', 'volcanoplot', 'tp_awk_tool', 'annotatemyids', 'fastq_filter', 'deg_annotate', 'tp_replace_in_column', 'gops_intersect_1', 'collection_column_join', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'samtool_filter2', 'addValue', 'bedtools_maskfastabed', 'trinity', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'sam_merge2', 'jbrowse', 'regexColumn1', 'XY_Plot_1', 'bedtools_mergebed', 'goenrichment', 'bcftools_norm', 'bg_uniq', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'tbl2gff3', 'dexseq_count', 'chipseeker', 'bedtools_subtractbed', 'ggplot2_heatmap', 'mergeCols1', 'histogram_rpy', 'naive_variant_caller', 'venn_list', 'cardinal_mz_images', 'join_files_on_column_fuzzy', 'cardinal_spectra_plots', 'gprofiler_gost', 'vcfbedintersect', 'column_remove_by_header', 'bg_column_arrange_by_header', 'cardinal_combine', 'uniprot', 'vegan_diversity', 'SnpEff-cds-report', 'gops_coverage_1', 'diapysef', 'Fetch Taxonomic Ranks', 'vegan_rarefaction', 'pampa_glmsp', 'stoceps_trend_indic', 'stoceps_glm', 'gops_cluster_1', 'fastqc', 'gtftobed12', 'samtools_slice_bam', 'tp_split_on_column', 'addName', 'filter_by_fasta_ids', 'cutadapt', 'cardinal_classification', 'plotly_ml_performance_plots', 'rna_star', 'goseq', 'picard_FilterSamReads', 'xarray_select', 'eggnog_mapper', 'picard_MarkDuplicates', 'snpfreqplot', 'circos_wiggle_to_stacked', 'cummeRbund', 'chipeakanno_annopeaks', 'taxonomy_krona_chart', 'gops_merge_1', 'metaspades', 'fgsea', 'rcas', 'blockbuster', 'snpSift_extractFields', 'msstats', 'bedtools_windowbed', 'gemini_query', 'egsea', 'regex_replace', 'macs2_bdgdiff', 'ncbi_blastdbcmd_wrapper', 'get_sequences', 'samtools_depth', 'ctb_multi_obgrep', 'regex1', 'kraken2', 'qualimap_bamqc', 'cardinal_quality_report', 'diffbind', 'bam_to_sam', 'heinz_bum', 'transit_gumbel', 'tp_uniq_tool', 'tp_sort_rows', 'deeptools_bam_coverage']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.1925925925925926\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.1111111111111111\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'fastqc', 'Grep1', 'datamash_transpose', 'Remove beginning1', 'addValue', 'join1', 'Cut1', 'Add_a_column1', 'Filter1', 'wc_gnu', 'collapse_dataset']\n",
      "\n",
      "Test batch 1, True tools: ['bed2gff1', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'CONVERTER_gz_to_uncompressed', '__MERGE_COLLECTION__', 'bg_uniq', 'snpEff_build_gb', 'tp_find_and_replace', 'mothur_venn', 'porechop', 'kraken2', 'Paste1', 'XY_Plot_1', 'rbc_mafft', 'random_lines1', 'query_tabular', 'meryl', 'fastq_stats', 'join1', '__ZIP_COLLECTION__', 'sucos_docking_scoring', 'mothur_metastats', 'nextclade', 'bedtools_intersectbed', 'barchart_gnuplot', 'fastq_to_fasta_python', 'tp_replace_in_line', 'stringtie_merge', 'ncbi_makeblastdb', 'quast', 'metaeuk_easy_predict', 'Grouping1', 'xchem_transfs_scoring', 'mothur_classify_rf', 'filter_by_fasta_ids', 'Grep1', 'tp_grep_tool', 'megahit', 'tabular_to_csv', 'add_line_to_file', 'tp_replace_in_column', 'mothur_get_communitytype', 'datamash_ops', 'trimmomatic', 'filter_tabular', 'snpEff', 'fasta2tab', 'bcftools_consensus', 'fastq_to_tabular', 'fasterq_dump', 'bwa_mem', 'Add_a_column1', 'fastqc', 'cutadapt', 'mothur_corr_axes', 'metagene_annotator', 'fastq_paired_end_joiner', 'export_remote', 'circos_aln_to_links', 'pangolin', 'structure_to_gspan', 'replace_column_with_key_value_file', 'openbabel_filter', 'Convert characters1', 'medaka_consensus_pipeline', 'openbabel_compound_convert', 'comp1', 'multiqc', 'Cut1', 'minimap2', 'seqtk_sample', 'metaspades', 'Filter1', 'prinseq', '__RELABEL_FROM_FILE__', 'rdock_sort_filter', 'Count1', 'sort1', 'ncbi_blastn_wrapper', 'unicycler', 'xchem_pose_scoring', 'split_file_to_collection', 'mothur_lefse', 'nanoplot', 'bedtools_maskfastabed', 'fastq_filter', 'hisat2']\n",
      "\n",
      "Test batch 1, Predicted top 87 tools: ['tp_easyjoin_tool', 'snpEff_build_gb', 'barchart_gnuplot', 'metaeuk_easy_predict', 'Grouping1', 'datamash_ops', 'fasta2tab', 'fastq_to_tabular', 'metaspades', '__RELABEL_FROM_FILE__', 'hisat2', 'rbc_mafft', 'xchem_transfs_scoring', 'mothur_classify_rf', 'Add_a_column1', 'fastq_paired_end_joiner', 'mothur_lefse', 'fastq_stats', 'stringtie_merge', 'tp_replace_in_column', 'trimmomatic', 'fastqc', 'mothur_corr_axes', 'sort1', 'mothur_venn', 'Paste1', 'sucos_docking_scoring', 'add_line_to_file', 'filter_tabular', 'metagene_annotator', 'circos_aln_to_links', 'structure_to_gspan', 'replace_column_with_key_value_file', 'openbabel_filter', 'Cut1', 'seqtk_sample', 'Filter1', 'fastq_filter', 'CONVERTER_gz_to_uncompressed', 'bg_uniq', 'XY_Plot_1', '__ZIP_COLLECTION__', 'mothur_metastats', 'bed2gff1', 'fastq_to_fasta_python', 'multiqc', 'bwa_mem', 'random_lines1', 'mothur_get_communitytype', 'cutadapt', 'openbabel_compound_convert', 'porechop', 'export_remote', 'xchem_pose_scoring', 'meryl', 'tp_replace_in_line', 'prinseq', 'bedtools_intersectbed', 'join1', 'Grep1', 'nanoplot', 'pangolin', '__MERGE_COLLECTION__', 'unicycler', 'query_tabular', 'filter_by_fasta_ids', 'medaka_consensus_pipeline', 'fasterq_dump', 'taxonomy_krona_chart', 'bedtools_maskfastabed', 'tp_find_and_replace', 'megahit', 'nextclade', 'comp1', 'rdock_sort_filter', 'minimap2', 'quast', 'Convert characters1', 'ncbi_blastn_wrapper', 'split_file_to_collection', 'snpEff', 'ncbi_makeblastdb', 'tabular_to_csv', 'kraken2', 'bcftools_consensus', 'tp_grep_tool', 'Count1']\n",
      "\n",
      "Test batch 1, Predicted top 87 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'tp_replace_in_line', 'bwa_mem', 'datamash_ops', 'replace_column_with_key_value_file', 'bcftools_consensus', 'fastqc', 'cutadapt', 'trimmomatic', 'hisat2', 'filter_tabular', 'multiqc', 'minimap2', 'kraken2', 'fastq_to_fasta_python', 'ncbi_blastn_wrapper', 'fasterq_dump', 'tp_easyjoin_tool', 'bedtools_intersectbed', 'porechop', 'fasta2tab', 'unicycler', 'snpEff', 'quast', 'tp_grep_tool', 'nextclade', 'fastq_filter', 'taxonomy_krona_chart', 'nanoplot', 'tp_replace_in_column', 'filter_by_fasta_ids', 'split_file_to_collection', 'ncbi_makeblastdb', 'bedtools_maskfastabed', 'pangolin', 'meryl', 'megahit', 'fastq_to_tabular', 'XY_Plot_1', 'fastq_paired_end_joiner', 'seqtk_sample', 'query_tabular', 'bg_uniq', 'rbc_mafft', 'stringtie_merge', 'metaspades', 'medaka_consensus_pipeline', 'fastq_stats', 'mothur_venn', 'add_line_to_file', 'snpEff_build_gb', 'prinseq', 'structure_to_gspan', 'metagene_annotator', 'xchem_transfs_scoring', 'circos_aln_to_links', 'openbabel_filter', 'rdock_sort_filter', 'sucos_docking_scoring', 'metaeuk_easy_predict', 'mothur_corr_axes', 'mothur_get_communitytype', 'mothur_metastats', 'mothur_lefse', 'trinity', 'rna_star', 'histogram_rpy', 'collapse_dataset', 'bowtie2', 'regexColumn1', 'ggplot2_pca', 'tp_sort_header_tool', 'cshl_fastx_trimmer', 'goenrichment', 'regex_replace', 'plink', 'scatterplot_rpy', 'snippy', 'tp_cat', 'tp_awk_tool', 'regex1', 'snpSift_filter', 'column_remove_by_header', 'pathview', 'bamtools_split', 'muscle']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.1724137931034483\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.12643678160919541\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['fastq_groomer', 'cutadapt', 'bg_sortmerna', 'fastq_paired_end_interlacer', 'metaphlan2', 'humann2']\n",
      "\n",
      "Test batch 1, True tools: ['group_humann2_uniref_abundances_to_go', 'humann2_renorm_table', 'humann2_genefamilies_genus_level', 'humann_regroup_table', 'combine_metaphlan2_humann2', 'humann2_regroup_table', 'humann2_split_table', 'humann_renorm_table', 'Grep1']\n",
      "\n",
      "Test batch 1, Predicted top 9 tools: ['Grep1', 'humann2_genefamilies_genus_level', 'combine_metaphlan2_humann2', 'humann_regroup_table', 'humann2_regroup_table', 'group_humann2_uniref_abundances_to_go', 'humann_renorm_table', 'humann2_split_table', 'humann2_renorm_table']\n",
      "\n",
      "Test batch 1, Predicted top 9 tools with weights: ['combine_metaphlan2_humann2', 'humann2_renorm_table', 'group_humann2_uniref_abundances_to_go', 'humann2_genefamilies_genus_level', 'humann_renorm_table', 'humann_regroup_table', 'humann2_regroup_table', 'humann2_split_table', 'humann2_join_tables']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.5555555555555556\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.4444444444444444\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['cutadapt', 'umi_tools_extract', 'rna_star', 'samtools_flagstat', 'tp_awk_tool', 'sam_to_bam']\n",
      "\n",
      "Test batch 1, True tools: ['pureclip', 'bamtools', 'samtools_mpileup', 'sam_pileup', 'samtools_fixmate', 'CONVERTER_Bam_Bai_0', 'filter', 'macs2_callpeak', 'samtools_sort', 'bedtools_bamtobed', 'pileometh', 'rgPicardMarkDups', 'deeptools_bam_coverage', 'picard_ARRG', 'bctools_extract_alignment_ends', 'ngsutils_bam_filter', 'bamleftalign', 'peakcalling_macs', 'cuffdiff', 'samtools_flagstat', 'samtools_rmdup', 'deeptools_compute_matrix', 'bedtools_bamtofastq', 'pilon', 'picard_SortSam', 'samtools_split', 'freebayes', 'umi_tools_dedup', 'bam_to_consensus', 'picard_ReorderSam']\n",
      "\n",
      "Test batch 1, Predicted top 30 tools: ['picard_ARRG', 'samtools_rmdup', 'samtools_mpileup', 'picard_ReorderSam', 'samtools_fixmate', 'samtools_flagstat', 'bamleftalign', 'freebayes', 'CONVERTER_Bam_Bai_0', 'deeptools_compute_matrix', 'samtools_split', 'deeptools_bam_coverage', 'rgPicardMarkDups', 'bedtools_bamtofastq', 'pileometh', 'pilon', 'bam_to_consensus', 'filter', 'bctools_extract_alignment_ends', 'macs2_callpeak', 'bedtools_bamtobed', 'picard_SortSam', 'sam_pileup', 'pureclip', 'peakcalling_macs', 'umi_tools_dedup', 'ngsutils_bam_filter', 'samtools_sort', 'cuffdiff', 'bamtools']\n",
      "\n",
      "Test batch 1, Predicted top 30 tools with weights: ['deeptools_bam_coverage', 'macs2_callpeak', 'samtools_rmdup', 'deeptools_compute_matrix', 'freebayes', 'samtools_flagstat', 'bamleftalign', 'bedtools_bamtobed', 'ngsutils_bam_filter', 'bedtools_bamtofastq', 'samtools_sort', 'pileometh', 'bamtools', 'picard_SortSam', 'cuffdiff', 'sam_pileup', 'umi_tools_dedup', 'pureclip', 'samtools_fixmate', 'pilon', 'samtools_mpileup', 'samtools_split', 'bctools_extract_alignment_ends', 'picard_ReorderSam', 'picard_ARRG', 'rgPicardMarkDups', 'featurecounts', 'samtools_idxstats', 'cufflinks', 'htseq_count']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.03333333333333333\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.03333333333333333\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'fastqc', 'Grep1', 'Add_a_column1', 'wc_gnu', 'Remove beginning1', 'datamash_ops', 'join1', 'Filter1', 'addValue', 'Grouping1', 'cat1', 'melt', 'Show beginning1', 'Cut1']\n",
      "\n",
      "Test batch 1, True tools: ['bg_column_arrange_by_header', 'cshl_word_list_grep', 'taxonomy_krona_chart', 'tp_easyjoin_tool', 'tp_split_on_column', 'get_flanks1', 'join_files_on_column_fuzzy', 'regexColumn1', 'collection_column_join', 'plotly_ml_performance_plots', 'bg_uniq', 'venn_list', 'scanpy_read_10x', 'tp_find_and_replace', 'bedtools_slopbed', 'Paste1', 'Show beginning1', 'ggplot_histogram', 'krona-text', 'stacks_denovomap', 'XY_Plot_1', 'random_lines1', 'unipept', 'query_tabular', 'ChangeCase', 'tp_tail_tool', 'collapse_dataset', 'annotatemyids', 'join1', 'goseq', 'heinz_scoring', 'tp_uniq_tool', 'export2graphlan', 'tp_cut_tool', 'seqtk_subseq', 'wc_gnu', 'bedtools_genomecoveragebed', 'wig_to_bigWig', 'clusterprofiler_bitr', 'bedtools_intersectbed', 'crossmap_bed', '__FLATTEN__', 'barchart_gnuplot', 'gtftobed12', 'Remove beginning1', 'macs2_callpeak', 'picard_FilterSamReads', 'pygenomeTracks', 'get_feature_info', 'bedtools_mergebed', 'ggplot_point', 'tp_replace_in_line', 'stringtie_merge', '__FILTER_FROM_FILE__', 'deseq2', 'mergeCols1', 'biotranslator', 'umi_tools_extract', 'tp_sort_header_tool', 'Grouping1', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'ggplot2_heatmap', 'tab2fasta', 'Grep1', 'msstats', 'table_compute', 'tp_grep_tool', 'egsea', 'add_line_to_file', 'fgsea', 'tp_replace_in_column', 'featurecounts', 'seq_filter_by_id', 'annotateMyIDs', 'bedtools_getfastabed', 'datamash_ops', 'tp_sorted_uniq', 'goenrichment', 'regex1', 'tp_multijoin_tool', 'CONVERTER_interval_to_bed_0', 'gops_concat_1', 'bedtools_intersectbed_bam', 'tp_cat', 'charts', 'bcftools_consensus', 'fasterq_dump', 'ggplot2_heatmap2', 'rna_star', 'combine_metaphlan2_humann2', 'stacks_populations', 'Add_a_column1', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'w4mclassfilter', 'get_sequences', 'Kraken2Tax', 'cshl_find_and_replace', 'ont_fast5_api_fast5_subset', 'interactions', 'gops_coverage_1', 'heinz_bum', 'addName', 'mass_spectrometry_imaging_ion_images', 'hgv_david', 'tp_sed_tool', 'Convert characters1', 'Show tail1', 'trimmer', 'comp1', 'secure_hash_message_digest', 'volcanoplot', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'circos', 'gops_subtract_1', 'cat1', 'scatterplot_rpy', 'regex_replace', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'histogram_rpy', 'diff', '__RELABEL_FROM_FILE__', '__TAG_FROM_FILE__', 'gops_join_1', 'Count1', 'sort1', 'deeptools_compute_matrix', 'tp_awk_tool', 'tp_head_tool', 'interactive_tool_simtext_app', 'sklearn_ensemble', 'addValue', 'dexseq_count', 'cardinal_filtering', 'MassCalculator', 'split_file_to_collection', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'CONVERTER_bed_to_gff_0', 'Krona', 'bedtools_sortbed', 'param_value_from_file', 'ncbi_acc_download', 'Summary_Statistics1']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools: ['tp_easyjoin_tool', 'get_flanks1', 'join_files_on_column_fuzzy', 'ggplot_histogram', 'tp_uniq_tool', 'barchart_gnuplot', 'deseq2', 'mergeCols1', 'Grouping1', 'ggplot2_heatmap', 'annotateMyIDs', 'bedtools_getfastabed', 'tp_sorted_uniq', 'goenrichment', 'rna_star', 'dropletutils_read_10x', 'checkFormat', 'datamash_transpose', 'openbabel_remIons', 'gops_coverage_1', 'comp1', 'regex_replace', '__TAG_FROM_FILE__', 'sklearn_ensemble', 'dexseq_count', 'regexColumn1', 'scanpy_read_10x', 'krona-text', 'tp_tail_tool', 'heinz_scoring', 'wig_to_bigWig', 'clusterprofiler_bitr', 'crossmap_bed', 'Remove beginning1', 'ggplot_point', 'tp_replace_in_line', 'tp_sort_header_tool', 'gops_intersect_1', 'filter_by_fasta_ids', 'fa-extract-sequence', 'tab2fasta', 'msstats', 'table_compute', 'featurecounts', 'tp_cat', 'Add_a_column1', 'w4mclassfilter', 'cshl_find_and_replace', 'addName', 'mass_spectrometry_imaging_ion_images', 'secure_hash_message_digest', 'MassCalculator', 'gtf_filter_by_attribute_values_list', 'limma_voom', 'param_value_from_file', 'plotly_ml_performance_plots', 'stacks_denovomap', 'random_lines1', 'query_tabular', 'ChangeCase', 'goseq', 'seqtk_subseq', 'bedtools_genomecoveragebed', 'bedtools_intersectbed', '__FLATTEN__', 'gtftobed12', 'picard_FilterSamReads', 'stringtie_merge', 'biotranslator', 'tp_grep_tool', 'egsea', 'tp_replace_in_column', 'seq_filter_by_id', 'fasterq_dump', 'combine_metaphlan2_humann2', 'stacks_populations', 'get_sequences', 'Kraken2Tax', 'ont_fast5_api_fast5_subset', 'heinz_bum', 'ggplot2_point', 'ncbi_blastdbcmd_wrapper', 'iReport', 'gops_join_1', 'sort1', 'tp_awk_tool', 'addValue', 'ncbi_acc_download', 'bedtools_slopbed', 'Paste1', 'join1', 'tp_cut_tool', 'get_feature_info', 'add_line_to_file', 'fgsea', 'regex1', 'interactions', 'circos', 'cat1', 'scatterplot_rpy', 'Extract genomic DNA 1', 'Filter1', 'pathview', 'diff', 'tp_head_tool', 'interactive_tool_simtext_app', 'CONVERTER_bed_to_gff_0', 'Krona', 'Summary_Statistics1', 'cshl_word_list_grep', 'collection_column_join', 'bg_uniq', 'XY_Plot_1', 'unipept', 'collapse_dataset', 'annotatemyids', 'export2graphlan', 'taxonomy_krona_chart', 'Show beginning1', 'tp_find_and_replace', 'bcftools_consensus', 'hgv_david', 'split_file_to_collection', 'venn_list', 'pygenomeTracks', 'trimmer', 'macs2_callpeak', 'Grep1', 'tp_multijoin_tool', 'Show tail1', 'volcanoplot', '__FILTER_FROM_FILE__', 'bedtools_intersectbed_bam', 'tp_sed_tool', 'Convert characters1', 'bedtools_sortbed', '__RELABEL_FROM_FILE__', 'umi_tools_extract', 'Count1', 'wc_gnu', 'datamash_ops', 'deeptools_compute_matrix', 'ggplot2_heatmap2', 'histogram_rpy', 'bg_column_arrange_by_header', 'bedtools_mergebed', 'CONVERTER_interval_to_bed_0', 'gops_subtract_1', 'cardinal_filtering', 'gops_concat_1', 'charts', 'tp_split_on_column']\n",
      "\n",
      "Test batch 1, Predicted top 152 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'tp_replace_in_line', 'datamash_ops', 'bedtools_genomecoveragebed', 'gops_concat_1', 'bcftools_consensus', 'gops_subtract_1', 'featurecounts', 'rna_star', 'fasterq_dump', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'tp_cat', 'deeptools_compute_matrix', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'collapse_dataset', 'volcanoplot', 'tp_grep_tool', 'circos', 'tp_awk_tool', 'annotatemyids', 'taxonomy_krona_chart', 'tp_replace_in_column', 'filter_by_fasta_ids', 'goseq', 'gops_intersect_1', 'collection_column_join', 'limma_voom', 'split_file_to_collection', 'tp_head_tool', 'tp_sed_tool', 'Kraken2Tax', 'addValue', 'combine_metaphlan2_humann2', 'pygenomeTracks', 'addName', 'seq_filter_by_id', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'ncbi_acc_download', 'bedtools_sortbed', 'tp_tail_tool', 'ggplot2_point', 'regexColumn1', 'seqtk_subseq', 'fgsea', 'XY_Plot_1', 'bedtools_mergebed', 'export2graphlan', 'goenrichment', 'query_tabular', 'bg_uniq', 'krona-text', 'tp_sorted_uniq', 'cardinal_filtering', 'pathview', 'gtftobed12', 'umi_tools_extract', 'stringtie_merge', 'dexseq_count', 'regex1', 'msstats', 'picard_FilterSamReads', 'ggplot2_heatmap', 'mergeCols1', 'tp_split_on_column', 'histogram_rpy', 'ncbi_blastdbcmd_wrapper', 'scatterplot_rpy', 'venn_list', 'egsea', 'join_files_on_column_fuzzy', 'add_line_to_file', 'bedtools_slopbed', 'dropletutils_read_10x', 'plotly_ml_performance_plots', 'get_flanks1', 'tp_uniq_tool', 'regex_replace', 'bg_column_arrange_by_header', 'sklearn_ensemble', 'stacks_denovomap', 'diff', 'gops_coverage_1', 'get_feature_info', 'checkFormat', 'get_sequences', 'heinz_scoring', 'heinz_bum', 'MassCalculator', 'gops_merge_1', 'macs2_bdgdiff', 'uniprot', 'edger', 'chipeakanno_annopeaks', 'xarray_select', 'deg_annotate', 'metilene', 'tabular_to_fastq', 'filter_tabular', 'bedtools_subtractbed', 'htseq_count', 'raceid_filtnormconf', 'tables_arithmetic_operations', 'Heatmap', 'chipseeker', 'cardinal_spectra_plots', 'column_remove_by_header', 'fastq_groomer', 'cummeRbund', 'macs2_bdgpeakcall', 'tbl2gff3', 'plink', 'humann2', 'scater_create_qcmetric_ready_sce', 'multiqc', 'macs2_bdgcmp', 'datamash_reverse', 'vegan_diversity', 'column_order_header_sort', 'gops_cluster_1', 'bedtools_multiintersectbed', 'melt', 'graphprot_predict_profile', 'idr', 'get_pdb', 'intervene_upset', 'hicexplorer_hicfindrestrictionsites', 'intervene_pairwise', 'rseqc_infer_experiment', 'stoceps_trend_indic', 'vegan_rarefaction', 'cor2', 'ctb_chemfp_mol2fps', 'samtools_idxstats', 'Fetch Taxonomic Ranks', 'sklearn_feature_selection', 'sam_merge2', 'sklearn_svm_classifier', 'climate_stripes', 'transit_gumbel', 'flair_correct', 'subtract_query1', 'kc-align', 'ggplot2_pca']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.3092105263157895\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.21710526315789475\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trim_galore', 'samtools_rmdup', 'deeptools_bam_coverage', 'bowtie2', 'macs2_callpeak', 'Remove beginning1', 'bedtools_subtractbed', 'tp_sort_header_tool', 'tp_head_tool', 'Extract genomic DNA 1', 'macs2_callpeak']\n",
      "\n",
      "Test batch 1, True tools: ['wig_to_bigWig', 'bed2gff1', 'comp1', 'bedtools_intersectbed', 'crossmap_bed', 'tp_easyjoin_tool', 'multiqc', 'gops_concat_1', 'Cut1', 'jbrowse', 'tp_cat', 'intervene_pairwise', 'pygenomeTracks', 'Remove beginning1', 'macs2_callpeak', 'cat1', 'deeptools_bam_coverage', 'bedtools_mergebed', 'Extract genomic DNA 1', 'bedtools_slopbed', 'Filter1', 'bedtools_multiintersectbed', 'Add_a_column1', 'bedtools_makewindowsbed', 'CONVERTER_bedgraph_to_bigwig', 'histogram_rpy', 'macs2_bdgdiff', 'idr', 'tp_sort_header_tool', 'gops_join_1', 'macs2_bdgpeakcall', 'gops_intersect_1', 'sort1', 'samtools_rmdup', 'deeptools_compute_matrix', 'tp_awk_tool', 'bowtie2', 'tp_head_tool', 'intervene_upset', 'chipseeker', 'addValue', 'bedtools_subtractbed', 'samtools_view', 'tp_grep_tool', 'join1', 'diffbind', 'deeptools_plot_enrichment', 'CONVERTER_interval_to_bedstrict_0', 'chipeakanno_annopeaks', 'macs2_bdgcmp', 'bedtools_sortbed', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'tp_cut_tool', 'Convert characters1', 'bundle_collection', 'Summary_Statistics1', 'bedtools_getfastabed']\n",
      "\n",
      "Test batch 1, Predicted top 57 tools: ['tp_easyjoin_tool', 'macs2_bdgdiff', 'comp1', 'crossmap_bed', 'tp_sort_header_tool', 'Add_a_column1', 'bedtools_makewindowsbed', 'intervene_pairwise', 'deeptools_plot_enrichment', 'CONVERTER_bedgraph_to_bigwig', 'idr', 'intervene_upset', 'addValue', 'tp_cut_tool', 'macs2_bdgcmp', 'Cut1', 'cat1', 'macs2_bdgpeakcall', 'bundle_collection', 'deeptools_compute_matrix', 'sort1', 'diffbind', 'Filter1', 'Summary_Statistics1', 'deeptools_bam_coverage', 'tp_awk_tool', 'bowtie2', 'join1', 'bedtools_multiintersectbed', 'Convert characters1', 'bedtools_getfastabed', 'chipeakanno_annopeaks', 'gops_intersect_1', 'bedtools_slopbed', 'CONVERTER_interval_to_bedstrict_0', 'bedtools_subtractbed', 'Extract genomic DNA 1', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', 'wig_to_bigWig', 'histogram_rpy', 'tp_cat', 'bedtools_intersectbed', 'tp_head_tool', 'bed2gff1', 'samtools_rmdup', 'gops_join_1', 'bedtools_mergebed', 'chipseeker', 'pygenomeTracks', 'tp_grep_tool', 'gops_concat_1', 'multiqc', 'jbrowse', 'bedtools_sortbed', 'macs2_callpeak', 'samtools_view', 'Remove beginning1']\n",
      "\n",
      "Test batch 1, Predicted top 57 tools with weights: ['Add_a_column1', 'samtools_view', 'gops_concat_1', 'bowtie2', 'multiqc', 'deeptools_bam_coverage', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'samtools_rmdup', 'tp_cat', 'deeptools_compute_matrix', 'tp_grep_tool', 'tp_awk_tool', 'gops_intersect_1', 'tp_head_tool', 'addValue', 'pygenomeTracks', 'Extract genomic DNA 1', 'bedtools_getfastabed', 'bedtools_sortbed', 'jbrowse', 'diffbind', 'bedtools_mergebed', 'chipseeker', 'bedtools_subtractbed', 'bedtools_multiintersectbed', 'histogram_rpy', 'macs2_bdgpeakcall', 'macs2_bdgcmp', 'bedtools_slopbed', 'idr', 'chipeakanno_annopeaks', 'bedtools_makewindowsbed', 'intervene_pairwise', 'intervene_upset', 'macs2_bdgdiff', 'deeptools_plot_enrichment', 'htseq_count', 'featurecounts', 'samtools_idxstats', 'deeptools_multi_bigwig_summary', 'deeptools_bam_compare', 'sklearn_pca', 'seacr', 'samtools_slice_bam', 'lofreq_call', 'bigwig_to_bedgraph', 'deeptools_bigwig_compare', 'tp_uniq_tool', 'bedtools_coveragebed', 'rseqc_infer_experiment', 'bamtools', 'sam_pileup', 'fastqc', 'query_tabular']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.07017543859649122\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.05263157894736842\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Test batch 1, Tool sequence: ['trimmomatic', 'fastqc', 'Grep1', '__FLATTEN__', 'collapse_dataset', 'join1', 'Add_a_column1', 'comp1', 'cat1', 'Cut1', 'collection_column_join', 'regex_replace', 'melt', 'Show beginning1']\n",
      "\n",
      "Test batch 1, True tools: ['ctb_ob_genProp', 'Show tail1', 'comp1', 'cardinal_spectra_plots', 'volcanoplot', 'tp_easyjoin_tool', 'Cut1', 'tp_cat', 'collection_column_join', 'barchart_gnuplot', 'Remove beginning1', 'cat1', 'cardinal_mz_images', 'plotly_ml_performance_plots', 'column_remove_by_header', 'sam_merge2', 'Extract genomic DNA 1', 'tp_find_and_replace', 'rna_star', 'ggplot2_heatmap2', 'Add_a_column1', 'fastqc', 'deseq2', 'Paste1', '__FILTER_FROM_FILE__', 'cutadapt', 'ctb_multi_obgrep', 'datamash_transpose', '__BUILD_LIST__', 'tp_sort_header_tool', 'random_lines1', 'ctb_chemfp_mol2fps', 'ctb_subsearch', 'ncbi_blastn_wrapper', 'bowtie2', 'tab2fasta', 'tables_arithmetic_operations', 'sklearn_ensemble', 'table_compute', 'addValue', 'join1', 'tophat', 'tp_replace_in_column', 'param_value_from_file', 'hgv_david', 'tp_cut_tool', 'Convert characters1']\n",
      "\n",
      "Test batch 1, Predicted top 47 tools: ['tables_arithmetic_operations', 'datamash_transpose', 'comp1', 'ctb_chemfp_mol2fps', 'Add_a_column1', 'plotly_ml_performance_plots', 'random_lines1', 'ctb_ob_genProp', 'Paste1', 'join1', 'tophat', 'cutadapt', 'ctb_subsearch', 'Convert characters1', 'ctb_multi_obgrep', 'deseq2', 'sklearn_ensemble', 'param_value_from_file', 'Cut1', 'collection_column_join', 'rna_star', 'cardinal_mz_images', 'tab2fasta', 'cardinal_spectra_plots', 'Remove beginning1', 'cat1', '__FILTER_FROM_FILE__', 'tp_cat', 'barchart_gnuplot', 'tp_sort_header_tool', 'tp_cut_tool', 'Extract genomic DNA 1', 'addValue', 'tp_find_and_replace', 'bowtie2', '__BUILD_LIST__', 'sam_merge2', 'volcanoplot', 'tp_replace_in_column', 'ggplot2_heatmap2', 'hgv_david', 'column_remove_by_header', 'table_compute', 'tp_easyjoin_tool', 'Show tail1', 'ncbi_blastn_wrapper', 'fastqc']\n",
      "\n",
      "Test batch 1, Predicted top 47 tools with weights: ['Add_a_column1', 'tp_find_and_replace', 'fastqc', 'rna_star', 'cutadapt', 'bowtie2', 'ncbi_blastn_wrapper', 'datamash_transpose', 'tp_easyjoin_tool', 'deseq2', 'tp_sort_header_tool', 'tp_cat', 'tab2fasta', 'ggplot2_heatmap2', 'table_compute', 'volcanoplot', 'tp_replace_in_column', 'collection_column_join', 'addValue', 'Extract genomic DNA 1', 'sam_merge2', 'cardinal_mz_images', 'cardinal_spectra_plots', 'plotly_ml_performance_plots', 'column_remove_by_header', 'sklearn_ensemble', 'ctb_chemfp_mol2fps', 'tables_arithmetic_operations', 'ctb_multi_obgrep', 'regexColumn1', 'bedtools_intersectbed', 'datamash_ops', 'gtftobed12', 'collapse_dataset', 'cor2', 'trimmomatic', 'macs2_callpeak', 'bg_uniq', 'tp_sed_tool', 'tp_awk_tool', 'seq_filter_by_id', 'gmx_solvate', 'ggplot2_point', 'addName', 'combine_metaphlan2_humann2', 'circos', 'Heatmap']\n",
      "\n",
      "Test batch 1, Precision: 1.0\n",
      "\n",
      "Test batch 1, Published precision: 0.1276595744680851\n",
      "\n",
      "Test batch 1, Published precision with weights: 0.10638297872340426\n",
      "=========================\n",
      "--------------------------\n",
      "Batch 1 prediction finished ...\n",
      "Precision@47: 0.9668367346938775\n",
      "Published Precision@47: 0.18473888092858975\n"
     ]
    }
   ],
   "source": [
    "predict_seq()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "16027192-45ed-4766-b6a1-1d114f23b244",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Low: test tool sequence: ['rbc_splitfasta', 'hhsearch', 'collection_element_identifiers', 'ffindex_dbkit_create', 'ffindex_dbkit_merge']\n",
      "\n",
      "Low: True labels: ['spring_model_all']\n",
      "\n",
      "Low: Predicted labels: ['spring_model_all'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['filtlong', 'flye', 'racon', 'maxbin2', 'prokka', 'roary']\n",
      "\n",
      "Low: True labels: ['taxonomy_krona_chart', 'barchart_gnuplot', 'CONVERTER_fasta_to_fai', 'iqtree', 'fasttree', 'raxml', 'gubbins', 'gd_raxml', 'graphlan']\n",
      "\n",
      "Low: Predicted labels: ['graphlan', 'raxml', 'barchart_gnuplot', 'CONVERTER_fasta_to_fai', 'fasttree', 'iqtree', 'taxonomy_krona_chart', 'gubbins', 'gd_raxml'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_groomer', '__ZIP_COLLECTION__']\n",
      "\n",
      "Low: True labels: ['__FILTER_EMPTY_DATASETS__', '__RELABEL_FROM_FILE__', 'hisat2', 'snippy_core', 'fastq_groomer', 'OMSSAAdapter', 'bwa_mem', 'trimmomatic', 'fastqc', 'salmon']\n",
      "\n",
      "Low: Predicted labels: ['fastq_groomer', 'fastqc', 'bwa_mem', 'salmon', 'hisat2', '__FILTER_EMPTY_DATASETS__', 'OMSSAAdapter', 'snippy_core', 'trimmomatic', '__RELABEL_FROM_FILE__'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['mspurity_puritya', 'abims_xcms_xcmsSet', 'xcms_merge']\n",
      "\n",
      "Low: True labels: ['abims_xcms_group', 'xcms_plot_chromatogram', 'quality_metrics']\n",
      "\n",
      "Low: Predicted labels: ['xcms_plot_chromatogram', 'quality_metrics', 'abims_xcms_group'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'bwa_mem', 'samtools_sort', 'picard_MarkDuplicates', 'gatk2_haplotype_caller', 'gatk2_variant_select']\n",
      "\n",
      "Low: True labels: ['gatk2_variant_filtration']\n",
      "\n",
      "Low: Predicted labels: ['gatk2_variant_filtration'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['sklearn_pca', 'sklearn_numeric_clustering']\n",
      "\n",
      "Low: True labels: ['macs2_callpeak', 'ggplot2_point', 'sklearn_nn_classifier']\n",
      "\n",
      "Low: Predicted labels: ['ggplot2_point', 'macs2_callpeak', 'sklearn_nn_classifier'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['porechop', 'filtlong', 'racon', 'rbc_splitfasta']\n",
      "\n",
      "Low: True labels: ['ncbi_blastp_wrapper', 'hhsearch', 'interproscan', 'mlst', 'mitos2', '__EXTRACT_DATASET__']\n",
      "\n",
      "Low: Predicted labels: ['__EXTRACT_DATASET__', 'mitos2', 'hhsearch', 'mlst', 'ncbi_blastp_wrapper', 'interproscan'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_awk_tool', 'regionalgam_ab_index', 'regionalgam_glmmpql', 'regionalgam_gls']\n",
      "\n",
      "Low: True labels: ['regionalgam_plot_trend', 'regionalgam_autocor_acf']\n",
      "\n",
      "Low: Predicted labels: ['regionalgam_plot_trend', 'regionalgam_autocor_acf'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['basic_illumination', 'ashlar', 'unet_coreograph']\n",
      "\n",
      "Low: True labels: ['unmicst', 'quantification', 's3segmenter']\n",
      "\n",
      "Low: Predicted labels: ['quantification', 'ConsensusID', 'rna_abbased_data'], Precision: 0.3333333333333333\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['FeatureFinderMultiplex', 'HighResPrecursorMassCorrector', 'MSGFPlusAdapter', 'PeptideIndexer', 'FalseDiscoveryRate', 'IDMapper', 'FileFilter', 'MapAlignerPoseClustering', 'MapRTTransformer']\n",
      "\n",
      "Low: True labels: ['FeatureLinkerUnlabeledQT', 'FeatureLinkerUnlabeledKD']\n",
      "\n",
      "Low: Predicted labels: ['FeatureLinkerUnlabeledQT', 'FeatureLinkerUnlabeledKD'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['FeatureFinderCentroided', 'IDMapper', 'QCCalculator']\n",
      "\n",
      "Low: True labels: ['QCExtractor']\n",
      "\n",
      "Low: Predicted labels: ['QCExtractor'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['CONVERTER_gz_to_uncompressed', 'cat1', 'cutadapt', 'chira_collapse', 'cshl_fastx_reverse_complement', 'chira_map', 'chira_merge', 'chira_quantify']\n",
      "\n",
      "Low: True labels: ['chira_extract']\n",
      "\n",
      "Low: Predicted labels: ['chira_extract'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cat1', '_ensembl_gtf2gene_list']\n",
      "\n",
      "Low: True labels: ['bg_column_arrange_by_header', 'scanpy_read_10x', 'scanpy_filter_genes', 'alevin', 'tp_replace_in_column', 'tp_awk_tool', 'join1', 'Cut1']\n",
      "\n",
      "Low: Predicted labels: ['scanpy_filter_genes', 'scanpy_read_10x', 'tp_awk_tool', 'bg_column_arrange_by_header', 'alevin', 'Cut1', 'join1', 'tp_replace_in_column'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.fasta.remove_desc', 'get_orfs_or_cdss', 'edu.tamu.cpt.gff3.cdsParents', 'edu.tamu.cpt.gff3.require_phage_start', 'edu.tamu.cpt.gff3.require_sd', 'edu.tamu.cpt.gff3.export_seq', 'edu.tamu.cpt.fasta.seq_trans']\n",
      "\n",
      "Low: True labels: ['LipoP', 'TMHMM_to_gff3', 'interproscan', 'ncbi_blastp_wrapper_new', 'SignalP']\n",
      "\n",
      "Low: Predicted labels: ['ncbi_blastp_wrapper_new', 'LipoP', 'SignalP', 'TMHMM_to_gff3', 'interproscan'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['1_create_conf']\n",
      "\n",
      "Low: True labels: ['conf4circos', '4_filter_sam', '2_map', '3_filter_single_pair', '5_calc_stat']\n",
      "\n",
      "Low: Predicted labels: ['3_filter_single_pair', 'diffbind-embl', 'ia_coordinates_of_roi', 'NSPDK_candidateClust', 'plotdexseq'], Precision: 0.2\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['sklearn_feature_selection']\n",
      "\n",
      "Low: True labels: ['sklearn_nn_classifier', 'keras_train_and_eval', 'sklearn_lightgbm']\n",
      "\n",
      "Low: Predicted labels: ['sklearn_nn_classifier', 'sklearn_lightgbm', 'keras_train_and_eval'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['__FILTER_FROM_FILE__', 'unzip']\n",
      "\n",
      "Low: True labels: ['graphicsmagick_image_convert', 'ip_convertimage', 'ip_imageinfo', 'ip_filter_standard', 'tab2fasta', 'ip_2d_feature_extraction', 'gdal_gdalwarp', 'megahit', 'ip_histogram_equalization', 'cp_cellprofiler', 'collection_element_identifiers', 'ip_threshold']\n",
      "\n",
      "Low: Predicted labels: ['collection_element_identifiers', 'cp_cellprofiler', 'ip_threshold', 'tab2fasta', 'ip_imageinfo', 'megahit', 'graphicsmagick_image_convert', 'ip_2d_feature_extraction', 'ip_filter_standard', 'gdal_gdalwarp', 'ip_histogram_equalization', 'ip_convertimage'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['sklearn_build_pipeline', 'sklearn_estimator_attributes', 'sklearn_searchcv']\n",
      "\n",
      "Low: True labels: ['sklearn_ensemble', 'plotly_ml_performance_plots', 'ggplot2_point', 'plotly_parallel_coordinates_plot', 'sklearn_stacking_ensemble_models']\n",
      "\n",
      "Low: Predicted labels: ['sklearn_ensemble', 'plotly_ml_performance_plots', 'ggplot2_point', 'sklearn_stacking_ensemble_models', 'plotly_parallel_coordinates_plot'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msconvert', 'CometAdapter', 'IDPosteriorErrorProbability', 'IDFileConverter']\n",
      "\n",
      "Low: True labels: ['FalseDiscoveryRate', 'proteomics_search_peptide_prophet_1', 'PeptideIndexer', 'FileInfo', 'seq_composition']\n",
      "\n",
      "Low: Predicted labels: ['FileInfo', 'proteomics_search_peptide_prophet_1', 'seq_composition', 'FalseDiscoveryRate', 'PeptideIndexer'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['seurat_read10x', 'seurat_filter_cells', 'seurat_normalise_data', 'seurat_find_variable_genes', 'seurat_run_pca', 'seurat_find_neighbours', 'seurat_find_clusters', 'seurat_run_tsne', 'sceasy_convert']\n",
      "\n",
      "Low: True labels: ['scmap_scmap_cell', 'export_to_genap2', 'scmap_preprocess_sce', 'scanpy_filter_genes', 'scmap_scmap_cluster', 'anndata_ops', 'anndata_manipulate']\n",
      "\n",
      "Low: Predicted labels: ['scmap_scmap_cell', 'scmap_preprocess_sce', 'scmap_scmap_cluster', 'anndata_ops', 'anndata_manipulate', 'scanpy_filter_genes', 'export_to_genap2'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_to_fasta_python', 'seqtk_mergepe']\n",
      "\n",
      "Low: True labels: ['fastq_to_fasta_python', 'seqtk_mergefa', 'mothur_merge_files', 'mummer_nucmer', 'mummer_mummer', 'qiime_align_seqs', 'lastz_wrapper_2']\n",
      "\n",
      "Low: Predicted labels: ['qiime_align_seqs', 'seqtk_mergefa', 'mummer_mummer', 'lastz_wrapper_2', 'mothur_merge_files', 'mummer_nucmer', 'fastq_to_fasta_python'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['retrieve_scxa']\n",
      "\n",
      "Low: True labels: ['scanpy_read_10x', 'seurat_read10x']\n",
      "\n",
      "Low: Predicted labels: ['kraken-report', 'kraken_taxonomy_report'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['OpenSwathDecoyGenerator', 'TargetedFileConverter', 'OpenSwathWorkflow']\n",
      "\n",
      "Low: True labels: ['pyprophet_merge']\n",
      "\n",
      "Low: Predicted labels: ['pyprophet_merge'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['vcfsort', 'vcfbedintersect']\n",
      "\n",
      "Low: True labels: ['bedtools_links', 'vcfannotate', 'bedtools_annotatebed', 'vcffilter2']\n",
      "\n",
      "Low: Predicted labels: ['bedtools_annotatebed', 'vcfannotate', 'vcffilter2', 'bedtools_links'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['stacks2_ustacks', 'stacks2_cstacks', 'stacks2_sstacks']\n",
      "\n",
      "Low: True labels: ['stacks2_tsv2bam']\n",
      "\n",
      "Low: Predicted labels: ['stacks2_tsv2bam'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_paired_end_joiner', 'trimmomatic', 'fastq_to_fasta_python', 'bowtie2', 'metaphlan', 'humann2', 'humann_regroup_table', 'humann2_rename_table']\n",
      "\n",
      "Low: True labels: ['Grep1', 'humann2_join_tables']\n",
      "\n",
      "Low: Predicted labels: ['humann2_join_tables', 'Grep1'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['stacks2_procrad', 'stacks2_denovomap']\n",
      "\n",
      "Low: True labels: ['stacks2_populations', 'stacks_populations']\n",
      "\n",
      "Low: Predicted labels: ['stacks2_populations', 'stacks_populations'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_groomer', 'trimmomatic', 'abyss-pe']\n",
      "\n",
      "Low: True labels: ['staramr_search', 'augustus', 'bandage_image', 'funannotate_sort', 'mlst', 'ncbi_makeblastdb', 'prokka', 'quast']\n",
      "\n",
      "Low: Predicted labels: ['staramr_search', 'augustus', 'quast', 'ncbi_makeblastdb', 'mlst', 'bandage_image', 'prokka', 'funannotate_sort'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['sklearn_build_pipeline', 'sklearn_searchcv', 'sklearn_stacking_ensemble_models']\n",
      "\n",
      "Low: True labels: ['sklearn_model_validation']\n",
      "\n",
      "Low: Predicted labels: ['sklearn_ensemble'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['TrimPrimer', 'Flash', 'Btrim64', 'uparse', 'pynast']\n",
      "\n",
      "Low: True labels: ['fasttree', 'FastTree']\n",
      "\n",
      "Low: Predicted labels: ['FastTree', 'fasttree'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['split_file_to_collection', 'param_value_from_file', 'sklearn_train_test_split']\n",
      "\n",
      "Low: True labels: ['tp_sorted_uniq', 'sklearn_ensemble', 'cardinal_filtering', 'addName', 'sklearn_train_test_eval', 'Cut1']\n",
      "\n",
      "Low: Predicted labels: ['sklearn_ensemble', 'sklearn_train_test_eval', 'cardinal_filtering', 'tp_sorted_uniq', 'Cut1', 'addName'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['_ensembl_gtf2gene_list', 'alevin']\n",
      "\n",
      "Low: True labels: ['_salmon_kallisto_mtx_to_10x', '_dropletBarcodePlot']\n",
      "\n",
      "Low: Predicted labels: ['_dropletBarcodePlot', '_salmon_kallisto_mtx_to_10x'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['collection_element_identifiers']\n",
      "\n",
      "Low: True labels: ['regex_replace', '__RELABEL_FROM_FILE__', 'tp_replace_in_line', 'ffindex_dbkit_create', 'tp_awk_tool', 'split_file_to_collection', 'Paste1', 'collapse_dataset']\n",
      "\n",
      "Low: Predicted labels: ['tp_replace_in_line', '__RELABEL_FROM_FILE__', 'Paste1', 'tp_awk_tool', 'collapse_dataset', 'split_file_to_collection', 'regex_replace', 'ffindex_dbkit_create'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msconvert', 'OpenSwathWorkflow', 'pyprophet_merge', 'pyprophet_score']\n",
      "\n",
      "Low: True labels: ['pyprophet_peptide']\n",
      "\n",
      "Low: Predicted labels: ['pyprophet_peptide'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['deseq2', 'tp_cut_tool', 'heinz_scoring', 'heinz']\n",
      "\n",
      "Low: True labels: ['heinz_visualization']\n",
      "\n",
      "Low: Predicted labels: ['heinz_visualization'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['samtools_view', 'param_value_from_file', 'compose_text_param']\n",
      "\n",
      "Low: True labels: ['rdock_sort_filter', 'hisat2', 'tp_find_and_replace', 'Add_a_column1', 'tp_replace_in_column', 'sdf_to_tab', 'rxdock_sort_filter', 'tp_awk_tool', 'bowtie2', 'add_line_to_file', 'tp_text_file_with_recurring_lines', 'filter_tabular', 'regex1', 'Filter1', 'snpSift_filter']\n",
      "\n",
      "Low: Predicted labels: ['rxdock_sort_filter', 'Filter1', 'sdf_to_tab', 'add_line_to_file', 'tp_text_file_with_recurring_lines', 'filter_tabular', 'snpSift_filter', 'rdock_sort_filter', 'hisat2', 'Add_a_column1', 'bowtie2', 'tp_awk_tool', 'tp_find_and_replace', 'regex1', 'tp_replace_in_column'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['PeakPickerHiRes', 'FileFilter', 'xcms-find-peaks', 'xcms-collect-peaks', 'xcms-group-peaks', 'xcms-blankfilter', 'xcms-dilutionfilter', 'camera-annotate-peaks', 'camera-group-fwhm', 'camera-find-adducts', 'camera-find-isotopes', 'msms2metfrag-multiple']\n",
      "\n",
      "Low: True labels: ['metfrag-cli-batch-multiple']\n",
      "\n",
      "Low: Predicted labels: ['metfrag-cli-batch-multiple'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bedtools_multiintersectbed', 'Filter1', 'gops_cluster_1']\n",
      "\n",
      "Low: True labels: ['gops_subtract_1', 'gops_basecoverage_1', 'deeptools_compute_matrix', 'tp_replace_in_column']\n",
      "\n",
      "Low: Predicted labels: ['gops_subtract_1', 'gops_basecoverage_1', 'tp_replace_in_column', 'deeptools_compute_matrix'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['thermo_raw_file_converter', 'search_gui', 'peptide_shaker', 'Show beginning1', 'cat1', 'proteomics_moff']\n",
      "\n",
      "Low: True labels: ['tp_replace_in_column', 'sort1', 'lfq_protein_quant', 'tp_sed_tool', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['tp_sed_tool', 'Filter1', 'sort1', 'lfq_protein_quant', 'tp_replace_in_column'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['xcms-find-peaks']\n",
      "\n",
      "Low: True labels: ['camera-annotate-peaks', 'xcms-collect-peaks']\n",
      "\n",
      "Low: Predicted labels: ['camera-annotate-peaks', 'xcms-collect-peaks'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['addName', 'tp_sorted_uniq', 'tp_cat', 'addValue', 'mass_spectrometry_imaging_combine', 'mass_spectrometry_imaging_preprocessing']\n",
      "\n",
      "Low: True labels: ['tp_easyjoin_tool', 'mass_spectrometry_imaging_combine', 'mass_spectrometry_imaging_segmentations', 'mass_spectrometry_imaging_preprocessing', 'tp_sort_header_tool', 'mass_spectrometry_imaging_ion_images', 'mass_spectrometry_imaging_qc', 'mass_spectrometry_imaging_classification', 'mass_spectrometry_imaging_mzplots', 'Cut1', 'collection_column_join']\n",
      "\n",
      "Low: Predicted labels: ['mass_spectrometry_imaging_combine', 'mass_spectrometry_imaging_preprocessing', 'mass_spectrometry_imaging_qc', 'mass_spectrometry_imaging_classification', 'mass_spectrometry_imaging_ion_images', 'mass_spectrometry_imaging_segmentations', 'tp_sort_header_tool', 'mass_spectrometry_imaging_mzplots', 'collection_column_join', 'Cut1', 'tp_easyjoin_tool'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['funannotate_clean', 'funannotate_sort', 'rna_star', 'funannotate_predict', 'interproscan', 'funannotate_annotate']\n",
      "\n",
      "Low: True labels: ['busco', 'jcvi_gff_stats', 'jbrowse', 'aegean_parseval', 'funannotate_compare']\n",
      "\n",
      "Low: Predicted labels: ['busco', 'jbrowse', 'jcvi_gff_stats', 'funannotate_compare', 'aegean_parseval'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_cut_tool', 'tp_tail_tool', 'fastq_dump']\n",
      "\n",
      "Low: True labels: ['predict_pipeline', 'cshl_fastx_trimmer', 'fastq_to_fasta_python', 'rna_star', 'fastq_paired_end_splitter', 'bwameth', 'multiqc', 'rna_starsolo', 'hisat2', 'khmer_abundance_distribution_single', 'fastq_manipulation', 'fastq_groomer', 'cshl_fastq_quality_filter', 'bwa_mem', 'gffcompare', 'snippy', 'sickle', 'fastq_stats', 'kallisto_quant', 'cufflinks', 'megahit', 'trimmomatic', 'fastqc', 'fastq_paired_end_deinterlacer', 'bismark_bowtie2', 'bowtie2', 'cshl_fastq_to_fasta', 'trim_galore', 'gffread', 'sample_seqs', 'hifiasm', 'regex1', 'cutadapt', 'cshl_fastx_clipper', 'minimap2', 'cat1', 'bowtie_wrapper', '__EXTRACT_DATASET__', 'fastq_trimmer', 'CONVERTER_gz_to_uncompressed', 'kraken2', 'fastp', 'spades']\n",
      "\n",
      "Low: Predicted labels: ['multiqc', 'cshl_fastq_quality_filter', 'bowtie2', 'hifiasm', 'CONVERTER_gz_to_uncompressed', 'fastq_paired_end_splitter', 'snippy', 'sickle', 'fastqc', 'bwameth', 'kallisto_quant', 'bismark_bowtie2', 'fastq_to_fasta_python', 'cshl_fastx_clipper', 'fastp', 'predict_pipeline', 'khmer_abundance_distribution_single', 'fastq_manipulation', 'sample_seqs', 'bwa_mem', 'cufflinks', 'kraken2', 'fastq_groomer', 'bowtie_wrapper', 'megahit', 'regex1', 'trimmomatic', 'fastq_trimmer', 'fastq_paired_end_deinterlacer', 'trim_galore', 'gffcompare', 'cshl_fastx_trimmer', '__EXTRACT_DATASET__', 'rna_starsolo', 'cshl_fastq_to_fasta', 'cat1', 'fastq_stats', 'rna_star', 'cutadapt', 'gffread', 'spades', 'minimap2', 'hisat2'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['usearch_derep_full']\n",
      "\n",
      "Low: True labels: ['cluster_otu', 'map_reads_otu', 'vsearch_sorting']\n",
      "\n",
      "Low: Predicted labels: ['vsearch_sorting', 'map_reads_otu', 'cluster_otu'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msgfplus', 'idpqonvert']\n",
      "\n",
      "Low: True labels: ['idpassemble', 'idpquery']\n",
      "\n",
      "Low: Predicted labels: ['IDFileConverter', 'percolator_input_converters'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastp', 'unicycler', 'mob_recon', 'progressivemauve']\n",
      "\n",
      "Low: True labels: ['newick_display', 'circos_aln_to_links', 'xmfa2gff3']\n",
      "\n",
      "Low: Predicted labels: ['newick_display', 'circos_aln_to_links', 'xmfa2gff3'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['FeatureFinderCentroided']\n",
      "\n",
      "Low: True labels: ['QCCalculator', 'FeatureLinkerUnlabeled', 'MapStatistics', 'IDMapper']\n",
      "\n",
      "Low: Predicted labels: ['IDMapper', 'MapStatistics', 'FeatureLinkerUnlabeled', 'QCCalculator'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['dropletutils_read_10x', 'scpred_preprocess_data', 'scpred_predict_labels', 'scpred_get_std_output']\n",
      "\n",
      "Low: True labels: ['ct_combine_tool_outputs']\n",
      "\n",
      "Low: Predicted labels: ['scanpy_compute_graph'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['umi_tools_extract', 'rna_starsolo']\n",
      "\n",
      "Low: True labels: ['multiqc', 'scanpy_read_10x', 'dropletutils', 'bamFilter', 'anndata_import', 'stringtie']\n",
      "\n",
      "Low: Predicted labels: ['multiqc', 'dropletutils', 'bamFilter', 'stringtie', 'scanpy_read_10x', 'anndata_import'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.fasta.remove_desc', 'edu.tamu.cpt.gff3.require_phage_start']\n",
      "\n",
      "Low: True labels: ['edu.tamu.cpt.genbank.shinefind', 'edu.tamu.cpt.gff3.require_sd']\n",
      "\n",
      "Low: Predicted labels: ['edu.tamu.cpt.genbank.shinefind', 'edu.tamu.cpt.gff3.require_sd'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['minfi_read450k', 'minfi_ppfun', 'minfi_dropsnp']\n",
      "\n",
      "Low: True labels: ['minfi_getanno', 'minfi_dmr', 'minfi_getM', 'minfi_getbeta', 'minfi_methcpg', 'minfi_dmp']\n",
      "\n",
      "Low: Predicted labels: ['minfi_dropsnp', 'vg_convert', 'muscle', 'Kneeplot', 'sansa_annotate', 'fromHicupToJuicebox'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'velvetoptimiser', 'glimmer_not-knowlegde-based']\n",
      "\n",
      "Low: True labels: ['ncbi_blastn_wrapper', 'EMBOSS: transeq101', 'edu.tamu.cpt2.util.glimmer3_to_gff3', 'ncbi_blastx_wrapper']\n",
      "\n",
      "Low: Predicted labels: ['edu.tamu.cpt2.util.glimmer3_to_gff3', 'ncbi_blastx_wrapper', 'ncbi_blastn_wrapper', 'EMBOSS: transeq101'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['rpExtractSink', 'retropath2', 'rp2paths']\n",
      "\n",
      "Low: True labels: ['rpCompletion']\n",
      "\n",
      "Low: Predicted labels: ['XTandemAdapter'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['schicexplorer_schicqualitycontrol', 'schicexplorer_schicnormalize', 'schicexplorer_schicclustersvl']\n",
      "\n",
      "Low: True labels: ['schicexplorer_schicconsensusmatrices', 'schicexplorer_schicplotclusterprofiles']\n",
      "\n",
      "Low: Predicted labels: ['schicexplorer_schicplotclusterprofiles', 'schicexplorer_schicclustersvl'], Precision: 0.5\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_paired_end_interlacer', 'spades', 'cisarunner']\n",
      "\n",
      "Low: True labels: ['prokka', 'cshl_fasta_formatter', 'check_contigs']\n",
      "\n",
      "Low: Predicted labels: ['prokka', 'cshl_fasta_formatter', 'check_contigs'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['retrieve_scxa', 'scanpy_read_10x', 'scanpy_filter_cells', 'scanpy_filter_genes', 'scanpy_normalise_data', 'scanpy_regress_variable', 'scanpy_scale_data', 'scanpy_find_variable_genes', 'scanpy_run_pca', 'scanpy_compute_graph', 'scanpy_find_cluster', 'scanpy_run_umap', 'run_sccaf', 'sccaf_asses']\n",
      "\n",
      "Low: True labels: ['sccaf_asses_merger']\n",
      "\n",
      "Low: Predicted labels: ['sccaf_asses'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['TargetedFileConverter', 'OpenSwathWorkflow', 'pyprophet_merge']\n",
      "\n",
      "Low: True labels: ['pyprophet_score']\n",
      "\n",
      "Low: Predicted labels: ['pyprophet_score'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['deeptools_bam_coverage', 'bigwig_to_bedgraph']\n",
      "\n",
      "Low: True labels: ['bedtools_unionbedgraph', 'bedtools_makewindowsbed', 'bedtools_intersectbed', 'bedtools_mergebedgraph']\n",
      "\n",
      "Low: Predicted labels: ['bedtools_intersectbed', 'bedtools_mergebedgraph', 'bedtools_unionbedgraph', 'bedtools_makewindowsbed'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['qiime_validate_mapping_file', 'qiime_split_libraries', 'qiime_pick_otus', 'qiime_pick_rep_set', 'qiime_assign_taxonomy', 'qiime_make_otu_table']\n",
      "\n",
      "Low: True labels: ['qiime_summarize_taxa_through_plots', 'qiime_make_otu_heatmap']\n",
      "\n",
      "Low: Predicted labels: ['qiime_make_otu_heatmap', 'qiime_summarize_taxa_through_plots'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'mothur_make_contigs']\n",
      "\n",
      "Low: True labels: ['mothur_get_seqs', 'mothur_screen_seqs', 'hmmer_hmmsearch', 'mothur_unique_seqs', 'mothur_summary_qual', 'mothur_merge_files', 'mothur_make_group', 'mothur_align_seqs', 'seq_filter_by_id', 'fastqc', 'mothur_sub_sample', 'mothur_trim_seqs', 'mothur_summary_seqs', 'mycrobiota_subsample_add_discarded_samples', 'fasta_filter_by_length', 'mothur_count_seqs']\n",
      "\n",
      "Low: Predicted labels: ['seq_filter_by_id', 'mothur_merge_files', 'fastqc', 'hmmer_hmmsearch', 'mothur_align_seqs', 'mothur_summary_seqs', 'mothur_count_seqs', 'mothur_unique_seqs', 'mothur_trim_seqs', 'mothur_make_group', 'mothur_get_seqs', 'mothur_screen_seqs', 'fasta_filter_by_length', 'mothur_summary_qual', 'mycrobiota_subsample_add_discarded_samples', 'mothur_sub_sample'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_common', 'cp_identify_primary_objects', 'cp_convert_objects_to_image']\n",
      "\n",
      "Low: True labels: ['cp_display_data_on_image', 'cp_measure_object_size_shape', 'cp_image_math', 'cp_gray_to_color']\n",
      "\n",
      "Low: Predicted labels: ['cp_display_data_on_image', 'cp_measure_object_size_shape', 'cp_gray_to_color', 'cp_image_math'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['methtools_calling', 'cshl_sort_header', 'cshl_uniq_tool', 'Filter1', 'Add_a_column1', 'Cut1', 'bedtools_intersectbed_bam']\n",
      "\n",
      "Low: True labels: ['methtools_plot', 'sort1', 'Cut1', 'methtools_dmr', 'wc_gnu']\n",
      "\n",
      "Low: Predicted labels: ['methtools_plot', 'sort1', 'methtools_dmr', 'Cut1', 'wc_gnu'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['make_window_bed', 'window_coverage']\n",
      "\n",
      "Low: True labels: ['find_qPCR_regions']\n",
      "\n",
      "Low: Predicted labels: ['find_qPCR_regions'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['make_window_bed']\n",
      "\n",
      "Low: True labels: ['window_coverage']\n",
      "\n",
      "Low: Predicted labels: ['hisat2'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['PeakPickerHiRes', 'FileFilter', 'xcms-find-peaks', 'xcms-collect-peaks', 'xcms-group-peaks', 'xcms-blankfilter', 'xcms-dilutionfilter', 'camera-annotate-peaks', 'camera-group-fwhm', 'camera-find-adducts', 'camera-find-isotopes', 'msms2metfrag-multiple', 'metfrag-cli-batch-multiple']\n",
      "\n",
      "Low: True labels: ['passatutto']\n",
      "\n",
      "Low: Predicted labels: ['passatutto'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_common', 'cp_identify_primary_objects', 'cp_convert_objects_to_image', 'cp_gray_to_color', 'cp_save_images', 'cp_measure_granularity', 'cp_measure_texture', 'cp_measure_object_intensity']\n",
      "\n",
      "Low: True labels: ['cp_measure_object_size_shape', 'cp_track_objects']\n",
      "\n",
      "Low: Predicted labels: ['cp_track_objects', 'cp_measure_object_size_shape'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['humann', 'humann_renorm_table']\n",
      "\n",
      "Low: True labels: ['combine_metaphlan2_humann2', 'humann_unpack_pathways', 'humann2_unpack_pathways']\n",
      "\n",
      "Low: Predicted labels: ['humann_unpack_pathways', 'humann2_unpack_pathways', 'combine_metaphlan2_humann2'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt.fasta.remove_desc']\n",
      "\n",
      "Low: True labels: ['edu.tamu.cpt.tRNAscan', 'MGA', 'edu.tamu.cpt2.util.mga_to_gff3', 'edu.tamu.cpt.fasta.lipory', 'ncbi_blastn_wrapper_new', 'edu.tamu.cpt.gff3.export_seq', 'jbrowse-iuc-cpt', 'edu.tamu.cpt.genbank.shinefind', 'aragorn_trna', 'get_orfs_or_cdss', 'glimmer_not-knowlegde-based', 'edu.tamu.cpt.gff3.require_sd', 'edu.tamu.cpt.util.wigToBigWig', 'edu.tamu.cpt.gff3.require_phage_start', 'edu.tamu.cpt2.util.glimmer3_to_gff3', 'TransTermHP']\n",
      "\n",
      "Low: Predicted labels: ['staramr_search', 'addValue', 'edu.tamu.cpt.util.wigToBigWig', 'quast', 'mergeCols1', 'mlst', 'collapse_dataset', 'infernal_cmscan', 'bandage_image', 'abricate', 'infernal_cmsearch', 'edu.tamu.cpt2.util.mga_to_gff3', 'bio_hansel', 'bowtie2', 'Extract genomic DNA 1', 'qualimap_bamqc'], Precision: 0.125\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['ebi_search_rest_results', 'regex1', 'tp_cut_tool', 'lftp']\n",
      "\n",
      "Low: True labels: ['graphicsmagick_image_convert', 'ip_convertimage', 'unzip', 'barchart_gnuplot', 'ip_scale_image', 'ip_imageinfo', 'maxquant', 'msnbase_readmsdata', 'scatterplot_rpy', 'collapse_dataset']\n",
      "\n",
      "Low: Predicted labels: ['ip_scale_image', 'collapse_dataset', 'unzip', 'barchart_gnuplot', 'graphicsmagick_image_convert', 'ip_convertimage', 'ip_imageinfo', 'msnbase_readmsdata', 'scatterplot_rpy', 'maxquant'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bowtie2', 'bamFilter', 'getTn5ExtendedCoverage']\n",
      "\n",
      "Low: True labels: ['splitbedGraphFromBed']\n",
      "\n",
      "Low: Predicted labels: ['splitbedGraphFromBed'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_common', 'cp_identify_primary_objects', 'cp_convert_objects_to_image', 'cp_image_math', 'cp_measure_granularity', 'cp_measure_texture', 'cp_measure_object_intensity', 'cp_measure_object_size_shape', 'cp_relate_objects', 'cp_measure_image_quality']\n",
      "\n",
      "Low: True labels: ['cp_measure_image_area_occupied']\n",
      "\n",
      "Low: Predicted labels: ['cp_measure_image_area_occupied'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Add_a_column1', 'Cut1', 'fgsea']\n",
      "\n",
      "Low: True labels: ['tp_find_and_replace', 'plotdexseq', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['tp_find_and_replace', 'Filter1', 'plotdexseq'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['unzip', 'ip_histogram_equalization', 'ip_convertimage']\n",
      "\n",
      "Low: True labels: ['ip_scale_image', 'ip_filter_standard', 'ip_overlay_segmentation']\n",
      "\n",
      "Low: Predicted labels: ['ip_filter_standard', 'ip_scale_image', 'ip_overlay_segmentation'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.gff3.reformat', 'edu.tamu.cpt.gff3.export_seq']\n",
      "\n",
      "Low: True labels: ['edu.tamu.cpt.fasta.seq_trans', 'edu.tamu.cpt.fasta.start_codons', 'edu.tamu.cpt.fasta.stop_codons']\n",
      "\n",
      "Low: Predicted labels: ['edu.tamu.cpt.fasta.start_codons', 'edu.tamu.cpt.fasta.stop_codons', 'edu.tamu.cpt.fasta.seq_trans'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['mspurity_puritya', 'abims_xcms_xcmsSet']\n",
      "\n",
      "Low: True labels: ['abims_xcms_refine', 'xcms_merge', 'abims_xcms_group', 'mspurity_purityx', 'xcms_plot_chromatogram', 'query_tabular', 'quality_metrics', 'xcms_export_samplemetadata']\n",
      "\n",
      "Low: Predicted labels: ['xcms_plot_chromatogram', 'quality_metrics', 'xcms_merge', 'abims_xcms_group', 'query_tabular', 'xcms_export_samplemetadata', 'mspurity_purityx', 'abims_xcms_refine'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['get_pdb']\n",
      "\n",
      "Low: True labels: ['bio3d_pca', 'Grep1', 'prepare_receptor', 'bio3d_rmsf', 'CONVERTER_Pdb_to_Gro_0', 'docking', 'bio3d_rmsd', 'tp_grep_tool', 'gmx_setup', 'fpocket']\n",
      "\n",
      "Low: Predicted labels: ['Grep1', 'fpocket', 'prepare_receptor', 'tp_grep_tool', 'gmx_setup', 'docking', 'CONVERTER_Pdb_to_Gro_0', 'bio3d_rmsf', 'bio3d_rmsd', 'bio3d_pca'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['chira_merge']\n",
      "\n",
      "Low: True labels: ['chira_quantify']\n",
      "\n",
      "Low: Predicted labels: ['chira_quantify'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['hca_matrix_downloader', 'seurat_read10x']\n",
      "\n",
      "Low: True labels: ['seurat_filter_cells', 'seurat_normalise_data', 'seurat_create_seurat_object', 'seurat_find_variable_genes']\n",
      "\n",
      "Low: Predicted labels: ['seurat_filter_cells', 'seurat_normalise_data', 'seurat_create_seurat_object', 'seurat_find_variable_genes'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_common', 'cp_identify_primary_objects', 'cp_convert_objects_to_image', 'cp_gray_to_color', 'cp_save_images', 'cp_measure_granularity', 'cp_measure_texture', 'cp_measure_object_intensity', 'cp_measure_object_size_shape']\n",
      "\n",
      "Low: True labels: ['cp_display_data_on_image', 'cp_relate_objects', 'cp_measure_object_intensity', 'cp_export_to_spreadsheet', 'cp_measure_image_quality']\n",
      "\n",
      "Low: Predicted labels: ['cp_display_data_on_image', 'cp_export_to_spreadsheet', 'cp_measure_image_quality', 'cp_relate_objects', 'cp_measure_object_intensity'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.fasta.remove_desc', 'get_orfs_or_cdss', 'edu.tamu.cpt.gff3.cdsParents', 'edu.tamu.cpt.gff3.require_phage_start', 'edu.tamu.cpt.gff3.require_sd']\n",
      "\n",
      "Low: True labels: ['edu.tamu.cpt.gff3.lipoP_to_gff3', 'edu.tamu.cpt.fasta.lipory', 'edu.tamu.cpt.gff3.export_seq', 'jbrowse-iuc-cpt', 'gff3.rebase']\n",
      "\n",
      "Low: Predicted labels: ['jbrowse-iuc-cpt', 'edu.tamu.cpt.gff3.export_seq', 'gff3.rebase', 'edu.tamu.cpt.fasta.lipory', 'edu.tamu.cpt.gff3.lipoP_to_gff3'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['ctb_remIons']\n",
      "\n",
      "Low: True labels: ['ctb_chemfp_mol2fps', 'ctb_compound_convert', 'ctb_filter']\n",
      "\n",
      "Low: Predicted labels: ['ctb_chemfp_mol2fps', 'ctb_filter', 'ctb_compound_convert'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['retrieve_scxa', 'scanpy_read_10x', 'scanpy_filter_cells', 'scanpy_filter_genes', 'scanpy_normalise_data', 'scanpy_find_variable_genes', 'scanpy_scale_data', 'scanpy_run_pca', 'scanpy_compute_graph', 'scanpy_find_cluster', 'sceasy_convert', 'scmap_preprocess_sce']\n",
      "\n",
      "Low: True labels: ['scmap_scmap_cell', 'scmap_scmap_cluster', 'scmap_select_features']\n",
      "\n",
      "Low: Predicted labels: ['scmap_select_features', 'scmap_scmap_cluster', 'scmap_scmap_cell'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_measure_granularity', 'cp_measure_texture', 'cp_measure_object_intensity', 'cp_measure_object_size_shape', 'cp_relate_objects', 'cp_measure_image_quality', 'cp_measure_image_area_occupied']\n",
      "\n",
      "Low: True labels: ['cp_measure_image_intensity', 'cp_export_to_spreadsheet']\n",
      "\n",
      "Low: Predicted labels: ['cp_export_to_spreadsheet', 'cp_measure_image_intensity'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_replace_in_line', 'clustalw']\n",
      "\n",
      "Low: True labels: ['hmmer_hmmbuild', 'rbc_rnacode', 'iqtree', 'rgweblogo3', 'fasttree', 'raxml', 'phyml', 'align_back_trans', 'rnaz', 'qiime_make_phylogeny', 'gd_raxml', 'wc_gnu']\n",
      "\n",
      "Low: Predicted labels: ['qiime_make_phylogeny', 'raxml', 'hmmer_hmmbuild', 'rnaz', 'gd_raxml', 'wc_gnu', 'iqtree', 'fasttree', 'align_back_trans', 'rbc_rnacode', 'rgweblogo3', 'phyml'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'kraken2', 'kraken_taxonomy_report']\n",
      "\n",
      "Low: True labels: ['vegan_fisher_alpha', 'vegan_diversity', 'vegan_rarefaction', 'Fetch Taxonomic Ranks']\n",
      "\n",
      "Low: Predicted labels: ['Fetch Taxonomic Ranks', 'vegan_diversity', 'vegan_rarefaction', 'vegan_fisher_alpha'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['gbk_to_orf', 'smf_utils_fix-fasta-headers']\n",
      "\n",
      "Low: True labels: ['smf_utils_find-boxes', 'fasttree', 'raxml', 'blast_reciprocal_best_hits', 'smf_utils_extract-boxed-sequences']\n",
      "\n",
      "Low: Predicted labels: ['smf_utils_extract-boxed-sequences', 'fasttree', 'raxml', 'smf_utils_find-boxes', 'blast_reciprocal_best_hits'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['rbc_splitfasta', 'hhsearch', 'ffindex_dbkit_create']\n",
      "\n",
      "Low: True labels: ['ffindex_dbkit_merge', 'spring_model_all']\n",
      "\n",
      "Low: Predicted labels: ['spring_model_all', 'ffindex_dbkit_merge'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Remove beginning1', 'split_file_to_collection', '__RELABEL_FROM_FILE__', 'Cut1', 'param_value_from_file', 'kc-align']\n",
      "\n",
      "Low: True labels: ['sarscov2formatter', 'sarscov2summary']\n",
      "\n",
      "Low: Predicted labels: ['sarscov2summary', 'sarscov2formatter'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_identify_primary_objects']\n",
      "\n",
      "Low: True labels: ['cp_display_data_on_image', 'cp_measure_object_size_shape', 'cp_overlay_outlines', 'cp_convert_objects_to_image', 'cp_cellprofiler']\n",
      "\n",
      "Low: Predicted labels: ['cp_cellprofiler', 'cp_display_data_on_image', 'cp_measure_object_size_shape', 'cp_overlay_outlines', 'cp_convert_objects_to_image'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['gops_join_1', 'CONVERTER_interval_to_bedstrict_0']\n",
      "\n",
      "Low: True labels: ['circos_wiggle_to_stacked', 'Grouping1', 'datamash_ops', 'tp_tail_tool', 'gops_intersect_1', 'CONVERTER_bed_to_fli_0', 'tp_replace_in_column', 'gops_join_1', 'circos_wiggle_to_scatter', 'join1', 'gops_concat_1']\n",
      "\n",
      "Low: Predicted labels: ['circos_wiggle_to_stacked', 'gops_intersect_1', 'CONVERTER_bed_to_fli_0', 'circos_wiggle_to_scatter', 'join1', 'gops_concat_1', 'gops_join_1', 'Grouping1', 'tp_replace_in_column', 'tp_tail_tool', 'datamash_ops'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['sed']\n",
      "\n",
      "Low: True labels: ['gtf_filter_by_attribute_values_list_b', 'awkscript']\n",
      "\n",
      "Low: Predicted labels: ['awkscript', 'gtf_filter_by_attribute_values_list_b'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Cut1', 'stacks_denovomap']\n",
      "\n",
      "Low: True labels: ['stacks_populations', 'stacks_genotypes']\n",
      "\n",
      "Low: Predicted labels: ['stacks_genotypes', 'stacks_populations'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['macs2_callpeak', 'macs2_bdgcmp']\n",
      "\n",
      "Low: True labels: ['macs2_bdgbroadcall', 'wig_to_bigWig']\n",
      "\n",
      "Low: Predicted labels: ['macs2_bdgbroadcall', 'wig_to_bigWig'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['keras_model_builder']\n",
      "\n",
      "Low: True labels: ['sklearn_model_validation', 'sklearn_estimator_attributes', 'keras_train_and_eval']\n",
      "\n",
      "Low: Predicted labels: ['keras_train_and_eval', 'sklearn_model_validation', 'sklearn_estimator_attributes'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['EMBOSS: transeq101', 'filter_by_fasta_ids', 'fasplit', 'ncbi_blastp_wrapper', 'tp_cat', 'blast_parser', 'hcluster_sg']\n",
      "\n",
      "Low: True labels: ['hcluster_sg_parser']\n",
      "\n",
      "Low: Predicted labels: ['hcluster_sg_parser'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['CONVERTER_gff_to_interval_index_0', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0']\n",
      "\n",
      "Low: True labels: ['CONVERTER_bed_to_bgzip_0', 'gops_intersect_1', 'featurecounts', 'Add_a_column1', 'gffcompare', 'deeptools_compute_matrix', 'stringtie_merge', 'CONVERTER_gff_to_interval_index_0', 'gops_concat_1', 'cat1', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['CONVERTER_gff_to_interval_index_0', 'CONVERTER_bed_to_bgzip_0', 'gops_intersect_1', 'deeptools_compute_matrix', 'stringtie_merge', 'featurecounts', 'Add_a_column1', 'gffcompare', 'Filter1', 'gops_concat_1', 'cat1'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['ip_threshold']\n",
      "\n",
      "Low: True labels: ['ip_convertimage', 'ip_count_objects', 'ip_binary_to_labelimage', 'ip_2d_split_binaryimage_by_watershed', 'imagej2_watershed_binary']\n",
      "\n",
      "Low: Predicted labels: ['ip_count_objects', 'ip_convertimage', 'ip_2d_split_binaryimage_by_watershed', 'ip_binary_to_labelimage', 'imagej2_watershed_binary'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['proteomiqon_peptidedb', 'proteomiqon_psmstatistics']\n",
      "\n",
      "Low: True labels: ['proteomiqon_proteininference', 'proteomiqon_psmbasedquantification']\n",
      "\n",
      "Low: Predicted labels: ['proteomiqon_psmbasedquantification', 'proteomiqon_proteininference'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cutadapt', 'hicup_hicup']\n",
      "\n",
      "Low: True labels: ['fromHicupToJuicebox']\n",
      "\n",
      "Low: Predicted labels: ['tophat2'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_sorted_uniq', 'join1', 'tp_cut_tool', 'sklearn_svm_classifier']\n",
      "\n",
      "Low: True labels: ['tp_tail_tool', 'plotly_ml_performance_plots', 'sklearn_clf_metrics']\n",
      "\n",
      "Low: Predicted labels: ['tp_tail_tool', 'plotly_ml_performance_plots', 'sklearn_clf_metrics'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trim_galore', 'iuc_pear']\n",
      "\n",
      "Low: True labels: ['cshl_fastq_quality_filter', 'fastqc', 'cutadapt', 'cat1', 'fastqStat', 'fastq_filter']\n",
      "\n",
      "Low: Predicted labels: ['fastqc', 'cat1', 'fastq_filter', 'cutadapt', 'cshl_fastq_quality_filter', 'fastqStat'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['preproc', 'viennarna_rnafold', 'structure_to_gspan', 'nspdk_sparse']\n",
      "\n",
      "Low: True labels: ['tp_cat', 'NSPDK_candidateClust']\n",
      "\n",
      "Low: Predicted labels: ['tp_cat', 'NSPDK_candidateClust'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'iuc_pear', 'cat1', 'uclust2otutable', 'biom_convert']\n",
      "\n",
      "Low: True labels: ['mothur_make_shared', 'biom_add_metadata']\n",
      "\n",
      "Low: Predicted labels: ['mothur_make_shared', 'biom_add_metadata'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['OpenSwathWorkflow', 'pyprophet_merge', 'pyprophet_score', 'pyprophet_peptide']\n",
      "\n",
      "Low: True labels: ['pyprophet_protein']\n",
      "\n",
      "Low: Predicted labels: ['pyprophet_protein'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['ctb_online_data_fetch', 'ctb_change_title', 'ctb_compound_convert', 'ctb_confab', 'ctb_ob_genProp']\n",
      "\n",
      "Low: True labels: ['ctb_filter', 'ctb_spectrophore_search']\n",
      "\n",
      "Low: Predicted labels: ['ctb_filter', 'ctb_spectrophore_search'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['EMBOSS: transeq101', 'ncbi_makeblastdb', 'ncbi_blastp_wrapper', 'blast_parser']\n",
      "\n",
      "Low: True labels: ['hcluster_sg']\n",
      "\n",
      "Low: Predicted labels: ['hcluster_sg'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msconvert', 'proteomiqon_mzmltomzlite', 'proteomiqon_peptidespectrummatching']\n",
      "\n",
      "Low: True labels: ['proteomiqon_psmstatistics']\n",
      "\n",
      "Low: Predicted labels: ['proteomiqon_psmstatistics'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['PeakPickerHiRes', 'FileFilter', 'xcms-find-peaks', 'xcms-collect-peaks', 'xcms-group-peaks', 'xcms-blankfilter', 'xcms-dilutionfilter', 'camera-annotate-peaks', 'camera-group-fwhm', 'camera-find-adducts', 'camera-find-isotopes']\n",
      "\n",
      "Low: True labels: ['map-msms2camera', 'camera-prepareoutput', 'msms2metfrag-multiple', 'camera-find-adducts']\n",
      "\n",
      "Low: Predicted labels: ['msms2metfrag-multiple', 'camera-prepareoutput', 'map-msms2camera', 'camera-find-adducts'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_groomer', 'fastq_filter', 'fastq_manipulation']\n",
      "\n",
      "Low: True labels: ['fastq_to_fasta_python', 'fastq_groomer', 'trimmomatic', 'fastqc', 'rbc_mirdeep2_mapper', 'cd_hit_dup']\n",
      "\n",
      "Low: Predicted labels: ['fastq_to_fasta_python', 'fastqc', 'cd_hit_dup', 'trimmomatic', 'fastq_groomer', 'rbc_mirdeep2_mapper'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['ncbi_acc_download', 'music_construct_eset', 'raceid_inspectclusters', 'scater_create_qcmetric_ready_sce']\n",
      "\n",
      "Low: True labels: ['scater_filter', 'scater_plot_pca', 'scater_normalize', 'scater_plot_dist_scatter', 'scater_plot_exprs_freq']\n",
      "\n",
      "Low: Predicted labels: ['scater_plot_dist_scatter', 'scater_filter', 'scater_plot_pca', 'scater_plot_exprs_freq', 'scater_normalize'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['encyclopedia_searchtolib']\n",
      "\n",
      "Low: True labels: ['encyclopedia_quantify']\n",
      "\n",
      "Low: Predicted labels: ['encyclopedia_quantify'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['make_families', 'correct_barcodes', 'align_families', 'dunovo']\n",
      "\n",
      "Low: True labels: ['fasta2fastq', 'sequence_content_trimmer']\n",
      "\n",
      "Low: Predicted labels: ['sequence_content_trimmer', 'fasta2fastq'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msnbase_readmsdata', 'abims_xcms_xcmsSet', 'xcms_merge', 'abims_xcms_group', 'abims_xcms_fillPeaks', 'abims_CAMERA_annotateDiffreport']\n",
      "\n",
      "Low: True labels: ['Univariate', 'Batch_correction', 'generic_filter', 'w4mclassfilter', 'lcmsmatching', 'lipidmaps', 'wsdl_hmdb', 'quality_metrics', 'Transformation', 'abims_CAMERA_combinexsAnnos']\n",
      "\n",
      "Low: Predicted labels: ['Univariate', 'generic_filter', 'quality_metrics', 'w4mclassfilter', 'Batch_correction', 'wsdl_hmdb', 'Transformation', 'abims_CAMERA_combinexsAnnos', 'lcmsmatching', 'lipidmaps'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['gops_coverage_1']\n",
      "\n",
      "Low: True labels: ['cshl_grep_tool', 'sort1', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['sort1', 'Filter1', 'cshl_grep_tool'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msnbase_readmsdata', 'abims_xcms_xcmsSet', 'xcms_merge', 'abims_xcms_group', 'abims_xcms_fillPeaks']\n",
      "\n",
      "Low: True labels: ['mspurity_frag4feature', 'abims_xcms_summary', 'Batch_correction', 'generic_filter', 'abims_CAMERA_annotateDiffreport', 'quality_metrics']\n",
      "\n",
      "Low: Predicted labels: ['generic_filter', 'mspurity_frag4feature', 'quality_metrics', 'abims_xcms_summary', 'abims_CAMERA_annotateDiffreport', 'Batch_correction'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trim_galore', 'make_families']\n",
      "\n",
      "Low: True labels: ['correct_barcodes', 'Cut1']\n",
      "\n",
      "Low: Predicted labels: ['Cut1', 'correct_barcodes'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['monocle3_reduceDim']\n",
      "\n",
      "Low: True labels: ['monocle3_partition']\n",
      "\n",
      "Low: Predicted labels: ['monocle3_partition'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['shovill', 'cat_contigs']\n",
      "\n",
      "Low: True labels: ['cat_add_names', 'prokka']\n",
      "\n",
      "Low: Predicted labels: ['prokka', 'cat_add_names'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Grep1', 'computeMatrix']\n",
      "\n",
      "Low: True labels: ['dt_profiler', 'heatmapper_deepTools']\n",
      "\n",
      "Low: Predicted labels: ['dt_profiler', 'heatmapper_deepTools'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msnbase_readmsdata', 'abims_xcms_xcmsSet', 'xcms_merge', 'abims_xcms_group', 'mspurity_frag4feature']\n",
      "\n",
      "Low: True labels: ['mspurity_filterfragspectra']\n",
      "\n",
      "Low: Predicted labels: ['mspurity_filterfragspectra'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['qiime_validate_mapping_file']\n",
      "\n",
      "Low: True labels: ['qiime_split_libraries', 'qiime_summarize_taxa_through_plots', 'qiime_make_otu_table', 'qiime_core_diversity', 'qiime_make_otu_heatmap', 'qiime_split_libraries_fastq']\n",
      "\n",
      "Low: Predicted labels: ['qiime_pick_rep_set', 'qiime_pick_otus', 'obi_annotate', 'obi_ngsfilter', 'obi_clean', 'hamronize_tool'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['OpenSwathDecoyGenerator', 'TargetedFileConverter', 'pyprophet_merge', 'pyprophet_score', 'pyprophet_peptide', 'pyprophet_protein']\n",
      "\n",
      "Low: True labels: ['pyprophet_export']\n",
      "\n",
      "Low: Predicted labels: ['pyprophet_export'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Extractor', 'GeneModelMapper']\n",
      "\n",
      "Low: True labels: ['GeMoMa_Annotation_Filter']\n",
      "\n",
      "Low: Predicted labels: ['GeMoMa_Annotation_Filter'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_groomer', 'trimmomatic', 'velvetoptimiser']\n",
      "\n",
      "Low: True labels: ['staramr_search', 'fasta-stats', 'abricate', 'filter_by_fasta_ids', 'PlasFlow', 'mothur_classify_seqs', 'exonerate', 'EMBOSS: getorf42', 'mlst', 'glimmer_not-knowlegde-based', 'bowtie2', 'quast', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['PlasFlow', 'glimmer_not-knowlegde-based', 'quast', 'staramr_search', 'EMBOSS: getorf42', 'mlst', 'abricate', 'mothur_classify_seqs', 'Filter1', 'exonerate', 'fasta-stats', 'bowtie2', 'filter_by_fasta_ids'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'iuc_pear', 'cat1', 'uclust2otutable', 'biom_convert', 'biom_add_metadata']\n",
      "\n",
      "Low: True labels: ['phyloseq_richness', 'phyloseq_taxonomy', 'phyloseq_DESeq2', 'phyloseq_abundance']\n",
      "\n",
      "Low: Predicted labels: ['phyloseq_DESeq2', 'phyloseq_abundance', 'phyloseq_richness', 'phyloseq_taxonomy'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bedtools_bamtobed', 'flair_correct']\n",
      "\n",
      "Low: True labels: ['flair_collapse']\n",
      "\n",
      "Low: Predicted labels: ['flair_collapse'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.fasta.remove_desc', 'get_orfs_or_cdss', 'edu.tamu.cpt.gff3.cdsParents', 'edu.tamu.cpt.gff3.require_phage_start', 'edu.tamu.cpt.gff3.require_sd', 'edu.tamu.cpt.gff3.export_seq', 'edu.tamu.cpt.fasta.seq_trans', 'ncbi_blastp_wrapper_new']\n",
      "\n",
      "Low: True labels: ['jbrowse-iuc-cpt', 'edu.tamu.cpt2.phage.intron_detection', 'edu.tamu.cpt.blastp_to_gff3']\n",
      "\n",
      "Low: Predicted labels: ['jbrowse-iuc-cpt', 'edu.tamu.cpt2.phage.intron_detection', 'edu.tamu.cpt.blastp_to_gff3'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['map_reads', 'call_sites', 'filter_regions', 'call_consensus']\n",
      "\n",
      "Low: True labels: ['snp_matrix', 'merge_vcfs']\n",
      "\n",
      "Low: Predicted labels: ['snp_matrix', 'merge_vcfs'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['obi_illumina_pairend', 'fastq_groomer', 'obi_grep']\n",
      "\n",
      "Low: True labels: ['obi_uniq', 'obi_ngsfilter', 'qiime_pick_otus', 'obi_clean']\n",
      "\n",
      "Low: Predicted labels: ['obi_uniq', 'obi_ngsfilter', 'qiime_pick_otus', 'obi_clean'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bowtie2', 'hicexplorer_hicbuildmatrix', 'hicexplorer_chicviewpointbackgroundmodel', 'hicexplorer_chicviewpoint']\n",
      "\n",
      "Low: True labels: ['hicexplorer_chicaggregatestatistic', 'hicexplorer_chicsignificantinteractions']\n",
      "\n",
      "Low: Predicted labels: ['hicexplorer_chicaggregatestatistic', 'hicexplorer_chicsignificantinteractions'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trim_galore', 'make_families', 'correct_barcodes']\n",
      "\n",
      "Low: True labels: ['align_families', 'Cut1']\n",
      "\n",
      "Low: Predicted labels: ['Cut1', 'correct_barcodes'], Precision: 0.5\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Extractor', 'ncbi_tblastn_wrapper']\n",
      "\n",
      "Low: True labels: ['datamash_ops', 'GeneModelMapper', 'Cut1', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['datamash_ops', 'Filter1', 'Cut1', 'GeneModelMapper'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.gff3.reformat', 'edu.tamu.cpt.gff3.export_seq', 'edu.tamu.cpt.fasta.seq_trans', 'TMHMM_to_gff3']\n",
      "\n",
      "Low: True labels: ['gff3.rebase', 'edu.tamu.cpt.wig.rebase']\n",
      "\n",
      "Low: Predicted labels: ['gff3.rebase', 'edu.tamu.cpt.wig.rebase'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['keras_model_config', 'keras_model_builder', 'keras_train_and_eval']\n",
      "\n",
      "Low: True labels: ['model_prediction', 'ml_visualization_ex']\n",
      "\n",
      "Low: Predicted labels: ['ml_visualization_ex', 'model_prediction'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['TrimPrimer', 'Flash', 'Btrim64', 'ExtractFASTAfromFASTQ', 'ChooseTag', 'TrimByN', 'TrimByLength']\n",
      "\n",
      "Low: True labels: ['uchime']\n",
      "\n",
      "Low: Predicted labels: ['Count1'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['TrimPrimer', 'Flash', 'Btrim64', 'ExtractFASTAfromFASTQ', 'ChooseTag', 'TrimByN', 'TrimByLength', 'uchime']\n",
      "\n",
      "Low: True labels: ['map_reads_otu', 'pynast', 'uparse']\n",
      "\n",
      "Low: Predicted labels: ['pynast', 'map_reads_otu', 'uparse'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['locarna_multiple', 'selectsequencesfrommsa']\n",
      "\n",
      "Low: True labels: ['rbc_rnacode', 'rnaz', 'viennarna_rnaalifold']\n",
      "\n",
      "Low: Predicted labels: ['rnaz', 'rbc_rnacode', 'viennarna_rnaalifold'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['FeatureFinderMetabo', 'MetaboliteAdductDecharger', 'AccurateMassSearch', 'HighResPrecursorMassCorrector']\n",
      "\n",
      "Low: True labels: ['FileFilter', 'MSGFPlusAdapter', 'AssayGeneratorMetabo', 'XTandemAdapter']\n",
      "\n",
      "Low: Predicted labels: ['AssayGeneratorMetabo', 'XTandemAdapter', 'MSGFPlusAdapter', 'FileFilter'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_sed_tool', 'csv_to_tabular']\n",
      "\n",
      "Low: True labels: ['tp_easyjoin_tool', 'deseq2', 'Grouping1', 'datamash_ops', 'datamash_transpose', 'sklearn_numeric_clustering', 'histogram_rpy', 'regexColumn1', 'Remove beginning1', 'tp_sort_header_tool', 'tp_cat', 'charts', 'limma_voom', 'edger', 'sklearn_generalized_linear', 'ggplot2_point', 'tp_awk_tool', 'tp_cut_tool', 'tp_sed_tool', 'Cut1', 'Filter1', 'Count1', 'annotatemyids']\n",
      "\n",
      "Low: Predicted labels: ['sklearn_numeric_clustering', 'sklearn_generalized_linear', 'limma_voom', 'tp_awk_tool', 'edger', 'tp_cut_tool', 'regexColumn1', 'annotatemyids', 'ggplot2_point', 'tp_cat', 'deseq2', 'datamash_transpose', 'histogram_rpy', 'tp_sed_tool', 'Grouping1', 'Remove beginning1', 'charts', 'Filter1', 'Cut1', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'datamash_ops', 'Count1'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['OpenSwathAssayGenerator']\n",
      "\n",
      "Low: True labels: ['OpenSwathDecoyGenerator']\n",
      "\n",
      "Low: Predicted labels: ['infernal_cmscan'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bam_to_sam', 'MergingDataClusterize']\n",
      "\n",
      "Low: True labels: ['removeIntrons', 'removeExonLines']\n",
      "\n",
      "Low: Predicted labels: ['removeIntrons', 'removeExonLines'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bioext_bealign', 'bioext_bam2msa', 'hyphy_strike_ambigs', 'tn93_cluster']\n",
      "\n",
      "Low: True labels: ['tn93_filter', 'hyphy_annotate']\n",
      "\n",
      "Low: Predicted labels: ['tn93_filter', 'hyphy_annotate'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['1_create_conf', '4_filter_sam', '5_calc_stat']\n",
      "\n",
      "Low: True labels: ['conf4circos', '6_parse_discord']\n",
      "\n",
      "Low: Predicted labels: ['6_parse_discord', 'conf4circos'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_find_and_replace', 'Convert characters1', 'tp_easyjoin_tool', 'Cut1', 'regex1', 'stoceps_maketablecarrer', 'stoceps_filteringsp', 'Filter1', 'stoceps_glm']\n",
      "\n",
      "Low: True labels: ['stoceps_glm_group']\n",
      "\n",
      "Low: Predicted labels: ['stoceps_glm_group'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low test prediction precision: 0.8934606481481482\n"
     ]
    }
   ],
   "source": [
    "predict_low_freq_seq()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2938e6ae-7a49-45b8-9b11-1af23fd36a07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tool sequence: ['snpEff_build_gb', 'bwa_mem', 'samtools_view']\n",
      "\n",
      "Compatible true tools: ['collapse_dataset', 'samtools_fastx', '__EXTRACT_DATASET__', 'lofreq_viterbi', 'lofreq_alnqual', 'samtools_stats', 'samtools_merge', 'macs2_callpeak', 'samtools_sort', 'bedtools_intersectbed', 'ivar_trim', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'lofreq_indelqual', 'racon', 'Cut1', 'bedtools_bamtobed', 'Grouping1', 'freebayes', 'bamtools', 'deeptools_plot_fingerprint', 'tp_replace_in_line', 'samtools_rmdup', 'qualimap_bamqc', 'seq_filter_by_mapping', '__FILTER_FROM_FILE__', 'picard_SamToFastq', 'stringtie', 'trinity', 'bamFilter', 'featurecounts', 'cufflinks', 'samtools_depth', 'samtools_flagstat', 'tp_sort_header_tool', 'samtools_idxstats', 'sam_merge2', 'bcftools_mpileup', 'fastqc', 'deeptools_alignmentsieve', 'bamleftalign', '__FILTER_FAILED_DATASETS__', 'rseqc_bam_stat', 'ngsutils_bam_filter', 'sam2interval', 'samtools_bedcov', 'strelka_somatic', 'deeptools_compute_matrix', 'picard_MergeSamFiles', 'param_value_from_file', 'tp_awk_tool', 'hicexplorer_hicbuildmatrix', 'picard_SortSam', 'CONVERTER_bam_to_coodinate_sorted_bam', 'htseq_count', 'picard_CollectInsertSizeMetrics', 'varscan_somatic', 'column_remove_by_header', 'bedtools_bamtofastq', 'tb_profiler_profile', 'extract_variants', 'macs2_predictd'], size: 62\n",
      "\n",
      "Predicted top 20 tools: ['CONVERTER_bam_to_coodinate_sorted_bam', 'strelka_somatic', 'bcftools_mpileup', 'macs2_callpeak', 'deeptools_bam_coverage', 'trinity', 'macs2_predictd', 'lofreq_alnqual', 'sam2interval', 'ivar_trim', 'picard_CollectInsertSizeMetrics', 'lofreq_indelqual', 'picard_SortSam', 'qualimap_bamqc', 'extract_variants', 'param_value_from_file', 'rseqc_bam_stat', 'bamFilter', 'deeptools_plot_fingerprint', 'bedtools_intersectbed']\n",
      "\n",
      "Predicted precision: 1.0\n",
      "\n",
      "Correctly predicted tools: ['deeptools_plot_fingerprint', 'picard_SortSam', 'bcftools_mpileup', 'bedtools_intersectbed', 'rseqc_bam_stat', 'param_value_from_file', 'lofreq_alnqual', 'CONVERTER_bam_to_coodinate_sorted_bam', 'lofreq_indelqual', 'picard_CollectInsertSizeMetrics', 'trinity', 'extract_variants', 'bamFilter', 'qualimap_bamqc', 'strelka_somatic', 'macs2_predictd', 'ivar_trim', 'macs2_callpeak', 'deeptools_bam_coverage', 'sam2interval']\n",
      "\n",
      "Predicted top 20 tools with weights: ['tp_replace_in_line', 'samtools_stats', 'qualimap_bamqc', 'ivar_trim', 'lofreq_indelqual', 'lofreq_viterbi', 'fastqc', 'featurecounts', 'stringtie', 'htseq_count', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'tp_sort_header_tool', 'bedtools_intersectbed', 'macs2_callpeak', 'samtools_rmdup', 'deeptools_compute_matrix', 'bamFilter', 'collapse_dataset', 'freebayes']\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_5947/1929167505.py:15: UserWarning: Tight layout not applied. tight_layout cannot make axes width small enough to accommodate all axes decorations\n",
      "  plt.tight_layout()\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAPMAAACUCAYAAABGBx/dAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAjtElEQVR4nO2deXyU1fX/3yd7SAIJ+07YEQERES24V62t27eurRXq0uqvVq378m1d2q/9FpdWqdpWLe61atVWi21dvorFrQqCgiyybwmQQCAL2XN+f5xnyDCZJPPMDMwk3M/r9bwyz33u3Hknec7c+9x7zrmiqjg5OXV8pSQawMnJKT5yxuzk1EnkjNnJqZPIGbOTUyeRM2Ynp04iZ8xOTp1EzpidnDqJnDE7OXUSpSUawKlZIvJN4GBgqar+w7E4Hj8S5wGWHBKRWcBIYBUwClipqlcd6CyOJ3K5njl5lKaq3wqciMhjjmWPHE8Ecs/MyaM1IecbE0JhSiYWcDwRyQ2zEywRedx7OQjIxG6U4QCqevSByuJ4/MsNsxOvJuCZkDIBLjzAWcDx+JLrmRMsEclV1UoRGaeqS0LLD1QWx+NfzpiTRCLyDjA3qEiB1cDzqtp0oLI4nsjlJsCSR1uAdGwyJR17LhsO3B9NYyLyexHJ6QQsjidCuWfm5NFiVf1V4EREblHVmSJya5TtZQI3ezfJB8DrqlrbAVkcT4Ryxpw8OlRE0lS1QUQygCleeWY0janqJQAiMgqYBcwGCjogi+OJUG6YnTyaA2wQkUXAeuBVEZkMdI2mMRG5SUTmY7OvbwPjOyiL44m0HTcBljwSke7Ys9dqVd0RVN5bVbf5bOsO4CDgNeBvqrq7o7I4ngjbccac/BKRGar6dJTvPQS4DWhU1fM7E4vj2VvumbljSHy/wbyVNgDnAEXAC52QxfEEyRlzx1A0w6cpwH+A41W1pJOyOJ4gOWPuGPL9bQ+cr6pfxp0kuVjA8eyRM+aOoS1RvGe3iLwCVACvA5tU9cNOxuJ4guSWphIsERncyjEzUEdV34ii6f/G1izXAn8Dzu1ILI7Hv1zPnHjNBdYB3YCe3utC7Fv6lhjaXa6q74nIEapaJyKR9BjJxOJ4fMr1zInXlap6AvAcUKiqxwJDgb/E2O4EETkSyBKRccCIDsbieHzK9cwJVlAyuL7qLfqrapOI+HbnC9HdwB+BCcDJwKUdicXx+JdzGkkSicgzQDmwEksS11VV4xb0LiL9VLW4o7E4Hh/vc8acHPIc9n+Al74V+GM0kTMicoiqfi4iM0Iuna6qkU70JA2L4/HRnjPm5JSIfCOamVERma2ql4rIu+wdQH+Mqn69o7M4njbac8acWHn/yNB/ggCDVXV4DO2OV9XFQecHt+eYkEwsjse/E4mbAEu8PgF+F1ImwI9ibPf3YkHz7wNEeHMkE4vj8SnXMyepRORIVf04hvfPAOqBY7Cb8LlonuuSjcXxtNGOM+bESkR+rKoPS3NOZrBv+/GqOjkO7ecCDwLfVNW+HYXF8bTPEyo3zE68AoHoAjwZVD49lkZFZDZQBZyGOTUc2cFYHI9fqao7kuAA8to6b+U9X2/j2grgEiBrf7C0xRMLi+OJ/HDD7CSRiBwTUjRdVX8YUueb2GRLLu3MoorIYaq6oJVrU7WNqJxIWPzwxMLieNrnCcgNs5NHs4CFeP90oDFMnVuBq4Eyr973W2ustZvD0wigrRskEpaIeWJkcTzt8wDOmJNJl6nqp4ETEfl/YerMV9VFQXVejfKz2gugj4QlXjyRBPM7ngjkjDlJFHJz5AJHAH8IqfaRiLyHbYUiWErWaGZR23y2ipAlXjztPuc5nsjkjDlJJCJlNA/HyoGHwlS7DpgJ7PTOo51FbfPbPkKWePG02/M4nggVy+yZO+J3YHmg2qszK+R8RIRt54ecD4+VJVoevyyOp32ewOFms5NI3uznwcCXqvrPMNefpnnHQTCH/BNbaet3wNPAYcANwMuqekO8WPzwxMrieCKTyzSSJBKRWcBV2OzoT0TkwTDVhgLvYFuirKd5+BZO69VcDKdjN92uOLP44YmaxfFELvfMnDxKU9VvBU5E5LEwdS5W1VVBdcL2CJ7yRORobPuU3SK+MsBGwuKHJxYWxxMpVFTvctoXWhNyvjFMnR0i8mtsXXMuNnxrbV+jIuC3wEUichq2h3A8WfzwxMLieCKUe2ZOsIKc9gdhW4KuwTYkQ1WPDqn7KLZ/70HAPcAvVPXKCD9njKoujxdLLDyRsDie9nlC5XrmxKsJ28ozVOGWMVao6lMicrOq7hCRFj2CiNzeyuccA4SdLIuSpV2eGFkcj085Y068rlHVytBCEVkQ9Drgn3uwiPQDVES6AcPCtDcB+HuY8p1xZiECnlhYHI9fRbqG5Y7EHcAM7+dRwCYsFG8dcGSYugNbaSNsebQskfDsaxbHs/fhlqY6hgRAVd9X1YFY9E0h0MJBX1U3AYjIaBH5QEQqRGQekBNPlkh49gOL4wmSM+aOIQUQ06HAWLEwvNB8VMG6CbgWGALcSGzbp7Rg8cmzr1gcT5DcM3OCJSIjNGgtsrVq3s+/Aw2YXzCY435rWq6qn3ivPxaR4+PM4ofHN4vj8S9nzInXnSIyHThCg5LCiUiBqgZugve9nztUdUZQnUPbaHeUd30ttnwSyf5Fflj88ETD4nh8yg2zE68NarMeJ4eU70nfqqoB3953ReR48bYSBU5vo91fAw9jEzCzsHXOeLL44YmGxfH4lHMaSbBE5HrgcqAXzcMxwfYv6hFS9ylsOSMQfhdT8vVYWBxP4nlaKF5T8u6Ieinje9jjzvSQ8gvC1H0x5Py4Ntodg3kfVQDzgNHxZPHDEw2L4/F/L7lhduJ1iKo2YEnQg7U0TN23RKQw6HxwG+3eiP8ZUj8sfniiYXE8PuUmwBKvAhE5DjhZRIqCyqcDoRkffwOUiEXVCNAVi4MNp2hmSP2w+OGJdrbW8fiQM+bE6zngAuBQ9l7WCLeM8XNVvS9wIiLntNFuNDOkflj88EQ7W+t4/Ciasbk74n8AR4ecT/P5/rNCzsdgKVorseWSiJ/DYmUJ5YmFxfFEfrieOcESkTOAr1R1XlDZROwb+gMfTeWFnO9Q1alee4OA7vuRJZTHN4vj8S83AZZ4nQXUicj3RWSGiPQBtgJTfLYTusYYnMu5AvjxfmQJ5YmGxfH4lOuZE6+1qrpGRHoBPwFeUdVKESmOpjERORY4DjhWmtPPpBBZ9opkYnE8PuWMOfFSAFX9j4hs0eZ4Wb/ePIG7YScWbjcRSyIHlrbmuf3IEuCJhcXx+G3QewB3SpBEZA7wD+/0bOBl7/U3VPXMdt57vKq+673OUdWqoGu9VLWktfN4s7TFEw2L42mfp0X7zpgTKxFZCCwKc6nFBt4iMgH4GdCTdtwDxbZNOYnmiZbTVfXceLH44YmGxfG0zxMqN8xOvG5S1bdCC0XkhDB1rwd+BZwPzCa8o0JAc7AbL+BDHMkMqR8WPzzRsDgen3LGnGAF3xwiMhYYCyxV1XfCVF+sqgtF5BRVXSkitW00vUpVrwlqe2ScWfzw+GZxPP7llqaSRCJyM/AXYAbwkoiE88+d5q1r5ovIz7Asjq3pDRG5WESOEctycVOcWfzwRM3ieCKX65mTRwer6sGBExEJN6N5HVCHxb/eAvx3G+1dDNTSnOmxrawk0bD44YmFxfFEKGfMyaPQpOefgw3nVDUQlXOeqt7tvb6unfZKNfKsJNGw+OGJhcXxRChnzMmjaWK7Bq7B8itniCVJD06IfrSI9MZ8eF9V1c/aaO9zseibQOaL04GFcWTxwxMLi+OJUG5pKkkkIn8F/hbm0umqeo5XJ1tVq0WkPzY7OkxVR7fSXjF79yARZ7mIhMUPTywsjseHwkVfuGP/H8CgkPMe3s+BQWXnAi9hPcKDtBGtA1wUcn5iPFn88MTC4ngiP9wwO3lUJiLfJshxADhXvUTpnr6FOSB8X4Mid8JJVZ8MKQqNqoqVJWKeGFkcT4Ryw+wkkYjMBZYAATe+Y1T162HqZQDfxG6g7ap6cyvtfR1zWAh4H4VNOhcLS6Q8sbA4nsjleubk0VoN2u5TRIaGVhCROzHH/O9g+aReaqO97wLfAC7D0tdcH08WnzyxsDieCOWcRpJH60TkJBEZIpZj+fth6vwAGA3coqpjVbW1LUHBthMtA9JUtR4oiDOLH55YWBxPhHLD7CSRN6O5jOZQxhYzmiJyFHA1nhshcLOqrm2lvTnYt/zxQD9gpKoeGy8WPzyxsDieyOWMOUkkIpcBr2FpXGcA5ao6O6TOq8CfgFXYN/73VPW0VtobD5QC6cBFwAuquiJeLH54YmFxPJHLDbOTR9OATMz1r593HqqFqvqiqn6mqn8GvmijvRu89n4B5AJhJ8piYPHDEwuL44lQzpiTRwuBImCMqt5ES5dBgE0iMgzA+1kWpk5we5vbaS8WFj88sbA4ngjlZrOTRxOA3wJvikg25iYIgIjswJzwBfipiKj3uitwr9/2YmGJkicWFscTqaLxNHFH/A9gFJYkrgs2EXJ20LXW9jIKW95ee7GwRMMTC4vjifxwE2BOTp1E7pnZyamTyBlzkslb9tjv9RxPx+dxxpx8ivSfGe96++tzHE986rWQM2Ynp04iNwGWQGVIpmaRs1dZPbWkk9nue8PVGzVhd4t6Jdsb6dUjda+yBV/UlqpqL8fTsXlC5daZE6gscjhCwkbORaU33lgUUb3UfqvWhyt3PB2LJ1RumO3k1Em0X3pmETkZuAdzgfvEKx6G7aL3N7EdAS7AvGY+B/oAPYAhwJ2q+lVQW3nAA0Cqql7kg+ESIEdVHxSRF7EsEP8O4VkPXIilRP26D56HgSOBDyNlEpFLMskG4Av9iFK2UEAvunrRb9VUkUUXtrCB4YyjjG2kkU4e+dRRQz11VLObYYwlR5oTU1RUNnHlrSV8vKCaqYdn88SsPhH/fZKVZ6V+wS52MFmO26tOqRaznIUcxrFkS07LRlpRRWUT195eQmMjEfN0BO0XY1bVN0XkAmC+qj4EICIjgAyvyneB51T1HRFJx8LFRmMucBrSVoWIPINFl/hheFxkz76ZN2GZFP8Rhudo4AzgMR88jwO9/fLkScFsgJFMoIxSetKXQTICgN1aSRNN7KSUUooYwDC6S2+atImPeJOpfCNsu3m5KVz83Ty2lTb4wUlqnoEMZxc7WtTpKf3I8h9cRF5uCheek8dTL1T4fm8yq11jFpHRWALvxcChwPvAz4FHsJ5qGHAakI/5l2YAH2I3/5eqOjNMm0diTuVPimXwPxxIF5G+HlMv4HbgTVX9MAzPzcBwEVmAbU59OHAfttftNGyvnmOBZ7G407do7vkPC2kvE3jVO30b6I8Z5oxwPMAIETkDc44f5/3+9d7fYqqIHI1llfgMqAnDsx3IArrLnnDYZjVpI4uwX7kHvamlhnpqKWY9dVpDE0o9taxhKT3ow26qKNUi8sjngh9tYc26OtLThfUbG/jo0xrmfVzNxs0NHDohk6xM4cNPawBSROT0/cGzZWsDF56TR+n2Rtasr2faaRuprFKOm5bN1yZnccXNJYH/wxTgUe9/cAIwOIXUsDxLWUAGmWSSRQNtf0kU6XpK2ZvngbtsLmlzcQP3/a6MtRvqSUsTZt3Vco7JW/e9DCCLLm1+VqIVyTPzN7Gs/A9hN/RrWFRHYEi5GDhJVTdg6Ud3qOpdqjodM4hxQW2dLSIPANcEClT139imWU+q6nOq+jSwU1XvDDXkIJ56LP/SOVivXgM0qAV07/Z4/u6xVajqbcCl2BfAXjzAP4HhwPNYgHgW1hO34AFWAHdjcam/xPIclwE7gF3e9aOAcixbxD1heFaran/gFEWp1F17YLaxmYXMo5pK+jKIwYyiiUZy6Eo/htBXBtNfhpBGOsPlYLqQxyoWM44pDJWDGD4knYL8VAryU+nWLYVRw9OZ99pA8vJSKCtr4sYfF5CdJWA5pvYLz7FTsxlWmE5dPfTpmcYHcwYxoF8aH35SQ15uCvldUwL3wSfefTBHVScB54Qz5s2sJZU0RskhDGYU9bS+3Vad1oblmTDWZpVrapQbrijg4Zm9WbmmntffrmrRhqo+qqqTVXVyJLPWiVQkxvwYsA2Yh/XI9V554LmxhL2zCa4Jer0ay7IQ0MtqG2RdBrwRBW+AZyfmnB7gqYY9X5tNITxFQa8rQnmwL6bHgTdUtQnL2dSahgMlqhq4g9YACzyegVgPHOApBw5ui0cQqijfc6E3AzxDKaQHfRARsmn9WbCaKjLIJEXsph86JJ1JEzIp6JbCpqIGduxsIj1NyM4S8vKEpSvqSLX/eLCV7FOegPr1Sd1zPmJoOpmZxuNXVZTThVzjjYEn9HzE0PSoeJJJkRjzEcBMVT0C2IplVoDWd4sPDt8agfVye0lVywGR6LbhOALbWX5FEE8XYKV3vRbrXcGMq3/Qe/PC8CzFhsgiIpOAwjY+ezXQR0QC7Q/DjPg5YCMWYB7gCfTwrfIoSg5d9/qAHLpSSzUA5VpGNS17i4CyyaGOGhq1EYC16+vZXNzAd76dy+AB6dTWKc/8pZzqaqWmVjloVAYZGQLN6W72OU9AxVsb95x/tbqO6pomDhqVQV5uCjTfh4NbbTyIZzeVxqsaNU/o+Ver6zhoVAYdWZFMgHUHfiMia7Bnx6+wm/8SEXkSm0gaLyKve/UzRORWLEbzWVVd4qUSnQD0EJF8r15/4BMRyfGuTReRcuAgoJvYrnm/V9XtITz9sQ2ss7ANqiu836MwKAD8Vm+WWbGh9Z3AdGzI29V7fwE2c/0v4GvezzVAIzZU3hWOB5s/eF5EPsFmuXOxuYJ+mJH8xCurxoy5VR5BaKCeNSyjgTqKWU8P+rKTUj5jHtnkIMBOSkkjnTRNp4oKe48uYyDDGMF4lvAfumoBQ0oaqahUrvnZdoq3NqAKsx7bSWVVE12y0jhopA3DS7Y39RORG/cHT6+eqTzx5wq2lTaws1yYfNIGqmuV46Zmc+qJXdhc3MAVN5cMFpEfYaOY6WKbmk9vopGdWkoJRdSwm1Itpj9DWcp8VugiYyCNTaxmJBNa3LgZkskI3ZtnYP80KiqbePalChT4+X3b2VTcyIihGZx6YnI/E7enuHqAichFQKH3fJlw7QseETlcVT/1Xt8OrFHVZ6Ph6SrdNVYnhF26g25ie3MfdcM7DB2czoXntJ1DPbXfqgWqOjnZeELLOyNPsN4oWhRRvdZ4QhW3pSkRGYQl+i4QkXGquiRebScZzwXeTHANNlL530TybGEDpVpMCimM3t7IrVdHlqU12Xj2lZKNZ18qbsasqhux2eG4yVv7PSqkeJWqvr+veUTkFKBvSPFyvO07PRVhziL7nKdUt1BHzV5lXcgjj/w955MmZPLxghqOOiK73faSjSdWhePpQV9Gy8Q957/9Zesz351BSeubLSLfAx5S1YR8larqv8Lw/DNRPD2l+XulWDewgoUcJ2eST/MuJhedv9PxHMBKWt9sVf0TtnabFEomnn4ymDTS26+4n5RsPAeqkrZnDsjzwBkATMKcMJ4HZqvq7d5s+j+BvwN/wYa7L2DOHx8C44Ffq+qiVtq+EbgD29vnRGwGei6WVK07cIaqNnpeXZcAXUVkHvARMBnb6GsE5nyh2HN0FTZTngb8NQzPWswvvIVH0TpdwVqWMpJD2MFWUkilgF6UUUI9dUxkGiJCmZZQTy2f60dUU0kGWfRnCNmDVzFqeAbHTcvm+b9WcMFZeTQ0Kn9+pZLTTs7h9uu7c/0dpWDLa89geZ+nEMbDyQ9LEetooJ5P9V260YMKyqijltyhFYwbk0FKCmRmCjnZQtnOJhoa4dvfyuG8M/P2GY8tYVUwgvE0UMeXzGfGlXUU5Kfw6cIarrk8n4/n1zB/US03X13AqSeGX6/ubB5gidY7qnoH5np4tfdzneeGORQ4X1V3Y0tmMzFvtV+o6j3A/cBPW2tYVe/FdhJ4W1XPxZbI1nveaw3ARM+f+3lsOasc++JIx9xZa7DlrEcwd9KJ2LryjzFDD8dzUGseRYUymnQy6U5vJsjXqGQX2XRhnExBECrYiaqyhP+QRgYjGc9gRiKYt9YJR3chNyeFe27rSe+eqfzp5Qoe/N/enHJCF7aXNZCRLtx2XQHY+vz9wE9b83DywzKccaSRTi/6oTQxgGGkkMrQwelcNqMb116ez6IltWRmpvDQzN5sL2vk5qsK9ilPoYwmn56sZRldJI8B2N5us+7qxU1XFvDw7F385he9mPXLXjzydOsDrs7mAZZoBTzKVmEeVS8A52O9223AJBEZAFSqrbPVA98RkZ969doN6lbV1d7LnZhjCNj6cB5mlN2xdepu3nmjxzMA+MD7nDFYjz0J809PiZani5iHUxrpZHveTulk0EA99dRRTx2NNLCVjVRRTjoZVFHOlEMz2VbawD/+r4ofXtiVmlplc3ED3QtSqapS0tPhhVcrwSb24sayhfU0UE8dtQhCF3KppZppUzJ58dUKVqyu42uTs1i4uJbauiYaGxUR2ac863Q5VZQT7G8+vNAGovndUhhWaI8FBfkpVFY2tfexHUIdwZgDHmWjgKWq+hnmKXQ65mI6B5hNc7DELZhh/9Irj1WlmMvqI9gz84vYUHwUNptd7PFMxXqYOdjjQMDZJa486WSQTiappNGHQQxhNBlkkUNXhgxKp7yiiTlvVTF+bAYD+6fxg+u2ceIxNpt894Nl5OakAGyJJ8sAhpFGOn0YRAG92E0FmWTTr08aGzc38NH8Gnr3TOXUk3K45a7t9Oyeus95CmUMQxhNQfvfD51GSWvM3uxxNyw441dYzxjYg+dlbONqxYbAw1V1ftC1k0Tkl8APgSGeB1q4zzgX8+66SESOxzzbLvLcOid4n5mB7a/7MhZN9SAWInkpZrDHAO9ibp0TsGipcUB3ETnJD89W3UQD9RTpOnboNmrYTRHrKNcyKtnFFjagNNGfQuqoZQULWcg8yiljMKOY82YVOV1SSBH408uVNDXCsq/qWLi4lvWbGhg0II2339sNNqKIC8t4juALPqKOGlawiBKK2cxa0slg3sc1HDs1m01FDSxZVs+h4zL4cnkdO3Y28dZ7uznr1Nx9wrOKxazT5azhSxslaA0lFDPv4xq+Wl3Hsy9VsHhpHZ99UcNTL5SzflMDb73XMoVPR5PLAZZAJZtHkePpWDyhStqe2cnJyZ+SfmkqHhKRXCz2OViVqvpSInmy6EKRrgOwZ2AZmDAWsKWXIl2XMJZk5OlISjpjFpGJQL6qzo3ivYOw/GCLg4MrVLUSeDKK9p7Aoqm27guertL9if5SGHF7X+qndKcPmWTRXXxlKQJg4+Z6rr29FILDHoP+Np2BZxAj6Cqd1/+6LSXjMHsilv7Htzx/47/HkeUSIDtZeMYyGaWJMkqiev+gAemcdnLkie86Is+BasgQh545TI6we4FfYZkvR2OJ+t4WkYu98l9js749gSeAbwAjsTxiWcB/AflejO0fMG+q6zGnkDHAvaq6TEQOCleOrVme6cUin4/tVn80toa8NAzPC8DFmEPIdZjDx0TgFMwRJBsoF5GHsXXQ9dhs9l+x5anx2FrzPZhDyxbvyMeWtNaG8NRh3mONqaSyXbfSQ/pQpOtYxWL6MJAi1pNGGqM4hI2spoKdTOQoVvIFShOppLFMP6OOWrLpQinF9KI/Z11czJJldUw7Iosbrsjn+1dupU/vNPr2TqVsVyO9e6RROCgN7+97XTLwHDk5kLcBROQK7z6YAGSE87gq1x0s4RO60YOD5XCKdB2bWcMYJrGCRfTXQvpLIZW6i3WsIJdu7KaCQsawZHkt3718K5MnZjL7/t78z292kJ2Vwk1XFnDJNVvJSBf+cO/eI4wDzQMsNEfYduB+Vb0buAEzGFT1CSzq6DPPw6oWyFPVS7HkeCep6jYsj9hcLwfYFiylzx88D6pHaF6PbK18AubU8RDm9PEv4ErMUMPxvOZdq/PqP4Il7uuPJQL8DFtXnoQNmY/C3DNHY/nAUoAN3ueXYl8mj2BJBm8Iw/MAtkZ+cja5rMYiD/tLIV3Ioxf9GcUhCCk00MBAhtGDPtRRQw/6kEc+BfSikp0MYSS72M4gRrKbSm6+qoAmVQYPSOOy67fRvXsqW7c1ctn0bsxfVMu9d/Rk8bI6MNfTpOAJ0WXA3ao6EPivcB5XXaU7QxhNqtcPpZDCCMaTJ/l7rSkvYwEDGUahjKYvg1nJF4wbk8kVl3Rj8IA0UlKEZSvredGcVhjYLy0czwHnARaaIywNOE5EbsP+OaGr9m15W4XTBPb2AjuknfK5mH/0PK/96cBTWC7s1nj+hK0hF2Hrxl+G4ZkAnIxl9xwGVHqf2xcLJZwIvIQZLkBfVa1oi6eOOupCEtJlk0tfBlNHLRlkUUYpOXSlMSQLZQW72M5WyrFUPqmkMaIwnS0ljbzyehWLltRx9qm5pKYCqmwtaSQvN4VjvpYNltEjKXhCdBFwuZfF5bDQiwH1ZTClFNOg9ZRRSoG0dAwJ8KzT5ZRRssf4zz41h1der2Lh4lpOPzmH1FRYsaqOyqqmcDwdSvGgD80RdhHQX1X/B+vB/KoRy8dVICKDsfjh4d61kVgGR9ooHw687/HUY73+/2GOHE+E+0BVrcYS+f0E+DTkchOWDmgJFiwxH/M2ez3oc4uwlETLsWH7AzR/UbXKk95KpFGqpJJNDhtZtScJfbPMPTGXPLrRg64U0Jv+9KQfK9fWc+i4TPr1TaVLtjBmZDrnnZnLNbeVMsxLXrfG8l5VJgtPiAap6gVYoMv1Sng3y1RJpQ8DWcGiMDymPLrRmwEUyhiGMoae9AOgd880+vVN5a77d3DmKTl7eKZO2fcx1/ta8ZjNDs0R9jjwqIjci6Wg7SYiZ2PPpAEPq9dozvsV6A3Hi8g/sR0mvoc9ez+MeVrdJCIrsaHtD7zPbVHuzR4fCwzyku6NAzZ7ZduwCa3WePDq3eh9/jTsGX43ZrQbgR9h/tePYRFaWR7PNOxL6GLvcw4DprTHU0/dkCYa2aqbSCN9j1dTL7XJ3TJKGMl4trCBXWynkUZSSGU3lWSRw2ZW043uLGUBBfTkD0/t4tH7evPBp9WkCDz1QgW9e6SycHEtH/1jIBs31/Pvj6rBcrH9Ohl45rxZBZArImOBMzzvOwVeEVKube2mG8hwPuEdxjAJgErdxU5KqaSc7tqHsUxmPV/RRXOppYY+NC9tnXdGLm/O3U1uTgrnnZHLXfeX8crjHT8e2nmAJVAiUoJNqAWrJ/bs3Z5iqTdEw+wq6Hg6Fk8Lqao7kujAtvDZ7/UcT8fnSTqnkX2pWHKK7Uee3ESwgOPp6DqgjFlVV2Ez0EmhcDwicmWCcBxPB1fHnovvnHo0QfX21+c4nvjUayE3Aebk1EnkemYnp04iZ8xOTp1EzpidnDqJnDHvZ4nIFBGZKyIfishUr+wcEVknIs95LqyxtP89ESlr5VqmiNwoItVeEoCE8YjpGRG5TUTuEJGnRSQnwX+fmSLyWxG5VUReE5FRieQJqvOYiMxvqw7gnEYScQB3AveFlM0FTotT++taKT8WCxJRIDeRPNiG7z8POv89cH2C/z630zwpfC3wWCJ5vGsXArOIwJnkgFpn7igSka5YsMZKLOvna6r6hogcCVyDhWVOBG5R1Q0i0gPzF1+OBbuEjWJQ1fe89hPOo6qN2G4iAaWAt4t6Ang8pl8EnY7A4t8TxuPF7I/FYuentcfhjDlxOl5EHgg6HxH0+lbMM+1XIpINLBeR4Vgo5a2qulZEzgKuwgJDbgH+o6p3e8PnmR2JR0QKsRHD1YnmEUu2cRMWNHRDonhEpAuWWvoyLCd7u3LGnDi9q6p7bhaRoL1HLaJsu4jc4p0vxqLTqoErRaQU25onw7t+MPBHsBxaXkBAh+ARkYFYwojzVTU4mDohPKq6ArhURC7B4rzPSxDPCVic/3XYF11fr/3H1ZJ4tJAz5uTU58AWVf0tgIhciGVweQT4q6o+LSInAxd49ZdiKYwC2S3jvY3DPuHxeq87gctVtVxEzlbVlxPIc6Pa/mNg6Z6Ghau3P3hUdQ62OwoichwwSVXbHOE4Y97PEpHJWPx2hogcqaofi8i3sdjq80Xkc6ynukdEfoZ9mxer7Ub5LHCViAwFBgETvPZmArNF5D5sC51yEblCVX8X8tmF2IQKWCz4c1jgwn7nEYvv/jcWT/2a9xy/UkTWJ+rvAxwutvNIFZYm6ieJ/H8F3S/TgX4icktbBu3cOZ2cOoncOrOTUyeRM2Ynp04iZ8xOTp1EzpidnDqJnDE7OXUSOWN2cuokcsbs5NRJ9P8BpLjCtPV021oAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1152x576 with 4 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "predict_custom_seq()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f496caf-c8e4-4020-8cf9-82497377da20",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
