{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8439389d-c89d-4d0a-90ae-fbe6706cd498",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-08-09 11:52:54.420790: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2022-08-09 11:52:54.420855: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import time\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import h5py\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "import random\n",
    "import json\n",
    "\n",
    "import tensorflow as tf\n",
    "from tensorflow import keras"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0256fa07-f9b1-4e18-bed2-a0ab93ef20f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_size = (15, 15)\n",
    "font = {'family': 'serif', 'size': 12}\n",
    "plt.rc('font', **font)\n",
    "\n",
    "batch_size = 100\n",
    "test_batches = 1\n",
    "n_topk = 2\n",
    "max_seq_len = 25\n",
    "\n",
    "base_path = \"../log_08_08_22_rnn/\"\n",
    "predict_rnn = True\n",
    "\n",
    "#base_path = \"log_08_08_22_rnn/\"\n",
    "#predict_rnn = True # set to True for RNN model\n",
    "\n",
    "# log_08_08_22_2 (finish time: 40,000 steps in 158683.60082054138 seconds)\n",
    "# log_08_08_22_rnn (finish time: 40,000 steps in 173480.83078551292 seconds)\n",
    "\n",
    "#\"log_03_08_22_1/\" Balanced data with really selection of low freq tools - random choice\n",
    "# RNN: log_01_08_22_3_rnn\n",
    "# Transformer: log_01_08_22_0\n",
    "\n",
    "model_number = 40000\n",
    "model_path = base_path + \"saved_model/\" + str(model_number) + \"/tf_model/\"\n",
    "\n",
    "\n",
    "def remove_file(file_path):\n",
    "    if os.path.exists(file_path):\n",
    "        os.remove(file_path)\n",
    "\n",
    "\n",
    "def read_file(file_path):\n",
    "    \"\"\"\n",
    "    Read a file\n",
    "    \"\"\"\n",
    "    with open(file_path, \"r\") as json_file:\n",
    "        file_content = json.loads(json_file.read())\n",
    "    return file_content\n",
    "\n",
    "\n",
    "def write_file(file_path, content):\n",
    "    \"\"\"\n",
    "    Write a file\n",
    "    \"\"\"\n",
    "    remove_file(file_path)\n",
    "    with open(file_path, \"w\") as json_file:\n",
    "        json_file.write(json.dumps(content))\n",
    "\n",
    "\n",
    "r_dict = read_file(base_path + \"data/rev_dict.txt\")\n",
    "f_dict = read_file(base_path + \"data/f_dict.txt\")\n",
    "\n",
    "path_test_data = base_path + \"saved_data/test.h5\"\n",
    "\n",
    "file_obj = h5py.File(path_test_data, 'r')\n",
    "\n",
    "test_input = np.array(file_obj[\"input\"])\n",
    "test_target = np.array(file_obj[\"target\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0524ee9c-2b57-432b-a05c-5533271ebf2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.]])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "27ae2b83-55f4-4075-8c63-babcf9b86d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generated_attention(attention_weights, i_names, f_dict, r_dict):\n",
    "\n",
    "    #print(attention_weights.shape)\n",
    "    attention_heads = tf.squeeze(attention_weights, 0)\n",
    "    n_heads = attention_heads.shape[1]\n",
    "    i_names = i_names.split(\",\")\n",
    "    in_tokens = i_names\n",
    "    out_tokens = i_names\n",
    "    fig = plt.figure(figsize=(16, 8))\n",
    "\n",
    "    for h, head in enumerate(attention_heads):\n",
    "      ax = fig.add_subplot(2, n_heads, h+1)\n",
    "      plot_attention_head(in_tokens, out_tokens, head)\n",
    "      ax.set_xlabel(f'Head {h+1}')\n",
    "    plt.tight_layout()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9390e063-d2c1-4a5b-83f2-1ec4235a81d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_attention_head(in_tokens, out_tokens, attention):\n",
    "  # The plot is of the attention when a token was generated.\n",
    "  # The model didn't generate `<START>` in the output. Skip it.\n",
    "  #translated_tokens = translated_tokens[1:]\n",
    "  #print(attention)\n",
    "  ax = plt.gca()\n",
    "  ax.matshow(attention[:len(in_tokens), :len(out_tokens)])\n",
    "  #ax.matshow(attention[:len(in_tokens), :])\n",
    "\n",
    "  ax.set_xticks(range(len(in_tokens)))\n",
    "  ax.set_yticks(range(len(out_tokens)))\n",
    "\n",
    "  ax.set_xticklabels(in_tokens, rotation=90)\n",
    "  ax.set_yticklabels(out_tokens)\n",
    "\n",
    "\n",
    "def visualize_loss_acc():\n",
    "    epo_tr_batch_loss = read_file(base_path + \"data/epo_tr_batch_loss.txt\").split(\",\")\n",
    "\n",
    "    #print(len(epo_tr_batch_loss))\n",
    "    epo_tr_batch_loss = [np.round(float(item), 4) for item in epo_tr_batch_loss]\n",
    "\n",
    "    epo_tr_batch_acc = read_file(base_path + \"data/epo_tr_batch_acc.txt\").split(\",\")\n",
    "    epo_tr_batch_acc = [np.round(float(item), 4) for item in epo_tr_batch_acc]\n",
    "\n",
    "    epo_te_batch_loss = read_file(base_path + \"data/epo_te_batch_loss.txt\").split(\",\")\n",
    "    epo_te_batch_loss = [np.round(float(item), 4) for item in epo_te_batch_loss]\n",
    "\n",
    "    epo_te_batch_acc = read_file(base_path + \"data/epo_te_precision.txt\").split(\",\")\n",
    "    epo_te_batch_acc = [np.round(float(item), 4) for item in epo_te_batch_acc]\n",
    "\n",
    "    plot_loss_acc(epo_tr_batch_loss, epo_tr_batch_acc, \"training\")\n",
    "    plot_loss_acc(epo_te_batch_loss, epo_te_batch_acc, \"test\")\n",
    "\n",
    "    epo_te_low_batch_acc = read_file(base_path + \"data/epo_low_te_precision.txt\").split(\",\")\n",
    "    epo_te_low_batch_acc = [np.round(float(item), 4) for item in epo_te_low_batch_acc]\n",
    "\n",
    "    plot_low_te_prec(epo_te_low_batch_acc, \"Low test\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fa590092-8a14-485f-ab29-612852d95e5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-08-09 11:53:35.145809: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory\n",
      "2022-08-09 11:53:35.145831: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)\n",
      "2022-08-09 11:53:35.145965: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (anupkumar-HP-ProBook-450-G5): /proc/driver/nvidia/version does not exist\n",
      "2022-08-09 11:53:35.146721: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    }
   ],
   "source": [
    "def load_model():\n",
    "    return tf.saved_model.load(model_path)\n",
    "tf_loaded_model = load_model()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad68747f-5586-46d3-bb27-ce601e6907b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_custom_seq():\n",
    "\n",
    "    n_topk_ind = 20\n",
    "    t_ip = np.zeros((25))\n",
    "    t_ip[0] = int(f_dict[\"snpEff_build_gb\"])\n",
    "    t_ip[1] = int(f_dict[\"bwa_mem\"])\n",
    "    t_ip[2] = int(f_dict[\"samtools_view\"])\n",
    "\n",
    "    last_tool_name = \"samtools_view\"\n",
    "    \n",
    "    t_ip = tf.convert_to_tensor(t_ip, dtype=tf.int64)\n",
    "    if predict_rnn is True:\n",
    "        prediction = tf_loaded_model([t_ip], training=False)\n",
    "    else:\n",
    "        prediction, att_weights = tf_loaded_model([t_ip], training=False)\n",
    "    prediction_cwts = tf.math.multiply(c_weights, prediction)\n",
    "\n",
    "    top_k = tf.math.top_k(prediction, k=n_topk_ind, sorted=True)\n",
    "    top_k_wts = tf.math.top_k(prediction_cwts, k=n_topk_ind, sorted=True)\n",
    "\n",
    "    t_ip = t_ip.numpy()\n",
    "    label_pos = np.where(t_ip > 0)[0]\n",
    "\n",
    "    i_names = \",\".join([r_dict[str(item)] for item in t_ip[label_pos]  if item not in [0, \"0\"]])\n",
    "\n",
    "    pred_tools = [r_dict[str(item)] for item in top_k.indices.numpy()[0]  if item not in [0, \"0\"]]\n",
    "    pred_tools_wts = [r_dict[str(item)] for item in top_k_wts.indices.numpy()[0]  if item not in [0, \"0\"]]\n",
    "\n",
    "    c_tools = [r_dict[str(item)] for item in compatible_tools[str(f_dict[last_tool_name])]]\n",
    "\n",
    "    pred_intersection = list(set(pred_tools).intersection(set(c_tools)))\n",
    "    prd_te_prec = len(pred_intersection) / float(n_topk_ind)\n",
    "\n",
    "    print(\"Tool sequence: {}\".format([r_dict[str(item)] for item in t_ip[label_pos]]))\n",
    "    print()\n",
    "    print(\"Compatible true tools: {}, size: {}\".format(c_tools, len(c_tools)))\n",
    "    print()\n",
    "    print(\"Predicted top {} tools: {}\".format(n_topk_ind, pred_tools))\n",
    "    print()\n",
    "    print(\"Predicted precision: {}\".format(prd_te_prec))\n",
    "    print()\n",
    "    print(\"Correctly predicted tools: {}\".format(pred_intersection))\n",
    "    print()\n",
    "    print(\"Predicted top {} tools with weights: {}\".format(n_topk_ind, pred_tools_wts))\n",
    "    print()\n",
    "    if predict_rnn is False:\n",
    "        generated_attention(att_weights, i_names, f_dict, r_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "991ec9f5-77a9-4278-903e-2ee8ac1d5ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_low_freq_seq():\n",
    "    te_lowest_t_ids = read_file(base_path + \"data/te_lowest_t_ids.txt\")\n",
    "    lowest_t_ids = [int(item) for item in te_lowest_t_ids.split(\",\")]\n",
    "    \n",
    "    low_te_data = test_input[lowest_t_ids]\n",
    "    low_te_labels = test_target[lowest_t_ids]\n",
    "    #print(\"Test lowest ids\", low_te_data.shape, low_te_labels.shape)\n",
    "    #low_te_pred_batch, low_att_weights = tf_loaded_model([low_te_data], training=False)\n",
    "    low_topk = 5\n",
    "    low_te_precision = list()\n",
    "    for i, (low_inp, low_tar) in enumerate(zip(low_te_data, low_te_labels)):\n",
    "        if predict_rnn is True:\n",
    "            low_prediction = tf_loaded_model([low_inp], training=False)\n",
    "        else:\n",
    "            low_prediction, att_weights = tf_loaded_model([low_inp], training=False)\n",
    "        #print(low_prediction.shape)\n",
    "        \n",
    "        low_label_pos = np.where(low_tar > 0)[0]\n",
    "        low_topk = len(low_label_pos)\n",
    "        low_topk_pred = tf.math.top_k(low_prediction, k=low_topk, sorted=True)\n",
    "        low_topk_pred = low_topk_pred.indices.numpy()[0]\n",
    "        \n",
    "        low_label_pos_tools = [r_dict[str(item)] for item in low_label_pos if item not in [0, \"0\"]]\n",
    "        low_pred_label_pos_tools = [r_dict[str(item)] for item in low_topk_pred if item not in [0, \"0\"]]\n",
    "\n",
    "        low_intersection = list(set(low_label_pos_tools).intersection(set(low_pred_label_pos_tools)))\n",
    "        low_pred_precision = float(len(low_intersection)) / len(low_label_pos)\n",
    "        low_te_precision.append(low_pred_precision)\n",
    "        \n",
    "        #low_inp = low_inp.numpy()\n",
    "        low_inp_pos = np.where(low_inp > 0)[0]\n",
    "\n",
    "        print(\"Low: test tool sequence: {}\".format([r_dict[str(int(item))] for item in low_inp[low_inp_pos]]))\n",
    "        print()\n",
    "        print(\"Low: True labels: {}\".format(low_label_pos_tools))\n",
    "        print()\n",
    "        print(\"Low: Predicted labels: {}, Precision: {}\".format(low_pred_label_pos_tools, low_pred_precision))\n",
    "        print(\"-----------------\")\n",
    "        print()\n",
    "    \n",
    "    if len(lowest_t_ids) > 0:\n",
    "        print(\"Low test prediction precision: {}\".format(np.mean(low_te_precision)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d23c48f7-8ff3-4342-85c2-efe1e3ffe42b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def verify_tool_in_tr(r_dict):\n",
    "    all_sel_tool_ids = read_file(base_path + \"data/all_sel_tool_ids.txt\").split(\",\")\n",
    "\n",
    "    freq_dict = dict()\n",
    "    freq_dict_names = dict()\n",
    "\n",
    "    for tool_id in all_sel_tool_ids:\n",
    "        if tool_id not in freq_dict:\n",
    "            freq_dict[tool_id] = 0\n",
    "\n",
    "        if tool_id not in freq_dict_names:\n",
    "            freq_dict_names[r_dict[str(int(tool_id))]] = 0\n",
    "\n",
    "        freq_dict[tool_id] += 1\n",
    "        freq_dict_names[r_dict[str(int(tool_id))]] += 1\n",
    "\n",
    "    s_freq = dict(sorted(freq_dict.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    s_freq_names = dict(sorted(freq_dict_names.items(), key=lambda kv: kv[1], reverse=True))\n",
    "\n",
    "\n",
    "    write_file(base_path + \"data/s_freq_names.txt\", s_freq_names)\n",
    "    write_file(base_path + \"data/s_freq.txt\", s_freq)\n",
    "\n",
    "    return s_freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0e7626fc-8703-4aac-afa0-c72a41aa6364",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_u_tr_labels(y_tr):\n",
    "    labels = list()\n",
    "    labels_pos_dict = dict()\n",
    "    for i, item in enumerate(y_tr):\n",
    "        label_pos = np.where(item > 0)[0]\n",
    "        labels.extend(label_pos)\n",
    "        for label in label_pos:\n",
    "            if label not in labels_pos_dict:\n",
    "                labels_pos_dict[label] = list()\n",
    "            labels_pos_dict[label].append(i)\n",
    "\n",
    "    u_labels = list(set(labels))\n",
    "    \n",
    "    for item in labels_pos_dict:\n",
    "        labels_pos_dict[item] = list(set(labels_pos_dict[item]))\n",
    "    return u_labels, labels_pos_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "06683e24-e729-458e-b9fb-a1d4cd1f650d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_seq():\n",
    "\n",
    "    #visualize_loss_acc()\n",
    "    \n",
    "    tool_tr_freq = read_file(base_path + \"data/all_sel_tool_ids.txt\")\n",
    "    \n",
    "    #sys.exit()\n",
    "\n",
    "    #verify_training_sampling(tool_tr_freq, r_dict)\n",
    "    \n",
    "    r_dict = read_file(base_path + \"data/rev_dict.txt\")\n",
    "    f_dict = read_file(base_path + \"data/f_dict.txt\")\n",
    "    \n",
    "    all_tr_label_tools = verify_tool_in_tr(r_dict)\n",
    "\n",
    "    all_tr_label_tools_ids = list(all_tr_label_tools.keys())\n",
    "    all_tr_label_tools_ids = [int(t) for t in all_tr_label_tools_ids]\n",
    "    \n",
    "    \n",
    "    print(test_input.shape, test_target.shape)\n",
    "\n",
    "    class_weights = read_file(base_path + \"data/class_weights.txt\")\n",
    "    compatible_tools = read_file(base_path + \"data/compatible_tools.txt\")\n",
    "\n",
    "    #tool_freq = utils.read_file(base_path + \"data/freq_dict_names.txt\")\n",
    "    published_connections = read_file(base_path + \"data/published_connections.txt\")\n",
    "\n",
    "    c_weights = list(class_weights.values())\n",
    "\n",
    "    c_weights = tf.convert_to_tensor(c_weights, dtype=tf.float32)\n",
    "\n",
    "    u_te_y_labels, u_te_y_labels_dict = get_u_tr_labels(test_target)\n",
    "\n",
    "    precision = list()\n",
    "    pub_prec_list = list()\n",
    "    error_label_tools = list()\n",
    "\n",
    "    for j in range(test_batches):\n",
    "        #te_x_batch, y_train_batch = sample_balanced(test_input, test_target, ulabels_te_dict)\n",
    "        #te_x_batch, y_train_batch, selected_label_tools, bat_ind = sample_balanced_tr_y(test_input, test_target, u_te_y_labels_dict)\n",
    "        print(j * batch_size, j * batch_size + batch_size)\n",
    "        te_x_batch = test_input[j * batch_size : j * batch_size + batch_size, :]\n",
    "        y_train_batch = test_target[j * batch_size : j * batch_size + batch_size, :]\n",
    "\n",
    "        for i, (inp, tar) in enumerate(zip(te_x_batch, y_train_batch)):\n",
    "\n",
    "            t_ip = inp\n",
    "            if len(np.where(inp > 0)[0]) < max_seq_len:\n",
    "                real_prediction = np.where(tar > 0)[0]\n",
    "                target_pos = list(set(all_tr_label_tools_ids).intersection(set(real_prediction)))\n",
    "                #print(\"Real predicted tools: {}, Actual trained labels: {}\".format(real_prediction, target_pos))\n",
    "                #print()\n",
    "                t_ip = tf.convert_to_tensor(t_ip, dtype=tf.int64)\n",
    "                if predict_rnn is True:\n",
    "                    prediction = tf_loaded_model([t_ip], training=False)\n",
    "                else:\n",
    "                    prediction, att_weights = tf_loaded_model([t_ip], training=False)\n",
    "                prediction_wts = tf.math.multiply(c_weights, prediction)\n",
    "\n",
    "                n_topk = len(target_pos)\n",
    "                top_k = tf.math.top_k(prediction, k=n_topk, sorted=True)\n",
    "                top_k_wts = tf.math.top_k(prediction_wts, k=n_topk, sorted=True)\n",
    "\n",
    "                t_ip = t_ip.numpy()\n",
    "                label_pos = np.where(t_ip > 0)[0]\n",
    "\n",
    "                i_names = \",\".join([r_dict[str(item)] for item in t_ip[label_pos]  if item not in [0, \"0\"]])\n",
    "                t_names = \",\".join([r_dict[str(int(item))] for item in target_pos  if item not in [0, \"0\"]])\n",
    "\n",
    "                last_i_tool = [r_dict[str(item)] for item in t_ip[label_pos]][-1]\n",
    "\n",
    "                true_tools = [r_dict[str(int(item))] for item in target_pos]\n",
    "\n",
    "                #print(\"Selected tool in true tools: {}\".format(s_label_tool in true_tools))\n",
    "\n",
    "                pred_tools = [r_dict[str(item)] for item in top_k.indices.numpy()[0]  if item not in [0, \"0\"]]\n",
    "                pred_tools_wts = [r_dict[str(item)] for item in top_k_wts.indices.numpy()[0]  if item not in [0, \"0\"]]\n",
    "\n",
    "                intersection = list(set(true_tools).intersection(set(pred_tools)))\n",
    "\n",
    "                pub_prec = 0.0\n",
    "                pub_prec_wt = 0.0\n",
    "\n",
    "                if last_i_tool in published_connections:\n",
    "                    true_pub_conn = published_connections[last_i_tool]\n",
    "                    #print(\"Test batch {}, True published tools: {}\".format(j+1, true_pub_conn)) \n",
    "                    #print()\n",
    "                    if len(pred_tools) > 0:\n",
    "                        intersection_pub = list(set(true_pub_conn).intersection(set(pred_tools)))\n",
    "                        intersection_pub_wt = list(set(true_pub_conn).intersection(set(pred_tools_wts)))\n",
    "                        pub_prec = float(len(intersection_pub)) / len(pred_tools)\n",
    "                        pub_prec_list.append(pub_prec)\n",
    "                        pub_prec_wt = float(len(intersection_pub_wt)) / len(pred_tools)\n",
    "                    else:\n",
    "                        pub_prec = False\n",
    "                        pub_prec_wt = False\n",
    "\n",
    "                if len(pred_tools) > 0:\n",
    "                    pred_precision = float(len(intersection)) / len(pred_tools)\n",
    "                    precision.append(pred_precision)\n",
    "\n",
    "                if pred_precision < 1.0:\n",
    "            \n",
    "                    print(\"Test batch {}, Tool sequence: {}\".format(j+1, [r_dict[str(item)] for item in t_ip[label_pos]]))\n",
    "                    print()\n",
    "                    print(\"Test batch {}, True tools: {}\".format(j+1, true_tools))\n",
    "                    print()\n",
    "                    print(\"Test batch {}, Predicted top {} tools: {}\".format(j+1, n_topk, pred_tools))\n",
    "                    print()\n",
    "                    print(\"Test batch {}, Predicted top {} tools with weights: {}\".format(j+1, n_topk, pred_tools_wts))\n",
    "                    print()\n",
    "                    print(\"Test batch {}, Precision: {}\".format(j+1, pred_precision)) \n",
    "                    print()\n",
    "                    print(\"Test batch {}, Published precision: {}\".format(j+1, pub_prec))\n",
    "                    print()\n",
    "                    print(\"Test batch {}, Published precision with weights: {}\".format(j+1, pub_prec_wt))\n",
    "                    #error_label_tools.append(select_tools[i])\n",
    "                    print(\"=========================\")\n",
    "                print(\"--------------------------\")\n",
    "                #generated_attention(att_weights, i_names, f_dict, r_dict)\n",
    "        \n",
    "                print(\"Batch {} prediction finished ...\".format(j+1))\n",
    "\n",
    "    if test_batches > 0:\n",
    "        print(\"Precision@{}: {}\".format(n_topk, np.mean(precision)))\n",
    "        print(\"Published Precision@{}: {}\".format(n_topk, np.mean(pub_prec_list)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "34f50e20-a934-463a-81b6-4393d36a9d25",
   "metadata": {},
   "outputs": [],
   "source": [
    "#predict_seq()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "16027192-45ed-4766-b6a1-1d114f23b244",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Low: test tool sequence: ['rbc_splitfasta', 'hhsearch', 'collection_element_identifiers', 'ffindex_dbkit_create', 'ffindex_dbkit_merge']\n",
      "\n",
      "Low: True labels: ['spring_model_all']\n",
      "\n",
      "Low: Predicted labels: ['spring_model_all'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['filtlong', 'flye', 'racon', 'maxbin2', 'prokka', 'roary']\n",
      "\n",
      "Low: True labels: ['taxonomy_krona_chart', 'barchart_gnuplot', 'CONVERTER_fasta_to_fai', 'iqtree', 'fasttree', 'raxml', 'gubbins', 'gd_raxml', 'graphlan']\n",
      "\n",
      "Low: Predicted labels: ['graphlan', 'raxml', 'barchart_gnuplot', 'CONVERTER_fasta_to_fai', 'fasttree', 'iqtree', 'taxonomy_krona_chart', 'gubbins', 'gd_raxml'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_groomer', '__ZIP_COLLECTION__']\n",
      "\n",
      "Low: True labels: ['__FILTER_EMPTY_DATASETS__', '__RELABEL_FROM_FILE__', 'hisat2', 'snippy_core', 'fastq_groomer', 'OMSSAAdapter', 'bwa_mem', 'trimmomatic', 'fastqc', 'salmon']\n",
      "\n",
      "Low: Predicted labels: ['fastq_groomer', 'fastqc', 'bwa_mem', 'salmon', 'hisat2', '__FILTER_EMPTY_DATASETS__', 'OMSSAAdapter', 'snippy_core', 'trimmomatic', '__RELABEL_FROM_FILE__'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['mspurity_puritya', 'abims_xcms_xcmsSet', 'xcms_merge']\n",
      "\n",
      "Low: True labels: ['abims_xcms_group', 'xcms_plot_chromatogram', 'quality_metrics']\n",
      "\n",
      "Low: Predicted labels: ['xcms_plot_chromatogram', 'quality_metrics', 'abims_xcms_group'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'bwa_mem', 'samtools_sort', 'picard_MarkDuplicates', 'gatk2_haplotype_caller', 'gatk2_variant_select']\n",
      "\n",
      "Low: True labels: ['gatk2_variant_filtration']\n",
      "\n",
      "Low: Predicted labels: ['gatk2_variant_filtration'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['sklearn_pca', 'sklearn_numeric_clustering']\n",
      "\n",
      "Low: True labels: ['macs2_callpeak', 'ggplot2_point', 'sklearn_nn_classifier']\n",
      "\n",
      "Low: Predicted labels: ['ggplot2_point', 'macs2_callpeak', 'sklearn_nn_classifier'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['porechop', 'filtlong', 'racon', 'rbc_splitfasta']\n",
      "\n",
      "Low: True labels: ['ncbi_blastp_wrapper', 'hhsearch', 'interproscan', 'mlst', 'mitos2', '__EXTRACT_DATASET__']\n",
      "\n",
      "Low: Predicted labels: ['__EXTRACT_DATASET__', 'mitos2', 'hhsearch', 'mlst', 'ncbi_blastp_wrapper', 'interproscan'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_awk_tool', 'regionalgam_ab_index', 'regionalgam_glmmpql', 'regionalgam_gls']\n",
      "\n",
      "Low: True labels: ['regionalgam_plot_trend', 'regionalgam_autocor_acf']\n",
      "\n",
      "Low: Predicted labels: ['regionalgam_plot_trend', 'regionalgam_autocor_acf'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['basic_illumination', 'ashlar', 'unet_coreograph']\n",
      "\n",
      "Low: True labels: ['unmicst', 'quantification', 's3segmenter']\n",
      "\n",
      "Low: Predicted labels: ['quantification', 'ConsensusID', 'rna_abbased_data'], Precision: 0.3333333333333333\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['FeatureFinderMultiplex', 'HighResPrecursorMassCorrector', 'MSGFPlusAdapter', 'PeptideIndexer', 'FalseDiscoveryRate', 'IDMapper', 'FileFilter', 'MapAlignerPoseClustering', 'MapRTTransformer']\n",
      "\n",
      "Low: True labels: ['FeatureLinkerUnlabeledQT', 'FeatureLinkerUnlabeledKD']\n",
      "\n",
      "Low: Predicted labels: ['FeatureLinkerUnlabeledQT', 'FeatureLinkerUnlabeledKD'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['FeatureFinderCentroided', 'IDMapper', 'QCCalculator']\n",
      "\n",
      "Low: True labels: ['QCExtractor']\n",
      "\n",
      "Low: Predicted labels: ['QCExtractor'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['CONVERTER_gz_to_uncompressed', 'cat1', 'cutadapt', 'chira_collapse', 'cshl_fastx_reverse_complement', 'chira_map', 'chira_merge', 'chira_quantify']\n",
      "\n",
      "Low: True labels: ['chira_extract']\n",
      "\n",
      "Low: Predicted labels: ['chira_extract'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cat1', '_ensembl_gtf2gene_list']\n",
      "\n",
      "Low: True labels: ['bg_column_arrange_by_header', 'scanpy_read_10x', 'scanpy_filter_genes', 'alevin', 'tp_replace_in_column', 'tp_awk_tool', 'join1', 'Cut1']\n",
      "\n",
      "Low: Predicted labels: ['scanpy_filter_genes', 'scanpy_read_10x', 'tp_awk_tool', 'bg_column_arrange_by_header', 'alevin', 'Cut1', 'join1', 'tp_replace_in_column'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.fasta.remove_desc', 'get_orfs_or_cdss', 'edu.tamu.cpt.gff3.cdsParents', 'edu.tamu.cpt.gff3.require_phage_start', 'edu.tamu.cpt.gff3.require_sd', 'edu.tamu.cpt.gff3.export_seq', 'edu.tamu.cpt.fasta.seq_trans']\n",
      "\n",
      "Low: True labels: ['LipoP', 'TMHMM_to_gff3', 'interproscan', 'ncbi_blastp_wrapper_new', 'SignalP']\n",
      "\n",
      "Low: Predicted labels: ['ncbi_blastp_wrapper_new', 'LipoP', 'SignalP', 'TMHMM_to_gff3', 'interproscan'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['1_create_conf']\n",
      "\n",
      "Low: True labels: ['conf4circos', '4_filter_sam', '2_map', '3_filter_single_pair', '5_calc_stat']\n",
      "\n",
      "Low: Predicted labels: ['3_filter_single_pair', 'diffbind-embl', 'ia_coordinates_of_roi', 'NSPDK_candidateClust', 'plotdexseq'], Precision: 0.2\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['sklearn_feature_selection']\n",
      "\n",
      "Low: True labels: ['sklearn_nn_classifier', 'keras_train_and_eval', 'sklearn_lightgbm']\n",
      "\n",
      "Low: Predicted labels: ['sklearn_nn_classifier', 'sklearn_lightgbm', 'keras_train_and_eval'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['__FILTER_FROM_FILE__', 'unzip']\n",
      "\n",
      "Low: True labels: ['graphicsmagick_image_convert', 'ip_convertimage', 'ip_imageinfo', 'ip_filter_standard', 'tab2fasta', 'ip_2d_feature_extraction', 'gdal_gdalwarp', 'megahit', 'ip_histogram_equalization', 'cp_cellprofiler', 'collection_element_identifiers', 'ip_threshold']\n",
      "\n",
      "Low: Predicted labels: ['collection_element_identifiers', 'cp_cellprofiler', 'ip_threshold', 'tab2fasta', 'ip_imageinfo', 'megahit', 'graphicsmagick_image_convert', 'ip_2d_feature_extraction', 'ip_filter_standard', 'gdal_gdalwarp', 'ip_histogram_equalization', 'ip_convertimage'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['sklearn_build_pipeline', 'sklearn_estimator_attributes', 'sklearn_searchcv']\n",
      "\n",
      "Low: True labels: ['sklearn_ensemble', 'plotly_ml_performance_plots', 'ggplot2_point', 'plotly_parallel_coordinates_plot', 'sklearn_stacking_ensemble_models']\n",
      "\n",
      "Low: Predicted labels: ['sklearn_ensemble', 'plotly_ml_performance_plots', 'ggplot2_point', 'sklearn_stacking_ensemble_models', 'plotly_parallel_coordinates_plot'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msconvert', 'CometAdapter', 'IDPosteriorErrorProbability', 'IDFileConverter']\n",
      "\n",
      "Low: True labels: ['FalseDiscoveryRate', 'proteomics_search_peptide_prophet_1', 'PeptideIndexer', 'FileInfo', 'seq_composition']\n",
      "\n",
      "Low: Predicted labels: ['FileInfo', 'proteomics_search_peptide_prophet_1', 'seq_composition', 'FalseDiscoveryRate', 'PeptideIndexer'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['seurat_read10x', 'seurat_filter_cells', 'seurat_normalise_data', 'seurat_find_variable_genes', 'seurat_run_pca', 'seurat_find_neighbours', 'seurat_find_clusters', 'seurat_run_tsne', 'sceasy_convert']\n",
      "\n",
      "Low: True labels: ['scmap_scmap_cell', 'export_to_genap2', 'scmap_preprocess_sce', 'scanpy_filter_genes', 'scmap_scmap_cluster', 'anndata_ops', 'anndata_manipulate']\n",
      "\n",
      "Low: Predicted labels: ['scmap_scmap_cell', 'scmap_preprocess_sce', 'scmap_scmap_cluster', 'anndata_ops', 'anndata_manipulate', 'scanpy_filter_genes', 'export_to_genap2'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_to_fasta_python', 'seqtk_mergepe']\n",
      "\n",
      "Low: True labels: ['fastq_to_fasta_python', 'seqtk_mergefa', 'mothur_merge_files', 'mummer_nucmer', 'mummer_mummer', 'qiime_align_seqs', 'lastz_wrapper_2']\n",
      "\n",
      "Low: Predicted labels: ['qiime_align_seqs', 'seqtk_mergefa', 'mummer_mummer', 'lastz_wrapper_2', 'mothur_merge_files', 'mummer_nucmer', 'fastq_to_fasta_python'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['retrieve_scxa']\n",
      "\n",
      "Low: True labels: ['scanpy_read_10x', 'seurat_read10x']\n",
      "\n",
      "Low: Predicted labels: ['kraken-report', 'kraken_taxonomy_report'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['OpenSwathDecoyGenerator', 'TargetedFileConverter', 'OpenSwathWorkflow']\n",
      "\n",
      "Low: True labels: ['pyprophet_merge']\n",
      "\n",
      "Low: Predicted labels: ['pyprophet_merge'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['vcfsort', 'vcfbedintersect']\n",
      "\n",
      "Low: True labels: ['bedtools_links', 'vcfannotate', 'bedtools_annotatebed', 'vcffilter2']\n",
      "\n",
      "Low: Predicted labels: ['bedtools_annotatebed', 'vcfannotate', 'vcffilter2', 'bedtools_links'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['stacks2_ustacks', 'stacks2_cstacks', 'stacks2_sstacks']\n",
      "\n",
      "Low: True labels: ['stacks2_tsv2bam']\n",
      "\n",
      "Low: Predicted labels: ['stacks2_tsv2bam'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_paired_end_joiner', 'trimmomatic', 'fastq_to_fasta_python', 'bowtie2', 'metaphlan', 'humann2', 'humann_regroup_table', 'humann2_rename_table']\n",
      "\n",
      "Low: True labels: ['Grep1', 'humann2_join_tables']\n",
      "\n",
      "Low: Predicted labels: ['humann2_join_tables', 'Grep1'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['stacks2_procrad', 'stacks2_denovomap']\n",
      "\n",
      "Low: True labels: ['stacks2_populations', 'stacks_populations']\n",
      "\n",
      "Low: Predicted labels: ['stacks2_populations', 'stacks_populations'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_groomer', 'trimmomatic', 'abyss-pe']\n",
      "\n",
      "Low: True labels: ['staramr_search', 'augustus', 'bandage_image', 'funannotate_sort', 'mlst', 'ncbi_makeblastdb', 'prokka', 'quast']\n",
      "\n",
      "Low: Predicted labels: ['staramr_search', 'augustus', 'quast', 'ncbi_makeblastdb', 'mlst', 'bandage_image', 'prokka', 'funannotate_sort'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['sklearn_build_pipeline', 'sklearn_searchcv', 'sklearn_stacking_ensemble_models']\n",
      "\n",
      "Low: True labels: ['sklearn_model_validation']\n",
      "\n",
      "Low: Predicted labels: ['sklearn_ensemble'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['TrimPrimer', 'Flash', 'Btrim64', 'uparse', 'pynast']\n",
      "\n",
      "Low: True labels: ['fasttree', 'FastTree']\n",
      "\n",
      "Low: Predicted labels: ['FastTree', 'fasttree'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['split_file_to_collection', 'param_value_from_file', 'sklearn_train_test_split']\n",
      "\n",
      "Low: True labels: ['tp_sorted_uniq', 'sklearn_ensemble', 'cardinal_filtering', 'addName', 'sklearn_train_test_eval', 'Cut1']\n",
      "\n",
      "Low: Predicted labels: ['sklearn_ensemble', 'sklearn_train_test_eval', 'cardinal_filtering', 'tp_sorted_uniq', 'Cut1', 'addName'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['_ensembl_gtf2gene_list', 'alevin']\n",
      "\n",
      "Low: True labels: ['_salmon_kallisto_mtx_to_10x', '_dropletBarcodePlot']\n",
      "\n",
      "Low: Predicted labels: ['_dropletBarcodePlot', '_salmon_kallisto_mtx_to_10x'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['collection_element_identifiers']\n",
      "\n",
      "Low: True labels: ['regex_replace', '__RELABEL_FROM_FILE__', 'tp_replace_in_line', 'ffindex_dbkit_create', 'tp_awk_tool', 'split_file_to_collection', 'Paste1', 'collapse_dataset']\n",
      "\n",
      "Low: Predicted labels: ['tp_replace_in_line', '__RELABEL_FROM_FILE__', 'Paste1', 'tp_awk_tool', 'collapse_dataset', 'split_file_to_collection', 'regex_replace', 'ffindex_dbkit_create'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msconvert', 'OpenSwathWorkflow', 'pyprophet_merge', 'pyprophet_score']\n",
      "\n",
      "Low: True labels: ['pyprophet_peptide']\n",
      "\n",
      "Low: Predicted labels: ['pyprophet_peptide'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['deseq2', 'tp_cut_tool', 'heinz_scoring', 'heinz']\n",
      "\n",
      "Low: True labels: ['heinz_visualization']\n",
      "\n",
      "Low: Predicted labels: ['heinz_visualization'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['samtools_view', 'param_value_from_file', 'compose_text_param']\n",
      "\n",
      "Low: True labels: ['rdock_sort_filter', 'hisat2', 'tp_find_and_replace', 'Add_a_column1', 'tp_replace_in_column', 'sdf_to_tab', 'rxdock_sort_filter', 'tp_awk_tool', 'bowtie2', 'add_line_to_file', 'tp_text_file_with_recurring_lines', 'filter_tabular', 'regex1', 'Filter1', 'snpSift_filter']\n",
      "\n",
      "Low: Predicted labels: ['rxdock_sort_filter', 'Filter1', 'sdf_to_tab', 'add_line_to_file', 'tp_text_file_with_recurring_lines', 'filter_tabular', 'snpSift_filter', 'rdock_sort_filter', 'hisat2', 'Add_a_column1', 'bowtie2', 'tp_awk_tool', 'tp_find_and_replace', 'regex1', 'tp_replace_in_column'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['PeakPickerHiRes', 'FileFilter', 'xcms-find-peaks', 'xcms-collect-peaks', 'xcms-group-peaks', 'xcms-blankfilter', 'xcms-dilutionfilter', 'camera-annotate-peaks', 'camera-group-fwhm', 'camera-find-adducts', 'camera-find-isotopes', 'msms2metfrag-multiple']\n",
      "\n",
      "Low: True labels: ['metfrag-cli-batch-multiple']\n",
      "\n",
      "Low: Predicted labels: ['metfrag-cli-batch-multiple'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bedtools_multiintersectbed', 'Filter1', 'gops_cluster_1']\n",
      "\n",
      "Low: True labels: ['gops_subtract_1', 'gops_basecoverage_1', 'deeptools_compute_matrix', 'tp_replace_in_column']\n",
      "\n",
      "Low: Predicted labels: ['gops_subtract_1', 'gops_basecoverage_1', 'tp_replace_in_column', 'deeptools_compute_matrix'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['thermo_raw_file_converter', 'search_gui', 'peptide_shaker', 'Show beginning1', 'cat1', 'proteomics_moff']\n",
      "\n",
      "Low: True labels: ['tp_replace_in_column', 'sort1', 'lfq_protein_quant', 'tp_sed_tool', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['tp_sed_tool', 'Filter1', 'sort1', 'lfq_protein_quant', 'tp_replace_in_column'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['xcms-find-peaks']\n",
      "\n",
      "Low: True labels: ['camera-annotate-peaks', 'xcms-collect-peaks']\n",
      "\n",
      "Low: Predicted labels: ['camera-annotate-peaks', 'xcms-collect-peaks'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['addName', 'tp_sorted_uniq', 'tp_cat', 'addValue', 'mass_spectrometry_imaging_combine', 'mass_spectrometry_imaging_preprocessing']\n",
      "\n",
      "Low: True labels: ['tp_easyjoin_tool', 'mass_spectrometry_imaging_combine', 'mass_spectrometry_imaging_segmentations', 'mass_spectrometry_imaging_preprocessing', 'tp_sort_header_tool', 'mass_spectrometry_imaging_ion_images', 'mass_spectrometry_imaging_qc', 'mass_spectrometry_imaging_classification', 'mass_spectrometry_imaging_mzplots', 'Cut1', 'collection_column_join']\n",
      "\n",
      "Low: Predicted labels: ['mass_spectrometry_imaging_combine', 'mass_spectrometry_imaging_preprocessing', 'mass_spectrometry_imaging_qc', 'mass_spectrometry_imaging_classification', 'mass_spectrometry_imaging_ion_images', 'mass_spectrometry_imaging_segmentations', 'tp_sort_header_tool', 'mass_spectrometry_imaging_mzplots', 'collection_column_join', 'Cut1', 'tp_easyjoin_tool'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['funannotate_clean', 'funannotate_sort', 'rna_star', 'funannotate_predict', 'interproscan', 'funannotate_annotate']\n",
      "\n",
      "Low: True labels: ['busco', 'jcvi_gff_stats', 'jbrowse', 'aegean_parseval', 'funannotate_compare']\n",
      "\n",
      "Low: Predicted labels: ['busco', 'jbrowse', 'jcvi_gff_stats', 'funannotate_compare', 'aegean_parseval'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_cut_tool', 'tp_tail_tool', 'fastq_dump']\n",
      "\n",
      "Low: True labels: ['predict_pipeline', 'cshl_fastx_trimmer', 'fastq_to_fasta_python', 'rna_star', 'fastq_paired_end_splitter', 'bwameth', 'multiqc', 'rna_starsolo', 'hisat2', 'khmer_abundance_distribution_single', 'fastq_manipulation', 'fastq_groomer', 'cshl_fastq_quality_filter', 'bwa_mem', 'gffcompare', 'snippy', 'sickle', 'fastq_stats', 'kallisto_quant', 'cufflinks', 'megahit', 'trimmomatic', 'fastqc', 'fastq_paired_end_deinterlacer', 'bismark_bowtie2', 'bowtie2', 'cshl_fastq_to_fasta', 'trim_galore', 'gffread', 'sample_seqs', 'hifiasm', 'regex1', 'cutadapt', 'cshl_fastx_clipper', 'minimap2', 'cat1', 'bowtie_wrapper', '__EXTRACT_DATASET__', 'fastq_trimmer', 'CONVERTER_gz_to_uncompressed', 'kraken2', 'fastp', 'spades']\n",
      "\n",
      "Low: Predicted labels: ['multiqc', 'cshl_fastq_quality_filter', 'bowtie2', 'hifiasm', 'CONVERTER_gz_to_uncompressed', 'fastq_paired_end_splitter', 'snippy', 'sickle', 'fastqc', 'bwameth', 'kallisto_quant', 'bismark_bowtie2', 'fastq_to_fasta_python', 'cshl_fastx_clipper', 'fastp', 'predict_pipeline', 'khmer_abundance_distribution_single', 'fastq_manipulation', 'sample_seqs', 'bwa_mem', 'cufflinks', 'kraken2', 'fastq_groomer', 'bowtie_wrapper', 'megahit', 'regex1', 'trimmomatic', 'fastq_trimmer', 'fastq_paired_end_deinterlacer', 'trim_galore', 'gffcompare', 'cshl_fastx_trimmer', '__EXTRACT_DATASET__', 'rna_starsolo', 'cshl_fastq_to_fasta', 'cat1', 'fastq_stats', 'rna_star', 'cutadapt', 'gffread', 'spades', 'minimap2', 'hisat2'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['usearch_derep_full']\n",
      "\n",
      "Low: True labels: ['cluster_otu', 'map_reads_otu', 'vsearch_sorting']\n",
      "\n",
      "Low: Predicted labels: ['vsearch_sorting', 'map_reads_otu', 'cluster_otu'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msgfplus', 'idpqonvert']\n",
      "\n",
      "Low: True labels: ['idpassemble', 'idpquery']\n",
      "\n",
      "Low: Predicted labels: ['IDFileConverter', 'percolator_input_converters'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastp', 'unicycler', 'mob_recon', 'progressivemauve']\n",
      "\n",
      "Low: True labels: ['newick_display', 'circos_aln_to_links', 'xmfa2gff3']\n",
      "\n",
      "Low: Predicted labels: ['newick_display', 'circos_aln_to_links', 'xmfa2gff3'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['FeatureFinderCentroided']\n",
      "\n",
      "Low: True labels: ['QCCalculator', 'FeatureLinkerUnlabeled', 'MapStatistics', 'IDMapper']\n",
      "\n",
      "Low: Predicted labels: ['IDMapper', 'MapStatistics', 'FeatureLinkerUnlabeled', 'QCCalculator'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['dropletutils_read_10x', 'scpred_preprocess_data', 'scpred_predict_labels', 'scpred_get_std_output']\n",
      "\n",
      "Low: True labels: ['ct_combine_tool_outputs']\n",
      "\n",
      "Low: Predicted labels: ['scanpy_compute_graph'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['umi_tools_extract', 'rna_starsolo']\n",
      "\n",
      "Low: True labels: ['multiqc', 'scanpy_read_10x', 'dropletutils', 'bamFilter', 'anndata_import', 'stringtie']\n",
      "\n",
      "Low: Predicted labels: ['multiqc', 'dropletutils', 'bamFilter', 'stringtie', 'scanpy_read_10x', 'anndata_import'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.fasta.remove_desc', 'edu.tamu.cpt.gff3.require_phage_start']\n",
      "\n",
      "Low: True labels: ['edu.tamu.cpt.genbank.shinefind', 'edu.tamu.cpt.gff3.require_sd']\n",
      "\n",
      "Low: Predicted labels: ['edu.tamu.cpt.genbank.shinefind', 'edu.tamu.cpt.gff3.require_sd'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['minfi_read450k', 'minfi_ppfun', 'minfi_dropsnp']\n",
      "\n",
      "Low: True labels: ['minfi_getanno', 'minfi_dmr', 'minfi_getM', 'minfi_getbeta', 'minfi_methcpg', 'minfi_dmp']\n",
      "\n",
      "Low: Predicted labels: ['minfi_dropsnp', 'vg_convert', 'muscle', 'Kneeplot', 'sansa_annotate', 'fromHicupToJuicebox'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'velvetoptimiser', 'glimmer_not-knowlegde-based']\n",
      "\n",
      "Low: True labels: ['ncbi_blastn_wrapper', 'EMBOSS: transeq101', 'edu.tamu.cpt2.util.glimmer3_to_gff3', 'ncbi_blastx_wrapper']\n",
      "\n",
      "Low: Predicted labels: ['edu.tamu.cpt2.util.glimmer3_to_gff3', 'ncbi_blastx_wrapper', 'ncbi_blastn_wrapper', 'EMBOSS: transeq101'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['rpExtractSink', 'retropath2', 'rp2paths']\n",
      "\n",
      "Low: True labels: ['rpCompletion']\n",
      "\n",
      "Low: Predicted labels: ['XTandemAdapter'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['schicexplorer_schicqualitycontrol', 'schicexplorer_schicnormalize', 'schicexplorer_schicclustersvl']\n",
      "\n",
      "Low: True labels: ['schicexplorer_schicconsensusmatrices', 'schicexplorer_schicplotclusterprofiles']\n",
      "\n",
      "Low: Predicted labels: ['schicexplorer_schicplotclusterprofiles', 'schicexplorer_schicclustersvl'], Precision: 0.5\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_paired_end_interlacer', 'spades', 'cisarunner']\n",
      "\n",
      "Low: True labels: ['prokka', 'cshl_fasta_formatter', 'check_contigs']\n",
      "\n",
      "Low: Predicted labels: ['prokka', 'cshl_fasta_formatter', 'check_contigs'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['retrieve_scxa', 'scanpy_read_10x', 'scanpy_filter_cells', 'scanpy_filter_genes', 'scanpy_normalise_data', 'scanpy_regress_variable', 'scanpy_scale_data', 'scanpy_find_variable_genes', 'scanpy_run_pca', 'scanpy_compute_graph', 'scanpy_find_cluster', 'scanpy_run_umap', 'run_sccaf', 'sccaf_asses']\n",
      "\n",
      "Low: True labels: ['sccaf_asses_merger']\n",
      "\n",
      "Low: Predicted labels: ['sccaf_asses'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['TargetedFileConverter', 'OpenSwathWorkflow', 'pyprophet_merge']\n",
      "\n",
      "Low: True labels: ['pyprophet_score']\n",
      "\n",
      "Low: Predicted labels: ['pyprophet_score'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['deeptools_bam_coverage', 'bigwig_to_bedgraph']\n",
      "\n",
      "Low: True labels: ['bedtools_unionbedgraph', 'bedtools_makewindowsbed', 'bedtools_intersectbed', 'bedtools_mergebedgraph']\n",
      "\n",
      "Low: Predicted labels: ['bedtools_intersectbed', 'bedtools_mergebedgraph', 'bedtools_unionbedgraph', 'bedtools_makewindowsbed'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['qiime_validate_mapping_file', 'qiime_split_libraries', 'qiime_pick_otus', 'qiime_pick_rep_set', 'qiime_assign_taxonomy', 'qiime_make_otu_table']\n",
      "\n",
      "Low: True labels: ['qiime_summarize_taxa_through_plots', 'qiime_make_otu_heatmap']\n",
      "\n",
      "Low: Predicted labels: ['qiime_make_otu_heatmap', 'qiime_summarize_taxa_through_plots'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'mothur_make_contigs']\n",
      "\n",
      "Low: True labels: ['mothur_get_seqs', 'mothur_screen_seqs', 'hmmer_hmmsearch', 'mothur_unique_seqs', 'mothur_summary_qual', 'mothur_merge_files', 'mothur_make_group', 'mothur_align_seqs', 'seq_filter_by_id', 'fastqc', 'mothur_sub_sample', 'mothur_trim_seqs', 'mothur_summary_seqs', 'mycrobiota_subsample_add_discarded_samples', 'fasta_filter_by_length', 'mothur_count_seqs']\n",
      "\n",
      "Low: Predicted labels: ['seq_filter_by_id', 'mothur_merge_files', 'fastqc', 'hmmer_hmmsearch', 'mothur_align_seqs', 'mothur_summary_seqs', 'mothur_count_seqs', 'mothur_unique_seqs', 'mothur_trim_seqs', 'mothur_make_group', 'mothur_get_seqs', 'mothur_screen_seqs', 'fasta_filter_by_length', 'mothur_summary_qual', 'mycrobiota_subsample_add_discarded_samples', 'mothur_sub_sample'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_common', 'cp_identify_primary_objects', 'cp_convert_objects_to_image']\n",
      "\n",
      "Low: True labels: ['cp_display_data_on_image', 'cp_measure_object_size_shape', 'cp_image_math', 'cp_gray_to_color']\n",
      "\n",
      "Low: Predicted labels: ['cp_display_data_on_image', 'cp_measure_object_size_shape', 'cp_gray_to_color', 'cp_image_math'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['methtools_calling', 'cshl_sort_header', 'cshl_uniq_tool', 'Filter1', 'Add_a_column1', 'Cut1', 'bedtools_intersectbed_bam']\n",
      "\n",
      "Low: True labels: ['methtools_plot', 'sort1', 'Cut1', 'methtools_dmr', 'wc_gnu']\n",
      "\n",
      "Low: Predicted labels: ['methtools_plot', 'sort1', 'methtools_dmr', 'Cut1', 'wc_gnu'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['make_window_bed', 'window_coverage']\n",
      "\n",
      "Low: True labels: ['find_qPCR_regions']\n",
      "\n",
      "Low: Predicted labels: ['find_qPCR_regions'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['make_window_bed']\n",
      "\n",
      "Low: True labels: ['window_coverage']\n",
      "\n",
      "Low: Predicted labels: ['hisat2'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['PeakPickerHiRes', 'FileFilter', 'xcms-find-peaks', 'xcms-collect-peaks', 'xcms-group-peaks', 'xcms-blankfilter', 'xcms-dilutionfilter', 'camera-annotate-peaks', 'camera-group-fwhm', 'camera-find-adducts', 'camera-find-isotopes', 'msms2metfrag-multiple', 'metfrag-cli-batch-multiple']\n",
      "\n",
      "Low: True labels: ['passatutto']\n",
      "\n",
      "Low: Predicted labels: ['passatutto'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_common', 'cp_identify_primary_objects', 'cp_convert_objects_to_image', 'cp_gray_to_color', 'cp_save_images', 'cp_measure_granularity', 'cp_measure_texture', 'cp_measure_object_intensity']\n",
      "\n",
      "Low: True labels: ['cp_measure_object_size_shape', 'cp_track_objects']\n",
      "\n",
      "Low: Predicted labels: ['cp_track_objects', 'cp_measure_object_size_shape'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['humann', 'humann_renorm_table']\n",
      "\n",
      "Low: True labels: ['combine_metaphlan2_humann2', 'humann_unpack_pathways', 'humann2_unpack_pathways']\n",
      "\n",
      "Low: Predicted labels: ['humann_unpack_pathways', 'humann2_unpack_pathways', 'combine_metaphlan2_humann2'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt.fasta.remove_desc']\n",
      "\n",
      "Low: True labels: ['edu.tamu.cpt.tRNAscan', 'MGA', 'edu.tamu.cpt2.util.mga_to_gff3', 'edu.tamu.cpt.fasta.lipory', 'ncbi_blastn_wrapper_new', 'edu.tamu.cpt.gff3.export_seq', 'jbrowse-iuc-cpt', 'edu.tamu.cpt.genbank.shinefind', 'aragorn_trna', 'get_orfs_or_cdss', 'glimmer_not-knowlegde-based', 'edu.tamu.cpt.gff3.require_sd', 'edu.tamu.cpt.util.wigToBigWig', 'edu.tamu.cpt.gff3.require_phage_start', 'edu.tamu.cpt2.util.glimmer3_to_gff3', 'TransTermHP']\n",
      "\n",
      "Low: Predicted labels: ['staramr_search', 'addValue', 'edu.tamu.cpt.util.wigToBigWig', 'quast', 'mergeCols1', 'mlst', 'collapse_dataset', 'infernal_cmscan', 'bandage_image', 'abricate', 'infernal_cmsearch', 'edu.tamu.cpt2.util.mga_to_gff3', 'bio_hansel', 'bowtie2', 'Extract genomic DNA 1', 'qualimap_bamqc'], Precision: 0.125\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['ebi_search_rest_results', 'regex1', 'tp_cut_tool', 'lftp']\n",
      "\n",
      "Low: True labels: ['graphicsmagick_image_convert', 'ip_convertimage', 'unzip', 'barchart_gnuplot', 'ip_scale_image', 'ip_imageinfo', 'maxquant', 'msnbase_readmsdata', 'scatterplot_rpy', 'collapse_dataset']\n",
      "\n",
      "Low: Predicted labels: ['ip_scale_image', 'collapse_dataset', 'unzip', 'barchart_gnuplot', 'graphicsmagick_image_convert', 'ip_convertimage', 'ip_imageinfo', 'msnbase_readmsdata', 'scatterplot_rpy', 'maxquant'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bowtie2', 'bamFilter', 'getTn5ExtendedCoverage']\n",
      "\n",
      "Low: True labels: ['splitbedGraphFromBed']\n",
      "\n",
      "Low: Predicted labels: ['splitbedGraphFromBed'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_common', 'cp_identify_primary_objects', 'cp_convert_objects_to_image', 'cp_image_math', 'cp_measure_granularity', 'cp_measure_texture', 'cp_measure_object_intensity', 'cp_measure_object_size_shape', 'cp_relate_objects', 'cp_measure_image_quality']\n",
      "\n",
      "Low: True labels: ['cp_measure_image_area_occupied']\n",
      "\n",
      "Low: Predicted labels: ['cp_measure_image_area_occupied'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Add_a_column1', 'Cut1', 'fgsea']\n",
      "\n",
      "Low: True labels: ['tp_find_and_replace', 'plotdexseq', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['tp_find_and_replace', 'Filter1', 'plotdexseq'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['unzip', 'ip_histogram_equalization', 'ip_convertimage']\n",
      "\n",
      "Low: True labels: ['ip_scale_image', 'ip_filter_standard', 'ip_overlay_segmentation']\n",
      "\n",
      "Low: Predicted labels: ['ip_filter_standard', 'ip_scale_image', 'ip_overlay_segmentation'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.gff3.reformat', 'edu.tamu.cpt.gff3.export_seq']\n",
      "\n",
      "Low: True labels: ['edu.tamu.cpt.fasta.seq_trans', 'edu.tamu.cpt.fasta.start_codons', 'edu.tamu.cpt.fasta.stop_codons']\n",
      "\n",
      "Low: Predicted labels: ['edu.tamu.cpt.fasta.start_codons', 'edu.tamu.cpt.fasta.stop_codons', 'edu.tamu.cpt.fasta.seq_trans'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['mspurity_puritya', 'abims_xcms_xcmsSet']\n",
      "\n",
      "Low: True labels: ['abims_xcms_refine', 'xcms_merge', 'abims_xcms_group', 'mspurity_purityx', 'xcms_plot_chromatogram', 'query_tabular', 'quality_metrics', 'xcms_export_samplemetadata']\n",
      "\n",
      "Low: Predicted labels: ['xcms_plot_chromatogram', 'quality_metrics', 'xcms_merge', 'abims_xcms_group', 'query_tabular', 'xcms_export_samplemetadata', 'mspurity_purityx', 'abims_xcms_refine'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['get_pdb']\n",
      "\n",
      "Low: True labels: ['bio3d_pca', 'Grep1', 'prepare_receptor', 'bio3d_rmsf', 'CONVERTER_Pdb_to_Gro_0', 'docking', 'bio3d_rmsd', 'tp_grep_tool', 'gmx_setup', 'fpocket']\n",
      "\n",
      "Low: Predicted labels: ['Grep1', 'fpocket', 'prepare_receptor', 'tp_grep_tool', 'gmx_setup', 'docking', 'CONVERTER_Pdb_to_Gro_0', 'bio3d_rmsf', 'bio3d_rmsd', 'bio3d_pca'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['chira_merge']\n",
      "\n",
      "Low: True labels: ['chira_quantify']\n",
      "\n",
      "Low: Predicted labels: ['chira_quantify'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['hca_matrix_downloader', 'seurat_read10x']\n",
      "\n",
      "Low: True labels: ['seurat_filter_cells', 'seurat_normalise_data', 'seurat_create_seurat_object', 'seurat_find_variable_genes']\n",
      "\n",
      "Low: Predicted labels: ['seurat_filter_cells', 'seurat_normalise_data', 'seurat_create_seurat_object', 'seurat_find_variable_genes'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_common', 'cp_identify_primary_objects', 'cp_convert_objects_to_image', 'cp_gray_to_color', 'cp_save_images', 'cp_measure_granularity', 'cp_measure_texture', 'cp_measure_object_intensity', 'cp_measure_object_size_shape']\n",
      "\n",
      "Low: True labels: ['cp_display_data_on_image', 'cp_relate_objects', 'cp_measure_object_intensity', 'cp_export_to_spreadsheet', 'cp_measure_image_quality']\n",
      "\n",
      "Low: Predicted labels: ['cp_display_data_on_image', 'cp_export_to_spreadsheet', 'cp_measure_image_quality', 'cp_relate_objects', 'cp_measure_object_intensity'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.fasta.remove_desc', 'get_orfs_or_cdss', 'edu.tamu.cpt.gff3.cdsParents', 'edu.tamu.cpt.gff3.require_phage_start', 'edu.tamu.cpt.gff3.require_sd']\n",
      "\n",
      "Low: True labels: ['edu.tamu.cpt.gff3.lipoP_to_gff3', 'edu.tamu.cpt.fasta.lipory', 'edu.tamu.cpt.gff3.export_seq', 'jbrowse-iuc-cpt', 'gff3.rebase']\n",
      "\n",
      "Low: Predicted labels: ['jbrowse-iuc-cpt', 'edu.tamu.cpt.gff3.export_seq', 'gff3.rebase', 'edu.tamu.cpt.fasta.lipory', 'edu.tamu.cpt.gff3.lipoP_to_gff3'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['ctb_remIons']\n",
      "\n",
      "Low: True labels: ['ctb_chemfp_mol2fps', 'ctb_compound_convert', 'ctb_filter']\n",
      "\n",
      "Low: Predicted labels: ['ctb_chemfp_mol2fps', 'ctb_filter', 'ctb_compound_convert'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['retrieve_scxa', 'scanpy_read_10x', 'scanpy_filter_cells', 'scanpy_filter_genes', 'scanpy_normalise_data', 'scanpy_find_variable_genes', 'scanpy_scale_data', 'scanpy_run_pca', 'scanpy_compute_graph', 'scanpy_find_cluster', 'sceasy_convert', 'scmap_preprocess_sce']\n",
      "\n",
      "Low: True labels: ['scmap_scmap_cell', 'scmap_scmap_cluster', 'scmap_select_features']\n",
      "\n",
      "Low: Predicted labels: ['scmap_select_features', 'scmap_scmap_cluster', 'scmap_scmap_cell'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_measure_granularity', 'cp_measure_texture', 'cp_measure_object_intensity', 'cp_measure_object_size_shape', 'cp_relate_objects', 'cp_measure_image_quality', 'cp_measure_image_area_occupied']\n",
      "\n",
      "Low: True labels: ['cp_measure_image_intensity', 'cp_export_to_spreadsheet']\n",
      "\n",
      "Low: Predicted labels: ['cp_export_to_spreadsheet', 'cp_measure_image_intensity'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_replace_in_line', 'clustalw']\n",
      "\n",
      "Low: True labels: ['hmmer_hmmbuild', 'rbc_rnacode', 'iqtree', 'rgweblogo3', 'fasttree', 'raxml', 'phyml', 'align_back_trans', 'rnaz', 'qiime_make_phylogeny', 'gd_raxml', 'wc_gnu']\n",
      "\n",
      "Low: Predicted labels: ['qiime_make_phylogeny', 'raxml', 'hmmer_hmmbuild', 'rnaz', 'gd_raxml', 'wc_gnu', 'iqtree', 'fasttree', 'align_back_trans', 'rbc_rnacode', 'rgweblogo3', 'phyml'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'kraken2', 'kraken_taxonomy_report']\n",
      "\n",
      "Low: True labels: ['vegan_fisher_alpha', 'vegan_diversity', 'vegan_rarefaction', 'Fetch Taxonomic Ranks']\n",
      "\n",
      "Low: Predicted labels: ['Fetch Taxonomic Ranks', 'vegan_diversity', 'vegan_rarefaction', 'vegan_fisher_alpha'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['gbk_to_orf', 'smf_utils_fix-fasta-headers']\n",
      "\n",
      "Low: True labels: ['smf_utils_find-boxes', 'fasttree', 'raxml', 'blast_reciprocal_best_hits', 'smf_utils_extract-boxed-sequences']\n",
      "\n",
      "Low: Predicted labels: ['smf_utils_extract-boxed-sequences', 'fasttree', 'raxml', 'smf_utils_find-boxes', 'blast_reciprocal_best_hits'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['rbc_splitfasta', 'hhsearch', 'ffindex_dbkit_create']\n",
      "\n",
      "Low: True labels: ['ffindex_dbkit_merge', 'spring_model_all']\n",
      "\n",
      "Low: Predicted labels: ['spring_model_all', 'ffindex_dbkit_merge'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Remove beginning1', 'split_file_to_collection', '__RELABEL_FROM_FILE__', 'Cut1', 'param_value_from_file', 'kc-align']\n",
      "\n",
      "Low: True labels: ['sarscov2formatter', 'sarscov2summary']\n",
      "\n",
      "Low: Predicted labels: ['sarscov2summary', 'sarscov2formatter'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cp_identify_primary_objects']\n",
      "\n",
      "Low: True labels: ['cp_display_data_on_image', 'cp_measure_object_size_shape', 'cp_overlay_outlines', 'cp_convert_objects_to_image', 'cp_cellprofiler']\n",
      "\n",
      "Low: Predicted labels: ['cp_cellprofiler', 'cp_display_data_on_image', 'cp_measure_object_size_shape', 'cp_overlay_outlines', 'cp_convert_objects_to_image'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['gops_join_1', 'CONVERTER_interval_to_bedstrict_0']\n",
      "\n",
      "Low: True labels: ['circos_wiggle_to_stacked', 'Grouping1', 'datamash_ops', 'tp_tail_tool', 'gops_intersect_1', 'CONVERTER_bed_to_fli_0', 'tp_replace_in_column', 'gops_join_1', 'circos_wiggle_to_scatter', 'join1', 'gops_concat_1']\n",
      "\n",
      "Low: Predicted labels: ['circos_wiggle_to_stacked', 'gops_intersect_1', 'CONVERTER_bed_to_fli_0', 'circos_wiggle_to_scatter', 'join1', 'gops_concat_1', 'gops_join_1', 'Grouping1', 'tp_replace_in_column', 'tp_tail_tool', 'datamash_ops'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['sed']\n",
      "\n",
      "Low: True labels: ['gtf_filter_by_attribute_values_list_b', 'awkscript']\n",
      "\n",
      "Low: Predicted labels: ['awkscript', 'gtf_filter_by_attribute_values_list_b'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Cut1', 'stacks_denovomap']\n",
      "\n",
      "Low: True labels: ['stacks_populations', 'stacks_genotypes']\n",
      "\n",
      "Low: Predicted labels: ['stacks_genotypes', 'stacks_populations'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['macs2_callpeak', 'macs2_bdgcmp']\n",
      "\n",
      "Low: True labels: ['macs2_bdgbroadcall', 'wig_to_bigWig']\n",
      "\n",
      "Low: Predicted labels: ['macs2_bdgbroadcall', 'wig_to_bigWig'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['keras_model_builder']\n",
      "\n",
      "Low: True labels: ['sklearn_model_validation', 'sklearn_estimator_attributes', 'keras_train_and_eval']\n",
      "\n",
      "Low: Predicted labels: ['keras_train_and_eval', 'sklearn_model_validation', 'sklearn_estimator_attributes'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['EMBOSS: transeq101', 'filter_by_fasta_ids', 'fasplit', 'ncbi_blastp_wrapper', 'tp_cat', 'blast_parser', 'hcluster_sg']\n",
      "\n",
      "Low: True labels: ['hcluster_sg_parser']\n",
      "\n",
      "Low: Predicted labels: ['hcluster_sg_parser'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['CONVERTER_gff_to_interval_index_0', 'CONVERTER_bed_gff_or_vcf_to_bigwig_0']\n",
      "\n",
      "Low: True labels: ['CONVERTER_bed_to_bgzip_0', 'gops_intersect_1', 'featurecounts', 'Add_a_column1', 'gffcompare', 'deeptools_compute_matrix', 'stringtie_merge', 'CONVERTER_gff_to_interval_index_0', 'gops_concat_1', 'cat1', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['CONVERTER_gff_to_interval_index_0', 'CONVERTER_bed_to_bgzip_0', 'gops_intersect_1', 'deeptools_compute_matrix', 'stringtie_merge', 'featurecounts', 'Add_a_column1', 'gffcompare', 'Filter1', 'gops_concat_1', 'cat1'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['ip_threshold']\n",
      "\n",
      "Low: True labels: ['ip_convertimage', 'ip_count_objects', 'ip_binary_to_labelimage', 'ip_2d_split_binaryimage_by_watershed', 'imagej2_watershed_binary']\n",
      "\n",
      "Low: Predicted labels: ['ip_count_objects', 'ip_convertimage', 'ip_2d_split_binaryimage_by_watershed', 'ip_binary_to_labelimage', 'imagej2_watershed_binary'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['proteomiqon_peptidedb', 'proteomiqon_psmstatistics']\n",
      "\n",
      "Low: True labels: ['proteomiqon_proteininference', 'proteomiqon_psmbasedquantification']\n",
      "\n",
      "Low: Predicted labels: ['proteomiqon_psmbasedquantification', 'proteomiqon_proteininference'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['cutadapt', 'hicup_hicup']\n",
      "\n",
      "Low: True labels: ['fromHicupToJuicebox']\n",
      "\n",
      "Low: Predicted labels: ['tophat2'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_sorted_uniq', 'join1', 'tp_cut_tool', 'sklearn_svm_classifier']\n",
      "\n",
      "Low: True labels: ['tp_tail_tool', 'plotly_ml_performance_plots', 'sklearn_clf_metrics']\n",
      "\n",
      "Low: Predicted labels: ['tp_tail_tool', 'plotly_ml_performance_plots', 'sklearn_clf_metrics'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trim_galore', 'iuc_pear']\n",
      "\n",
      "Low: True labels: ['cshl_fastq_quality_filter', 'fastqc', 'cutadapt', 'cat1', 'fastqStat', 'fastq_filter']\n",
      "\n",
      "Low: Predicted labels: ['fastqc', 'cat1', 'fastq_filter', 'cutadapt', 'cshl_fastq_quality_filter', 'fastqStat'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['preproc', 'viennarna_rnafold', 'structure_to_gspan', 'nspdk_sparse']\n",
      "\n",
      "Low: True labels: ['tp_cat', 'NSPDK_candidateClust']\n",
      "\n",
      "Low: Predicted labels: ['tp_cat', 'NSPDK_candidateClust'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'iuc_pear', 'cat1', 'uclust2otutable', 'biom_convert']\n",
      "\n",
      "Low: True labels: ['mothur_make_shared', 'biom_add_metadata']\n",
      "\n",
      "Low: Predicted labels: ['mothur_make_shared', 'biom_add_metadata'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['OpenSwathWorkflow', 'pyprophet_merge', 'pyprophet_score', 'pyprophet_peptide']\n",
      "\n",
      "Low: True labels: ['pyprophet_protein']\n",
      "\n",
      "Low: Predicted labels: ['pyprophet_protein'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['ctb_online_data_fetch', 'ctb_change_title', 'ctb_compound_convert', 'ctb_confab', 'ctb_ob_genProp']\n",
      "\n",
      "Low: True labels: ['ctb_filter', 'ctb_spectrophore_search']\n",
      "\n",
      "Low: Predicted labels: ['ctb_filter', 'ctb_spectrophore_search'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['EMBOSS: transeq101', 'ncbi_makeblastdb', 'ncbi_blastp_wrapper', 'blast_parser']\n",
      "\n",
      "Low: True labels: ['hcluster_sg']\n",
      "\n",
      "Low: Predicted labels: ['hcluster_sg'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msconvert', 'proteomiqon_mzmltomzlite', 'proteomiqon_peptidespectrummatching']\n",
      "\n",
      "Low: True labels: ['proteomiqon_psmstatistics']\n",
      "\n",
      "Low: Predicted labels: ['proteomiqon_psmstatistics'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['PeakPickerHiRes', 'FileFilter', 'xcms-find-peaks', 'xcms-collect-peaks', 'xcms-group-peaks', 'xcms-blankfilter', 'xcms-dilutionfilter', 'camera-annotate-peaks', 'camera-group-fwhm', 'camera-find-adducts', 'camera-find-isotopes']\n",
      "\n",
      "Low: True labels: ['map-msms2camera', 'camera-prepareoutput', 'msms2metfrag-multiple', 'camera-find-adducts']\n",
      "\n",
      "Low: Predicted labels: ['msms2metfrag-multiple', 'camera-prepareoutput', 'map-msms2camera', 'camera-find-adducts'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_groomer', 'fastq_filter', 'fastq_manipulation']\n",
      "\n",
      "Low: True labels: ['fastq_to_fasta_python', 'fastq_groomer', 'trimmomatic', 'fastqc', 'rbc_mirdeep2_mapper', 'cd_hit_dup']\n",
      "\n",
      "Low: Predicted labels: ['fastq_to_fasta_python', 'fastqc', 'cd_hit_dup', 'trimmomatic', 'fastq_groomer', 'rbc_mirdeep2_mapper'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['ncbi_acc_download', 'music_construct_eset', 'raceid_inspectclusters', 'scater_create_qcmetric_ready_sce']\n",
      "\n",
      "Low: True labels: ['scater_filter', 'scater_plot_pca', 'scater_normalize', 'scater_plot_dist_scatter', 'scater_plot_exprs_freq']\n",
      "\n",
      "Low: Predicted labels: ['scater_plot_dist_scatter', 'scater_filter', 'scater_plot_pca', 'scater_plot_exprs_freq', 'scater_normalize'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['encyclopedia_searchtolib']\n",
      "\n",
      "Low: True labels: ['encyclopedia_quantify']\n",
      "\n",
      "Low: Predicted labels: ['encyclopedia_quantify'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['make_families', 'correct_barcodes', 'align_families', 'dunovo']\n",
      "\n",
      "Low: True labels: ['fasta2fastq', 'sequence_content_trimmer']\n",
      "\n",
      "Low: Predicted labels: ['sequence_content_trimmer', 'fasta2fastq'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msnbase_readmsdata', 'abims_xcms_xcmsSet', 'xcms_merge', 'abims_xcms_group', 'abims_xcms_fillPeaks', 'abims_CAMERA_annotateDiffreport']\n",
      "\n",
      "Low: True labels: ['Univariate', 'Batch_correction', 'generic_filter', 'w4mclassfilter', 'lcmsmatching', 'lipidmaps', 'wsdl_hmdb', 'quality_metrics', 'Transformation', 'abims_CAMERA_combinexsAnnos']\n",
      "\n",
      "Low: Predicted labels: ['Univariate', 'generic_filter', 'quality_metrics', 'w4mclassfilter', 'Batch_correction', 'wsdl_hmdb', 'Transformation', 'abims_CAMERA_combinexsAnnos', 'lcmsmatching', 'lipidmaps'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['gops_coverage_1']\n",
      "\n",
      "Low: True labels: ['cshl_grep_tool', 'sort1', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['sort1', 'Filter1', 'cshl_grep_tool'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msnbase_readmsdata', 'abims_xcms_xcmsSet', 'xcms_merge', 'abims_xcms_group', 'abims_xcms_fillPeaks']\n",
      "\n",
      "Low: True labels: ['mspurity_frag4feature', 'abims_xcms_summary', 'Batch_correction', 'generic_filter', 'abims_CAMERA_annotateDiffreport', 'quality_metrics']\n",
      "\n",
      "Low: Predicted labels: ['generic_filter', 'mspurity_frag4feature', 'quality_metrics', 'abims_xcms_summary', 'abims_CAMERA_annotateDiffreport', 'Batch_correction'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trim_galore', 'make_families']\n",
      "\n",
      "Low: True labels: ['correct_barcodes', 'Cut1']\n",
      "\n",
      "Low: Predicted labels: ['Cut1', 'correct_barcodes'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['monocle3_reduceDim']\n",
      "\n",
      "Low: True labels: ['monocle3_partition']\n",
      "\n",
      "Low: Predicted labels: ['monocle3_partition'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['shovill', 'cat_contigs']\n",
      "\n",
      "Low: True labels: ['cat_add_names', 'prokka']\n",
      "\n",
      "Low: Predicted labels: ['prokka', 'cat_add_names'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Grep1', 'computeMatrix']\n",
      "\n",
      "Low: True labels: ['dt_profiler', 'heatmapper_deepTools']\n",
      "\n",
      "Low: Predicted labels: ['dt_profiler', 'heatmapper_deepTools'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['msnbase_readmsdata', 'abims_xcms_xcmsSet', 'xcms_merge', 'abims_xcms_group', 'mspurity_frag4feature']\n",
      "\n",
      "Low: True labels: ['mspurity_filterfragspectra']\n",
      "\n",
      "Low: Predicted labels: ['mspurity_filterfragspectra'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['qiime_validate_mapping_file']\n",
      "\n",
      "Low: True labels: ['qiime_split_libraries', 'qiime_summarize_taxa_through_plots', 'qiime_make_otu_table', 'qiime_core_diversity', 'qiime_make_otu_heatmap', 'qiime_split_libraries_fastq']\n",
      "\n",
      "Low: Predicted labels: ['qiime_pick_rep_set', 'qiime_pick_otus', 'obi_annotate', 'obi_ngsfilter', 'obi_clean', 'hamronize_tool'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['OpenSwathDecoyGenerator', 'TargetedFileConverter', 'pyprophet_merge', 'pyprophet_score', 'pyprophet_peptide', 'pyprophet_protein']\n",
      "\n",
      "Low: True labels: ['pyprophet_export']\n",
      "\n",
      "Low: Predicted labels: ['pyprophet_export'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Extractor', 'GeneModelMapper']\n",
      "\n",
      "Low: True labels: ['GeMoMa_Annotation_Filter']\n",
      "\n",
      "Low: Predicted labels: ['GeMoMa_Annotation_Filter'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['fastq_groomer', 'trimmomatic', 'velvetoptimiser']\n",
      "\n",
      "Low: True labels: ['staramr_search', 'fasta-stats', 'abricate', 'filter_by_fasta_ids', 'PlasFlow', 'mothur_classify_seqs', 'exonerate', 'EMBOSS: getorf42', 'mlst', 'glimmer_not-knowlegde-based', 'bowtie2', 'quast', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['PlasFlow', 'glimmer_not-knowlegde-based', 'quast', 'staramr_search', 'EMBOSS: getorf42', 'mlst', 'abricate', 'mothur_classify_seqs', 'Filter1', 'exonerate', 'fasta-stats', 'bowtie2', 'filter_by_fasta_ids'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trimmomatic', 'iuc_pear', 'cat1', 'uclust2otutable', 'biom_convert', 'biom_add_metadata']\n",
      "\n",
      "Low: True labels: ['phyloseq_richness', 'phyloseq_taxonomy', 'phyloseq_DESeq2', 'phyloseq_abundance']\n",
      "\n",
      "Low: Predicted labels: ['phyloseq_DESeq2', 'phyloseq_abundance', 'phyloseq_richness', 'phyloseq_taxonomy'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bedtools_bamtobed', 'flair_correct']\n",
      "\n",
      "Low: True labels: ['flair_collapse']\n",
      "\n",
      "Low: Predicted labels: ['flair_collapse'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.fasta.remove_desc', 'get_orfs_or_cdss', 'edu.tamu.cpt.gff3.cdsParents', 'edu.tamu.cpt.gff3.require_phage_start', 'edu.tamu.cpt.gff3.require_sd', 'edu.tamu.cpt.gff3.export_seq', 'edu.tamu.cpt.fasta.seq_trans', 'ncbi_blastp_wrapper_new']\n",
      "\n",
      "Low: True labels: ['jbrowse-iuc-cpt', 'edu.tamu.cpt2.phage.intron_detection', 'edu.tamu.cpt.blastp_to_gff3']\n",
      "\n",
      "Low: Predicted labels: ['jbrowse-iuc-cpt', 'edu.tamu.cpt2.phage.intron_detection', 'edu.tamu.cpt.blastp_to_gff3'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['map_reads', 'call_sites', 'filter_regions', 'call_consensus']\n",
      "\n",
      "Low: True labels: ['snp_matrix', 'merge_vcfs']\n",
      "\n",
      "Low: Predicted labels: ['snp_matrix', 'merge_vcfs'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['obi_illumina_pairend', 'fastq_groomer', 'obi_grep']\n",
      "\n",
      "Low: True labels: ['obi_uniq', 'obi_ngsfilter', 'qiime_pick_otus', 'obi_clean']\n",
      "\n",
      "Low: Predicted labels: ['obi_uniq', 'obi_ngsfilter', 'qiime_pick_otus', 'obi_clean'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bowtie2', 'hicexplorer_hicbuildmatrix', 'hicexplorer_chicviewpointbackgroundmodel', 'hicexplorer_chicviewpoint']\n",
      "\n",
      "Low: True labels: ['hicexplorer_chicaggregatestatistic', 'hicexplorer_chicsignificantinteractions']\n",
      "\n",
      "Low: Predicted labels: ['hicexplorer_chicaggregatestatistic', 'hicexplorer_chicsignificantinteractions'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['trim_galore', 'make_families', 'correct_barcodes']\n",
      "\n",
      "Low: True labels: ['align_families', 'Cut1']\n",
      "\n",
      "Low: Predicted labels: ['Cut1', 'correct_barcodes'], Precision: 0.5\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['Extractor', 'ncbi_tblastn_wrapper']\n",
      "\n",
      "Low: True labels: ['datamash_ops', 'GeneModelMapper', 'Cut1', 'Filter1']\n",
      "\n",
      "Low: Predicted labels: ['datamash_ops', 'Filter1', 'Cut1', 'GeneModelMapper'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['edu.tamu.cpt2.gff3.splitGff', 'edu.tamu.cpt.gff3.reformat', 'edu.tamu.cpt.gff3.export_seq', 'edu.tamu.cpt.fasta.seq_trans', 'TMHMM_to_gff3']\n",
      "\n",
      "Low: True labels: ['gff3.rebase', 'edu.tamu.cpt.wig.rebase']\n",
      "\n",
      "Low: Predicted labels: ['gff3.rebase', 'edu.tamu.cpt.wig.rebase'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['keras_model_config', 'keras_model_builder', 'keras_train_and_eval']\n",
      "\n",
      "Low: True labels: ['model_prediction', 'ml_visualization_ex']\n",
      "\n",
      "Low: Predicted labels: ['ml_visualization_ex', 'model_prediction'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['TrimPrimer', 'Flash', 'Btrim64', 'ExtractFASTAfromFASTQ', 'ChooseTag', 'TrimByN', 'TrimByLength']\n",
      "\n",
      "Low: True labels: ['uchime']\n",
      "\n",
      "Low: Predicted labels: ['Count1'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['TrimPrimer', 'Flash', 'Btrim64', 'ExtractFASTAfromFASTQ', 'ChooseTag', 'TrimByN', 'TrimByLength', 'uchime']\n",
      "\n",
      "Low: True labels: ['map_reads_otu', 'pynast', 'uparse']\n",
      "\n",
      "Low: Predicted labels: ['pynast', 'map_reads_otu', 'uparse'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['locarna_multiple', 'selectsequencesfrommsa']\n",
      "\n",
      "Low: True labels: ['rbc_rnacode', 'rnaz', 'viennarna_rnaalifold']\n",
      "\n",
      "Low: Predicted labels: ['rnaz', 'rbc_rnacode', 'viennarna_rnaalifold'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['FeatureFinderMetabo', 'MetaboliteAdductDecharger', 'AccurateMassSearch', 'HighResPrecursorMassCorrector']\n",
      "\n",
      "Low: True labels: ['FileFilter', 'MSGFPlusAdapter', 'AssayGeneratorMetabo', 'XTandemAdapter']\n",
      "\n",
      "Low: Predicted labels: ['AssayGeneratorMetabo', 'XTandemAdapter', 'MSGFPlusAdapter', 'FileFilter'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_sed_tool', 'csv_to_tabular']\n",
      "\n",
      "Low: True labels: ['tp_easyjoin_tool', 'deseq2', 'Grouping1', 'datamash_ops', 'datamash_transpose', 'sklearn_numeric_clustering', 'histogram_rpy', 'regexColumn1', 'Remove beginning1', 'tp_sort_header_tool', 'tp_cat', 'charts', 'limma_voom', 'edger', 'sklearn_generalized_linear', 'ggplot2_point', 'tp_awk_tool', 'tp_cut_tool', 'tp_sed_tool', 'Cut1', 'Filter1', 'Count1', 'annotatemyids']\n",
      "\n",
      "Low: Predicted labels: ['sklearn_numeric_clustering', 'sklearn_generalized_linear', 'limma_voom', 'tp_awk_tool', 'edger', 'tp_cut_tool', 'regexColumn1', 'annotatemyids', 'ggplot2_point', 'tp_cat', 'deseq2', 'datamash_transpose', 'histogram_rpy', 'tp_sed_tool', 'Grouping1', 'Remove beginning1', 'charts', 'Filter1', 'Cut1', 'tp_easyjoin_tool', 'tp_sort_header_tool', 'datamash_ops', 'Count1'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['OpenSwathAssayGenerator']\n",
      "\n",
      "Low: True labels: ['OpenSwathDecoyGenerator']\n",
      "\n",
      "Low: Predicted labels: ['infernal_cmscan'], Precision: 0.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bam_to_sam', 'MergingDataClusterize']\n",
      "\n",
      "Low: True labels: ['removeIntrons', 'removeExonLines']\n",
      "\n",
      "Low: Predicted labels: ['removeIntrons', 'removeExonLines'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['bioext_bealign', 'bioext_bam2msa', 'hyphy_strike_ambigs', 'tn93_cluster']\n",
      "\n",
      "Low: True labels: ['tn93_filter', 'hyphy_annotate']\n",
      "\n",
      "Low: Predicted labels: ['tn93_filter', 'hyphy_annotate'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['1_create_conf', '4_filter_sam', '5_calc_stat']\n",
      "\n",
      "Low: True labels: ['conf4circos', '6_parse_discord']\n",
      "\n",
      "Low: Predicted labels: ['6_parse_discord', 'conf4circos'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low: test tool sequence: ['tp_find_and_replace', 'Convert characters1', 'tp_easyjoin_tool', 'Cut1', 'regex1', 'stoceps_maketablecarrer', 'stoceps_filteringsp', 'Filter1', 'stoceps_glm']\n",
      "\n",
      "Low: True labels: ['stoceps_glm_group']\n",
      "\n",
      "Low: Predicted labels: ['stoceps_glm_group'], Precision: 1.0\n",
      "-----------------\n",
      "\n",
      "Low test prediction precision: 0.8934606481481482\n"
     ]
    }
   ],
   "source": [
    "predict_low_freq_seq()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2938e6ae-7a49-45b8-9b11-1af23fd36a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "#predict_custom_seq()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f496caf-c8e4-4020-8cf9-82497377da20",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
