{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading workflows...\n",
      "Reading workflows finished\n",
      "Processing workflows...\n",
      "Workflows processed: 18659\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import random\n",
    "import numpy as np\n",
    "import json\n",
    "import warnings\n",
    "import operator\n",
    "\n",
    "import h5py\n",
    "from keras.models import model_from_json\n",
    "from keras import backend as K\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "\n",
    "def format_tool_id(tool_link):\n",
    "    \"\"\"\n",
    "    Extract tool id from tool link\n",
    "    \"\"\"\n",
    "    tool_id_split = tool_link.split(\"/\")\n",
    "    tool_id = tool_id_split[-2] if len(tool_id_split) > 1 else tool_link\n",
    "    return tool_id\n",
    "\n",
    "def read_workflow(wf_id, workflow_rows):\n",
    "    \"\"\"\n",
    "    Read all connections for a workflow\n",
    "    \"\"\"\n",
    "    tool_parents = dict()\n",
    "    for connection in workflow_rows:\n",
    "        in_tool = connection[0]\n",
    "        out_tool = connection[1]\n",
    "        if out_tool not in tool_parents:\n",
    "            tool_parents[out_tool] = list()\n",
    "        if in_tool not in tool_parents[out_tool]:\n",
    "            tool_parents[out_tool].append(in_tool)\n",
    "    return tool_parents\n",
    "\n",
    "def get_roots_leaves(graph):\n",
    "    roots = list()\n",
    "    leaves = list()\n",
    "    all_parents = list()\n",
    "    for item in graph:\n",
    "        all_parents.extend(graph[item])\n",
    "    all_parents = list(set(all_parents))\n",
    "    children = graph.keys()\n",
    "    roots = list(set(all_parents).difference(set(children)))\n",
    "    leaves = list(set(children).difference(set(all_parents)))\n",
    "    return roots, leaves\n",
    "\n",
    "def find_tool_paths_workflow(graph, start, end, path=[]):\n",
    "    path = path + [end]\n",
    "    if start == end:\n",
    "        return [path]\n",
    "    path_list = list()\n",
    "    if end in graph:\n",
    "        for node in graph[end]:\n",
    "            if node not in path:\n",
    "                new_tools_paths = find_tool_paths_workflow(graph, start, node, path)\n",
    "                for tool_path in new_tools_paths:\n",
    "                    path_list.append(tool_path)\n",
    "    return path_list\n",
    "\n",
    "def read_tabular_file(raw_file_path):\n",
    "    \"\"\"\n",
    "    Read tabular file and extract workflow connections\n",
    "    \"\"\"\n",
    "    print(\"Reading workflows...\")\n",
    "    workflows = {}\n",
    "    workflow_paths_dup = \"\"\n",
    "    workflow_parents = dict()\n",
    "    workflow_paths = list()\n",
    "    unique_paths = dict()\n",
    "    tool_frequency = dict()\n",
    "    standard_connections = dict()\n",
    "    with open(raw_file_path, 'rt') as workflow_connections_file:\n",
    "        workflow_connections = csv.reader(workflow_connections_file, delimiter='\\t')\n",
    "        for index, row in enumerate(workflow_connections):\n",
    "            wf_id = str(row[0])\n",
    "            in_tool = format_tool_id(row[3])\n",
    "            out_tool = format_tool_id(row[6])\n",
    "            if wf_id not in workflows:\n",
    "                workflows[wf_id] = list()\n",
    "            if out_tool and in_tool and out_tool != in_tool:\n",
    "                workflows[wf_id].append((out_tool, in_tool))\n",
    "            if out_tool != \"\":\n",
    "                if out_tool not in tool_frequency:\n",
    "                    tool_frequency[out_tool] = 0\n",
    "                tool_frequency[out_tool] += 1\n",
    "            if in_tool != \"\":\n",
    "                if in_tool not in tool_frequency:\n",
    "                    tool_frequency[in_tool] = 0\n",
    "                \n",
    "                tool_frequency[in_tool] += 1\n",
    "    print(\"Reading workflows finished\")\n",
    "    print(\"Processing workflows...\")\n",
    "    wf_ctr = 0\n",
    "    for wf_id in workflows:\n",
    "        wf_ctr += 1\n",
    "        workflow_parents[wf_id] = read_workflow(wf_id, workflows[wf_id])\n",
    "\n",
    "    for wf_id in workflow_parents:\n",
    "        flow_paths = list()\n",
    "        parents_graph = workflow_parents[wf_id]\n",
    "        roots, leaves = get_roots_leaves(parents_graph)\n",
    "        for root in roots:\n",
    "            for leaf in leaves:\n",
    "                paths = find_tool_paths_workflow(parents_graph, root, leaf)\n",
    "                # reverse the paths as they are computed from leaves to roots leaf\n",
    "                paths = [tool_path for tool_path in paths]\n",
    "                if len(paths) > 0:\n",
    "                    flow_paths.extend(paths)\n",
    "        workflow_paths.extend(flow_paths)\n",
    "    print(\"Workflows processed: %d\" % wf_ctr)\n",
    "\n",
    "    # remove slashes from the tool ids\n",
    "    wf_paths_no_slash = list()\n",
    "    for path in workflow_paths:\n",
    "        path_no_slash = [format_tool_id(tool_id) for tool_id in path]\n",
    "        wf_paths_no_slash.append(path_no_slash)\n",
    "\n",
    "    # collect duplicate paths\n",
    "    for path in wf_paths_no_slash:\n",
    "        workflow_paths_dup += \",\".join(path) + \"\\n\"\n",
    "    unique_paths = list(workflow_paths_dup.split(\"\\n\"))\n",
    "    unique_paths = list(filter(None, unique_paths))\n",
    "    random.shuffle(unique_paths)\n",
    "    paths_freq = dict()\n",
    "    for path in unique_paths:\n",
    "        if path not in paths_freq:\n",
    "            paths_freq[path] = 0\n",
    "        paths_freq[path] += 1\n",
    "        \n",
    "    return tool_frequency, paths_freq\n",
    "\n",
    "\n",
    "wf_path = \"../data/worflow-connection-20-04.tsv\"\n",
    "t_freq, p_freq = read_tabular_file(wf_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "last_tool_freq = dict()\n",
    "for path in p_freq:\n",
    "    last_t = path.split(\",\")[-1]\n",
    "    if last_t not in last_tool_freq:\n",
    "        last_tool_freq[last_t] = 0\n",
    "    last_tool_freq[last_t] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "t_freq  = dict(sorted(t_freq.items(), key=lambda kv: kv[1], reverse=False))\n",
    "p_freq = dict(sorted(p_freq.items(), key=lambda kv: kv[1], reverse=False))\n",
    "last_tool_freq = dict(sorted(last_tool_freq.items(), key=lambda kv: kv[1], reverse=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_file(file_path):\n",
    "    with open(file_path, 'r') as data_file:\n",
    "        data = json.loads(data_file.read())\n",
    "    return data\n",
    "\n",
    "\n",
    "def create_model(model_path):\n",
    "    reverse_dictionary = dict((str(v), k) for k, v in dictionary.items())\n",
    "    model_weights = list()\n",
    "    weight_ctr = 0\n",
    "    for index, item in enumerate(trained_model.keys()):\n",
    "        if \"weight_\" in item:\n",
    "            d_key = \"weight_\" + str(weight_ctr)\n",
    "            weights = trained_model.get(d_key).value\n",
    "            model_weights.append(weights)\n",
    "            weight_ctr += 1\n",
    "    # set the model weights\n",
    "    loaded_model.set_weights(model_weights)\n",
    "    return loaded_model, dictionary, reverse_dictionary\n",
    "\n",
    "\n",
    "def get_predicted_tools(base_tools, predictions, topk):\n",
    "    \"\"\"\n",
    "    Get predicted tools. If predicted tools are less in number, combine them with published tools\n",
    "    \"\"\"\n",
    "    precision = np.nan\n",
    "    intersection = list()\n",
    "    if len(base_tools) > 0:\n",
    "        intersection = list(set(predictions).intersection(set(base_tools)))\n",
    "        precision = len(intersection) / float(len(predictions))\n",
    "    #print(base_tools)\n",
    "    #print(intersection)\n",
    "    #print(precision)\n",
    "    #print()\n",
    "    return intersection[:topk], precision\n",
    "\n",
    "\n",
    "def sort_by_usage(t_list, class_weights, d_dict):\n",
    "    \"\"\"\n",
    "    Sort predictions by usage/class weights\n",
    "    \"\"\"\n",
    "    tool_dict = dict()\n",
    "    for tool in t_list:\n",
    "        t_id = d_dict[tool]\n",
    "        tool_dict[tool] = class_weights[str(t_id)]\n",
    "    #tool_dict = dict(sorted(tool_dict.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    return list(tool_dict.keys()), list(tool_dict.values())\n",
    "\n",
    "\n",
    "def separate_predictions(base_tools, predictions, last_tool_name, weight_values, topk):\n",
    "    \"\"\"\n",
    "    Get predictions from published and normal workflows\n",
    "    \"\"\"\n",
    "    last_base_tools = list()\n",
    "    predictions = predictions * weight_values\n",
    "    prediction_pos = np.argsort(predictions, axis=-1)\n",
    "    topk_prediction_pos = prediction_pos[-topk:]\n",
    "    # get tool ids\n",
    "    pred_tool_ids = [reverse_dictionary[str(tool_pos)] for tool_pos in topk_prediction_pos]\n",
    "    if last_tool_name in base_tools:\n",
    "        last_base_tools = base_tools[last_tool_name]\n",
    "        if type(last_base_tools).__name__ == \"str\":\n",
    "            # get published or compatible tools for the last tool in a sequence of tools\n",
    "            last_base_tools = last_base_tools.split(\",\")\n",
    "    # get predicted tools\n",
    "    p_tools, precision = get_predicted_tools(last_base_tools, pred_tool_ids, topk)\n",
    "    sorted_c_t, sorted_c_v = sort_by_usage(p_tools, class_weights, dictionary)\n",
    "    return sorted_c_t, sorted_c_v, precision\n",
    "\n",
    "\n",
    "def compute_recommendations(model, tool_sequence, labels, dictionary, reverse_dictionary, class_weights, topk=10, max_seq_len=25):\n",
    "    tl_seq = tool_sequence.split(\",\")\n",
    "    tl_seq_ids = [str(dictionary[t]) for t in tl_seq]\n",
    "    last_tool_name = tl_seq[-1]\n",
    "    sample = np.zeros(max_seq_len)\n",
    "    weight_val = list(class_weights.values())\n",
    "    weight_val = np.reshape(weight_val, (len(weight_val),))\n",
    "    for idx, tool_id in enumerate(tl_seq_ids):\n",
    "        sample[idx] = int(tool_id)\n",
    "    sample_reshaped = np.reshape(sample, (1, max_seq_len))\n",
    "    tool_sequence_names = [reverse_dictionary[str(tool_pos)] for tool_pos in tl_seq_ids]\n",
    "    # predict next tools for a test path\n",
    "    prediction = model.predict(sample_reshaped, verbose=0)\n",
    "    nw_dimension = prediction.shape[1]\n",
    "    prediction = np.reshape(prediction, (nw_dimension,))\n",
    "    \n",
    "    half_len = int(nw_dimension / 2)\n",
    "    \n",
    "    pub_t, pub_v, pub_prec = separate_predictions(standard_connections, prediction[:half_len], last_tool_name, weight_val, topk)\n",
    "    # get recommended tools from normal workflows\n",
    "    c_t, c_v, c_prec = separate_predictions(compatible_tools, prediction[half_len:], last_tool_name, weight_val, topk)\n",
    "    \n",
    "    return pub_prec, c_prec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_path = \"../data/tool_recommendation_model_20_04.hdf5\"\n",
    "trained_model = h5py.File(model_path, 'r')\n",
    "model_config = json.loads(trained_model.get('model_config').value)\n",
    "dictionary = json.loads(trained_model.get('data_dictionary').value)\n",
    "class_weights = json.loads(trained_model.get('class_weights').value)\n",
    "standard_connections = json.loads(trained_model.get('standard_connections').value)\n",
    "compatible_tools = json.loads(trained_model.get('compatible_tools').value)\n",
    "loaded_model = model_from_json(model_config)\n",
    "model, dictionary, reverse_dictionary = create_model(model_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# get tools with lowest frequency\n",
    "t_lowest_freq = dict()\n",
    "for t in last_tool_freq:\n",
    "    if last_tool_freq[t] < 10:\n",
    "        t_lowest_freq[t] = last_tool_freq[t]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mtbls520_23_seasons_rda': 1,\n",
       " 'mtbls520_08c_species_variability': 1,\n",
       " 'keras_batch_models': 1,\n",
       " 'metagene_annotator': 1,\n",
       " 'rseqc_RNA_fragment_size': 1,\n",
       " 'mtbls520_19b_seasons_unique': 1,\n",
       " 'get_sequences': 1,\n",
       " 'minfi_getM': 1,\n",
       " 'mummer_nucmer': 1,\n",
       " 'fpocket': 1,\n",
       " 'ambertools_antechamber': 1,\n",
       " 'tag_stat2': 1,\n",
       " 'ip_scale_image': 1,\n",
       " 'mlocarna': 1,\n",
       " 'cp_image_math': 1,\n",
       " 'aggregate': 1,\n",
       " 'collection_element_identifiers': 1,\n",
       " 'bed_to_bigBed': 1,\n",
       " 'ggplot2_pca': 1,\n",
       " 'mtbls520_08d_concentration': 1,\n",
       " 'ip_histogram_equalization': 1,\n",
       " 'rseqc_clipping_profile': 1,\n",
       " 'cp_relate_objects': 1,\n",
       " 'gops_basecoverage_1': 1,\n",
       " 'Nucleosome': 1,\n",
       " 'gemini_de_novo': 1,\n",
       " 'vcfaddinfo': 1,\n",
       " 'mtbls520_09_species_venn': 1,\n",
       " 'pileup_interval': 1,\n",
       " 'mothur_phylotype': 1,\n",
       " 'vsearch_sorting': 1,\n",
       " 'tp_uniq_tool': 1,\n",
       " 'interactive_tool_phinch': 1,\n",
       " 'p_clip_peaks': 1,\n",
       " 'mtbls520_08b_species_unique': 1,\n",
       " 'scater_plot_tsne': 1,\n",
       " 'deeptools_alignmentsieve': 1,\n",
       " 'ctb_filter': 1,\n",
       " 'ip_binary_to_labelimage': 1,\n",
       " 'picard_SortSam': 1,\n",
       " 'hgv_linkToGProfile': 1,\n",
       " 'peakcalling_sicer': 1,\n",
       " 'velvetg_jgi': 1,\n",
       " 'mothur_align_seqs': 1,\n",
       " 'gemini_comp_hets': 1,\n",
       " 'cshl_fasta_nucleotides_changer': 1,\n",
       " 'ConsensusID': 1,\n",
       " 'CONVERTER_gz_to_uncompressed': 1,\n",
       " 'keras_train_and_eval': 1,\n",
       " 'diffbind-embl': 1,\n",
       " 'mtbls520_19e_seasons_features': 1,\n",
       " 'express': 1,\n",
       " 'get_feature_info': 1,\n",
       " 'bedtools_multiintersectbed': 1,\n",
       " 'deeptools_bigwigCompare': 1,\n",
       " 'maxquant': 1,\n",
       " 'climate_stripes': 1,\n",
       " 'idba_hybrid': 1,\n",
       " 'signalp3': 1,\n",
       " 'mtbls520_08a_species_shannon': 1,\n",
       " 'lastz_wrapper_2': 1,\n",
       " 'interactive_tool_fairdom_send': 1,\n",
       " 'maldi_quant_preprocessing': 1,\n",
       " 'cd_hit_est': 1,\n",
       " '_ensembl_gtf2gene_list': 1,\n",
       " 'bedtools_groupbybed': 1,\n",
       " 'convert_bc_to_binary_RY.py': 1,\n",
       " 'gemini_interactions': 1,\n",
       " '__EXTRACT_DATASET__': 1,\n",
       " 'Remove_ending': 1,\n",
       " 'samtools_markdup': 1,\n",
       " 'xpath': 1,\n",
       " 'htseq-count': 1,\n",
       " 'blastxml_to_tabular': 1,\n",
       " 'subtract_query1': 1,\n",
       " 'rseqc_insertion_profile': 1,\n",
       " 'mothur_make_group': 1,\n",
       " 'rbc_mirdeep2_mapper': 1,\n",
       " 'modencode_peakcalling_spp': 1,\n",
       " 'samtools_merge': 1,\n",
       " 'minfi_methcpg': 1,\n",
       " 'ctb_chemfp_mol2fps': 1,\n",
       " 'mtbls520_18_phylogeny': 1,\n",
       " 'microsat_pal_finder': 1,\n",
       " 'flexbar_no_split': 1,\n",
       " 'raceid_clustering': 1,\n",
       " 'xchem_transfs_scoring': 1,\n",
       " 'picard_NormalizeFasta': 1,\n",
       " 'CONVERTER_bed_to_bgzip_0': 1,\n",
       " 'sailfish': 1,\n",
       " 'scanpy_compute_graph': 1,\n",
       " 'genomespace_exporter': 1,\n",
       " 'macs2_bdgcmp': 1,\n",
       " 'kallisto_pseudo': 1,\n",
       " 'snp_sites': 1,\n",
       " 'rgPicFixMate': 1,\n",
       " 'fastq_stats': 1,\n",
       " 'snpSift_dbnsfp_generic': 1,\n",
       " 'Batch_correction': 1,\n",
       " 'gmx_editconf': 1,\n",
       " 'ivar_consensus': 1,\n",
       " 'gemini_pathways': 1,\n",
       " 'correctGCBias': 1,\n",
       " 'enhanced_bowtie_wrapper': 1,\n",
       " 'shyft_longitudes': 1,\n",
       " 'fastp': 1,\n",
       " 'mothur_make_contigs': 1,\n",
       " 'SequenceCoverageCalculator': 1,\n",
       " 'metaspades': 1,\n",
       " 'seqtk_mergefa': 1,\n",
       " 'mummer_mummer': 1,\n",
       " 'mtbls520_10_species_varpart': 1,\n",
       " 'gemini_stats': 1,\n",
       " 'gemini_windower': 1,\n",
       " 'cufflinks_prok': 1,\n",
       " 'vsearch_dereplication': 1,\n",
       " 'vcfselectsamples': 1,\n",
       " 'sklearn_feature_selection': 1,\n",
       " 'scater_plot_exprs_freq': 1,\n",
       " 'bedtools_slopbed': 1,\n",
       " 'FROGS_affiliations_stat': 1,\n",
       " 'picard_CleanSam': 1,\n",
       " 'progressivemauve': 1,\n",
       " 'hicexplorer_hicplotviewpoint': 1,\n",
       " 'extract_aln_ends.py': 1,\n",
       " 'hicexplorer_hicmergematrixbins': 1,\n",
       " 'secure_hash_message_digest': 1,\n",
       " 'mummer_mummerplot': 1,\n",
       " 'tabular_to_csv': 1,\n",
       " 'cd_hit_protein': 1,\n",
       " 'mtbls520_12_species_marchantia': 1,\n",
       " 'IDMassAccuracy': 1,\n",
       " 'mtbls520_08e_species_features': 1,\n",
       " 'minfi_dmp': 1,\n",
       " 'repeatexplorer2': 1,\n",
       " 'OpenSwathWorkflow': 1,\n",
       " 'gatk2_variant_filtration': 1,\n",
       " 'tp_sort_rows': 1,\n",
       " 'igvtools_count': 1,\n",
       " 'bamSplit': 1,\n",
       " 'numeric_clustering': 1,\n",
       " 'tarean_exp': 1,\n",
       " 'smf_utils_find-boxes': 1,\n",
       " 'mtbls520_19d_seasons_concentration': 1,\n",
       " 'sklearn_clf_metrics': 1,\n",
       " 'cp_export_to_spreadsheet': 1,\n",
       " 'eggnog_mapper': 1,\n",
       " 'gd_calc_freq': 1,\n",
       " 'metaphlan2krona': 1,\n",
       " 'ucsc-twobittofa': 1,\n",
       " 'cp_convert_objects_to_image': 1,\n",
       " 'methtools_tiling': 1,\n",
       " 'FeatureFinderMultiplex': 1,\n",
       " 'mothur_summary_qual': 1,\n",
       " 'bamtools_split': 1,\n",
       " 'IDPosteriorErrorProbability': 1,\n",
       " 'rseqc_mismatch_profile': 1,\n",
       " 'ip_2d_split_binaryimage_by_watershed': 1,\n",
       " 'Determine_BC': 1,\n",
       " 'exonerate': 1,\n",
       " 'ctb_ob_genProp': 1,\n",
       " 'percolator_input_converters': 1,\n",
       " 'stacks_refmap': 1,\n",
       " 'padel': 1,\n",
       " 'metaphlan2_hutlab': 1,\n",
       " 'FeatureLinkerUnlabeled': 1,\n",
       " 'Fetch Taxonomic Ranks': 1,\n",
       " 'bed_to_bigwig': 1,\n",
       " 'countKeggs': 1,\n",
       " 'cast': 1,\n",
       " 'vsearch_search': 1,\n",
       " 'param_value_from_file': 1,\n",
       " 'mtbls520_19a_seasons_shannon': 1,\n",
       " 'bgchem_fragment_merger': 1,\n",
       " 'CONVERTER_fasta_to_tabular': 1,\n",
       " 'EMBOSS: wordcount109': 1,\n",
       " 'CONVERTER_bed_to_gff_0': 1,\n",
       " 'extract_bcs.py': 1,\n",
       " 'TargetedFileConverter': 1,\n",
       " 'sRNAPipe': 1,\n",
       " 'Kraken2Tax': 1,\n",
       " 'mtbls520_16_ecology_rda': 1,\n",
       " 'minfi_getanno': 1,\n",
       " 'flye': 1,\n",
       " 'sm_mirdeep2core_without_Randfold': 1,\n",
       " 'AnnovarShed': 1,\n",
       " 'sam2interval': 1,\n",
       " 'pureclip': 1,\n",
       " 'cshl_easyjoin': 1,\n",
       " 'bg_diamond': 1,\n",
       " 'EMBOSS: water107': 1,\n",
       " 'gemini_burden': 1,\n",
       " 'trnascan': 1,\n",
       " 'vsearch_clustering': 1,\n",
       " 'dada2_filterAndTrim': 1,\n",
       " 'abims_CAMERA_combinexsAnnos': 1,\n",
       " 'gdal_gdalinfo': 1,\n",
       " 'deeptools_plot_enrichment': 1,\n",
       " 'miranda': 1,\n",
       " 'gtf2bedgraph': 1,\n",
       " 'w4mclassfilter': 1,\n",
       " 'MapStatistics': 1,\n",
       " 'minfi_getsnp': 1,\n",
       " 'cshl_princeton_fastx_barcode_splitter': 1,\n",
       " 'ip_filter_standard': 1,\n",
       " 'sample_seqs': 1,\n",
       " 'EMBOSS: textsearch98': 1,\n",
       " 'sequence_content_trimmer': 1,\n",
       " 'hmmer_hmmscan': 1,\n",
       " 'diffexp': 1,\n",
       " 'Kneeplot': 1,\n",
       " 'ncbi_tblastn_wrapper': 1,\n",
       " 'maker_map_ids': 1,\n",
       " 'stoceps_glm': 1,\n",
       " 'react_cal_pipeline': 1,\n",
       " 'hicexplorer_hicsummatrices': 1,\n",
       " 'stacks2_populations': 1,\n",
       " 'ProteinResolver': 1,\n",
       " 'gff_to_sequence': 1,\n",
       " 'mimodd_info': 1,\n",
       " 'miniasm': 1,\n",
       " 'vcfbreakcreatemulti': 1,\n",
       " 'picard_MarkDuplicatesWithMateCigar': 1,\n",
       " 'abims_xcms_summary': 1,\n",
       " 'ctb_osra': 1,\n",
       " 'fastq_paired_end_splitter': 1,\n",
       " 'reactome_analysis': 1,\n",
       " 'Krona': 1,\n",
       " 'fastq_filter': 1,\n",
       " 'sklearn_train_test_eval': 1,\n",
       " 'Jvenn': 1,\n",
       " 'vigiechiro_bilanenrichirp': 1,\n",
       " 'gops_subtract_1': 1,\n",
       " 'rdock_rbcavity': 1,\n",
       " 'rmcontamination': 1,\n",
       " 'dt_profiler': 1,\n",
       " 'nonpareil': 1,\n",
       " 'pyprophet_export': 1,\n",
       " 'mass_spectrometry_imaging_segmentations': 1,\n",
       " 'sceasy_convert': 1,\n",
       " 'stacks_genotypes': 1,\n",
       " '_dropletBarcodePlot': 1,\n",
       " 'gatk4_mutect2': 1,\n",
       " 'ip_imageinfo': 1,\n",
       " 'openbabel_compound_convert': 1,\n",
       " 'minfi_mset': 1,\n",
       " 'ambertools_acpype': 1,\n",
       " 'vg_deconstruct': 1,\n",
       " 'minfi_ppfun': 1,\n",
       " 'psy_maps': 1,\n",
       " 'minfi_qc': 1,\n",
       " 'ctb_obgrep': 1,\n",
       " 'fgsea': 1,\n",
       " 'dropletutils': 1,\n",
       " 'mummer_dnadiff': 1,\n",
       " 'hyphy_gard': 1,\n",
       " 'EMBOSS: revseq82': 1,\n",
       " 'deeptools_bigwig_compare': 1,\n",
       " 'qiime_count_seqs': 1,\n",
       " 'vigiechiro_bilanenrichipf': 1,\n",
       " 'vegan_rarefaction': 2,\n",
       " 'mcClust': 2,\n",
       " 'cshl_fastx_trimmer': 2,\n",
       " 'picard_DownsampleSam': 2,\n",
       " 'sklearn_stacking_ensemble_models': 2,\n",
       " '16Sclassifier': 2,\n",
       " 'seurat_export_cellbrowser': 2,\n",
       " 'bwa': 2,\n",
       " 'varscan_copynumber': 2,\n",
       " 'pick_open_reference_otus': 2,\n",
       " 'mothur_collect_single': 2,\n",
       " 'picard_FixMateInformation': 2,\n",
       " 'extract': 2,\n",
       " 'ggplot2_heatmap': 2,\n",
       " 'abims_xcms_fillPeaks': 2,\n",
       " 'ete_homology_classifier': 2,\n",
       " 'gd_raxml': 2,\n",
       " 'w4mclstrpeakpics': 2,\n",
       " 'iframe': 2,\n",
       " 'vcfannotate': 2,\n",
       " 'minfi_dmr': 2,\n",
       " 'lofreq_call': 2,\n",
       " 'mothur_remove_seqs': 2,\n",
       " 'bedtools_subtractbed': 2,\n",
       " 'mmpbsa_mmgbsa': 2,\n",
       " 'vcf2pgSnp': 2,\n",
       " 'map_peptides_to_bed': 2,\n",
       " 'ncbi_makeblastdb': 2,\n",
       " 'megahit_contig2fastg': 2,\n",
       " 'ctb_rdkit_descriptors': 2,\n",
       " 'mothur_merge_files': 2,\n",
       " 'ip_2d_filter_segmentation_by_features': 2,\n",
       " 'w4mcorcov': 2,\n",
       " 'bg_statistical_hypothesis_testing': 2,\n",
       " 'heinz': 2,\n",
       " 'vsearch_chimera_detection': 2,\n",
       " 'schicexplorer_schicplotconsensusmatrices': 2,\n",
       " 'mothur_chimera_vsearch': 2,\n",
       " 'biom_convert': 2,\n",
       " 'EMBOSS: fuzzpro38': 2,\n",
       " 'filter_tabular': 2,\n",
       " 'tbvcfreport': 2,\n",
       " 'modencode_peakcalling_macs2': 2,\n",
       " 'find_in_reference': 2,\n",
       " 'graphicsmagick_image_montage': 2,\n",
       " 'schicexplorer_schicplotclusterprofiles': 2,\n",
       " 'rgPicardMarkDups': 2,\n",
       " 'vegan_diversity': 2,\n",
       " 'rdpmulticlassifier': 2,\n",
       " 'scmap_scmap_cluster': 2,\n",
       " 'iuc_deseq2': 2,\n",
       " 'heatmapper': 2,\n",
       " 'dada2_dada': 2,\n",
       " 'md_converter': 2,\n",
       " 'bg_sortmerna': 2,\n",
       " 'maker': 2,\n",
       " 'custom_pro_db': 2,\n",
       " 'picard_ValidateSamFile': 2,\n",
       " 'clusterprofiler_go': 2,\n",
       " 'regexColumn1': 2,\n",
       " 'viennarna_rnafold': 2,\n",
       " 'mothur_chimera_uchime': 2,\n",
       " 'FidoAdapter': 2,\n",
       " 'minimap2': 2,\n",
       " '__MERGE_COLLECTION__': 2,\n",
       " 'idpquery': 2,\n",
       " 'rseqc_read_duplication': 2,\n",
       " 'samtools_slice_bam': 2,\n",
       " 'snpsift_vartype': 2,\n",
       " 'basil': 2,\n",
       " 'picard_EstimateLibraryComplexity': 2,\n",
       " 'sqlite_to_tabular': 2,\n",
       " 'picard_AddOrReplaceReadGroups': 2,\n",
       " 'gene2exon1': 2,\n",
       " 'seurat': 2,\n",
       " 'kraken-mpa-report': 2,\n",
       " 'checkFormat': 2,\n",
       " 'pieplots_macs': 2,\n",
       " 'gatk_indel_realigner': 2,\n",
       " 'FastTree': 2,\n",
       " 'sam_pileup': 2,\n",
       " 'gmx_setup': 2,\n",
       " 'trinity_abundance_estimates_to_matrix': 2,\n",
       " 'predict_pipeline': 2,\n",
       " 'scmap_scmap_cell': 2,\n",
       " 'qiime_align_seqs': 2,\n",
       " 'bcftools_stats': 2,\n",
       " '__FILTER_FAILED_DATASETS__': 2,\n",
       " 'openbabel_svg_depiction': 2,\n",
       " 'blockclust': 2,\n",
       " 'gatk2_variant_select': 2,\n",
       " 'minfi_getbeta': 2,\n",
       " 'cshl_awk_replace_in_column': 2,\n",
       " 'raceid_inspecttrajectory': 2,\n",
       " 'openms_id_mapper': 2,\n",
       " 'MultiplexResolver': 2,\n",
       " 'fastq_trimmer': 2,\n",
       " 'methtools_calling': 2,\n",
       " 'cshl_fastq_to_fasta': 2,\n",
       " 'bedtools_genomecoveragebed_histogram': 2,\n",
       " 'rseqc_junction_annotation': 2,\n",
       " 'bismark_bowtie': 2,\n",
       " 'shovill': 2,\n",
       " 'flashlfq': 2,\n",
       " 'vcfgenotypes': 2,\n",
       " 'rna_starsolo': 2,\n",
       " 'gatk2_realigner_target_creator': 2,\n",
       " 'deeptools_bam_pe_fragmentsize': 2,\n",
       " 'CONVERTER_bedgraph_to_bigwig': 2,\n",
       " 'EMBOSS: geecee41': 2,\n",
       " 'seq_composition': 2,\n",
       " 'mothur_unique_seqs': 2,\n",
       " 'gdal_ogr2ogr': 2,\n",
       " 'kraken-translate': 2,\n",
       " 'heinz_scoring': 2,\n",
       " 'sklearn_searchcv': 2,\n",
       " 'blockbuster': 2,\n",
       " 'gatk_unified_genotyper': 2,\n",
       " 'cummerbund_to_cuffdiff': 2,\n",
       " 'rdock_rbdock': 2,\n",
       " 'smf_utils_filter-by-energy': 2,\n",
       " 'ia_coordinates_of_roi': 2,\n",
       " 'filtlong': 2,\n",
       " 'Psortb': 2,\n",
       " 'openbabel_remDuplicates': 2,\n",
       " 'velvetoptimiser': 2,\n",
       " 'sklearn_model_validation': 2,\n",
       " 'iedb_api': 2,\n",
       " 'sccaf_regress_out': 2,\n",
       " 'cluter_profiler': 2,\n",
       " 'nspdk_sparse': 2,\n",
       " 'goenrichment': 2,\n",
       " 'xchem_pose_scoring': 2,\n",
       " 'plotly_parallel_coordinates_plot': 2,\n",
       " 'sccaf_asses_merger': 2,\n",
       " 'deeptools_correct_gc_bias': 2,\n",
       " 'picard_ReorderSam': 2,\n",
       " 'decoyfasta': 2,\n",
       " 'metaquantome_viz': 2,\n",
       " 'bg_protein_properties': 2,\n",
       " 'cat_bins': 2,\n",
       " 'nn_classifier': 2,\n",
       " 'snpfinder': 2,\n",
       " 'resistome_analyzer': 2,\n",
       " 'infernal_cmstat': 2,\n",
       " 'hyphy_absrel': 2,\n",
       " 'fastq_paired_end_interlacer': 2,\n",
       " 'qiime_core_diversity': 2,\n",
       " 'sixgill_build': 3,\n",
       " 'bcftools_norm': 3,\n",
       " 'bedtools_bed12tobed6': 3,\n",
       " 'cshl_multijoin': 3,\n",
       " 'rseqc_junction_saturation': 3,\n",
       " 'sam_bw_filter': 3,\n",
       " 'rseqc_FPKM_count': 3,\n",
       " 'trinity_align_and_estimate_abundance': 3,\n",
       " 'get_flanks1': 3,\n",
       " 'cshl_fastx_nucleotides_distribution': 3,\n",
       " 'smooth_running_window': 3,\n",
       " 'scater_plot_dist_scatter': 3,\n",
       " 'samtools_calmd': 3,\n",
       " 'cshl_find_and_replace': 3,\n",
       " 'mass_spectrometry_imaging_filtering': 3,\n",
       " 'dexseq_annotate': 3,\n",
       " 'scater_plot_pca': 3,\n",
       " 'fastq_groomer': 3,\n",
       " 'search_gui': 3,\n",
       " 'uniprot': 3,\n",
       " 'table_annovar': 3,\n",
       " 'cat_summarise': 3,\n",
       " 'velvetg': 3,\n",
       " 'raceid_trajectory': 3,\n",
       " 'gatk2_base_recalibrator': 3,\n",
       " 'bcftools_plugin_counts': 3,\n",
       " 'rgweblogo3': 3,\n",
       " 'merge_pcr_duplicates.py': 3,\n",
       " 'mimodd_map': 3,\n",
       " 'picard_BamIndexStats': 3,\n",
       " 'ip_projective_transformation_points': 3,\n",
       " 'tophat_fusion_post': 3,\n",
       " 'vcfbedintersect': 3,\n",
       " 'bcftools_view': 3,\n",
       " 'gmx_trj': 3,\n",
       " 'peakcalling_macs': 3,\n",
       " 'gffread': 3,\n",
       " 'fasterq_dump': 3,\n",
       " 'EMBOSS: getorf42': 3,\n",
       " 'picard_MergeSamFiles': 3,\n",
       " 'ip_2d_feature_extraction': 3,\n",
       " 'circgraph': 3,\n",
       " 'cardinal_data_exporter': 3,\n",
       " 'bwa_wrapper': 3,\n",
       " 'vcftools_slice': 3,\n",
       " 'bam_to_bigwig': 3,\n",
       " 'umi_tools_group': 3,\n",
       " 'ctb_remove_protonation_state': 3,\n",
       " 'jcvi_gff_stats': 3,\n",
       " 'mass_spectrometry_imaging_mzplots': 3,\n",
       " 'gatk_analyze_covariates': 3,\n",
       " 'bismark_pretty_report': 3,\n",
       " 'segemehl': 3,\n",
       " 'vcftools_isec': 3,\n",
       " 'picard_CollectWgsMetrics': 3,\n",
       " 'deeptools_compute_matrix_operations': 3,\n",
       " 'mass_spectrometry_imaging_classification': 3,\n",
       " 'EMBOSS: tranalign100': 3,\n",
       " 'tp_split_on_column': 3,\n",
       " 'bedtools_sortbed': 3,\n",
       " 'cutadapt': 3,\n",
       " 'snpSift_int': 3,\n",
       " 'snpeff_to_peptides': 3,\n",
       " 'unipept': 3,\n",
       " 'cor2': 3,\n",
       " 'snippy_core': 3,\n",
       " 'umi_tools_extract': 3,\n",
       " 'gtftobed12': 3,\n",
       " 'trimmer': 3,\n",
       " 'Datamash': 3,\n",
       " 'trinity_samples_qccheck': 3,\n",
       " 'canu': 3,\n",
       " 'aggregate_scores_in_intervals2': 3,\n",
       " 'trinotate': 3,\n",
       " 'lfq_protein_quant': 3,\n",
       " 'metfrag-cli-batch-multiple': 3,\n",
       " 'gmx_restraints': 3,\n",
       " 'medaka_consensus': 3,\n",
       " 'crosscontamination_barcode_filter': 3,\n",
       " 'deeptools_computeGCBias': 3,\n",
       " 'gemini_annotate': 3,\n",
       " 'ococo': 3,\n",
       " 'ip_viz_overlay_moving_and_fixed_image': 3,\n",
       " 'cmv': 3,\n",
       " 'megablast_wrapper': 4,\n",
       " 'bamtools': 4,\n",
       " 'naive_variant_caller': 4,\n",
       " 'tabular_to_fastq': 4,\n",
       " 'split_file_to_collection': 4,\n",
       " 'gmx_em': 4,\n",
       " 'translate_bed_sequences': 4,\n",
       " 'kallisto_quant': 4,\n",
       " 'gatk_count_covariates': 4,\n",
       " 'megablast_xml_parser': 4,\n",
       " 'abims_CAMERA_annotateDiffreport': 4,\n",
       " 'bedtools_bedtobam': 4,\n",
       " 'hicexplorer_hicpca': 4,\n",
       " 'maxbin2': 4,\n",
       " 'bundle_collection': 4,\n",
       " '__FLATTEN__': 4,\n",
       " 'ip_count_objects': 4,\n",
       " 'sklearn_ensemble': 4,\n",
       " 'datamash_transpose': 4,\n",
       " 'cshl_awk_tool': 4,\n",
       " 'cardinal_preprocessing': 4,\n",
       " 'xcms_merge': 4,\n",
       " 'deeptools_profiler': 4,\n",
       " 'qiime_assign_taxonomy': 4,\n",
       " 'bedtools_genomecoveragebed_bedgraph': 4,\n",
       " 'proteomics_search_protein_prophet_1': 4,\n",
       " 'regionalgam_autocor_acf': 4,\n",
       " 'MetaProSIP': 4,\n",
       " 'bcftools_call': 4,\n",
       " 'samtools_bedcov': 4,\n",
       " 'raceid_inspectclusters': 4,\n",
       " 'ctb_shapeit': 4,\n",
       " 'paralyzer': 4,\n",
       " 'scanpy_run_umap': 4,\n",
       " 'circos': 4,\n",
       " 'GeMoMa_Annotation_Filter': 4,\n",
       " 'snpSift_geneSets': 4,\n",
       " 'qualimap_rnaseq': 4,\n",
       " 'rarefaction': 4,\n",
       " 'cshl_fastq_quality_boxplot': 4,\n",
       " 'deseq2_single': 4,\n",
       " 'bamCoverage_deepTools': 4,\n",
       " 'seq_filter_by_mapping': 4,\n",
       " 'ctb_compound_convert': 4,\n",
       " 'regionalgam_glmmpql': 4,\n",
       " 'allele_counts_1': 4,\n",
       " 'seurat_dim_plot': 4,\n",
       " 'gops_join_1': 4,\n",
       " 'gmx_solvate': 4,\n",
       " '__BUILD_LIST__': 4,\n",
       " '__TAG_FROM_FILE__': 4,\n",
       " 'tmhmm2': 4,\n",
       " 'fastq_to_tabular': 4,\n",
       " 'hicexplorer_hicfindtads': 4,\n",
       " 'mass_spectrometry_imaging_combine': 4,\n",
       " 'tb_variant_filter': 4,\n",
       " 'bedtools_bamtofastq': 4,\n",
       " 'deeptools_computeMatrix': 4,\n",
       " 'stoceps_glm_group': 4,\n",
       " 'diffbind': 4,\n",
       " 'rbc_rnacode': 4,\n",
       " 'bamleftalign': 4,\n",
       " 'mdanalysis_cosine_analysis': 4,\n",
       " 'bio3d_pca_visualize': 4,\n",
       " 'vmd_hbonds': 4,\n",
       " 'flexbar': 4,\n",
       " 'seqtk_subseq': 4,\n",
       " 'viz_overlay_moving_and_fixed_image': 4,\n",
       " 'metaphlan_hclust_heatmap': 5,\n",
       " 'umi_tools_dedup': 5,\n",
       " 'ivar_variants': 5,\n",
       " 'cshl_fastq_quality_filter': 5,\n",
       " 'quality_metrics': 5,\n",
       " 'ChangeCase': 5,\n",
       " 'bam2wig': 5,\n",
       " 'roary': 5,\n",
       " 'massbank_ws_searchspectrum': 5,\n",
       " 'metilene': 5,\n",
       " 'bedtools_multicovtbed': 5,\n",
       " 'bedtools_coveragebed': 5,\n",
       " 'picard_SamToFastq': 5,\n",
       " 'cp_cellprofiler': 5,\n",
       " 'bg_column_arrange_by_header': 5,\n",
       " 'tp_tac': 5,\n",
       " 'macs2_predictd': 5,\n",
       " 'sdf_to_tab': 5,\n",
       " 'samtools_depth': 5,\n",
       " 'limma_voom': 5,\n",
       " 'addName': 5,\n",
       " 'dexseq_count': 5,\n",
       " 'aragorn_trna': 5,\n",
       " 'PeptideIndexer': 5,\n",
       " 'trinity_run_de_analysis': 5,\n",
       " 'mothur_screen_seqs': 5,\n",
       " 'vcftools_merge': 5,\n",
       " 'CONVERTER_bed_gff_or_vcf_to_bigwig_0': 5,\n",
       " 'rbc_mirdeep2': 5,\n",
       " 'cuffquant': 5,\n",
       " 'bismark_methylation_extractor': 5,\n",
       " 'heinz_visualization': 5,\n",
       " 'annotateMyIDs': 5,\n",
       " 'trinity_contig_exn50_statistic': 5,\n",
       " 'methtools_destrand': 5,\n",
       " 'IDScoreSwitcher': 5,\n",
       " 'IDFilter': 5,\n",
       " 'gops_concat_1': 5,\n",
       " 'mergeCols1': 5,\n",
       " 'stacks_populations': 5,\n",
       " 'picard_CollectRnaSeqMetrics': 5,\n",
       " 'CONVERTER_Bam_Bai_0': 5,\n",
       " 'rbc_mirdeep2_quantifier': 5,\n",
       " 'viennarna_rnaalifold': 5,\n",
       " 'regex1': 5,\n",
       " 'plotly_regression_performance_plots': 5,\n",
       " 'mimodd_vcf_filter': 5,\n",
       " 'augustus': 5,\n",
       " 'table_compute': 5,\n",
       " 'varscan_mpileup': 6,\n",
       " 'cuffnorm': 6,\n",
       " 'racon': 6,\n",
       " 'gatk2_print_reads': 6,\n",
       " 'ctb_subsearch': 6,\n",
       " 'xcms_plot_chromatogram': 6,\n",
       " 'muscle': 6,\n",
       " 'mimodd_varreport': 6,\n",
       " 'gatk2_unified_genotyper': 6,\n",
       " 'stringtie_merge': 6,\n",
       " 'bismark_bowtie2': 6,\n",
       " 'bowtie_wrapper': 6,\n",
       " 'samtools_view': 6,\n",
       " 'varscan_somatic': 6,\n",
       " 'build_ucsc_custom_track_1': 6,\n",
       " 'PicardASMetrics': 6,\n",
       " 'vcfvcfintersect': 6,\n",
       " 'snippy': 6,\n",
       " 'PicardInsertSize': 6,\n",
       " 'cshl_sort_header': 6,\n",
       " 'EMBOSS: transeq101': 6,\n",
       " 'regionalgam_plot_trend': 6,\n",
       " 'CONVERTER_gff_to_bed_0': 6,\n",
       " 'rseqc_geneBody_coverage': 6,\n",
       " 'cshl_word_list_grep': 6,\n",
       " 'cardinal_combine': 6,\n",
       " 'IDMapper': 6,\n",
       " 'sucos_max_score': 6,\n",
       " 'vcf_filter': 6,\n",
       " 'raceid_main': 6,\n",
       " 'venn_list': 6,\n",
       " 'cuffmerge': 6,\n",
       " 'filter_by_fasta_ids': 6,\n",
       " 'blast2go': 6,\n",
       " 'cloudmap_variant_discovery_mapping': 6,\n",
       " 'humann2': 6,\n",
       " 'hmmer_jackhmmer': 6,\n",
       " 'gemini_inheritance': 6,\n",
       " 'umi_tools_count': 6,\n",
       " 'gmx_npt': 6,\n",
       " 'trinity_define_clusters_by_cutting_tree': 6,\n",
       " 'ctb_chemfp_butina_clustering': 6,\n",
       " 'cshl_fasta_formatter': 6,\n",
       " 'humann2_unpack_pathways': 6,\n",
       " 'lofreq_filter': 7,\n",
       " 'QCCalculator': 7,\n",
       " 'freebayes_wrapper': 7,\n",
       " 'salmon': 7,\n",
       " 'fasta_merge_files_and_filter_unique_sequences': 7,\n",
       " 'vcffixup': 7,\n",
       " 'clustalw': 7,\n",
       " 'bio3d_pca': 7,\n",
       " 'gatk2_depth_of_coverage': 7,\n",
       " 'DecoyDatabase': 7,\n",
       " 'cshl_fastx_reverse_complement': 7,\n",
       " '__APPLY_RULES__': 7,\n",
       " 'bio3d_rmsd': 7,\n",
       " 'rseqc_bam2wig': 7,\n",
       " 'trinity': 7,\n",
       " 'rnaz': 7,\n",
       " 'scanpy_run_tsne': 7,\n",
       " 'bedtools_bamtobed': 7,\n",
       " 'ctb_spectrophore_search': 7,\n",
       " 'docking': 7,\n",
       " 'bio3d_rmsf': 7,\n",
       " 'maldi_quant_peak_detection': 7,\n",
       " 'hicexplorer_hicbuildmatrix': 7,\n",
       " 'qiime_pick_closed_reference_otus': 7,\n",
       " 'pilon': 7,\n",
       " 'Extract_features1': 7,\n",
       " 'valet': 7,\n",
       " 'deeptools_bamFingerprint': 7,\n",
       " 'tp_find_and_replace': 8,\n",
       " 'samtool_filter2': 8,\n",
       " 'snpSift_filter': 8,\n",
       " 'snpSift_annotate': 8,\n",
       " 'deeptools_bamCorrelate': 8,\n",
       " 'ctb_np-likeness-calculator': 8,\n",
       " 'bedtools_intersectbed_bam': 8,\n",
       " 'trimmomatic': 8,\n",
       " 'ctb_alignit': 8,\n",
       " 'fasta_compute_length': 8,\n",
       " 'IDMerger': 8,\n",
       " 't2t_report': 8,\n",
       " 'deeptools_bam_compare': 8,\n",
       " 'transdecoder': 8,\n",
       " 'openms_protein_quantifier': 8,\n",
       " 'deeptools_plot_coverage': 8,\n",
       " 'FalseDiscoveryRate': 8,\n",
       " 'ctb_rdkit_describtors': 8,\n",
       " 'PlasFlow': 8,\n",
       " 'random_lines1': 8,\n",
       " 'ip_overlay_segmentation': 8,\n",
       " 'egsea': 8,\n",
       " 'bwa_mem': 8,\n",
       " 'ctb_stripit': 8,\n",
       " 'cardinal_spectra_plots': 8,\n",
       " 'cshl_cut_tool': 8,\n",
       " '__ZIP_COLLECTION__': 9,\n",
       " 'deeptools_compute_matrix': 9,\n",
       " 'regex_replace': 9,\n",
       " 'ctb_silicos_qed': 9,\n",
       " 'draw_circos': 9,\n",
       " 'unicycler': 9,\n",
       " 'mothur_filter_seqs': 9,\n",
       " 'fraggenescan': 9,\n",
       " 'fastq_to_fasta_python': 9,\n",
       " 'gffcompare': 9,\n",
       " 'samtools_fastx': 9,\n",
       " 'vcfallelicprimitives': 9,\n",
       " 'rseqc_inner_distance': 9,\n",
       " 'deg_annotate': 9,\n",
       " 'tophat2': 9,\n",
       " 'CONVERTER_interval_to_bedstrict_0': 9,\n",
       " 'pipmir': 9,\n",
       " 'rseqc_read_distribution': 9,\n",
       " 'scanpy_find_markers': 9,\n",
       " 'cardinal_segmentations': 9,\n",
       " 'cshl_uniq_tool': 9,\n",
       " 'openbabel_filter': 9,\n",
       " 'gmx_sim': 9,\n",
       " 'glimmer_knowlegde-based': 9}"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_tool_freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of tool sequences: 198459\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-161-459ee921e2f7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mt_seq\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mt_lowest_freq\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m     \u001b[0mp_prec\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc_prec\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcompute_recommendations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mt_seq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdictionary\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreverse_dictionary\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mclass_weights\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtopk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m     \u001b[0mcomplete_p_prec\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp_prec\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m     \u001b[0mcomplete_c_prec\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc_prec\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-119-32812f2c9012>\u001b[0m in \u001b[0;36mcompute_recommendations\u001b[0;34m(model, tool_sequence, labels, dictionary, reverse_dictionary, class_weights, topk, max_seq_len)\u001b[0m\n\u001b[1;32m     81\u001b[0m     \u001b[0mtool_sequence_names\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mreverse_dictionary\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtool_pos\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mtool_pos\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtl_seq_ids\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     82\u001b[0m     \u001b[0;31m# predict next tools for a test path\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 83\u001b[0;31m     \u001b[0mprediction\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_reshaped\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     84\u001b[0m     \u001b[0mnw_dimension\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprediction\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     85\u001b[0m     \u001b[0mprediction\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprediction\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mnw_dimension\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/tool_prediction_gru_wc/lib/python3.6/site-packages/keras/engine/training.py\u001b[0m in \u001b[0;36mpredict\u001b[0;34m(self, x, batch_size, verbose, steps, callbacks, max_queue_size, workers, use_multiprocessing)\u001b[0m\n\u001b[1;32m   1460\u001b[0m                                             \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mverbose\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1461\u001b[0m                                             \u001b[0msteps\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msteps\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1462\u001b[0;31m                                             callbacks=callbacks)\n\u001b[0m\u001b[1;32m   1463\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1464\u001b[0m     def train_on_batch(self, x, y,\n",
      "\u001b[0;32m~/miniconda3/envs/tool_prediction_gru_wc/lib/python3.6/site-packages/keras/engine/training_arrays.py\u001b[0m in \u001b[0;36mpredict_loop\u001b[0;34m(model, f, ins, batch_size, verbose, steps, callbacks)\u001b[0m\n\u001b[1;32m    322\u001b[0m             \u001b[0mbatch_logs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m'batch'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbatch_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'size'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbatch_ids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    323\u001b[0m             \u001b[0mcallbacks\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_batch_hook\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'predict'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'begin'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_logs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 324\u001b[0;31m             \u001b[0mbatch_outs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mins_batch\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    325\u001b[0m             \u001b[0mbatch_outs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mto_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbatch_outs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    326\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mbatch_index\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/tool_prediction_gru_wc/lib/python3.6/site-packages/tensorflow/python/keras/backend.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, inputs)\u001b[0m\n\u001b[1;32m   3074\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3075\u001b[0m     fetched = self._callable_fn(*array_vals,\n\u001b[0;32m-> 3076\u001b[0;31m                                 run_metadata=self.run_metadata)\n\u001b[0m\u001b[1;32m   3077\u001b[0m     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_fetch_callbacks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfetched\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_fetches\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3078\u001b[0m     return nest.pack_sequence_as(self._outputs_structure,\n",
      "\u001b[0;32m~/miniconda3/envs/tool_prediction_gru_wc/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1437\u001b[0m           ret = tf_session.TF_SessionRunCallable(\n\u001b[1;32m   1438\u001b[0m               \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_session\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_handle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstatus\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1439\u001b[0;31m               run_metadata_ptr)\n\u001b[0m\u001b[1;32m   1440\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mrun_metadata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1441\u001b[0m           \u001b[0mproto_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtf_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTF_GetBuffer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrun_metadata_ptr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "topk = 1\n",
    "complete_p_prec = list()\n",
    "complete_c_prec = list()\n",
    "\n",
    "freq = list()\n",
    "print(\"Number of tool sequences: %d\" % len(p_freq))\n",
    "ctr = 0\n",
    "for t_seq in p_freq:\n",
    "    last_tool = t_seq.split(\",\")[-1]\n",
    "    if last_tool in t_lowest_freq.keys():\n",
    "        p_prec, c_prec = compute_recommendations(model, t_seq, \"\", dictionary, reverse_dictionary, class_weights, topk)\n",
    "        complete_p_prec.append(p_prec)\n",
    "        complete_c_prec.append(c_prec)\n",
    "        freq.append(p_freq[t_seq])\n",
    "        ctr+= 1\n",
    "\n",
    "for t_seq in t_lowest_freq:\n",
    "    p_prec, c_prec = compute_recommendations(model, t_seq, \"\", dictionary, reverse_dictionary, class_weights, topk)\n",
    "    complete_p_prec.append(p_prec)\n",
    "    complete_c_prec.append(c_prec)\n",
    "    freq.append(t_lowest_freq[t_seq])\n",
    "    ctr+= 1\n",
    "    \n",
    "print(\"Published precision: %s\" % str(np.nanmean(complete_p_prec)))\n",
    "print(\"Normal precision: %s\" % str(np.nanmean(complete_c_prec)))\n",
    "print(\"Number of paths used : %s\" % ctr)\n",
    "print(\"Mean frequency: %s\" % str(np.mean(freq)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " 0.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " 0.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " 1.0,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " 0.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 0.0,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 1.0,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_p_prec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
