{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tool recommendation \n",
    "## (Gated recurrent units neural network with weighted cross-entropy loss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import json\n",
    "import warnings\n",
    "import operator\n",
    "\n",
    "import h5py\n",
    "from keras.models import model_from_json\n",
    "from keras import backend as K\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "\n",
    "def read_file(file_path):\n",
    "    with open(file_path, 'r') as data_file:\n",
    "        data = json.loads(data_file.read())\n",
    "    return data\n",
    "\n",
    "\n",
    "def create_model(model_path):\n",
    "    reverse_dictionary = dict((str(v), k) for k, v in dictionary.items())\n",
    "    model_weights = list()\n",
    "    weight_ctr = 0\n",
    "    for index, item in enumerate(trained_model.keys()):\n",
    "        if \"weight_\" in item:\n",
    "            d_key = \"weight_\" + str(weight_ctr)\n",
    "            weights = trained_model.get(d_key).value\n",
    "            model_weights.append(weights)\n",
    "            weight_ctr += 1\n",
    "    # set the model weights\n",
    "    loaded_model.set_weights(model_weights)\n",
    "    return loaded_model, dictionary, reverse_dictionary\n",
    "\n",
    "def get_predicted_tools(base_tools, predictions, topk):\n",
    "    \"\"\"\n",
    "    Get predicted tools. If predicted tools are less in number, combine them with published tools\n",
    "    \"\"\"\n",
    "    intersection = list(set(predictions).intersection(set(base_tools)))\n",
    "    print(base_tools)\n",
    "    print(predictions)\n",
    "    print(intersection)\n",
    "    print()\n",
    "    return intersection[:topk]\n",
    "\n",
    "def sort_by_usage(t_list, class_weights, d_dict):\n",
    "    \"\"\"\n",
    "    Sort predictions by usage/class weights\n",
    "    \"\"\"\n",
    "    tool_dict = dict()\n",
    "    for tool in t_list:\n",
    "        t_id = d_dict[tool]\n",
    "        tool_dict[tool] = class_weights[str(t_id)]\n",
    "    tool_dict = dict(sorted(tool_dict.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    return list(tool_dict.keys()), list(tool_dict.values())\n",
    "\n",
    "def separate_predictions(base_tools, predictions, last_tool_name, weight_values, topk):\n",
    "    \"\"\"\n",
    "    Get predictions from published and normal workflows\n",
    "    \"\"\"\n",
    "    last_base_tools = list()\n",
    "    predictions = predictions * weight_values\n",
    "    prediction_pos = np.argsort(predictions, axis=-1)\n",
    "    topk_prediction_pos = prediction_pos[-topk:]\n",
    "    # get tool ids\n",
    "    pred_tool_ids = [reverse_dictionary[str(tool_pos)] for tool_pos in topk_prediction_pos]\n",
    "    if last_tool_name in base_tools:\n",
    "        last_base_tools = base_tools[last_tool_name]\n",
    "        if type(last_base_tools).__name__ == \"str\":\n",
    "            # get published or compatible tools for the last tool in a sequence of tools\n",
    "            last_base_tools = last_base_tools.split(\",\")\n",
    "    # get predicted tools\n",
    "    p_tools = get_predicted_tools(last_base_tools, pred_tool_ids, topk)\n",
    "    sorted_c_t, sorted_c_v = sort_by_usage(p_tools, class_weights, dictionary)\n",
    "    return sorted_c_t, sorted_c_v\n",
    "\n",
    "def compute_recommendations(model, tool_sequence, labels, dictionary, reverse_dictionary, class_weights, topk=10, max_seq_len=25):\n",
    "    tl_seq = tool_sequence.split(\",\")\n",
    "    tl_seq_ids = [str(dictionary[t]) for t in tl_seq] #\",\".join([str(dictionary[t]) for t in tl_seq])\n",
    "    last_tool_name = tl_seq[-1]\n",
    "    sample = np.zeros(max_seq_len)\n",
    "    weight_val = list(class_weights.values())\n",
    "    weight_val = np.reshape(weight_val, (len(weight_val),))\n",
    "    for idx, tool_id in enumerate(tl_seq_ids):\n",
    "        sample[idx] = int(tool_id)\n",
    "    sample_reshaped = np.reshape(sample, (1, max_seq_len))\n",
    "    tool_sequence_names = [reverse_dictionary[str(tool_pos)] for tool_pos in tl_seq_ids]\n",
    "    # predict next tools for a test path\n",
    "    prediction = model.predict(sample_reshaped, verbose=0)\n",
    "    nw_dimension = prediction.shape[1]\n",
    "    prediction = np.reshape(prediction, (nw_dimension,))\n",
    "    \n",
    "    half_len = int(nw_dimension / 2)\n",
    "    \n",
    "    pub_t, pub_v = separate_predictions(standard_connections, prediction[:half_len], last_tool_name, weight_val, topk)\n",
    "    # get recommended tools from normal workflows\n",
    "    c_t, c_v = separate_predictions(compatible_tools, prediction[half_len:], last_tool_name, weight_val, topk)\n",
    "    # combine predictions coming from different workflows\n",
    "    # promote recommended tools coming from published workflows\n",
    "    # to the top and then show other recommendations\n",
    "    print()\n",
    "    tool_seq_name = \",\".join(tool_sequence_names)\n",
    "    print(\"Current tool sequence: \")\n",
    "    print()\n",
    "    print(tool_seq_name)\n",
    "    print()\n",
    "    print(\"Overall recommendations: \")\n",
    "    print()\n",
    "    pub_t.extend(c_t)\n",
    "    pub_v.extend(c_v)\n",
    "    # remove duplicates if any\n",
    "    pub_t = list(dict.fromkeys(pub_t))\n",
    "    pub_v = list(dict.fromkeys(pub_v))\n",
    "    print(pub_t)\n",
    "    ids_tools = dict()\n",
    "    for key in pub_t:\n",
    "        ids_tools[key] = dictionary[key]\n",
    "    print()\n",
    "    print(\"Recommended tool ids:\")\n",
    "    print()\n",
    "    for i in ids_tools:\n",
    "        rev_id = dictionary[i]\n",
    "        wt = class_weights[str(rev_id)]\n",
    "        print(i + \"(\" + str(ids_tools[i]) + \")\" + \"(\" + str(wt) + \")\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unpack trained model for prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_path = \"data/tool_recommendation_model_no_reg.hdf5\"\n",
    "trained_model = h5py.File(model_path, 'r')\n",
    "model_config = json.loads(trained_model.get('model_config').value)\n",
    "dictionary = json.loads(trained_model.get('data_dictionary').value)\n",
    "class_weights = json.loads(trained_model.get('class_weights').value)\n",
    "standard_connections = json.loads(trained_model.get('standard_connections').value)\n",
    "compatible_tools = json.loads(trained_model.get('compatible_tools').value)\n",
    "loaded_model = model_from_json(model_config)\n",
    "model, dictionary, reverse_dictionary = create_model(model_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indices of tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#print(reverse_dictionary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Recommended tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['multiqc', 'umi_tools_dedup', 'bamFilter', 'deeptools_plot_fingerprint', 'deeptools_multi_bam_summary', 'featurecounts', 'rseqc_infer_experiment', 'samtools_stats', 'htseq_count', 'picard_CollectInsertSizeMetrics', 'samtools_flagstat', 'bam_to_sam']\n",
      "['deeptools_plot_fingerprint', 'umi_tools_dedup', 'picard_CollectInsertSizeMetrics', 'samtools_flagstat', 'bamFilter', 'samtools_stats', 'rseqc_infer_experiment', 'htseq_count', 'multiqc', 'featurecounts']\n",
      "['deeptools_plot_fingerprint', 'samtools_flagstat', 'bamFilter', 'picard_CollectInsertSizeMetrics', 'umi_tools_dedup', 'htseq_count', 'multiqc', 'samtools_stats', 'featurecounts', 'rseqc_infer_experiment']\n",
      "\n",
      "['qualimap_rnaseq', 'rseqc_read_duplication', 'cufflinks', 'collection_column_join', 'deeptools_bam_coverage', 'bam_to_sam', 'htseq_count', 'sam_merge2', 'Add_a_column1', 'rseqc_read_distribution', 'tp_cut_tool', 'featurecounts', 'samtools_merge', 'samtools_sort', 'samtools_stats', 'rseqc_infer_experiment', 'picard_SortSam', 'rgPicardMarkDups', 'umi_tools_dedup', 'je_markdupes', 'rseqc_geneBody_coverage', 'multiqc', '__FILTER_FAILED_DATASETS__', 'dexseq_count', 'samtools_rmdup', 'rseqc_bam_stat', 'cuffdiff', 'bamFilter', 'samtools_flagstat', '__RELABEL_FROM_FILE__', 'deeptools_compute_matrix', 'deeptools_multi_bam_summary', 'samtool_filter2', 'picard_CollectInsertSizeMetrics', 'picard_MarkDuplicates', 'merge_pcr_duplicates.py', 'stringtie', 'deseq2', 'deeptools_compute_gc_bias', 'picard_CASM', 'samtools_idxstats', 'rseqc_junction_saturation', 'freebayes', 'deeptools_plot_fingerprint', 'fastqc', 'ngsutils_bam_filter', 'rseqc_junction_annotation']\n",
      "['freebayes', 'tp_cut_tool', 'Add_a_column1', 'deeptools_compute_matrix', 'deseq2', 'stringtie', 'samtools_rmdup', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'fastqc']\n",
      "['deseq2', 'Add_a_column1', 'picard_MarkDuplicates', 'deeptools_bam_coverage', 'deeptools_compute_matrix', 'freebayes', 'samtools_rmdup', 'tp_cut_tool', 'stringtie', 'fastqc']\n",
      "\n",
      "\n",
      "Current tool sequence: \n",
      "\n",
      "rna_star\n",
      "\n",
      "Overall recommendations: \n",
      "\n",
      "['featurecounts', 'multiqc', 'htseq_count', 'rseqc_infer_experiment', 'samtools_stats', 'bamFilter', 'samtools_flagstat', 'picard_CollectInsertSizeMetrics', 'umi_tools_dedup', 'deeptools_plot_fingerprint', 'fastqc', 'deeptools_bam_coverage', 'picard_MarkDuplicates', 'samtools_rmdup', 'stringtie', 'deseq2', 'deeptools_compute_matrix', 'Add_a_column1', 'tp_cut_tool', 'freebayes']\n",
      "\n",
      "Recommended tool ids:\n",
      "\n",
      "featurecounts(1532)(7.958838)\n",
      "multiqc(589)(7.415196)\n",
      "htseq_count(1500)(6.760737)\n",
      "rseqc_infer_experiment(234)(5.963322)\n",
      "samtools_stats(1355)(5.755742)\n",
      "bamFilter(649)(5.551221)\n",
      "samtools_flagstat(303)(5.465816)\n",
      "picard_CollectInsertSizeMetrics(146)(5.111015)\n",
      "umi_tools_dedup(728)(4.698675)\n",
      "deeptools_plot_fingerprint(1109)(4.258446)\n",
      "fastqc(526)(9.16016)\n",
      "deeptools_bam_coverage(226)(6.958448)\n",
      "picard_MarkDuplicates(734)(6.876187)\n",
      "samtools_rmdup(217)(6.865856)\n",
      "stringtie(264)(6.511049)\n",
      "deseq2(1138)(6.417875)\n",
      "deeptools_compute_matrix(706)(6.353451)\n",
      "Add_a_column1(939)(6.215933)\n",
      "tp_cut_tool(135)(6.183317)\n",
      "freebayes(780)(6.142499)\n"
     ]
    }
   ],
   "source": [
    "########### \n",
    "####### Tools from training material\n",
    "# Assembly: \n",
    "# (https://training.galaxyproject.org/training-material/topics/assembly/tutorials/debruijn-graph-assembly/tutorial.html)\n",
    "# (504) Spades -> 'bandage_info', 'fasta-stats', 'bandage_image', 'fasta_filter_by_length', 'abricate', 'quast', 'mlst' ... \n",
    "# (1113) Velveth -> velvetg\n",
    "# (https://training.galaxyproject.org/training-material/topics/assembly/tutorials/unicycler-assembly/tutorial.html)\n",
    "# (35) Unicycler -> 'bandage_info', 'glimmer_build-icm', 'glimmer_knowlegde-based', 'bandage_image', 'transdecoder', 'minimap2', 'antismash', 'fasta_filter_by_length' ...\n",
    "\n",
    "## Computational chemistry\n",
    "# (84) ctb_remDuplicates -> ctb_remIons \n",
    "# (84,1204) ctb_remDuplicates,ctb_remIons -> 'ctb_chemfp_mol2fps', 'ctb_compound_convert'\n",
    "# (84,1204,626) ctb_remDuplicates,ctb_remIons,ctb_chemfp_mol2fps -> 'ctb_chemfp_butina_clustering', 'ctb_simsearch', 'ctb_chemfp_nxn_clustering', 'comp1'\n",
    "# (https://training.galaxyproject.org/training-material/topics/computational-chemistry/tutorials/cheminformatics/tutorial.html)\n",
    "\n",
    "## RAD-seq\n",
    "# (https://training.galaxyproject.org/training-material/topics/ecology/tutorials/ref-based-rad-seq/tutorial.html)\n",
    "# (583) stacks_procrad -> 'bwa', 'bwa_wrapper', 'Grep1', 'stacks_denovomap', 'fastqc', 'fastq_filter'\n",
    "\n",
    "## Epigenetics\n",
    "# (https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/atac-seq/tutorial.html)\n",
    "# (1344,775) cutadapt,bowtie2 => samtools_flagstat', 'picard_MarkDuplicates', 'picard_AddOrReplaceReadGroups', 'macs2_callpeak','bg_sortmerna', 'multiqc', 'hisat2', 'trim_galore', 'bowtie2' ...\n",
    "# (1344,775,292) cutadapt,bowtie2,picard_MarkDuplicates -> 'picard_ReorderSam', 'gatk4_mutect2', 'samtools_rmdup'\n",
    "# (1344,775,292,473) cutadapt,bowtie2,picard_MarkDuplicates,genrich -> 'pygenomeTracks'\n",
    "#\n",
    "# (https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/methylation-seq/tutorial.html)\n",
    "# (639) bwameth -> 'samtools_rmdup', 'samtools_sort', 'bam_to_sam', 'pileometh' ..\n",
    "# (639,1451) bwameth,pileometh -> 'deeptools_compute_matrix', 'tp_sed_tool', 'Filter1', 'metilene', 'Remove beginning1', 'wig_to_bigWig'\n",
    "# ()\n",
    "# (775) bowtie2 -> samtools_flagstat', 'picard_MarkDuplicates', 'picard_AddOrReplaceReadGroups ... \n",
    "# (775,1231) bowtie2,deeptools_multi_bam_summary -> 'deeptools_plot_pca', 'r_correlation_matrix' ...\n",
    "# (https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/formation_of_super-structures_on_xi/tutorial.html)\n",
    "\n",
    "# (775,170) bowtie2,hicexplorer_hicbuildmatrix -> 'hicexplorer_hicsummatrices', 'hicexplorer_hicplotviewpoint', 'tp_sed_tool', 'hicexplorer_hiccorrectmatrix', 'hicexplorer_hicmergematrixbins', 'hicexplorer_hicpca' ..\n",
    "# (775,170,669) bowtie2,hicexplorer_hicbuildmatrix,hicexplorer_hicmergematrixbins -> 'hicexplorer_hiccorrectmatrix', 'hicexplorer_hicplottads', 'hicexplorer_hicplotmatrix'\n",
    "# (https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/hicexplorer/tutorial.html)\n",
    "\n",
    "# (232) minfi_read450k -> 'minfi_getbeta'\n",
    "\n",
    "\n",
    "## Genome annotation\n",
    "# (https://training.galaxyproject.org/training-material/topics/genome-annotation/tutorials/annotation-with-maker/tutorial.html)\n",
    "# (1157) maker -> 'gffread', 'maker_map_ids', 'jcvi_gff_stats'\n",
    "# (https://training.galaxyproject.org/training-material/topics/genome-annotation/tutorials/annotation-with-prokka/tutorial.html)\n",
    "# (94) prokka -> 'mlst', 'jbrowse', 'taxonomy_krona_chart' ...\n",
    "\n",
    "\n",
    "## Imaging\n",
    "# (https://training.galaxyproject.org/training-material/topics/imaging/tutorials/hela-screen-analysis/tutorial.html)\n",
    "# (415) ip_filter_standard -> 'ip_histogram_equalization', 'ip_threshold', 'ip_count_objects\n",
    "# (415,309) ip_filter_standard,ip_threshold -> ip_binary_to_labelimage', 'ip_2d_split_binaryimage_by_watershed', 'ip_count_objects', 'ip_convertimage'\n",
    "# (415,309,992) ip_filter_standard,ip_threshold,ip_2d_split_binaryimage_by_watershed -> 'ip_2d_filter_segmentation_by_features', 'ip_2d_feature_extraction'\n",
    "\n",
    "\n",
    "## Mass spectrometry\n",
    "# (630) mass_spectrometry_imaging_preprocessing -> 'mass_spectrometry_imaging_combine', 'mass_spectrometry_imaging_preprocessing'\n",
    "# (630,1386) mass_spectrometry_imaging_preprocessing,mass_spectrometry_imaging_combine -> 'maldi_quant_preprocessing', 'mass_spectrometry_imaging_preprocessing', 'mass_spectrometry_imaging_qc' ...\n",
    "# (711) search_gui -> peptide_shaker\n",
    "# (711,1411) search_gui,peptide_shaker -> 'mz_to_sqlite', 'Remove beginning1', 'tp_replace_in_column', 'unipept', proteomics_moff ...\n",
    "\n",
    "\n",
    "## Single cell\n",
    "# (531) raceid_main -> 'seurat','raceid_trajectory'\n",
    "# (531,691) raceid_main,raceid_trajectory -> 'raceid_inspecttrajectory'\n",
    "# (141,685) raceid_inspectclusters,__BUILD_LIST__ -> 'picard_MarkDuplicates', 'hisat2', 'stringtie_merge', 'cutadapt', 'tp_cat'\n",
    "# (89) raceid_filtnormconf -> 'raceid_clustering', '__BUILD_LIST__'\n",
    "# (89,270) raceid_filtnormconf,raceid_clustering -> 'raceid_trajectory', 'raceid_inspectclusters'\n",
    "# (638) scanpy_regress_variable -> scanpy_scale_data \n",
    "# (638,739) scanpy_regress_variable,scanpy_scale_data -> 'scanpy_run_pca', 'scanpy_find_variable_genes'\n",
    "# (638,739,505) scanpy_regress_variable,scanpy_scale_data,scanpy_run_pca -> 'scanpy_compute_graph', 'scanpy_plot', 'scanpy_run_tsne', 'scanpy_plot_embed'\n",
    "\n",
    "# (https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/scrna-preprocessing-tenx/tutorial.html)\n",
    "# (728) rna_starsolo -> 'dropletutils', 'multiqc'\n",
    "# (728,306) rna_starsolo,dropletutils -> 'scanpy_read_10x', 'scanpy_cluster_reduce_dimension', 'seurat_read10x', 'anndata_import', 'scanpy_plot', 'raceid_filtnormconf'\n",
    "\n",
    "\n",
    "## Variant calling\n",
    "# (https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/microbial-variants/tutorial.html)\n",
    "# (1076) snippy -> 'bedtools_intersectbed', 'vcfvcfintersect', 'Remove beginning1', 'snippy_core', 'qualimap_bamqc', 'freebayes', 'jbrowse', 'vcfcombine'\n",
    "\n",
    "# https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/somatic-variants/tutorial.html\n",
    "# (57,712,991,629) trimmomatic,bwa_mem,samtools_rmdup,bamleftalign -> 'samtool_filter2', 'ivar_variants', 'freebayes', 'varscan_somatic', 'deeptools_bam_coverage', 'fastqc', 'ngsutils_bam_filter', 'bamFilter', 'rgPicFixMate', 'samtools_calmd'\n",
    "# (57,712,991,629,273) trimmomatic,bwa_mem,samtools_rmdup,bamleftalign,varscan_somatic -> 'bcftools_norm', 'gemini_annotate', 'vt_normalize', 'vcffilter2', 'vcfallelicprimitives', 'snpEff'\n",
    "\n",
    "# (https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/dip/tutorial.html)\n",
    "# (988) freebayes -> 'vcfallelicprimitives', 'bcftools_norm', 'custom_pro_db', 'vcfvcfintersect'\n",
    "# (988,436) freebayes,vcfallelicprimitives -> 'vt_normalize', 'snpSift_filter', 'snpSift_annotate', 'snpEff'\n",
    "# (988,436,32) freebayes,vcfallelicprimitives,snpEff -> 'gemini_load', 'mimodd_varreport', 'snpSift_extractFields\n",
    "\n",
    "\n",
    "# Transcriptomics\n",
    "#(https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/small_ncrna_clustering/tutorial.html)\n",
    "# (1459,1473,1522) samtools_sort,blockclust,sort1 -> 'cshl_awk_tool', 'Show beginning1', 'tp_awk_tool', 'blockbuster'\n",
    "\n",
    "# (https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/ref-based/tutorial.html)\n",
    "# (1344) cutadapt -> 'umi_tools_extract', 'fastq_paired_end_interlacer', 'chira_collapse', 'rna_star',\n",
    "# (1344,484) cutadapt,rna_star -> 'featurecounts', 'multiqc', 'htseq_count', 'rseqc_infer_experiment', 'samtools_stats', 'bamFilter',\n",
    "# (1344,484,1209) cutadapt,rna_star,featurecounts -> 'multiqc', 'deseq2', 'tp_sort_header_tool', 'bamFilter', 'collection_column_join ...\n",
    "\n",
    "# Single-cell HiC\n",
    "# 508 schicexplorer_schicqualitycontrol -> 'schicexplorer_schicnormalize'\n",
    "# schicexplorer_schicnormalize -> 'schicexplorer_schicclustersvl', 'schicexplorer_schicconsensusmatrices', 'schicexplorer_schicplotclusterprofiles'\n",
    "# (13,743) schicexplorer_schicnormalize,schicexplorer_schicclustersvl -> 'schicexplorer_schicplotclusterprofiles', 'schicexplorer_schicconsensusmatrices'\n",
    "\n",
    "\n",
    "# Animal detection on acoustic recording\n",
    "# (1224) vigiechiro_idvalid -> 'vigiechiro_bilanenrichipf', 'vigiechiro_bilanenrichirp'\n",
    "\n",
    "topk = 10 # set the maximum number of recommendations \n",
    "\n",
    "tool_seq = \"rna_star\" # give tools ids in a sequence and see the recommendations. To know all the tool ids, \n",
    "                     # please print the variable 'reverse_dictionary'\n",
    "compute_recommendations(model, tool_seq, \"\", dictionary, reverse_dictionary, class_weights, topk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'fastqc': 9.16016, 'CONVERTER_gz_to_uncompressed': 7.994104, 'featurecounts': 7.958838, 'trimmomatic': 7.873924, 'rna_star': 7.677528, 'bowtie2': 7.67546, 'sklearn_searchcv': 7.634176, 'multiqc': 7.415196, 'cshl_fastx_collapser': 7.35617, 'hisat2': 7.322985, 'trim_galore': 7.306017, 'minimap2': 7.294264, 'bwa_mem': 7.163771, 'fastq_to_fasta_python': 7.144794, 'Filter1': 7.142097, 'Cut1': 7.094403, 'cutadapt': 7.086727, 'deeptools_bam_coverage': 6.958448, 'fasta_filter_by_length': 6.918154, 'picard_MarkDuplicates': 6.876187, 'bedtools_intersectbed': 6.870884, 'samtools_rmdup': 6.865856, 'samtools_slice_bam': 6.781136, 'htseq_count': 6.760737, 'fastq_groomer': 6.644561, 'prokka': 6.636952, 'stringtie': 6.511049, 'deseq2': 6.417875, 'unicycler': 6.413295, 'infernal_cmbuild': 6.412702, 'infernal_cmsearch': 6.412702, 'cmFinder': 6.406118, 'locarna_best_subtree': 6.401151, 'deeptools_plot_heatmap': 6.385925, 'cat1': 6.370149, 'macs2_callpeak': 6.36366, 'abricate': 6.356281, 'deeptools_compute_matrix': 6.353451, 'join1': 6.335054, 'Add_a_column1': 6.215933, 'Grep1': 6.204356, 'fastq_dump': 6.199123, 'tp_cut_tool': 6.183317, 'tp_easyjoin_tool': 6.144343, 'freebayes': 6.142499, 'tp_sort_header_tool': 6.116251, 'deeptools_plot_profile': 6.110254, 'rseqc_read_distribution': 6.059884, 'kraken2': 6.054479, 'ncbi_blastp_wrapper': 6.000393, 'rseqc_infer_experiment': 5.963322, 'rseqc_geneBody_coverage': 5.95804, 'samtool_filter2': 5.952984, 'spades': 5.940434, 'volcanoplot': 5.930984, 'snippy': 5.916382, 'salmon': 5.901188, 'ggplot2_heatmap2': 5.874623, 'tab2fasta': 5.843448, 'quast': 5.8097, 'bandage_image': 5.803295, 'fastp': 5.801364, 'addValue': 5.766569, 'mothur_taxonomy_to_krona': 5.757013, 'samtools_stats': 5.755742, 'samtools_idxstats': 5.74789, 'table_compute': 5.745157, 'ngsutils_bam_filter': 5.724483, 'Remove beginning1': 5.720349, 'nanoplot': 5.718104, 'edger': 5.706279, 'wig_to_bigWig': 5.661528, 'mothur_summary_seqs': 5.632203, 'annotatemyids': 5.620401, 'kallisto_quant': 5.611332, 'mothur_screen_seqs': 5.603331, 'fasta2tab': 5.595269, 'sam_merge2': 5.58312, 'Grouping1': 5.5827, 'samtools_sort': 5.581364, 'bamFilter': 5.551221, 'cardinal_preprocessing': 5.550021, 'CONVERTER_interval_to_bedstrict_0': 5.53799, 'comp1': 5.531499, 'tp_grep_tool': 5.531322, 'query_tabular': 5.493187, 'filter_by_fasta_ids': 5.465894, 'samtools_flagstat': 5.465816, 'tp_cat': 5.441543, 'tophat2': 5.436339, 'porechop': 5.428907, 'gemini_query': 5.413633, 'ncbi_blastn_wrapper': 5.394081, 'bwa': 5.391579, 'tp_replace_in_column': 5.380343, 'pygenomeTracks': 5.367407, 'vcf2tsv': 5.364255, 'snpEff': 5.352332, 'racon': 5.350707, 'tp_awk_tool': 5.338109, 'bamleftalign': 5.333399, 'mothur_unique_seqs': 5.332744, 'sort1': 5.330623, 'miniasm': 5.316143, 'bamtools': 5.315959, 'mothur_align_seqs': 5.302466, 'varscan_mpileup': 5.294697, 'canu': 5.2901, 'trinity': 5.289132, 'fastq_filter': 5.280662, 'deg_annotate': 5.27409, 'CONVERTER_interval_to_bed_0': 5.262224, 'mothur_classify_seqs': 5.245038, 'fastq_join': 5.240384, 'bcftools_norm': 5.236423, 'deeptools_bam_compare': 5.228967, 'gops_join_1': 5.206476, 'goseq': 5.203182, 'samtools_mpileup': 5.183468, 'cshl_fastq_to_fasta': 5.181543, 'hicexplorer_hicplotmatrix': 5.170902, 'shovill': 5.167846, 'msconvert': 5.145305, 'gfa_to_fa': 5.145121, 'picard_CollectInsertSizeMetrics': 5.111015, 'split_file_to_collection': 5.106003, 'vcffilter2': 5.104669, 'staramr_search': 5.104284, 'limma_voom': 5.075231, 'collapse_dataset': 5.053056, 'bedtools_getfastabed': 5.050817, 'bedtools_mergebed': 5.036303, 'genrich': 5.035115, 'metaphlan2': 5.026759, 'bowtie_wrapper': 5.012338, 'cardinal_filtering': 5.005511, 'fastq_paired_end_joiner': 4.992294, 'Extract genomic DNA 1': 4.986901, 'chipseeker': 4.980473, 'cshl_fastq_quality_filter': 4.976767, 'abims_xcms_xcmsSet': 4.952244, 'tp_multijoin_tool': 4.938216, 'pileometh': 4.936673, 'mothur_cluster_split': 4.918155, 'mothur_count_seqs': 4.917162, 'roary': 4.910338, 'cardinal_segmentations': 4.899828, 'Convert characters1': 4.894894, 'fastq_paired_end_interlacer': 4.892519, 'mlst': 4.889782, 'umi_tools_extract': 4.883372, 'mergeCols1': 4.881315, 'mothur_classify_otu': 4.875107, 'trinity_abundance_estimates_to_matrix': 4.867471, 'PlasFlow': 4.845648, 'lastz_wrapper_2': 4.832078, 'jbrowse': 4.829735, 'csv_to_tabular': 4.807134, 'fasta-stats': 4.80238, 'CONVERTER_bam_to_bigwig_0': 4.801559, 'metilene': 4.796609, 'collection_column_join': 4.787815, 'dexseq_count': 4.781833, 'busco': 4.775225, 'cardinal_data_exporter': 4.759321, 'mothur_pre_cluster': 4.748078, 'cufflinks': 4.73792, 'taxonomy_krona_chart': 4.735321, 'tp_replace_in_line': 4.730387, 'mothur_filter_seqs': 4.727592, 'gemini_load': 4.727307, 'newick_display': 4.716643, 'CONVERTER_bam_to_coodinate_sorted_bam': 4.710515, 'tp_find_and_replace': 4.699571, 'umi_tools_dedup': 4.698675, 'gtftobed12': 4.697055, 'mothur_make_contigs': 4.675673, 'bedtools_sortbed': 4.669544, 'bedtools_bamtobed': 4.649848, 'openbabel_compound_convert': 4.646458, 'sam_to_bam': 4.632434, 'vcfallelicprimitives': 4.627501, 'cshl_fastx_artifacts_filter': 4.599638, 'trinity_align_and_estimate_abundance': 4.599008, 'tp_tail_tool': 4.597667, 'interproscan': 4.593249, 'tabular_to_csv': 4.574893, 'CONVERTER_fasta_to_tabular': 4.568852, 'datamash_ops': 4.568506, 'bedtools_subtractbed': 4.559542, 'Show beginning1': 4.556282, 'sklearn_build_pipeline': 4.552824, 'blast_reciprocal_best_hits': 4.541804, 'clustalw': 4.540837, 'deeptools_plot_correlation': 4.527101, 'bedtools_coveragebed': 4.522882, 'flye': 4.519893, 'gops_intersect_1': 4.508683, 'mothur_sub_sample': 4.494239, 'mothur_make_biom': 4.480776, 'scanpy_plot': 4.479543, 'maker': 4.471825, 'cshl_fastx_trimmer': 4.458602, 'mothur_chimera_vsearch': 4.451065, 'fasta_merge_files_and_filter_unique_sequences': 4.449685, 'tp_head_tool': 4.447531, 'bcftools_consensus': 4.435258, 'charts': 4.433591, 'bg_sortmerna': 4.429259, 'sklearn_estimator_attributes': 4.426102, 'filtlong': 4.417635, 'search_gui': 4.415712, 'scanpy_inspect': 4.411489, 'augustus': 4.402588, 'samtools_depth': 4.395502, 'bwameth': 4.391925, 'regex1': 4.383834, 'bedtools_map': 4.367978, 'idpquery': 4.367296, 'humann2': 4.365874, 'peptide_shaker': 4.35201, 'gffread': 4.346399, 'meme_chip': 4.335047, 'velvetoptimiser': 4.311821, 'snpSift_extractFields': 4.311316, 'mothur_tree_shared': 4.298947, 'sklearn_generalized_linear': 4.296163, 'poretools_extract': 4.292913, 'CONVERTER_bed_gff_or_vcf_to_bigwig_0': 4.287563, 'mothur_remove_seqs': 4.280824, 'ncbi_makeblastdb': 4.275276, 'snpEff_build_gb': 4.266028, 'fastq_paired_end_deinterlacer': 4.262328, 'bigwig_to_bedgraph': 4.258522, 'deeptools_plot_fingerprint': 4.258446, 'je_markdupes': 4.256121, 'bp_genbank2gff3': 4.252167, 'bam_to_sam': 4.250531, 'megahit': 4.243759, 'goenrichment': 4.243143, 'viennarna_rnafold': 4.232596, 'anndata_manipulate': 4.218889, 'ncbi_tblastn_wrapper': 4.214829, 'scanpy_filter': 4.208064, 'bg_diamond': 4.203716, 'dropletutils': 4.203023, 'diffbind': 4.198705, 'trinity_run_de_analysis': 4.190343, 'trimmer': 4.185606, 'scanpy_cluster_reduce_dimension': 4.18274, 'addName': 4.178601, 'cummeRbund': 4.178116, 'pathview': 4.171485, 'gemini_annotate': 4.171113, 'gemini_inheritance': 4.147719, 'rbc_mafft': 4.144348, 'deeptools_multi_bam_summary': 4.142597, 'cshl_fastx_reverse_complement': 4.138899, 'varscan_somatic': 4.131961, 'metaspades': 4.129048, 'stringtie_merge': 4.12552, 'bedtools_windowbed': 4.123904, 'sam_pileup': 4.120894, 'anndata_inspect': 4.118292, 'je_clip': 4.111842, 'gmx_setup': 4.110499, 'velvetg': 4.091391, 'gffcompare': 4.073575, 'format_metaphlan2_output': 4.069898, 'CONVERTER_gff_to_bed_0': 4.06436, 'cardinal_combine': 4.054108, 'sklearn_feature_selection': 4.034959, 'fasta_compute_length': 4.030497, 'mothur_rarefaction_single': 4.030102, 'stacks2_populations': 4.017642, 'CONVERTER_bedgraph_to_bigwig': 4.007001, 'fastq_trimmer': 3.998119, 'fastq_stats': 3.994349, 'fastq_manipulation': 3.989964, 'tp_sed_tool': 3.986666, 'cardinal_quality_report': 3.975318, 'scatterplot_rpy': 3.971149, 'hicexplorer_hicbuildmatrix': 3.949669, 'CONVERTER_interval_to_bed12_0': 3.944732, 'fastq_quality_trimmer': 3.93645, 'sklearn_ensemble': 3.934318, 'bedtools_multiintersectbed': 3.933288, 'abyss-pe': 3.914441, 'velveth': 3.906779, 'CONVERTER_fasta_to_fai': 3.904279, 'bedtools_genomecoveragebed_bedgraph': 3.89751, 'mass_spectrometry_imaging_filtering': 3.892716, 'bedtools_genomecoveragebed': 3.882764, 'fgsea': 3.879822, 'bismark_bowtie2': 3.876678, 'docking': 3.872189, 'pilon': 3.869116, 'augustus_training': 3.858349, '__BUILD_LIST__': 3.851602, 'cardinal_mz_images': 3.84962, 'ChangeCase': 3.831527, 'iqtree': 3.826273, 'rnaspades': 3.823972, 'ctb_chemfp_mol2fps': 3.82227, 'prinseq': 3.820553, 'bedtools_bedtobam': 3.815574, 'ncbi_blastx_wrapper': 3.807526, 'deeptools_plot_pca': 3.803106, 'mothur_make_shared': 3.782855, 'qiime_pick_rep_set': 3.773397, 'mothur_count_groups': 3.767952, 'deeptools_alignmentsieve': 3.7621, 'mothur_summary_single': 3.7612, 'mothur_dist_shared': 3.753027, 'deeptools_bigwig_compare': 3.749391, 'segemehl': 3.744305, 'deeptools_multi_bigwig_summary': 3.727391, 'XY_Plot_1': 3.726358, 'mothur_heatmap_sim': 3.725007, 'cuffmerge': 3.719982, 'naive_variant_caller': 3.715694, 'metaphlan2krona': 3.71548, 'unipept': 3.710505, 'venn_list': 3.705682, 'humann2_renorm_table': 3.699552, 'raceid_filtnormconf': 3.684552, 'gmx_solvate': 3.665392, 'ggplot2_pca': 3.664262, 'cardinal_classification': 3.663562, 'fasttree': 3.662621, 'cuffdiff': 3.661778, 'tabular_to_fastq': 3.657561, 'bedtools_groupbybed': 3.653252, 'egsea': 3.648925, 'vigiechiro_bilanenrichipf': 3.645563, 'CONVERTER_bed_to_gff_0': 3.643957, 'get_pdb': 3.63557, 'snippy_core': 3.634951, 'vcfvcfintersect': 3.621671, 'vigiechiro_idvalid': 3.618123, 'vigiechiro_idcorrect_2ndlayer': 3.618123, 'exonerate': 3.615785, 'snpSift_filter': 3.608032, 'bedtools_multicovtbed': 3.599189, 'varscan': 3.598002, 'gdal_ogr2ogr': 3.583284, 'picard_NormalizeFasta': 3.5778, 'stacks2_denovomap': 3.576059, 'dbbuilder': 3.576035, 'datamash_transpose': 3.573823, 'pileup_interval': 3.571975, 'sailfish': 3.570401, 'anndata_import': 3.570328, 'gops_coverage_1': 3.568962, 'hicexplorer_hicfindtads': 3.566404, 'stacks_denovomap': 3.56361, 'bg_find_subsequences': 3.558743, 'unzip': 3.552841, 'qiime_assign_taxonomy': 3.552487, 'antismash': 3.548659, 'picard_SortSam': 3.544576, 'rna_starsolo': 3.541405, 'snpEff_download': 3.535492, 'pileup_parser': 3.531239, 'meme_meme': 3.521066, 'glob_report': 3.520171, 'nspdk_sparse': 3.519433, 'picard_CASM': 3.51468, 'preproc': 3.50893, 'mummer_mummer': 3.497114, 'preMloc': 3.494298, 'NSPDK_candidateClust': 3.494298, 'qiime_make_phylogeny': 3.491299, 'rseqc_bam2wig': 3.478289, 'sklearn_stacking_ensemble_models': 3.477358, 'combine_metaphlan2_humann2': 3.472424, 'gmx_md': 3.472298, 'gemini_db_info': 3.469237, 'progressivemauve': 3.468104, 'gmx_npt': 3.465607, 'signalp3': 3.460565, 'sqlite_to_tabular': 3.460483, 'flash': 3.452102, 'picard_FilterSamReads': 3.452006, 'prepare_ligands_for_docking': 3.427074, 'dexseq': 3.42388, 'gmx_nvt': 3.411331, 'umi_tools_count': 3.390514, 'qiime_make_otu_heatmap': 3.373434, 'sklearn_svm_classifier': 3.372923, 'structure_to_gspan': 3.357768, 'bandage_info': 3.356897, 'qiime_summarize_taxa_through_plots': 3.354915, 'join_files_on_column_fuzzy': 3.346389, 'bedtools_shufflebed': 3.343948, 'keras_model_config': 3.332562, 'mummer_nucmer': 3.329612, 'mothur_remove_lineage': 3.326594, 'umi_tools_group': 3.322894, 'samtools_filter': 3.321248, 'mira_assembler': 3.312201, 'hicexplorer_hicplottads': 3.307451, 'gmx_em': 3.306887, '__UNZIP_COLLECTION__': 3.304065, 'picard_AddOrReplaceReadGroups': 3.299534, 'mothur_dist_seqs': 3.297183, 'IDPosteriorErrorProbability': 3.289278, 'raceid_clustering': 3.279657, 'meme_fimo': 3.279209, 'je_demultiplex': 3.277605, 'flaimapper': 3.272014, 'fastq_combiner': 3.269068, 'wolf_psort': 3.26844, 'sklearn_nn_classifier': 3.264318, 'wtdbg': 3.263821, 'seq_filter_by_id': 3.263556, 'jcvi_gff_stats': 3.260402, 'graphlan': 3.26002, 'mummer_dnadiff': 3.258749, 'raceid_inspectclusters': 3.257366, 'sed_stream_editor': 3.252548, 'peakcalling_macs': 3.246379, 'tmhmm2': 3.245963, 'vsearch_sorting': 3.242035, 'umi_tools_whitelist': 3.238679, 'tp_uniq_tool': 3.237557, 'cshl_fasta_nucleotides_changer': 3.234379, 'bedtools_bamtofastq': 3.233979, 'mothur_trim_seqs': 3.228347, 'gd_raxml': 3.22123, 'maldi_quant_preprocessing': 3.214552, 'maxquant': 3.20278, 'hicexplorer_hiccorrectmatrix': 3.202393, 'kallisto_pseudo': 3.201007, 'EMBOSS: transeq101': 3.190092, 'seqtk_mergefa': 3.182212, 'mothur_remove_groups': 3.178433, 'samtools_bedcov': 3.172401, 'peakachu': 3.167949, 'rbc_mirdeep2_mapper': 3.166178, 'picard_CollectRnaSeqMetrics': 3.165017, 'filter_tabular': 3.16454, 'muscle': 3.162853, 'vcfcombine': 3.162402, 'graphicsmagick_image_convert': 3.161905, 'mothur_venn': 3.149407, 'peakcalling_sicer': 3.144383, 'raceid_inspecttrajectory': 3.141654, 'picard_SamToFastq': 3.141189, 'bedtools_annotatebed': 3.13449, 'vsearch_clustering': 3.133625, 'secure_hash_message_digest': 3.129769, 'export2graphlan': 3.126761, 'subtract_query1': 3.12365, 'picard_MergeSamFiles': 3.118826, 'mothur_cluster': 3.116474, 'gops_cluster_1': 3.112111, 'qiime_pick_closed_reference_otus': 3.10968, 'qiime_make_otu_table': 3.109467, 'chembl': 3.109012, 'aragorn_trna': 3.108798, 'IDScoreSwitcher': 3.10818, 'nanopolish_variants': 3.107843, 'transdecoder': 3.105798, 'gff2bed1': 3.101593, 'trinity_filter_low_expr_transcripts': 3.099847, 'bg_uniq': 3.098152, 'qiime_pick_otus': 3.09318, 'histogram_rpy': 3.0909, 'bg_column_arrange_by_header': 3.090629, 'prepare_receptor': 3.09021, 'cshl_fastx_barcode_splitter': 3.088167, 'hmmer_hmmscan': 3.086997, 'gops_basecoverage_1': 3.081271, 'mothur_get_groups': 3.060494, 'sklearn_train_test_eval': 3.031998, 'mothur_unifrac_weighted': 3.019027, 'rbc_rnacode': 3.017851, 'selectsequencesfrommsa': 3.017681, 'FidoAdapter': 3.014685, 'fraggenescan': 3.014141, 'stacks_refmap': 3.008279, 'prepare_box': 3.005862, 'varscan_copynumber': 3.003017, 'cardinal_spectra_plots': 2.960459, 'maldi_quant_peak_detection': 2.960217, 'ggplot2_histogram': 2.960105, 'EMBOSS: geecee41': 2.959011, 'samtools_calmd': 2.958468, 'blastxml_to_top_descr': 2.958253, 'pureclip': 2.954437, 'mothur_make_group': 2.953721, 'PicardInsertSize': 2.952969, 'stacks_procrad': 2.950519, 'graphprot_predict_profile': 2.949688, 'datamash_reverse': 2.941791, 'cd_hit_est': 2.940988, 'metaphlan_hclust_heatmap': 2.932405, 'uniprotxml_downloader': 2.928524, 'flanking_features_1': 2.928523, 'ruvseq': 2.9188, 'hicexplorer_hicpca': 2.906758, 'vsearch_dereplication': 2.903499, 'plotly_ml_performance_plots': 2.892838, 'jq': 2.889509, 'seq_filter_by_mapping': 2.881604, 'cshl_fastx_renamer': 2.880586, 'fasplit': 2.877252, 'cshl_fasta_formatter': 2.869683, 'bctools_extract_alignment_ends': 2.860738, 'md_converter': 2.856912, 'Paste1': 2.85647, 'gffcompare_to_bed': 2.854553, 'w4mclassfilter': 2.845357, '__FLATTEN__': 2.843103, 'random_lines1': 2.841647, 'mothur_heatmap_bin': 2.825443, 'humann2_unpack_pathways': 2.82255, 'abims_CAMERA_annotateDiffreport': 2.821063, 'ggplot2_point': 2.820296, 'blockclust': 2.819393, 'pick_open_reference_otus': 2.817596, 'translate_bed': 2.809101, 'qiime_align_seqs': 2.804635, 'piranha': 2.80336, 'bedtools_slopbed': 2.803094, 'megablast_wrapper': 2.797448, 'ggplot2_heatmap': 2.790146, 'bedtools_unionbedgraph': 2.788093, 'qiime_pick_open_reference_otus': 2.775709, 'mothur_get_seqs': 2.775068, 'deeptools_plot_coverage': 2.774541, 'group_humann2_uniref_abundances_to_go': 2.771656, 'snpSift_int': 2.769392, 'IDMerger': 2.757045, 'hmmer_hmmbuild': 2.750399, 'gops_merge_1': 2.745893, 'heinz_scoring': 2.744808, 'picard_ARRG': 2.736998, 'Nucleosome': 2.734596, 'scpipe': 2.734546, 'rawtools': 2.730847, 'gops_subtract_1': 2.727012, 'xpath': 2.725979, 'graphclust_align_cluster': 2.725437, '__MERGE_COLLECTION__': 2.725241, 'mimodd_info': 2.720638, 'trinity_analyze_diff_expr': 2.717756, 'tp_sorted_uniq': 2.717628, 'idpqonvert': 2.716529, 'flashlfq': 2.713681, 'nn_classifier': 2.711902, 'mimodd_convert': 2.679283, 'rsem_prepare_reference': 2.676695, 'seurat': 2.674149, 'cds_essential_variability': 2.672705, 'cummerbund_to_cuffdiff': 2.668786, 'mothur_merge_files': 2.668722, 'mummer_mummerplot': 2.661336, 'vsearch_search': 2.657044, 'column_remove_by_header': 2.653242, 'raceid_trajectory': 2.650107, 'mothur_seq_error': 2.642896, 'phyml': 2.640334, 'graphlan_annotate': 2.639759, 'seqtk_subseq': 2.629634, 'gemini_stats': 2.627576, 'bio3d_rmsd': 2.615924, 'chipeakanno_annopeaks': 2.615774, 'paralyzer': 2.608379, 'FileFilter': 2.603472, 'svm_classifier': 2.597288, 'mimodd_varcall': 2.595255, 'deeptools_compute_matrix_operations': 2.594358, 'hicexplorer_hicmergematrixbins': 2.593217, 'trinity_contig_exn50_statistic': 2.592021, 'TextExporter': 2.588255, 'crosscontamination_barcode_filter': 2.587764, 'hicexplorer_hicplotviewpoint': 2.584593, 'uniprot': 2.582785, 'blast2go': 2.574307, 'tp_text_file_with_recurring_lines': 2.569718, 'scanpy_remove_confounders': 2.566831, 'collection_element_identifiers': 2.566372, 'rseqc_bam_stat': 2.556145, 'fastq_to_tabular': 2.553432, 'rbc_mirdeep2_quantifier': 2.547881, 'valet': 2.54553, 'spocc_occ': 2.544223, 'mothur_get_oturep': 2.537298, 'bismark_methylation_extractor': 2.535783, 'cuffcompare': 2.53013, 'krona-text': 2.527329, 'bcftools_call': 2.527258, 'Show tail1': 2.52424, 'trinity_super_transcripts': 2.517696, 'bio3d_pca_visualize': 2.513351, 'fasta_concatenate0': 2.511076, 'EMBOSS: getorf42': 2.493205, 'scater_filter': 2.485923, 'mothur_corr_axes': 2.485752, 'vcfsort': 2.477399, 'mothur_metastats': 2.473336, 'picard_BamIndexStats': 2.465271, 'tb_profiler_profile': 2.463527, 'vcfbreakcreatemulti': 2.460514, 'mdanalysis_cosine_analysis': 2.458205, 'bedtools_complementbed': 2.45612, 'ip_landmark_registration': 2.454397, 'blast_parser': 2.454173, 'trnascan': 2.451005, 'bio3d_pca': 2.443725, 'gemini_burden': 2.441594, 'mimodd_map': 2.440606, 'PicardASMetrics': 2.440606, 'mothur_lefse': 2.438678, 'fastq_paired_end_splitter': 2.428519, 'macs2_predictd': 2.416708, 'sam2interval': 2.416176, 'ctb_online_data_fetch': 2.413476, 'snpsift_vartype': 2.407534, 'cd_hit_protein': 2.406945, 'bio3d_rmsf': 2.4047, 'enasearch_search_data': 2.36556, 'ssake': 2.347083, 'hca_matrix_downloader': 2.345911, 'trim_reads': 2.341536, 'mimodd_vcf_filter': 2.340473, 'tp_tac': 2.339206, 'percolator_input_converters': 2.339206, 'humann2_genefamilies_genus_level': 2.331932, 'basil': 2.326248, 'maker_map_ids': 2.320452, 'myrimatch': 2.319946, 'filter_bed_on_splice_junctions': 2.319816, 'allele_counts_1': 2.313614, 'resize_coordinate_window': 2.308688, 'regexColumn1': 2.30772, 'ete_homology_classifier': 2.306476, 'mothur_summary_tax': 2.305245, 'ete_genetree_splitter': 2.304939, 'mothur_classify_rf': 2.304895, 'meme_dreme': 2.298022, 'infernal_cmstat': 2.296697, 'smf_utils_fix-fasta-headers': 2.296589, 'Count1': 2.280384, 'ococo': 2.27321, 'snpeff_to_peptides': 2.265351, 'gemini_interactions': 2.261763, 'deeptools_bam_pe_fragmentsize': 2.259018, 'ProteinQuantifier': 2.253286, 'heinz': 2.248538, 'mothur_get_relabund': 2.236751, 'bedtools_bed12tobed6': 2.214526, 'tp_split_on_column': 2.208262, 'heinz_visualization': 2.198562, 'smf_utils_estimate-energy': 2.197249, 'mimodd_reheader': 2.197249, 'rseqc_FPKM_count': 2.195094, 'openbabel_remIons': 2.194962, '__FILTER_FROM_FILE__': 2.194046, '__ZIP_COLLECTION__': 2.188055, 'CONVERTER_fasta_to_2bit': 2.185396, 'openbabel_genProp': 2.185136, 'cshl_fastx_clipper': 2.183732, '__EXTRACT_DATASET__': 2.182131, 'vcfannotate': 2.181218, 'disco': 2.179835, 'rgweblogo3': 2.179604, 'qiime_filter_alignment': 2.179603, 'liftOver1': 2.175732, 'gops_concat_1': 2.166258, 'plotly_regression_performance_plots': 2.166192, 'picard_CleanSam': 2.165601, 'rseqc_clipping_profile': 2.163876, 'lca1': 2.163848, 'vegan_diversity': 2.162153, 'sickle': 2.157823, 'seqtk_sample': 2.154378, 'bed2gff1': 2.141623, 'vcf_filter': 2.135985, 'picard_DownsampleSam': 2.124724, 'imagecoordinates_flipaxis': 2.124654, 'nonpareil': 2.120621, 'blockbuster': 2.108191, 'psy_maps': 2.100864, 'vcfbedintersect': 2.089806, 'bismark_deduplicate': 2.074281, 'hcluster_sg_parser': 2.071966, 'xcms_plot_chromatogram': 2.064401, 'fasterq_dump': 2.060589, 'circos': 2.060589, 'mothur_split_abund': 2.051556, 'plotly_parallel_coordinates_plot': 2.045932, 'vcfselectsamples': 2.040768, 'miranda': 2.04072, 'DecoyDatabase': 2.03688, 'species_tree_generator': 2.034372, 'ctb_chemfp_butina_clustering': 2.029333, 'gemini_pathways': 2.023594, 'vcfgenotypes': 2.017873, 'viennarna_rnaalifold': 2.011481, 'trinity_define_clusters_by_cutting_tree': 1.991781, 'sklearn_numeric_clustering': 1.98716, 'wsdl_hmdb': 1.984381, 'vcffixup': 1.983029, 'goslimmer': 1.97786, 'mimodd_rebase': 1.977565, 'hicexplorer_hictransform': 1.968976, 'picard_ValidateSamFile': 1.950187, 'bed_to_protein_map': 1.939387, 'rbc_mirdeep2': 1.938087, 'picard_MarkDuplicatesWithMateCigar': 1.931521, 'tp_sort_rows': 1.929545, 'vcfcheck': 1.928933, 'PeakPickerHiRes': 1.915234, 'CONVERTER_Bam_Bai_0': 1.909542, 'sklearn_clf_metrics': 1.904094, 'smf_utils_find-boxes': 1.895639, 'SnpEff-cds-report': 1.894788, 'xcms_merge': 1.89085, 'vcfflatten2': 1.887485, 'trinotate': 1.885785, 'clusterprofiler_go': 1.884556, 'generic_filter': 1.884552, 'rdock_rbdock': 1.884002, 'gmx_sim': 1.883991, 'qualimap_bamqc': 1.883991, 'picard_FastqToSam': 1.863375, 'sklearn_model_validation': 1.856009, 'gatk2_realigner_target_creator': 1.855624, 'cor2': 1.853463, 'sample_seqs': 1.851579, 'picard_CollectWgsMetrics': 1.851406, 'get_flanks1': 1.85027, 'mummer_show_coords': 1.848453, 'ctb_multi_obgrep': 1.844392, 'ip_2d_split_binaryimage_by_watershed': 1.844389, 'clusterprofiler_bitr': 1.843441, 'macs2_bdgbroadcall': 1.84226, 'megablast_xml_parser': 1.840749, 'cshl_fastx_quality_statistics': 1.8298, 'feelnc': 1.822242, 'mothur_phylotype': 1.82014, 'get_sequences': 1.816452, 'sdf_to_tab': 1.815025, 'trinity_gene_to_trans_map': 1.808289, 'replace_chromosome_names': 1.806358, 'deeptools_compute_gc_bias': 1.804645, 'gmx_merge_topology_files': 1.802398, 'ctb_chemfp_nxn_clustering': 1.802388, 'mothur_make_lefse': 1.801708, 'meme_psp_gen': 1.801559, 'vt_normalize': 1.80127, 'keras_batch_models': 1.787621, 'stacks_populations': 1.764048, 'rgPicFixMate': 1.759972, 'retrieve_scxa': 1.759817, 'metaquantome_viz': 1.759586, 'scanpy_read_10x': 1.757729, 'vcfaddinfo': 1.754571, 'gemini_actionable_mutations': 1.754009, 'mothur_get_communitytype': 1.751973, 'humann2_regroup_table': 1.750835, 'heinz_bum': 1.744297, 'viennarna_rnaplot': 1.736304, 'numeric_clustering': 1.735646, 'alevin': 1.73051, 'scanpy_normalize': 1.726923, 'scanpy_scale_data': 1.724179, 'scanpy_normalise_data': 1.724179, 'scanpy_run_umap': 1.724179, 'scanpy_run_pca': 1.724179, 'scanpy_compute_graph': 1.724179, 'scanpy_find_markers': 1.724179, 'scanpy_filter_genes': 1.724179, 'scanpy_run_tsne': 1.724179, 'scanpy_find_variable_genes': 1.724179, 'scanpy_find_cluster': 1.724179, 'bg_protein_properties': 1.722697, 'msnbase_readmsdata': 1.71761, 'msgfplus': 1.704973, 'mothur_pcoa': 1.697991, 'samtools_split': 1.68841, 'CONVERTER_gff_to_interval_index_0': 1.687045, 'mimodd_varreport': 1.686451, 'seqtk_mergepe': 1.68198, 'ip_imageinfo': 1.670393, 'hgv_david': 1.669924, 'metaquantome_stat': 1.668732, 'ctb_confab': 1.66756, 'ctb_sdf2fps': 1.667517, 'regionalgam_ab_index': 1.666649, 'graphicsmagick_image_montage': 1.666516, 'ctb_np-likeness-calculator': 1.664345, 'mothur_pairwise_seqs': 1.66434, 'samtools_fastx': 1.664125, 'samtools_fixmate': 1.664125, 'dada2_filterAndTrim': 1.664125, 'snp_sites': 1.664125, 'sklearn_train_test_split': 1.664125, 'sucos_max_score': 1.664125, 'samtools_view': 1.664125, 'samtools_merge': 1.664125, 'maxbin2': 1.664125, 'picard_CollectGcBiasMetrics': 1.656481, 'get_feature_info': 1.655727, 'cshl_fastq_quality_boxplot': 1.651959, 'cmv': 1.650734, 'iframe': 1.647818, 'hmmer_hmmsearch': 1.647079, 'quality_metrics': 1.639508, 'mimodd_varextract': 1.636017, 'deeptools_correct_gc_bias': 1.628455, 'mothur_chimera_uchime': 1.622683, 'regionalgam_glmmpql': 1.617273, 'gafa': 1.616791, 'EMBOSS: fuzztran39': 1.616785, 'massbank_ws_searchspectrum': 1.616784, 'mothur_parsimony': 1.616767, 'melt': 1.614903, 'EMBOSS: wordcount109': 1.6149, 'Summary_Statistics1': 1.608449, 'Transformation': 1.607663, 'regionalgam_autocor_acf': 1.604609, 'scanpy_filter_cells': 1.602094, 'gemini_windower': 1.602083, 'openbabel_filter': 1.601722, 'lfq_protein_quant': 1.596708, 'bctools_remove_tail': 1.596705, 'hmmer_jackhmmer': 1.596705, 'rseqc_inner_distance': 1.589234, 'column_order_header_sort': 1.586749, 'ip_filter_standard': 1.582685, 'openbabel_svg_depiction': 1.582684, 'mothur_split_groups': 1.582683, 'macs2_filterdup': 1.582682, 'ip_viz_overlay_moving_and_fixed_image': 1.581786, 'picard_EstimateLibraryComplexity': 1.5815, 'length_and_gc_content': 1.580903, 'picard_ReorderSam': 1.578578, 'qiime_validate_mapping_file': 1.57521, 'graphclust_aggregate_alignments': 1.574796, 'biosigner': 1.573679, 'qualimap_rnaseq': 1.573384, 't2t_report': 1.573106, 'keras_model_builder': 1.564521, 'FalseDiscoveryRate': 1.563797, 'gatk4_mutect2': 1.563748, 'fpocket': 1.563748, 'picard_FixMateInformation': 1.556642, 'ctb_simsearch': 1.556035, 'idpassemble': 1.555409, 't_coffee': 1.539196, 'abims_xcms_group': 1.537909, 'EMBOSS: newseq59': 1.536449, 'snpSift_rmInfo': 1.536151, 'graphclust_glob_report_no_align': 1.531221, 'rseqc_mismatch_profile': 1.525173, 'mz_to_sqlite': 1.525169, 'rseqc_RNA_fragment_size': 1.524257, 'custom_pro_db': 1.520202, 'gd_calc_freq': 1.510611, 'treebest_best': 1.508275, 'ip_threshold': 1.506163, 'ucsc-twobittofa': 1.502674, 'describe_samples': 1.499855, 'ctb_osra': 1.495811, 'ip_count_objects': 1.49581, 'ucsc_cell_browser': 1.494071, 'megahit_contig2fastg': 1.490242, 'cast': 1.485317, 'tp_unfold_column_tool': 1.484441, 'dexseq_annotate': 1.481605, 'CONVERTER_bed_to_fli_0': 1.470257, 'gff_filter_by_attribute': 1.466789, 'hicexplorer_hicsummatrices': 1.462115, 'xcms_export_samplemetadata': 1.461114, 'rseqc_insertion_profile': 1.459399, 'cuffnorm': 1.454254, 'glimmer_knowlegde-based': 1.453953, 'vcftools_slice': 1.450121, 'vcf2pgSnp': 1.449269, 'gatk2_depth_of_coverage': 1.440747, 'Extract_features1': 1.437008, 'abims_CAMERA_combinexsAnnos': 1.43485, 'ip_2d_filter_segmentation_by_features': 1.434802, 'bctools_remove_spurious_events': 1.419881, 'checkFormat': 1.415021, 'FileConverter': 1.406689, 'rnaz': 1.396878, 'eggnog_mapper': 1.396502, 'tb_variant_filter': 1.387438, 'ncbi_acc_download': 1.384901, 'lofreq_call': 1.384901, 'keras_train_and_eval': 1.384901, 'gatk_count_covariates': 1.38022, 'mothur_make_design': 1.378248, 'CONVERTER_bed_to_bgzip_0': 1.374994, 'mothur_unifrac_unweighted': 1.371499, 'hicexplorer_hicnormalize': 1.362981, 'tables_arithmetic_operations': 1.361806, 'rseqc_read_duplication': 1.360977, 'samtools_markdup': 1.359558, 'rseqc_junction_saturation': 1.357307, 'qiime_core_diversity': 1.351874, 'thermo_raw_file_converter': 1.348282, 'mothur_homova': 1.342089, 'mothur_summary_qual': 1.336652, '__APPLY_RULES__': 1.333931, 'find_in_reference': 1.331735, 'rbc_splitfasta': 1.32619, 'scater_plot_pca': 1.32295, 'metaquantome_expand': 1.322462, 'openbabel_remDuplicates': 1.320806, 'mothur_nmds': 1.316831, 'ambertools_antechamber': 1.314042, 'tsne': 1.310904, 'build_ucsc_custom_track_1': 1.305476, 'ip_binary_to_labelimage': 1.303409, 'vcftools_annotate': 1.302309, 'mummer_delta_filter': 1.300582, 'intarna': 1.289082, 'gdal_gdalinfo': 1.287836, 'bamSplit': 1.287105, 'TargetedFileConverter': 1.284213, 'ip_projective_transformation_points': 1.261662, 'lofreq_indelqual': 1.251241, 'lofreq_alnqual': 1.251241, 'bamtools_split': 1.251241, 'qiime_count_seqs': 1.244566, 'scater_plot_dist_scatter': 1.241924, 'glimmer_build-icm': 1.236754, 'hcluster_sg': 1.227571, 'vsearch_chimera_detection': 1.219686, 'macs2_bdgcmp': 1.217361, 'Psortb': 1.191641, 'bctools_extract_crosslinked_nucleotides': 1.18058, 'picard_QualityScoreDistribution': 1.162723, 'cshl_fastx_nucleotides_distribution': 1.148636, 'OpenSwathWorkflow': 1.142431, 'shyft_longitudes': 1.141616, 'scater_normalize': 1.141337, 'Fetch Taxonomic Ranks': 1.138332, 'scanpy_parameter_iterator': 1.109202, 'ConsensusID': 1.105646, 'padel': 1.087132, 'dada2_dada': 1.087087, 'vcftools_isec': 1.084224, 'create_or_update': 1.078656, 'seqtk_trimfq': 1.060336, 'abims_xcms_fillPeaks': 1.054002, 'gene2exon1': 1.0456, 'rcas': 1.032861, 'rseqc_junction_annotation': 1.026478, 'EMBOSS: water107': 1.025132, 'Multivariate': 1.020433, 'Heatmap': 1.018424, 'mothur_collect_single': 1.010236, 'mimodd_header': 1.008772, 'replace_column_with_key_value_file': 1.003427, '_ensembl_gtf2gene_list': 1.003041, 'seurat_run_tsne': 1.002723, 'Univariate': 0.999623, 'ip_scale_image': 0.997359, 'regex_replace': 0.993033, 'abims_xcms_retcor': 0.97456, 'mothur_amova': 0.969833, 'hicexplorer_chicqualitycontrol': 0.957327, 'deeptools_plot_enrichment': 0.942563, 'param_value_from_file': 0.920374, 'FeatureFinderMultiplex': 0.919696, 'bismark_bowtie': 0.913126, 'sucos_docking_scoring': 0.908686, 'express': 0.905208, 'idba_hybrid': 0.904722, '__SORTLIST__': 0.903579, 'ctb_spectrophore_search': 0.897192, 'ctb_subsearch': 0.897121, 'dada2_learnErrors': 0.897107, 'climate_stripes': 0.897107, 'scater_plot_tsne': 0.897107, 'rdock_rbcavity': 0.897107, 'snpSift_annotate': 0.887112, 'stacks_genotypes': 0.865049, 'gff_to_sequence': 0.863155, 'hyphy_absrel': 0.861753, 'hyphy_gard': 0.861753, 'vmd_hbonds': 0.857667, 'vcftools_compare': 0.855604, 'metaquantome_filter': 0.854305, 'blastxml_to_tabular': 0.846198, 'ambertools_acpype': 0.84405, 'bismark_pretty_report': 0.823888, 'regionalgam_flight_curve': 0.819444, 'seurat_read10x': 0.816677, 'tbvcfreport': 0.816152, 'openbabel_change_title': 0.808189, 'ip_coordinates_of_roi': 0.807745, 'ip_histogram_equalization': 0.79817, 'ctb_rdkit_descriptors': 0.789603, 'translate_bed_sequences': 0.732184, 'wiggle2simple1': 0.72055, 'glimmer_not-knowlegde-based': 0.718949, 'gatk2_haplotype_caller': 0.707488, 'gatk2_unified_genotyper': 0.671346, 'interactive_tool_phinch': 0.654252, 'qiime_split_libraries_fastq': 0.652422, 'ip_overlay_segmentation': 0.650753, 'HighResPrecursorMassCorrector': 0.650389, 'gatk2_base_recalibrator': 0.649885, 'rnbeads': 0.642165, 'wc_gnu': 0.641854, 'methtools_dmr': 0.641854, 'vigiechiro_bilanenrichirp': 0.638303, 'ctb_mordred_descriptors': 0.628821, 'trinity_samples_qccheck': 0.625591, 'regionalgam_plot_trend': 0.616936, 'MultiplexResolver': 0.612909, 'minfi_read450k': 0.601911, 'r_correlation_matrix': 0.595007, 'ctb_remIons': 0.576613, 'ctb_pubchem_download_as_smiles': 0.574748, '__TAG_FROM_FILE__': 0.572669, 'gspan': 0.571782, 'bcftools_view': 0.551645, 'CONVERTER_sam_to_unsorted_bam': 0.528932, 'EMBOSS: revseq82': 0.528516, 'gatk2_indel_realigner': 0.517258, 'mass_spectrometry_imaging_preprocessing': 0.517097, 'hgv_linkToGProfile': 0.510734, 'ncbi_rpsblast_wrapper': 0.510061, 'metams_runGC': 0.507412, 'map_peptides_to_bed': 0.491201, 'methtools_calling': 0.488017, 'Batch_correction': 0.480431, 'abims_xcms_summary': 0.479918, 'hgv_snpFreq': 0.479918, 'smf_utils_filter-by-energy': 0.479918, 'smf_utils_extract-boxed-sequences': 0.464428, 'proteomics_moff': 0.446717, 'mass_spectrometry_imaging_mzplots': 0.432459, 'locarna_multiple': 0.421994, 'pipmir': 0.418736, 'minfi_dmr': 0.412249, 'gatk2_variant_select': 0.411048, 'mass_spectrometry_imaging_qc': 0.405764, 'MAF_To_Fasta1': 0.403417, 'minfi_qc': 0.399759, 'genbank_to_gff': 0.393246, 'mass_spectrometry_imaging_combine': 0.38549, 'seurat_create_seurat_object': 0.384026, '__RELABEL_FROM_FILE__': 0.382519, 'regionalgam_gls': 0.381489, 'bedtools_genomecoveragebed_histogram': 0.360841, 'mass_spectrometry_imaging_segmentations': 0.355038, 'mass_spectrometry_imaging_ion_images': 0.354149, 'scater_create_qcmetric_ready_sce': 0.340402, 'rsem_calculate_expression': 0.326823, 'ip_2d_feature_extraction': 0.32402, 'sam_bw_filter': 0.314406, 'vegan_rarefaction': 0.297459, 'rgPicardMarkDups': 0.270732, 'gbk_to_orf': 0.262364, 'CONVERTER_interval_to_bgzip_0': 0.259699, 'ip_convertimage': 0.257914, 'MzTabExporter': 0.250928, 'mlocarna': 0.24011, 'seurat_find_variable_genes': 0.234742, 'Determine_BC': 0.228208, 'scater_plot_exprs_freq': 0.228173, 'seurat_scale_data': 0.21546, 'seurat_normalise_data': 0.21546, 'ip_projective_transformation': 0.213135, 'OpenSwathAssayGenerator': 0.192684, 'OpenSwathDecoyGenerator': 0.192684, 'cat_summarise': 0.192684, 'schicexplorer_schicplotconsensusmatrices': 0.192684, 'schicexplorer_schicclustersvl': 0.192684, 'hicexplorer_chicsignificantinteractions': 0.192684, 'hicexplorer_chicviewpointbackgroundmodel': 0.192684, 'hicexplorer_chicdifferentialtest': 0.192684, 'schicexplorer_schicconsensusmatrices': 0.192684, 'hicexplorer_chicaggregatestatistic': 0.192684, 'hicexplorer_chicviewpoint': 0.192684, 'cat_contigs': 0.192684, 'lofreq_viterbi': 0.192684, 'lofreq_filter': 0.192684, 'chira_map': 0.192684, 'proteinortho': 0.192683, 'xchem_transfs_scoring': 0.192683, 'rdock_sort_filter': 0.192683, 'pyprophet_score': 0.192683, 'pyprophet_merge': 0.192682, 'cp_measure_image_intensity': 0.192682, 'cp_enhance_or_suppress_features': 0.192682, 'mmpbsa_mmgbsa': 0.192682, 'cp_identify_primary_objects': 0.192682, 'cp_export_to_spreadsheet': 0.192682, 'cp_measure_object_size_shape': 0.192682, 'cp_measure_granularity': 0.192682, 'diapysef': 0.192682, 'enumerate_charges': 0.192682, 'ctb_silicos_qed': 0.192682, 'cp_display_data_on_image': 0.192682, 'pyprophet_export': 0.192682, 'cp_cellprofiler': 0.192682, 'cp_measure_object_intensity': 0.192682, 'iedb_api': 0.192682, 'cp_gray_to_color': 0.192682, 'cp_relate_objects': 0.192682, 'biom_convert': 0.192682, 'cp_convert_objects_to_image': 0.192682, 'pyprophet_protein': 0.192682, 'cp_mask_image': 0.192682, 'cp_measure_image_area_occupied': 0.192682, 'cp_measure_image_quality': 0.192682, 'cp_common': 0.192682, 'cp_measure_texture': 0.192682, 'pyprophet_peptide': 0.192682, 'cp_save_images': 0.192682, 'seurat_dim_plot': 0.187445, 'raceid_main': 0.182322, 'minfi_rset': 0.174375, 'msconvert_nix': 0.173137, 'gtf_filter_by_attribute_values_list': 0.169041, 'sRNAPipe': 0.163629, 'metagene_annotator': 0.163629, 'gmx_editconf': 0.163629, 'gmx_restraints': 0.163629, 'chira_quantify': 0.163629, 'sixgill_build': 0.163629, 'chira_merge': 0.163629, 'idr_download_by_ids': 0.163629, 'gmx_trj': 0.163629, 'cat_bins': 0.163629, 'medaka_consensus': 0.163629, 'gmx_makendx': 0.163629, 'chira_extract': 0.163629, 'chira_collapse': 0.163629, 'seq2hla': 0.163629, 'seurat_export_cellbrowser': 0.163416, 'extract_bcs.py': 0.158711, 'annotateMyIDs': 0.157811, 'minfi_maptogenome': 0.156906, 'extract_aln_ends.py': 0.154419, 'flexbar_no_split': 0.154419, 'rm_spurious_events.py': 0.154419, 'coords2clnt.py': 0.154419, 'remove_tail.py': 0.154419, 'merge_pcr_duplicates.py': 0.154419, 'createInterval': 0.152766, 'gemini_de_novo': 0.150962, 'seq_composition': 0.137781, 'ivar_variants': 0.119801, 'stoceps_trend_indic': 0.119801, 'stoceps_glm': 0.119801, 'stoceps_maketablecarrer': 0.119801, 'stoceps_glm_group': 0.119801, 'ivar_consensus': 0.119801, 'stoceps_filteringsp': 0.119801, 'gemini_recessive_and_dominant': 0.114031, 'gemini_comp_hets': 0.110718, 'proteomics_search_tandem_1': 0.110363, 'seurat_find_clusters': 0.10907, 'msconvert_win': 0.107412, 'FileInfo': 0.10661, 'Draw_phylogram': 0.105361, 'FileMerger': 0.10478, 'seurat_run_pca': 0.099724, 'predict_pipeline': 0.095757, 'deeptools_bamCoverage': 0.095653, 'MapStatistics': 0.0954, 'qiime_split_libraries': 0.09535, 'gatk2_variant_annotator': 0.095344, 'gtf2bedgraph': 0.095328, 'RNAlien': 0.095321, 'gatk2_print_reads': 0.095318, 'Interval_Maf_Merged_Fasta2': 0.095314, 'ctb_change_title': 0.09531, 'EMBOSS: extractseq35': 0.09531, 'EMBOSS: shuffleseq87': 0.09531, 'compose_text_param': 0.09531, 'sceasy_convert': 0.09531, 'Sff_extractor': 0.09531, 'cuffquant': 0.09531, 'minfi_dmp': 0.09531, 'dropletutils_empty_drops': 0.09531, 'kraken-translate': 0.09531, 'dropletutils_read_10x': 0.09531, 'methtools_tiling': 0.09531, 'seurat_filter_cells': 0.09531, 'align_back_trans': 0.09531, '_dropletBarcodePlot': 0.09531, 'tophat_fusion_post': 0.09531, 'minfi_dropsnp': 0.09531, 'proteomics_search_protein_prophet_1': 0.09531, 'minfi_getM': 0.09531, 'w4mjoinpn': 0.09531, 'minfi_getbeta': 0.09531, 'barchart_gnuplot': 0.09531, 'ia_coordinates_of_roi': 0.095309, 'scale_image': 0.095309, 'FeatureFinderMetabo': 0.095308, 'w4mcorcov': 0.095308, 'bg_statistical_hypothesis_testing': 0.095308, 'QCCalculator': 0.095307, 'gatk2_variant_filtration': 0.095307, 'smooth_running_window': 0.095307, 'IsobaricAnalyzer': 0.095305, 'bwa_wrapper': 0.095305, 'viz_overlay_moving_and_fixed_image': 0.095304, 're_he_maldi_image_registration': 0.095303, 'EMBOSS: tranalign100': 0.095299, 'snpSift_dbnsfp_generic': 0.095295, 'ctb_compound_convert': 0.095291, 'prepare_ligand': 0.095291, 'minfi_ppfun': 0.095287, 'gatk2_reduce_reads': 0.09528, 'convert_bc_to_binary_RY.py': 0.095279, 'flexbar_split_RR_bcs': 0.095279, 'flexbardsc': 0.095277, 'FeatureFinderCentroided': 0.095277, 'EMBOSS: fuzzpro38': 0.095276, 'methtools_plot': 0.095269, 'EMBOSS: seqret84': 0.095268, 'decoyfasta': 0.095245, 'scanpy_plot_embed': 0.095243, 'Digestor': 0.095215, 'EMBOSS: textsearch98': 0.094976, 'Kraken2Tax': 0.094829, 'kraken-report': 0.094235, 'eden_sequence_converter': 0.093463, 'interactive_tool_fairdom_send': 0.093143, 'cat_add_names': 0.092557, 'ctb_opsin': 0.091962, 'ProteinInference': 0.090904, 'CometAdapter': 0.090802, 'genomespace_exporter': 0.088586, 'xchem_pose_scoring': 0.088514, 'cshl_find_and_replace': 0.088261, 'minfi_getsnp': 0.088237, 'cshl_awk_tool': 0.085592, 'ProteinResolver': 0.08434, 'scanpy_regress_variable': 0.083862, 'kraken-filter': 0.083862, 'kraken-mpa-report': 0.081719, 'IDMassAccuracy': 0.079791, 'gd_rank_terms': 0.078823, 'get_read_pipeline': 0.078823, 'cshl_word_list_grep': 0.078807, 'deeptools_bamCompare': 0.078807, 'FeatureLinkerUnlabeledQT': 0.078609, 'FeatureLinkerUnlabeled': 0.078609, 'ConsensusMapNormalizer': 0.078609, 'p_clip_peaks': 0.078392, 'MassCalculator': 0.078023, 'aggregate_scores_in_intervals2': 0.077675, 'proteomics_search_peptide_prophet_1': 0.076117, 'ctb_alignit': 0.075026, 'ctb_alignit_create_db': 0.075026, 'methtools_filter': 0.075026, 'qual_stats_boxplot': 0.074854, 'term_id_vs_term_name': 0.074208, 'get_subontology_from': 0.074208, 'SequenceCoverageCalculator': 0.073626, 'gatk2_variant_recalibrator': 0.072487, 'DatamashOps': 0.069029, '_salmon_kallisto_mtx_to_10x': 0.057805, 'sccaf_regress_out': 0.057805, 'vcftools_merge': 0.055273, 'methtools_destrand': 0.048262, 'kraken': 0.038444, 'MetaProSIP': 0.034643, 'ctb_ob_genProp': 0.034563, 'schicexplorer_schicqualitycontrol': 0.029896, 'schicexplorer_schicplotclusterprofiles': 0.029896, 'schicexplorer_schicnormalize': 0.029896, '__FILTER_FAILED_DATASETS__': 0.025786, 'fragmenter': 0.022181, 'mass_spectrometry_imaging_classification': 0.009427}\n"
     ]
    }
   ],
   "source": [
    "sorted_cw = dict(sorted(class_weights.items(), key=lambda kv: kv[1], reverse=True))\n",
    "cw_tools = dict()\n",
    "for i in sorted_cw:\n",
    "    if i in reverse_dictionary:\n",
    "        if sorted_cw[i] > 0:\n",
    "            cw_tools[reverse_dictionary[i]] = sorted_cw[i]\n",
    "print(cw_tools)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
