{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tool recommendation \n",
    "## (Random forest with weighted cross-entropy loss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import json\n",
    "import warnings\n",
    "import operator\n",
    "\n",
    "import h5py\n",
    "from joblib import load\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "\n",
    "def read_file(file_path):\n",
    "    with open(file_path, 'r') as data_file:\n",
    "        data = json.loads(data_file.read())\n",
    "    return data\n",
    "\n",
    "\n",
    "def create_model(model_path):\n",
    "    loaded_model = load(model_path) \n",
    "    return loaded_model\n",
    "\n",
    "\n",
    "def get_predicted_tools(base_tools, predictions, topk):\n",
    "    \"\"\"\n",
    "    Get predicted tools. If predicted tools are less in number, combine them with published tools\n",
    "    \"\"\"\n",
    "    intersection = list(set(predictions).intersection(set(base_tools)))\n",
    "    return intersection[:topk]\n",
    "\n",
    "\n",
    "def sort_by_usage(t_list, class_weights, d_dict):\n",
    "    \"\"\"\n",
    "    Sort predictions by usage/class weights\n",
    "    \"\"\"\n",
    "    tool_dict = dict()\n",
    "    for tool in t_list:\n",
    "        t_id = d_dict[tool]\n",
    "        tool_dict[tool] = class_weights[str(t_id)]\n",
    "    tool_dict = dict(sorted(tool_dict.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    return list(tool_dict.keys()), list(tool_dict.values())\n",
    "\n",
    "\n",
    "def separate_predictions(base_tools, predictions, last_tool_name, weight_values, topk):\n",
    "    \"\"\"\n",
    "    Get predictions from published and normal workflows\n",
    "    \"\"\"\n",
    "    last_base_tools = list()\n",
    "    predictions = predictions * weight_values\n",
    "    prediction_pos = np.argsort(predictions, axis=-1)\n",
    "    topk_prediction_pos = prediction_pos[-topk:]\n",
    "    # get tool ids\n",
    "    pred_tool_ids = [reverse_dictionary[str(tool_pos)] for tool_pos in topk_prediction_pos]\n",
    "    if last_tool_name in base_tools:\n",
    "        last_base_tools = base_tools[last_tool_name]\n",
    "        if type(last_base_tools).__name__ == \"str\":\n",
    "            # get published or compatible tools for the last tool in a sequence of tools\n",
    "            last_base_tools = last_base_tools.split(\",\")\n",
    "    # get predicted tools\n",
    "    p_tools = get_predicted_tools(last_base_tools, pred_tool_ids, topk)\n",
    "    sorted_c_t, sorted_c_v = sort_by_usage(p_tools, class_weights, dictionary)\n",
    "    return sorted_c_t, sorted_c_v\n",
    "\n",
    "\n",
    "def compute_recommendations(model, tool_sequence, labels, dictionary, reverse_dictionary, class_weights, topk=10, max_seq_len=25):\n",
    "    tl_seq = tool_sequence.split(\",\")\n",
    "    tl_seq_ids = [str(dictionary[t]) for t in tl_seq]\n",
    "    last_tool_name = tl_seq[-1]\n",
    "    sample = np.zeros(max_seq_len)\n",
    "    weight_val = list(class_weights.values())\n",
    "    weight_val = np.reshape(weight_val, (len(weight_val),))\n",
    "    for idx, tool_id in enumerate(tl_seq_ids):\n",
    "        sample[idx] = int(tool_id)\n",
    "    sample_reshaped = np.reshape(sample, (1, max_seq_len))\n",
    "    tool_sequence_names = [reverse_dictionary[str(tool_pos)] for tool_pos in tl_seq_ids]\n",
    "    # predict next tools for a test path\n",
    "    prediction = model.predict(sample_reshaped)\n",
    "    nw_dimension = prediction.shape[1]\n",
    "    prediction = np.reshape(prediction, (nw_dimension,))\n",
    "    \n",
    "    half_len = int(nw_dimension / 2)\n",
    "    \n",
    "    pub_t, pub_v = separate_predictions(standard_connections, prediction[:half_len], last_tool_name, weight_val, topk)\n",
    "    # get recommended tools from normal workflows\n",
    "    c_t, c_v = separate_predictions(compatible_tools, prediction[half_len:], last_tool_name, weight_val, topk)\n",
    "    # combine predictions coming from different workflows\n",
    "    # promote recommended tools coming from published workflows\n",
    "    # to the top and then show other recommendations\n",
    "    print()\n",
    "    tool_seq_name = \",\".join(tool_sequence_names)\n",
    "    print(\"Current tool sequence: \")\n",
    "    print()\n",
    "    print(tool_seq_name)\n",
    "    print()\n",
    "    print(\"Overall recommendations: \")\n",
    "    print()\n",
    "    pub_t.extend(c_t)\n",
    "    pub_v.extend(c_v)\n",
    "    # remove duplicates if any\n",
    "    pub_t = list(dict.fromkeys(pub_t))\n",
    "    pub_v = list(dict.fromkeys(pub_v))\n",
    "    print(pub_t)\n",
    "    ids_tools = dict()\n",
    "    for key in pub_t:\n",
    "        ids_tools[key] = dictionary[key]\n",
    "    print()\n",
    "    print(\"Recommended tool ids:\")\n",
    "    print()\n",
    "    for i in ids_tools:\n",
    "        rev_id = dictionary[i]\n",
    "        wt = class_weights[str(rev_id)]\n",
    "        print(i + \"(\" + str(ids_tools[i]) + \")\" + \"(\" + str(wt) + \")\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unpack trained model for prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = \"../data/tool_recommendation_model.hdf5\"\n",
    "model_path = \"../data/model.joblib\"\n",
    "\n",
    "output_model = h5py.File(output_path, 'r')\n",
    "dictionary = json.loads(output_model.get('data_dictionary').value)\n",
    "reverse_dictionary = dict((str(v), k) for k, v in dictionary.items())\n",
    "class_weights = json.loads(output_model.get('class_weights').value)\n",
    "standard_connections = json.loads(output_model.get('standard_connections').value)\n",
    "compatible_tools = json.loads(output_model.get('compatible_tools').value)\n",
    "loaded_model = create_model(model_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Assembly: \n",
    "# (https://training.galaxyproject.org/training-material/topics/assembly/tutorials/debruijn-graph-assembly/tutorial.html)\n",
    "# spades -> 'bandage_info', 'fasta-stats', 'bandage_image', 'fasta_filter_by_length', 'abricate', 'quast', 'mlst' ... \n",
    "# velveth -> velvetg\n",
    "# (https://training.galaxyproject.org/training-material/topics/assembly/tutorials/unicycler-assembly/tutorial.html)\n",
    "# unicycler -> 'bandage_info', 'glimmer_build-icm', 'glimmer_knowlegde-based', 'bandage_image', 'transdecoder', 'minimap2', 'antismash', 'fasta_filter_by_length' ...\n",
    "\n",
    "\n",
    "## Computational chemistry\n",
    "# ctb_remDuplicates -> ctb_remIons \n",
    "# ctb_remDuplicates,ctb_remIons -> 'ctb_chemfp_mol2fps', 'ctb_compound_convert'\n",
    "# ctb_remDuplicates,ctb_remIons,ctb_chemfp_mol2fps -> 'ctb_chemfp_butina_clustering', 'ctb_simsearch', 'ctb_chemfp_nxn_clustering', 'comp1'\n",
    "# (https://training.galaxyproject.org/training-material/topics/computational-chemistry/tutorials/cheminformatics/tutorial.html)\n",
    "\n",
    "\n",
    "## RAD-seq\n",
    "# (https://training.galaxyproject.org/training-material/topics/ecology/tutorials/ref-based-rad-seq/tutorial.html)\n",
    "# stacks_procrad -> 'bwa', 'bwa_wrapper', 'Grep1', 'stacks_denovomap', 'fastqc', 'fastq_filter'\n",
    "\n",
    "\n",
    "## Epigenetics\n",
    "# (https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/atac-seq/tutorial.html)\n",
    "# cutadapt,bowtie2 => samtools_flagstat', 'picard_MarkDuplicates', 'picard_AddOrReplaceReadGroups', 'macs2_callpeak','bg_sortmerna', 'multiqc', 'hisat2', 'trim_galore', 'bowtie2' ...\n",
    "# cutadapt,bowtie2,picard_MarkDuplicates -> 'picard_ReorderSam', 'gatk4_mutect2', 'samtools_rmdup'\n",
    "# cutadapt,bowtie2,picard_MarkDuplicates,genrich -> 'pygenomeTracks'\n",
    "#\n",
    "# (https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/methylation-seq/tutorial.html)\n",
    "# bwameth -> 'samtools_rmdup', 'samtools_sort', 'bam_to_sam', 'pileometh' ..\n",
    "# bwameth,pileometh -> 'deeptools_compute_matrix', 'tp_sed_tool', 'Filter1', 'metilene', 'Remove beginning1', 'wig_to_bigWig'\n",
    "#\n",
    "# bowtie2 -> samtools_flagstat', 'picard_MarkDuplicates', 'picard_AddOrReplaceReadGroups ... \n",
    "# bowtie2,deeptools_multi_bam_summary -> 'deeptools_plot_pca', 'r_correlation_matrix' ...\n",
    "# (https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/formation_of_super-structures_on_xi/tutorial.html)\n",
    "\n",
    "# bowtie2,hicexplorer_hicbuildmatrix -> 'hicexplorer_hicsummatrices', 'hicexplorer_hicplotviewpoint', 'tp_sed_tool', 'hicexplorer_hiccorrectmatrix', 'hicexplorer_hicmergematrixbins', 'hicexplorer_hicpca' ..\n",
    "# bowtie2,hicexplorer_hicbuildmatrix,hicexplorer_hicmergematrixbins -> 'hicexplorer_hiccorrectmatrix', 'hicexplorer_hicplottads', 'hicexplorer_hicplotmatrix'\n",
    "# (https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/hicexplorer/tutorial.html)\n",
    "\n",
    "# minfi_read450k -> 'minfi_getbeta'\n",
    "\n",
    "\n",
    "## Genome annotation\n",
    "# (https://training.galaxyproject.org/training-material/topics/genome-annotation/tutorials/annotation-with-maker/tutorial.html)\n",
    "# maker -> 'gffread', 'maker_map_ids', 'jcvi_gff_stats'\n",
    "# (https://training.galaxyproject.org/training-material/topics/genome-annotation/tutorials/annotation-with-prokka/tutorial.html)\n",
    "# prokka -> 'mlst', 'jbrowse', 'taxonomy_krona_chart' ...\n",
    "\n",
    "\n",
    "## Imaging\n",
    "# (https://training.galaxyproject.org/training-material/topics/imaging/tutorials/hela-screen-analysis/tutorial.html)\n",
    "# ip_filter_standard -> 'ip_histogram_equalization', 'ip_threshold', 'ip_count_objects\n",
    "# ip_filter_standard,ip_threshold -> ip_binary_to_labelimage', 'ip_2d_split_binaryimage_by_watershed', 'ip_count_objects', 'ip_convertimage'\n",
    "# ip_filter_standard,ip_threshold,ip_2d_split_binaryimage_by_watershed -> 'ip_2d_filter_segmentation_by_features', 'ip_2d_feature_extraction'\n",
    "\n",
    "\n",
    "## Mass spectrometry\n",
    "# mass_spectrometry_imaging_preprocessing -> 'mass_spectrometry_imaging_combine', 'mass_spectrometry_imaging_preprocessing'\n",
    "# mass_spectrometry_imaging_preprocessing,mass_spectrometry_imaging_combine -> 'maldi_quant_preprocessing', 'mass_spectrometry_imaging_preprocessing', 'mass_spectrometry_imaging_qc' ...\n",
    "# search_gui -> peptide_shaker\n",
    "# search_gui,peptide_shaker -> 'mz_to_sqlite', 'Remove beginning1', 'tp_replace_in_column', 'unipept', proteomics_moff ...\n",
    "\n",
    "\n",
    "## Single cell\n",
    "# raceid_main -> 'seurat','raceid_trajectory'\n",
    "# raceid_main,raceid_trajectory -> 'raceid_inspecttrajectory'\n",
    "# raceid_inspectclusters,__BUILD_LIST__ -> 'picard_MarkDuplicates', 'hisat2', 'stringtie_merge', 'cutadapt', 'tp_cat'\n",
    "# raceid_filtnormconf -> 'raceid_clustering', '__BUILD_LIST__'\n",
    "# raceid_filtnormconf,raceid_clustering -> 'raceid_trajectory', 'raceid_inspectclusters'\n",
    "# scanpy_regress_variable -> scanpy_scale_data \n",
    "# scanpy_regress_variable,scanpy_scale_data -> 'scanpy_run_pca', 'scanpy_find_variable_genes'\n",
    "# scanpy_regress_variable,scanpy_scale_data,scanpy_run_pca -> 'scanpy_compute_graph', 'scanpy_plot', 'scanpy_run_tsne', 'scanpy_plot_embed'\n",
    "\n",
    "# (https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/scrna-preprocessing-tenx/tutorial.html)\n",
    "# rna_starsolo -> 'dropletutils', 'multiqc'\n",
    "# rna_starsolo,dropletutils -> 'scanpy_read_10x', 'scanpy_cluster_reduce_dimension', 'seurat_read10x', 'anndata_import', 'scanpy_plot', 'raceid_filtnormconf'\n",
    "\n",
    "\n",
    "## Variant calling\n",
    "# (https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/microbial-variants/tutorial.html)\n",
    "# snippy -> 'bedtools_intersectbed', 'vcfvcfintersect', 'Remove beginning1', 'snippy_core', 'qualimap_bamqc', 'freebayes', 'jbrowse', 'vcfcombine'\n",
    "\n",
    "# https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/somatic-variants/tutorial.html\n",
    "# trimmomatic,bwa_mem,samtools_rmdup,bamleftalign -> 'samtool_filter2', 'ivar_variants', 'freebayes', 'varscan_somatic', 'deeptools_bam_coverage', 'fastqc', 'ngsutils_bam_filter', 'bamFilter', 'rgPicFixMate', 'samtools_calmd'\n",
    "# trimmomatic,bwa_mem,samtools_rmdup,bamleftalign,varscan_somatic -> 'bcftools_norm', 'gemini_annotate', 'vt_normalize', 'vcffilter2', 'vcfallelicprimitives', 'snpEff'\n",
    "\n",
    "# (https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/dip/tutorial.html)\n",
    "# freebayes -> 'vcfallelicprimitives', 'bcftools_norm', 'custom_pro_db', 'vcfvcfintersect'\n",
    "# freebayes,vcfallelicprimitives -> 'vt_normalize', 'snpSift_filter', 'snpSift_annotate', 'snpEff'\n",
    "# freebayes,vcfallelicprimitives,snpEff -> 'gemini_load', 'mimodd_varreport', 'snpSift_extractFields\n",
    "\n",
    "\n",
    "# Transcriptomics\n",
    "#(https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/small_ncrna_clustering/tutorial.html)\n",
    "# samtools_sort,blockclust,sort1 -> 'cshl_awk_tool', 'Show beginning1', 'tp_awk_tool', 'blockbuster'\n",
    "\n",
    "# (https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/ref-based/tutorial.html)\n",
    "# cutadapt -> 'umi_tools_extract', 'fastq_paired_end_interlacer', 'chira_collapse', 'rna_star',\n",
    "# cutadapt,rna_star -> 'featurecounts', 'multiqc', 'htseq_count', 'rseqc_infer_experiment', 'samtools_stats', 'bamFilter',\n",
    "# cutadapt,rna_star,featurecounts -> 'multiqc', 'deseq2', 'tp_sort_header_tool', 'bamFilter', 'collection_column_join ...\n",
    "\n",
    "\n",
    "# Single-cell HiC\n",
    "# schicexplorer_schicqualitycontrol -> 'schicexplorer_schicnormalize'\n",
    "# schicexplorer_schicnormalize -> 'schicexplorer_schicclustersvl', 'schicexplorer_schicconsensusmatrices', 'schicexplorer_schicplotclusterprofiles'\n",
    "# schicexplorer_schicnormalize,schicexplorer_schicclustersvl -> 'schicexplorer_schicplotclusterprofiles', 'schicexplorer_schicconsensusmatrices'\n",
    "\n",
    "\n",
    "# Animal detection on acoustic recording\n",
    "# vigiechiro_idvalid -> 'vigiechiro_bilanenrichipf', 'vigiechiro_bilanenrichirp'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Recommended tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Current tool sequence: \n",
      "\n",
      "Cut1\n",
      "\n",
      "Overall recommendations: \n",
      "\n",
      "[]\n",
      "\n",
      "Recommended tool ids:\n",
      "\n"
     ]
    }
   ],
   "source": [
    "topk = 10 # set the maximum number of recommendations\n",
    "\n",
    "# Give tools ids in a sequence and see the recommendations. # To know all the tool ids, \n",
    "# please print the variable 'reverse_dictionary'\n",
    "tool_seq = \"Cut1\"\n",
    "\n",
    "compute_recommendations(loaded_model, tool_seq, \"\", dictionary, reverse_dictionary, class_weights, topk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CONVERTER_Bam_Bai_0': 1,\n",
       " 'fastqc': 2,\n",
       " 'gemini_interactions': 3,\n",
       " 'IDMerger': 4,\n",
       " 'tophat2': 5,\n",
       " 'Grouping1': 6,\n",
       " 'deseq2_single': 7,\n",
       " 'bwa_wrapper': 8,\n",
       " 'find_in_reference': 9,\n",
       " 'varscan': 10,\n",
       " 'ncbi_blastp_wrapper': 11,\n",
       " 'mergeCols1': 12,\n",
       " 'Count1': 13,\n",
       " 'cshl_find_and_replace': 14,\n",
       " 'fastq_filter': 15,\n",
       " 'gops_coverage_1': 16,\n",
       " 'eukaryotic_ncbi_submission': 17,\n",
       " 'Grep1': 18,\n",
       " 'sam_bw_filter': 19,\n",
       " 'deseq2': 20,\n",
       " 'bwa_mem': 21,\n",
       " 'Convert characters1': 22,\n",
       " 'trim_galore': 23,\n",
       " 'CONVERTER_bedgraph_to_bigwig': 24,\n",
       " 'charts': 25,\n",
       " 'sample_seqs': 26,\n",
       " 'IDFilter': 27,\n",
       " 'fragmenter': 28,\n",
       " 'trimmer': 29,\n",
       " 'comp1': 30,\n",
       " 'blockbuster': 31,\n",
       " 'EMBOSS: transeq101': 32,\n",
       " 'ctb_remSmall': 33,\n",
       " 'naive_variant_caller': 34,\n",
       " 'ncbi_tblastn_wrapper': 35,\n",
       " 'bismark_bowtie2': 36,\n",
       " 'computeMatrix': 37,\n",
       " 'cshl_multijoin': 38,\n",
       " 'PicardASMetrics': 39,\n",
       " 'Remove_ending': 40,\n",
       " 'blastxml_to_tabular': 41,\n",
       " 'bam2wig': 42,\n",
       " 'gops_join_1': 43,\n",
       " 'bowtie2': 44,\n",
       " 'heatmapper_deepTools': 45,\n",
       " 'tab2fasta': 46,\n",
       " 'CONVERTER_bed_to_bgzip_0': 47,\n",
       " 'fasta_filter_by_length': 48,\n",
       " 'bedtools_bamtofastq': 49,\n",
       " 'cshl_fastq_to_fasta': 50,\n",
       " 'deeptools_computeMatrix': 51,\n",
       " 'wc_gnu': 52,\n",
       " 'snpSift_annotate': 53,\n",
       " 'fasta2tab': 54,\n",
       " 'vcftools_merge': 55,\n",
       " 'cuffmerge': 56,\n",
       " 'freebayes': 57,\n",
       " 'ssake': 58,\n",
       " 'heatmapper': 59,\n",
       " 'glimmer_build-icm': 60,\n",
       " 'vcffilter': 61,\n",
       " 'fasta_compute_length': 62,\n",
       " 'cshl_sort_header': 63,\n",
       " 'bcftools_view': 64,\n",
       " 'EMBOSS: geecee41': 65,\n",
       " 'samtools_mpileup': 66,\n",
       " 'Cut1': 67,\n",
       " 'CONVERTER_interval_to_bgzip_0': 68,\n",
       " 'smooth_running_window': 69,\n",
       " 'sam_merge2': 70,\n",
       " 'Extract genomic DNA 1': 71,\n",
       " 'FidoAdapter': 72,\n",
       " 'cshl_cut_tool': 73,\n",
       " 'vt_normalize': 74,\n",
       " 'FalseDiscoveryRate': 75,\n",
       " 'gemini_burden': 76,\n",
       " 'ProteinQuantifier': 77,\n",
       " 'deeptools_bamCoverage': 78,\n",
       " 'proteomics_search_protein_prophet_1': 79,\n",
       " 'ctb_change_title': 80,\n",
       " 'silac_analyzer': 81,\n",
       " 'translate_bed_sequences': 82,\n",
       " 'bedtools_multiintersectbed': 83,\n",
       " 'openms_id_mapper': 84,\n",
       " 'modencode_peakcalling_macs2': 85,\n",
       " 'XTandemAdapter': 86,\n",
       " 'cufflinks': 87,\n",
       " 'CONVERTER_bed_to_gff_0': 88,\n",
       " 'hgv_david': 89,\n",
       " 'Filter1': 90,\n",
       " 'openms_protein_quantifier': 91,\n",
       " 'IDMapper': 92,\n",
       " 'CONVERTER_gff_to_bed_0': 93,\n",
       " 'openms_id_file_converter': 94,\n",
       " 'ctb_remIons': 95,\n",
       " 'augustus': 96,\n",
       " 'rsem_prepare_reference': 97,\n",
       " 'samtools_sort': 98,\n",
       " 'proteomics_search_tandem_1': 99,\n",
       " 'Datamash': 100,\n",
       " 'sort1': 101,\n",
       " 'infernal_cmsearch': 102,\n",
       " 'deeptools_correctGCBias': 103,\n",
       " 'CONVERTER_bed_gff_or_vcf_to_bigwig_0': 104,\n",
       " 'fastq_to_fasta_python': 105,\n",
       " 'cshl_fasta_formatter': 106,\n",
       " 'gemini_load': 107,\n",
       " 'cshl_sed_tool': 108,\n",
       " 'snpEff': 109,\n",
       " 'cshl_fastx_trimmer': 110,\n",
       " 'bismark_bowtie': 111,\n",
       " 'gatk2_variant_annotator': 112,\n",
       " 'cshl_awk_replace_in_column': 113,\n",
       " 'cshl_grep_tool': 114,\n",
       " 'CONVERTER_interval_to_bedstrict_0': 115,\n",
       " 'snpSift_filter': 116,\n",
       " 'bamCompare_deepTools': 117,\n",
       " 'gatk2_reduce_reads': 118,\n",
       " 'filter_bed_on_splice_junctions': 119,\n",
       " 'gemini_query': 120,\n",
       " 'ctb_simsearch': 121,\n",
       " 'FeatureFinderMultiplex': 122,\n",
       " 'get_flanks1': 123,\n",
       " 'proteomics_search_peptide_prophet_1': 124,\n",
       " 'Show beginning1': 125,\n",
       " 'blockclust': 126,\n",
       " 'bedtools_bamtobed': 127,\n",
       " 'ncbi_makeblastdb': 128,\n",
       " 'EMBOSS: water107': 129,\n",
       " 'methtools_calling': 130,\n",
       " 'samtools_rmdup': 131,\n",
       " 'EMBOSS: newseq59': 132,\n",
       " 'bed_to_bigBed': 133,\n",
       " 'rgPicardMarkDups': 134,\n",
       " 'peakcalling_macs14': 135,\n",
       " 'Show tail1': 136,\n",
       " 'random_lines1': 137,\n",
       " 'cshl_fastx_clipper': 138,\n",
       " 'bedtools_mergebed': 139,\n",
       " 'scaffold2fasta': 140,\n",
       " 'samtools_flagstat': 141,\n",
       " 'cshl_word_list_grep': 142,\n",
       " 'ctb_compound_convert': 143,\n",
       " 'meme_meme': 144,\n",
       " 'bam_to_sam': 145,\n",
       " 'ctb_filter': 146,\n",
       " 'Remove beginning1': 147,\n",
       " 'cuffdiff': 148,\n",
       " 'CONVERTER_interval_to_bed_0': 149,\n",
       " 'bgchem_fragment_merger': 150,\n",
       " 'rseqc_bam2wig': 151,\n",
       " 'gemini_comp_hets': 152,\n",
       " 'Extract_features1': 153,\n",
       " 'vcftools_isec': 154,\n",
       " 'tophat': 155,\n",
       " 'msconvert3_raw': 156,\n",
       " 'sam2interval': 157,\n",
       " 'glimmer_knowlegde-based': 158,\n",
       " 'vcfallelicprimitives': 159,\n",
       " 'methtools_plot': 160,\n",
       " 'cshl_awk_tool': 161,\n",
       " 'r_correlation_matrix': 162,\n",
       " 'antismash': 163,\n",
       " 'gatk2_variant_select': 164,\n",
       " 'Summary_Statistics1': 165,\n",
       " 'deeptools_profiler': 166,\n",
       " 'bedtools_intersectbed_bam': 167,\n",
       " 'correctGCBias': 168,\n",
       " 'peakcalling_macs': 169,\n",
       " 'gemini_pathways': 170,\n",
       " 'bamFilter': 171,\n",
       " 'cat1': 172,\n",
       " 'gops_subtract_1': 173,\n",
       " 'EMBOSS: shuffleseq87': 174,\n",
       " 'deeptools_computeGCBias': 175,\n",
       " 'fastq_quality_trimmer': 176,\n",
       " 'cshl_uniq_tool': 177,\n",
       " 'picard_ReorderSam': 178,\n",
       " 'gatk2_haplotype_caller': 179,\n",
       " 'gtf_filter_by_attribute_values_list': 180,\n",
       " 'cor2': 181,\n",
       " 'macs2_callpeak': 182,\n",
       " 'picard_ARRG': 183,\n",
       " 'bams2ratio': 184,\n",
       " 'barchart_gnuplot': 185,\n",
       " 'gatk2_realigner_target_creator': 186,\n",
       " 'PeptideIndexer': 187,\n",
       " 'XY_Plot_1': 188,\n",
       " 'methtools_tiling': 189,\n",
       " 'ctb_online_data_fetch': 190,\n",
       " 'gatk2_print_reads': 191,\n",
       " 'bedtools_intersectBed': 192,\n",
       " 'prokaryotic_ncbi_submission': 193,\n",
       " 'dt_profiler': 194,\n",
       " 'gff2bed1': 195,\n",
       " 'sam_to_bam': 196,\n",
       " 'sed_stream_editor': 197,\n",
       " 'htseq_count': 198,\n",
       " 'cshl_easyjoin': 199,\n",
       " 'gemini_windower': 200,\n",
       " 'DatamashTranspose': 201,\n",
       " 'bamCoverage_deepTools': 202,\n",
       " 'rmcontamination': 203,\n",
       " 'deeptools_bamCorrelate': 204,\n",
       " 'gatk2_variant_recalibrator': 205,\n",
       " 'gemini_de_novo': 206,\n",
       " 'addValue': 207,\n",
       " 'methtools_filter': 208,\n",
       " 'htseq-count': 209,\n",
       " 'methtools_destrand': 210,\n",
       " 'bedtools_coveragebed_counts': 211,\n",
       " 'bed2gff1': 212,\n",
       " 'seq_filter_by_id': 213,\n",
       " 'allele_counts_1': 214,\n",
       " 'methtools_dmr': 215,\n",
       " 'gatk2_indel_realigner': 216,\n",
       " 'Add_a_column1': 217,\n",
       " 'gatk2_variant_apply_recalibration': 218,\n",
       " 'bedtools_subtractbed': 219,\n",
       " 'EMBOSS: fuzztran39': 220,\n",
       " 'gtf2bedgraph': 221,\n",
       " 'EMBOSS: fuzzpro38': 222,\n",
       " 'ctb_chemfp_mol2fps': 223,\n",
       " 'IDPosteriorErrorProbability': 224,\n",
       " 'blastxml_to_top_descr': 225,\n",
       " 'histogram_rpy': 226,\n",
       " 'bg_uniq': 227,\n",
       " 'deeptools_heatmapper': 228,\n",
       " 'gatk2_base_recalibrator': 229,\n",
       " 'join1': 230,\n",
       " 'fastq_groomer': 231,\n",
       " 'deeptools_bigwigCompare': 232,\n",
       " 'bedtools_genomecoveragebed_bedgraph': 233,\n",
       " 'rsem_calculate_expression': 234,\n",
       " 'gops_intersect_1': 235,\n",
       " 'deeptools_bamCompare': 236,\n",
       " 'snpSift_geneSets': 237,\n",
       " 'wig_to_bigWig': 238,\n",
       " 'deeptools_bamFingerprint': 239,\n",
       " 'Paste1': 240,\n",
       " 'ncbi_blastn_wrapper': 241}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
