{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': 'gemini_comp_hets', '2': 'nanoplot', '3': 'mothur_screen_seqs', '4': 'rbc_mirdeep2_quantifier', '5': 'deeptools_bamCoverage', '6': 'abims_xcms_fillPeaks', '7': 'fasta2tab', '8': 'cshl_grep_tool', '9': 'wc_gnu', '10': 'Show tail1', '11': 'sklearn_build_pipeline', '12': 'bamCompare_deepTools', '13': 'fastq_to_fasta_python', '14': 'tp_sort_header_tool', '15': 'plotly_regression_performance_plots', '16': 'IDScoreSwitcher', '17': 'varscan_mpileup', '18': 'bedtools_multiintersectbed', '19': 'gemini_recessive_and_dominant', '20': 'ggplot2_heatmap2', '21': 'varscan_somatic', '22': 'snpsift_vartype', '23': 'blastxml_to_tabular_selectable', '24': 'ctb_remSmall', '25': 'egsea', '26': 'tp_split_on_column', '27': 'Heatmap', '28': 'gafa', '29': 'enhanced_bowtie_wrapper', '30': 'deeptools_correct_gc_bias', '31': 'samtools_slice_bam', '32': 'glimmer_build-icm', '33': 'eden_vectorizer', '34': 'proteomics_search_protein_prophet_1', '35': 'Cut1', '36': 'sam_bw_filter', '37': 'rseqc_junction_saturation', '38': 'mothur_heatmap_bin', '39': 'picard_CleanSam', '40': 'ProteinInference', '41': 'prokaryotic_ncbi_submission', '42': 'gbk_to_orf', '43': 'trimmer', '44': 'bg_column_arrange_by_header', '45': 'deeptools_bam_coverage', '46': 'tp_head_tool', '47': 'qiime_pick_open_reference_otus', '48': 'gops_cluster_1', '49': 'hmmer_hmmsearch', '50': 'datamash_ops', '51': '16Saligner', '52': 'fastq_quality_trimmer', '53': 'w4mjoinpn', '54': 'raceid_main', '55': 'deeptools_plot_correlation', '56': 'pureclip', '57': 'vcftools_isec', '58': 'cmv', '59': 'idpquery', '60': 'vcfflatten2', '61': 'gemini_pathways', '62': 'mothur_classify_otu', '63': 'FalseDiscoveryRate', '64': 'mycrobiota-correct-replicates', '65': 'EMBOSS: seqret84', '66': '__RELABEL_FROM_FILE__', '67': 'CONVERTER_Bam_Bai_0', '68': 'minfi_methcpg', '69': 'naive_variant_caller', '70': 'tp_sort_rows', '71': 'hgv_linkToGProfile', '72': 'Interval_Maf_Merged_Fasta2', '73': 'rseqc_geneBody_coverage', '74': 'w4mclstrpeakpics', '75': 'bctools_remove_tail', '76': 'rbc_mirdeep2', '77': 'limma_voom', '78': 'viennarna_rnafold', '79': 'mothur_remove_groups', '80': 'ggplot2_point', '81': 'ggplot_histogram', '82': 'newick_display', '83': 'cshl_find_and_replace', '84': 'minfi_getanno', '85': 'bamFilter', '86': 't2t_report', '87': 'blockclust', '88': 'antismash', '89': 'picard_NormalizeFasta', '90': 'umi_tools_group', '91': 'column_order_header_sort', '92': 'filter_by_fasta_ids', '93': 'deeptools_plot_enrichment', '94': 'minfi_dmr', '95': 'bedtools_coveragebed', '96': 'disco', '97': 'cuffcompare', '98': 'CometAdapter', '99': 'fastq collapser', '100': 'gtftobed12', '101': 'IsobaricAnalyzer', '102': 'megablast_xml_parser', '103': 'cardinal_quality_report', '104': 'FastTree', '105': 'metaspades', '106': 'shovill', '107': 'openms_protein_quantifier', '108': 'gatk2_reduce_reads', '109': 'locarna_multiple', '110': 'trim_reads', '111': 'multiqc', '112': 'p_clip_peaks', '113': 'gemini_query', '114': 'bedtools_windowbed', '115': 'mothur_summary_single', '116': 'sailfish', '117': 'raceid_trajectory', '118': 'GeneModelMapper', '119': 'stacks_denovomap', '120': 'picard_CollectRnaSeqMetrics', '121': 'mothur_unique_seqs', '122': 'rsem_calculate_expression', '123': 'pileup_parser', '124': 'join1', '125': 'openms_id_mapper', '126': 'annotateMyIDs', '127': 'FeatureFinderMultiplex', '128': 'deeptools_bigwig_compare', '129': 'Summary_Statistics1', '130': 'muscle', '131': 'cummeRbund', '132': 'allele_counts_1', '133': 'msconvert_nix', '134': 'myrimatch', '135': 'replace_column_with_key_value_file', '136': 'ncbi_blastx_wrapper', '137': 'piranha', '138': 'mothur_rarefaction_single', '139': 'map_peptides_to_bed', '140': 'gffcompare', '141': 'correctGCBias', '142': 'samtools_bedcov', '143': 'goenrichment', '144': 'peakcalling_macs14', '145': 'abims_xcms_xcmsSet', '146': 'bwa', '147': 'samtools_sort', '148': 'hicexplorer_hicbuildmatrix', '149': 'join_files_on_column_fuzzy', '150': 'transpose', '151': 'FROGS_preprocess', '152': 'htseq_count', '153': 'w4mclassfilter', '154': 'meme_chip', '155': 'mothur_make_biom', '156': 'bwa_mem', '157': 'mycrobiota-qc-report', '158': 'minfi_mset', '159': 'FeatureFinderMetabo', '160': 'lca1', '161': 'cshl_fastx_reverse_complement', '162': 'sklearn_svm_classifier', '163': 'mimodd_varextract', '164': 'mothur_dist_seqs', '165': 'Show beginning1', '166': 'meme_meme', '167': 'ggplot2_heatmap', '168': 'picard_MarkDuplicatesWithMateCigar', '169': 'minfi_getM', '170': 'bed2gff1', '171': 'regex1', '172': 'featurecounts', '173': 'DatamashTranspose', '174': 'get_read_pipeline', '175': 'rcas', '176': 'fastq_paired_end_joiner', '177': 'mothur_seq_error', '178': 'mass_spectrometry_imaging_classification', '179': 'deeptools_plot_profile', '180': 'sklearn_numeric_clustering', '181': 'modencode_peakcalling_macs2', '182': 'cardinal_segmentations', '183': 'tp_find_and_replace', '184': 'cd_hit_dup', '185': 'createInterval', '186': 'Remove beginning1', '187': 'xpath', '188': 'blastxml_to_top_descr', '189': 'mass_spectrometry_imaging_preprocessing', '190': 'subtract_query1', '191': 'datamash_transpose', '192': 'tp_replace_in_column', '193': 'mothur_get_oturep', '194': 'salmon', '195': 'cardinal_spectra_plots', '196': 'heatmapper', '197': 'get_sequences', '198': 'vcf_filter', '199': 'deeptools_plot_pca', '200': 'vcftools_annotate', '201': 'tp_sorted_uniq', '202': 'bedtools_genomecoveragebed', '203': 'heinz_visualization', '204': 'histogram_rpy', '205': 'stacks_populations', '206': 'gops_basecoverage_1', '207': 'correct_barcodes', '208': 'Paste1', '209': 'vegan_rarefaction', '210': 'rgPicFixMate', '211': 'vcfbedintersect', '212': 'rbc_mirdeep2_mapper', '213': 'wig_to_bigWig', '214': 'gatk2_variant_apply_recalibration', '215': 'samtools_idxstats', '216': 'uniprot', '217': 'TextExporter', '218': 'rnaz', '219': 'spades', '220': 'metaphlan2krona', '221': 'collection_column_join', '222': 'hicexplorer_hictransform', '223': 'bg_uniq', '224': 'thermo_raw_file_converter', '225': 'gff_to_sequence', '226': 'deeptools_multi_bigwig_summary', '227': 'Digestor', '228': 'cshl_fastx_collapser', '229': 'Multivariate', '230': 'bedtools_shufflebed', '231': 'genbank_to_gff', '232': 'CONVERTER_gz_to_uncompressed', '233': 'cshl_fastx_renamer', '234': 'picard_FixMateInformation', '235': 'proteomics_search_peptide_prophet_1', '236': 'porechop', '237': 'mothur_taxonomy_to_krona', '238': 'sklear_numeric_clustering', '239': 'Nucleosome', '240': 'mothur_venn', '241': 'mothur_pre_cluster', '242': 'IDFilter', '243': 'extract_element_from_collection', '244': 'gatk_realigner_target_creator', '245': 'hicexplorer_hicfindtads', '246': 'mothur_lefse', '247': 'metaphlan_hclust_heatmap', '248': 'fastq_dump', '249': 'gff2bed1', '250': 'fastq_paired_end_interlacer', '251': 'FROGS_clustering', '252': '__SORTLIST__', '253': 'igvtools_count', '254': 'ctb_online_data_fetch', '255': 'picard_ReorderSam', '256': 'goseq', '257': 'cat1', '258': 'Count1', '259': 'cshl_fastx_clipper', '260': 'selectsequencesfrommsa', '261': 'Batch_correction', '262': 'pileometh', '263': 'ProteinQuantifier', '264': 'gops_merge_1', '265': 'rarefaction', '266': 'iReport', '267': 'mothur_merge_files', '268': 'tophat2', '269': 'qiime_pick_otus', '270': 'circgraph', '271': 'lastz_wrapper_2', '272': '__FILTER_FAILED_DATASETS__', '273': 'meme_psp_gen', '274': 'methtools_destrand', '275': 'cshl_fastx_artifacts_filter', '276': 'cor2', '277': 'tp_unfold_column_tool', '278': 'bedtools_genomecoveragebed_bedgraph', '279': 'dexseq_annotate', '280': 'EMBOSS: revseq82', '281': 'bundle_collection', '282': 'fastq_paired_end_splitter', '283': 'ctb_filter', '284': 'glob_report', '285': 'gene2exon1', '286': 'fastq_manipulation', '287': 'ngsutils_bam_filter', '288': 'hisat2', '289': 'smooth_running_window', '290': 'FileConverter', '291': 'FROGS_remove_chimera', '292': 'CONVERTER_interval_to_bgzip_0', '293': 'table_annovar', '294': 'qiime_make_otu_heatmap', '295': 'rnbeads', '296': 'nanopolish_variants', '297': 'deeptools_computeMatrix', '298': 'NFS_transfer', '299': 'bedtools_bedtobam', '300': 'tp_cat', '301': 'rseqc_infer_experiment', '302': '__APPLY_RULES__', '303': 'resistome_analyzer', '304': 'Extractor', '305': 'silac_analyzer', '306': 'mergeCols1', '307': 'mass_spectrometry_imaging_mzplots', '308': 'humann2', '309': 'pieplots_macs', '310': 'fasta_compute_length', '311': 'gatk_indel_realigner', '312': 'random_lines1', '313': 'peakcalling_macs', '314': 'PeakPickerHiRes', '315': 'cshl_fastx_nucleotides_distribution', '316': 'mothur_filter_seqs', '317': 'cshl_easyjoin', '318': 'rgweblogo3', '319': 'ncbi_blastp_wrapper', '320': 'sam_pileup', '321': 'deeptools_bam_compare', '322': 'gmx_em', '323': 'tp_uniq_tool', '324': 'qiime_validate_mapping_file', '325': 're_he_maldi_image_registration', '326': 'vt_normalize', '327': 'gemini_load', '328': 'svm_classifier', '329': 'qiime_pick_rep_set', '330': 'cshl_fastq_quality_boxplot', '331': 'fastq_to_tabular', '332': 'mass_spectrometry_imaging_filtering', '333': 'methtools_plot', '334': 'EMBOSS: shuffleseq87', '335': 'EMBOSS: getorf42', '336': 'dt_profiler', '337': 'ggplot_point', '338': 'snpSift_dbnsfp_generic', '339': 'mothur_parsimony', '340': 'coords2clnt.py', '341': 'sklearn_ensemble', '342': 'CONVERTER_bedgraph_to_bigwig', '343': 'flexbardsc', '344': 'heinz_bum', '345': 'MapStatistics', '346': 'msconvert2_raw', '347': 'chipsequtil_maptoknowngenes', '348': 'ctb_remIons', '349': 'IDMapper', '350': 'minfi_dmp', '351': 'aragorn_trna', '352': 'cshl_sort_header', '353': 'Grep1', '354': 'aggregate_scores_in_intervals2', '355': 'bamtools', '356': 'smf_utils_extract-boxed-sequences', '357': 'samtools_calmd', '358': 'bedtools_annotatebed', '359': 'fastq_join', '360': 'mothur_summary_seqs', '361': 'gemini_burden', '362': 'dbbuilder', '363': 'cshl_fasta_nucleotides_changer', '364': 'cshl_awk_tool', '365': 'mlocarna', '366': 'cummerbund_to_cuffdiff', '367': 'proteomics_moff', '368': 'gff_filter_by_attribute', '369': 'bedtools_genomecoveragebed_histogram', '370': 'fasttree', '371': 'mimodd_vcf_filter', '372': 'cshl_fastq_to_fasta', '373': 'flexbar_no_split', '374': 'MapRTTransformer', '375': 'hmmer_hmmbuild', '376': 'sed_stream_editor', '377': 'react_cal_pipeline', '378': 'cardinal_combine', '379': 'sam_to_bam', '380': 'mothur_align_seqs', '381': 'PicardASMetrics', '382': 'dexseq', '383': 'term_id_vs_term_def', '384': 'EMBOSS: transeq101', '385': 'mass_spectrometry_imaging_segmentations', '386': 'cufflinks', '387': 'EMBOSS: tranalign100', '388': 'mothur_chimera_vsearch', '389': 'flexbar_split_RYYR_bcs', '390': 'fgsea', '391': 'Extract_features1', '392': 'picard_CollectWgsMetrics', '393': 'EMBOSS: fuzztran39', '394': 'methtools_dmr', '395': 'picard_MergeSamFiles', '396': 'rm_spurious_events.py', '397': 'deeptools_plot_heatmap', '398': 'raceid_clustering', '399': 'checkFormat', '400': 'jbrowse', '401': '16Sclassifier', '402': 'kallisto_quant', '403': 'velveth_jgi', '404': 'hicexplorer_hicplottads', '405': 'mothur_make_shared', '406': 'locarna_best_subtree', '407': 'Convert characters1', '408': 'mothur_corr_axes', '409': 'preMloc', '410': 'extract_bcs.py', '411': 'vsearch_dereplication', '412': 'cuffnorm', '413': 'cuffquant', '414': 'minimap2', '415': 'mothur_get_groups', '416': 'krona-text', '417': 'bctools_extract_alignment_ends', '418': 'picard_CollectInsertSizeMetrics', '419': 'bcftools_call', '420': 'deeptools_compute_matrix_operations', '421': 'maldi_quant_peak_detection', '422': 'ncbi_rpsblast_wrapper', '423': 'bandage_image', '424': 'stacks_refmap', '425': 'get_feature_info', '426': 'gatk2_variant_annotator', '427': 'sklearn_nn_classifier', '428': 'hcluster_sg_parser', '429': 'segemehl', '430': 'viz_overlay_moving_and_fixed_image', '431': 'heatmapper_deepTools', '432': 'vcfannotate', '433': 'deeptools_profiler', '434': 'picard_MarkDuplicates', '435': 'tabular_to_fastq', '436': 'mothur_metastats', '437': 'CONVERTER_fastqgz_to_fastq', '438': 'tp_tail_tool', '439': 'tag_stat2', '440': 'gd_raxml', '441': 'rbc_mafft', '442': 'ncbi_makeblastdb', '443': 'raceid_inspectclusters', '444': 'mothur_classify_rf', '445': 'infernal_cmstat', '446': 'extract_aln_ends.py', '447': '__UNZIP_COLLECTION__', '448': 'kraken', '449': 'EMBOSS: newseq59', '450': 'gtf_filter_by_attribute_values_list', '451': 'gemini_interactions', '452': 'bamleftalign', '453': 'RNAlien', '454': 'plotly_ml_performance_plots', '455': 'TrimByLength', '456': 'mothur_tree_shared', '457': 'sample_seqs', '458': 'ssake', '459': 'convert_bc_to_binary_RY.py', '460': 'fastq_trimmer', '461': 'scaffold2fasta', '462': 'EMBOSS: textsearch98', '463': '__MERGE_COLLECTION__', '464': 'unipept', '465': 'infernal_cmbuild', '466': 'idr-embl', '467': 'bismark_bowtie', '468': 'rbc_rnacode', '469': 'Add_a_column1', '470': 'ncbi_tblastn_wrapper', '471': 'cshl_word_list_grep', '472': 'bedtools_multicovtbed', '473': 'ucsc-twobittofa', '474': 'bowtie2', '475': 'addValue', '476': 'translate_bed_sequences', '477': 'CONVERTER_sam_to_unsorted_bam', '478': 'gatk2_variant_recalibrator', '479': 'tp_awk_tool', '480': 'mothur_phylotype', '481': 'structure_to_gspan', '482': 'varscan', '483': 'mothur_chimera_uchime', '484': 'cshl_fastq_quality_filter', '485': 'bedtools_unionbedgraph', '486': 'picard_CASM', '487': 'graphclust_aggregate_alignments', '488': 'nn_classifier', '489': 'stacks_procrad', '490': 'bedtools_intersectBed', '491': 'get_subontology_from', '492': 'numeric_clustering', '493': 'methtools_tiling', '494': 'gspan', '495': 'FROGS_affiliation_OTU', '496': 'proteomics_search_tandem_1', '497': 'CONVERTER_interval_to_bed_0', '498': 'smf_utils_fix-fasta-headers', '499': 'fasplit', '500': 'qiime_split_libraries_fastq', '501': 'mothur_sub_sample', '502': 'gops_coverage_1', '503': 'FidoAdapter', '504': 'MSGFPlusAdapter', '505': 'qiime_core_diversity', '506': 'flexbar_split_RR_bcs', '507': 'signalp3', '508': 'tab2fasta', '509': 'snpeff_to_peptides', '510': 'ctb_chemfp_mol2fps', '511': 'metams_runGC', '512': 'umi_tools_whitelist', '513': 'FROGS_filters', '514': 'sklearn_searchcv', '515': 'sklearn_feature_selection', '516': 'goslimmer', '517': 'FROGS_affiliations_stat', '518': 'bam_to_sam', '519': 'hicexplorer_hicmergematrixbins', '520': 'collapse_dataset', '521': 'NSPDK_candidateClust', '522': 'abricate', '523': 'length_and_gc_content', '524': 'je_markdupes', '525': 'tp_replace_in_line', '526': 'bg_diamond', '527': 'cmFinder', '528': 'comp1', '529': 'seurat', '530': 'cardinal_filtering', '531': 'deeptools_bamCorrelate', '532': 'trimmomatic', '533': 'w4mcorcov', '534': 'diffbind-embl', '535': 'tp_grep_tool', '536': 'picard_ARRG', '537': 'bedtools_subtractbed', '538': 'viennarna_rnaplot', '539': 'fastq_filter', '540': 'modencode_peakcalling_spp', '541': 'sickle', '542': 'bam2wig', '543': 'vcfcombine', '544': 'MzTabExporter', '545': 'remove_tail.py', '546': 'samtools_filter', '547': 'IDMassAccuracy', '548': 'glimmer_not-knowlegde-based', '549': 'Extract genomic DNA 1', '550': 'IDPosteriorErrorProbability', '551': 'mothur_make_group', '552': 'ncbi_blastn_wrapper', '553': 'CONVERTER_bed_to_gff_0', '554': 'ete_homology_classifier', '555': 'cast', '556': 'heinz', '557': 'picard_BamIndexStats', '558': 'picard_SamToFastq', '559': 'hicexplorer_hicpca', '560': 'gatk2_unified_genotyper', '561': 'methtools_filter', '562': 'mothur_heatmap_sim', '563': 'gatk_depth_of_coverage', '564': 'samtools_rmdup', '565': 'smf_utils_filter-by-energy', '566': 'gmx_solvate', '567': 'raceid_inspecttrajectory', '568': 'picard_EstimateLibraryComplexity', '569': 'mothur_make_contigs', '570': 'term_id_vs_term_name', '571': 'htseq-count', '572': 'hcluster_sg', '573': 'plotly_parallel_coordinates_plot', '574': 'bedtools_bamtofastq', '575': 'mothur_dist_shared', '576': 'minfi_qc', '577': 'uparse', '578': 'CONVERTER_fasta_to_fai', '579': 'MAF_To_Fasta1', '580': 'rmcontamination', '581': 'hgv_snpFreq', '582': 'hicexplorer_hicsummatrices', '583': 'mass_spectrometry_imaging_combine', '584': 'gemini_windower', '585': 'Flash', '586': 'sequence_content_trimmer', '587': 'Filter1', '588': 'ctb_ob_genProp', '589': 'bcftools_view', '590': 'idpqonvert', '591': 'bg_statistical_hypothesis_testing', '592': 'deeptools_plot_fingerprint', '593': 'cardinal_mz_images', '594': 'blockbuster', '595': 'Remove_ending', '596': 'mothur_homova', '597': 'vcfsort', '598': 'gops_concat_1', '599': 'trim_galore', '600': 'deeptools_plot_coverage', '601': 'ctb_change_title', '602': 'minfi_maptogenome', '603': 'pilon', '604': 'bed_to_bigBed', '605': 'bedtools_sortbed', '606': 'wolf_psort', '607': 'bam-filter', '608': 'ip_projective_transformation', '609': 'methtools_calling', '610': 'FeatureLinkerUnlabeledKD', '611': 'bowtie_wrapper', '612': 'mothur_classify_seqs', '613': 'nspdk_sparse', '614': 'gatk2_depth_of_coverage', '615': 'snpEff', '616': 'filtlong', '617': 'gemini_actionable_mutations', '618': 'samtools_split', '619': 'make_families', '620': 'nonpareil', '621': 'megablast_wrapper', '622': 'pileup_interval', '623': 'qiime_assign_taxonomy', '624': 'snippy_core', '625': 'je_clip', '626': 'gatk_unified_genotyper', '627': 'mycrobiota-make-multi-otutable', '628': 'bedtools_slopbed', '629': 'msgfplus', '630': 'EMBOSS: wordcount109', '631': 'graphclust_align_cluster', '632': 'IDConflictResolver', '633': 'maldi_quant_preprocessing', '634': 'ConsensusMapNormalizer', '635': 'EMBOSS: water107', '636': 'addName', '637': 'SequenceCoverageCalculator', '638': 'scale_image', '639': 'EMBOSS: extractseq35', '640': 'Psortb', '641': 'bedtools_intersectbed', '642': 'Determine_BC', '643': 'samtool_filter2', '644': 'ia_coordinates_of_roi', '645': 'CONVERTER_interval_to_bedstrict_0', '646': 'sam_merge2', '647': 'fragmenter', '648': 'eden_sequence_converter', '649': 'mothur_get_seqs', '650': 'metilene', '651': 'bctools_extract_crosslinked_nucleotides', '652': 'bgchem_fragment_merger', '653': 'umi_tools_count', '654': 'eukaryotic_ncbi_submission', '655': 'TrimByN', '656': 'vcffilter2', '657': 'smf_utils_estimate-energy', '658': 'paralyzer', '659': 'graphlan', '660': 'PeptideIndexer', '661': 'bwameth', '662': 'seq_composition', '663': 'graphicsmagick_image_convert', '664': 'MapAlignerPoseClustering', '665': 'humann2_genefamilies_genus_level', '666': 'bwa_wrapper', '667': 'find_in_reference', '668': 'cshl_sed_tool', '669': 'TrimPrimer', '670': 'vcfgenotypes', '671': 'tp_easyjoin_tool', '672': 'gatk2_haplotype_caller', '673': 'vcf2tsv', '674': 'kraken-report', '675': 'tsne', '676': 'picard_FilterSamReads', '677': 'rsem_prepare_reference', '678': 'preproc', '679': 'je_demultiplex', '680': 'xcms_merge', '681': 'gmx_nvt', '682': 'fastq_paired_end_deinterlacer', '683': 'get_flanks1', '684': '__FLATTEN__', '685': 'vcfvcfintersect', '686': 'hisat', '687': 'biosigner', '688': 'meme_dreme', '689': 'hgv_david', '690': 'Draw_phylogram', '691': 'rseqc_bam2wig', '692': 'mothur_remove_lineage', '693': 'idpassemble', '694': 'cshl_awk_replace_in_column', '695': 'cutadapt', '696': 'format_metaphlan2_output', '697': 'Extract_RNA_seq_Evidence', '698': 'picard_AddOrReplaceReadGroups', '699': 'bamCoverage_deepTools', '700': 'DatamashOps', '701': 'deeptools_bamCompare', '702': 'megahit', '703': 'viennarna_rnaalifold', '704': 'rna_star', '705': 'QCCalculator', '706': 'gemini_de_novo', '707': 'augustus', '708': 'bedtools_map', '709': 'deseq2_single', '710': 'vcfallelicprimitives', '711': 'ProteinResolver', '712': 'ctb_simsearch', '713': 'minfi_dropsnp', '714': 'MultiplexResolver', '715': 'snpfinder', '716': 'ChooseTag', '717': 'picard_SortSam', '718': 'deeptools_heatmapper', '719': 'minfi_rset', '720': 'tp_text_file_with_recurring_lines', '721': 'vcftools_merge', '722': 'mass_spectrometry_imaging_ion_images', '723': 'minfi_ppfun', '724': 'XTandemAdapter', '725': 'heinz_scoring', '726': 'mothur_cluster_split', '727': 'qiime_count_seqs', '728': 'bcftools_norm', '729': 'column_remove_by_header', '730': 'vegan_diversity', '731': 'deeptools_bamFingerprint', '732': 'ExtractFASTAfromFASTQ', '733': 'samtools_flagstat', '734': 'Fetch Taxonomic Ranks', '735': 'regex_replace', '736': 'dunovo', '737': 'snpEff_download', '738': 'ConsensusID', '739': 'bam_to_bigwig', '740': 'mothur_cluster', '741': 'fastq_combiner', '742': 'tp_cut_tool', '743': 'blastxml_to_tabular', '744': 'PicardInsertSize', '745': 'FileInfo', '746': 'varscan_copynumber', '747': 'deeptools_compute_gc_bias', '748': 'iuc_pear', '749': 'gops_join_1', '750': 'mothur_count_seqs', '751': 'bg_find_subsequences', '752': 'mycrobiota-split-multi-otutable', '753': 'Datamash', '754': 'FeatureFinderCentroided', '755': 'racon', '756': 'snpSift_annotate', '757': 'abims_CAMERA_annotateDiffreport', '758': 'get_child_terms', '759': 'picard_ValidateSamFile', '760': 'FileFilter', '761': 'gmx_md', '762': 'gatk2_variant_select', '763': 'export2graphlan', '764': 'scpipe', '765': 'unicycler', '766': 'macs2_bdgcmp', '767': 'umi_tools_dedup', '768': 'humann2_regroup_table', '769': 'metaphlan2', '770': 'tp_multijoin_tool', '771': 'seq_filter_by_mapping', '772': 'rbc_splitfasta', '773': 'vcffilter', '774': 'samtools_flag_filter', '775': 'mothur_nmds', '776': 'bedtools_bamtobed', '777': 'pynast', '778': 'cardinal_preprocessing', '779': 'snippy', '780': 'FeatureLinkerUnlabeledQT', '781': 'meme_fimo', '782': 'FeatureLinkerUnlabeled', '783': 'blast_parser', '784': 'tp_sed_tool', '785': 'bedtools_intersectbed_bam', '786': 'combine_metaphlan2_humann2', '787': 't_coffee', '788': 'Univariate', '789': 'cshl_fastx_barcode_splitter', '790': 'tophat_fusion_post', '791': 'abims_xcms_group', '792': 'mcClust', '793': 'EMBOSS: geecee41', '794': 'fraggenescan', '795': 'minfi_getbeta', '796': 'samtools_stats', '797': 'ctb_silicos_qed', '798': 'bed_to_bigwig', '799': 'FileMerger', '800': 'gffread', '801': 'vcffixup', '802': 'CONVERTER_gff_to_interval_index_0', '803': 'abims_xcms_summary', '804': 'gatk2_indel_realigner', '805': 'interproscan', '806': 'stacks_genotypes', '807': 'deeptools_bigwigCompare', '808': 'annotatemyids', '809': 'XY_Plot_1', '810': 'freebayes', '811': 'mimodd_varcall', '812': 'sort1', '813': 'MassCalculator', '814': 'humann2_renorm_table', '815': 'heatmap2', '816': 'valet', '817': 'fasta-stats', '818': 'tp_tac', '819': 'uchime', '820': 'snpSift_filter', '821': 'CONVERTER_bed_to_bgzip_0', '822': 'mothur_amova', '823': 'stringtie', '824': 'fasta_merge_files_and_filter_unique_sequences', '825': 'mycrobiota-krona-mothur', '826': 'clustalw', '827': 'vcfselectsamples', '828': 'query_tabular', '829': 'DecoyDatabase', '830': 'bg_protein_properties', '831': 'gops_intersect_1', '832': 'gtf2bedgraph', '833': 'bg_sortmerna', '834': 'deseq2', '835': 'gatk2_base_recalibrator', '836': 'bcftools_consensus', '837': 'rseqc_inner_distance', '838': 'macs2_callpeak', '839': 'ggplot2_histogram', '840': 'fastq_groomer', '841': 'flexbar', '842': 'qiime_align_seqs', '843': 'graphclust_glob_report_no_align', '844': 'flye', '845': 'Grouping1', '846': 'GeMoMa_Annotation_Filter', '847': 'mothur_remove_seqs', '848': 'raceid_filtnormconf', '849': 'group_humann2_uniref_abundances_to_go', '850': 'IDMerger', '851': 'edger', '852': 'bedtools_getfastabed', '853': 'mothur_trim_seqs', '854': 'cshl_fastx_quality_statistics', '855': 'seq_filter_by_id', '856': 'bedtools_coveragebed_counts', '857': 'r_correlation_matrix', '858': 'graphlan_annotate', '859': 'gmx_setup', '860': 'ip_projective_transformation_points', '861': 'stringtie_merge', '862': 'gmx_npt', '863': 'gemini_db_info', '864': 'infernal_cmsearch', '865': 'deeptools_compute_matrix', '866': 'gatk2_variant_filtration', '867': 'velvet', '868': 'deeptools_computeGCBias', '869': 'kraken-translate', '870': 'mothur_get_communitytype', '871': 'lfq_protein_quant', '872': 'barchart_gnuplot', '873': 'quast', '874': 'rseqc_read_duplication', '875': 'cshl_princeton_fastx_barcode_splitter', '876': 'gatk2_print_reads', '877': 'cshl_uniq_tool', '878': 'charts', '879': 'umi_tools_extract', '880': 'cshl_multijoin', '881': 'snpSift_geneSets', '882': 'cshl_fasta_formatter', '883': 'cardinal_data_exporter', '884': 'rnafold', '885': 'dexseq_count', '886': 'computeMatrix', '887': 'taxonomy_krona_chart', '888': 'ctb_compound_convert', '889': 'filter_bed_on_splice_junctions', '890': 'mycrobiota-otutable_add_blast', '891': 'chipseeker', '892': 'HighResPrecursorMassCorrector', '893': 'predict_pipeline', '894': 'mz_to_sqlite', '895': 'rseqc_read_distribution', '896': 'gemini_annotate', '897': 'tmhmm2', '898': 'bedtools_mergebed', '899': 'qiime_pick_closed_reference_otus', '900': 'trinity', '901': 'venn_list', '902': 'align_families', '903': 'sam2interval', '904': 'treebest_best', '905': 'cuffdiff', '906': 'kraken-filter', '907': 'openms_id_file_converter', '908': 'rarefaction_analyzer', '909': 'melt', '910': 'Btrim64', '911': 'wsdl_hmdb', '912': 'mimodd_map', '913': 'gatk2_realigner_target_creator', '914': 'velvetoptimiser', '915': 'bigwig_to_bedgraph', '916': 'wiggle2simple1', '917': 'blast2go', '918': 'samtools_mpileup', '919': 'prokka', '920': 'flanking_features_1', '921': 'deeptools_correctGCBias', '922': 'mothur_pcoa', '923': 'peptide_shaker', '924': 'EMBOSS: fuzzpro38', '925': 'ruvseq', '926': 'mimodd_varreport', '927': 'fastqc', '928': 'bams2ratio', '929': 'gops_subtract_1', '930': 'bedtools_complementbed', '931': 'transdecoder', '932': 'secretbt2test', '933': 'sqlite_to_tabular', '934': 'vcf2pgSnp', '935': 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', '936': 'hicexplorer_hiccorrectmatrix', '937': 'CONVERTER_fasta_to_tabular', '938': 'qiime_make_phylogeny', '939': 'bismark_bowtie2', '940': 'search_gui', '941': 'rseqc_bam_stat', '942': 'roary', '943': 'velvetg_jgi', '944': 'smf_utils_find-boxes', '945': 'ete_genetree_splitter', '946': 'merge_pcr_duplicates.py', '947': 'cuffmerge', '948': 'deg_annotate', '949': 'msconvert3_raw', '950': 'macs2_predictd', '951': 'CONVERTER_gff_to_bed_0', '952': '__FILTER_FROM_FILE__', '953': 'IDFileConverter', '954': 'cshl_cut_tool', '955': 'minfi_read450k', '956': 'deeptools_multi_bam_summary', '957': 'fasta_filter_by_length', '958': 'mothur_count_groups', '959': 'hicexplorer_hicplotmatrix', '960': 'peakachu', '961': 'bctools_remove_spurious_events', '962': 'rgPicardMarkDups', '963': 'decoyfasta', '964': '__ZIP_COLLECTION__', '965': 'glimmer_knowlegde-based', '966': 'cufflinks_prok', '967': 'tophat', '968': 'mass_spectrometry_imaging_qc', '969': 'ChangeCase', '970': 'cshl_fastx_trimmer', '971': 'secure_hash_message_digest'}\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import json\n",
    "import warnings\n",
    "import operator\n",
    "\n",
    "import h5py\n",
    "from keras.models import model_from_json\n",
    "from keras import backend as K\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "\n",
    "def read_file(file_path):\n",
    "    with open(file_path, 'r') as data_file:\n",
    "        data = json.loads(data_file.read())\n",
    "    return data\n",
    "\n",
    "\n",
    "def create_model(model_path):\n",
    "    reverse_dictionary = dict((str(v), k) for k, v in dictionary.items())\n",
    "    model_weights = list()\n",
    "    weight_ctr = 0\n",
    "    for index, item in enumerate(trained_model.keys()):\n",
    "        if \"weight_\" in item:\n",
    "            d_key = \"weight_\" + str(weight_ctr)\n",
    "            weights = trained_model.get(d_key).value\n",
    "            model_weights.append(weights)\n",
    "            weight_ctr += 1\n",
    "    # set the model weights\n",
    "    loaded_model.set_weights(model_weights)\n",
    "    return loaded_model, dictionary, reverse_dictionary\n",
    "\n",
    "\n",
    "def compute_recommendations(model, tool_sequence, labels, dictionary, reverse_dictionary, class_weights, topk=20, max_seq_len=25):\n",
    "    tl_seq = tool_sequence.split(\",\")\n",
    "    last_tool_name = reverse_dictionary[str(tl_seq[-1])]\n",
    "    sample = np.zeros(max_seq_len)\n",
    "    for idx, tool_id in enumerate(tl_seq):\n",
    "        sample[idx] = int(tool_id)\n",
    "    sample_reshaped = np.reshape(sample, (1, max_seq_len))\n",
    "\n",
    "    tool_sequence_names = [reverse_dictionary[str(tool_pos)] for tool_pos in tool_sequence.split(\",\")]\n",
    "    \n",
    "    model.summary()\n",
    "    \n",
    "    # predict next tools for a test path\n",
    "    prediction = model.predict(sample_reshaped, verbose=0)\n",
    "    \n",
    "    weight_val = list(class_weights.values())\n",
    "    weight_val = np.reshape(weight_val, (len(weight_val),))\n",
    "    \n",
    "    prediction = np.reshape(prediction, (prediction.shape[1],))\n",
    "    \n",
    "    prediction = prediction * weight_val\n",
    "    \n",
    "    prediction = prediction / float(np.max(prediction))\n",
    "    \n",
    "    prediction_pos = np.argsort(prediction, axis=-1)\n",
    "\n",
    "    # get topk prediction\n",
    "    topk_prediction_pos = prediction_pos[-topk:]\n",
    "    topk_prediction_val = [int(prediction[pos] * 100) for pos in topk_prediction_pos]\n",
    "    \n",
    "    # read tool names using reverse dictionary\n",
    "    pred_tool_ids = [reverse_dictionary[str(tool_pos)] for tool_pos in topk_prediction_pos]\n",
    "\n",
    "    pred_tool_ids_sorted = dict()\n",
    "    for (tool_pos, tool_pred_val) in zip(topk_prediction_pos, topk_prediction_val):\n",
    "        tool_name = reverse_dictionary[str(tool_pos)]\n",
    "        pred_tool_ids_sorted[tool_name] = tool_pred_val\n",
    "    pred_tool_ids_sorted = dict(sorted(pred_tool_ids_sorted.items(), key=lambda kv: kv[1], reverse=True))\n",
    "\n",
    "    ids_tools = dict()\n",
    "    keys = list(pred_tool_ids_sorted.keys())\n",
    "    tool_seq_name = \",\".join(tool_sequence_names)\n",
    "    print(\"Current tool sequence: \")\n",
    "    print()\n",
    "    print(tool_seq_name)\n",
    "    print()\n",
    "    print(\"Recommended tools for the tool sequence '%s' with their scores in decreasing order:\" % tool_seq_name)\n",
    "    print()\n",
    "    for i in pred_tool_ids_sorted:\n",
    "        print(i + \"(\" + str(pred_tool_ids_sorted[i]) + \"%)\")\n",
    "    for key in pred_tool_ids_sorted:\n",
    "        ids_tools[key] = dictionary[key]\n",
    "    print()\n",
    "    print(\"Tool ids:\")\n",
    "    print()\n",
    "    for i in ids_tools:\n",
    "        print(i + \"(\" + str(ids_tools[i]) + \")\")\n",
    "\n",
    "\n",
    "model_path = \"../output_files/data/models/model_rnn_custom_loss_19_03.hdf5\"\n",
    "\n",
    "trained_model = h5py.File(model_path, 'r')\n",
    "model_config = json.loads(trained_model.get('model_config').value)\n",
    "dictionary = json.loads(trained_model.get('data_dictionary').value)\n",
    "class_weights = json.loads(trained_model.get('class_weights').value)\n",
    "loaded_model = model_from_json(model_config)\n",
    "\n",
    "model, dictionary, reverse_dictionary = create_model(model_path)\n",
    "\n",
    "print(reverse_dictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"sequential_31\"\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "embedding_31 (Embedding)     (None, None, 244)         237168    \n",
      "_________________________________________________________________\n",
      "spatial_dropout1d_31 (Spatia (None, None, 244)         0         \n",
      "_________________________________________________________________\n",
      "gru_61 (GRU)                 (None, None, 359)         650508    \n",
      "_________________________________________________________________\n",
      "dropout_61 (Dropout)         (None, None, 359)         0         \n",
      "_________________________________________________________________\n",
      "gru_62 (GRU)                 (None, 359)               774363    \n",
      "_________________________________________________________________\n",
      "dropout_62 (Dropout)         (None, 359)               0         \n",
      "_________________________________________________________________\n",
      "dense_31 (Dense)             (None, 972)               349920    \n",
      "=================================================================\n",
      "Total params: 2,011,959\n",
      "Trainable params: 2,011,959\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "Current tool sequence: \n",
      "\n",
      "bowtie2\n",
      "\n",
      "Recommended tools for the tool sequence 'bowtie2' with their scores in decreasing order:\n",
      "\n",
      "fastqc(100%)\n",
      "freebayes(91%)\n",
      "featurecounts(81%)\n",
      "deeptools_bam_coverage(80%)\n",
      "hisat2(70%)\n",
      "macs2_callpeak(70%)\n",
      "htseq_count(70%)\n",
      "multiqc(68%)\n",
      "picard_MarkDuplicates(66%)\n",
      "samtools_flagstat(63%)\n",
      "stringtie(63%)\n",
      "bamFilter(63%)\n",
      "bedtools_intersectbed(62%)\n",
      "samtools_rmdup(62%)\n",
      "samtool_filter2(61%)\n",
      "samtools_idxstats(60%)\n",
      "bedtools_genomecoveragebed_bedgraph(58%)\n",
      "spades(56%)\n",
      "rseqc_infer_experiment(56%)\n",
      "deeptools_bam_compare(54%)\n",
      "\n",
      "Tool ids:\n",
      "\n",
      "fastqc(927)\n",
      "freebayes(810)\n",
      "featurecounts(172)\n",
      "deeptools_bam_coverage(45)\n",
      "hisat2(288)\n",
      "macs2_callpeak(838)\n",
      "htseq_count(152)\n",
      "multiqc(111)\n",
      "picard_MarkDuplicates(434)\n",
      "samtools_flagstat(733)\n",
      "stringtie(823)\n",
      "bamFilter(85)\n",
      "bedtools_intersectbed(641)\n",
      "samtools_rmdup(564)\n",
      "samtool_filter2(643)\n",
      "samtools_idxstats(215)\n",
      "bedtools_genomecoveragebed_bedgraph(278)\n",
      "spades(219)\n",
      "rseqc_infer_experiment(301)\n",
      "deeptools_bam_compare(321)\n"
     ]
    }
   ],
   "source": [
    "topk = 20 # set the maximum number of recommendations\n",
    "tool_seq = \"474\" # give tools ids in a sequence and see the recommendations. To know all the tool ids, \n",
    "                     # please print the variable 'reverse_dictionary'\n",
    "compute_recommendations(model, tool_seq, \"\", dictionary, reverse_dictionary, class_weights, topk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
